{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Reference"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(reshape2)\n",
    "library(ggplot2)\n",
    "library(rtracklayer)\n",
    "library(BiocParallel)\n",
    "library(dplyr)\n",
    "setwd(\"/NAS/wg_looking/gRNA_ONT\")\n",
    "\n",
    "\n",
    "\n",
    "# loading gtf file\n",
    "gtfFile = \"/NAS/wg_looking/human_ref/gencode.v37.chr_patch_hapl_scaff.annotation.gtf\"\n",
    "gtf <- rtracklayer::import(gtfFile)\n",
    "gtf <- as.data.frame(gtf)\n",
    "transcript_gtf <- gtf[gtf$type == \"transcript\",]\n",
    "coding_gene =  gtf[gtf$type==\"gene\" & gtf$gene_type==\"protein_coding\", ]\n",
    "\n",
    "hg38.gtf <- gtf\n",
    "save(hg38.gtf, file=\"stat/hg38.gtf.RData\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# loading gff file\n",
    "gffFile = \"/NAS/wg_looking/human_ref/gencode.v37.chr_patch_hapl_scaff.annotation.gff3\"\n",
    "gff <- rtracklayer::import(gffFile)\n",
    "gff <- as.data.frame(gff)\n",
    "\n",
    "gff.sub <- gff[!(gff$type %in% \"gene\"), ]\n",
    "gff.split <- split(gff.sub, factor(gff.sub$transcript_id))\n",
    "\n",
    "\n",
    "# Mapping genomic coordinates to transcript-relative coordinates\n",
    "map.list <- lapply(gff.split, function(df){\n",
    "    df.list <- split(df, factor(df$seqnames))\n",
    "    b <- lapply(df.list, function(df){\n",
    "            df <- df[, c(\"seqnames\", \"start\", \"end\", \"type\", \"ID\", \"transcript_id\", \"gene_id\", \"gene_type\", \"gene_name\")]\n",
    "            t_start <- (df$start - df$start[1] + 1)\n",
    "            t_end <- (df$end - df$start[1] + 1)\n",
    "            re.df <- data.frame(df[, c(\"transcript_id\", \"ID\")], t_start, t_end, df[, -c(5,6)]) \n",
    "    })\n",
    "    c <- rbindlist(b)\n",
    "})\n",
    "\n",
    "gff.anno <- rbindlist(map.list)\n",
    "\n",
    "save(gff.anno, file=\"stat/gff.anno.RData\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# the transcript length\n",
    "hg38_trans_length = \"/NAS/wg_looking/human_ref/gencode.v37.chr_patch_hapl_scaff.transcript.length\"\n",
    "L <- read.table(hg38_trans_length, stringsAsFactors=F)\n",
    "colnames(L) <- c(\"transcript_id\", \"length\")\n",
    "\n",
    "p1 <- ggplot(L, aes(x=length)) +\n",
    "    geom_density(size=2,  color=\"#4678BA\") + \n",
    "    theme_bw() +\n",
    "    theme(\n",
    "      panel.grid=element_blank(),\n",
    "      axis.text.x = element_text(angle = 30, color = \"black\", \n",
    "                                 size = 8, hjust = 0.5, vjust=0.5),\n",
    "      plot.title=element_text(size=14, face=\"bold\", hjust=0.5)\n",
    "    ) +\n",
    "    # scale_color_manual(values = cols) +\n",
    "    scale_x_log10() +\n",
    "    ggtitle(\"the length distribution of ref transcript\") \n",
    "  \n",
    "ggsave(paste0(\"stat/ref_trans_length_density.pdf\"),\n",
    "      width=6, height=5)\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "################\n",
    "#  Distance of Alu from gene endpoints\n",
    "RM_file <- paste0(\"database/gencode.v37.chr_patch_hapl_scaff.transcript.fa.out\")\n",
    "RM.tb <- read_rm(RM_file)\n",
    "RM.tb <- data.frame(RM.tb)\n",
    "RM.tb <- RM.tb[as.numeric(RM.tb$sw_score)>=225 & as.numeric(RM.tb$perc_div)<=18, ]\n",
    "RM.tb <- RM.tb[grepl(\"L1HS\", RM.tb$repeat_id) | \n",
    "                   grepl(\"L1P\", RM.tb$repeat_id) | \n",
    "                   RM.tb$matching_class %in% c(\"SINE/Alu\"), ]\n",
    "dim(RM.tb) \n",
    "TEs <- c(\"SINE/Alu\", \"LINE/L1\")\n",
    "\n",
    "tb <- RM.tb[RM.tb$matching_class ==\"SINE/Alu\",]\n",
    "start <- tb$qry_start\n",
    "pp <- data.frame(start)\n",
    "\n",
    "ggplot(pp, aes(x=start)) + \n",
    "  geom_density(color=\"#4678BA\", size=2)+\n",
    "  theme_classic() +\n",
    "  scale_x_log10() +\n",
    "  ggtitle(\"The start of Alu in reference transcripts\")\n",
    "\n",
    "ggsave(\"stat/The_start_of_Alu_in_reference_transcripts.pdf\", width = 7, height = 5)\n",
    "\n",
    "\n",
    "##########\n",
    "# Alu\n",
    "tb <- RM.tb[RM.tb$matching_class ==\"SINE/Alu\",]\n",
    "colnames(tb)[5] <- \"transcript_id\"\n",
    "length(unique(tb$transcript_id)) # 36945\n",
    "tb.merge <- merge(tb, transcript_gtf[,c(\"transcript_id\", \"gene_id\", \"gene_name\", \"gene_type\")], by=\"transcript_id\")\n",
    "# length(unique(tb.merge$gene_id[tb.merge$gene_type==\"protein_coding\"]))\n",
    "length(unique(tb.merge$gene_id)) # 16311\n",
    "\n",
    "start <- tb$qry_start\n",
    "pp <- data.frame(start)\n",
    "\n",
    "ggplot(pp, aes(x=start)) + \n",
    "  geom_density(color=\"#4678BA\", size=2)+\n",
    "  theme_classic() +\n",
    "  scale_x_log10() +\n",
    "  ggtitle(\"The start of Alu in reference transcripts\")\n",
    "\n",
    "ggsave(\"stat/The_start_of_Alu_in_reference_transcripts.pdf\", width = 7, height = 5)\n",
    "\n",
    "\n",
    "#########\n",
    "# L1\n",
    "tb <- RM.tb[RM.tb$matching_class ==\"LINE/L1\",]\n",
    "colnames(tb)[5] <- \"transcript_id\"\n",
    "length(unique(tb$transcript_id))\n",
    "tb.merge <- merge(tb, transcript_gtf[,c(\"transcript_id\", \"gene_id\", \"gene_name\", \"gene_type\")], by=\"transcript_id\")\n",
    "length(unique(tb.merge$gene_id[tb.merge$gene_type==\"protein_coding\"]))\n",
    "length(unique(tb.merge$gene_id))\n",
    "\n",
    "\n",
    "start <- tb$qry_start\n",
    "pp <- data.frame(start)\n",
    "\n",
    "ggplot(pp, aes(x=start)) + \n",
    "  geom_density(color=\"#4678BA\", size=2)+\n",
    "  theme_classic() +\n",
    "  scale_x_log10() +\n",
    "  ggtitle(\"The start of L1 in reference transcripts\")\n",
    "\n",
    "ggsave(\"stat/The_start_of_L1_in_reference_transcripts.pdf\", width = 7, height = 5)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "## Alu and L1 related genes/transcript\n",
    "# Alu\n",
    "tb <- RM.tb[RM.tb$matching_class %in% c(\"SINE/Alu\", \"LINE/L1\"),]\n",
    "colnames(tb)[5] <- \"transcript_id\"\n",
    "tb.merge <- merge(tb, transcript_gtf[,c(\"transcript_id\", \"gene_id\", \"gene_name\", \"gene_type\")], by=\"transcript_id\")\n",
    "write.table(tb.merge, \"stat/Ref_Alu_L1_related_transcripts.txt\", sep=\"\\t\", quote=F, row.names=F, col.names=T)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "## Alu and L1 related genes/transcript\n",
    "num.list <- lapply(c(\"SINE/Alu\", \"LINE/L1\"), function(TE){\n",
    "    tb <- RM.tb[RM.tb$matching_class %in% TE,]\n",
    "    colnames(tb)[5] <- \"transcript_id\"\n",
    "    tb.merge <- merge(tb, transcript_gtf[,c(\"transcript_id\", \"gene_id\", \"gene_name\", \"gene_type\")], by=\"transcript_id\")\n",
    "    coding_trans <- unique(tb.merge[tb.merge$gene_type %in% \"protein_coding\", \"transcript_id\"])\n",
    "    noncoding_trans <- unique(tb.merge[!(tb.merge$gene_type %in% \"protein_coding\"), \"transcript_id\"])\n",
    "    coding_gene <- unique(tb.merge[tb.merge$gene_type %in% \"protein_coding\", \"gene_id\"])\n",
    "    noncoding_gene <- unique(tb.merge[!(tb.merge$gene_type %in% \"protein_coding\"), \"gene_id\"])\n",
    "\n",
    "    n1_coding <- data.frame(Repeat=TE, \n",
    "                     coding_type=\"protein_coding\",\n",
    "                    type=c(\"gene\", \"transcript\"),\n",
    "                    num=c(length(coding_gene), length(coding_trans)))\n",
    "    n1_noncoding <- data.frame(Repeat=TE, \n",
    "                     coding_type=\"non_coding\",\n",
    "                    type=c(\"gene\", \"transcript\"),\n",
    "                    num=c(length(noncoding_gene), length(noncoding_trans)))\n",
    "    n1 <- rbind(n1_coding, n1_noncoding)\n",
    "\n",
    "\n",
    "})\n",
    "\n",
    "num <- rbindlist(num.list)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(bedtoolsr)\n",
    "load(\"stat/gff.anno.RData\")\n",
    "\n",
    "RM_file <- paste0(\"database/ref_transcript_with_intron/RepeatMasker_res/gencode.v37.chr_patch_hapl_scaff.transcript.with_intron.fa.out\")\n",
    "RM.tb <- read_rm(RM_file)\n",
    "RM.tb <- data.frame(RM.tb)\n",
    "RM.tb <- RM.tb[as.numeric(RM.tb$sw_score)>=225 & as.numeric(RM.tb$perc_div)<=18, ]\n",
    "tb <- RM.tb[RM.tb$matching_class %in% c(\"SINE/Alu\", \"LINE/L1\"),]\n",
    "RM.bed <- tb[, c(\"qry_id\", \"qry_start\", \"qry_end\", \"repeat_id\")]\n",
    "\n",
    "gff.bed <- gff.anno[, c(\"transcript_id\", \"t_start\", \"t_end\", \"type\")]\n",
    "\n",
    "res <- bt.intersect(gff.bed, RM.bed, wa=T, wb=T)\n",
    "res <- res[, -5]\n",
    "colnames(res) <- c(\"transcript_id\", \"start\", \"end\", \"type\", \"repeat_start\", \"repeat_end\", \"repeat_id\")\n",
    "f <- paste(res$transcript_id, res$repeat_start, res$repeat_end, res$repeat_id, sep=\"_\")\n",
    "res$split_f <- f\n",
    "res.list <- split(res, factor(res$split_f))\n",
    "res.list2 <- lapply(res.list, function(df){\n",
    "    if(dim(df)[1]==1){\n",
    "    d <- df}\n",
    "    else if(dim(df)[1]==2){\n",
    "        d <- df[df$type %in% \"exon\", ]\n",
    "    }\n",
    "    else{\n",
    "        if(\"three_prime_UTR\" %in% df$type){\n",
    "            d <- df[df$type %in% \"three_prime_UTR\", ]\n",
    "        }\n",
    "        else if(\"five_prime_UTR\" %in% df$type){\n",
    "            d <- df[df$type %in% \"five_prime_UTR\", ]\n",
    "        }\n",
    "        else(\n",
    "             d <- df[df$type %in% \"exon\", ]\n",
    "        )\n",
    "    }\n",
    "    return(d)\n",
    "})\n",
    "\n",
    "res2 <- rbindlist(res.list2)\n",
    "res.out <- res2[,1:7] \n",
    "res.out <- merge(res.out, transcript_gtf[,c(\"transcript_id\", \"gene_id\", \"gene_name\", \"gene_type\")], by=\"transcript_id\")\n",
    "res.out$type <- as.character(res.out$type)\n",
    "res.out$type[res.out$type %in% \"transcript\"] <- \"intron\"\n",
    "\n",
    "write.table(res.out, \"stat/Ref_Alu_L1_related_transcripts_with_intron.txt\", \n",
    "            sep=\"\\t\", quote=F, row.names=F, col.names=T)\n",
    "  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(\"ensembldb\")\n",
    "library(EnsDb.Hsapiens.v86)\n",
    "\n",
    "tb <- RM.tb[RM.tb$matching_class ==c(\"SINE/Alu\", \"LINE/L1\"),]\n",
    "rng_tx <- IRanges(start=tb$qry_start, end=tb$qry_end, names=tb$qry_id)\n",
    "rng_tx <- IRanges(bed)\n",
    "rng_gnm <- transcriptToGenome(rng_tx, EnsDb.Hsapiens.v86)\n",
    "edbx <- filter(EnsDb.Hsapiens.v86, filter = ~ seq_name == \"X\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## guide RNA in transcript"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(RFLPtools)\n",
    "setwd(\"/NAS/wg_looking/gRNA_ONT\")\n",
    "\n",
    "gRNA_out <- read.blast(\"stat/gencode.v37.chr_patch_hapl_scaff.transcript.fa.guideRNA.out\", sep=\"\\t\")\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# cleavage sites (from Biopython results)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(ggplot2)\n",
    "setwd(\"/NAS/wg_looking/gRNA_ONT\")\n",
    "\n",
    "# for Alu\n",
    "# c(\"sgRNA_20210527\", \"sgRNA_20210617\", \"sgRNA_20210802\", \"DNA_20210425\", \"sgRNA_20210913\", \"sgRNA_20210916_sg14\", \"sgRNA_20210916_sg24\")\n",
    "# IDs <- c(\"sgRNA_HCT116_1\", \"sgRNA_HCT116_2\", \"sgRNA_HCT116_3\", \"sgRNA_NCM460_1\", \"sgRNA_NCM460_2\", \"sgRNA_NCM460_3\")\n",
    "# IDs <- c(\"sgRNA_K562_1\", \"sgRNA_MDAMB231_1\", \"sgRNA_MDAMB231_2\", \"sgRNA_MDAMB231_3\")\n",
    "IDs <- c(\"sgRNA_MCF10A_1\", \"sgRNA_MCF10A_2\", \"sgRNA_MCF10A_3\")\n",
    "for(ID in IDs){\n",
    "  Forward = \"GGCCGGGCGCGGTGGCTCACGCCTGTAATCCCAGCACTTTGGGAGGCCGAGGCGGGCGGA\"\n",
    "  # reverse_complete = \"TCCGCCCGCCTCGGCCTCCCAAAGTGCTGGGATTACAGGCGTGAGCCACCGCGCCCGGCC\"\n",
    "  file = paste0(ID, \"/read_start_cutsite_Alu.txt\")\n",
    "  file2 = paste0(ID, \"/read_start_cutsite_Alu_for_plot.txt\")\n",
    "  system(paste0(\"grep Alu \", file, \" > \", file2))\n",
    "  site.tb <- read.table(file2, blank.lines.skip=T,sep = '\\t',fill=T, stringsAsFactors = F)\n",
    "  site.tb <- site.tb[site.tb$V3==\"Alu\",]\n",
    "  Forward.tb <- site.tb[grepl(\"Forward\",site.tb$V1),]\n",
    "  Reverse.tb <- site.tb[grepl(\"Reverse\",site.tb$V1),]\n",
    "  \n",
    "  # Forward\n",
    "  Forward.pp <- data.frame(site=1:60, number=0)\n",
    "  t <- table(Forward.tb$V2)[names(table(Forward.tb$V2)) %in% 1:60]\n",
    "  Forward.pp[names(t),\"number\"] <- t\n",
    "  Forward.pp$site <- factor(Forward.pp$site)\n",
    "  Forward.pp$number <- as.numeric(Forward.pp$number)\n",
    "  # reverse\n",
    "  Reverse.pp <- data.frame(site=1:60, number=0)\n",
    "  t <- table(Reverse.tb$V2)[names(table(Reverse.tb$V2)) %in% 1:60]\n",
    "  Reverse.pp[names(t),\"number\"] <- t\n",
    "  Reverse.pp$site <- factor(Reverse.pp$site)\n",
    "  Reverse.pp$number <- as.numeric(Reverse.pp$number)\n",
    "  \n",
    "  pp <- rbind(data.frame(Forward.pp, Orientation=\"Forward\"),\n",
    "              data.frame(Reverse.pp, Orientation=\"Reverse\"))\n",
    "  pp[pp$Orientation==\"Reverse\", \"number\"] <- -pp[pp$Orientation==\"Reverse\", \"number\"]\n",
    "  \n",
    "  \n",
    "  ggplot(pp, aes(x=site, y=number, fill=Orientation)) +\n",
    "    geom_bar(stat=\"identity\") +\n",
    "    # facet_wrap(~class, nrow = 2, scales=\"free\") +\n",
    "    scale_fill_manual(values = c(\"#4678BA\", \"#EF8354\")) +\n",
    "    ggtitle(ID) +\n",
    "    theme_bw() +\n",
    "    theme(\n",
    "      panel.grid=element_blank(),\n",
    "      axis.text.x = element_text(angle = 0, color = \"black\",\n",
    "                                 size = 8, hjust = 0.5, vjust=0.5),\n",
    "      plot.title=element_text(size=14, face=\"bold\")\n",
    "    ) +\n",
    "    scale_x_discrete(labels=strsplit(Forward, \"\")[[1]])+\n",
    "    ylab(\"the number of reads\")\n",
    "  \n",
    "  \n",
    "  ggsave(paste0(\"stat/\", ID, \".Alu_cleavage_sites.pdf\"), width=8, height=5)\n",
    "  \n",
    "  \n",
    "}\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for  L1\n",
    "for(ID in IDs){\n",
    "  Forward = \"NNNNNNNNNNTCTGAGATCAAACTGCAAGGNGGNNNNNNNNNNNNNNCCNTTGTAGGGACATGGATGAAANNNNNNNNNN\"\n",
    "  # reverse_complete = \"\"\n",
    "  file = paste0(ID, \"/read_start_cutsite_L1.txt\")\n",
    "  file2 = paste0(ID, \"/read_start_cutsite_L1_for_plot.txt\")\n",
    "  system(paste0(\"grep L1 \", file, \" > \", file2))\n",
    "  site.tb <- read.table(file2, blank.lines.skip=T,sep = '\\t',fill=T, stringsAsFactors = F)\n",
    "  site.tb <- site.tb[site.tb$V3==\"L1\",]\n",
    "  Forward.tb <- site.tb[grepl(\"Forward\",site.tb$V1),]\n",
    "  Reverse.tb <- site.tb[grepl(\"Reverse\",site.tb$V1),]\n",
    "  \n",
    "  # Forward\n",
    "  Forward.pp <- data.frame(site=c(389:428, 5703:5742), number=0)\n",
    "  t <- table(Forward.tb$V2)[names(table(Forward.tb$V2)) %in% c(389:428, 5703:5742)]\n",
    "  Forward.pp$site <- factor(Forward.pp$site)\n",
    "  Forward.pp[Forward.pp$site %in% names(t),\"number\"] <- t\n",
    "  Forward.pp$number <- as.numeric(Forward.pp$number)\n",
    "  # reverse\n",
    "  Reverse.pp <- data.frame(site=c(389:428, 5703:5742), number=0)\n",
    "  t <- table(Reverse.tb$V2)[names(table(Reverse.tb$V2)) %in% c(389:428, 5703:5742)]\n",
    "  Reverse.pp$site <- factor(Reverse.pp$site)\n",
    "  Reverse.pp[Reverse.pp$site %in% names(t),\"number\"] <- t\n",
    "  \n",
    "  Reverse.pp$number <- as.numeric(Reverse.pp$number)\n",
    "  \n",
    "  pp <- rbind(data.frame(Forward.pp, Orientation=\"Forward\"),\n",
    "              data.frame(Reverse.pp, Orientation=\"Reverse\"))\n",
    "  pp[pp$Orientation==\"Reverse\", \"number\"] <- -pp[pp$Orientation==\"Reverse\", \"number\"]\n",
    "  \n",
    "  \n",
    "  ggplot(pp, aes(x=site, y=number, fill=Orientation)) +\n",
    "    geom_bar(stat=\"identity\") +\n",
    "    # facet_wrap(~class, nrow = 2, scales=\"free\") +\n",
    "    scale_fill_manual(values = c(\"#4678BA\", \"#EF8354\")) +\n",
    "    ggtitle(ID) +\n",
    "    theme_bw() +\n",
    "    theme(\n",
    "      panel.grid=element_blank(),\n",
    "      axis.text.x = element_text(angle = 0, color = \"black\",\n",
    "                                 size = 8, hjust = 0.5, vjust=0.5),\n",
    "      plot.title=element_text(size=14, face=\"bold\")\n",
    "    ) +\n",
    "    scale_x_discrete(labels=strsplit(Forward, \"\")[[1]])+\n",
    "    ylab(\"the number of reads\")\n",
    "  \n",
    "  \n",
    "  ggsave(paste0(\"stat/\", ID, \".L1_cleavage_sites.pdf\"), width=8, height=5)\n",
    "  \n",
    "  \n",
    "}\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "# TE and Gene anno"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(biomartr)\n",
    "library(data.table)\n",
    "library(ggplot2)\n",
    "setwd(\"/NAS/wg_looking/gRNA_ONT\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## function"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### RM + uniquely mapping res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# RM + uniquely mapping res\n",
    "readAnno_func3 <- function(ID, RM_file, bamFile, div, w, sw_score=225){\n",
    "  # w : the region length of head or tail\n",
    "  library(rtracklayer)\n",
    "  library(BiocParallel)\n",
    "  library(dplyr)\n",
    "  library(biomartr)\n",
    "  library(data.table)\n",
    "  \n",
    "  \n",
    "  # loading gtf file\n",
    "  gtfFile = \"/NAS/wg_looking/human_ref/gencode.v37.chr_patch_hapl_scaff.annotation.gtf\"\n",
    "  gtf <- rtracklayer::import(gtfFile)\n",
    "  gtf <- as.data.frame(gtf)\n",
    "  transcript_gtf <- gtf[gtf$type == \"transcript\",]\n",
    "  coding_gene =  gtf[gtf$type==\"gene\" & gtf$gene_type==\"protein_coding\", ]\n",
    "  \n",
    "  \n",
    "  \n",
    "  # bam file\n",
    "  library(Rsamtools)\n",
    "  bamFile <- paste0(ID, \"/\", ID, \".transcriptom.sort.unique.bam\")\n",
    "  bam <- scanBam(bamFile)\n",
    "  bam.tb <- data.frame(bam[[1]])\n",
    "  bam.tb <- bam.tb[bam.tb$flag %in% c(0, 16), c(\"qname\", \"rname\")]\n",
    "  # tail(sort(table(bam.tb$qname)))\n",
    "  colnames(bam.tb) <- c(\"qry_id\", \"transcript_id\")\n",
    "  bam.anno <- merge(bam.tb, transcript_gtf[, c(\"transcript_id\", \"gene_id\", \"gene_name\", \"gene_type\")], by=\"transcript_id\")\n",
    "  rownames(bam.anno) <- bam.anno$qry_id\n",
    "  \n",
    "  # RepeatMasker results\n",
    "  RM.tb <- read_rm(RM_file)\n",
    "  RM.tb <- data.frame(RM.tb)\n",
    "  RM.tb <- RM.tb[as.numeric(RM.tb$sw_score)>=sw_score & as.numeric(RM.tb$perc_div)<=div, ]\n",
    "  RM.tb <- RM.tb[grepl(\"L1HS\", RM.tb$repeat_id) | \n",
    "                   grepl(\"L1P\", RM.tb$repeat_id) | \n",
    "                   RM.tb$matching_class %in% c(\"SINE/Alu\"), ]\n",
    "\n",
    "  dim(RM.tb) \n",
    "  TEs <- c(\"SINE/Alu\", \"LINE/L1\")\n",
    "  \n",
    "  \n",
    "  \n",
    "  readTE.list <- lapply(setNames(TEs, TEs), function(TE){\n",
    "    \n",
    "    sub <- RM.tb[RM.tb$matching_class %in% TE, ]\n",
    "    label <- apply(sub, 1, function(x){\n",
    "      start <- as.integer(x[6])\n",
    "      end <- as.integer(x[7])\n",
    "      left <- x[8]\n",
    "      left <- gsub(\"\\\\)\", \"\", gsub(\"\\\\(\", \"\", left))\n",
    "      left <- as.integer(left)\n",
    "      # repeat_id <- x[10]\n",
    "      # repeat_class <- x[11]\n",
    "      # p0 <- paste(start, end, left, sep= \",\")\n",
    "      # rep0 <- paste(repeat_class, repeat_id, sep=\",\")\n",
    "      if(start <= w & left <= w){\n",
    "        label0 = \"cross\"\n",
    "      }else if(start <= w & left > w){\n",
    "        label0 = \"head\"\n",
    "      }else if(start > w & left <= w){\n",
    "        label0 = \"tail\"\n",
    "      }else{\n",
    "        label0 = \"middle\"\n",
    "      }\n",
    "      return(label0)\n",
    "    })\n",
    "    sub$label <- label\n",
    "    sub.list  <- split(sub, factor(sub$qry_id))\n",
    "    \n",
    "    a <- lapply(sub.list, function(df){\n",
    "      read_name <- df[1,5]\n",
    "      p <- paste(df$qry_start, df$qry_end, df$qry_left,sep=\",\")\n",
    "      rep  <- paste(df$matching_class, df$repeat_id, sep=\",\")\n",
    "      label <- df$label\n",
    "      \n",
    "      p <- paste(p, collapse = \":\")\n",
    "      rep <- paste(rep, collapse = \":\")\n",
    "      label <- paste(label, collapse = \":\")\n",
    "      \n",
    "      \n",
    "      #\n",
    "      read_info <- data.frame(read_name, p, rep, label, dim(df)[1])\n",
    "      colnames(read_info) <- c(\"read_name\", \"position\", \"repeat\", \"label\", \"number\")\n",
    "      \n",
    "      return(read_info)\n",
    "    })\n",
    "    a <- rbindlist(a)\n",
    "    a <- data.frame(a)\n",
    "    rownames(a) <- a$read_name\n",
    "    # a <- merge(a, talon.tb, by=\"row.names\", all.x=TRUE)\n",
    "    a <- merge(a, bam.anno, by=\"row.names\", all.x=TRUE)\n",
    "    a$gene_novelty <- a$gene_id\n",
    "    a$gene_novelty[!is.na(a$gene_id)] <- \"Known\"\n",
    "    a$protein_coding[a$gene_type ==\"protein_coding\"] <- \"YES\"\n",
    "    \n",
    "    return(a)\n",
    "  })\n",
    "  \n",
    "  \n",
    "  \n",
    "  \n",
    "  # write.table(readTE.list[[\"SINE/Alu\"]], \"test_20210527.Alu.gene.txt\", sep=\"\\t\", quote = F, col.names = T, row.names = F)\n",
    "  # write.table(readTE.list[[\"LINE/L1\"]], \"test_20210527.Alu.gene.txt\", sep=\"\\t\", quote = F, col.names = T, row.names = F)\n",
    "  \n",
    "  \n",
    "  \n",
    "  return(readTE.list)\n",
    "}\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### reads anno with uniquely mapping res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# uniquely mapping res\n",
    "bam_anno_func <- function(ID){\n",
    "  # w : the region length of head or tail\n",
    "  library(rtracklayer)\n",
    "  library(BiocParallel)\n",
    "  library(dplyr)\n",
    "  library(biomartr)\n",
    "  library(data.table)\n",
    "  \n",
    "  \n",
    "  # loading gtf file\n",
    "  gtfFile = \"/NAS/wg_looking/human_ref/gencode.v37.chr_patch_hapl_scaff.annotation.gtf\"\n",
    "  gtf <- rtracklayer::import(gtfFile)\n",
    "  gtf <- as.data.frame(gtf)\n",
    "  transcript_gtf <- gtf[gtf$type == \"transcript\",]\n",
    "  coding_gene =  gtf[gtf$type==\"gene\" & gtf$gene_type==\"protein_coding\", ]\n",
    "  \n",
    "  \n",
    "  \n",
    "  # bam file\n",
    "  library(Rsamtools)\n",
    "  bamFile <- paste0(ID, \"/\", ID, \".transcriptom.sort.unique.bam\")\n",
    "  bam <- scanBam(bamFile)\n",
    "  bam.tb <- data.frame(bam[[1]])\n",
    "  bam.tb <- bam.tb[bam.tb$flag %in% c(0, 16), c(\"qname\", \"rname\")]\n",
    "  # tail(sort(table(bam.tb$qname)))\n",
    "  colnames(bam.tb) <- c(\"qry_id\", \"transcript_id\")\n",
    "  bam.anno <- merge(bam.tb, transcript_gtf[, c(\"transcript_id\", \"gene_id\", \"gene_name\", \"gene_type\")], by=\"transcript_id\")\n",
    "  rownames(bam.anno) <- bam.anno$qry_id\n",
    "  \n",
    "  \n",
    "  return(bam.anno)\n",
    "  \n",
    "}\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### The percentage of TE on the head of reads "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Reads_head_PCT_plot_func <- function(ID, Test_anno, plot=F){\n",
    "    library(data.table)\n",
    "    library(ggplot2)\n",
    "    library(dplyr)\n",
    "    library(reshape2)\n",
    "    library(scales)\n",
    "    library(RColorBrewer)\n",
    "  \n",
    "  Sample_info <- read.table(\"stat/Sample_info2.txt\", header = T, stringsAsFactors = F)\n",
    "  rownames(Sample_info) <- Sample_info$Samples\n",
    "  \n",
    "  # total reads\n",
    "  pa <- system(paste0(\"grep '>' QC/\", ID, \".trimming.Q7.L300.fa | wc -l\"), intern=T)\n",
    "  pa <- as.integer(tstrsplit(pa, \" \")[[1]])\n",
    "  \n",
    "  \n",
    "  ## Alu\n",
    "  tb <- Test_anno[[\"SINE/Alu\"]]\n",
    "  # Alu_target <- tb[which(grepl(\"head\",tb$label) | grepl(\"tail\",tb$label)), \"read_name\"]\n",
    "  # tb <- tb[complete.cases(tb$transcript_id),]\n",
    "  Alu_target <- tb[grepl(\"head\",tb$label), \"read_name\"]\n",
    "  \n",
    "  \n",
    "  ## L1\n",
    "  tb <- Test_anno[[\"LINE/L1\"]]\n",
    "  # L1_target <- tb[which(grepl(\"head\",tb$label) | grepl(\"tail\",tb$label)), \"read_name\"]\n",
    "  # tb <- tb[complete.cases(tb$transcript_id),]\n",
    "  L1_target <- tb[grepl(\"head\",tb$label), \"read_name\"]\n",
    "  \n",
    "  \n",
    "  \n",
    "  \n",
    "  # pie\n",
    "  Alu_tartget_reads = length(setdiff(Alu_target, L1_target))\n",
    "  L1_tartget_reads = length(setdiff(L1_target, Alu_target))\n",
    "  sgRNA_num <- c(\"Alu-target Sequence\"=Alu_tartget_reads, \"L1-target Sequence\"=L1_tartget_reads, \n",
    "                 \"Alu&L1-target Sequence\"=length(intersect(Alu_target, L1_target)), \n",
    "                 \"Not Found\"=pa-length(union(Alu_target, L1_target)))\n",
    "  \n",
    "  if(plot==T){\n",
    "      pp <- data.frame(sgRNA_num)\n",
    "  pp <- melt(t(pp))\n",
    "  colnames(pp) <- c(\"Method\", \"classify\", \"count\")\n",
    "  p <- pp %>%\n",
    "    group_by(Method) %>%\n",
    "    mutate(Percentage = count / sum(count)*100)\n",
    "  \n",
    "  \n",
    "  plot <- ggplot(p, aes(x=\"\", y=Percentage, fill=classify))+\n",
    "    geom_bar(width = 1, stat = \"identity\") +\n",
    "    coord_polar(\"y\", start=0) +\n",
    "    scale_fill_manual(values = c(\"#F4B540\", \"#BA748A\", \"#87B07A\", \"#4678BA\", \"#FDA56C\"))+\n",
    "    theme_minimal() +\n",
    "    theme(\n",
    "      axis.title.x = element_blank(),\n",
    "      axis.title.y = element_blank(),\n",
    "      panel.border = element_blank(),\n",
    "      panel.grid=element_blank(),\n",
    "      axis.ticks = element_blank(),\n",
    "      plot.title=element_text(size=14, face=\"bold\", hjust=0.5)\n",
    "    ) +\n",
    "    theme(axis.text.x=element_blank())+\n",
    "    geom_text(aes(x = 1.6, label = scales::percent(Percentage/100, accuracy = .1)), position = position_stack(vjust = .5)) +\n",
    "    # facet_wrap(~Method)\n",
    "    ggtitle(ID)\n",
    "  \n",
    "  \n",
    "  # ggsave(paste0(\"stat/\", ID, \"_Alu_L1_in_head_tail_pie.pdf\"), width = 5, height = 5)\n",
    "  ggsave(paste0(\"stat/\", ID, \"_Alu_L1_in_head_pie.pdf\"), width = 5, height = 5)\n",
    "\n",
    "  }\n",
    "    \n",
    "  return(sgRNA_num)\n",
    "  \n",
    "}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### the position of TE in reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#\n",
    "position_plot_func <- function(ID, Test_anno){\n",
    "    TE.min.pos <- lapply(names(Test_anno), function(TE){\n",
    "        TE.name <- strsplit(TE, \"/\")[[1]][2]\n",
    "        tb <- Test_anno[[TE]]\n",
    "        position.list <- strsplit(as.character(tb$position), \":\")\n",
    "        min.pos <- lapply(position.list, function(x){\n",
    "            x <- x[complete.cases(x)]\n",
    "            start <- as.integer(tstrsplit(x, \",\")[[1]])\n",
    "            min.value <- min(start)\n",
    "          })\n",
    "        min.pos <- unlist(min.pos)\n",
    "        min.pos <- data.frame(min.pos, \"Repeat\"=TE)\n",
    "        \n",
    "    })\n",
    "     \n",
    "    \n",
    "    p <- rbindlist(TE.min.pos)\n",
    "    ggplot(p, aes(x=min.pos, color=Repeat)) +\n",
    "        geom_density(size=2)+\n",
    "        theme_classic() +\n",
    "        scale_x_log10() +\n",
    "        scale_color_manual(values=c(\"#4678BA\", \"#FDA56C\")) +\n",
    "        ggtitle(paste0(ID, \": the position of TE in reads\"))\n",
    "    \n",
    "    ggsave(paste0(\"stat/\", ID, \"_the_positon_of_TE_in_reads.pdf\"),\n",
    "           width = 7, height = 4)\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "position_plot_func2 <- function(ID, Test_anno){\n",
    "    TE.start.pos <- lapply(names(Test_anno), function(TE){\n",
    "        TE.name <- strsplit(TE, \"/\")[[1]][2]\n",
    "        tb <- Test_anno[[TE]]\n",
    "        position.list <- strsplit(as.character(tb$position), \":\")\n",
    "        start.pos <- lapply(position.list, function(x){\n",
    "            x <- x[complete.cases(x)]\n",
    "            start <- as.integer(tstrsplit(x, \",\")[[1]])\n",
    "            return(start)\n",
    "          })\n",
    "        start.pos <- unlist(start.pos)\n",
    "        start.pos <- data.frame(start.pos, \"Repeat\"=TE)\n",
    "        \n",
    "    })\n",
    "     \n",
    "    \n",
    "    p <- rbindlist(TE.start.pos)\n",
    "    ggplot(p, aes(x=start.pos, fill=Repeat, color=Repeat)) +\n",
    "        geom_histogram(alpha=0.5)+\n",
    "        theme_classic() +\n",
    "        scale_x_log10() +\n",
    "        theme(strip.background = element_blank()) +\n",
    "        scale_color_manual(values=c(\"#4A89D5\", \"#EE9248\")) +\n",
    "        scale_fill_manual(values=c(\"#4A89D5\", \"#FDA56C\")) +\n",
    "        facet_wrap(~Repeat, nrow = 2, scales=\"free\") + \n",
    "        ggtitle(paste0(ID, \": the position of TE in reads\"))\n",
    "    \n",
    "    ggsave(paste0(\"stat/\", ID, \"_the_positon_of_TE_in_reads_histogram.pdf\"),\n",
    "           width = 7, height = 5)\n",
    "}\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### The percentage of TE-reads "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Reads_PCT_plot_func <- function(ID, Test_anno, plot=F){\n",
    "    library(ggplot2)\n",
    "    library(dplyr)\n",
    "    library(reshape2)\n",
    "    library(scales)\n",
    "    library(RColorBrewer)\n",
    "\n",
    "    Sample_info <- read.table(\"stat/Sample_info.txt\", header = T, stringsAsFactors = F)\n",
    "    rownames(Sample_info) <- Sample_info$Samples\n",
    "\n",
    "  # total reads\n",
    "    pa <- system(paste0(\"grep '>' QC/\", ID, \".trimming.Q7.L300.fa | wc -l\"), intern=T)\n",
    "    pa <- as.integer(tstrsplit(pa, \" \")[[1]])\n",
    "\n",
    "\n",
    "    ## Alu\n",
    "    tb <- Test_anno[[\"SINE/Alu\"]]\n",
    "    Alu_target <- tb$read_name\n",
    "\n",
    "\n",
    "    ## L1\n",
    "    tb <- Test_anno[[\"LINE/L1\"]]\n",
    "    L1_target <- tb$read_name\n",
    "\n",
    "\n",
    "  \n",
    "  # pie\n",
    "    Alu_tartget_reads = length(setdiff(Alu_target, L1_target))\n",
    "    L1_tartget_reads = length(setdiff(L1_target, Alu_target))\n",
    "    sgRNA_num <- c(\"Alu-target Sequence\"=Alu_tartget_reads, \"L1-target Sequence\"=L1_tartget_reads, \n",
    "                 \"Alu&L1-target Sequence\"=length(intersect(Alu_target, L1_target)), \n",
    "                 \"Not Found\"=pa-length(union(Alu_target, L1_target)))\n",
    "    \n",
    "    if(plot==T){\n",
    "        pp <- data.frame(sgRNA_num)\n",
    "        pp <- melt(t(pp))\n",
    "        colnames(pp) <- c(\"Method\", \"classify\", \"count\")\n",
    "        p <- pp %>%\n",
    "        group_by(Method) %>%\n",
    "        mutate(Percentage = count / sum(count)*100)\n",
    "\n",
    "\n",
    "        plot <- ggplot(p, aes(x=\"\", y=Percentage, fill=classify))+\n",
    "        geom_bar(width = 1, stat = \"identity\") +\n",
    "        coord_polar(\"y\", start=0) +\n",
    "        scale_fill_manual(values = c(\"#F4B540\", \"#BA748A\", \"#87B07A\", \"#4678BA\", \"#FDA56C\"))+\n",
    "        theme_minimal() +\n",
    "        theme(\n",
    "          axis.title.x = element_blank(),\n",
    "          axis.title.y = element_blank(),\n",
    "          panel.border = element_blank(),\n",
    "          panel.grid=element_blank(),\n",
    "          axis.ticks = element_blank(),\n",
    "          plot.title=element_text(size=14, face=\"bold\", hjust=0.5)\n",
    "        ) +\n",
    "        theme(axis.text.x=element_blank())+\n",
    "        geom_text(aes(x = 1.6, label = scales::percent(Percentage/100, accuracy = .1)), position = position_stack(vjust = .5)) +\n",
    "        # facet_wrap(~Method)\n",
    "        ggtitle(ID)\n",
    "\n",
    "\n",
    "        ggsave(paste0(\"stat/\", ID, \"_Alu_L1_in_reads_pie.pdf\"), width = 5, height = 5)\n",
    "\n",
    "        \n",
    "    }\n",
    "  \n",
    "  \n",
    "    \n",
    "    return(sgRNA_num)\n",
    "  \n",
    "}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### the PCT of Alu and L1 related genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Genes_PCT_plot_func <- function(ID, totalID, Test_anno.list){\n",
    "  library(ggplot2)\n",
    "  library(dplyr)\n",
    "  library(reshape2)\n",
    "  library(scales)\n",
    "  library(RColorBrewer)\n",
    "  library(VennDiagram)\n",
    "  library(\"ggvenn\")\n",
    "  setwd(\"/NAS/wg_looking/gRNA_ONT\")\n",
    "  \n",
    "\n",
    "  RNA.list <- Test_anno.list[c(ID, totalID)]\n",
    "  \n",
    "  for(TE in c(\"SINE/Alu\", \"LINE/L1\")){\n",
    "    #\n",
    "    TE.name <- strsplit(TE, \"/\")[[1]][2]\n",
    "    \n",
    "    x <- lapply(RNA.list, function(sub){\n",
    "      tb <- sub[[TE]]\n",
    "      related_coding_gene = tb$gene_id[which(tb$gene_novelty == \"Known\" & tb$protein_coding == \"YES\")]\n",
    "      related_coding_gene <- as.character(unique(related_coding_gene))\n",
    "      return(related_coding_gene)\n",
    "      })\n",
    "    \n",
    "  \n",
    "    # plot\n",
    "    ggvenn(x, \n",
    "           fill_color = c(\"#4678BA\", \"#EF8354\"),\n",
    "           stroke_color = NA,\n",
    "            fill_alpha = 0.8,\n",
    "           stroke_size = 0.5, \n",
    "           set_name_size = 4,\n",
    "           show_outside = \"always\"\n",
    "        ) \n",
    "    \n",
    "\n",
    "    ggsave(paste0('stat/', ID, '_VS_', totalID, \"_\",TE.name, '_coding_gene_venn.pdf'),\n",
    "           width=10, height=5)\n",
    "    \n",
    "  }\n",
    "  \n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Transcripts_PCT_plot_func <- function(ID, totalID, Test_anno.list, plot=F){\n",
    "    library(ggplot2)\n",
    "    library(dplyr)\n",
    "    library(reshape2)\n",
    "    library(scales)\n",
    "    library(RColorBrewer)\n",
    "    library(VennDiagram)\n",
    "    library(venn)\n",
    "    library(\"ggvenn\")\n",
    "    setwd(\"/NAS/wg_looking/gRNA_ONT\")\n",
    "\n",
    "\n",
    "    RNA.list <- Test_anno.list[c(ID, totalID)]\n",
    "    \n",
    "    res_trans <- lapply(setNames(c(\"SINE/Alu\", \"LINE/L1\"), c(\"SINE/Alu\", \"LINE/L1\")), function(TE){\n",
    "        TE.name <- strsplit(TE, \"/\")[[1]][2]\n",
    "        x <- lapply(RNA.list, function(sub){\n",
    "          tb <- sub[[TE]]\n",
    "          # related_coding_gene = tb$transcript_id[which(tb$gene_novelty == \"Known\" & tb$protein_coding == \"YES\")]\n",
    "          related_coding_gene = tb$transcript_id[which(tb$gene_novelty == \"Known\")]\n",
    "          related_coding_gene <- as.character(unique(related_coding_gene))\n",
    "          return(related_coding_gene)\n",
    "         })\n",
    "        \n",
    "         if(plot==T){\n",
    "            plot <- venn.diagram(\n",
    "              x = x,\n",
    "              # category.names = c(paste0(\"sgRNA: \", TE.name, \"-target related coding genes\") , \n",
    "              #                    paste0(\"Total RNA: include \", TE.name, \" coding genes\")),\n",
    "              filename = NULL,\n",
    "              # filename = paste0('stat/', ID, '_', TE.name, '_coding_gene_venn_diagramm.png'),\n",
    "              # imagetype = \"svg\",\n",
    "              # output=TRUE,\n",
    "              height = 1000 , \n",
    "              width = 1000 , \n",
    "              sub.cex = 0.5,\n",
    "              resolution = 300,\n",
    "              cat.pos = c(0, 1.05),\n",
    "              fill = c(\"#4678BA\", \"#EF8354\"),\n",
    "              col = c(\"#4678BA\", \"#EF8354\"),\n",
    "            )\n",
    "    \n",
    "            pdf(file=paste0('stat/', ID, '_VS_', totalID, \"_\",TE.name, '_transcript_venn.pdf'))\n",
    "            grid.draw(plot)\n",
    "            dev.off()    \n",
    "     }\n",
    "        \n",
    "        return(x)\n",
    "\n",
    " })\n",
    "    return(res_trans)\n",
    "}\n",
    "\n",
    "        \n",
    "  \n",
    "    # plot\n",
    "#     ggvenn(x, \n",
    "#            fill_color = c(\"#4678BA\", \"#EF8354\"),\n",
    "#            stroke_color = NA,\n",
    "#            fill_alpha = 0.8,\n",
    "#            stroke_size = 0.5, \n",
    "#            set_name_size = 4,\n",
    "#            show_outside = \"auto\",\n",
    "#            auto_scale = FALSE\n",
    "#         ) \n",
    "    \n",
    "\n",
    "#     ggsave(paste0('stat/', ID, '_VS_', totalID, \"_\",TE.name, '_transcript_venn.pdf'),\n",
    "#            width=10, height=5)\n",
    "   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#\n",
    "Genes_PCT_plot_func2 <- function(ID, Test_anno.list, CellLine){\n",
    "  library(ggplot2)\n",
    "  library(dplyr)\n",
    "  library(reshape2)\n",
    "  library(scales)\n",
    "  library(RColorBrewer)\n",
    "  library(VennDiagram)\n",
    "  library(patchwork) # p <- p1 + p3 + plot_layout(ncol = 1, heights = c(1, 1))\n",
    "  library(\"ggvenn\")\n",
    "  setwd(\"/NAS/wg_looking/gRNA_ONT\")\n",
    "  \n",
    "  \n",
    "  # load Sample info\n",
    "  Sample_info <- read.table(\"stat/Sample_info2.txt\", header = T, stringsAsFactors = F)\n",
    "  rownames(Sample_info) <- Sample_info$Samples\n",
    "  totalRNA_IDs <- Sample_info$Samples[Sample_info$SeqType==\"totalRNA\" & Sample_info$CellLine==CellLine]\n",
    "  RNA.list <- Test_anno.list[c(ID,totalRNA_IDs)]\n",
    "\n",
    "  \n",
    "  for(TE in c(\"SINE/Alu\", \"LINE/L1\")){\n",
    "    #\n",
    "    TE.name <- strsplit(TE, \"/\")[[1]][2]\n",
    "    x <- lapply(RNA.list, function(sub){\n",
    "      tb <- sub[[TE]]\n",
    "      related_coding_gene = tb$gene_id[which(tb$gene_novelty == \"Known\" & tb$protein_coding == \"YES\")]\n",
    "      related_coding_gene <- as.character(unique(related_coding_gene))\n",
    "      return(related_coding_gene)\n",
    "    })\n",
    "    \n",
    "    \n",
    "    \n",
    "    # plot\n",
    "    ggvenn(x, \n",
    "           fill_color = c(\"#4678BA\", \"#EF8354\", \"#87B07A\", \"#F4B540\"),\n",
    "           stroke_color = NA,\n",
    "           fill_alpha = 0.8,\n",
    "           stroke_size = 0.5, \n",
    "           set_name_size = 4,\n",
    "           show_outside = \"always\"\n",
    "    ) \n",
    "    \n",
    "    \n",
    "    ggsave(paste0('stat/', ID, '_', TE.name, '_coding_gene_venn.pdf'),\n",
    "           width=10, height=5)\n",
    "    \n",
    "    \n",
    "    # venn.diagram(\n",
    "    #   x = list(as.character(unique(related_coding_gene)), as.character(unique(related_coding_gene_c))),\n",
    "    #   category.names = c(paste0(\"sgRNA: \", TE.name, \"-target related coding genes\") , \n",
    "    #                      paste0(\"Total RNA: include \", TE.name, \" coding genes\")),\n",
    "    #   # filename = paste0('stat/', ID, '_Alu_coding_gene_venn_diagramm.tiff'),\n",
    "    #   filename = paste0('stat/', ID, '_', TE.name, '_coding_gene_venn_diagramm.png'),\n",
    "    #   # imagetype = \"svg\",\n",
    "    #   # output=TRUE,\n",
    "    #   height = 1000 , \n",
    "    #   width = 1000 , \n",
    "    #   sub.cex = 0.5,\n",
    "    #   resolution = 300,\n",
    "    #   cat.pos = c(0, 1.05),\n",
    "    #   fill = brewer.pal(12,\"Paired\")[c(1,7)],\n",
    "    #   col = brewer.pal(12,\"Paired\")[c(1,7)]\n",
    "    # )\n",
    "    \n",
    "  }\n",
    "  \n",
    "}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Main"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### loading Data: RM, annotation, gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error in setwd(\"/NAS/wg_looking/gRNA_ONT\"): 无法改变工作目录\n",
     "output_type": "error",
     "traceback": [
      "Error in setwd(\"/NAS/wg_looking/gRNA_ONT\"): 无法改变工作目录\nTraceback:\n",
      "1. setwd(\"/NAS/wg_looking/gRNA_ONT\")"
     ]
    }
   ],
   "source": [
    "#################\n",
    "# loading Data: RM, annotation, gene\n",
    "#################\n",
    "setwd(\"/NAS/wg_looking/gRNA_ONT\")\n",
    "Sample_info <- read.table(\"stat/Sample_info2.txt\", header = T, stringsAsFactors = F)\n",
    "rownames(Sample_info) <- Sample_info$Samples\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "### CellLine\n",
    "CellLine_anno.list <- list()\n",
    "CellLine_bamAnno.list <- list()\n",
    "for(ID in Sample_info$Samples){\n",
    "  CellLine_anno.list[[ID]] <- readAnno_func3(ID, \n",
    "                                         RM_file=paste0(ID, \"/RepeatMasker_res_Q7_L300/\", ID, \".trimming.Q7.L300.fa.out\"), \n",
    "                                         bamFile <- paste0(ID, \"/\", ID, \".transcriptom.sort.unique.bam\"), \n",
    "                                         div=18, w=50, sw_score=225)\n",
    "  CellLine_bamAnno.list[[ID]] <- bam_anno_func(ID)\n",
    "}\n",
    "\n",
    "save(CellLine_anno.list, CellLine_bamAnno.list, file=\"stat/CellLine_anno2.RData\")\n",
    "\n",
    "\n",
    "\n",
    "### K562 test\n",
    "# totalRNA_K562_anno.list <- list()\n",
    "totalRNA_K562_bamAnno.list <- list()\n",
    "for(ID in c('totalRNA_K562_20211026_BC21', 'totalRNA_K562_20211026_BC22')){\n",
    "  totalRNA_K562_anno.list[[ID]] <- readAnno_func3(ID, \n",
    "                                         RM_file=paste0(ID, \"/RepeatMasker_res_Q7_L300/\", ID, \".trimming.Q7.L300.fa.out\"), \n",
    "                                         bamFile <- paste0(ID, \"/\", ID, \".transcriptom.sort.unique.bam\"), \n",
    "                                         div=18, w=50, sw_score=225)\n",
    "    totalRNA_K562_bamAnno.list[[ID]] <- bam_anno_func(ID)\n",
    "}\n",
    "\n",
    "save(totalRNA_K562_anno.list, file=\"stat/totalRNA_K562_anno.RData\")\n",
    "\n",
    "\n",
    "\n",
    "###\n",
    "for(ID in names(Test_anno.list)){\n",
    "  for(TE in names(Test_anno.list[[ID]])){\n",
    "    Test_anno <- Test_anno.list[[ID]][[TE]]\n",
    "    TE <- strsplit(TE, \"/\")[[1]][2]\n",
    "    write.table(Test_anno, paste0(\"stat/\", ID, \".\", TE, \".info.txt\"), quote=F, sep=\"\\t\", col.names=T, row.names=F)\n",
    "  }\n",
    "}\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### The percentage of Alu and L1 reads "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "####\n",
    "library(plyr)\n",
    "setwd(\"/NAS/wg_looking/gRNA_ONT\")\n",
    "load(\"stat/CellLine_anno2.RData\")\n",
    "\n",
    "\n",
    "Sample_info <- read.table(\"stat/Sample_info2.txt\", header = T, stringsAsFactors = F)\n",
    "rownames(Sample_info) <- Sample_info$Samples\n",
    "samples <- Sample_info$Samples[Sample_info$CellLine %in% c(\"K562\", \"MDAMB231\", \"MCF10A\") & \n",
    "                              Sample_info$SeqType %in% c(\"sgRNA\", \"totalRNA\")]\n",
    "\n",
    "pp <- lapply(samples, function(ID){\n",
    "    a <- Reads_head_PCT_plot_func(ID, CellLine_anno.list[[ID]], plot=F)\n",
    "    aa <- data.frame(Count=c(sum(a)-a[\"Not Found\"], a[\"Not Found\"]),\n",
    "                     Class=c(\"TE in head\\n(<50bp)\", \"Not Found\"),\n",
    "                     Facet=\"TE in head (<50bp)\")\n",
    "    b <- Reads_PCT_plot_func(ID, CellLine_anno.list[[ID]] , plot=F)\n",
    "    bb <- data.frame(Count=c(sum(b)-b[\"Not Found\"], b[\"Not Found\"]),\n",
    "                     Class=c(\"with TE\", \"Not Found\"),\n",
    "                     Facet=\"with TE\")\n",
    "    \n",
    "    c <- data.frame(rbind(aa, bb), ID=ID) \n",
    "})\n",
    "\n",
    "\n",
    "p <- rbindlist(pp)\n",
    "p$SeqType <- Sample_info[as.character(p$ID), \"SeqType\"]\n",
    "p <- ddply(p, .(ID, Facet), mutate, Percentage = Count / sum(Count) * 100)\n",
    "\n",
    "\n",
    "cols <- c(\"with TE\"='#39C6AA', \"TE in head\\n(<50bp)\"='#5655FE')\n",
    "# cols <- c('#39C6AA', '#5655FE')\n",
    "p2 <- p[!(p$Class %in% \"Not Found\"), ]\n",
    "ggplot(p2, aes(x=SeqType, y=Percentage, fill=Class, color=Class)) +\n",
    "    stat_boxplot(geom='errorbar', linetype=1, width=0.3,  position = position_dodge(width = 0.8)) +\n",
    "    geom_boxplot(position = position_dodge(width = 0.8))+\n",
    "    stat_summary(fun=median, geom=\"point\", pch=18, size=3, color = \"#AFAFAF\", position = position_dodge(width = 0.8)) +\n",
    "    theme_classic() +\n",
    "    theme(\n",
    "      panel.grid=element_blank(),\n",
    "      axis.text.x = element_text(angle = 0, color = \"black\", \n",
    "                                 size = 8, hjust = 0.5, vjust=0.5),\n",
    "      plot.title=element_text(size=14, face=\"bold\", hjust=0.5)\n",
    "    ) +\n",
    "    # facet_wrap(~Facet, nrow = 1, scales=\"free\") +\n",
    "    scale_fill_manual(values = cols) +\n",
    "    scale_color_manual(values = cols) +\n",
    "    ggtitle(\"Targeting efficiency\") +\n",
    "    ylim(0,100)\n",
    "  \n",
    "ggsave(paste0(\"stat/Fig1D.v2.Targeting_efficiency.pdf\"),\n",
    "         width = 5, height = 4)\n",
    "\n",
    "\n",
    "write.table(p2[,c(1,3,4,5,6)], \"stat/Fig1D.v2.Targeting_efficiency.txt\", sep=\"\\t\", quote=F, col.names=T, row.names=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "Sample_info <- read.table(\"stat/Sample_info2.txt\", header = T, stringsAsFactors = F)\n",
    "rownames(Sample_info) <- Sample_info$Samples\n",
    "samples <- Sample_info$Samples[Sample_info$CellLine %in% c(\"K562\", \"MDAMB231\", \"MCF10A\") & \n",
    "                              Sample_info$SeqType %in% \"totalRNA\"]\n",
    "\n",
    "\n",
    "pp <- lapply(samples, function(ID){\n",
    "    a <- Reads_head_PCT_plot_func(ID, CellLine_anno.list[[ID]], plot=F)\n",
    "    aa <- data.frame(Count=c(sum(a)-a[\"Not Found\"], a[\"Not Found\"]),\n",
    "                     Class=c(\"TE in head\\n(<50bp)\", \"Not Found\"),\n",
    "                     Facet=\"TE in head (<50bp)\")\n",
    "    b <- Reads_PCT_plot_func(ID, CellLine_anno.list[[ID]] , plot=F)\n",
    "    bb <- data.frame(Count=c(sum(b)-b[\"Not Found\"], b[\"Not Found\"]),\n",
    "                     Class=c(\"with TE\", \"Not Found\"),\n",
    "                     Facet=\"with TE\")\n",
    "    \n",
    "    c <- data.frame(rbind(aa, bb), ID=ID) \n",
    "})\n",
    "\n",
    "\n",
    "p <- rbindlist(pp)\n",
    "p <- ddply(p, .(ID, Facet), mutate, Percentage = Count / sum(Count) * 100)\n",
    "p2 <- p[!(p$Class %in% \"Not Found\"), ]\n",
    "\n",
    "write.table(p2, \"stat/FigS2.A.txt\", col.names=T, row.names=F, sep=\"\\t\", quote=F)\n",
    "\n",
    "\n",
    "cols <- c(\"with TE\"='#39C6AA', \"TE in head\\n(<50bp)\"='#5655FE')\n",
    "ggplot(p2, aes(x=Class, y=Percentage, fill=Class, color=Class)) +\n",
    "    stat_boxplot(geom='errorbar', linetype=1, width=0.3) +\n",
    "    geom_boxplot()+\n",
    "    stat_summary(fun=median, geom=\"point\", pch=18, size=3, color = \"#AFAFAF\") +\n",
    "    theme_bw() +\n",
    "    theme(\n",
    "      panel.grid=element_blank(),\n",
    "      axis.text.x = element_text(angle = 0, color = \"black\", \n",
    "                                 size = 8, hjust = 0.5, vjust=0.5),\n",
    "      plot.title=element_text(size=14, face=\"bold\", hjust=0.5)\n",
    "    ) +\n",
    "    # facet_wrap(~Facet, nrow = 1, scales=\"free\") +\n",
    "    # ylim(0,100) +\n",
    "    xlab(\"totalRNA\") +\n",
    "    scale_fill_manual(values = cols) +\n",
    "    scale_color_manual(values = cols) \n",
    "    # ggtitle(\"Targeting efficiency\") +\n",
    " \n",
    "  \n",
    "ggsave(paste0(\"stat/FigS2.A.Targeting_efficiency.pdf\"),\n",
    "         width = 5, height = 4)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 只含有L1的reads中，L1亚型的占比"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Test_anno <- CellLine_anno.list[[ID]]\n",
    "\n",
    "## ALU\n",
    "tb <- Test_anno[[\"SINE/Alu\"]]\n",
    "Alu_target <- tb[grepl(\"head\",tb$label), \"read_name\"] \n",
    "## L1\n",
    "tb <- Test_anno[[\"LINE/L1\"]]\n",
    "L1_target <- tb[grepl(\"head\",tb$label), \"read_name\"]\n",
    "#\n",
    "i <- intersect(Alu_target, L1_target)\n",
    "L1_specific <- setdiff(L1_target, i)\n",
    "\n",
    "sub.tb <- Test_anno[[\"LINE/L1\"]][Test_anno[[\"LINE/L1\"]]$read_name %in% L1_specific, ]\n",
    "L1_repeat <- tstrsplit(sub.tb$repeat., \":\")[[1]]\n",
    "L1_repeat <- tstrsplit(L1_repeat, \",\")[[2]]\n",
    "L1_repeat <- substring(L1_repeat, 1, 4)\n",
    "L1_repeat[grepl(\"L1M\", L1_repeat)] <- \"L1M\"\n",
    "L1_repeat[grepl(\"L1P\", L1_repeat)] <- \"L1P\"\n",
    "\n",
    "pp <- data.frame(table(L1_repeat))\n",
    "pp <- pp %>% mutate(Percentage = Freq / sum(Freq)*100)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### the position of TE in reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    " \n",
    "# Fig2S.A-C\n",
    "for(ID in names(CellLine_anno.list)){\n",
    "  position_plot_func2(ID, CellLine_anno.list[[ID]])\n",
    "}\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### the percentage of Alu and L1 related genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "load(\"stat/CellLine_anno2.RData\") # CellLine_anno.list    CellLine_bamAnno.list\n",
    "\n",
    "\n",
    "\n",
    "Sample_info <- read.table(\"stat/Sample_info2.txt\", header = T, stringsAsFactors = F)\n",
    "rownames(Sample_info) <- Sample_info$Samples\n",
    "\n",
    "IDs <- Sample_info$Samples[Sample_info$SeqType==\"sgRNA\"]\n",
    "for(ID in IDs){\n",
    "  Genes_PCT_plot_func2(ID, CellLine_anno.list, Sample_info[ID,\"CellLine\"])\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "# Transcripts: Fig2A\n",
    "library(RFLPtools)\n",
    "setwd(\"/NAS/wg_looking/gRNA_ONT\")\n",
    "gRNA_out <- read.blast(\"stat/gencode.v37.chr_patch_hapl_scaff.transcript.fa.guideRNA.out\", sep=\"\\t\")\n",
    "\n",
    "\n",
    "for(totalID in c(\"totalRNA_K562_1\")){\n",
    "    # Genes_PCT_plot_func(\"sgRNA_K562_1\", totalID, CellLine_anno.list)\n",
    "    # trans <- Transcripts_PCT_plot_func(\"sgRNA_K562_1\", totalID, CellLine_anno.list, plot=T) \n",
    "    trans <- Transcripts_PCT_plot_func(\"sgRNA_K562_1\", totalID, CellLine_anno.list, plot=F) \n",
    "    \n",
    "    aa <- lapply(names(trans), function(TE){\n",
    "        TE_name <- strsplit(TE, \"/\")[[1]][2]\n",
    "        sg <- trans[[TE]][[\"sgRNA_K562_1\"]]\n",
    "        total <- trans[[TE]][[totalID]]\n",
    "        sg.specific <- setdiff(sg, total)\n",
    "        total.specific <- setdiff(total, sg)\n",
    "        common <- intersect(sg, total)\n",
    "        \n",
    "        sub <- gRNA_out[grepl(TE_name, gRNA_out$subject.id), ]\n",
    "        t1 <- table(sg.specific %in% sub$query.id)\n",
    "        t2 <- table(total.specific %in% sub$query.id)\n",
    "        t3 <- table(common %in% sub$query.id)\n",
    "        \n",
    "        res <- rbindlist(list(data.frame(t1, name=\"sg.specific\"),\n",
    "              data.frame(t2, name=\"total.specific\"),\n",
    "              data.frame(t3, name=\"common\")))\n",
    "        res$Var1 <- as.character(res$Var1)\n",
    "        res$Var1[res$Var1 %in% \"FALSE\"] <- \"notFound_gRNA\"\n",
    "        res$Var1[res$Var1 %in% \"TRUE\"] <- \"Found_gRNA\"\n",
    "        res <- data.frame(res, TE=TE)\n",
    "        \n",
    "        return(res)\n",
    "    })\n",
    "    \n",
    "    aa <- rbindlist(aa)\n",
    "    aa <- ddply(aa, .(TE, name), mutate, pct=Freq/sum(Freq)*100)\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    \n",
    "### the count of TE-reads related coding genes \n",
    "anno.list <- CellLine_anno.list\n",
    "c <- lapply(c(\"SINE/Alu\", \"LINE/L1\"), function(TE){\n",
    "  #\n",
    "  TE.name <- strsplit(TE, \"/\")[[1]][2]\n",
    "  x <- lapply(anno.list, function(sub){\n",
    "    tb <- sub[[TE]]\n",
    "    related_coding_gene = tb$gene_id[which(tb$gene_novelty == \"Known\" & tb$protein_coding == \"YES\")]\n",
    "    related_coding_gene <- as.character(unique(related_coding_gene))\n",
    "    return(length(related_coding_gene))\n",
    "  })\n",
    "  \n",
    "})\n",
    "names(c) <- c(\"SINE/Alu\", \"LINE/L1\")\n",
    "c <- melt(c)\n",
    "colnames(c) <- c(\"protein_coding_genes\", \"ID\", \"TE\")\n",
    "\n",
    "\n",
    "c$TE <- factor(c$TE, levels=c(\"SINE/Alu\", \"LINE/L1\"))\n",
    "ggplot(c, aes(x=ID, y=protein_coding_genes, fill=TE)) +\n",
    "    geom_bar(stat=\"identity\", position=position_dodge())+\n",
    "    theme_bw() +\n",
    "    theme(\n",
    "      panel.grid=element_blank(),\n",
    "      axis.text.x = element_text(angle = 90, color = \"black\", \n",
    "                                 size = 8, hjust = 1, vjust=0.5),\n",
    "      plot.title=element_text(size=14, face=\"bold\")\n",
    "    ) +\n",
    "    # facet_wrap(~class, nrow = 2, scales=\"free\") +\n",
    "    scale_fill_manual(values = c(\"#4678BA\", \"#FDA56C\")) +\n",
    "    ggtitle(\"Genes\")\n",
    "  \n",
    "ggsave(paste0(\"stat/CellLine_TE_reads_related_coding_genes_bar.pdf\"),\n",
    "         width = 8, height = 5)\n",
    "\n",
    "\n",
    "\n",
    "### the count of TE-reads related genes/transcripts (K562)\n",
    "anno.list <- CellLine_anno.list\n",
    "cc <- lapply(c(\"SINE/Alu\", \"LINE/L1\"), function(TE){\n",
    "    #\n",
    "    modif_tb <- function(tb){\n",
    "        sub.tb <- tb[grepl(\"L1HS\", tb$repeat.) | grepl(\"L1P\", tb$repeat.),]\n",
    "    }\n",
    "    #\n",
    "    sg.tb <- anno.list[[\"sgRNA_K562_1\"]][[TE]]\n",
    "    total.tb <- anno.list[[\"totalRNA_K562_1\"]][[TE]]\n",
    "    if(TE == \"LINE/L1\"){\n",
    "        sg.tb <- modif_tb(sg.tb)\n",
    "        total.tb <- modif_tb(total.tb)\n",
    "    }\n",
    "\n",
    "    \n",
    "    sg.tb <- sg.tb[sg.tb$gene_novelty %in% \"Known\", ]\n",
    "    total.tb <- total.tb[total.tb$gene_novelty %in% \"Known\", ]\n",
    "    \n",
    "    # genes\n",
    "    sg.coding <- unique(sg.tb$gene_id[sg.tb$gene_type %in% \"protein_coding\"])\n",
    "    sg.non_coding <- unique(sg.tb$gene_id[!(sg.tb$gene_type %in% \"protein_coding\")])\n",
    "    total.coding <- unique(total.tb$gene_id[total.tb$gene_type %in% \"protein_coding\"])\n",
    "    total.non_coding <- unique(total.tb$gene_id[!(total.tb$gene_type %in% \"protein_coding\")])\n",
    "    coding_intersect <- intersect(sg.coding, total.coding)\n",
    "    non_coding_intersect <- intersect(sg.non_coding, total.non_coding)\n",
    "    coding <- data.frame(sgRNA_K562_1=length(sg.coding),\n",
    "               totalRNA_K562_1=length(total.coding),\n",
    "               intersect = length(coding_intersect),\n",
    "               coding_type=\"protein_coding\",\n",
    "               type=\"gene\",\n",
    "               Repeat=TE\n",
    "              )\n",
    "    non_coding <- data.frame(sgRNA_K562_1=length(sg.non_coding),\n",
    "               totalRNA_K562_1=length(total.non_coding),\n",
    "               intersect = length(non_coding_intersect),\n",
    "               coding_type=\"non_coding\",\n",
    "               type=\"gene\",\n",
    "               Repeat=TE\n",
    "              )\n",
    "    c1 <- rbind(coding, non_coding)\n",
    "\n",
    "    \n",
    "    # transcripts\n",
    "    sg.coding <- unique(sg.tb$transcript_id[sg.tb$gene_type %in% \"protein_coding\"])\n",
    "    sg.non_coding <- unique(sg.tb$transcript_id[!(sg.tb$gene_type %in% \"protein_coding\")])\n",
    "    total.coding <- unique(total.tb$transcript_id[total.tb$gene_type %in% \"protein_coding\"])\n",
    "    total.non_coding <- unique(total.tb$transcript_id[!(total.tb$gene_type %in% \"protein_coding\")])\n",
    "    coding_intersect <- intersect(sg.coding, total.coding)\n",
    "    non_coding_intersect <- intersect(sg.non_coding, total.non_coding)\n",
    "    coding <- data.frame(sgRNA_K562_1=length(sg.coding),\n",
    "               totalRNA_K562_1=length(total.coding),\n",
    "               intersect = length(coding_intersect),\n",
    "               coding_type=\"protein_coding\",\n",
    "               type=\"transcript\",\n",
    "               Repeat=TE\n",
    "              )\n",
    "    non_coding <- data.frame(sgRNA_K562_1=length(sg.non_coding),\n",
    "               totalRNA_K562_1=length(total.non_coding),\n",
    "               intersect = length(non_coding_intersect),\n",
    "               coding_type=\"non_coding\",\n",
    "               type=\"transcript\",\n",
    "               Repeat=TE\n",
    "              )\n",
    "    c2 <- rbind(coding, non_coding)\n",
    "\n",
    "    \n",
    "    c <- rbind(c1, c2)\n",
    "    \n",
    "  })\n",
    "  \n",
    "\n",
    "pp <- rbindlist(cc)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### the count of TE-reads related coding transcripts \n",
    "anno.list <- CellLine_anno.list\n",
    "c <- lapply(c(\"SINE/Alu\", \"LINE/L1\"), function(TE){\n",
    "  #\n",
    "  TE.name <- strsplit(TE, \"/\")[[1]][2]\n",
    "  x <- lapply(anno.list, function(sub){\n",
    "    tb <- sub[[TE]]\n",
    "    related_coding_transcript = tb$transcript_id[which(tb$gene_novelty == \"Known\" & tb$protein_coding == \"YES\")]\n",
    "    related_coding_transcript <- as.character(unique(related_coding_transcript))\n",
    "    return(length(related_coding_transcript))\n",
    "  })\n",
    "  \n",
    "})\n",
    "names(c) <- c(\"SINE/Alu\", \"LINE/L1\")\n",
    "c <- melt(c)\n",
    "colnames(c) <- c(\"protein_coding_transcripts\", \"ID\", \"TE\")\n",
    "\n",
    "\n",
    "c$TE <- factor(c$TE, levels=c(\"SINE/Alu\", \"LINE/L1\"))\n",
    "ggplot(c, aes(x=ID, y=protein_coding_transcripts, fill=TE)) +\n",
    "    geom_bar(stat=\"identity\", position=position_dodge())+\n",
    "    theme_bw() +\n",
    "    theme(\n",
    "      panel.grid=element_blank(),\n",
    "      axis.text.x = element_text(angle = 90, color = \"black\", \n",
    "                                 size = 8, hjust = 1, vjust=0.5),\n",
    "      plot.title=element_text(size=14, face=\"bold\")\n",
    "    ) +\n",
    "    # facet_wrap(~class, nrow = 2, scales=\"free\") +\n",
    "    scale_fill_manual(values = c(\"#4678BA\", \"#FDA56C\")) +\n",
    "    ggtitle(\"Transcripts\")\n",
    "  \n",
    "ggsave(paste0(\"stat/CellLine_TE_reads_related_coding_transcripts_bar.pdf\"),\n",
    "         width = 8, height = 5)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "### the count of reads related genes\n",
    "anno.list <- totalRNA_K562_bamAnno.list\n",
    "c <- lapply(anno.list, function(tb){\n",
    "  length(unique(tb$gene_id))\n",
    "})\n",
    "c <- melt(c)\n",
    "colnames(c) <- c(\"gene_count\", \"ID\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### total reads include sgRNA\n",
    "library(RFLPtools)\n",
    "setwd(\"/NAS/wg_looking/gRNA_ONT\")\n",
    "gRNA_out <- read.table(\"/NAS/wg_looking/gRNA_ONT/stat/totalRNA_K562_1.trimming.Q7.L300.fa.guideRNA.out\", sep=\"\\t\")\n",
    "\n",
    "colnames(gRNA_out) <- c(\"query.id\", \"subject.id\", \"identity\", \"alignment.length\", \"mismatches\", \"gap.opens\", \"q.start\", \n",
    "     \"q.end\",  \"s.start\" , \"s.end\", \"evalue\", \"bit.score\")\n",
    "\n",
    "\n",
    "anno.list <- CellLine_anno.list\n",
    "cc <- lapply(c(\"SINE/Alu\", \"LINE/L1\"), function(TE){\n",
    "    TE_name <- strsplit(TE, \"/\")[[1]][2]\n",
    "    #\n",
    "    modif_tb <- function(tb){\n",
    "        sub.tb <- tb[grepl(\"L1HS\", tb$repeat.) | grepl(\"L1P\", tb$repeat.),]\n",
    "    }\n",
    "    #\n",
    "    sg.tb <- anno.list[[\"sgRNA_K562_1\"]][[TE]]\n",
    "    total.tb <- anno.list[[\"totalRNA_K562_1\"]][[TE]]\n",
    "    if(TE == \"LINE/L1\"){\n",
    "        sg.tb <- modif_tb(sg.tb)\n",
    "        total.tb <- modif_tb(total.tb)\n",
    "    }\n",
    "\n",
    "    \n",
    "    sg.tb <- sg.tb[sg.tb$gene_novelty %in% \"Known\", ]\n",
    "    total.tb <- total.tb[total.tb$gene_novelty %in% \"Known\", ]\n",
    "    sub <- gRNA_out[grepl(TE_name, gRNA_out$subject.id), ]\n",
    "    \n",
    "    # common transcripts\n",
    "    common <- intersect(sg.tb$transcript_id, total.tb$transcript_id)\n",
    "    total.read_name <- unique(total.tb[total.tb$transcript_id %in% common, \"read_name\"])\n",
    "    t1 <- table(total.read_name %in% sub$query.id)\n",
    "    \n",
    "    # total transcripts\n",
    "    total.specific <- setdiff(total.tb$transcript_id, sg.tb$transcript_id)\n",
    "    total.read_name <- unique(total.tb[total.tb$transcript_id %in% total.specific, \"read_name\"])\n",
    "    t2 <- table(total.read_name %in% sub$query.id)\n",
    "    \n",
    "    res <- rbindlist(list(data.frame(t1, name=\"common\"),\n",
    "              data.frame(t2, name=\"total.specific\")))\n",
    "    res$Var1 <- as.character(res$Var1)\n",
    "    res$Var1[res$Var1 %in% \"FALSE\"] <- \"notFound_gRNA\"\n",
    "    res$Var1[res$Var1 %in% \"TRUE\"] <- \"Found_gRNA\"\n",
    "    res <- data.frame(res, TE=TE)\n",
    "    \n",
    "  })\n",
    "  \n",
    "\n",
    "\n",
    "cc <- rbindlist(cc)\n",
    "cc <- ddply(cc, .(TE, name), mutate, pct=Freq/sum(Freq)*100)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### TE type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#\n",
    "TE_type_func <- function(ID, Test_anno){\n",
    "  library(reshape2)\n",
    "  \n",
    "  a <- lapply(Test_anno, function(tb){\n",
    "    index <- which(colnames(tb)==\"repeat.\")\n",
    "    reps <- apply(tb, 1, function(x){\n",
    "      rep <- as.character(x[index])\n",
    "      rep_tmp <- strsplit(rep, \":\")[[1]]\n",
    "      return(rep_tmp)\n",
    "    })\n",
    "    reps <- unlist(reps)\n",
    "    reps_num <- table(reps)\n",
    "    \n",
    "    return(reps_num)\n",
    "  })\n",
    "  a.tb <- melt(a)\n",
    "  colnames(a.tb) <- c(\"repeat\", \"count\", \"class\")\n",
    "  a.tb <- data.frame(a.tb, ID=ID)\n",
    "  return(a.tb)\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "load(\"stat/CellLine_anno2.RData\") # \n",
    "\n",
    "## absolute Percentage\n",
    "pp <- lapply(names(CellLine_anno.list), function(ID){TE_type_func(ID, CellLine_anno.list[[ID]])})\n",
    "pp <- rbindlist(pp)\n",
    "pp$repeat_class <- tstrsplit(pp$repeat., \",\")[[2]]\n",
    "pp$repeat_class <- substring(pp$repeat_class, 1, 4)\n",
    "# pp$repeat_class[grepl(\"L1M\", pp$repeat_class)] <- \"L1M\"\n",
    "pp$repeat_class[grepl(\"L1P\", pp$repeat_class)] <- \"L1P\"\n",
    "\n",
    "#\n",
    "p <- pp[pp$ID %in% c(\"sgRNA_K562_1\", \"totalRNA_K562_1\"),]\n",
    "p <- aggregate(count ~ ID + class + repeat_class, data = p, sum)\n",
    "p <- ddply(p, .(ID, class), mutate, Percentage = count / sum(count) * 100)\n",
    "p <- data.frame(p)\n",
    "p <- p[order(p$ID, p$Percentage, decreasing = T),]\n",
    "# p <- p[!p$Repeat %in% c(\"Alu\", \"AluYi6\", \"FRAM\", \"AluYc3\", \"AluYi6_4d\", \"FAM\", \"FLAM_A\", \"FRAM\", \"7SLRNA\", \"BC200\", \"FRAM_short\"),]\n",
    "p$CellLine <- Sample_info[as.character(p$ID), \"CellLine\"]\n",
    "# p$ID_RNAtype <- factor(p$ID_RNAtype, levels = rev(c(\"test_20210520_sgRNA\", \"test_20210527_sgRNA\", \"test_20210617_sgRNA\", \"test_20210624_sgRNA\", \"test_20210623_totalRNA\")))\n",
    "# p <- p[p$Percentage>1,] # remian: Percentage > 1%\n",
    "# Repeat <- p[p$RNAtype == \"totalRNA\", \"Repeat\"]\n",
    "p$repeat_class <- factor(p$repeat_class, levels = unique(p$repeat_class))\n",
    "\n",
    "\n",
    "for(CellLine in c(\"K562\")){\n",
    "  sub.p <- p[p$CellLine %in% CellLine, ]\n",
    "  # cols <- c(\"#F4B540\", \"#EF8354\", \"#BA748A\", \"#7D6185\", \"#87B07A\", \"#4678BA\")\n",
    "  cols <- c(\"#66C2A5\", \"#FB8C61\")\n",
    "  ggplot(sub.p, aes(x=repeat_class, y=Percentage, fill=ID)) +\n",
    "    geom_bar(stat=\"identity\", position=position_dodge())+\n",
    "    theme_bw() +\n",
    "    theme(\n",
    "      panel.grid=element_blank(),\n",
    "      axis.text.x = element_text(angle = 0, color = \"black\", \n",
    "                                 size = 8, hjust = 0.5, vjust=0.5),\n",
    "      plot.title=element_text(size=14, face=\"bold\")\n",
    "    ) +\n",
    "    # facet_wrap(~class, nrow = 1, scales=\"free_x\") +\n",
    "    facet_grid(~class, scales=\"free\", space=\"free\") +\n",
    "    scale_fill_manual(values = cols)\n",
    "  \n",
    "  ggsave(paste0(\"stat/Fig2S.the_percentage_of_Alu_and_L1_type_\", CellLine, \".pdf\"),\n",
    "         width = 7, height = 3)\n",
    "  \n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#################\n",
    "# the ratio of Alu and L1 type (sgRNA vs totalRNA)\n",
    "#################\n",
    "library(data.table)\n",
    "library(plyr)\n",
    "library(reshape2)\n",
    "library(RColorBrewer)\n",
    "setwd(\"/NAS/wg_looking/gRNA_ONT\")\n",
    "load(\"stat/test_anno.RData\")\n",
    "\n",
    "sgIDs <- c(\"test_20210520\", \"test_20210527\", \"test_20210617\", \"test_20210624\")\n",
    "cols <- setNames(brewer.pal(12,\"Paired\")[c(2,4,6,8,10)], names(Test_anno.list))\n",
    "pp <- lapply(names(Test_anno.list), function(ID){TE_type_func(ID, Test_anno.list[[ID]])})\n",
    "names(pp) <- names(Test_anno.list)\n",
    "pp <- lapply(sgIDs, function(ID){\n",
    "  test <- pp[[ID]]\n",
    "  control <- pp[[\"test_20210623\"]]\n",
    "  #\n",
    "  test <- ddply(test, .(class), mutate, Percentage = count / sum(count) * 100)\n",
    "  control <- ddply(control, .(class), mutate, Percentage = count / sum(count) * 100)\n",
    "  \n",
    "  # test <- test %>%\n",
    "  #   group_by(class) %>%\n",
    "  #   mutate(Percentage = count / sum(count)*100)\n",
    "  # control <- control %>%\n",
    "  #   group_by(class) %>%\n",
    "  #   mutate(Percentage = count / sum(count)*100)\n",
    "  \n",
    "  a <- merge(test, control, by=\"repeat.\")\n",
    "  a$Ratio <- a$Percentage.x/a$Percentage.y\n",
    "  a <- a[order(a$class.x, a$Ratio, decreasing = T),]\n",
    "  a <- a[, c(\"repeat.\", \"Percentage.x\", \"Ratio\", \"class.x\", \"ID.x\")]\n",
    "  colnames(a) <- c(\"repeat.\", \"Percentage\", \"Ratio\", \"class\", \"ID\")\n",
    "  \n",
    "  return(a)\n",
    "})\n",
    "p <- rbindlist(pp)\n",
    "# p <- p[p$Ratio>2,]\n",
    "# p <- p[p$Percentage>0.1, ]\n",
    "p$Repeat <- tstrsplit(p$repeat., \",\")[[2]]\n",
    "p <- p[!p$Repeat %in% c(\"Alu\", \"AluYi6\", \"FRAM\", \"AluYc3\", \"AluYi6_4d\", \"FAM\", \"FLAM_A\", \"FRAM\", \"7SLRNA\", \"BC200\", \"FRAM_short\"),]\n",
    "p$Repeat <- factor(p$Repeat, levels = Repeat) # From the previous part  \n",
    " \n",
    "\n",
    "ggplot(p, aes(x=Repeat, y=Ratio, fill=ID)) +\n",
    "  geom_bar(stat=\"identity\", position=position_dodge())+\n",
    "  theme_bw() +\n",
    "  theme(\n",
    "    # panel.border = element_blank(),\n",
    "    panel.grid=element_blank(),\n",
    "    axis.text.x = element_text(angle = 90, color = \"black\", \n",
    "                               size = 8, hjust = 1, vjust = 0.5),\n",
    "    # axis.ticks = element_blank(),\n",
    "    plot.title=element_text(size=14, face=\"bold\")\n",
    "  ) +\n",
    "  facet_wrap(~class, nrow = 2, scales=\"free\") +\n",
    "  geom_hline(yintercept = 1, color=\"red\", linetype=\"dashed\") +\n",
    "  scale_fill_manual(values = cols)\n",
    "  \n",
    "ggsave(\"stat/the_Ratio_of_Alu_and_L1_type.pdf\",\n",
    "         width = 18, height = 6)\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 基因组de靶向效率(L1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(biomartr)\n",
    "setwd(\"/NAS/wg_looking/gRNA_ONT\")\n",
    "\n",
    "# total reads\n",
    "pa <- system(paste0(\"grep '>' QC/DNA_20210624_BC23.trimming.Q7.L300.fa | wc -l\"), intern=T)\n",
    "pa <- as.integer(tstrsplit(pa, \" \")[[1]])\n",
    "\n",
    "\n",
    "# RepeatMasker results\n",
    "RM_file <- \"DNA_20210624_BC23/RepeatMasker_res_Q7_L300/DNA_20210624_BC23.trimming.Q7.L300.fa.out\"\n",
    "RM.tb <- read_rm(RM_file)\n",
    "RM.tb <- data.frame(RM.tb)\n",
    "RM.tb <- RM.tb[as.numeric(RM.tb$sw_score)>=225 & as.numeric(RM.tb$perc_div)<=18, ]\n",
    "dim(RM.tb) \n",
    "\n",
    "sub.RM.tb <- RM.tb[RM.tb$matching_class %in% \"LINE/L1\", ]\n",
    "\n",
    "w <- 100 \n",
    "label <- apply(sub.RM.tb, 1, function(x){\n",
    "      start <- as.integer(x[6])\n",
    "      end <- as.integer(x[7])\n",
    "      left <- x[8]\n",
    "      left <- gsub(\"\\\\)\", \"\", gsub(\"\\\\(\", \"\", left))\n",
    "      left <- as.integer(left)\n",
    "\n",
    "      if(start <= w & left <= w){\n",
    "        label0 = \"cross\"\n",
    "      }else if(start <= w & left > w){\n",
    "        label0 = \"head\"\n",
    "      }else if(start > w & left <= w){\n",
    "        label0 = \"tail\"\n",
    "      }else{\n",
    "        label0 = \"middle\"\n",
    "      }\n",
    "      return(label0)\n",
    "})\n",
    "\n",
    "sub.RM.tb$label <- label\n",
    "target_reads <- unique(sub.RM.tb[sub.RM.tb$label %in% c(\"head\", \"tail\"), \"qry_id\"])\n",
    "# target_reads <- unique(sub.RM.tb[sub.RM.tb$label %in% c(\"head\"), \"qry_id\"])\n",
    "PCT <- length(target_reads)/pa\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Saturation analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(parallel)\n",
    "library(data.table)\n",
    "library(dplyr)\n",
    "library(plyr)\n",
    "library(reshape2)\n",
    "library(RColorBrewer)\n",
    "library(ggplot2)\n",
    "library(patchwork) # p <- p1 + p3 + plot_layout(ncol = 1, heights = c(1, 1))\n",
    "options(digits = 7)         \n",
    "options(scipen = 100)   \n",
    "setwd(\"/NAS/wg_looking/gRNA_ONT\")\n",
    "load(\"stat/CellLine_anno2.RData\") # CellLine_anno.list    CellLine_bamAnno.list\n",
    "\n",
    "\n",
    "# speed up\n",
    "random_func <- function(frac, times, type){\n",
    "  library(data.table)\n",
    "  setwd(\"/NAS/wg_looking/gRNA_ONT\")\n",
    "  \n",
    "  \n",
    "  #\n",
    "  if(frac==1){\n",
    "    frac = 0.9999\n",
    "  }\n",
    "  \n",
    "  # loading length\n",
    "  sample.tb <- read.table(paste0(\"Saturation_analysis/\", type, \"_subsample_frac\", \n",
    "                                 frac, \"_times\", times, \".length.txt\"), \n",
    "                          stringsAsFactors = F, fill=T)\n",
    "  colnames(sample.tb) <- c(\"read_name\", \"Length\")\n",
    "  sample.tb$Length <- as.numeric(sample.tb$Length)\n",
    "  sample.tb <- sample.tb[complete.cases(sample.tb),]\n",
    "  \n",
    "  #\n",
    "  Sample_info <- read.table(\"stat/Sample_info2.txt\", header = T, stringsAsFactors = F)\n",
    "  rownames(Sample_info) <- Sample_info$Samples\n",
    "\n",
    "  \n",
    "  if(type==\"totalRNA\"){\n",
    "      total.list <- CellLine_anno.list[c(\"totalRNA_K562_1\", \"totalRNA_K562_2\", \"totalRNA_K562_3\")]\n",
    "      tb <- rbindlist(lapply(total.list, rbindlist))\n",
    "     }\n",
    "  if(type==\"sgRNA\"){\n",
    "      tb <- rbindlist(CellLine_anno.list[[\"sgRNA_K562_1\"]])  \n",
    "  }\n",
    "  sub.tb <- tb[as.character(tb$read_name) %in% sample.tb$read_name, ] \n",
    "  sub.total_genes <- as.character(unique(sub.tb$gene_id[which(sub.tb$gene_novelty == \"Known\")]))\n",
    "  sub.coding_genes <- as.character(unique(sub.tb$gene_id[which(sub.tb$gene_novelty == \"Known\" &\n",
    "                                                                   sub.tb$protein_coding == \"YES\")]))\n",
    "  sub.non_coding_genes <- as.character(unique(sub.tb$gene_id[which(sub.tb$gene_novelty == \"Known\" &\n",
    "                                                                   is.na(sub.tb$protein_coding))]))\n",
    "  sub.total_transcripts <- as.character(unique(sub.tb$transcript_id[which(sub.tb$gene_novelty == \"Known\")]))\n",
    "\n",
    "\n",
    "  stat_num <- data.frame(class=c(\"total_genes\", \"coding_genes\", \"non_coding_genes\", \"total_transcripts\"),\n",
    "                           num=c(length(sub.total_genes),length(sub.coding_genes),\n",
    "                                 length(sub.non_coding_genes),length(sub.total_transcripts)))\n",
    "  c <- list(stat=data.frame(seqBases=sum(sample.tb$Length), \n",
    "                            fraction=frac,\n",
    "                            random_times=times, \n",
    "                            stat_num,\n",
    "                            type=type),\n",
    "            total_transcripts=sub.total_transcripts)\n",
    "  \n",
    "  \n",
    "  return(c)\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## totalRNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "library(dplyr)\n",
    "library(plyr)\n",
    "library(ggpmisc)\n",
    "load(\"stat/CellLine_anno2.RData\") # CellLine_anno.list    CellLine_bamAnno.list\n",
    "\n",
    "\n",
    "type = \"totalRNA\"\n",
    "step <- seq(5,100,5)\n",
    "random.list <- lapply(step, function(frac){\n",
    "  a <- lapply(1:50, function(times){\n",
    "    random_func(frac/100, times, type)\n",
    "  })\n",
    "})\n",
    "\n",
    "save(random.list, file=\"stat/Saturation_analysis_totalRNA_random2.RData\")\n",
    "\n",
    "\n",
    "#####\n",
    "load(\"stat/Saturation_analysis_totalRNA_random2.RData\")\n",
    "numDf.list <- lapply(random.list, function(x){lapply(x, function(y){y[[1]]})})\n",
    "pp <- rbindlist(lapply(numDf.list, rbindlist))\n",
    "pp$fraction <- factor(pp$fraction)\n",
    "# p <- pp %>%\n",
    "#   group_by(fraction) %>%\n",
    "#   mutate(Gigabase = mean(seqBases)/10^9)\n",
    "# p <- ddply(pp, .(fraction), mutate, Gigabase = mean(seqBases)/10^9)\n",
    "p <- pp\n",
    "p$Gigabase <- p$seqBases/10^9\n",
    "table(p$Gigabase)\n",
    "p$Gigabase <- round(p$Gigabase,2)\n",
    "# p$Gigabase <- factor(p$Gigabase)\n",
    "p <- p[p$class %in% c(\"total_genes\", \"total_transcripts\"),]\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "### regression\n",
    "# for genes\n",
    "sub.p <- p[p$class %in% \"total_genes\", ]\n",
    "lm.test <- lm(num~log(Gigabase), sub.p)\n",
    "summary(lm.test) \n",
    "\n",
    "y <- 2917.529*log(x) + 3473.080\n",
    "y <- 10000\n",
    "x <- exp((y- 3473.080)/2917.529)\n",
    "\n",
    "\n",
    "# for transcripts\n",
    "sub.p <- p[p$class %in% \"total_transcripts\", ]\n",
    "lm.test <- lm(num~log(Gigabase), sub.p)\n",
    "summary(lm.test) \n",
    "\n",
    "y <- 5908.07*log(x) + 2341.46\n",
    "y <- 20000\n",
    "x <- exp((y-2341.46)/5908.07)\n",
    "\n",
    "\n",
    "\n",
    "# plot\n",
    "ggplot(p, aes(x=Gigabase, y=num, color=class)) +\n",
    "    # geom_boxplot(position = position_dodge(width = 0)) +\n",
    "    geom_point() +\n",
    "    geom_smooth(method = \"lm\", se=F, formula = y ~ log(x)) +\n",
    "    stat_poly_eq(formula = y ~ log(x), \n",
    "               aes(label = paste(..eq.label..))) +   \n",
    "    # facet_wrap(~class, nrow = 2, scales=\"free\") +\n",
    "    # geom_hline(yintercept = c(0.75), color=\"red\", linetype=\"dashed\") +\n",
    "    # scale_fill_manual(values = c(\"#4678BA\", \"#87B07A\", \"#EF8354\", \"#F4B540\")) +\n",
    "    scale_color_manual(values = c(\"#EE9248\", \"#4A89D5\"))+\n",
    "    theme_classic()+\n",
    "    theme(\n",
    "        panel.grid=element_blank(),\n",
    "        axis.text.x = element_text(angle = 0, color = \"black\", \n",
    "                                   size = 8, hjust = 0.5, vjust = 0.5),\n",
    "        legend.position=\"bottom\",\n",
    "        plot.title=element_text(size=14, face=\"bold\")\n",
    "    ) + \n",
    "    geom_hline(yintercept = c(10000, 20000), color=\"red\", linetype=\"dashed\") +\n",
    "    geom_vline(xintercept = c( exp((10000- 3473.080)/2917.529), exp((20000-2341.46)/5908.07) ), \n",
    "               color=\"red\", linetype=\"dashed\") +\n",
    "    ggtitle(\"totalRNA\")\n",
    "\n",
    "\n",
    "ggsave(\"stat/Saturation_analysis_totalRNA_3.pdf\",\n",
    "       width = 6, height = 4)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## sgRNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(dplyr)\n",
    "library(plyr)\n",
    "library(ggpmisc)\n",
    "load(\"stat/CellLine_anno2.RData\") # CellLine_anno.list    CellLine_bamAnno.list\n",
    "\n",
    "\n",
    "type = \"sgRNA\"\n",
    "step <- seq(5,100,5)\n",
    "random.list <- lapply(step, function(frac){\n",
    "  a <- lapply(1:50, function(times){\n",
    "    random_func(frac/100, times, type)\n",
    "  })\n",
    "})\n",
    "\n",
    "save(random.list, file=\"stat/Saturation_analysis_sgRNA_random2.RData\")\n",
    "\n",
    "##\n",
    "# random.list <- list()\n",
    "# for(frac in 1:10){\n",
    "#   random.list[[frac]] <- list()\n",
    "#   for(times in 1:100){\n",
    "#     random.list[[frac]][[times]] <- random_func(frac/10, times, type)\n",
    "#   }\n",
    "# }\n",
    "\n",
    "load(\"stat/Saturation_analysis_sgRNA_random2.RData\")\n",
    "numDf.list <- lapply(random.list, function(x){lapply(x, function(y){y[[1]]})})\n",
    "pp <- rbindlist(lapply(numDf.list, rbindlist))\n",
    "pp$fraction <- factor(pp$fraction)\n",
    "# p <- pp %>%\n",
    "#   group_by(fraction) %>%\n",
    "#   mutate(Gigabase = mean(seqBases)/10^9)\n",
    "# p <- ddply(pp, .(fraction), mutate, Gigabase = mean(seqBases)/10^9)\n",
    "p <- pp\n",
    "p$Gigabase <- p$seqBases/10^9\n",
    "table(p$Gigabase)\n",
    "p$Gigabase <- round(p$Gigabase,2)\n",
    "# p$Gigabase <- factor(p$Gigabase)\n",
    "\n",
    "\n",
    "\n",
    "p <- p[p$class %in% c(\"total_genes\", \"total_transcripts\"),]\n",
    "### regression\n",
    "# for genes\n",
    "sub.p <- p[p$class %in% \"total_genes\", ]\n",
    "lm.test <- lm(num~log(Gigabase), sub.p)\n",
    "summary(lm.test) \n",
    "\n",
    "y <- 3128.599*log(x) + 9381.414\n",
    "y <- 10000\n",
    "x <- exp((y- 9381.414)/3128.599)\n",
    "\n",
    "# for transcripts\n",
    "sub.p <- p[p$class %in% \"total_transcripts\", ]\n",
    "lm.test <- lm(num~log(Gigabase), sub.p)\n",
    "summary(lm.test) \n",
    "\n",
    "y <- 7415.11*log(x) + 15033.22\n",
    "y <- 20000\n",
    "x <- exp((y-15033.22)/7415.11)\n",
    "\n",
    "\n",
    "\n",
    "# plot\n",
    "ggplot(p, aes(x=Gigabase, y=num, color=class)) +\n",
    "    # geom_boxplot(position = position_dodge(width = 0)) +\n",
    "    geom_point() +\n",
    "    geom_smooth(method = \"lm\", se=F, formula = y ~ log(x)) +\n",
    "    stat_poly_eq(formula = y ~ log(x), \n",
    "               aes(label = paste(..eq.label..))) +   \n",
    "    # facet_wrap(~class, nrow = 2, scales=\"free\") +\n",
    "    # geom_hline(yintercept = c(0.75), color=\"red\", linetype=\"dashed\") +\n",
    "    # scale_fill_manual(values = c(\"#4678BA\", \"#87B07A\", \"#EF8354\", \"#F4B540\")) +\n",
    "    scale_color_manual(values = c(\"#EE9248\", \"#4A89D5\"))+\n",
    "    theme_classic() +\n",
    "    theme(\n",
    "        panel.grid=element_blank(),\n",
    "        axis.text.x = element_text(angle = 0, color = \"black\", \n",
    "                                   size = 8, hjust = 0.5, vjust = 0.5),\n",
    "        legend.position=\"bottom\",\n",
    "        plot.title=element_text(size=14, face=\"bold\")\n",
    "    ) +  \n",
    "    geom_hline(yintercept = c(10000, 20000), color=\"red\", linetype=\"dashed\") +\n",
    "    geom_vline(xintercept = c( exp((10000- 9381.414)/3128.599), exp((20000-15033.22)/7415.11)), \n",
    "               color=\"red\", linetype=\"dashed\") +\n",
    "    ggtitle(\"sgRNA\")\n",
    "\n",
    "\n",
    "\n",
    "ggsave(\"stat/Saturation_analysis_sgRNA_3.pdf\",\n",
    "       width = 6, height = 4)\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 测序量totalRNA/sgRNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pp <- lapply(seq(5000, 30000, 1000), function(y){\n",
    "    g.sg <- exp((y- 9381.414)/3128.599)\n",
    "    t.sg <- exp((y-15033.22)/7415.11) \n",
    "\n",
    "    g.total <- exp((y-3473.080)/2917.529)\n",
    "    t.total <- exp((y-2341.46)/5908.07)\n",
    "    \n",
    "    g.ratio <- g.total/g.sg\n",
    "    t.ratio <- t.total/t.sg\n",
    "    \n",
    "    a <- rbind(data.frame(type=\"genes\", ratio=g.ratio, number=y),\n",
    "               data.frame(type=\"transcripts\", ratio=t.ratio, number=y))\n",
    "    return(a)\n",
    "})\n",
    "\n",
    "pp <- rbindlist(pp)\n",
    "ggplot(pp, aes(x=number, y=ratio, color=type)) +\n",
    "    geom_point() +\n",
    "    geom_line() +\n",
    "\n",
    "    # facet_wrap(~class, nrow = 2, scales=\"free\") +\n",
    "    # geom_hline(yintercept = c(0.75), color=\"red\", linetype=\"dashed\") +\n",
    "    # scale_fill_manual(values = c(\"#4678BA\", \"#87B07A\", \"#EF8354\", \"#F4B540\")) +\n",
    "    scale_color_manual(values = c(\"#EE9248\", \"#4A89D5\"))+\n",
    "    theme_classic() +\n",
    "    theme(\n",
    "        panel.grid=element_blank(),\n",
    "        axis.text.x = element_text(angle = 0, color = \"black\", \n",
    "                                   size = 8, hjust = 0.5, vjust = 0.5),\n",
    "        legend.position=\"bottom\",\n",
    "        plot.title=element_text(size=14, face=\"bold\")\n",
    "    ) + \n",
    "    ggtitle(\"\")\n",
    "\n",
    "\n",
    "\n",
    "ggsave(\"stat/Fig2B.totalRNA_vs_sgRNA_relative_data.pdf\",\n",
    "       width = 6, height = 4)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": [],
    "toc-hr-collapsed": true
   },
   "source": [
    "# Gene or transcript quantification"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(DESeq2)\n",
    "library(parallel)\n",
    "library(RColorBrewer)\n",
    "library(plyr)\n",
    "library(pheatmap)\n",
    "library(ggplot2)\n",
    "library(DESeq2)\n",
    "library(patchwork) # p <- p1 + p3 + plot_layout(ncol = 1, heights = c(1, 1))\n",
    "library(rtracklayer)\n",
    "library(BiocParallel)\n",
    "library(dplyr)\n",
    "setwd(\"/NAS/wg_looking/gRNA_ONT\")\n",
    "\n",
    "\n",
    "# loading gtf file\n",
    "gtfFile = \"/NAS/wg_looking/human_ref/gencode.v37.chr_patch_hapl_scaff.annotation.gtf\"\n",
    "gtf <- rtracklayer::import(gtfFile)\n",
    "gtf <- as.data.frame(gtf)\n",
    "gene_gtf <- gtf[gtf$type==\"gene\",]\n",
    "rownames(gene_gtf) <- gene_gtf$gene_id\n",
    "transcript_gtf <- gtf[gtf$type==\"transcript\",]\n",
    "rownames(transcript_gtf) <- transcript_gtf$transcript_id\n",
    "\n",
    "save(gene_gtf, transcript_gtf, file=\"stat/hg38.gene.transcript.gtf.RData\")\n",
    "\n",
    "# load Sample info\n",
    "Sample_info <- read.table(\"stat/Sample_info2.txt\", header = T, stringsAsFactors = F)\n",
    "rownames(Sample_info) <- Sample_info$Samples\n",
    "SampleIDs <- Sample_info$Samples[Sample_info$CellLine %in% c(\"HCT116\", \"NCM460\")]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "## map ENSEMBL to ENTREZID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "setwd(\"/NAS/wg_looking/gRNA_ONT\")\n",
    "\n",
    "file <- paste0(\"database/GeneID_ENTREZID_mart_export.txt\")\n",
    "mart <- read.table(file, header=T, fill=T)\n",
    "mart <- mart[, c(2,4,5)]\n",
    "colnames(mart) <- c(\"gene_id\", \"transcript_id\", \"ENTREZID\")\n",
    "\n",
    "# for gene merge\n",
    "mart.tmp <- mart[, c(\"gene_id\", \"ENTREZID\")]\n",
    "mart.tmp <- mart.tmp[order(mart.tmp$gene_id),]\n",
    "mart.tmp <- mart.tmp[!duplicated(mart.tmp$gene_id), ]\n",
    "write.table(mart.tmp, \"database/ENSEMBL_ID_Entrez_ID_all.txt\", row.names=F, col.names=T, quote=F, sep=\"\\t\")\n",
    "\n",
    "\n",
    "# merge.tb <- merge(gene_res, mart.tmp, by=\"gene_id\", all.x=T)\n",
    "# dim(merge.tb)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## salmon"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Correlation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# load Sample info\n",
    "Sample_info <- read.table(\"stat/Sample_info2.txt\", header = T, stringsAsFactors = F)\n",
    "rownames(Sample_info) <- Sample_info$Samples\n",
    "SampleIDs <- Sample_info$Samples\n",
    "\n",
    "\n",
    "\n",
    "# transcripts count\n",
    "quant.list <- lapply(SampleIDs, function(ID){\n",
    "  sf <- paste0(ID, \"/\", ID, \".salmon_quant/quant.sf\")\n",
    "  quant <- read.table(sf, header = T)\n",
    "  quant <- quant[, c(\"Name\", \"NumReads\")]\n",
    "  colnames(quant) <- c(\"transcript_id\", ID)\n",
    "  return(quant)\n",
    "})\n",
    "\n",
    "\n",
    "quant <- join_all(quant.list)\n",
    "rownames(quant) <- quant$transcript_id\n",
    "quant <- quant[,-1]\n",
    "\n",
    "# sgRNA count > 1 at least one sample\n",
    "logic <- apply(quant, 1, function(x){any(x[Sample_info[SampleIDs, \"SeqType\"] %in% \"sgRNA\"] >1)})\n",
    "sub.quant <- quant[logic,]\n",
    "cor(sub.quant)\n",
    "\n",
    "\n",
    "pheatmap(cor(sub.quant, method = \"pearson\"),\n",
    "         display_numbers=T,\n",
    "         main=\"Count correlation\",\n",
    "         # col=colorRampPalette(rev(brewer.pal(11,\"RdYlBu\")[3:10]))(21),\n",
    "         col=colorRampPalette(c(\"#4678BA\", \"#F7F7F7\", \"#EF8354\"))(100),\n",
    "         border=NA,\n",
    "         filename=\"stat/CRC_count_correlation.pdf\")\n",
    "\n",
    "\n",
    "# sgRNA count > 10 at all sgRNA samples\n",
    "logic <- apply(quant, 1, function(x){all(x[Sample_info[SampleIDs, \"SeqType\"] %in% \"sgRNA\"] >10)})\n",
    "sub.quant <- quant[logic,]\n",
    "colMeans(sub.quant)\n",
    "apply(sub.quant, 2, median)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#########  \n",
    "#transcripts TPM\n",
    "quant.list <- lapply(SampleIDs, function(ID){\n",
    "  sf <- paste0(ID, \"/\", ID, \".salmon_quant/quant.sf\")\n",
    "  quant <- read.table(sf, header = T)\n",
    "  quant <- quant[, c(\"Name\", \"TPM\")]\n",
    "  colnames(quant) <- c(\"transcript_id\", ID)\n",
    "  return(quant)\n",
    "})\n",
    "\n",
    "\n",
    "quant <- join_all(quant.list)\n",
    "rownames(quant) <- quant$transcript_id\n",
    "quant <- quant[,-1]\n",
    "\n",
    "# sgRNA TPM > 1 at least one sample\n",
    "logic <- apply(quant, 1, function(x){all(x[Sample_info[SampleIDs, \"SeqType\"] %in% \"sgRNA\"] >1)})\n",
    "logic <- apply(quant, 1, function(x){any(x[Sample_info[SampleIDs, \"SeqType\"] %in% \"sgRNA\"] >1)})\n",
    "sub.quant <- quant[logic,]\n",
    "cor(sub.quant)\n",
    "\n",
    "\n",
    "pheatmap(cor(sub.quant, method = \"spearman\"),\n",
    "         display_numbers=T,\n",
    "         main=\"Transcripts: TPM correlation\",\n",
    "         # col=colorRampPalette(rev(brewer.pal(11,\"RdYlBu\")[3:10]))(21),\n",
    "         col=colorRampPalette(c(\"#4678BA\", \"#F7F7F7\", \"#EF8354\"))(100),\n",
    "         border=NA,\n",
    "         filename=\"stat/CellLine_TPM_correlation_spearman.pdf\")\n",
    "\n",
    "\n",
    "pheatmap(cor(sub.quant, method = \"pearson\"),\n",
    "         display_numbers=T,\n",
    "         main=\"Transcripts: TPM correlation\",\n",
    "         # col=colorRampPalette(rev(brewer.pal(11,\"RdYlBu\")[3:10]))(21),\n",
    "         col=colorRampPalette(c(\"#4678BA\", \"#F7F7F7\", \"#EF8354\"))(100),\n",
    "         border=NA,\n",
    "         filename=\"stat/CellLine_TPM_correlation_pearson.pdf\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": [],
    "toc-hr-collapsed": true
   },
   "source": [
    "### HCT116 vs NCM460"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Differential Transcripts "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(DESeq2)\n",
    "library(tximport)\n",
    "\n",
    "SampleIDs <- Sample_info$Samples[(Sample_info$CellLine %in% c(\"HCT116\", \"NCM460\")) & (Sample_info$SeqType %in% \"sgRNA\")]\n",
    "\n",
    "\n",
    "files <- file.path(SampleIDs, paste0(SampleIDs,\".salmon_quant\"), \"quant.sf\")\n",
    "names(files) <- SampleIDs\n",
    "txi <- tximport(files, type = \"salmon\", txOut = TRUE)\n",
    "names(txi)\n",
    "dds <- DESeqDataSetFromTximport(txi, \n",
    "                                Sample_info[SampleIDs, c(\"CellLine\"), drop=F], \n",
    "                                ~CellLine)\n",
    "Object <- DESeq(dds\n",
    "                # parallel = T, BPPARAM=MulticoreParam(10)\n",
    "                )\n",
    "norlCount <- counts(Object, normalized=T)\n",
    "res <- results(Object, contrast=c(\"CellLine\", \"HCT116\", \"NCM460\"))\n",
    "res <- res[complete.cases(res),]\n",
    "dif_res <- res\n",
    "dif_res <- res[res$padj<0.05 & abs(res$log2FoldChange)>2 & res$baseMean>10, ]\n",
    "anno_colors <- list(CellLine=c(NCM460=\"#4678BA\", HCT116=\"#F4B540\"))\n",
    "colInfo <- Sample_info[SampleIDs, c(\"CellLine\"), drop=F]\n",
    "pheatmap(norlCount[rownames(dif_res),],\n",
    "         # display_numbers=T,\n",
    "         scale=\"row\",\n",
    "         annotation_col = colInfo,\n",
    "         annotation_colors = anno_colors,\n",
    "         main=\"differential transcript expression\",\n",
    "         col=colorRampPalette(rev(brewer.pal(11,\"RdBu\")))(21),\n",
    "         # col=colorRampPalette(c(\"#4678BA\", \"#F7F7F7\", \"#EF8354\"))(11),\n",
    "         border=NA,\n",
    "         show_rownames=F,\n",
    "         show_colnames=F,\n",
    "         width=6.5,\n",
    "         height=6,\n",
    "         filename=\"stat/CRC_trans_dif_heatmap.pdf\")\n",
    "\n",
    "out_res <- dif_res[, c(\"baseMean\", \"log2FoldChange\", \"padj\")]\n",
    "out_res$dif <- \"HCT116_over\"\n",
    "out_res$dif[out_res$log2FoldChange<0] <- \"NCM460_over\"\n",
    "out_res <- out_res[order(out_res$dif),]\n",
    "out_res <- data.frame(gene_id=rownames(out_res), out_res)\n",
    "out_res <- data.frame(out_res, gene_name=transcript_gtf[rownames(out_res), \"gene_name\"])\n",
    "transcript_res <- out_res\n",
    "# write.table(out_res, \"stat/CRC_differential_transcripts.txt\", row.names = F, col.names = T, quote = F, sep=\"\\t\")\n",
    "write.table(out_res, \"stat/CRC_all_transcripts.txt\", row.names = F, col.names = T, quote = F, sep=\"\\t\")\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Differential Genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "library(DESeq2)\n",
    "library(tximport)\n",
    "\n",
    "\n",
    "\n",
    "# load Sample info\n",
    "Sample_info <- read.table(\"stat/Sample_info2.txt\", header = T, stringsAsFactors = F)\n",
    "rownames(Sample_info) <- Sample_info$Samples\n",
    "SampleIDs <- Sample_info$Samples[(Sample_info$CellLine %in% c(\"HCT116\", \"NCM460\")) & (Sample_info$SeqType %in% \"sgRNA\")]\n",
    "\n",
    "#\n",
    "files <- file.path(SampleIDs, paste0(SampleIDs,\".salmon_quant\"), \"quant.genes.sf\")\n",
    "names(files) <- SampleIDs\n",
    "txi <- tximport(files, type = \"salmon\", txOut = TRUE)\n",
    "names(txi)\n",
    "dds <- DESeqDataSetFromTximport(txi, \n",
    "                                Sample_info[SampleIDs, c(\"CellLine\"), drop=F], \n",
    "                                ~CellLine)\n",
    "Object <- DESeq(dds\n",
    "                # parallel = T, BPPARAM=MulticoreParam(10)\n",
    ")\n",
    "norlCount <- counts(Object, normalized=T)\n",
    "res <- results(Object, contrast=c(\"CellLine\", \"HCT116\", \"NCM460\"))\n",
    "res <- res[complete.cases(res),]\n",
    "gene_res <- res\n",
    "dif_res <- res[res$padj<0.05 & abs(res$log2FoldChange)>1 & res$baseMean>10, ]\n",
    "\n",
    "\n",
    "## for write out\n",
    "out_res <- dif_res[, c(\"baseMean\", \"log2FoldChange\", \"padj\")]\n",
    "out_res$dif <- \"HCT116_over\"\n",
    "out_res$dif[out_res$log2FoldChange<0] <- \"NCM460_over\"\n",
    "out_res <- out_res[order(out_res$dif),]\n",
    "out_res <- data.frame(gene_id=rownames(out_res), out_res)\n",
    "out_res <- data.frame(out_res, gene_name=gene_gtf[rownames(out_res), \"gene_name\"])\n",
    "gene_res <- out_res\n",
    "# write.table(gene_res, \"stat/CRC_differential_genes.txt\", row.names = F, col.names = T, quote = F, sep=\"\\t\")\n",
    "write.table(gene_res, \"stat/CRC_all_genes.txt\", row.names = F, col.names = T, quote = F, sep=\"\\t\")\n",
    "\n",
    "\n",
    "# plotting \n",
    "anno_colors <- list(CellLine=c(NCM460=\"#4678BA\", HCT116=\"#F4B540\"))\n",
    "colInfo <- Sample_info[SampleIDs, c(\"CellLine\"), drop=F]\n",
    "pheatmap(norlCount[rownames(dif_res),],\n",
    "         # display_numbers=T,\n",
    "         scale=\"row\",\n",
    "         annotation_col = colInfo,\n",
    "         annotation_colors = anno_colors,\n",
    "         main=\"differential gene expression\",\n",
    "         col=colorRampPalette(rev(brewer.pal(11,\"RdBu\")))(21),\n",
    "         # col=colorRampPalette(c(\"#4678BA\", \"#F7F7F7\", \"#EF8354\"))(11),\n",
    "         border=NA,\n",
    "         show_rownames=F,\n",
    "         show_colnames=F,\n",
    "         width=6.5,\n",
    "         height=6,\n",
    "         filename=\"stat/HCT116_vs_NCM460_genes_dif_heatmap.pdf\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Enrichment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(rtracklayer)\n",
    "library(reshape2)\n",
    "library(ggplot2)\n",
    "library(bedtoolsr)\n",
    "library(biomartr)\n",
    "library(data.table)\n",
    "setwd(\"/NAS/wg_looking/gRNA_ONT\")\n",
    " \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(gprofiler2)\n",
    "set_base_url(\"https://biit.cs.ut.ee/gprofiler_archive3/e102_eg49_p15\")\n",
    "\n",
    "# for gene_res\n",
    "gene_res <- res\n",
    "totalGenes <- rownames(res)\n",
    "totalGenes <- tstrsplit(totalGenes, \"[.]\")[[1]]\n",
    " \n",
    "diff_genes <- lapply(setNames(c(\"hyper\", \"hypo\"), c(\"hyper\", \"hypo\")), function(x){\n",
    "    d_res <- res[res$padj<0.05 & res$baseMean>10, ]   \n",
    "    if(x==\"hyper\"){\n",
    "        d <- d_res[d_res$log2FoldChange>1, ]   \n",
    "    }\n",
    "     if(x==\"hypo\"){\n",
    "        d <- d_res[d_res$log2FoldChange< -1, ]\n",
    "    }\n",
    "    \n",
    "    g <- tstrsplit(rownames(d), \"[.]\")[[1]]\n",
    "    \n",
    "    return(g)\n",
    "})\n",
    "\n",
    "\n",
    "gostres.list <- lapply(diff_genes, function(g){\n",
    "    gostres <- gost(query=g,\n",
    "                    custom_bg=totalGenes,\n",
    "                    organism = \"hsapiens\",\n",
    "                    user_threshold = 0.01,\n",
    "                    correction_method=\"gSCS\")\n",
    "    gostres <- data.frame(gostres$result)\n",
    "    return(gostres)\n",
    "})\n",
    "\n",
    "\n",
    "r <- gostres.list[[\"hyper\"]]\n",
    "r[r$source %in% c(\"GO:BP\", \"KEGG\"), c(\"p_value\", \"term_name\", \"source\")]\n",
    "\n",
    "\n",
    "r2 <- gostres.list[[\"hypo\"]]\n",
    "r2[r2$source %in% c(\"GO:BP\", \"GO:MF\"), c(\"term_name\", \"source\")]\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### term plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "### term plot\n",
    "plot_gostres_func <- function(class, gostres, height = 7, width = 7){\n",
    "    r <- gostres\n",
    "    pp <- r[(r$source %in% c(\"GO:BP\", \"KEGG\")) , c(\"p_value\", \"term_name\", \"source\")]\n",
    "    order = pp$term_name[order(pp$p_value, decreasing=T)]\n",
    "    pp$term_name <- factor(pp$term_name, levels = order)\n",
    "\n",
    "    ggplot(pp, aes(x=term_name, y=-log10(p_value), fill=source)) +\n",
    "      geom_bar(stat=\"identity\", width = 0.8) +\n",
    "      theme_bw() +\n",
    "      theme(panel.grid = element_blank(),\n",
    "            plot.title = element_text(hjust = 0, size = 12),\n",
    "            axis.text.x  = element_text(angle = 0, hjust=1, vjust=0.5, \n",
    "                                        colour = \"black\", size = 11),\n",
    "            axis.text.y  = element_text(colour = \"black\"),\n",
    "            # legend.position = c(0.8,0.8),\n",
    "            strip.background = element_rect(fill=\"white\", colour = \"black\", size=1) ,\n",
    "            legend.title = element_text(face=\"bold\")) +\n",
    "      # geom_text(aes(label=symbol), vjust=0.5, hjust=0,size=4, colour=\"black\") +\n",
    "      # annotate(\"text\", x=pp$NAME[2], y=-1.5,size=3,\n",
    "      #          label=\"nominal P values: \\n*P < 0.05 \\n**P < 0.005 \\n***P < 0.0005 \\nNS: not significant\") +\n",
    "      # scale_x_discrete(limits=c(pp$term.name))+\n",
    "      scale_fill_manual(values=c(\"GO:BP\"=\"#4678BA\", \"KEGG\"=\"#EF8354\")) +\n",
    "      coord_flip() +\n",
    "      xlab(\"\") +\n",
    "      ggtitle(paste0(\"HCT116_VS_NCM460: \", class))\n",
    "    # facet_grid(.~module, scales = \"free_x\", space = \"free_x\") \n",
    "\n",
    "\n",
    "    ggsave(paste0(\"stat/\", \"HCT116_VS_NCM460_\", class, \"_function_bar.pdf\"),\n",
    "           height = height, width = width)\n",
    "\n",
    "}\n",
    "\n",
    "\n",
    "plot_gostres_func(\"hyper\", gostres.list[[\"hyper\"]], height = 7, width = 6)\n",
    "plot_gostres_func(\"hypo\", gostres.list[[\"hypo\"]], height = 2, width = 6)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### MDAMB231 vs MCF10A"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Differential Transcripts "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(DESeq2)\n",
    "library(tximport)\n",
    "c # default\n",
    "\n",
    "Sample_info <- read.table(\"stat/Sample_info2.txt\", header = T, stringsAsFactors = F)\n",
    "rownames(Sample_info) <- Sample_info$Samples\n",
    "SampleIDs <- Sample_info$Samples[(Sample_info$CellLine %in% c(\"MDAMB231\", \"MCF10A\")) & (Sample_info$SeqType %in% \"sgRNA\")]\n",
    "\n",
    "\n",
    "files <- file.path(SampleIDs, paste0(SampleIDs,\".salmon_quant\"), \"quant.sf\")\n",
    "names(files) <- SampleIDs\n",
    "txi <- tximport(files, type = \"salmon\", txOut = TRUE)\n",
    "names(txi)\n",
    "dds <- DESeqDataSetFromTximport(txi, \n",
    "                                Sample_info[SampleIDs, c(\"CellLine\"), drop=F], \n",
    "                                ~CellLine)\n",
    "Object <- DESeq(dds\n",
    "                # parallel = T, BPPARAM=MulticoreParam(10)\n",
    "                )\n",
    "norlCount <- counts(Object, normalized=T)\n",
    "res <- results(Object, contrast=c(\"CellLine\", \"MDAMB231\", \"MCF10A\"))\n",
    "res <- res[complete.cases(res),]\n",
    "dif_res <- res\n",
    "# dif_res <- res[res$padj<0.05 & abs(res$log2FoldChange)>2 & res$baseMean>10, ]\n",
    "\n",
    "\n",
    "out_res <- dif_res[, c(\"baseMean\", \"log2FoldChange\", \"padj\")]\n",
    "out_res$dif <- \"MDAMB231_hyper\"\n",
    "out_res$dif[out_res$log2FoldChange<0] <- \"MCF10A_hyper\"\n",
    "out_res <- out_res[order(out_res$dif),]\n",
    "out_res <- data.frame(gene_id=rownames(out_res), out_res)\n",
    "out_res <- data.frame(out_res, gene_name=transcript_gtf[rownames(out_res), \"gene_name\"])\n",
    "transcript_res <- out_res\n",
    "out_res <- transcript_res[transcript_res$padj<0.05 & abs(transcript_res$log2FoldChange)>2 & transcript_res$baseMean>10, ]\n",
    "write.table(out_res, \"stat/MDAMB231_vs_MCF10A_differential_transcripts.txt\", row.names = F, col.names = T, quote = F, sep=\"\\t\")\n",
    "# write.table(out_res, \"stat/MDAMB231_vs_MCF10A_all_transcripts.txt\", row.names = F, col.names = T, quote = F, sep=\"\\t\")\n",
    "\n",
    "\n",
    "\n",
    "anno_colors <- list(CellLine=c(NCM460=\"#4678BA\", HCT116=\"#F4B540\"))\n",
    "colInfo <- Sample_info[SampleIDs, c(\"CellLine\"), drop=F]\n",
    "pheatmap(norlCount[rownames(dif_res),],\n",
    "         # display_numbers=T,\n",
    "         scale=\"row\",\n",
    "         annotation_col = colInfo,\n",
    "         annotation_colors = anno_colors,\n",
    "         main=\"differential transcript expression\",\n",
    "         col=colorRampPalette(rev(brewer.pal(11,\"RdBu\")))(21),\n",
    "         # col=colorRampPalette(c(\"#4678BA\", \"#F7F7F7\", \"#EF8354\"))(11),\n",
    "         border=NA,\n",
    "         show_rownames=F,\n",
    "         show_colnames=F,\n",
    "         width=6.5,\n",
    "         height=6,\n",
    "         filename=\"stat/CRC_trans_dif_heatmap.pdf\")\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### Differential Genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(DESeq2)\n",
    "library(tximport)\n",
    "\n",
    "\n",
    "\n",
    "# load Sample info\n",
    "Sample_info <- read.table(\"stat/Sample_info2.txt\", header = T, stringsAsFactors = F)\n",
    "rownames(Sample_info) <- Sample_info$Samples\n",
    "SampleIDs <- Sample_info$Samples[(Sample_info$CellLine %in% c(\"MDAMB231\", \"MCF10A\")) & (Sample_info$SeqType %in% \"sgRNA\")]\n",
    "\n",
    "#\n",
    "files <- file.path(SampleIDs, paste0(SampleIDs,\".salmon_quant\"), \"quant.genes.sf\")\n",
    "names(files) <- SampleIDs\n",
    "txi <- tximport(files, type = \"salmon\", txOut = TRUE)\n",
    "names(txi)\n",
    "dds <- DESeqDataSetFromTximport(txi, \n",
    "                                Sample_info[SampleIDs, c(\"CellLine\"), drop=F], \n",
    "                                ~CellLine)\n",
    "Object <- DESeq(dds\n",
    "                # parallel = T, BPPARAM=MulticoreParam(10)\n",
    ")\n",
    "norlCount <- counts(Object, normalized=T)\n",
    "res <- results(Object, contrast=c(\"CellLine\", \"MDAMB231\", \"MCF10A\"))\n",
    "res <- res[complete.cases(res),]\n",
    "# dif_res <- res\n",
    "dif_res <- res[res$padj<0.05 & abs(res$log2FoldChange)>1 & res$baseMean>10, ]\n",
    "\n",
    "\n",
    "## for write out\n",
    "out_res <- dif_res[, c(\"baseMean\", \"log2FoldChange\", \"padj\")]\n",
    "out_res$dif <- \"MDAMB231_hyper\"\n",
    "out_res$dif[out_res$log2FoldChange<0] <- \"MCF10A_hyper\"\n",
    "out_res <- out_res[order(out_res$dif),]\n",
    "out_res <- data.frame(gene_id=rownames(out_res), out_res)\n",
    "out_res <- data.frame(out_res, gene_name=gene_gtf[rownames(out_res), \"gene_name\"])\n",
    "gene_res <- out_res\n",
    "out_res <- gene_res[gene_res$padj<0.05 & abs(gene_res$log2FoldChange)>2 & gene_res$baseMean>10, ]\n",
    "write.table(out_res, \"stat/MDAMB231_vs_MCF10A_differential_genes.txt\", row.names = F, col.names = T, quote = F, sep=\"\\t\")\n",
    "# write.table(gene_res, \"stat/MDAMB231_vs_MCF10A_all_genes.txt\", row.names = F, col.names = T, quote = F, sep=\"\\t\")\n",
    "\n",
    "\n",
    "# plotting \n",
    "anno_colors <- list(CellLine=c(MCF10A=\"#4678BA\", MDAMB231=\"#F4B540\"))\n",
    "colInfo <- Sample_info[SampleIDs, c(\"CellLine\"), drop=F]\n",
    "pheatmap(norlCount[rownames(dif_res),],\n",
    "         # display_numbers=T,\n",
    "         scale=\"row\",\n",
    "         annotation_col = colInfo,\n",
    "         annotation_colors = anno_colors,\n",
    "         main=\"differential gene expression\",\n",
    "         col=colorRampPalette(rev(brewer.pal(11,\"RdBu\")))(21),\n",
    "         # col=colorRampPalette(c(\"#4678BA\", \"#F7F7F7\", \"#EF8354\"))(11),\n",
    "         border=NA,\n",
    "         show_rownames=F,\n",
    "         show_colnames=F,\n",
    "         width=6.5,\n",
    "         height=6,\n",
    "         filename=\"stat/MDAMB231_vs_MCF10A_genes_dif_heatmap.pdf\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Enrichment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(rtracklayer)\n",
    "library(reshape2)\n",
    "library(ggplot2)\n",
    "library(bedtoolsr)\n",
    "library(biomartr)\n",
    "library(data.table)\n",
    "setwd(\"/NAS/wg_looking/gRNA_ONT\")\n",
    " \n",
    "\n",
    "library(gprofiler2)\n",
    "set_base_url(\"https://biit.cs.ut.ee/gprofiler_archive3/e102_eg49_p15\")\n",
    "\n",
    "# for gene_res\n",
    "gene_res <- res\n",
    "totalGenes <- rownames(res)\n",
    "totalGenes <- tstrsplit(totalGenes, \"[.]\")[[1]]\n",
    " \n",
    "diff_genes <- lapply(setNames(c(\"hyper\", \"hypo\"), c(\"hyper\", \"hypo\")), function(x){\n",
    "    d_res <- res[res$padj<0.05 & res$baseMean>10, ]   \n",
    "    if(x==\"hyper\"){\n",
    "        d <- d_res[d_res$log2FoldChange>1, ]   \n",
    "    }\n",
    "     if(x==\"hypo\"){\n",
    "        d <- d_res[d_res$log2FoldChange< -1, ]\n",
    "    }\n",
    "    \n",
    "    g <- tstrsplit(rownames(d), \"[.]\")[[1]]\n",
    "    \n",
    "    return(g)\n",
    "})\n",
    "\n",
    "\n",
    "gostres.list <- lapply(diff_genes, function(g){\n",
    "    gostres <- gost(query=g,\n",
    "                    custom_bg=totalGenes,\n",
    "                    organism = \"hsapiens\",\n",
    "                    user_threshold = 0.01,\n",
    "                    correction_method=\"gSCS\")\n",
    "    gostres <- data.frame(gostres$result)\n",
    "    return(gostres)\n",
    "})\n",
    "\n",
    "\n",
    "# r <- gostres.list[[\"hyper\"]]\n",
    "# r[r$source %in% c(\"GO:BP\", \"KEGG\"), c(\"p_value\", \"term_name\", \"source\")]\n",
    "\n",
    "\n",
    "# r2 <- gostres.list[[\"hypo\"]]\n",
    "# r2[r2$source %in% c(\"GO:BP\", \"GO:MF\"), c(\"term_name\", \"source\")]\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### term plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### term plot\n",
    "plot_gostres_func <- function(class, gostres, height = 7, width = 7){\n",
    "    r <- gostres\n",
    "    pp <- r[(r$source %in% c(\"GO:BP\", \"KEGG\")) , c(\"p_value\", \"term_name\", \"source\")]\n",
    "    order = pp$term_name[order(pp$p_value, decreasing=T)]\n",
    "    pp$term_name <- factor(pp$term_name, levels = order)\n",
    "\n",
    "    ggplot(pp, aes(x=term_name, y=-log10(p_value), fill=source)) +\n",
    "      geom_bar(stat=\"identity\", width = 0.8) +\n",
    "      theme_bw() +\n",
    "      theme(panel.grid = element_blank(),\n",
    "            plot.title = element_text(hjust = 0, size = 12),\n",
    "            axis.text.x  = element_text(angle = 0, hjust=1, vjust=0.5, \n",
    "                                        colour = \"black\", size = 11),\n",
    "            axis.text.y  = element_text(colour = \"black\"),\n",
    "            # legend.position = c(0.8,0.8),\n",
    "            strip.background = element_rect(fill=\"white\", colour = \"black\", size=1) ,\n",
    "            legend.title = element_text(face=\"bold\")) +\n",
    "      # geom_text(aes(label=symbol), vjust=0.5, hjust=0,size=4, colour=\"black\") +\n",
    "      # annotate(\"text\", x=pp$NAME[2], y=-1.5,size=3,\n",
    "      #          label=\"nominal P values: \\n*P < 0.05 \\n**P < 0.005 \\n***P < 0.0005 \\nNS: not significant\") +\n",
    "      # scale_x_discrete(limits=c(pp$term.name))+\n",
    "      scale_fill_manual(values=c(\"GO:BP\"=\"#4678BA\", \"KEGG\"=\"#EF8354\")) +\n",
    "      coord_flip() +\n",
    "      xlab(\"\") +\n",
    "      ggtitle(paste0(\"MDAMB231_VS_MCF10A: \", class))\n",
    "    # facet_grid(.~module, scales = \"free_x\", space = \"free_x\") \n",
    "\n",
    "\n",
    "    ggsave(paste0(\"stat/\", \"MDAMB231_VS_MCF10A_\", class, \"_function_bar.pdf\"),\n",
    "           height = height, width = width)\n",
    "\n",
    "}\n",
    "\n",
    "\n",
    "plot_gostres_func(\"hyper\", gostres.list[[\"hyper\"]], height = 7, width = 6)\n",
    "plot_gostres_func(\"hypo\", gostres.list[[\"hypo\"]], height = 2, width = 6)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### other enrichment function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(clusterProfiler)\n",
    "library(org.Hs.eg.db)\n",
    "\n",
    "\n",
    "ego <- enrichGO(gene          = rownames(hyper),\n",
    "                # universe      = names(geneList), # background genes.\n",
    "                OrgDb         = org.Hs.eg.db,\n",
    "                keyType       = 'ENSEMBL',\n",
    "                ont           = \"ALL\",\n",
    "                pAdjustMethod = \"BH\",\n",
    "                pvalueCutoff  = 0.01,\n",
    "                qvalueCutoff  = 0.05,\n",
    "                readable      = TRUE)\n",
    "head(ego)\n",
    "\n",
    "map <- read.table(\"database/ENSEMBL_ID_Entrez_ID_all.txt\", header=T. drop=F)\n",
    "rownames(map) <- map[,1] \n",
    "kk <- enrichKEGG(gene         = map[rownames(hyper), 2],\n",
    "                 organism     = 'hsa',\n",
    "                 pvalueCutoff = 0.05)\n",
    "data.frame(kk)\n",
    "kk <- enrichKEGG(gene         = map[rownames(hypo), 2],\n",
    "                 organism     = 'hsa',\n",
    "                 pvalueCutoff = 0.05)\n",
    "data.frame(kk)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Featurecount"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(dplyr)\n",
    "library(limma)\n",
    "setwd(\"/NAS/wg_looking/gRNA_ONT\")\n",
    "load(\"stat/assembly_preData.RData\")\n",
    "\n",
    "  \n",
    "Sample_info <- read.table(\"stat/Sample_info2.txt\", header = T, stringsAsFactors = F)\n",
    "rownames(Sample_info) <- Sample_info$Samples\n",
    "\n",
    "IDs <- setNames(Sample_info$Samples, Sample_info$Samples)\n",
    "\n",
    "## for transcripts\n",
    "trans.count.list <- lapply(IDs, function(ID){\n",
    "    cFile = file.path(ID, paste0(ID, \".hg38.featureCounts\"))\n",
    "    tb <- read.table(cFile, header=T)\n",
    "    colnames(tb)[7] <- ID\n",
    "    \n",
    "    return(tb)\n",
    "})\n",
    "\n",
    "trans.count <- join_all(trans.count.list)\n",
    "colnames(trans.count)[1] <- \"transcript_id\"\n",
    "rownames(trans.count) <- trans.count$transcript_id\n",
    "\n",
    "\n",
    "## for genes\n",
    "gene.count.list <- lapply(IDs, function(ID){\n",
    "    cFile = file.path(ID, paste0(ID, \".hg38.gene.featureCounts\"))\n",
    "    tb <- read.table(cFile, header=T)\n",
    "    colnames(tb)[7] <- ID\n",
    "    \n",
    "    return(tb)\n",
    "})\n",
    "\n",
    "gene.count <- join_all(gene.count.list)\n",
    "colnames(gene.count)[1] <- \"gene_id\"\n",
    "rownames(gene.count) <- gene.count$gene_id\n",
    "\n",
    "\n",
    "save(trans.count, gene.count, file=\"stat/hg38.quantify.FeatureCount.RData\")\n",
    "\n",
    "\n",
    "load(\"stat/hg38.quantify.FeatureCount.RData\")\n",
    "v <- c(\"totalRNA_HCT116\", \"totalRNA_NCM460\", \"totalRNA_NCM460_2\")\n",
    "write.table(trans.count[, !(colnames(trans.count) %in% v)], \n",
    "            \"stat/transcripts.FeatureCount.txt\", col.names=T, row.names=F, sep=\"\\t\", quote=F)\n",
    "write.table(gene.count[, !(colnames(gene.count) %in% v)], \n",
    "            \"stat/genes.FeatureCount.txt\", col.names=T, row.names=F, sep=\"\\t\", quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Read Count: sgRNA_K562_1 vs totalRNA_K562_1\n",
    "\n",
    "sub.count.tb <- trans.count[, c(\"totalRNA_K562_1\", \"sgRNA_K562_1\")]\n",
    "logic <- apply(sub.count.tb, 1, function(x){all(x[2] >1)})\n",
    "sub.count.tb <- sub.count.tb[logic,]\n",
    "\n",
    "\n",
    "\n",
    "ggplot(sub.count.tb, aes(x=log2(sgRNA_K562_1 + 1), y=log2(totalRNA_K562_1+1)) ) + \n",
    "    stat_binhex() +\n",
    "    theme_bw() +\n",
    "    theme(\n",
    "      panel.grid=element_blank(),\n",
    "      axis.text.x = element_text(angle = 0, color = \"black\", \n",
    "                                 size = 8, hjust = 0.5, vjust=0.5),\n",
    "      plot.title=element_text(size=14, face=\"bold\", hjust=0.5)\n",
    "    ) +\n",
    "    scale_x_continuous(breaks=c(0,5,10,15)) +\n",
    "    scale_y_continuous(breaks=c(0,5,10,15)) +\n",
    "    scale_fill_gradientn(colours=c(\"#FEE6CE\", \"#C95603\", \"#7F2704\"), name =\"Frequency\",na.value=NA) +\n",
    "    # guides(fill=F) +\n",
    "    geom_abline(slope = 1, intercept = 0, color=\"red\") +\n",
    "    ylim(0,15) +\n",
    "    xlim(0,15)\n",
    "\n",
    "\n",
    "ggsave(\"stat/sg_vs_total_counts_per_transcript.pdf\", width=5, height=4)\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 乳腺癌"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### correlation using transcript"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# correlation using transcript\n",
    "library(DESeq2)\n",
    "library(pheatmap)\n",
    "setwd(\"/NAS/wg_looking/gRNA_ONT\")\n",
    "load(\"stat/hg38.quantify.FeatureCount.RData\")\n",
    "\n",
    "  \n",
    "Sample_info <- read.table(\"stat/Sample_info2.txt\", header = TRUE, stringsAsFactors = FALSE)\n",
    "rownames(Sample_info) <- Sample_info$Samples\n",
    "IDs <- Sample_info$Samples[Sample_info$CellLine %in% c(\"MDAMB231\", \"MCF10A\")]\n",
    "\n",
    "# sgRNA count >= 3 at least three sample\n",
    "sub.trans.count <- trans.count[, IDs]\n",
    "# logic <- apply(sub.trans.count, 1, function(x){any(x >= 3)})\n",
    "logic <- apply(sub.trans.count, 1, function(x){sum(x >= 3)>=3})\n",
    "sub.trans.count <- sub.trans.count[logic,]\n",
    "dim(sub.trans.count) # 55962\n",
    "sub.trans.count <- round(sub.trans.count)\n",
    "\n",
    "\n",
    "# normalization\n",
    "colInfo <- Sample_info[IDs,]\n",
    "dds <- DESeqDataSetFromMatrix(sub.trans.count, \n",
    "                                colInfo, \n",
    "                                ~CellLine)\n",
    "Object <- DESeq(dds\n",
    "                # parallel = T, BPPARAM=MulticoreParam(10)\n",
    "                )\n",
    "norlCount <- counts(Object, normalized=TRUE) # 55962\n",
    "res <- results(Object, contrast=c(\"CellLine\", \"MDAMB231\", \"MCF10A\"))\n",
    "save(\"\")\n",
    "\n",
    "\n",
    "CV <- apply(norlCount, 1, function(x){CV = sd(x, na.rm = T)/mean(x, na.rm = T) * 100})\n",
    "CV <- sort(CV, decreasing=T)\n",
    "\n",
    "\n",
    "## plot correlation\n",
    "anno_colors <- list(CellLine=c(MDAMB231=\"#87B07A\", MCF10A=\"#F4B540\"))\n",
    "colInfo <- Sample_info[IDs, c(\"CellLine\"), drop=FALSE]\n",
    "pheatmap(cor(log2(norlCount+1), method = \"spearman\"),\n",
    "         display_numbers=TRUE,\n",
    "         annotation_col = colInfo,\n",
    "         annotation_row = colInfo,\n",
    "         annotation_colors = anno_colors,\n",
    "         main=\"Correlation of Transcript normalized Count \",\n",
    "         # col=colorRampPalette(rev(brewer.pal(11,\"RdYlBu\")[3:10]))(21),\n",
    "         col=colorRampPalette(c(\"#4678BA\", \"#F7F7F7\", \"#EF8354\"))(100),\n",
    "         border=\"white\",\n",
    "         width=6.5,\n",
    "         height=6,\n",
    "         show_rownames=F,\n",
    "         show_colnames=F,\n",
    "         filename=\"stat/Fig5S.A.MDAMB231_MCF10A_FeatureCount_spearman_correlation.pdf\")\n",
    "\n",
    "\n",
    "pheatmap(cor(log2(norlCount+1), method = \"pearson\"),\n",
    "         display_numbers=TRUE,\n",
    "         annotation_col = colInfo,\n",
    "         annotation_row = colInfo,\n",
    "         annotation_colors = anno_colors,\n",
    "         main=\"Correlation of Transcript normalized Count \",\n",
    "         # col=colorRampPalette(rev(brewer.pal(11,\"RdYlBu\")[3:10]))(21),\n",
    "         col=colorRampPalette(c(\"#4678BA\", \"#F7F7F7\", \"#EF8354\"))(100),\n",
    "         border=\"white\",\n",
    "         width=6.5,\n",
    "         height=6,\n",
    "         show_rownames=F,\n",
    "         show_colnames=F,\n",
    "         filename=\"stat/Fig3A.MDAMB231_MCF10A_FeatureCount_pearson_correlation.pdf\")\n",
    "\n",
    "\n",
    "pheatmap(cor(norlCount[names(CV)[1:10000],], method = \"spearman\"),\n",
    "         display_numbers=T,\n",
    "         annotation_col = colInfo,\n",
    "         annotation_row = colInfo,\n",
    "         annotation_colors = anno_colors,\n",
    "         main=\"Correlation of Transcript normalized Count \",\n",
    "         # col=colorRampPalette(rev(brewer.pal(11,\"RdYlBu\")[3:10]))(21),\n",
    "         # col=colorRampPalette(c(\"#4678BA\", \"#F7F7F7\", \"#EF8354\"))(100),\n",
    "         color=colorRampPalette(c(\"navy\", \"white\", \"firebrick3\"))(100),\n",
    "         border=\"white\",\n",
    "         width=6.5,\n",
    "         height=6,\n",
    "         show_rownames=F,\n",
    "         show_colnames=F,\n",
    "         filename=\"stat/Fig3A.MDAMB231_MCF10A_FeatureCount_spearman_correlation_top.pdf\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# scatter plot\n",
    "library(ggpubr)\n",
    "\n",
    "norlCount <- data.frame(norlCount)\n",
    "\n",
    "combs <- list(\"1\"=c(\"sgRNA_MDAMB231_1\", \"sgRNA_MDAMB231_2\"),\n",
    "             \"2\"=c(\"sgRNA_MDAMB231_1\", \"sgRNA_MDAMB231_3\"),\n",
    "             \"3\"=c(\"sgRNA_MDAMB231_2\", \"sgRNA_MDAMB231_3\"),\n",
    "             \"4\"=c(\"sgRNA_MCF10A_1\", \"sgRNA_MCF10A_2\"),\n",
    "             \"5\"=c(\"sgRNA_MCF10A_1\", \"sgRNA_MCF10A_3\"),\n",
    "             \"6\"=c(\"sgRNA_MCF10A_2\", \"sgRNA_MCF10A_3\"))\n",
    "\n",
    "\n",
    "pdf(\"stat/Fig5S.B.repeat_scatter_plot.pdf\", width = 4, height = 3.8)\n",
    "for(i in 1:6){\n",
    "    comb <- combs[[i]]\n",
    "    p <- ggplot(norlCount, aes(x=log2(get(comb[1])+1), y=log2(get(comb[2])+1))) +\n",
    "        geom_point(size=0.7, alpha=0.4, color=\"#4678BA\")+\n",
    "        stat_cor(method=\"pearson\") +\n",
    "        # scale_fill_manual(values = cols) +\n",
    "        # ggtitle(paste0(comb[1], \" vs \",comb[2])) +\n",
    "        xlab(comb[1]) +\n",
    "        ylab(comb[2]) +\n",
    "        theme_bw() +\n",
    "        theme(\n",
    "          panel.grid=element_blank(),\n",
    "          axis.text.x = element_text(angle = 0, color = \"black\", \n",
    "                                     size = 8, hjust = 0.5, vjust=0.5),\n",
    "          plot.title=element_text(size=8, face=\"bold\")\n",
    "        ) \n",
    "     \n",
    "    plot(p)\n",
    "  # ggsave(paste0(\"stat/sgRNA_MDAMB231_1_vs_sgRNA_MDAMB231_2_scatter_plot.pdf\"),\n",
    "  #        width = 4, height = 4)\n",
    "}\n",
    "   \n",
    "    \n",
    "dev.off()\n",
    "\n",
    "  \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### differential genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(DESeq2)\n",
    "load(\"stat/hg38.quantify.FeatureCount.RData\")\n",
    "load(\"stat/hg38.gene.transcript.gtf.RData\")\n",
    "\n",
    "\n",
    "Sample_info <- read.table(\"stat/Sample_info2.txt\", header = T, stringsAsFactors = F)\n",
    "rownames(Sample_info) <- Sample_info$Samples\n",
    "IDs <- Sample_info$Samples[(Sample_info$CellLine %in% c(\"MDAMB231\", \"MCF10A\")) & (Sample_info$SeqType %in% \"sgRNA\")]\n",
    "\n",
    "# sgRNA count >= 3 at least three sample\n",
    "sub.gene.count <- gene.count[, IDs]\n",
    "# logic <- apply(sub.gene.count, 1, function(x){any(x > 3)})\n",
    "logic <- apply(sub.gene.count, 1, function(x){sum(x >= 3)>=3})\n",
    "sub.gene.count <- sub.gene.count[logic,]\n",
    "dim(sub.gene.count) # 17208\n",
    "sub.gene.count <- round(sub.gene.count)\n",
    "\n",
    "# run DESeq2\n",
    "colInfo <- Sample_info[IDs, c(\"CellLine\"), drop=F]\n",
    "dds <- DESeqDataSetFromMatrix(sub.gene.count, \n",
    "                                colInfo, \n",
    "                                ~CellLine)\n",
    "Object <- DESeq(dds\n",
    "                # parallel = T, BPPARAM=MulticoreParam(10)\n",
    "                )\n",
    "gene.norlCount <- counts(Object, normalized=TRUE)\n",
    "res <- results(Object, contrast=c(\"CellLine\", \"MDAMB231\", \"MCF10A\"))\n",
    "# res <- res[complete.cases(res),]\n",
    "# dif_res <- res\n",
    "# dif_res <- res[res$padj<0.05 & abs(res$log2FoldChange)>1 & res$baseMean>10, ]\n",
    "save(Object, gene.norlCount, res, file=\"stat/MDAMB231_vs_MCF10A_gene_object.RData\")\n",
    "\n",
    "\n",
    "out_res <- data.frame(gene.norlCount, res[, c(\"baseMean\", \"log2FoldChange\", \"pvalue\", \"padj\")])\n",
    "out_res$dif <- \"ns\"\n",
    "out_res$dif[res$padj<0.05 & res$log2FoldChange>1] <- \"Up\"\n",
    "out_res$dif[res$padj<0.05 & res$log2FoldChange < -1] <- \"Down\"\n",
    "table(out_res$dif)\n",
    "out_res <- out_res[order(out_res$dif, decreasing=TRUE),]\n",
    "out_res <- data.frame(gene_id=rownames(out_res), out_res)\n",
    "out_res <- data.frame(out_res, gene_name=gene_gtf[rownames(out_res), \"gene_name\"])\n",
    "gene_res <- out_res\n",
    "# out_res <- gene_res[transcript_res$padj<0.05 & abs(transcript_res$log2FoldChange)>2 & transcript_res$baseMean>10, ]\n",
    "# write.table(out_res, \"stat/MDAMB231_vs_MCF10A_differential_genes.txt\", row.names = F, col.names = T, quote = F, sep=\"\\t\")\n",
    "write.table(out_res, \"stat/MDAMB231_vs_MCF10A_all_genes.txt\", \n",
    "            row.names = FALSE, col.names = TRUE, quote = FALSE, sep=\"\\t\")\n",
    "\n",
    "\n",
    "\n",
    "## plot diff genes\n",
    "anno_colors <- list(CellLine=c(MDAMB231=\"#87B07A\", MCF10A=\"#F4B540\"))\n",
    "colInfo <- Sample_info[IDs, c(\"CellLine\"), drop=F]\n",
    "pheatmap(gene.norlCount[rownames(dif_res),],\n",
    "         # display_numbers=T,\n",
    "         scale=\"row\",\n",
    "         annotation_col = colInfo,\n",
    "         annotation_colors = anno_colors,\n",
    "         main=\"MDAMB231_vs_MCF10A: differential gene expression\",\n",
    "         col=colorRampPalette(rev(brewer.pal(11,\"RdBu\")))(21),\n",
    "         # color = colorRampPalette(c(\"navy\", \"white\", \"firebrick3\"))(100))\n",
    "         # col=colorRampPalette(c(\"#4678BA\", \"#F7F7F7\", \"#EF8354\"))(11),\n",
    "         border=NA,\n",
    "         show_rownames=F,\n",
    "         show_colnames=F,\n",
    "         width=6.5,\n",
    "         height=6,\n",
    "         filename=\"stat/MDAMB231_vs_MCF10A_genes_dif_heatmap.pdf\")\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "load(\"stat/MDAMB231_vs_MCF10A_gene_object.RData\")\n",
    "\n",
    "\n",
    "# volcano plot\n",
    "res <- res[complete.cases(res),]\n",
    "res$dif <- \"ns\"\n",
    "res$dif[res$log2FoldChange>1 & res$padj<0.05] <- \"Up\"\n",
    "res$dif[res$log2FoldChange< -1 & res$padj<0.05] <- \"Down\"\n",
    "res <- data.frame(res)\n",
    "\n",
    "ggplot(res, aes(x=log2FoldChange, y=-log10(padj), color=dif)) +\n",
    "    geom_point(size=0.3, alpha=0.4) +\n",
    "    scale_color_manual(values = c(Down=\"#327FFF\", Up=\"#FD7E29\", ns=\"#AFAFAF\")) +\n",
    "    scale_y_log10() +\n",
    "    theme_bw() +\n",
    "    theme(\n",
    "    panel.grid=element_blank(),\n",
    "    axis.text.x = element_text(angle = 0, color = \"black\", \n",
    "                               size = 8, hjust = 0.5, vjust = 0.5),\n",
    "    plot.title=element_text(size=14, face=\"bold\")\n",
    "    ) + \n",
    "    ggtitle(\"\")\n",
    "\n",
    "\n",
    "ggsave(\"stat/Fig5S.C.MDAMB231_VS_MCF10A_gene_volcano_plot.pdf\",\n",
    "       width = 5, height = 3.2)\n",
    "\n",
    "\n",
    "\n",
    "# pie\n",
    "pp <- data.frame(table(res$dif))\n",
    "\n",
    "p <- pp %>%\n",
    "    mutate(Percentage = Freq / sum(Freq)*100)\n",
    "  \n",
    "  \n",
    "plot <- ggplot(p, aes(x=\"\", y=Percentage, fill=Var1))+\n",
    "    geom_bar(width = 1, stat = \"identity\") +\n",
    "    coord_polar(\"y\", start=0) +\n",
    "    scale_fill_manual(values = c(Down=\"#327FFF\", Up=\"#FD7E29\", ns=\"#AFAFAF\"))+\n",
    "    geom_text(aes(x = 1.6, label = paste0(Freq, \"(\",scales::percent(Percentage/100, accuracy = .1), \")\")), \n",
    "              position = position_stack(vjust = .5)) +\n",
    "    # facet_wrap(~Method)\n",
    "    ggtitle(\"\") +\n",
    "    theme_minimal() +\n",
    "    theme(\n",
    "      axis.title.x = element_blank(),\n",
    "      axis.title.y = element_blank(),\n",
    "      panel.border = element_blank(),\n",
    "      panel.grid=element_blank(),\n",
    "      axis.ticks = element_blank(),\n",
    "      plot.title=element_text(size=14, face=\"bold\", hjust=0.5),\n",
    "      axis.text.x=element_blank()\n",
    "    ) \n",
    "\n",
    "\n",
    "ggsave(\"stat/Fig5S.C.MDAMB231_VS_MCF10A_gene_dif_pie.pdf\",\n",
    "       width = 6, height = 3)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### enrichment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 5</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>baseMean</th><th scope=col>log2FoldChange</th><th scope=col>padj</th><th scope=col>dif</th><th scope=col>gene_name</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>ENSG00000238009.6</th><td> 19.558448</td><td>-0.4091211</td><td>4.851132e-01</td><td>MCF10A_hyper</td><td>AL627309.1</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000241860.7</th><td> 73.922692</td><td>-0.3861753</td><td>1.559391e-01</td><td>MCF10A_hyper</td><td>AL627309.5</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000222623.1</th><td>  3.029967</td><td>-1.0892758</td><td>4.599809e-01</td><td>MCF10A_hyper</td><td>RNU6-1100P</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000241599.1</th><td>  4.412292</td><td>-1.8146595</td><td>1.127330e-01</td><td>MCF10A_hyper</td><td>AL627309.4</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000237094.12</th><td>114.626694</td><td>-0.9564438</td><td>6.357017e-07</td><td>MCF10A_hyper</td><td>AL732372.2</td></tr>\n",
       "\t<tr><th scope=row>ENSG00000278757.1</th><td>  2.683375</td><td>-0.8313035</td><td>6.107363e-01</td><td>MCF10A_hyper</td><td>U6        </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 5\n",
       "\\begin{tabular}{r|lllll}\n",
       "  & baseMean & log2FoldChange & padj & dif & gene\\_name\\\\\n",
       "  & <dbl> & <dbl> & <dbl> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\tENSG00000238009.6 &  19.558448 & -0.4091211 & 4.851132e-01 & MCF10A\\_hyper & AL627309.1\\\\\n",
       "\tENSG00000241860.7 &  73.922692 & -0.3861753 & 1.559391e-01 & MCF10A\\_hyper & AL627309.5\\\\\n",
       "\tENSG00000222623.1 &   3.029967 & -1.0892758 & 4.599809e-01 & MCF10A\\_hyper & RNU6-1100P\\\\\n",
       "\tENSG00000241599.1 &   4.412292 & -1.8146595 & 1.127330e-01 & MCF10A\\_hyper & AL627309.4\\\\\n",
       "\tENSG00000237094.12 & 114.626694 & -0.9564438 & 6.357017e-07 & MCF10A\\_hyper & AL732372.2\\\\\n",
       "\tENSG00000278757.1 &   2.683375 & -0.8313035 & 6.107363e-01 & MCF10A\\_hyper & U6        \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 5\n",
       "\n",
       "| <!--/--> | baseMean &lt;dbl&gt; | log2FoldChange &lt;dbl&gt; | padj &lt;dbl&gt; | dif &lt;chr&gt; | gene_name &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|\n",
       "| ENSG00000238009.6 |  19.558448 | -0.4091211 | 4.851132e-01 | MCF10A_hyper | AL627309.1 |\n",
       "| ENSG00000241860.7 |  73.922692 | -0.3861753 | 1.559391e-01 | MCF10A_hyper | AL627309.5 |\n",
       "| ENSG00000222623.1 |   3.029967 | -1.0892758 | 4.599809e-01 | MCF10A_hyper | RNU6-1100P |\n",
       "| ENSG00000241599.1 |   4.412292 | -1.8146595 | 1.127330e-01 | MCF10A_hyper | AL627309.4 |\n",
       "| ENSG00000237094.12 | 114.626694 | -0.9564438 | 6.357017e-07 | MCF10A_hyper | AL732372.2 |\n",
       "| ENSG00000278757.1 |   2.683375 | -0.8313035 | 6.107363e-01 | MCF10A_hyper | U6         |\n",
       "\n"
      ],
      "text/plain": [
       "                   baseMean   log2FoldChange padj         dif         \n",
       "ENSG00000238009.6   19.558448 -0.4091211     4.851132e-01 MCF10A_hyper\n",
       "ENSG00000241860.7   73.922692 -0.3861753     1.559391e-01 MCF10A_hyper\n",
       "ENSG00000222623.1    3.029967 -1.0892758     4.599809e-01 MCF10A_hyper\n",
       "ENSG00000241599.1    4.412292 -1.8146595     1.127330e-01 MCF10A_hyper\n",
       "ENSG00000237094.12 114.626694 -0.9564438     6.357017e-07 MCF10A_hyper\n",
       "ENSG00000278757.1    2.683375 -0.8313035     6.107363e-01 MCF10A_hyper\n",
       "                   gene_name \n",
       "ENSG00000238009.6  AL627309.1\n",
       "ENSG00000241860.7  AL627309.5\n",
       "ENSG00000222623.1  RNU6-1100P\n",
       "ENSG00000241599.1  AL627309.4\n",
       "ENSG00000237094.12 AL732372.2\n",
       "ENSG00000278757.1  U6        "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "library(gprofiler2)\n",
    "res <- read.table(\"E:/项目/转座子/sgRNA/results/Figure/MDAMB231_vs_MCF10A_all_genes.txt\", header=T, row.names=1)\n",
    "head(res)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Detected custom background input, domain scope is set to 'custom'\n",
      "\n",
      "Detected custom background input, domain scope is set to 'custom'\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# library(rtracklayer)\n",
    "# library(reshape2)\n",
    "# library(ggplot2)\n",
    "# library(bedtoolsr)\n",
    "# library(biomartr)\n",
    "library(data.table)\n",
    "# setwd(\"/NAS/wg_looking/gRNA_ONT\")\n",
    " \n",
    "\n",
    "# library(gprofiler2)\n",
    "# set_base_url(\"https://biit.cs.ut.ee/gprofiler_archive3/e103_eg50_p15/\")\n",
    "\n",
    "\n",
    "# # for gene_res\n",
    "gene_res <- res\n",
    "totalGenes <- rownames(res)\n",
    "totalGenes <- tstrsplit(totalGenes, \"[.]\")[[1]]\n",
    " \n",
    "diff_genes <- lapply(setNames(c(\"hyper\", \"hypo\"), c(\"hyper\", \"hypo\")), function(x){\n",
    "    d_res <- res[res$padj<0.05 & res$baseMean>10, ]   \n",
    "    if(x==\"hyper\"){\n",
    "        d <- d_res[d_res$log2FoldChange>1, ]   \n",
    "    }\n",
    "     if(x==\"hypo\"){\n",
    "        d <- d_res[d_res$log2FoldChange< -1, ]\n",
    "    }\n",
    "    \n",
    "    g <- tstrsplit(rownames(d), \"[.]\")[[1]]\n",
    "    \n",
    "    return(g)\n",
    "})\n",
    "\n",
    "lengths(diff_genes)\n",
    "\n",
    "gostres.list <- lapply(diff_genes, function(g){\n",
    "    gostres <- gost(query=g,\n",
    "                    custom_bg=totalGenes,\n",
    "                    organism = \"hsapiens\",\n",
    "                    user_threshold = 0.05,\n",
    "                    correction_method=\"gSCS\")\n",
    "    gostres <- data.frame(gostres$result)\n",
    "    return(gostres)\n",
    "})\n",
    "\n",
    "\n",
    "# r <- gostres.list[[\"hyper\"]]\n",
    "# r[r$source %in% c(\"GO:BP\", \"KEGG\"), c(\"p_value\", \"term_name\", \"source\")]\n",
    "\n",
    "\n",
    "# r2 <- gostres.list[[\"hypo\"]]\n",
    "# r2[r2$source %in% c(\"GO:BP\", \"GO:MF\"), c(\"term_name\", \"source\")]\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 3 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>p_value</th><th scope=col>term_name</th><th scope=col>source</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>82</th><td>6.198851e-08</td><td>Herpes simplex virus 1 infection</td><td>KEGG</td></tr>\n",
       "\t<tr><th scope=row>83</th><td>4.770909e-05</td><td>KEGG root term                  </td><td>KEGG</td></tr>\n",
       "\t<tr><th scope=row>84</th><td>3.766876e-02</td><td>Morphine addiction              </td><td>KEGG</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 3 × 3\n",
       "\\begin{tabular}{r|lll}\n",
       "  & p\\_value & term\\_name & source\\\\\n",
       "  & <dbl> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t82 & 6.198851e-08 & Herpes simplex virus 1 infection & KEGG\\\\\n",
       "\t83 & 4.770909e-05 & KEGG root term                   & KEGG\\\\\n",
       "\t84 & 3.766876e-02 & Morphine addiction               & KEGG\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 3 × 3\n",
       "\n",
       "| <!--/--> | p_value &lt;dbl&gt; | term_name &lt;chr&gt; | source &lt;chr&gt; |\n",
       "|---|---|---|---|\n",
       "| 82 | 6.198851e-08 | Herpes simplex virus 1 infection | KEGG |\n",
       "| 83 | 4.770909e-05 | KEGG root term                   | KEGG |\n",
       "| 84 | 3.766876e-02 | Morphine addiction               | KEGG |\n",
       "\n"
      ],
      "text/plain": [
       "   p_value      term_name                        source\n",
       "82 6.198851e-08 Herpes simplex virus 1 infection KEGG  \n",
       "83 4.770909e-05 KEGG root term                   KEGG  \n",
       "84 3.766876e-02 Morphine addiction               KEGG  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 11 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>p_value</th><th scope=col>term_name</th><th scope=col>source</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>245</th><td>1.804825e-05</td><td>KEGG root term                                          </td><td>KEGG</td></tr>\n",
       "\t<tr><th scope=row>246</th><td>1.282189e-03</td><td>Pathways in cancer                                      </td><td>KEGG</td></tr>\n",
       "\t<tr><th scope=row>247</th><td>1.300911e-03</td><td>PI3K-Akt signaling pathway                              </td><td>KEGG</td></tr>\n",
       "\t<tr><th scope=row>248</th><td>1.378893e-03</td><td>ECM-receptor interaction                                </td><td>KEGG</td></tr>\n",
       "\t<tr><th scope=row>249</th><td>3.275763e-03</td><td>AGE-RAGE signaling pathway in diabetic complications    </td><td>KEGG</td></tr>\n",
       "\t<tr><th scope=row>250</th><td>8.410845e-03</td><td>Focal adhesion                                          </td><td>KEGG</td></tr>\n",
       "\t<tr><th scope=row>251</th><td>1.080766e-02</td><td>Amoebiasis                                              </td><td>KEGG</td></tr>\n",
       "\t<tr><th scope=row>252</th><td>1.213418e-02</td><td>Small cell lung cancer                                  </td><td>KEGG</td></tr>\n",
       "\t<tr><th scope=row>253</th><td>1.760866e-02</td><td>One carbon pool by folate                               </td><td>KEGG</td></tr>\n",
       "\t<tr><th scope=row>254</th><td>2.754254e-02</td><td>Signaling pathways regulating pluripotency of stem cells</td><td>KEGG</td></tr>\n",
       "\t<tr><th scope=row>255</th><td>4.329117e-02</td><td>Relaxin signaling pathway                               </td><td>KEGG</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 11 × 3\n",
       "\\begin{tabular}{r|lll}\n",
       "  & p\\_value & term\\_name & source\\\\\n",
       "  & <dbl> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t245 & 1.804825e-05 & KEGG root term                                           & KEGG\\\\\n",
       "\t246 & 1.282189e-03 & Pathways in cancer                                       & KEGG\\\\\n",
       "\t247 & 1.300911e-03 & PI3K-Akt signaling pathway                               & KEGG\\\\\n",
       "\t248 & 1.378893e-03 & ECM-receptor interaction                                 & KEGG\\\\\n",
       "\t249 & 3.275763e-03 & AGE-RAGE signaling pathway in diabetic complications     & KEGG\\\\\n",
       "\t250 & 8.410845e-03 & Focal adhesion                                           & KEGG\\\\\n",
       "\t251 & 1.080766e-02 & Amoebiasis                                               & KEGG\\\\\n",
       "\t252 & 1.213418e-02 & Small cell lung cancer                                   & KEGG\\\\\n",
       "\t253 & 1.760866e-02 & One carbon pool by folate                                & KEGG\\\\\n",
       "\t254 & 2.754254e-02 & Signaling pathways regulating pluripotency of stem cells & KEGG\\\\\n",
       "\t255 & 4.329117e-02 & Relaxin signaling pathway                                & KEGG\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 11 × 3\n",
       "\n",
       "| <!--/--> | p_value &lt;dbl&gt; | term_name &lt;chr&gt; | source &lt;chr&gt; |\n",
       "|---|---|---|---|\n",
       "| 245 | 1.804825e-05 | KEGG root term                                           | KEGG |\n",
       "| 246 | 1.282189e-03 | Pathways in cancer                                       | KEGG |\n",
       "| 247 | 1.300911e-03 | PI3K-Akt signaling pathway                               | KEGG |\n",
       "| 248 | 1.378893e-03 | ECM-receptor interaction                                 | KEGG |\n",
       "| 249 | 3.275763e-03 | AGE-RAGE signaling pathway in diabetic complications     | KEGG |\n",
       "| 250 | 8.410845e-03 | Focal adhesion                                           | KEGG |\n",
       "| 251 | 1.080766e-02 | Amoebiasis                                               | KEGG |\n",
       "| 252 | 1.213418e-02 | Small cell lung cancer                                   | KEGG |\n",
       "| 253 | 1.760866e-02 | One carbon pool by folate                                | KEGG |\n",
       "| 254 | 2.754254e-02 | Signaling pathways regulating pluripotency of stem cells | KEGG |\n",
       "| 255 | 4.329117e-02 | Relaxin signaling pathway                                | KEGG |\n",
       "\n"
      ],
      "text/plain": [
       "    p_value      term_name                                               \n",
       "245 1.804825e-05 KEGG root term                                          \n",
       "246 1.282189e-03 Pathways in cancer                                      \n",
       "247 1.300911e-03 PI3K-Akt signaling pathway                              \n",
       "248 1.378893e-03 ECM-receptor interaction                                \n",
       "249 3.275763e-03 AGE-RAGE signaling pathway in diabetic complications    \n",
       "250 8.410845e-03 Focal adhesion                                          \n",
       "251 1.080766e-02 Amoebiasis                                              \n",
       "252 1.213418e-02 Small cell lung cancer                                  \n",
       "253 1.760866e-02 One carbon pool by folate                               \n",
       "254 2.754254e-02 Signaling pathways regulating pluripotency of stem cells\n",
       "255 4.329117e-02 Relaxin signaling pathway                               \n",
       "    source\n",
       "245 KEGG  \n",
       "246 KEGG  \n",
       "247 KEGG  \n",
       "248 KEGG  \n",
       "249 KEGG  \n",
       "250 KEGG  \n",
       "251 KEGG  \n",
       "252 KEGG  \n",
       "253 KEGG  \n",
       "254 KEGG  \n",
       "255 KEGG  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "\n",
    "r <- gostres.list[[\"hyper\"]]\n",
    "r[r$source %in% c(\"KEGG\"), c(\"p_value\", \"term_name\", \"source\")]\n",
    "\n",
    "r <- gostres.list[[\"hypo\"]]\n",
    "r[r$source %in% c(\"KEGG\"), c(\"p_value\", \"term_name\", \"source\")]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### term plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0gAAANICAMAAADKOT/pAAAAOVBMVEUAAAAzMzNGeLpNTU1o\naGh8fHyMjIyampqnp6eysrK9vb3Hx8fQ0NDZ2dnh4eHp6envg1Tw8PD///9x/L4bAAAACXBI\nWXMAABJ0AAASdAHeZh94AAAgAElEQVR4nO3d6WIiudJuYe0DGIwHPrj/iz1kAiJTDBVURccr\nifX8sLGNATtiNYMpOh0A/LOkvgBADwgJcEBIgANCAhwQEuCAkAAHhAQ4ICTAASEBDggJcPCu\nIS0+trvxwG77sRjep9Fys8/HSIvnx139nL68XabF6dv265TWv5cT2Ba/25+0zIeX6ft49M3y\neCrbP160w+F3vUjr79PhdFGcy2ZxuRizCz+R3nXYEd71d3tcxPV4YH1eyct6Lk5LfPg+Hv7+\nw3HHZjanbxtWeHH97HH5b/Z2kc7pHXbDnu8X6fqtzy7a+SzScjc98zQ/l9XpKDcXfnbC5l8P\nXvWuv9vjzp3+m71YXuIY3u5WaXU6wjptzvv86Lib8ai/ab0frhfWwyfGNx/jcX8XN3v7mTbn\nQ5vh0DqtduM5buZHuzm7z7QYrr8+T5HPTjafy09a/A4f/ZQXfnbCtt8N/sK7/m7TcdWGq47f\n4/tJHOcbXcPHi+M1yPPjju8+rocXaZ+/uE2rm73dX28spt1wxNPtwfJ45dntLteS6zGO6dGv\n57IZL/ZX+iwv/OyEDb8Y/J13/d0eb/qk4d7JNn3NQ/o+/af863g9sUlfT487vR9yXdLTZ4/f\nfru3q3OkP+N12aO9Ls9uc67jsP/YFt92PZePIc1jex+3F356GTdp8XlM93QLcHh/+tz5OnG4\nt7c9HXO/PF+zwuZ9Q9qPm3JcwHlI5y1bHW8k/Zxv5j047iZdHyfYX24RXj77ey+U73zn53s8\n5npnuWir9Dv/8vXw9Vxm7+YXfvKdH8P9qO3xlMdbgMP1V0qf4wMnh/EMLwfHYxa3OPHU+4Z0\nvBF3GK8/5iGdDpxuhZ1uq9077miyadvzdc3X5LN3rnEWs61fDY8S/twcqTy74nTmD9rdC2l2\n4WffudofL+nyEvT6eC2WzvetvobMj1/ej9ea4zHxijcOaTP+d3t9N6SvsYfzzaM7xz39tztf\nU+wW59tB24/F5YbYvZBOJ/h1ie17eFxuUT68Vp7dyyHNLvzsO3/OR1me7s0tx9uRwwUZrgI/\nxk+O14Yp3faNp944pOGWzedx2+6FtBw36fd0M+/Bcb/zw9n7xep6yuvLLb47If2ON5ymt9V+\nPhfl0pZndxPSvQ+n72YX/uaow9vtUPvP8GbyfZPH1XlY4mXv+hs7rsruuNWr8dbN+ROj4bPH\nNxe7J8f9vdwxWk1XNj82d28bh6uC/XzDbxa+PLuP3N339VHB6bEHkxuN8wt/c9Th7XgZP08P\nHeYvEdK/eNff2Pnh6mGh5iGNt4s+81J9Pjvu+W9Py9WuPOnp+6nteFWznX359uHv+dl9Xm4s\nnp4acT+k06N2u+F22fzC3xz1/Ejg92E5ObWiHkJ62bv+xoZVOf/Zsvw70s/w5vwknby69457\nuu75zg/Yne7e7y7XMPe2cbg2Ot1BOa7+dnIqTy5a/jvSKpUPf18//Dzf19mUF/7mqOPb45Xp\n7/k/Ej+H88OJH9cnQxDSy971NzasytfxP9pf8zhOz2z4zX9DGe/OPDjufnxOwi53dHpmw/7j\nyX2k8ZHv82PgPyltj0X9rFLxbLubs1uPD0jsPs5/obob0vWZDcWFvznq6e0yLc43W0/f9z2c\n5/Hg8Trzg5D+wrv+xi53JnbX/bo+126T/9s8/if+8XH3pyfEXR5FW+Q/xFzO4sbkWXCby4N/\nf7po5+fR5efa3Rx7sLyc1p0LPzvq+ZGSdHkgZTzxj+vZLHaE9Bfe9Td2XvzFYR7Harxbsbje\n2FosHh739FzryV308RnYy+3sLG5Mnpc9PKc7rcoHqW/P7nhl8XE94oOQ9uOzvx9c+OlRz1en\n53tQx48+8mXeHmtc3zyfDxb8xt7T9/kxPZJxwu/xPV3umRGSE36P/7GUZjf+/uFIvhdqdTkU\ncob94/f4H6sypEV+ZI+QnPB7BBwQEuCAkAAHhAQ4ICTAASEBDggJcPCeIf0PXVHv0+FtQ1Jf\nAHiqYZyEhObVME5CQvNqGCchoXk1jJOQ0LwaxklIaF4N4yQkNK+GcRISmlfDOAkJzathnISE\n5tUwTkJC82oYJyGheTWMk5DQvBrGSUhoXg3jJCQ0r4ZxEhKaV8M4CQnNq2GchITm1TBOQkLz\nahgnIaF5NYyTkNC8GsZJSGheDeMkJDSvhnESEppXwzgJCc2rYZyEhObVME5CQvNqGCchoXk1\njJOQ0LwaxklIaF4N4yQkNK+GcRISmlfDOAkJzathnISE5tUwTkJC82oYJyGheTWMk5DQvBrG\nSUhoXg3jJCQ0r4ZxEhKaV8M4CQnNq2GchITm1TBOQkLzahgnIaF5NYyTkNC8GsZJSGheDeMk\npOz/UD/7OIMRUqbeERjYxxmMkDL1jsDAPs5ghJSpdwQG9nEGI6RMvSMwsI8zGCFl6h2BgX2c\nwQgpU+8IDOzjDEZImXpHYGAfZzBCytQ7AgP7OIMRUqbeERjYxxmMkDL1jsDAPs5ghJSpdwQG\n9nEGI6RMvSMwsI8zGCFl6h2BgX2cwQgpU+8IDOzjDEZImXpHYGAfZzBCytQ7AgP7OIMRUqbe\nERjYxxmMkDL1jsDAPs5ghJSpdwQG9nEGI6RMvSMwsI8zGCFl6h2BgX2cwQgpU+8IDOzjDEZI\nmXpHYGAfZzBCytQ7AgP7OIMRUqbeERjYxxmMkDL1jsDAPs5ghJSpdwQG9nEGI6RMvSMwsI8z\nGCFl6h2BgX2cwQgpU+8IDOzjDEZImXpHYGAfZzBCytQ7AgP7OIMRUqbeERjYxxmMkDL1jsDA\nPs5ghJSpdwQG9nEGI6RMvSMwsI8zGCFl6h2BgX2cwQgpU+8IDOzjDEZImXpHYGAfZzBCytQ7\nAgP7OIMRUqbeERjYxxmMkDL1jsDAPs5ghJSpdwQG9nEGI6RMvSMwsI8zGCFl6h2BgX2cwQgp\nU+8IDOzjDEZImXpHYGAfZzBCytQ7AgP7OIMRUqbeERjYxxmMkDL1jsDAPs5ghJSpdwQG9nEG\nI6RMvSMwsI8zGCFl6h2BgX2cwQgpU+8IDOzjDEZImXpHYGAfZzBCytQ7AgP7OIMRUqbeERjY\nxxmMkDL1jsDAPs5ghJSpdwQG9nEGI6RMvSMwsI8zGCFl6h2BgX2cwQgpU+8IDOzjDEZImXpH\nYGAfZzBCytQ7AgP7OIMRUqbeERjYxxmMkDL1jsDAPs5ghJSpdwQG9nEGI6RMvSMwsI8zGCFl\n6h2BgX2cwQgpU+8IDOzjDEZImXpHYGAfZzBCytQ7AgP7OIMRUqbeERjYxxmMkDL1jsDAPs5g\nhJSpdwQG9nEGI6RMvSMwsI8zGCFl6h2BgX2cwQgpU+8IDOzjDEZImXpHYGAfZzBCytQ7AgP7\nOIMRUqbeERjYxxmMkDL1jsDAPs5ghJSpdwQG9nEGI6RMvSMwsI8zGCFl6h2BgX2cwQgpU+8I\nDOzjDEZImXpHYGAfZzBCytQ7AgP7OIP9MaRkPN4Lp/30tP5wRunmwHD4/jc9PilCapV9nMH+\ny5D++jyffOu977V+7oqQWmUfZ7C2QkrmaAipT/ZxBnslpJTGK4Q0XC8Mb8YvpeL9YXrEy9Gu\n35hmp5WuZ3E5mPIZzM+1vLyzE0/l5RjfFpd5ciqE1Kq7O9pYSJfVPm1mKj45P+7lc/PjXr9n\n+m2Tms4hpcP9E54emp/ezdHLi5eP87/RvZ9UvSMwuLujbYR0dm+1y08Wp/nkuNfrm+mxyy+W\nJzy9vPdqTIc7p3YvR0Jq1d0dbSOk/K4MavbJ2cMAKV/JPAwp5Zti0zO6E1Lx+MLDkKYXbnZq\nhNSTuzvaWkjTz9y5wrhJ6U/XSMUFSOX259t70xOeHLi5RireE1J/7g2ut5Bubt9ZQkrlwTsh\n3e2OkN7SvcE1GVJ+/2RTy2CuNwNvPj9t4Pa00m0CT0O6/RwhdebujjYW0uRR7fHqw/bw92G6\nzU8e/j4UW38oHsAuL+9t2bPLcfPw9+xnJaRW3d3RJkJy/L6/PS9/hNQq+ziDEVKm3hEY2McZ\njJAy9Y7AwD7OYPUsdyRCapV9nMEIKVPvCAzs4wxGSJl6R2BgH2cwQsrUOwID+ziDEVKm3hEY\n2McZjJAy9Y7AwD7OYISUqXcEBvZxBiOkTL0jMLCPMxghZeodgYF9nMEIKVPvCAzs4wxGSJl6\nR2BgH2cwQsrUOwID+ziDEVKm3hEY2McZjJAy9Y7AwD7OYISUqXcEBvZxBiOkTL0jMLCPMxgh\nZeodgYF9nMEIKVPvCAzs4wxGSJl6R2BgH2cwQsrUOwID+ziDEVKm3hEY2McZjJAy9Y7AwD7O\nYISUqXcEBvZxBiOkTL0jMLCPMxghZeodgYF9nMEIKVPvCAzs4wxGSJl6R2BgH2cwQsrUOwID\n+ziDEVKm3hEY2McZjJAy9Y7AwD7OYISUqXcEBvZxBiOkTL0jMLCPMxghZeodgYF9nMEIKVPv\nCAzs4wxGSJl6R2BgH2cwQsrUOwID+ziDEVKm3hEY2McZjJAy9Y7AwD7OYISUqXcEBvZxBiOk\nTL0jMLCPMxghZeodgYF9nMEIKVPvCAzs4wxGSJl6R2BgH2cwQsrUOwID+ziDEVKm3hEY2McZ\njJAy9Y7AwD7OYISUqXcEBvZxBiOkTL0jMLCPMxghZeodgYF9nMEIKVPvCAzs4wxGSJl6R2Bg\nH2cwQsrUOwID+ziDEVKm3hEY2McZjJAy9Y7AwD7OYISUqXcEBvZxBiOkTL0jMLCPMxghZeod\ngYF9nMEIKVPvCAzs4wxGSJl6R2BgH2cwQsrUOwID+ziDEVKm3hEY2McZjJAy9Y7AwD7OYISU\nqXcEBvZxBiOkTL0jMLCPMxghZeodgYF9nMEIKVPvCAzs4wxGSJl6R2BgH2cwQsrUOwID+ziD\nEVKm3hEY2McZjJAy9Y7AwD7OYISUqXcEBvZxBiOkTL0jMLCPMxghZeodgYF9nMEIKVPvCAzs\n4wxGSJl6R2BgH2cwQsrUOwID+ziDEVKm3hEY2McZjJAy9Y7AwD7OYISUqXcEBvZxBiOkTL0j\nMLCPMxghZeodgYF9nMEIKVPvCAzs4wxGSJl6R2BgH2cwQkLzahgnIaF5NYyTkNC8GsZJSGhe\nDeMkJDSvhnESEppXwzgJCc2rYZyEhObVME5CQvNqGCchoXk1jJOQ0LwaxklIaF4N4yQkNK+G\ncRISmlfDOAkJzathnISE5tUwTkJC82oYJyGheTWMk5DQvBrGSUhoXg3jJCQ0r4ZxEhKaV8M4\nCQnNq2GchITm1TBOQkLzahgnIaF5NYyTkNC8GsZJSGheDeMkJDSvhnESEppXwzgJCc2rYZyE\nhObVME5CQvNqGCchoXk1jJOQ0LwaxklIaF4N4yQkNK+GcRISmlfDOAkJzathnISE5tUwTkJC\n82oYJyGheTWMk5DQvBrGSUhoXg3jJCQ0r4ZxEhKaV8M4CQnNq2GchJT93xuLHoAvQlIhpEL0\nAHwRkgohFaIH4IuQVAipED0AX4SkQkiF6AH4IiQVQipED8AXIakQUiF6AL4ISYWQCtED8EVI\nKoRUiB6AL0JSIaRC9AB8EZIKIRWiB+CLkFQIqRA9AF+EpEJIhegB+CIkFUIqRA/AFyGpEFIh\negC+CEmFkArRA/BFSCqEVIgegC9CUiGkQvQAfBGSCiEVogfgi5BUCKkQPQBfhKRCSIXoAfgi\nJBVCKkQPwBchqRBSIXoAvghJhZAK0QPwRUgqhFSIHoAvQlIhpEL0AHwRkgohFaIH4IuQVAip\nED0AX4SkQkiF6AH4IiQVQipED8AXIakQUiF6AL4ISYWQCtED8EVIKoRUiB6AL0JSIaRC9AB8\nEZIKIRWiB+CLkFQIqRA9AF+EpEJIhegB+CIkFUIqRA/AFyGpEFIhegC+CEmFkArRA/BFSCqE\nVIgegC9CUiGkQvQAfBGSCiEVogfgi5BUCKkQPQBfhKRCSIXoAfgiJBVCKkQPwBchqRBSIXoA\nvghJhZAK0QPwRUgqhFSIHoAvQlIhpEL0AHwRkgohFaIH4IuQVAipED0AX4SkQkiF6AH4IiQV\nQipED8AXIakQUiF6AL4ISYWQCtED8EVIKoRUiB6AL0JSIaRC9AB8EZIKIRWiB+DraUjbZUqL\n7Xhws0zL8VBK6fL2+OY7LQ+H/Tql1c/paIu02Lx6GQgpUy+zUvQAfD0L6TONhjBW46HVoQwp\npY/DYTEeGEo6HVq8eBkIKVMvs1L0AHw9C2moZTcms01pfThe72zLkDaH/WEzvNsOV03HQ5/D\ncbevXQZCytTLrBQ9AF/PQjpev6z346HjFdJuaGpZhrQ7Hj7e/tsf9h+/46Hxax+vXQZCytTL\nrBQ9AF/PQtqON92OgRT1zA9e3h3Ot/XS9WMjQsrUy6wUPQBfTx9s+F0PVXwT0n+BkArRA/D1\np4e/v8bHDpbFTbv9NKTypt3LCClTL7NS9AB8/eHBht3pwYbP64MNxztOX8OjCteQxgcbPofI\nxscZfriPZEJIhegB+HoW0sfpdtrnYfrw92Z8gHsa0uTh79PRFrvXLgMhZeplVooegK/nf5A9\nhrH4HA9+Xv4gO1wxrXezkKZ/kF0OX33xMhBSpl5mpegB+OIpQiqEVIgegC9CUiGkQvQAfBGS\nCiEVogfgi5BUCKkQPQBfhKRCSIXoAfgiJBVCKkQPwBchqRBSIXoAvghJhZAK0QPwRUgqhFSI\nHoAvQlIhpEL0AHwRkgohFaIH4OtZSP/vMdfLQEiZepmVogfgi5BUCKkQPQBfhKRCSIXoAfgi\nJBVCKkQPwBchqRBSIXoAvghJhZAK0QPwRUgqhFSIHoAvQlIhpEL0AHwRkgohFaIH4MsjpOlr\nQs5eHvL6cpFPXziyjpBSfjd9mcvJxf6Ll758hpAKfr9aBYeQ0vVtmrw9zHZz9olCTSGlw+zy\nTH+2ySeenoTxByKkguWXVq9/D2m6Pen2U4emQpokMz2UbJEQ0r+w/NLq5RfS9GAqu2okpPJa\n597Pdvnw8n+IGj9KDz8aj5hmL5Z+QkgFjyHq/GchXd/duaqaqyakm0v5JKT8U6Xi/WHyE6fL\nEdPsW/43unch1Mus9I/zE3MK6fwf4dnnmnuwIV3Kz5e2/NkmjzxM3zwN6ebYZ4RU8Bulgut9\npPvXSLOv3FNJSNP1v3zu+v72ft8fQpq1OH8zIqTCP85PjJAuXnuwwXaNdHuyhPTQX0+uCrGP\n2t1XUUizBggp1F9Prgqxf0e6r6aQigs7/dmmx+I+kr9/GZ7ef/bMhuv9hPl/5O+oKqRDMj2z\nIeVPTd7nn3328PftmxEhFbwGqcFz7VQIqRA9AF+EpEJIhegB+CIkFUIqRA/AFyGpEFIhegC+\nCEmFkArRA/BFSCqEVIgegC9CUiGkQvQAfPHa3yqEVIgegC9CUiGkQvQAfD0LKeqHJqQsbGsr\nFD0AX4SkQkiF6AH4IiQVQipED8AXIakQUiF6AL4ISYWQCtED8EVIKoRUiB6AL0JSIaRC9AB8\nEZIKIRWiB+CLkFQIqRA9AF//HtLktQ7+8sXnCSkL29oKRQ/Al1dIxQszPHzRkHsIKQvb2gpF\nD8CXU0jm14O7h5CysK2tUPQAfPmEVF7r3Fw3PUVIWdjWVih6AL5cQipeF5KQLAipED0AXx4h\nvfLi8/cQUha2tRWKHoAvh5DmL4N4mB7iPtJjhFSIHoAvHmxQIaRC9AB8/TcPfxPSnxFSIXoA\nvlz/IFt+kr8jPUFIhegB+HILyfji8/cQUha2tRWKHoAvnmunQkiF6AH4IiQVQipED8AXIakQ\nUiF6AL4ISYWQCtED8EVIKoRUiB6AL0JSIaRC9AB88ZLFKoRUiB6AL0JSIaRC9AB8EZIKIRWi\nB+CLkFQIqRA9AF+EpEJIhegB+CIkFUIqRA/AFyGpEFIhegC+CEmFkArRA/BFSCqEVIgegC9C\nUiGkQvQAfBGSCiEVogfgi5BUCKkQPQBfhKRCSIXoAfgiJBVCKkQPwBchqRBSIXoAvghJhZAK\n0QPwRUgqhFSIHoAvQlIhpEL0AHwRkgohFaIH4IuQVAipED0AX4SkQkiF6AH4IiQVQipED8AX\nIakQUiF6AL4ISYWQCtED8EVIKoRUiB6AL0JSIaRC9AB8EZIKIRWiB+CLkFQIqRA9AF+EpEJI\nhegB+CIkFUIqRA/AFyGpEFIhegC+CEmFkArRA/BFSCqEVIgegC9CUiGkQvQAfBGSCiEVogfg\ni5BUCKkQPQBfhKRCSIXoAfgiJBVCKkQPwBchqRBSIXoAvghJhZAK0QPwRUgqhFSIHoAvQlIh\npEL0AHwRkgohFaIH4IuQVAipED0AX4SkQkiF6AH4IiQVQipED8AXIakQUiF6AL4ISYWQCtED\n8EVIKoRUiB6AL0JSIaRC9AB8EZIKIRWiB+CLkFRq+M3DTQ3jJCQ0r4ZxEhKaV8M4CQnNq2Gc\nhITm1TBOQkLzahgnIaF5NYyTkNC8GsZJSGheDeMkJDSvhnESEppXwzgJCc2rYZyEhObVME5C\nQvNqGCchoXk1jJOQ0LwaxklIaF4N4yQkNK+GcRISmlfDOAkJzathnISE5tUwTkJC82oYJyGh\neTWMk5DQvBrGSUhoXg3jJCQ0r4ZxEhKaV8M4CQnNq2GchITm1TBOQkLzahgnIaF5NYyTkNC8\nGsZJSGheDeMkJDSvhnESEppXwzgJCc2rYZyEhObVME5CQvNqGCchoXk1jJOQ0LwaxklIaF4N\n4yQkNK+GcRISmlfDOAkJzathnISE5tUwTkJC82oYJyGheTWMk5Cy//MX/XO9KUJSIaSuEJIK\nIXWFkFQIqSuEpEJIXSEkFULqCiGpEFJXCEmFkLpCSCqE1BVCUiGkrhCSCiF1hZBUCKkrhKRC\nSF0hJBVC6gohqRBSVwhJhZC6QkgqhNQVQlIhpK4QkgohdYWQVAipK4SkQkhdISQVQuoKIakQ\nUlcISYWQukJIKoTUFUJSIaSuEJIKIXWFkFQIqSuEpEJIXSEkFULqCiGpEFJXCEmFkLpCSCqE\n1BVCUiGkrhCSCiF1hZBUCKkrhKRCSF0hJBVC6gohqRBSVwhJhZC6QkgqhNQVQlIhpK4Qkgoh\ndYWQVAipK4SkQkhdISQVQuoKIakQUlcISYWQukJIKoTUFUJSIaSuEJIKIXWFkFQIqSuEpEJI\nXSEkFULqCiGpEFJXCEmFkLpCSCqE1BVCUiGkrhCSCiF1hZBUCKkrhKRCSF0hJBVC6gohqRBS\nVwhJhZC6QkgqhNQVQlIhpK4QkgohdYWQVAipK4SkQkhdISQVQuoKIakQUlcISYWQukJIKoTU\nFUJSIaSuEJIKIXWFkFQIqSuEpEJIXSEkFULqCiGpEFJXCEmFkLpCSCqE1BVCUiGkrhCSCiF1\nhZBUCKkrhKRCSF0hJBVC6gohqRBSVwhJhZC6QkgqhNSV9w0pFe9f/fqD46b7X7lFSF1545DS\n9N2dr79yWne/5/kpEFJXCImQ4OCNQxrP9/w2pfMHaXxz+eD0Ph+8HO8wPZxSPo3LTbvTl9Lk\nK9MTPCOkrhDS5X5NOl875XpyRdOD+dLm7zl/y+T97EvlqQ8f/G9070IRUqsIqdj8mw/ufj2f\nxsOjH25Dmn83IXXlnUPKFZlDSre37QgJA0J69Rpp8v1cI+GCkP4ypDtHI6Q39tYh3V/120/e\nSyGdj3rn/fTohPQmCOlUwPUhtcnD3mVI5cPfl89N309Prnz4+0BI3XrfkB6JujSE1BVCKhES\n/gIhlQgJf4GQVAipK4SkQkhdISQVQuoKIakQUlcISYWQukJIKoTUFUJSIaSuEJIKIXWFkFQI\nqSuEpEJIXSEkFULqCiGpEFJXCEmFkLpCSCqE1BVCUiGkrhCSCiF1hZBUCKkrhKRCSF0hJBVC\n6gohqRBSVwhJhZC6QkgqhNQVQlIhpK4QkgohdYWQVAipK4SkQkhdISQVQuoKIakQUlcISYWQ\nukJIKoTUFUJSIaSuEJIKIXWFkFQIqSuEpEJIXSEkFULqCiGpEFJXCEmFkLpCSCqE1BVCUiGk\nrhCSCiF1hZBUCKkrhKRCSF0hJBVC6gohqRBSVwhJhZC6QkgqhNQVQlIhpK4QkgohdYWQVAip\nK4SkQkhdISQVQuoKIakQUlcISYWQukJIKoTUFUJSIaSuEJIKIXWFkFQIqSuEpEJIXSEkFULq\nCiGpEFJXCEmFkLpCSCqE1BVCUiGkrhCSCiF1hZBUCKkrhKRCSF0hJBVC6gohqRBSVwhJhZC6\nQkgqhNQVQlIhpK4QkgohdYWQVAipK4SkQkhdISQVQuoKIakQUlcISYWQukJIKoTUFUJSIaSu\nEJIKIXWFkFQIqSuEpEJIXSEkFULqCiGp1PCbh5saxklIaF4N4yQkNK+GcRISmlfDOAkJzath\nnISE5tUwTkJC82oYJyGheTWMk5DQvBrGSUhoXg3jJCQ0r4ZxEhKaV8M4CQnNq2GchITm1TBO\nQkLzahgnIaF5NYyTkNC8GsZJSGheDeMkJDSvhnESEppXwzgJCc2rYZyEhObVME5CQvNqGCch\noXk1jJOQ0LwaxklIaF4N4yQkNK+GcRISmlfDOAkJzathnISE5tUwTkJC82oYJyGheTWMk5DQ\nvBrGSUhoXg3jJCQ0r4ZxEhKaV8M4CQnNq2GchITm1TBOQkLzahgnIaF5NYyTkNC8GsZJSGhe\nDeMkJDSvhnESEppXwzgJCc2rYZyEhObVME5CQvNqGOebhoSuqPfp8K4h3SUZh+JM3+YHjURI\n2dvs19v8oJEIKXub/XqbHzQSIQEOCAlwQEiAA0ICHBAS4ICQAAeEBDggJMABIR1+Vil9/ASf\n6W6d0ir6TA+HVfy8t8u03IafazhC2qfRb+iZ7k5nGl3SNoXPez3+oP2XREjbtBnG/Rl6puvh\nTDdpGXqmh98UHtJvWu0PX9E/qAAhfQxXRr/pI/RMF+PvPXqtl4vwkDbpK/gcRQhJs9Oj4P9Q\nf6av8B9zlXxRTAsAAANPSURBVHbB5yhCSEkW0nfsXYfhWjf8xzye4UdafgefqwAhyULaLVah\n57dY7BUhrRSPqsQjJFVI0R2t07fgx0zDoyrbFPujKhDSOaRF8Nn+Bnd0SBeh5yq8Bxqr/5/w\nT5aKR+0OX9FnKArpg5DexSath9s9sX8y/BHd2Anf6K/TTbt18NnGI6Tf03+n96FnulJcOxwU\nVw2nn7T/x8AJaXyuXfTT3iQ3sw6S21ibRVrFPv9KgpAAB4QEOCAkwAEhAQ4ICXBASIADQgIc\nEFIDnv31Z3P62vSFJ77NzyN4cMLrN/hnD94IqQFPQtqcvjZ94Ykf+/NvH53wov9/9uCNkBrw\nOKT1+dkR0xeeWNifNvjohLfRz4VvHyE1YNz34QW81uNz1nartPweP7dY/JxamLzwxGZxOP/D\n1NX0GW77879rH57s/vOR0mJzPuHJv8faD2cxPulw/CpeQEgNGLZ8vxhuuy32l0Pj5m8uHVz/\n2c8+nQpZn4+dfYzPHN0de/o+ff/mJqTxhMfeNsFP4m0fITVg2PLN8A8vVsP2fx4P7VeXW2XT\nf+A7vP0c/iHs8C+896djZ9/jR5vjl5fDK/v8XiKafuv4GmHb8cixL0/WPkJqwLDly+EKZbg+\nuRx6ENLpimd83GE3f5Wi5XCbbzHe+9l9f65uQ1qeDg23D3fh/+6wdYRUs/NNuPm+z19kogyp\n/Phqm34OP+M1zSo9OOHrv+14g3/T6ovfV81eDWnxLKR9Wp/u+6zTcvu9IyRX/L4a8OebdtcX\nnrh0MB5n/u/Z1+l0i+30CN0spN31pt30ZGHG76sB8wcbzodmIV1feOJyH2l8QGL+iMHP+QXm\nhrf7fB9pkb7OH22GU/8a6+M+0qsIqQEPH/4+XEK6vvBEftRuOHZxOsvLY9vT+0jjR5/Xsxif\nHsGjdq8ipAbc+YPs1zyk6wtPXP6OdDzSunzJke35Fe3H/zdTvlm3WRyryWdxevUK/o70KkJq\n1MNXtByfxfqvd3ESz2x4ESE1Z7yjs3n8UnHDc+3+MSSea/cyQmrO+S7Ow5eK+1lMQsqvr/rS\noHn298sIqT3bZbq9/zPxvf7HkPj3SK8jJMABIQEOCAlwQEiAA0ICHBAS4ICQAAeEBDj4/3rc\nrAWz03uvAAAAAElFTkSuQmCC",
      "text/plain": [
       "plot without title"
      ]
     },
     "metadata": {
      "image/png": {
       "height": 420,
       "width": 420
      }
     },
     "output_type": "display_data"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0gAAANICAMAAADKOT/pAAAAOVBMVEUAAAAzMzNGeLpNTU1o\naGh8fHyMjIyampqnp6eysrK9vb3Hx8fQ0NDZ2dnh4eHp6envg1Tw8PD///9x/L4bAAAACXBI\nWXMAABJ0AAASdAHeZh94AAAgAElEQVR4nO3di3aiSrdA4dpHjcZc/OX9H/Zwr8sqkIUoq9Lz\nG2PHGyKprtkgsbNdBeBpbu8NAP4CQgI2QEjABggJ2AAhARsgJGADhARsgJCADRASsAFCAjZA\nSOYcPq639srt+nFoLl3reLmPS7jD/LKnn+7h69Eduqfdz86df4cVXJM/9h93HK8f3Xe9+OVY\nr+X6cNOq6vd8cOfv7robJK9yOQybEW18wBU/D4v/Bv6eeiKe2yvnfkoO0/PQTeLqu77+/WDZ\ntplL97RmCh/8vfXkF/P24Pr0qlszz+8H5586t2n9S7jjLXxxF7/KqVtEbHy04sXDY1Tx38Df\nU8+57u/sw3GIo/l6O7lTt8DZXfr5PLXspV30153vzX7h3NzRfvlol/09iHn76S79tUtz7exO\nt/YVL/Fi4uU+3aHZf312kUerHV/lxx1+m1s/6cZHK142NnYV/w38Pa6eas2u47e+DOLoD7qa\n24d6DzK/bHvx4a8f3H188OpOYt7e/cGiuzULdseD6XLpy92GveS5jSNc3L/Kpd3sL/eZbny0\n4gUDY1rx38DfUx/6uObdydV9xSF9d3+Vf9X7iYv7ml02fB/iJ2l3b/10OW9PfaQ/7b5sal6n\nL3fp66juH9fkaf5VPpo06/Y+5MaH23hxh8863e4IsLns7uv3ifW7vaN4y2YJIZlT7w7aSVdP\nwDikfpad6oOkn/4wb2LZi/OT7j4cEQ73/uZC+R7f/Hy3S55vSzbt5H7jh/3133jvOFzEGx88\n86N5H3Wt19weATb7L+c+2xMn3dPGq0YRkjn1/Do2fyz1/iMOqbvSHYV1x2q5ZVvBe5trv6/5\nCu7N7HEO0aw/NWcJf8RC6csl64lP2uVCijY+eubpXm/pcQj6XO/FXP/e6qvZ9vGqVYRkTntI\n0/y9fc6G9NX20B8eZZbt/u4e9xS3Q3eGobp+HIYDsVxI3Qq/hti+m/Nyh/T0Wvpy6pCijY+e\n+dMvcuzezR3b48hmQ5pd4Ed/1fAuiZDMqadTc2TzWc+2XEjHds79dod5E8t+j6ez74dg8p2H\nI75MSL/tLA2P1X4+/UnxiU0TIeVuhhfRxotFm6/Xpvaf5kvwvHhNJhnetH9VPV1u9aw+tUc3\n/R2t5t76y+A2s+zv8Jf3KZyy47m53IxsdgX3eIaLCZ++3MfY3bc/Kxgu3QgOGuONF4s2X9tt\n/OxOHY4PERL0+tPVzYSKQ2qPiz7Hufg5t2z/s6fj6ZauOrwMXdtdzTV6WJ7+jl/uczhY7D4a\nkQ+pO2t3aw7R4o0Xi/ZnAr+rY7A2QsI6zXTpf2yZ/hzpp/nSf0hnnLq5Zbt9j39T0b29vw17\nmNyMbPZG3RuUeupfg7XMbNr4c6STS09/+5uf/RucS7rxYtH2a70z/e3/kvip+tOJH/7tklWE\nZE4znb7qv7S/4ji6Tzb8jpOpfTszsey9/UzCzb85b+f+/WPmPVJ75rs/B/7j3LUu6ufkkh/d\niJc7tyckbh/9T6iyIflPNiQbLxbtvh7doT9s7Z73zVk7rDG8mbj5+eU/a3cZP6jW/hU/vey9\n+0DccBbtEP0gJhtS8Cm4y3Dy79Gm9T/gGT9rJ5ZuHId1ZTY+WrQ/U+KGEyntyj/8yxg+aUdI\n9vQT/1DFcZzatxUHf7B1OEwu233WOvoM6eUQfDQg/2Yj+Fx285ludxI7APFy9c7iwy84EdK9\n/fT3xMaHi/a70/4dVH3rY9zm64FPNgAa3/05PcunFqSiNhb/guGdGSGhDM5FB39PLLTtRp2G\na295wY0UtbHYlMmQDuOZPUIC/jmEBGyAkIANEBKwAUICNkBIwAYICdgAIRnzH/T2/kOrCMkc\nC5OiNBbGjJCMsTApSmNhzAjJGAuTojQWxoyQjLEwKUpjYcwIyRgLk6I0FsaMkIyxMClKY2HM\nCMkYC5OiNBbGjJCMsTApSmNhzAjJGAuTojQWxoyQjLEwKUpjYcwIyRgLk6I0FsaMkIyxMClK\nY2HMCMmYeFL8D1lzY7YPQjKGkJaYG7N9EJIxhLTE3Jjtg5CMIaQl5sZsH4RkDCEtMTdm+yAk\nYwhpibkx2wchGUNIS8yN2T4IyRhCWmJuzPZBSMYQ0hJzY7YPQjKGkJaYG7N9EJIxhLTE3Jjt\ng5CMIaQl5sZsH4RkDCEtMTdm+yAkYwhpibkx2wchGUNIS8yN2T4IyRhCWmJuzPZBSMYQ0hJz\nY7YPQjKGkJaYG7N9EJIxhLTE3Jjtg5CMIaQl5sZsH4RkDCEtMTdm+yAkYwhpibkx2wchGUNI\nS8yN2T4I6QE3Xrhee3O4El3dAiEtMTdm+yCkB9z41eXuDK/Or2LhWBPSEnNjtg9CesDFX6rw\nmlsWCSFtbm7M9kFID/h9kQypEuPnKtcfBLa33OStdsHmS3eXXwMhLRGPOiEVwO91FoU0dueS\nyypo0g0Luugp/7Wi9e09Ya2KR52QCtDuN7pr48mGcAfjr3ZLVwtDEkv3CGmJ+M+IkAoQ7WOG\n+/xl8v7ncUhRi/GXFiEtEf8ZEVIBdCcblu2R5GoJSSf+MyKkAuROfxPS3uI/I0IqgC9EnGJw\nydVqaUgc2j0r/jMipAL4CpZ8ssGflgguh8fj09/yS4uQloj/jAgJAiEtMTdm+yAkYwhpibkx\n2wchGUNIS8yN2T4IyRhCWmJuzPZBSMYQ0hJzY7YPQjKGkJaYG7N9EJIxhLTE3Jjtg5CMIaQl\n5sZsH4RkDCEtMTdm+yAkYwhpibkx2wchGUNIS8yN2T4IyRhCWmJuzPZBSMYQ0hJzY7YPQjKG\nkJaYG7N9EJIxhLTE3Jjtg5CMIaQl5sZsH4RkDCEtMTdm+yAkYwhpibkx2wchGUNIS8yN2T4I\nyRhCWmJuzPZBSMZYmBSlsTBmhGSMhUlRGgtjRkjGWJgUpbEwZoRkjIVJURoLY0ZIxliYFKWx\nMGaEZIyFSVEaC2NGSMZYmBSlsTBmhGSMhUlRGgtjRkjGWJgUpbEwZoRkjIVJURoLY0ZIxliY\nFKWxMGaEZIyFSVEaC2NGSMZYmBSlsTBmhGRMqZ/+3mm4WoQEgZD0CAkCIekREgRC0iMkCISk\nR0gQCEmPkCAQkh4hQSAkPUKCQEh6hASBkPQICQIh6RESBELSIyQIhKRHSBAISY+QIBCSHiFB\nICQ9QoJASHqEBIGQ9AgJAiHpERIEQtIjJAiEpEdIEAhJj5AgEJIeIVniOum9unU8vxmEpEdI\nlrjoYrj69vEhJD1CsoSQnvLucQoRkiVBSN0hXvvVVd31cQk3PlxFN8dbVXB8ODzUX46PZe4f\nnkJIeoRkiQ/JBZfDRE++Dhf+5vhkF92MV+jkC4z31/OhFW3U3nkst/mfhwIhWZKcbBhC8tfd\nOPejh8MRjPZdyeCmAVXydoOQ9AjJkiiI8eAtG1L8cHCmbzIkFx0guuBeR0hPIyRL5FmGOKTw\n6K4KblVBSlMh5Q7pKnnZICQ9QrLExdemQxpTcXEp1WRI4j1RlXm8R0h6hGRJHFJ6aDeeW8ic\ng6hmQ3LxCrMnG4LNICQ9QrIkPLRz3a7HuVxI8f2VOP0drCs4zR3vy9LT3x4h6RFS4V4xeISk\nR0iFIyTvBUOxGCEVTXzAdROEpEdIEAhJj5AgEJIeIUEgJD1CgkBIeoQEgZD0CAkCIekREgRC\n0iMkCISkR0gQCEmPkCAQkh4hQSAkPUKCQEh6hASBkPQICQIh6RESBELSIyQIhKRHSBAsTIrS\nWBgzQjLGwqQojYUxIyRjLEyK0lgYM0IyxsKkKI2FMSMkYyxMitJYGDNCMsbCpCiNhTEjJGMs\nTIrSWBgzQjLGwqQojYUxIyRjLEyK0lgYM0IyxsKkKI2FMSMkYyxMitJYGDNCMsbCpCiNhTEj\nJGMsTIrSWBgzQjLGwqQojYUxIyRj+GcUeoQEgZD0CAkCIekREgRC0iMkCISkR0gQCEmPkCAQ\nkh4hQSAkPUKCQEh6hASBkPQICQIh6RESBELSIyQIhKRHSBAISY+QIBCSHiFBICQ9QoJASHqE\nBIGQ9AgJAiHpERIEQtIjJAiEpEdIEAhJj5BeynWq7psMrlbRt+3SO6J1zKxeNXiLFyUkPUJ6\nKd/MMOddJhs3PpQ8sHD1C5ckpBcipJeaDslFSxHSFhaPxgsQ0ktFIYVXXbyUD2k4FKyiI8Hm\nWvvM/kH/UHKvf8VoSdc91F33rzU+HG4sIa1BSC81FZJLFwr+c+Fz/eJdD8Ei40PhvT5CseRw\nX/p1uGi//teKvoe981hugz+w1QjppeKTDZUb53J8ZBdFFIVUiRxkkrlHM0vG+c3tJAlJj5Be\nyolrwR4kemSY3cEjLnikSlIZD8kehuQP3rIhJQ+3CEmPkF4qH9Jw279/6a8kEzp7gJbswB6F\nlO7qxmBd+HC8jyQkPUJ6qSikaMIHO57wcZcMRz6VaN8yG5KL152GFPRGSE8hpJeaCSm58HsI\nFz7Xp+LSLLIh5U42pId2QcnOL0pITyGkl4oP7eJPNiR7Jn8Kbe70t7iduTdazbCkc7mQ4vsH\nhKRHSMVYNkzJfm7V2glJj5BKIPYZD5cipDcjpCL4jzssXEo5pvHaCUmPkCAQkh4hQSAkPUKC\nQEh6hASBkPQICQIh6RESBELSIyQIhKRHSBAISY+QIBCSHiFBICQ9QoJASHqEBIGQ9AgJAiHp\nERIEQtIjJAiEpEdIEAhJj5AgEJIeIUGwMClKY2HMCMkYC5OiNBbGjJCMsTApSmNhzAjJGAuT\nojQWxoyQjLEwKUpjYcwIyRgLk6I0FsaMkIyxMClKY2HMCMkYC5OiNBbGjJCMsTApSmNhzAjJ\nGAuTojQWxoyQjLEwKUpjYcwIyRgLk6I0FsaMkIyxMClKY2HMCMmYMj/9vdNg9QgJAiHpERIE\nQtIjJAiEpEdIEAhJj5AgEJIeIUEgJD1CgkBIeoQEgZD0CAkCIekREgRC0iMkCISkR0gQCEmP\nkCAQkh4hQSAkPUKCQEh6hASBkPQICQIh6RESBELSIyQIhKRHSBAISY+QIBCSHiFBICQ9Qnof\n1xtvtJfDg8+vfrPlCUmPkN4n+D59RGNXK9e06vEHCxCSHiG9jxNXHSFt5+GovRQhvY/LXKuv\nuvFrf9sFB37xZfPg+Ghw37DS6cfjFbVfnVjvgJD0COl9loXkhsdd5tL529F9jx7PrTCz3no+\ntKLN3juQpTb4E3oCIb2PP9kwu0eqknmf3k5D8WuafzxdYSbUHiHpEdL7TOyRuoriZpLgottL\nQ4rXkaxQLk9ITyGk93kQUtxKNXmp2SONly5emJC2Rkjv48RVN7VHSp7xdEi5Y0VC2hIhvU8Y\n0jh1w//GhXLvidrLIADVeyTXv2ayACFthpDeZ/KTDZkjsdnT3+FJ7nxI4vH0fHd6+rsipOcQ\nUkneNFKEpEdIJSGkGe8ZmymEVBJCmvGesZlCSBAISY+QIBCSHiFBICQ9QoJASHqEBIGQ9AgJ\nAiHpERIEQtIjJAiEpEdIEAhJj5AgEJIeIUEgJD1CgkBIeoQEgZD0CAkCIekREgRC0iMkCISk\nR0gQLEyK0lgYM0IyxsKkKI2FMSMkYyxMitJYGDNCMsbCpCiNhTEjJGMsTIrSWBgzQjLGwqQo\njYUxIyRjLEyK0lgYM0IyxsKkKI2FMSMkYyxMitJYGDNCMsbCpCiNhTEjJGMsTIrSWBgzQjLG\nwqQojYUxIyRjLEyK0lgYM0IyxsKkKI2FMSMkY6z/M4qdhmUWIUEgJD1CgkBIeoQEgZD0CAkC\nIekREgRC0iMkCISkR0gQCEmPkCAQkh4hQSAkPUKCQEh6hASBkPQICQIh6RESBELSIyQIhKRH\nSBAISY+QIBCSHiFBICQ9QoJASHqEBIGQ9AgJAiHpERIEQtIrKKRhMdcYr/TXxc1HSz9+badK\nfNmiuaXi15lYQm166x+vjZD0ygnJ9cv5rxNNuEVLb7VZq1c/8TrpetauN/+8ZWsjJL3iQgpy\nWRDS9NJbbdawJCG9z8ohealiQ5LPmwsps3R3lOe6q11pwUFgNcQX3jusI1nSdQ9114MtGB4O\nXj5cYRW+dPCquXWPD4fbnjwlfn681om1pd9yj5D0ignJBf/ln+cyS04uPc71Yb51/427rvGh\neFb2T5ZLDvelX+Uxpgu2a3zpML7MusNA44f8U5x4vsssmntmuPb/WtFQ7Z2NVBlUZkj+79eJ\nkw0Plw6qSydbcP/Eo5kl4/xc/uH5JPJb4+IlkldO1ij/GkiGMFOpWIw90hqlhOT/Qvd3uMwi\n6V//00vHy8Tz0bl4Kk6G5A/esiElD8+HFBwIjt1vF5JcW/jM+GQmIekVE1I4sfonufDBqKHZ\npf2zqolZVaVTcSqkYA1+ue6WixZJtyMTUm5rkruT8dKEFG9p5pviPdJziglp+DqdRhjS46X9\nvZkpuDCktIMkpKC3xyFlt2J8/rMhJVua+6aClyAkvUJCEkG4SjzPJYvMLp1MYZfMrvyhXe5k\nQ3pol7xicA4i3KIgJJdsRWauL+jY3/BPyH9Pc3939AhJr7yQpj/ZMF4uWjpz+lvcztzbv0J0\nSnx8KA4pvn+4O94LRKe/u6vJue0lp7+DL9Onv8MtTZ/J6e9nFRLSyy3bBhddbL32TWzxUoSk\nR0jpcc2CpQhpXxt8i5sjpPS4ZsFSyg1esvbNENI+CAkCIekREgRC0iMkCISkR0gQCEmPkCAQ\nkh4hQSAkPUKCQEh6hASBkPQICQIh6RESBELSIyQIhKRHSBAISY+QIBCSHiFBICQ9QoJASHqE\nBIGQ9AgJAiHpERIEC5OiNBbGjJCMsTApSmNhzAjJGAuTojQWxoyQjLEwKUpjYcwIyRgLk6I0\nFsaMkIyxMClKMztm16Nzh2t79XJ0x/aa/0W6zZdvd6yq+9m500+32MEdLtptICRjCElvbsw+\nu1+W3YRxaq+dqjQk5z6q6tBeaUrqrh2U20BIxhCS3tyYNbXc2mSuzp2rer9zTUO6VPfq0lxc\nm11Tfe2zWfaq2wZCMoaQ9ObGrN6/nO/ttXqHdGuaOqYh3err9fHfvbp//LbX2sc+dNtASMYQ\nkt7cmF3bQ7c6kKSe+GrwP+ke/s8p2t+NvWK78UKEpDc7Zr/npopvQvrHEJLeozH7as8dHJND\nu3sYUnpop0ZIxhCS3oOTDbfuZMOnP9lQv3H6as4q+JDakw2fTWTteYYf3iOVzsqnv3f69leZ\nC+mjO077rMLT35f2BHcYUnD6u1vscNNtAyEZQ0h68z+QrcM4fLZXP4cfyDY7pvMtCin8geyx\neVS5DYRkDCHpWTgcJiRjCEmPkCAQkh4hQSAkPUKCQEh6hASBkPQICQIh6RESBELSIyQIhKRH\nSBAISY+QIBCSHiFBICS9uZD+b9qm20BIxhCSHiFBICQ9QoJASHqEBIGQ9AgJAiHpERIEQtIj\nJAiEpEdIEAhJj5AgEJLeFiGFvxMy+vWQ/tdFzv7iSEJa5OEv38w95KaHd3pNhKS3QUjOf3XB\n1/GR4A8z/2dHSIs8HCZlSNMISe/5kMJqnLyrIqSNEJJl24UUXnVpV4T0vPiQOXOZHl43F+3Y\nTy8+Xo2OGQlJ72Uh+YvMripGSIvEo+smLuOb3e/xnFx8vDr+Gf3Xil6XkJbYKKTubzYREicb\nNuQHs6rkMVtye6qzYLHMQz1C0tv0PVJ+j/TwOJ2QFsnvkbp7kr/H/M1otzP+rValD0XrJ6QV\nCKkY0yG5+Pb0odz4hIlFeoSk996zdnmEtMhkSDPvkSZCEu+RovUT0grv/TlSHiEtkg+pv3Rx\nSC69O39JSJt52Scb3PBAdGY2i5AWSd8SVfI89/jwMPxRJXOnvytCeg6ftYNASHqEBIGQ9AgJ\nAiHpERIEQtIjJAiEpEdIEAhJj5AgEJIev/sbAiHpERIEQtKbC+ld3yIhGUNIeoQEgZD0CAkC\nIekREgRC0iMkCISkR0gQCEmPkCAQkh4hQSAkPUKCQEh6z4cU/GKG4LfXBf/8/MGvfickcwhJ\nb6uQkl/MEP5ClOCOPEIyhpD0Ngop+e0Zwa8RSn6jUA4hGWPhA5il2SakdK8j9k2zCMkYQtLb\nJKTk90ISUuEISW+LkNywRxpPNgS/Pi36BWx5hGQMIeltEFLmNwzyHqlohKTHyQYIhKT3mtPf\nhFQ0QtLb9Aey6Z38HKlMhKS3WUiV45MNfwUh6fFZOwiEpEdIEAhJj5AgEJIeIUEgJD1CgrD/\np793+safQEgQCEnPwl6ckIwhJD1CgkBIeoQEgZD0CAkCIekREgRC0iMkCISkR0gQCEmPkCAQ\nkh4hQSAkPUKCQEh6hASBkPQICQIh6RESBELSIyQIhKRHSBAISY+QIBCSHiFBICQ9QoJASHqE\nBIGQ9AgJAiHpERIEQtIjpD9A8X9FzCwq7yIkPUIqn1swgqoxJiQ9QiofIRlASOVzw/+Iqvuf\n9vr/r05w2T/Y/Q/h0ofS/8kvIekRUvF8DOMX//94Gy/9/eKh8P8I918rWj8hLUFIxXPBf7lK\n5OXk1R4h6RFS6cb/tWg+pP7/lxjf7/zeK9ibjQhJj5BK59JWpvZA4+1x9zUe8FVxSoSkR0il\ne7BHCpdK3xBlFmkRkh4hFc5nMP8eKT7ZkBztRW0R0hqEVLiJkB6c/g4X4fT3FggJAiHpERIE\nQtIjJAiEpEdIEAhJj5AgEJIeIUEgJD1CgkBIeoQEgZD0CAkCIekREgRC0iMkCISkR0gQCEmP\nkCAQkh4hQSAkPUKCQEh6hASBkPQICQIh6RESBELSIyQIhKRHSBAsTIrSWBgzQjLGwqQojYUx\nIyRjLEyK0lgYM0IyxsKkKI2FMSMkYyxMitJYGDNCMsbCpCiNhTEjJGMsTIrSWBgzQjLGwqQo\njYUxIyRjLEyK0lgYM0IyxsKkKI2FMSMkYyxMitJYGDNCMsbCpCiNhTEjJGMsTIrSWBgzQjLG\nwqQojYUxIyRjdv5nFDt9188hJAiEpEdIEAhJj5AgEJIeIUEgJD1CgkBIeoQEgZD0CAkCIekR\nEgRC0iMkCISkR0gQCEmPkCAQkh4hQSAkPUKCQEh6hASBkPQICQIh6RESBELSIyQIhKRHSBAI\nSY+QIBCSHiFBICQ9QoJASHqEtDNXm35wcnD8Iy8YPULSI6R9ufHLxKOEVAZC2tWDGAipGIS0\nKxdca4/x+gO9/oAvCMlNPeL8bs0FT59+Ynw7vH/YFkLSI6Rd+XdI3WwOG3FivzP1yHhn8HQ3\n9UT5Ev7+ej60om0kpCUIaXfjLiZ5xyRDqh6FlFyGT8wuW8nbDULSIyQLnAip21VlQnKZR5aF\nND5RHDI6QnoaIVkgQvLHX7k9UrUmpNwhXWbFDULSI6RdyRjm3iNVT4SUXSaz4gYh6RHSrly0\nD4pmeebQbiqxYPHo9sQqp+PrEZIeIe1sPAsdfenPXyfvZYLLODG/eHQ7fWK6TLLCESHpEdJf\n9cSoEpIeIf1B4XukNQhJj5D+otlPwj5GSHqEBIGQ9AgJAiHpERIEQtIjJAiEpEdIEAhJj5Ag\nEJIeIUEgJD1CgkBIeoQEgZD0CAkCIekREgRC0iMkCISkR0gQCEmPkCAQkh4hQSAkPUKCQEh6\nhASBkPQICYKFSVEaC2NGSMZYmBSlsTBmhGSMhUlRGgtjRkjGWJgUpbEwZoRkjIVJURoLY0ZI\nxliYFKWxMGaEZIyFSVEaC2NGSMZYmBSlsTBmhGSMhUlRGgtjRkjGWJgUpbEwZoRkjIVJURoL\nY0ZIxliYFKWxMGaEZIyFSVEaC2NGSMbs8unvnb7XrRASBELSIyQIhKRHSBAISY+QIBCSHiFB\nICQ9QoJASHqEBIGQ9AgJAiHpERIEQtIjJAiEpEdIEAhJj5AgEJIeIUEgJD1CgkBIeoQEgZD0\nCAkCIekREgRC0iMkCISkR0gQCEmPkCAQkh4hQSAkPUKCQEh6hPQ2rqZ8xppXmX5q8PLzayYk\nPUJ6k24W61LaNiSXv55ZkpD0COlNZvYUj56z0esQ0isR0nu44Iqr3LB/8nsoF9/XLCMWHG9X\nmZWMy4QPO//i3ePBUuEd8WEnIekR0ntEIfVT2QUPDNddtIzL3dk/IX0svBwfDkKqXPwy4RP8\nlvzXijadkJYgpPeI90jxZfS4eGzuCTMhBTeDu5KlZEgtQtIjpPfIhuT8EVVwQs0fk704JP9K\n4ZZUhLQGIb3H5B4p8/iWe6Tx8E6G5OKrAULSI6Q3mTm+Ch7ONPOqkDL7rAEh6RHSmwQ/R3rU\nhT+0S3ZP0RMmHkseng9pPLQLVlwR0hqE9Dbj2xC/t8if/g66yJzaHstbcvq7CkY3XWp4JU5/\nb4GQCrV40FaMLiHpEVKhCMkWQirUwkHTfk62RUh6hASBkPQICQIh6RESBELSIyQIhKRHSBAI\nSY+QIBCSHiFBICQ9QoJASHqEBIGQ9AgJAiHpERIEQtIjJAiEpEdIEAhJj5AgEJIeIUEgJD1C\ngkBIeoQEwcKkKI2FMSMkYyxMitJYGDNCMsbCpCiNhTEjJGMsTIrSWBgzQjLGwqQojYUxIyRj\nLEyK0lgYM0IyxsKkKI2FMSMkYyxMitJYGDNCMsbCpCiNhTEjJGMsTIrSWBgzQjLGwqQojYUx\nIyRjLEyK0lgYM0IyxsKkKI2FMSMkYyxMitJYGDNCMuZ9/4xip2/wBQgJAiHpERIEQtIjJAiE\npEdIEAhJj5AgEJIeIUEgJD1CgkBIeoQEgZD0CAkCIekREgRC0iMkCISkR0gQCEmPkCAQkh4h\nQSAkPUKCQEh6hASBkPQICQIh6RESBELSIyQIhKRHSBAISY+QIBCSXiEhudqDRZ2qyA3jdbM3\nH73wBhvinlhHdtQISa+MkB5FtGwt4ZKvCWnJXwmbLDK3uOrphLSRIkJyS5YjpDVPJ6SNlBRS\nf605ymv//IfDve7SVem9PsBoSdc91F0fV+/Gh6vo5riG4Phy2IDxVeM1z25E9MTsdxJ+FY+F\n4xEtOPt0/19HyQwAAAl1SURBVJ2E30SwOcH3WhHSGmWE5OfK8OfevzOI53Nyr59gYsnhvvTr\ncOFvRmuINyB6Yf/Qw43w4crFg7XkHgtDChdPb8dPH+9JnuSqZIvq+dCKRp+QligipHah3HQL\nJ5Gc0unkCZaMJ5LLPRxulasquerMRJ7ZxInspi5nXiTbRJVZJNoCJxdKRs0jJL1iQqqmUxmP\neh6G5I+PsiHFDwcHO+OjLvMS8qlbhJR7sWp8hWSN/R3OL5I8XYbkwlFLTvwRkt4fCMmFD82G\nFMxzv1x3y0WLDE9M//6ObkTzPrq6RUi5F6vCjZMPVf4gTm54GlI0puH3WhHSGkWE5CfB/PR7\nPIfnQgp6C6ZttIap1SVr3i2kyac/DilcDyHplRFSMMe7i3TaZUPKnWxID+2Gv4+r4CkuXiK6\nnq7G5db8YCPEd5B7WiXvSypPQ5p7eq7nYHOivygIaY0iQqqiTzYkZ5+D25l7u5vJeWsXnJPz\nIcX3+9cUdwzLBRsSr3luIzLfgTiRH52/9s8JNjXcHpfezj3d58bp75coJKR1T3HRxaZrX77O\nxxuh274Vo6VFSHrlhZQchixYauOQlm2AYiMI6Q8oL6TkMGTBUur1b7IByzdCs33PfEB1MULS\nKzAkvBoh6RESBELSIyQIhKRHSBAISY+QIBCSHiFBICQ9QoJASHqEBIGQ9AgJAiHpERIEQtIj\nJAiEpEdIEAhJj5AgEJIeIUEgJD1CgkBIeoQEgZD0CAkCIekREgQLk6I0FsaMkIyxMClKY2HM\nCMkYC5OiNBbGjJCMsTApSmNhzAjJGAuTojQWxoyQjLEwKUpjYcwIyRgLk6I0FsaMkIyxMClK\nY2HMCMkYC5OiNBbGjJCMsTApSmNhzAjJGAuTojQWxoyQjLEwKUpjYcwIyRgLk6I0FsaMkIx5\n7ae/d/qmXoyQIBCSHiFBICQ9QoJASHqEBIGQ9AgJAiHpERIEQtIjJAiEpEdIEAhJj5AgEJIe\nIUEgJD1CgkBIeoQEgZD0CAkCIekREgRC0iMkCISkR0gQCEmPkCAQkh4hQSAkPUKCQEh6hASB\nkPQICQIh6RESBELSI6S3cL3wLrnQxHNnVqsavMWLEpIeIb2FSy6rbb7r5etwmqUJSY+Q3oKQ\nCOn1/rGQ+kM8f9X1D7VXg+O/eMHoYf9Qcu/wSsmSrnvIv9zwJd6CHiHpEdJbBCE5fxlcHe4J\n9hzJgi54OH4ovHfIMLPkcF/6dbhov/7XiradkJYgpLcITjaEU7u/Os7rYG8xFV96/8SjmSXj\n/Fz+4RYh6RHSWwTzeigqPESbKGXcUciHx3U8DMm/Wjak5OEWIekR0lsEb0LiPVH6liVcIn+A\nFj5xQUjhYd+4XLQXDI/yOoSkR0hvkQ8pmO2ZQ7tgqTSVaN8yG1J4mQkp2YIeIekR0lsE0zuO\noAoOrKZbCBdMbmfurYZL+RLhQdy4MyKkDRDSW0THc+lZ7WFep3uk3OlvcTtzb7+26JT4+FAc\nUnz/gJD0CKkYy4bJRRfr1k5IeoRUArHPeLgUIb0ZIRUh+cDr46WUYxqvnZD0CAkCIekREgRC\n0iMkCISkR0gQCEmPkCAQkh4hQSAkPUKCQEh6hASBkPQICQIh6RESBELSIyQIhKRHSBAISY+Q\nIBCSHiFBICQ9QoJASHqEBIGQ9AgJAiHpERIEC5OiNBbGjJCMsTApSmNhzAjJGAuTojQWxoyQ\njLEwKUpjYcwIyRgLk6I0FsaMkIyxMClKY2HMCMkYC5OiNBbGjJCMsTApSmNhzAjJGAuTojQW\nxoyQjLEwKUpjYcwIyRgLk6I0FsaMkIyxMClKY2HMCMkYC5OiNBbGjJCM+Q96e/+hVYRk3BNT\n5F96qgGEZFpxU5qQYFFxU5qQAKxHSMAGCAnYACEBGyAkYAOEBGyAkIANEBKwAUIy6efk3MfP\n2mdfVv6p3s7Onda96tfRHa/rXrV2Kn8alv8d/EV31/pd9+yLW/eneutedU1JX+0z15Z0XbnB\nlpT/HfxFV3epqrP7XPXkerey7k/13LzqxR1XPPVY5/ftDqtetvpdu8GWlP8d/EUfzc7o132s\nee7h8LNyXh7ap62f1GufeTwQEl7imSl9eaaF5lXX7JEa380ObYVP90VIeAn33L7hmXn5vfad\nzse6HWi74yUkvMR+Id0Op5XPvBzXHoreCQmvsVtI6zuqmlN3K47tzu77yWNRG8r/Dv6iPqSV\nZ8HWz8vfZzpa97pu8MwLG1D69v9NxyfO2lXrQ/pa/YoHd1/5uoSE17m4c3PQs/YHnCtn5Y9b\nvT+6ND/zujZbvUrxGRGSTb/dX9L3lU9fOS9P6/cN90PzxMNt1esSEl6l+azdyk+9Vavn5TMH\nWc3H9M5rOyIkAC1CAjZASMAGCAnYACEBGyAkYAOEBGyAkP45cz+06X/bQ/grI74Xf1xhYsXn\n7+XbVi5C+ufMhNT/tofwV0b8LP/k7NSKD6t/slwQQvrnTIc0/LaH8FdGHJZ/4G9qxde1n2Iv\nCSH9c9r57j/Sczu543d73/jbHoJfGXE5dM/4cKfwA0D3/h+kNx9T//lw7nDpVxz8S6p78xLt\nxwUP6/4NelEI6Z/TzPL+Q6b34Vo788ff9uB/ZcTddYWc+6VHH67p6lb39N09/yJCalfc9nZZ\n/fHbchDSP6eZ5ZfmX0ycmtn/WV+7n4ajsvCf5jZfP5t/v1pfO927pUfdbzq51A8f3Vf7cfU0\npM/ul3td24XX/WKxkhDSP6eZ5cdmh9LsT4ZrEyF1O572vMMt/vVCx+aY79C++7l9f55kSMfu\nWnN8eFv9DwbLQUj/jv4QLp7v8a+HSENKb3tX91P9tHuak5tYsf9HGX/gn0k88ve/Qwy0IR3m\nQrq7c/fe5+yO1+8bIe29AXi3x4d2/ldGDB20y8T/EP3suiO27gxdFNLNH9qFq/3T/v53iER8\nsqG/FoXkf2XE8B6pPSERnzH46X/ffvP1Pr5HOriv/talWftXWx/vkfAHTZ7+roaQ/K+MGM/a\nNUsn6zkO57bD90jtrU//Eu3HIzhrhz8o8wPZrzgk/ysjhp8j1QuJ38hwbc57V+3nIU4/42Hd\n5VBXM75E93sn+DkS/hGTv4uy/RTrs29xVv52/aIQ0j+ufaNzmf6NdM1n7Z4Mic/a4e/r3+JM\n/iatn0MQ0vhrUVXThk9/4x9wPc7/Rrrv85Mh8e+RACxESMAGCAnYACEBGyAkYAOEBGyAkIAN\nEBKwgf8H0LltmvbD50sAAAAASUVORK5CYII=",
      "text/plain": [
       "plot without title"
      ]
     },
     "metadata": {
      "image/png": {
       "height": 420,
       "width": 420
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "### term plot\n",
    "library(ggplot2)\n",
    "\n",
    "plot_gostres_func <- function(class, gostres, height = 7, width = 7){\n",
    "    r <- gostres\n",
    "    pp <- r[(r$source %in% c(\"KEGG\")) , c(\"p_value\", \"term_name\", \"source\")]\n",
    "    order = pp$term_name[order(pp$p_value, decreasing=T)]\n",
    "    pp$term_name <- factor(pp$term_name, levels = order)\n",
    "\n",
    "    ggplot(pp, aes(x=term_name, y=-log10(p_value), fill=source)) +\n",
    "      geom_bar(stat=\"identity\", width = 0.8) +\n",
    "      theme_bw() +\n",
    "      theme(panel.grid = element_blank(),\n",
    "            plot.title = element_text(hjust = 0, size = 12),\n",
    "            axis.text.x  = element_text(angle = 0, hjust=1, vjust=0.5, \n",
    "                                        colour = \"black\", size = 11),\n",
    "            axis.text.y  = element_text(colour = \"black\"),\n",
    "            # legend.position = c(0.8,0.8),\n",
    "            strip.background = element_rect(fill=\"white\", colour = \"black\", size=1) ,\n",
    "            legend.title = element_text(face=\"bold\")) +\n",
    "      # geom_text(aes(label=symbol), vjust=0.5, hjust=0,size=4, colour=\"black\") +\n",
    "      # annotate(\"text\", x=pp$NAME[2], y=-1.5,size=3,\n",
    "      #          label=\"nominal P values: \\n*P < 0.05 \\n**P < 0.005 \\n***P < 0.0005 \\nNS: not significant\") +\n",
    "      # scale_x_discrete(limits=c(pp$term.name))+\n",
    "      scale_fill_manual(values=c(\"GO:BP\"=\"#4678BA\", \"KEGG\"=\"#EF8354\")) +\n",
    "      coord_flip() +\n",
    "      xlab(\"\") +\n",
    "      ggtitle(paste0(\"MDAMB231_VS_MCF10A: \", class))\n",
    "    # facet_grid(.~module, scales = \"free_x\", space = \"free_x\") \n",
    "\n",
    "\n",
    "#     ggsave(paste0(\"stat/\", \"MDAMB231_VS_MCF10A_\", class, \"_function_bar.pdf\"),\n",
    "#            height = height, width = width)\n",
    "\n",
    "}\n",
    "\n",
    "\n",
    "plot_gostres_func(\"hyper\", gostres.list[[\"hyper\"]], height = 7, width = 6)\n",
    "plot_gostres_func(\"hypo\", gostres.list[[\"hypo\"]], height = 2, width = 6)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0gAAANICAIAAAByhViMAAAACXBIWXMAABJ0AAASdAHeZh94\nAAAgAElEQVR4nOzdcXxb533f+x8a12mTOuSalk5qhUwzl4ptOoCcG4myVnsBuqTKdqC+UisF\ndIembUgOXO1cacR6r9ODuR3R7aY9bLWIm3BB5rZ7cVfALCevFmdXajMRvXRLi/cmkoCIskPW\nS0OmbiKsqXHqJk3cdrh/PPPZMQiAIAnygAef91/kwTkPftSrdb99nvP8Hl+1WhUAAADsf9/l\ndgEAAABoD4IdAACARxDsAAAAPIJgBwAA4BEEOwAAAI8g2AEAAHgEwQ4AAMAjCHYAAAAeQbAD\nAADwCIIdAACARxDsAAAAPIJgBwAA4BEEOwAAAI8g2AEAAHgEwQ4AAMAjCHYAAAAeQbADAADw\nCIIdAACARxDsAAAAPIJgBwAA4BEEOwAAAI8g2AEAAHgEwQ4AAMAjCHYAAAAeQbADAADwCIId\nAACARxDsAAAAPIJgBwAA4BEEOwAAAI8g2AEAAHgEwQ4AAMAjCHYAAAAeQbADAADwCIIdAACA\nRxDsAAAAPIJgBwAA4BEEOwAAAI8g2AEAAHgEwQ4AAMAjCHYAAAAeQbADAADwCIIdAACARxDs\nAAAAPIJgBwAA4BEEOwAAAI8g2AEAAHgEwQ4AAMAjCHYAAAAeQbADAADwCIIdAACARxDsAAAA\nPIJgBwAA4BEEOwAAAI8g2AEAAHgEwQ4AAMAjCHYAAAAeQbADAADwCIIdAACARxDsAAAAPIJg\nBwAA4BEEOwAAAI8g2AEAAHgEwQ4AAMAjCHYAAAAeQbADAADwCIIdAACARxDsAAAAPIJgBwAA\n4BEEOwAAAI8g2AEAAHgEwQ4AAMAjCHYAAAAeQbADAADwCIIdAACARxDsAAAAPIJgBwAA4BEE\nOwAAAI8g2AEAAHgEwQ4AAMAjCHYAAAAeQbADAADwCIIdAACARxDsAAAAPIJgBwAA4BEEOwAA\nAI8g2AEAAHgEwQ4AAMAjCHYAAAAeQbADAADwCIIdAACARxDsAAAAPIJgBwAA4BEEOwAAAI8g\n2AEAAHgEwQ4AAMAjCHYAAAAeQbADAADwCIIdAACARxDsAAAAPIJgBwAA4BEEOwAAAI8g2AEA\nAHgEwQ4AAMAjCHYAAAAeQbADAADwCIIdAACARxDsAAAAPIJgBwAA4BEEOwAAAI8g2AEAAHgE\nwQ4AAMAjCHYAAAAeQbADAADwCIIdAACARxDsAAAAPIJgBwAA4BEEOwAAAI8g2AEAAHgEwQ4A\nAMAjCHYAAAAeQbADAADwCIIdAACARxDsAAAAPOIOtwtoZmZmZmxszDCMiYmJujdMTU0lEolM\nJjM6Ourz+USkWq06b7As6/LlywsLC+l0WkTi8fijjz56/Pjxnp6eVgpQY9bQdT0Wiw0ODrby\nbE09IrK8vNzKVwMA9t7Q0JDbJQA74tuYPDrK+Ph4Op2en58PBoM1HxUKhVAoFI/Hz58/L/WC\n1NLS0tGjR+sOWywW/X7/pt9eN9gpKysrzbNdk2A3NPSdTb/aRcvLbxz6oWtuV9HQ8p++94E7\n59yuoqFbr8Ye+OMn3a6imVs//K/v/9y421U09PwHzt/3iz/ldhUNvfAr/7H3g8fdrqKZyu9d\nLj/4k25X0VDfzc98poPL+8mbnyHYYb/r9KXYJ598UkRCoVC5XHZeL5fLoVDIvmGjUqmkUl02\nm719+3a1Wq1Wq7dv385msyISCARKpVKLNVRfT40wN9e52QIAAHSnTg92/f39+XxeRM6dO+e8\nrn7N5/P9/f0bn7IsKxAIiEixWIxEIn19fep6X19fJBIpFosikkwmLcvaRkmRSEREUqnUNp4F\nAADYPZ0e7ERE0zRd11OplGma6oppmqlUStd1TdPqPvLss8+KSDabrbve6vf7s9msaZrXrv33\nBUefz+fz+SzLGh8f9/l8U1NT26gzl8uFw2Gfz5fL5bbxOAAAwA7tg2AnIk888YSIhMNhEbEs\nS/2gLta1uroqIg8//HCjGx566CERuXHjhvPi7Oys2mNxzz33NClG5TZd150Xk8lkNBpV0TMa\njW4vGgIAAOzE/gh2fX198/PzImKa5uXLl0Vkfn7eXmDdKJFIiEjdVVpF7XtQt9kqlUqlUqlW\nq2qx1eZ7vWg0KiKjo6P2DUtLS6lUStO0tbW1arW6tra2sLCwvb8UAABg2/ZHsBORYDCo63o4\nHI5Go7qub9wku3OxWKyVNii6rq+trTlT4+LioogYhqEu9vf3T05Otr08AACA5vZNsBPHBFvN\nTFu7NGpfYu+HjcfjmqY99thjNXOBqh7n4630UgEAAGiv/RTs7Om0TefV1Atw6+vrjW5QHxmG\nsaUCzpw5Y5pmIBCo6b0CAADQCfZTsGvd4cOHReS5555rdIP6aNPTI2oMDg6qLEgTOwAA0IG8\nGeweeeQREYlGo3W7EJdKpWg0qmmaum1LTp48KSKJRMI5HajSnvO71LZcAACAvdTpR4rVaHRO\n18aPSqWS6lGczWaDwaDaQlsulwuFgtrWevXq1eHh4ebD1r2uDqh1nmCrvkvTtOnp6f7+/vX1\n9ccff1y1PuGsWADYRzhSDPudZ4OdOLLdRs5U12TYutcty+rt7RUR595YlfbsezKZzNjYWKNS\nAQAAdoM3l2IVv99fqVTy+Xw8HldX4vF4Pp+vVCrOVLdVPT096rjYixcv2hcnJiby+bw6CSOb\nzTq73AEAAOyNfTZj5wEsxQJAx2IpFvvdHW4XsAUzMzNjY2POl9tqqPXQTCYzOjraaBX18uXL\nCwsL6uiweDz+6KOPHj9+vJW+xPLayqyTruuxWGyru2uHhr5/S/fvseXlPx+697bbVTS0/OLd\nD/QW3K6ioVuV4AOWvvl97rnVk3rguY+4XUVDtx5++u2ffr/bVTT0tY/9/nee7OgulW/816Xn\nfvaQ21U09PBv3jj34fe6XUVDT3z2mtslADu1n5ZiR0dH4/F4IpEoFOr83/VCoZBIJOLxeKNl\n0KWlpd7e3mg0qlKdiKTT6Wg02tvbW3fzbCtSqdTBgwebNMwDAADYM/sp2InIk08+KSKhUKim\nRXC5XA6FQvYNG5VKpaNHj4pINpu9ffu2Okni9u3b6m25QCDQerarOqjHZ2ZmdvA3AQAAtMc+\nC3b9/f35fF5Ezp0757yufs3n8zWHfSmWZantscViMRKJqNYnItLX1xeJRIrFoogkk0nLsrZa\nTyQSEZFUKrXlvwQAAKDd9lmwExFN03RdT6VSqlGciJimmUqldF1Xm1I3evbZZ0Ukm83WPcLV\n7/dns1nTNK9d++9vV/h8Pp/PZ1nW+Pi4z+ebmpratCT7560+CwAA0C77L9iJyBNPPCEi4XBY\nRCzLUj+oi3WpcyAefvjhRjc89NBDInLjxg3nxdnZWfU23j333NPowVwuJyKnT5+uud7KswAA\nAO21n3bF2vr6+ubn50OhkGma3/zmN0Vkfn7eXmDdSLUOrrtKq6htrYlEwrnftlKpVCqVjRtm\na/bGqpMtau5p9CwAAMDu2ZfBTkSCwaCu62quTtf1jdFq52KxWCvJ7NatW+vr6zWpscVnAQAA\n2mhfLsUq9hFezrO82qhRdzrnrthisVgqlQYGBtRq76bPAgAA7J59HOzsKbFN58Z0XReRJt3m\n1EeGYWy1Br/fPz09LSJzc3NbfRYAAKC99nGwa93hw4dF5Lnnnmt0g/poe9NsahGWjicAAMB1\nXRHsHnnkERGJRqN1uxCXSqVoNKppmrptq1T3u0adVgAAAPaMr+Y01f2l7oGwdT8qlUqqR7Ha\nxKq20JbL5UKhEI1GReTq1avDw8PNh9143bIswzBSqVQ2m1XNipuUpCwvL2/zrwUA7LKhoSG3\nSwB2pFuCnTiy3UbOVNdk2JpGJzZN0+bm5tSrfpsGOwAAgF3SFUuxit/vr1Qq+Xw+Ho+rK/F4\nPJ/PVyoVZ6rbEk3TstmsneoAAABc1Ll97GZmZsbGxgzDcDYNdlKndWUyGak3T1atVi3LKhQK\nN27csPuhGIZx6NCh8+fPnz9/vslX18y3bZyo03U9FovV3Wyx6VwdS7EA0LFYisV+19FLsePj\n4+l0en5+fmP/4UKhEAqF4vG4imh1F16TyaR9nqyTpmmzs7NNTqqo0WgFdm1trclpFo0sLy8P\nDd251af20vLyqz/0I192u4qG/vSP3vWXfWtuV9HQ95UHnnn1625X0cxjd77t41+13K6ioU+9\no+dDhW+4XUVDl4JvPfFrt9yuopnf+ecPfPhnF9yuoqHP/uaj7/vQr7hdRUOfv/SLBDvsdx29\nFPvkk0+KSCgUKpfLzuvlcjkUCtk3bKRepzNNM5PJrKys2P2EV1ZWDMMwTXNkZGSrxTj7Emez\nWRGZmZnZzl8FAACwOzo62PX39+fzeRE5d+6c87r6NZ/PN5owSyaTIlIsFkdHR50LpoODgxMT\nE5lMxjTNQqGw7cLUBlh61wEAgI7S0cFORDRN03U9lUrZi6qmaaZSKV3XG7WOW1paMk3TMAy/\n31/3hhMnTmSz2QMHDqhffT6fz+crl8tTU1M+ny8cDtddwK1bm/2zGsT56cYrAAAAu6rTg52I\nPPHEEyISDodFxLIs9YO6WNfi4qI07Rjc19cXiURqtj6MjIyoPRamaW6a7XK5nIicPn16S38I\nAADArtoHwa6vr29+fl5ETNO8fPmyiMzPzzfZ+qDy2VbPB1PNUKrVqvquS5cu1dzgc4hGo6rR\n8Vb/FgAAgN2zD4KdiASDQV3Xw+FwNBrVdX0bicpXj/OGWCymetGpwdPpdPMBb926tb6+vtUy\nAAAAds/+CHby2jyc84f22nSGz7krtlgslkqlgYGB1dXV3SgGAABgG/ZNsLOPdtj0jAfDMESk\nJnJVX2+Hxfj9/unpaRGZm5vb4VAAAADtsm+CXeuOHTsmIgsLu9uiU3VaoeMJAADoHB4MdsPD\nw5qmjY2NlUqlujdYVhva7qtBGu29ZYkWAADsvY4+UqzGxnPDGn2kTp4QEcMwNE2z359bXV29\nfv16NBoVEU3TVPfjjcPWXNl4g2VZhmGkUqlsNquaFTtPP7M/rVstZ8UCQMfiSDHsd94MdiKy\nurqaSCQataPLZDKjo6ONnq0b7DbSNG1ubk6986fOrrU/MgxDbfLYR/+8AABgv/NssFOWlpYW\nFxcXFhZUwtN1/fDhw0eOHHG2wdtGsNM07dSpU8ePH3fu5DBNc2ZmRh1QOzo62qRaAACA3bCf\ngp03sBQLAB2LpVjsd3e4XcDumpmZGRsbMwxjYmKi7g1TU1OJRKLJHJtlWdeuXbtx44bdP88w\njEOHDu3k2Imhoe/f9rN7YHn5z++970/drqKhF1/4oTvfUXS7ioZe/Wpg9U3/2e0qmhn81j/4\nv1654HYVDf3Pd536+Jd+1e0qGvrUu3/hJ36no88S/O0TZx/8tyNuV9HQzZ+fffCffcTtKhq6\n+etPu10CsFMeD3ajo6PXr19PJBJ1o1ihUEgkEvF43H7frkapVEomkzUv6qmEp2na7Oxsk5PN\nAAAA9pgH253UePLJJ0UkFAqVy2Xn9XK5rLY7qBs2Ultr1TtzKysrdnPjlZUVwzBM0xwZ6dz/\ntxgAAHQh7we7/v5+1dbk3Llzzuvq13w+r1oNb5RMJkWkWCyOjo46DxwbHBycmJjIZDKmaRYK\nhV0sHQAAYCu8H+xERNM0XddTqZS9qGqaZiqV0nW9UYfhpaUl0zQNw/D7/XVvOHHiRDabPXDg\ngPrV5/P5fL5yuTw1NeXz+cLhcKNOKwAAALukK4KdiDzxxBMiEg6HRcSyLPWDuljX4uKiND5Y\nQkT6+voikYhzJk9ERkZG1Bt4pmmS7QAAwB7rlmDX19c3Pz8vIqZpXr58WUTm5+ebbH1Q+awm\nt23K7/dXKpVqtaq+69KlSzsqGgAAYCu6JdiJSDAY1HU9HA5Ho1Fd17fRr8RXj/OGWCymWhar\nwdPpdLuKBwAA2FQXBTt5bR7O+UN7bXWGDwAAoI26K9jZJ4A5jwKryzAMEVldXXVerL7eLhUJ\nAACwPd0V7Fp37NgxEVlYWHC7EAAAgFYR7OobHh7WNG1sbKxUKtW9wbKsPS4JAACgOV+3LSnW\nPRC27kfq5AkRMQxD0zT7/bnV1dXr169Ho1ER0TRNdT/eOGyjL1peXm7znwQAaJOhoSG3SwB2\nhGDX7KPV1dVEItGoHV0mk7EPmW092AEAAOwSgt3mHy0tLS0uLi4sLKiEp+v64cOHjxw54myD\nR7ADAACu83iwq+kzp+i6HovFdt6aZHvRjaVYAOhYLMViv+vGYKesrKzsMNttO9gNvesdO/ne\n3bb85a8+cFfn/k/FrVd89339uttVNPTC2x667/d/0+0qmnnh/T/77vP/q9tVNPSl8U8eiEfd\nrqKhP0lnv/0PPuR2Fc18z3++9MW/t+Xu63vmPX9YmH7wYberaOjxm88R7LDfdcWu2Jr+c9ls\nVkTm5ubcrgsAAKCduiLY1YhEIiKSSqXcLgQAAKCdujHYNVIoFMbHx30+XzgcLhQKG2/I5XLh\ncNjn8+VyuY2flkqlZDKpDpBNJpONGuABAADskjvcLsAFKpbpuu68ODU1ZR8ga5qmaZq6rk9O\nTto3JJNJe5IvGo2+9NJLzscLhUIoFLJ/TaVSqVRqfn4+GOzcl10AAIDHdMWMne/1VG9huwWd\niJRKpUQioet6pVKpVquVSkXX9VQqZc+6LS0tpVIpTdPW1taq1era2lrNaWNnz54VEfVptVq9\nevWqiFy8eHHv/kgAAND1uiLY1dB1fW1trb+/375y5coVEUkkEj09PSLS09OjZu/UdRFZXFwU\nEcMw1FP9/f3OyTwRUS3uXn75ZfXr8PBwtVo9f/78Hvw5AAAASlcEO3s/bDwe1zTtsccec6Y6\nEVExrre3157V6+3tta/bPzjbo/j9fucIaqdtIBCYmppaX19fX1/f5b8JAACgVlcEO9uZM2dM\n0wwEAuVyub0jRyKRfD6vaVoikRgYGBgYGAiHw23/FgAAgCa6K9gNDg4ahiENmthV62l9cE3T\n8vl8sVjMZrPxeNw0zaeeeqptpQMAAGymu4KdiJw8eVJEEomEc7VUpb0mDUo23rC6ulr3Tr/f\nH4lE1Nt16XS6TVUDAABsriuOFKv5G1VnE8MwJiYm1JWlpaWjR49qmjY9Pa1ev1PtS+x7SqVS\nIBCwb1hfX3/88cfVhgk1+Pj4eDqdvnr16vDwsD1gPB7fuH+Cs2IBoGNxpBj2u24MdpZlqb0R\nzr2xzjZ1iqZps7OzfX196ldnozsRyWQyY2Nj9uAqydV8+86PowUAAGhd1y3FikhPT4/axOrs\nMzc5OTk/Px+Px9WvmUzGmepEZGJiQm2PEJFsNutsgyciw8PDxWLRbnqs6zqpDgAA7DGPz9h1\nIJZiAaBjsRSL/c6zR4qpRdgauq7HYrFNJ9JqFnA3DqVpmqZpJ06ccE7ptW7o3ndu46k9s/zi\nV+6/601uV9HQ8698676/+IbbVTT0wlveevALHd3FcOV/6r/33153u4qGXvz5h970s3/odhUN\nfes3/94fvaejT5T5kS+ezL7nP7hdRUPRL/7jX3iwczu3/+rNcbdLAHaqu5ZiU6nUwYMHG21o\nbZ1pmmNjYyMjI3SqAwAAncPjwa6mKZ16ta5uE7stDaUOkzVNs1AotLtkAACAbfJ4sKsRiURE\npGb36zbYh8leuHChDWUBAAC0Q3cFu0ZyuVw4HPb5fLlcrsVHenp6RES1shMRdcKs84aNVwAA\nAHaVZzdP1KVym92URHF2sItGoy+99FIrQ1mWJSKq+wkAAEAn8PiMne/1otGoiDhb0C0tLaVS\nKU3T1tbWqtXq2trawsLCpsOWy2V1yNipU6d2r3gAAIAt6a4ZO13XR0dH7dMmRGRxcVFEDMNQ\nF/v7+ycnJ+0FVqe6TU+OHz++m/UCAABsgceDnd2Lbnx8/KWXXnrsscecqU5E1B4IZ2c7v9+/\n6bB2Hzv1ph0AAEAn8Hiws505c+bgwYOmad6+fXt7XYU5ogMAAHQ4j79jZxscHFRvxW2viR0A\nAEDn65ZgJyInT54UkUQisb7+P058UmmvVCrZV3Z+LkW7BgEAANgSn1dXGGvOe1WmpqYSiYRh\nGBMTE+pKqVQKBAKapk1PT/f396+vrz/++ONq80TNWbHN/6HGx8fT6fT8/HwwGLQsyzAM1UJl\n41PLy8vt+QsBAO02NDTkdgnAjuyDYFd3O6rau9Dkbbm6acyyrN7eXhFZW1uzd1GotGffk8lk\nxsbG7F+r1Worwa5QKIRCIftXwzDUmJ3/zwsAADxjXy7FmqY5NjY2MjJSLpe39GBPT486Lvbi\nxYv2xYmJiXw+r1oNZ7NZZ5e71gWDQXuQTCZjzwgCAADsmX0zY+es017rzGaz6vjXvfnetmAp\nFgA6Fkux2O/2ZbCT1xZVNU3L5/N7+b07t7y8fN+772/vmO31wpeeP/BD73K7iob+5E+//MY3\nvs3tKhr6zne+frv8vW5X0czdfX/1+avfdLuKht539M2fefprblfR0E9+5O3n/0XR7SqaGf+X\ngV9+7IrbVTT01DM/9viDT7tdRUPTNz9CsMN+ty+XYkVEdQa2j4hQJ4Y5b3BeUT9bljU+Pu7z\n+aamppw35HI5n88XDofVSbJNFAoFNUI4HC4UCjWflkqlZDKphk0mk86dtgAAAHtgvwY7y7JE\nRL3T1qLZ2dl0Oi0i99xzj33RNE11gKz6IZlMNnp8amoqFAqpEUzTDIVCzpsLhUIgEFA7YUUk\nlUoFAoGN4Q8AAGD37MtgVy6XVf+5U6dOtf5UpVKpVCrVatX5Wt7MzMza2lq1Wl1bW9M0LZVK\nLS0tbXy2VColEgld19UIlUpF1/VUKmVPy509e1ZE1FDVavXq1avy+i0aAAAAu23fBDufw913\n351KpTRNO378eOsjxGKxjUe7Tk5Oqr4n/f39k5OTIrK4uLjx2StXrohIIpFQI/T09KhuJuq6\nvLYo/PLLL6tfh4eHq9Xq+fPnt/hXAgAAbN++CXZOmqZlMpnZ2dmNQa2JwcHBjRf9fn/Nz86e\ndjZ1sbe31w6Xqh+efbPqohIIBKamptbX152HWwAAAOyNO9wuoFUdvns3Eom8+c1vnpmZSSQS\nKu1pmjY7O9ukhTIAAEB77csZO7dU67E/Vb1XisViNpuNx+OmaT711FMuVgsAALqNN4Pd6urq\nNu5UOyEymczG29RejVY6mPj9/kgkot6uU1toAQAA9oZHgl08HhcR1V7Esqy5ubkWH0wkEup9\nuPX1ddW+5PDhwxtvO3bsmIgkk0n75blCoWC3xBMR1d/O3lGrflBVAQAA7A2PBLuTJ0+KSCgU\nUtsa1M6GVvj9/oGBAZ/PNzAwYJqmYRjO7RS24eFhXddN01Q3+3y+UCikaVosFlM3fPSjHxWR\no0ePqk+PHj0qImfOnGnPnwcAANCC/Xqk2Eamac7MzJimmclkRkdHnU/VHcG+ODU1lUgk4vH4\nyZMng8Fgk+8tFAoXL15UC6yZTObEiRPOvRGlUumZZ55RPYp1XY/FYnX34XJWLAB0LI4Uw363\nD4LdLtmlo2ABAADc4pGlWAAAAOybPnaewVIsAHQslmKx33k22KmVVidN0zRNq3kxbquDSNP3\n5zY+W3ep994f3uRZd734x6s/8JYDblfR0J/9xZ/4/vSNblfRUPWHvvOthb90u4pm3vTo933f\np/6L21U09Jcf/7s/8JHLblfR0J89ffyvH510u4pmvnshufjgaberaOjYzbPTD/4jt6to6PGb\n/8ntEoCd6qKlWNM0x8bGRkZGyuWyvNZteBvjpFKpgwcPtt4qDwAAYG94PNg5j4ioVCqqZYlq\nd7e9QarVqjoWtvVWeQAAAHvD48HOqaenR53ieuHChZ2ME4lERES1NQEAAOgcXRTsRKSnp0dE\nTNNUv6pmws4bNl5pXS6XC4fDPp8vl8vtsE4AAIBt8OzmibosyxIRTdN2MojKbbquOy8mk0l7\nDi8ajb700ks7+QoAAIBt6KIZu3K5bBiGiJw6dWpLD/peLxqNisjo6Kh9w9LSUiqV0jRtbW2t\nWq2ura0tLCy0t3gAAIBNeXzGrm7Tk+PHj+9kTF3XR0dH+/v77SuLi4siYhiGutjf3z85OWkv\n+AIAAOwNjwc7J7uPnXrTrnV2V5Tx8fGXXnrpsccec6Y6EVF7Mpyd7fx+/47rBQAA2BqPB7v2\nHgV75syZgwcPmqZ5+/bt1rscAwAA7I0uesdu5wYHB9VbejSxAwAAHYhg9z+0cpjEyZMnRSSR\nSKyvr9sXVdorlUpbGgoAAKC9fO1drOwcTY5qtY2Pj6fT6fn5+WAwaFmWYRiqZYn9VN1Bpqam\nEomEYRgTExPqSqlUCgQCmqZNT0/39/evr68//vjjavPExgKWl5fb8xcCANptaGjI7RKAHenq\nYFcoFEKhkP2rYRhqG0TzYGdZVm9vr4isra3ZuyhU2rPvyWQyY2NjmxYAAADQRl29FBsMBvP5\nvOpXnMlk7Bm45np6etRxsRcvXrQvTkxM2ENls1lnlzsAAIC90XEzdnU7z6k2Ja1vRN04iK7r\nsVjM2ZGkRZZlXbt27caNG/aEnGEYhw4dCgaDWx1KYSkWADoWS7HY7/ZBsFM0TZudnW0x2zUa\nxLl42opSqZRMJuu2Gt5SPU7Ly8tD9757q0/tpeUXv/T2nne4XUVDX7O++rf/7a1uV9HQG77r\nGzfL3+N2Fc082Pft6ef/xu0qGnr8/js+Nv8Nt6to6NOht/o/e9ntKpopffj4e/6Pf+V2FQ19\n8Z984j0TH3O7ioa+OPVpgh32uw5diq06VCoVXddN0ywUCtseRC2ezszMtP642hJhmmYmk1lZ\nWbGHWllZMQzDNM2RkZGt/VUAAAC7qUODnVNPT49aBr1w4cK2B4lEIiKiNr22KJlMikixWBwd\nHXWu4Q4ODk5MTGQymW1kTQAAgN2zD4KdiKhDwOwlUZ/PV7PYuvFKXWpzQyuDLP+8n20AACAA\nSURBVC0tmaZpGEajw8FOnDiRzWYPHDjgfLZcLk9NTfl8vnA4zFmxAABgj+2PYGdZlrw+lm1V\nLpcTkdOnT7d4/+LiYvNv7Ovri0QiNbsxRkZG1OSiaZpkOwAAsMf2QbArl8vqaIdTp05t6UGf\nQzQazWazrW9lVflsq7to/X5/pVKpVqvz8/MicunSpS09DgAAsBN3uF1AfXWbnhw/fnwnY966\ndWt9fX1Lu2I3rUpe34I4FoupVWOVINPp9Pnz57f9dQAAAFuyD2bsNE3LZDKzs7MqM7XOuSu2\nWCyWSqWBgYFdPcV1G33yAAAA2qVDg50zk+Xz+dHR0W10jHPy+/3T09MiMjc318r9avG3JgVW\nX28n9QAAALRdhwa73aAWYVvseHLs2DERWVhY2N2aAAAA2scLwa7F1dXmW2trBhkeHtY0bWxs\nrFQqNRkNAACgc3To5onm4vF4Op0uFArBYNCyrFZWVy3Lqtlau+kgk5OTpmkGAgHDMDRNs9+f\nW11dvX79ejQale12YFl+8UvbeGovfc36qtslNPOG7+rcI6dE5MG+b7tdwiYev7+j/xf/06HO\nPTJOREof3tEurj3wxX/yCbdLaOaLU592uwTAyzr0rNjmVRUKhVAoZP9qGIbqTmI/1eTA2bm5\nObUJY9NBRGR1dTWRSDRqR5fJZEZHRxuV3cofAgAA0Eb7MtiJiGmaMzMz6iDX0dHRmqfqdks5\nderU8ePHnVtrmw9iW1paWlxcXFhYUAlP1/XDhw8fOXLEuZ+DYAcAAFzXccHO85aXl90uAQBQ\n39DQkNslADvS0a/a7FzdqTtN006cONF6/5SNg+i6HovFtt21bujed2/vwb2x/OKX3v4DA25X\n0dDX/mztb/9b576D9Ybv+sbN8ve4XUUzD/Z9+98//zduV9HQR++/42PznfsO5adDb33o0//F\n7Sqauf6xv3voX3/O7SoauvHkB/yjv+J2FQ2VZn7R7RKAnfLCrtgtMU1zbGxsZGSkXC5ve5BU\nKnXw4MH19fU2FgYAALBDXRHsnF2FK5WKruumaRYKhW0Pks1mRWRmZmZ36gUAANiOrgh2Tj09\nPWr364ULF7Y9SCQSkZZ7HQMAAOyNrgt2IqI2xtpNTHw+X81bdBuv1OVsYrftQQAAANqlG4Nd\n8yMoWpHL5UTk9OnTbasJAABgxzy+K3ajcrl87tw5cRxB0aKa6bdsNhsMBttZGQAAwM50RbCr\n2/Tk+PEdnQt069at9fX1/v7+nQwCAADQRl23FKtpWiaTmZ2ddR5B0QrnrthisVgqlQYGBlZX\nV3epTgAAgK3qihm7tp+u4ff7p6enTdOcm5ubnJxs7+AAAADb03Uzdu2iFmHpeAIAADoHwa5W\ni6urzbfWskQLAAD2XlcsxTYXj8fT6XShUAgGg5Zlzc3NbfqIZVmGYYhja+2WBll+8UttqXz3\nfO3P1twuoZk3fFfnniUqIg/2fdvtEjbx0fs7+n/xPx3q3LOAReT6x/6u2yVs4saTH3C7hGY4\njxXYVb62v3/WUdR+2OZ/Y6FQCIVC9q+GYaijKeynGvUZ1jRtbm5ObcLYdBAAAIDdxlKsBIPB\nfD6vFlUzmczExMSmj2ials1m7VS3vUEAAADay+MzdpZlXb58eWFhIZ1Oi0g8Hn/00UePHz/e\nYq+TunN1uq7HYrHBwcFWnt34z7u8vNxS6QCAPTc0NOR2CcCOeDnYLS0tHT16tO5HxWLR7/dv\nOkKTw15XVlaaZ7smwW7o3rdv+tUuWn7xaw/84DfdrqKhW//1zff7Pu92FQ09X33f/V//Hber\naOb5t52473P/yu0qGnrhA5848Kl/7nYVDf3Jx3/tjtg/cruKZv5m7j995cEH3a6ioXfevPlM\nB5f32M2bBDvsd55dii2VSirVZbPZ27dvq8bCt2/fzmazIhIIBEqlUotDVV9PjdDKHgsAAIC9\n5M1gZ1lWIBAQkWKxGIlE+vr61PW+vr5IJFIsFkUkmUyqliVbFYlEhA52AACg83gz2D377LMi\nks1m6663+v3+bDZrmua1a9fUFZ/P5/P5LMsaHx/3+XxTU1Pb+NJcLhcOh30+Xy6X20nxAAAA\n29PR7ay2TfUHfvjhhxvd8NBDD4nIjRs3gsGgfXF2dlbtsbjnnnuaDK5ym67rzovJZNKew4tG\noy+99NL2qwcAANgWb87YqR5y6tSvutS+B3WbrVKpVCqVarWqFlttvteLRqMiMjo6at+wtLSU\nSqU0TVtbW6tWq2trawsLC+39iwAAADblzWC3PbFYrJU2KLqur62tOVPj4uKiiBiGoS729/dP\nTk7uXp0AAAB1eXMpdnsatS+xW5aMj4+/9NJLjz32WM1coJr5cz7eSi8VAACA9vLmjJ16AW59\nfb3RDeojdd5r686cOWOaZiAQKJfLO6wQAACg7bwZ7A4fPiwizz33XKMb1Eebnh5RY3BwUGVB\nmtgBAIAO5M1g98gjj4hINBqt24W4VCpFo1FN09RtW3Ly5EkRSSQSzulAlfac36W25QIAAOwl\nzx4pViqVVI/ibDYbDAZVj+JyuVwoFNS21qtXrw4PD6ubGx3/Vff61NRUIpEwDGNiYsL5XZqm\nTU9P9/f3r6+vP/7446Zp1h2Ts2IBoGNxpBj2O88GO3Fku42cqU62GOwsy+rt7RUR595Ylfbs\nezKZzNjYWN0xAQAAdkkHLcValpXL5dTZDz6fb3x8PJfLbe/UL8Xv91cqlXw+H4/H1ZV4PJ7P\n5yuVijPVbVVPT486LvbixYv2xYmJiXw+r2maiGSzWWeXOwAAgL3RKTN2S0tLR48erftRsVj0\nUvcQlmIBoGOxFIv9riP62JVKJZXq6r4PFwgEPJbtht71NrdLaGb5y19/4K1/7XYVDd36xnff\n/7c33K6ioeffcOj+P0q7XUUzz/9I/P7PjbtdRUPPf+B8/7/7+25X0dD6P/1/7vzn73G7imZe\n/bUv3vxY/VdQOsGDny4+89ght6to6LFnOve/LUCL3F+KtSxLvQlXLBYjkYhKdSLS19cXiUSK\nxaKIJJPJnazJAgAAdAP3g92zzz4rItlstu6cnN/vz2azpmleu3ZNXVFv4FmWpd7Gm5qaUtdL\npdLU1JT6NBwO53I55zjqerlcVveEw2G1a9Upl8uFw2Gfz6eeVY84bygUCupLw+FwoVCoebxU\nKiWTSfVUMpms22kFAABg97i/FKtavj388MONbnjooYdE5MaNG8Fg0L44OzubTqdF5J577hER\n0zTD4bD9qWmaKrdFIhHnUCMjI+q6usHe7iAiyWQylUqpn6PR6CuvvFJThnPfq3pc13X7TNhC\noRAKheybU6lUKpWan5931gwAALCr3J+xU2mp5vRVJ3U+hLOZiIhUKpVKpVKtVlV0U6nu6tWr\n1Wq1Wq1evXpVRFS/Oie1T7Zarc7Pz4vIpUuX1PVSqZRKpTRNW1tbq1ara2trNfN5pVIpkUjo\nuq4er1Qquq6nUil7Wu7s2bMioh63C3BumwUAANht7ge77YnFYj09PfavKk7ZTUwadTOxn1IT\naWrOT0SuXLkiIpOTkypf9vf321NxzhsSiYR6vKenRwVNdV1EVBB8+eWX7QKq1er58+fb89cC\nAAC0YL8Gu7rHvJbL5VKpZJpmMpls/Sl5bTrQ+ZJfzQt/6obe3l7fa1SPYnseUXW2CwQCU1NT\n6+vrzgPHAAAA9ob7wU7XdRFpkoTUR+o81iaSyeTdd98dCATC4bD9ttyeiUQi6o29RCIxMDAw\nMDAQDofL5fIelwEAALqZ+8Hu8OHDIvLcc881ukF91GiyTZmZmUmlUvF4fH5+vlgs3r59u+11\nymsLvjXsTzVNy+fzxWIxm83G43HTNJ966qndKAMAAKAu94PdI488IiLRaLRuf5BSqRSNRjVN\nU7c1og5mPX/+fDAY3EYrYzUd6CygppiNNzTi9/sjkYh6u85+hw8AAGAPuB/senp6VBfiQCCQ\ny+Xs5ctyuZzL5VTv4k984hPOrRKNqM4p6+vr586d21INP/ZjPyYiyWRSLfuur6/XvKV37Ngx\n5w0iUigUnF30VH+7paUl9av6wT6jFgAAYA+438dORPx+f7FYDAQCGxuUiMjVq1cb7XK1ZbPZ\naDR68ODBmuurq6vN13DtAlT7ko1di5Xh4eGNN2iaFovF1M8f/ehH0+l0zXG3Z86c2fSrAQAA\n2sXnfEvMXZZlPfvss5cuXVIrmPF4/EMf+tAjjzxSM1enToPYWPbMzIxakNV1PRaL/dVf/VUg\nEDAMY2Jiou5TG6/kcrkLFy6ozsOxWEzFROcNhULh4sWLqrxMJnPixAn7ADQRKZVKzzzzjNq3\noUaomymXl5e39+8DANhtQ0NDbpcA7EgHBbudsyzr2rVrN27csLuQGIZx6NCh7R3/4PP5nGdL\nAAAAdLiOWIptC3VUa81aqkp4mqbNzs46Z9dqqNk7e83XsqzZ2VkReeCBB3a3aAAAgPbxyIxd\nqVRS2ywymcyjjz5qr4Gurq6applIJFQvkkaP1xw1qzR/ZNtYigWAjsVSLPY7jwS7cDhsmmax\nWKzb60S9fjc/P99kTbZQKPz+7/++ekMuHo8/+uij6hTatlteXu7w/3AsLy+/4x3vcLuKhr76\n1a/eeeedblfR0Kuvvtrhjan7+vr+4A/+wO0qGvrRH/3R3/qt33K7ioZ+5md+psP7U/7yL//y\nhz/8YberaOizn/3sgw8+6HYVDd28ebPD//sMbMr9dic7t7S0ZJqmYRiNOtidOHEim80eOHBA\n/arOBLMsS/UoUS1LgsHg5OTk/Px8PB5Pp9MXLlwoFArqfsuyfD7f+Ph4zbA119VasBo8mUy2\n0vQOAACgjbwQ7BYXF0VE07RGN/T19UUikZo9qrOzs2p/6z333KOuTE1NhUIhddE0zVAopLrZ\n9fT0GIaRTqdVnzzbCy+8ICIf+tCHRKRQKAQCAfsos1QqFQgE7GgIAACwB7wQ7NQOiVb61TlV\nKpVKpVKtVtWSa6lUSiQSuq6ri5VKRTWuUxNvqoPxwsKCc4SbN2+KyJEjR0Tk7NmzIrK2tqbO\nGbt69aqIXLx4sS1/IAAAQCu8EOzq8tXjvCEWizk75F25ckVEEomEutjT06Pyorru9/vj8fjY\n2JhlWfYjY2Nj8XhcbbZVu3Fffvll9dHw8HC1WlUHiwEAAOwNzwa7TdXM8KkY19vba6fA3t5e\n+7qInDx5UkSeffZZ9atallXrsCKSzWZFJBAITE1Nra+v2yePAQAA7BkvBDvDMOS1pGWrvt7O\nv+W9732viMzMzKhfV1ZW5LV1WBGJRCL5fF7TtEQiMTAwMDAwEA6HO3x3JAAA8BgvBLtjx47J\nhhfgtqdaj/qop6cnm82apqneurt06ZK9DquovnfFYjGbzcbjcdM0O7wtAgAA8BgvBLvh4WFN\n08bGxhp1GHG+GNeImvZr3qPk4YcfFpFnnnlmfX09nU7b67BOfr8/Eomot+vUBlsAAIC94YVg\nJyLqRFf1iptzTXZ1dTWXy6m35Zr0Q5HXpv2SyaT9elyhULC73Cn9/f3xeDyVSj3++OPiWIcV\nEdUSb2lpSf2qfojH4236+wAAADbnkWDn9/tXVlbUK24HDx60N0AcPHgwGo2KSCaTaX4+2PDw\nsK7rpmkODAyoZ0OhkKZpsVjMeZvKaqZp1qzDfvSjHxWRo0ePqmePHj0qImfOnNmNPxYAAKAu\njxwpZltaWlpcXFxYWFD9R3RdP3z48JEjR5whTPU9qfuHFwqFixcvqiXUTCZz4sQJ54MiYlmW\nmv9TWyWcH5VKpWeeeUb1KNZ1PRaL1W2tx1mxANCxOFIM+53Xgt1uK5VKgUBARG7fvl2T+QAA\nANzlkaXYPfPMM8+IyPz8PKkOAAB0mjvcLmDfsA+u0HU9GAxuexyWYgGgY7EUi/1uPwU7y7Ku\nXbt248YN+zQIwzAOHTq0k5jVOk3TTNM0DGNiYmKHQ91737vaUtIuefGFL3//u+5yu4qG/vzL\nr/xl3zfcrqKh7yu/9fqdS25X0cxDrw7/e2vO7Soa+mhP7ONf+lW3q2joU+/+hZ/43X/mdhXN\n/PaP//qDmajbVTR0cyz74Jkfd7uKhm7+xu+6XQKwU/sm2JVKpWQyqbZE2FTC0zRtdnZ2t9dG\nm2+qBQAAcN3+eMdObVkwTTOTyaysrNhnQqysrBiGYZrmyMiI2zUCAAC4bH/M2CWTSREpFot+\nv995fXBwcGJi4i1vecvY2FihUNibNVkAAIDOtA9m7JaWltTLbTWpznbixIlsNnvgwAH1q2oR\nbFmWOg3CeXREoVBQF8PhcKFQ2DhUkxvUsOVyeWpqSt1Qsy6sFovVbclksvnpZAAAAG23D4Ld\n4uKiND0QrK+vLxKJ1HQDnp2dVX2G77nnHnVlamoqFAqpi6ZphkIhNRFo2/QGERkZGVEv9pmm\n6cx2hUIhEAio7sQikkqlAoFA3ewIAACwS/ZBsFNBqu4pDk1UKpVKpVKtViORiIiUSqVEIqHr\nurpYqVR0XU+lUva82qY3KH6/X90wPz8vIpcuXVLXz549KyJra2vq5b+rV6+KyMWLF3f6xwMA\nALRsHwS7unz1OG+IxWI9PT32r1euXBGRRCKhLvb09Ki8qK63ckPNsOp9PjW9JyJq6u7ll19W\nvw4PD1er1fPnz7f/LwcAAGhgvwa7TdXM8KmU1tvba6dAdeSr3RJv0xvqDmvLZrMiEggEpqam\n1tfX19fX2/8nAQAANLUPgp1hGCKyurrqvFh9PZdK+x8ikUg+n9c0LZFIDAwMDAwMhMPhcrns\ndl0AAKCL7INgd+zYMRFZWFjY+VDVerZ0QxOapuXz+WKxmM1m4/G4aZpPPfXUzmsGAABo0T4I\ndsPDw5qmjY2NNWogYlnWpoOoab8mLUg2vaFFfr8/Eomot+vsN/AAAAD2gK8T1jE3pU6eEBHD\nMDRNs190W11dvX79ejQaldcmzERE7aKo+buWlpaOHj2qadr09HR/f7+IFAqFUChkn/266Q0b\nh3VeGR8fT6fTV69eHR4etkeLx+Mb908sLy/vxj8RAGDnhoaG3C4B2JH9EexEZHV1NZFI1PQE\ntmUymdHRUfVz3WAnIslk0u4zp9QcMtv8hubBTiW5mm9cWVnZapcWAACAbds3wU5ZWlpaXFxc\nWFhQCU/X9cOHDx85csQOZ9I42IlIoVC4ePGiWiHNZDInTpxwPtj8hubBTkRKpdIzzzyjoqGu\n67FYjFQHAAD20j4LdltiWda1a9du3LhhtywxDOPQoUNbPVK2pkOe7Cy3sRQLAB2LpVjsd54N\ndurk1rpLtzUrsJvaGOyUtbU19TbeliwvL9973zu2+tReevGFr971zu9xu4qGXvnKt7/ztr9w\nu4qG3vj1t9x6U0d3MXzgW/2572y+38gtkTf2/OJX3C6isV95p0R+77+6XUUzuQ/+4Pv+7Yrb\nVTT0+Z8/+L4nPut2FQ19/tyHCXbY7/bBrthtUJstTNPMZDIrKyt245KVlRXDMEzTHBkZ2eqY\nzgYoqh3xzMzMLtQOAACwTd4MdslkUkSKxeLo6KhzwXRwcHBiYiKTyZimWSgUtj2+On+2ZqcF\nAACAuzwY7JaWlkzTNAzD7/fXveHEiRPZbPbAgQPqV3WAWLlcnpqa8vl84XC40d7bGpqm2T9v\nPKx24xUAAIBd5cFgt7i4KK9PXTX6+voikUjN1oeRkRG1x8I0zU2zXS6XE5HTp0+3p2IAAIB2\n8GCwU/lsq1tW/X5/pVKpVqvz8/MicunSpZobfA7RaDSbzW51dy0AAMCu8mCwq8tXj/OGWCzW\n09MjIiqubXoa2K1bt9bXO3rzIwAA6DbdEuw2tekMn3NXbLFYLJVKAwMDq6ure1MeAADApjwY\n7AzDEJGayFV9vR1+hd/vn56eFpG5ubkdDgUAANAuHgx2x44dE5GFhYVd/RbVmpiOJwAAoHN4\nMNgNDw9rmjY2NlYqlereYFltaLuvBmm095YlWgAAsPe8eaSYOnlCRAzD0DTNfn9udXX1+vXr\n0WhURDRNy+fz8tqJYc5/h5orG2+wLMswjFQqlc1mVbPi8fHxdDo9Pz8fDAbtT2ueUjgrFgA6\nFkeKYb/zZrATkdXV1UQi0agdXSaTGR0dVT+3GOw20jRtbm5O7aUtFAqhUMj+yDAM1XXFq/+8\nAACgA3k22ClLS0uLi4sLCwsq4em6fvjw4SNHjvT19dn3bCPYaZp26tSp48ePq1SnmKY5MzOj\nDqgdHR3dOCwAAMCu8niw60AsxQJAx2IpFvvdHW4XsLs2Trbpuh6LxbZ0LkXdGTtN006cOOGc\n+Wvdvffdv42n9syLLzx/19t/2O0qGnrla3/8Z/JWt6to6AfkG5+//Ua3q2jmfXd/5/f+4BW3\nq2jogz961++lv+x2FQ19MP6u//vj/5/bVTTzDz91+LN/7z+6XUVDH/7Dn/rEg/+b21U09K9u\n/u9ulwDslAd3xTaXSqUOHjy4w0MjTNMcGxsbGRkpl8vtKgwAAGCHuiLYOVsTZ7NZEZmZmdnJ\nIJVKRdd10zQLhcIu1AsAALAdXRHsnFR3kh02Fu7p6VGbXi9cuNCesgAAAHas64Kd4mws7PP5\nat6i23hlI7Uf1m6nsr1BAAAA2qjrgl0ulxOR06dP73Cc5idPAAAA7D2P74pVambOstlsMBjc\nyYDlcvncuXMicurUqR1VBgAA0D5dEexq3Lp1a319vb+/f0tP1W16cvz48fbVBQAAsCNdsRTr\n3NBaLBZLpdLAwMDq6uq2B9Q0LZPJzM7OOk+eAAAAcFfXzdj5/f7p6WnTNOfm5iYnJ1t/kCM6\nAABAh+uKGbsaahF2hx1PAAAAOk03BrvmG1p3skTb3kEAAAC2pOuWYi3LMgxDHBta4/F4Op0u\nFArBYNCyrLm5uW0Mu6VBXnzh+W18xV565Wt/7HYJzfyAfMPtEpp5393fcbuETXzwR+9yu4Rm\nPhh/l9slNPMPP3XY7RI28eE//Cm3S2iG81iBXeXz9qtjjVoEa5o2Nzentj4UCoVQKGR/ZBiG\nOlXC/pdRgzT/h9p0EAAAgN3WdUuxmqZls1k71YlIMBjM5/NqZTaTyUxMTGxj2LYMAgAAsBPe\nnLHbOFGn63osFhscHNzqUJZlXbt27caNG2oGTkQMwzh06NC2WxwvLy9v70EAwG4bGhpyuwRg\nR7ol2Clra2tb6ktcKpWSyaR9IKyTpmmzs7N9fX1brW15efne++7f6lN76cUXnv/+d205Ae+Z\nP//y6nfe/la3q2jojV/7xnPf/S23q2jm4b9+07mXS25X0dATf8f/8T/aznuue+NTPxL7id/9\nZ25X0cxv//ivP5iJul1FQzfHsg+e+XG3q2jo5m/8LsEO+52Xl2KdfYmz2ayIzMzMtP54qVQK\nBAKmaWYymZWVFXuolZUVwzBM0xwZGdm12gEAALbMy8HOKRKJyBZ71yWTSREpFoujo6PONdzB\nwcGJiYlMJmOaZqFQaHupAAAA29MtwU5x9q7z+Xw1K7bOK0tLS6ZpGobh9/vrDnXixIlsNnvg\nwAHns+VyeWpqyufzhcPhugu4AAAAu6dbgl0ulxOR06dPt3j/4uKiNG5iLCJ9fX2RSKRmN8bI\nyIjaY2GaJtkOAADsMS8HO59DNBrNZrOtb2VV+Wyru2j9fn+lUqlWq/Pz8yJy6dKlrdYMAACw\nbV4OdjVu3bq1vr6+kxF89ThviMViqj2eSpDpdHonXwcAALAlXg52zl2xxWKxVCoNDAzs6imu\n2+iTBwAA0C5eDnZOfr9/enpaRFo8CladJ1uTAquvtxt1AgAAbFu3BDsRUa2JW+x4cuzYMRFZ\nWFjY3ZoAAADap4uCnWVZ0nija83k3PDwsKZpY2NjpVL9Hv1qNAAAgM7h5SPFnH+aZVmGYaRS\nqWw2q5oVj4+Pp9Pp+fn5YDBof+p8Sp08ISKGYWiaZr8/t7q6ev369Wg0KiKapuXz+brfuPGK\nwlmxANCxOFIM+52Xg91GmqbNzc2pjauFQiEUCtkfGYahWpw4/0FWV1cTiUSjdnSZTGZ0dNT5\nja0EOwAAgF3SLcFO07RTp04dP35cpTrFNM2ZmRl1Guzo6GijKLa0tLS4uLiwsKASnq7rhw8f\nPnLkSF9fX803EuwAAICLvBnsOhlLsQDQsViKxX53h9sF7JaNk3a6rsdiMWerObcm1e59d0f/\nh+PFLy2/+W33ul1FQ9/8+ouvfM8Bt6to6K5v/8nn/+otblfRzPu+9y/Sf/QGt6toKP4jf/u/\n/P4rblfR0L95/13h/3NHfc53W/7n+t/3L/5ft6to6PP/8sj7Yp3buf3zc3G3SwB2qot2xaZS\nqYMHD+7w8AkAAICO5fFg5+wnnM1mRWRmZsbtogAAAHaFx4Odk+py0mKDYgAAgH2ni4Kd0qhB\nsc/nq3ktb+OVQqEwPj7u8/nC4XChUKgZoVQqJZNJ9VQymWzU2RgAAGCXdFGwy+VyInL69Ont\nPT41NRUKhdLptIiYphkKhZLJpP1poVAIBAL2dGAqlQoEAhvDHwAAwO7xeLDzOUSj0Ww2GwwG\ntzFOqVRKJBK6rlcqlWq1WqlUdF1PpVL2tNzZs2dFZG1tTb3Pd/XqVRG5ePFiG/8WAACA5jwe\n7GrcunVre7tir1y5IiKJREL1N+7p6VHHVKjrIqJ6F7/88svq1+Hh4Wq1ev78+baUDQAA0AqP\nBzvnrthisVgqlQYGBlZXV7c6jopxvb299vxfb2+vfV1E1JbbQCAwNTW1vr5OUxUAALD3PB7s\nnPx+//T0tIjMzc21ffBIJJLP5zVNSyQSAwMDAwMD4XC4XC63/YsAAAAa6aJgJyL9/f2yg44n\n1XrsTzVNy+fzxWIxm83G43HTNJ966qn21A0AANCC7gp2lmVJ444nTjXLtYZhiEgrHUz8fn8k\nElFv16kttAAAAHvDs2fFbmRZlspnp06d2vhpPB5Pp9OFQiEYDFqWVbNcTmZTYAAAIABJREFU\ne+zYMRFJJpPT09Nq2q9QKIRCIcMwJiYmRGR8fDydTl+9enV4eFhElpaW1Jh1K3nxS8tt/tva\n7Ztff9HtEpq569t/4nYJzbzve//C7RI2Ef+Rv3W7hGb+zfvvcruEZvI/1+92CZv4/L884nYJ\nzXAeK7CrfM7FRC+p6S1s0zRtbm5ObW5V96h/ARXU7NsMw1AbI+x/n2QyWbOGq2na7OxsX1+f\niCwtLR09erTmu1ZWVgYHB9v1FwEAADTXRUuxmqZls1k71dUIBoNq94OIZDIZNQ/nNDk5OT8/\nb0/CZTIZO9WJyPDwcLFY1HVd/arrOqkOAADsMc/O2AEAAHSbLpqxAwAA8DaCHQAAgEcQ7AAA\nADyCYAcAAOARBDsAAACPINgBAAB4BMEOAADAIwh2AAAAHkGwAwAA8AiCHQAAgEcQ7AAAADyC\nYAcAAOARBDsAAACPINgBAAB4xB1uF9B1lpeX3S4BAIBuNDQ05HYJu67rgp1lWZcvX15YWEin\n0yKi6/r73//+YDC4pUFKpZLf768Z9umnnx4bGxORTCbzwQ9+sL+/v9Hjn5O/3Fbte+QD8n1D\nQ19xu4qGlpff+V+/9ka3q2joB9/+nXd/5jfdrqKZL/3kz37my24X0dhPvkvuX/sNt6to6PmB\nM195879zu4pm3vnNf3rfq9/vdhUNvXDnn995551uV9HQq6++evszZberaObun+z7ztv+wu0q\nGnrj198y9EOzblfR0PKfjrhdwl7orqVYy7JisVg0GlWpTkRSqVQoFAqHw5ZltTjI+Ph4IBCo\nuRiLxVSqE5GxsbGBgYH19fV2lQ0AANCK7gp2Tz/9tGmauq7fvn27Wq1Wq9Xbt2/rum6a5uXL\nl1scxA6FtlwuZ5pmJpNRY2azWRG5ePFim6sHAABoqruCnZpUSyQSfX196kpfX18ikRCRaDS6\n7WEvXLggIh/5yEfUr5FIRH3LDqsFAADYku4KdkpPT0/Nr2qmzXmxUCiMj4/7fL5wOFwoFOzr\nPp/P/sH+OZ/PV6vVmmE1TduV6gEAABrormCnFklnZmbK5Wavx05NTYVCIbXkappmKBRKJpOt\nf0sulxOR06dP76xYAACAremuYBeJRHRdHxsbu/vuu8fHx3O53Orqas09pVIpkUjoul6pVKrV\naqVS0XU9lUqVSiURsSf2Nk7yicjS0pLP54tGo9lsdqs7bQEAAHaou4KdiExOThaLxUwmk06n\no9HowYMHw+GwCm3KlStXRCSRSKil1Z6eHvW2nLre3Le+9S3DMDRNi0ajpmnu2h8BAABQR9cF\nOxHx+/2jo6PVanVlZSWfz4tIIBCwp+5UjOvt7fW9pre3V1rbDBEMBicmJvL5fCaTqXk5DwAA\nYLd1Y7CzDQ4Oapo2PT0tIr/xG+3siap2yJ49e7aNYwIAADTXRcHOuY/VSR0RUdOdrlpP69+l\nlnFZjQUAAHupi4KdYRgisnF5VJ05YXcnUbc537rbVDgc9vl8zrMr1K7beDy+s5IBAAC2wLel\niah9bXV19eDBgyIyPz//3ve+V02qra6uzs3NpVKp+fl5tY91aWnp6NGjaolWTeYVCoVQKGQY\nxsTEhLzWyu727dt2l+NcLheNRjOZzOjoqIhYljU7O5tIJOwxnZaXl/fubwYAAK8ZGhpyu4Td\nV3fN0atUH7uNdF133qbres0NmqbZp5DZ83CaptmPbGxHXDMmAADAbnvDL/3SL7U7K3auoaGh\nn/u5n7v//vtFRG2D1XX9k5/85E//9E87bwsGg4888sib3vSmL3zhCyKSyWR+4Rd+wZ6fGxwc\n/Ju/+ZsvfOELBw8etA8ii0aj7373u1999dXV1dV4PP7JT37yYx/72J7+bQAAoOvd4XYBe62n\np+euu+6655577Cvf+ta3Nt4WDAaDweD58+c3fjQ4OBiPxzd+1NfXp4Z96aWXyuWyZVk1h4wp\nLMUC6GZdsRYGuKeL3rETEcuyYrHYxs2qmqbNzc3VzWEbjY+Pp9Ppmn839ZpdK2MuLy+fe/mL\nW6997zzxd95z3+fa2fylvV74wJnPXfu221U09IH3fs8PXK64XUUzf3a89+Z3/7LbVTT04F8/\nNXTnP3a7ioaWX/0P328tuF1FM3/e8+hXvvIVt6to6J3vfCfBDthVXbQrVkSefvpp0zR1Xbdf\nmLt9+7au66ZpXr58ucVBahqjiMj6+rraPKFOIatWq/Pz86ZpPv300+3+CwAAABrqrmA3NjYm\nIolEwn5hrq+vTx0pUTPftiUXL14UkdHRUXt+Tm2GVV8HAACwN7or2Ck1y6M9PT1qms15sVAo\njI+P+3y+mpPB7BbHznbHExMTXbWiDQAAOlN3BTvV7mRmZkY1EG5kamoqFAqpJVfTNEOhUDKZ\n3NIXqWbFG9umAAAA7J7uCnaRSETX9bGxsbvvvnt8fDyXy6mmJ06lUimRSOi6rl6Yq1Qquq6n\nUil1FoU9M7dxks/p2rVrIvLYY4/t2p8CAABQq7uCnYhMTk4Wi8VMJpNOp6PR6MGDB8PhsPMA\nsStXrohIIpFQK7Y9PT3qJTx1vRXlcvns2bOGYfj9/l34CwAAAOrrumAnIn6/f3R0tFqtrqys\n5PN5EQkEAvbUnYpxvb29vtf09vb+/+3df3Aj533n+S8SR3Jiy6RjhbOpkUnJK2Fc8siApIrE\n0VzZZ8I5e2w3eNmMfODsMpu6kDzgVnOrMbF1kbZxLh17dbnahkd7M3vDA5mrONwd4ERlc8W2\nJesyQG5URxNJaiTAA8kmrJVMeKcUwj9ERFZsK4lxfzxRbwu/BvzZYPP9+sNF9i98wbLkj5+n\nn+9jH7+uWq02MTERCATU/mMAAAB75iAGO5vf71d7worI2bM707nNNM3Dhw/PzMzsyNMAAAC6\nd4CCnXMdq9Pg4KA0dadruf9a5+dXq9VYLCYijz3Wu91fAQCAhx2gYGeapog4e5coagWrpmnO\ny5xv3XWjWCweOnTo5ptvnpmZsZvkAQAA7KUDtKVYuVw+cuSIiGSz2XvvvVetjSiXywsLC4Zh\nZLNZ1VU4n88fO3ZMTdGqwbxcLhcKhUzTVK/NqWG/9fV1O8BVq9VDhw7put7NDCx7xQI4yNhS\nDNhVByjYSasdXZWGTJZIJAzDcF6gadr8/LxKcmqvWHVQrb2Ym5trt8nEgfrzAgAAdx2sYCci\nlUrl2WeftSzLsiwR0XX9s5/97PDwcMNluVxucXFRBbhUKjU6OmqPz5XL5bNnz87OztrBruWr\ne8pB+/MCAAAXHbhg5zqmYgGgZzFTjP3uXW4XsNdqtdozzzxz+fJlNRqn6/onPvEJ9XZd94rF\nYrvmw+qFvM5x+f2fOrGpj9tjrz/7zMTiNberaGv+wcOPvvw/ul1FW4/f/r8u/8uC21V0cvxf\nBY8+dZfbVbRVOnn16NHfdruKtkqlP7g686DbVXRyV2Lx9v/wfberaOvlf3Tz17/+dberaOuB\nBx5wuwRguw7QqlgRqdVq4+PjY2NjdnMTwzBCoVA4HFZrY7sRi8WCwWDLU5VKJRQK7UytAAAA\nm3Swgt2TTz5pWZau6+vr66o13fr6uq7rlmU988wzXT6koeOdrVKpPPTQQztXLAAAwOYcrGCn\n1q7G43F7JcTAwIDaK6zlatnu5XK5oaEhNocFAAAuOljBTlEd7Jy/Nm8skcvlYrGYz+cLh8PO\nnsb2AtiGfSxCoVA6nWYnMQAA4KKDFezS6bSIzM3NVavVDpclk8lQKKSmXC3LCoVCiUSi85PX\n1tYikcgOlgoAALBZByvYRSIRXdenpqYOHToUi8UymUy5XG64plgsxuNxXdc3Njbq9frGxoau\n64ZhqE3G7IG9hkE+tUcFAACAiw5WsBORmZmZQqGQSqVmZ2fHxsaOHDkSDoedO8NeunRJROLx\nuJqx7evrUy/hqeMAAAA968AFOxEJBAKTk5P1en11dVVtHREMBu2hOxXj+vv7fW/r7++3jwMA\nAPSsgxjsbH6/X9O08+fPi8jZs2fdLgcAAGBbDlCwa1jHalOvxzV0p6u3skeFAgAAbMkBCnam\naYqIs3eJovac0DTNeZnzrTsAAIB9wXdwBqLK5fKRI0dEJJvN3nvvvWptRLlcXlhYMAwjm82q\nHWPz+fyxY8fUFK0azFPbv5qmOT09LW+3sltfX7e7HDupsx3+qqVSaVe+HgBg244ePep2CcC2\nHKBgJyKZTKblDhO6rjt7CycSCcMwnBdomjY/P6+SXCwWU/O2mqaptRdO1w12AAAAu+QATcWK\nSCQSWVtbS6VS9sSrrusrKysNO0bMzMxks9loNKp+TaVSdqoTkTNnztinAAAAesc+GLFrXvGg\naZqmaaOjoy0nQ7emWCxuaqfX5pG5XC63uLg4OzuradqpU6dOnDjRsHeZwlQsgIOMuU5gV+3L\nYKc4p0e3Sc2ubupP0RDsmid5NU1bWFhoznalUuk7fW9tr97ddWvthjvfWnW7irZeuuHIT37u\nV9yuoq13/+x7t/34/W5X0cmrv/j6Bz70125X0dYPXvmlo0PfdruKtkprdwz2/dTtKjqp1G78\n6U97t8Ibb7yRYAfsqn0zFetsO6K2+bIsq3mJ69Y09DrZrEqlMjY2lkql1C5k9Xo9m81alvXk\nk0/uSHkAAADd2DfBzsne5uvixYtu1yIisri4KCKTk5P2+JxaYDs1NeVmWQAA4IDZl8FORFSE\nsizLPlIsFpPJpOpCHA6HM5lMwy3FYjGRSKgLEomE3anOnupt7mCcy+VisZh6YIfRwenp6d6f\n0QYAAJ63X4NdQ1dhy7KCwaC9natlWWNjY85sl8vlgsGg3cTEMIxgMNh5JjeZTIZCITVLa1lW\nKBRKJBKbKk/X9c19KwAAgG3Yl8GuWq2q/SFOnTqljoTDYRFZWVlRr7itrKyIiHM1wxNPPCEi\na2trzgvUFKo92ObcN6xYLMbjcV3X1Wtz6q0+wzC63JHiypUrInLy5Mmd+cIAAABdeJfbBXSr\nZdOTEydOqJ8bZkKHh4cbLlaTtq+//rraTGJ4eLjz5OmlS5dEJB6Pqzlf9VafYRiXLl26bleU\narX6xBNPmKa5qf4pAAAA27QvR+w0TVNNgxuaiVSr1WKxaFlW85xpOp0WkWAwmEwmK5VKpVLp\n/BFqVre/v9/3tv7+fvt4B7VabWJiIhAIqP3HAAAA9sy+GbG77uqE5n3AnCKRyHve8565ubl4\nPK7C2Q62wXMyTfPw4cMNW1kAAADsgX05Ytdsbm7OMIxoNJrNZguFwvr6evM1amvXQqGQTqej\n0ahlWV/84hc7P7beSruLq9VqLBYTkccee2ybXwcAAGALPBLsVMe4CxcujIyMdH6zLRAIRCKR\nCxcuSMe+xGpxRpdLJdSVhw4duvnmm2dmZnZ8FBAAAKAb+2Yqthvlctnv91cqlbm5uYZTatOw\nlZUVta4in8+LSDQadV5TrVbtTHb8+HERSSQS58+fV+stcrlcKBQyTbP55blqtRoMBnVd73IG\n9tbaDVv5envopRuOuF1CJ+/+2ffcLqGTV3/xdbdLuI4fvPJLbpfQSWntDrdL6KRSu9HtEq7j\nxht7vUIAu2ff7BXbuc7mrVqV1dVVv98vIvl8/tixY+3Oqtgnb0/XqrPNL+05X8tzVjU3N9du\nk4ne//MCAADP8MhUbCQSSaVS6mdd11dXVwuFgji2phgeHi4UCnbHYHWNSnUicubMmYbROxGZ\nmZnJZrP2cbUOt+U0K1uHAQCAXrAPRuw8plQquV0CAKC1o0ePul0CsC374B27lq2JNU0bHR3d\nwWUKxWJxU/2E200Qq1fxOsflD699ewsV7plvDd3xnbf63a6irVtv2PjpP/grt6to68a/fN97\nfvJ+t6vo5M13v/6RG666XUVbL75119GjP3W7irZKpRt/+ReqblfRyQ//ZuDOG3/V7Sraeumn\nrw0NDbldRVtra2tulwBs176cirUsa2pqamJiolrdmX/DxmKxYDC4/edUKpVQKLT95wAAAGzB\nvgl2zk5yaudWy7JyudyOPLxD35PuVSqVhx56aPvPAQAA2Jp9E+yc1M6tInLx4kW3a/l7uVxu\naGiIzWEBAICL9mWwExG1S6y96FVEisViMplU+7qGw+FMJtNwS7FYTCQS6oJEImE3H7bf4VOn\nnLfkcrlYLKYe2Hl0MBQKpdNpdhIDAAAu2geLJ1qq1Woiomma+tWyrHA4bJ+1LEtlvkgkoo6o\nNQ32BYZhGIaRzWZHRkbafUQymVTjgvYDO7QgXltbU32MAQAA3LIvR+yq1ara8uvUqVPqiEp1\nKysr6iW8lZUVEXG2LH7iiSdEZG1tzXnB4uKiOFa2OreCLRaL8Xhc1/WNjQ37rT7DMNptMkaq\nAwAArts3I3Ytm56cOHFC/dzQXkTtG+akBvBef/11lcCGh4c7dyS5dOmSiMTjcTXnq97qMwzj\n0qVLvEgHAAB6074csdM0Te0DoVKXrVqtFotFy7ISiUTDLel0WkSCwWAymaxUKpVKpfNHqEnY\n/v5+39v6+/vt4wAAAD1o34zYXXeHjOatXZ0ikch73vOeubm5eDyuwplz41cAAAAP2Jcjds3m\n5uYMw4hGo9lstlAorK+vN1+jadrS0lKhUEin09Fo1LKsL37xi50fW29ld74BAADAdnkk2E1N\nTYnIhQsXRkZGOr8DFwgEIpHIhQsXpGNfYrU4o91SCQAAgB60b6Ziu1Eul/1+f6VSmZubazgV\ni8VmZ2dXVlbUuop8Pi8i0WjUeU21WrVnZo8fPy4iiUTi/Pnzar2Faphimub09PQ26/zW0B3b\nfMJuu/WGDbdL6OTGv3yf2yV08ua7X3e7hOt48a273C6hk1LpRrdL6OSHf9Pr72+89NPX3C6h\nE/ZjBXaVr/fnFtV62M51ZjIZZ3MT2+rqqt/vF5F8Pn/s2LF2Z1Xsk7ena9XZ5pf2nK/ltauq\nm2oBAAB2g0emYiORSCqVUj/rur66ulooFMSxNcXw8HChUNB13XmNSnUicubMmYbROxGZmZnJ\nZrP2cbUOl8UWAACgZ+2DEbsOWja30zRtdHS0+wS2S2Ns7R5bKpV29oMAAEA3jh496nYJu85r\nwU7ZVCuTvQ92V/+mf2c/a2fd9Qsbn/v3vfsSzFf+8dCr733G7Srauu1HJ377Sou3AnrHH9yb\n/vDK192uoq1vHXug/1Mn3K6irY1nn7nz+/NuV9HJSzdPHH2qd9+hLJ28+uNXf97tKtr6xdv+\n7sJ/v+J2FZ3E/vdjR4/27j+/pdIDrzz+oNtVtPWhRxcPQrDzwlSssxeJ2vvLsqxcLrep23e1\nQgAAgD3ghWDnpPb+EpGLFy+6XQsAAMCe8lqwExG1z5i9bEJEisViMplUO4OFw+FMJuO8Xh13\n/lyr1WKxmM/nSyaT9mW5XE4dDIfDLYcDM5lMOBz2+XwNzwcAANgbnupjp9RqNRHRNE39allW\nOBy2z1qWpTJfJBJp94T5+XnV/eTw4cPqSDKZtHeJVU/QdX1mZsa+xdkbZWxs7Nq1azv4jQAA\nALrhtRG7arWqNo04deqUOqJS3crKinqXbmVlRURaNr2zbWxsbGxs1Ot1Ff6KxWI8Htd1XR1U\nr/EZhmHvS5HP5w3D0DRtbW2tXq+vra1dvnx5V78mAABAMy8EO5/DoUOHVMY6ceLvF9apPKc2\nnBAR+4cOxsfH1XyucunSJRGJx+PqoP0anzouIsvLyyJimqbao2JwcNA5mAcAALA3vDYVa/ex\ncyYzEalWq6+99lqlUvnzP//z6z7EblysqBjX39/YoyQej6vtxdQFzrs671cLAACwG7wQ7K7b\nrKR5czAAAADv8cJUbGdzc3OGYUSj0Ww2WygU1tfXt/aceis7WyoAAMB2eD/YTU1NiciFCxdG\nRka2NkOqVmPYSyW6uaBcLm/hgwAAALbDC1uKdf4K6prV1VW/31+pVNQAnvMu50NaPjCfzx87\ndkzTtPPnz6vlEblcLhQKmaap3rErFovBYNC+oFKpPPTQQ6qpCnvFAgDQIw7ClmLeD3aZTKZl\ncxMV9aSLYCet3tJr2I7W2ehORFKplBop3Nd/XgAAsL94fyo2EomkUin1s67rq6urhUJB3rk1\nxXXNzMxks9loNKp+TaVSzlQnItPT00tLS6orcjqdnpyc3LEvAAAA0J39PWK3HzEVC+AgOwhz\nYYCLvNDuRLH3e7VFo9GPf/zjJ06caOhpt+WHdw7BuVxucXFxdnZW07RTp051+NxP/8W/22Y9\nu+prv/ZPbjB7tw/fW/HiH/7+d9yuoq3f+p1bb/78M25X0cn3nzxx7upn3a6irdN3ffUD3+rd\n//r94MPFqwufcbuKTu4af/rcuXNuV9HW6dOn3S4B8DgvT8XOzs6OjY2Nj49Xq9Xd/qxMJhMK\nhdQOs5Zlqc9Vu9YCAADsDa8FO2eTufX1dV3XLcva7f//WqlUxsbGUqmU2ky2Xq9ns1nLsp58\n8sld/VwAAAAnrwU7p4GBgZmZmWg0ahhGhy5027e4uCgik5OT9tzryMiIvN1CDwAAYG94Odgp\naimrc4vYYrGYTCZ9Pp/P5wuHw5lMxnm9Ol6r1WKxmM/nSyaTLR+byWR8Pp9aWjs9Pc0aFAAA\n4DrvLJ5o59ZbbxWR559/Xv1qWVY4HLbPWpalwlkkEnHeNT8/r16YO3z4cPMzVW88XddVf5Nm\n6u06Xdd35CsAAAB0w/sjdmp6VKU0EVGpbmVlRb0Mt7KyIiLNHYw3NjbUC3MNgU8cqW5mZqbd\nh165ckVETp48uXPfAwAA4Dq8P2LXoGHOdHh4uOVl4+PjLZuVdJPqqtXqE088YZrm1ramBQAA\n2JoDF+yUarX62muvVSoV57t3Tmq3sQb2xmLj4+Ptnlyr1SYmJgKBgNpGFgAAYM94fypWNbFz\nvu6WSCQOHToUDAbD4XDDDrCdGYahXqo7e/Zsu2tM0zx8+HCH8TwAAIBd4v1g98orr4jIRz7y\nEfXr3NycYRjRaDSbzRYKhfX19e4ftbq6ev78eRGZnZ3N5/MNZ6vVaiwWE5HHHntsZ0oHAADY\nDO8Huy9/+csics8996hfVW+5CxcujIyMbPYdOL/fPzg4uLS0JCKPP/6481SxWDx06NDNN988\nMzMzMDCwM6UDAABshs8zDdiat3OtVqvnzp0zDMO51kFdtrq66vf7K5WKGsBz3thyW9iGg7FY\nbHZ2dmlpSc3MVqvVQ4cOdV5RYSuVStv9qgCwbx09etTtEgAv81qwa6Zp2vz8vD2Kppa1Nl+m\nop50F+wqlcrQ0JCIbGxs9PX1zc3NtdtkwjN/XgAA0Pu8PBWr6/rS0tLS0pJzbjQSiaRSKfuC\n1dXVQqEgIqpNcZfsCdn5+Xlh6zAAANAb9veIXTgctixL0zQVs3pHsVhs9wIfU7EADjKmYoFd\ntY+DXbFYDAaD6udCodA73YDVG3jt/rClUumpv35qj0valJO/dPJ73249r90LfuWO+p0/2MTw\n6h576QPa0aNfd7uKTkqlByb/j//odhVtzf13//CDgd79B+S7xZPf/71Pul1FJzf/7qVz5865\nXUVbp0+fJtgBu2ofT8Wq3sLpdNr+uUfY25cBAADspX0c7KampjRNi0QimqbxlhsAAMB+DXaq\nP/CpU6fs/2zuGCwimUwmHA77fL5EIqG2oOj+bC6Xi8ViPp8vHA7ncjnnKZ/P5/P5qtVqMplU\nF9hrL+zFueqaHfiqAAAA3dmvwe6rX/2qvN12WP2nOuKUSCTGxsZU5DIMY2JiolardXk2mUyG\nQiE1qWpZVigUSiQSDc+fmJiIx+PqAme2AwAAcMW+DHa1Wk1t26o6z/n9fk3TDMNwJrN8Pq9a\nE29sbNTr9VQqZVnWM888083ZYrEYj8ftsxsbG7quG4ZRLBadZQQCAXVBNpsVkaefflocjevq\n9fr+XZgCAAD2o30Z7L75zW/K2zOwivpZHVeWl5dFZHx8vK+vT0Q+//nPi8jFixe7OXvp0iUR\nicfj6mxfX58amVPHbfbtIyMjwpoJAADgtne5XcBWNGz/KiIPPPCAOj48PKyOqCimhvREpK+v\nzzl+1s3Z/v7+hs+Nx+PT09P2r/btAAAAvWD/jdhVq1U1NnbkyBHf29QGX7Ozs81rIAAAAA6I\n/Rfs/uzP/mzLZzel3spOPRwAAGDH7b9gp9YorK+vN0Su9fV1+6yI6LouIuVyWf1aq9VUX5Ju\nzpqmKSINSyUAAAB63D7bUqxSqQwNDZmm6XzXzZZIJAzDWFtbGxwczOVyoVBI13W1BiKTyYyN\njaXT6UgkIiKdz+bz+WPHjmmadv78+cHBQft6+3NVgzrnn855RP28vr4+MDDQXCR7xQI4yNhS\nDNhdLScce5baQKxQKLQ8WygURCSdTqtf1bCcTdM01Z1kC2fVBfYwYfOfznkkGo3at+zgdwcA\nAOhsn43YqdnSpaWlLi9QQ3EiYprm+Ph4wxBa57O5XG5xcVEt1EilUqOjo/YFnUfsyuXy2bNn\nZ2dnNU3rUCoAAMDO2mfBDgAAAO3sv8UTAAAAaIlgBwAA4BEEOwAAAI8g2AEAAHgEwQ4AAMAj\nCHYAAAAeQbADAADwCIIdAACARxDsAAAAPIJgBwAA4BEEOwAAAI8g2AEAAHgEwQ4AAMAjei7Y\n1Wq1TCYTi8V8Pp/P50skErlcbo9rKBaL23+Iqn/7zwEAAOiSr16vu13Df1ar1cbHxy3Lajiu\nadrCwkJfX98e1BCLxWZnZ7f/Z1Gprqf+vAAAwNt6K9jNzc1NTU3pun769OmBgQERqVar586d\nMwwjnU5HIpE9qGGnAlm755RKpW0+GQCA3nT06FG3SzjoeivYqTC0sbHhHJyr1Wr9/f2yV6Nf\nexDs/uKj/3ybD99Vv/aNf/Nz/0x3u4q2fvZvjdtSD7tdRVuvTj1x57e/5HYVnbx0xxc+ctPX\n3K6irRff+PTRo993u4q2SqWbP/ChZber6OQHrxz/1i/twMsku+R0144xAAAgAElEQVTDfx04\n9/o33K6irdPv/+g//WrV7So6+fJnB/61/oLbVbT1L4y7CXau67l37ESkYcq1r6+vXq/bCcl+\ndy2Tyfh8vnA4nMlkmh+Sy+XUi3rhcLjlW3qZTCYcDqvX+KrVqv3whk+xf67VauqByWRSHS8W\ni8lkUp1tVwYAAMCe6a1gl06nRWRubs5OWu1YljU2Nmb/kEgknGeTyWQoFJqdnVUXhEKhhgsS\nicTY2Jh6mc8wjImJiVqt1vkT5+fn1QMPHz6sHhsMBuPxuLMesh0AAHBRbwW7SCSi6/rU1NSh\nQ4disVgmkymXyy2vnJubW1tbq9fra2trmqYZhpHP59WpYrEYj8d1Xd/Y2KjX6xsbG7quG4Zh\nr3XN5/OGYdgXpFIpy7KeeeYZccycOscIlY2NDXW9etUvHA6LyMrKirpyZWVFRFTWBAAAcEVv\nBTsRmZmZKRQKqVRqdnZ2bGzsyJEj4XC4uf/IzMzM4OCgiAwODs7MzIjI8vLfv/hy6dIlEYnH\n42pKt6+vT42rqeP2lePj4+qCz3/+8yJy8eLFzoXZ1ysqzw0PD6tf7R8AAADc0nPBTkQCgcDk\n5GS9Xl9dXV1aWhKRYDDYMHQXCAQafrZnRdUP/f39vreptRcNF/j9fvWreodPfVAH9vVO1Wq1\nWCxaltUw1QsAALD33uV2AZ34/X6/3x8IBCzLOnv27IULF9yu6B0SiYRhGG5XAQAA8Pd6aMSu\n3VYNaspVLVzoXr2VnSlURETm5uYMw4hGo9lstlAorK+v7+DDAQAAtqCHgp1pmiLS3JpErVfV\nNM150Dkzq97AS6VSzud02BZM13XnE2q1mupXsqlqp6amROTChQsjIyPOeWEAAAC39FCwU9Et\nFArlcjm7+Ui5XFZB7eGH39GTNh6PVyoVEalUKur9tvvuu0+dOn78uIgkEgl1gYjkcjln/7lP\nfOITIrKwsKA+Ra2HPXXqlPP51+23Ypenajh37twWvjIAAMAO6qF37Px+fzqdHhsbC4VCDad0\nXR8ZGXEeCQQCQ0ND9q+madrDZsPDw6q/iXPPWU3TxsfH1c8jIyPqAvsNOU3TTpw4oX6ORqOz\ns7OHDh3SNK3DigpV6pEjRxqOl8vllsssAAAAdltvbSkmIpVK5dlnn7UsS8UyXdc/+9nPOpuJ\n2Ft1JZPJeDwejUYffPDBhtgnIrlcbnFxUb2Zl0qlRkdH1eaztkwmo9rOmaY5Pj5uny2Xy2fP\nnp2dnbWDXbvNwdTOtqrI8fHxH//4x8Fg0DTN6enpDnexVywAwKvYUsx1PRfsrmun9nIFAADw\nmB56xw4AAADb0UPv2DVo2fpELWjt8t56va5mS+3p0WZqPjeVSk1OTrYcC6zValeuXHnhhRfs\n/samad59993Nk79dYioWAOBVTMW6rnenYlsGO1vnsp0RLRaLzc7OZrPZlu/hhUKhaDSqWh83\nB7tisZhIJJyLMGyaps3Pzze8t9eNUqn0jbt+c7N37aWPXv2jd42fdLuKtv524anbUg9f/zqX\nvDr1xJ3f/pLbVXTy0h1f+MhNX3O7irZefOPTR49+3+0q2iqVbr7hlt7964nIW//p01dv6K1e\n7k53vRU7V33C7SraOj3w8KeWe3oboWePz9z1byfcrqKtq/9snmDnul6fim3oMJxOp6XrcTvl\nkUceEZFQKNTQwaRararlt+qCZsViMRgMWpaVSqVWV1ftGlZXV03TtCxrYqJ3/+kCAAAHUK8H\nuwaRSERENrWR1+DgoFrc2tBqTv26tLSkdrZoptrjFQqFyclJZwcTv98/PT2dSqUsy2pupwwA\nAOCWfRbs2slkMuFw2OfzZTKZ5rOapjV0trMsyzAMXdcbNrSw5fN5y7Kc7fEajI6OptPpW265\nRf2q9kOrVqvJZFLtY9FyAhcAAGD39O7iiZZUbmuYik0kEvYY3tjY2LVr15pvPH36tGEY4XC4\nXq/XajW1gdjp06fbfdDy8rI07WPmNDAwoIYPnSYmJlSeU334lpaWOjwBAABgZ/X6iJ3vnVRL\n4cnJSfuCfD5vGIamaWtra/V6fW1t7fLly83PGRgYyGazImJZltpDLJvNdlj6oNbAbnYPiUAg\nsLGxUa/X1Wc9/fTTm7odAABgO3o92DXQdX1tbc35VpwaWjNNUx0cHBycmZlpea/aSSwcDo+N\njTXvUdYNXyvOC8bHx/v6+tRniYja9wIAAGBv9Hqws9eiRqNRTdNOnjzZsNaheWit3Vtx9sXO\nH3YWu8QCAAAX9Xqws505c8ayrGAw2NC1ZFPUcJrzh3ZM0xSRcrnsPNjQe2XLZQAAAOyGfRPs\n/H6/ClsLCwt78HHHjx8XkZav6wEAAPSmfRPsROTBBx8UkXg8XqlU7IMq7RWLRftIwzDb1gwP\nD2uaNjU15XyyU61W2/6nAAAA7KD9FOwGBwdVjFtcXLQPfvKTnxSRRCKh0l6lUtmp9+fUIoxg\nMJhMJp1hsVwuZzKZ/v5+6dgPBQAAYI/1+l6xDeXVajWVqJxrY5PJpDPMpVKpqamp5ns7PLbd\nqXK5HI/H27UaTqVSduOV5nvbfVCpVGr5NAAA9jv2inXdPgt2IpLJZMbGxkzTnJ6etg9aljU3\nN2dZVjqdjkQim0pv1z2Vz+eXl5cvX76sEp6u6/fdd9/999/vbIPXfbADAADYJb0b7AAAALAp\n+2BLsYYmwIqu6+Pj49ftG2cPm83NzU1NTTWM8zmp+Vw1u9ow2NZcQJef3hJTsQAAr2Iq1nX7\nYMSuZbBTVldXO6crZ0SLxWKzs7PZbLZ5z4lcLhcKhaLR6IULF6RpFrVdAQ17YHSpVCp9467f\n3Oxde+mjV//oXeMn3a6irb9deOq21MNuV9HWq1NP3PntL7ldRScv3fGFj9yUdbuKtl58I3T0\n6PfdrqKtUunmm269w+0qOnnjO9/+1g1tN0t03Yffqn65/DO3q2jrn/p/7gt/uOZ2FZ186beG\nfut3nnO7irb+8Pc/RrBz3b5ZFdvQHDidTssme9o98sgjIhIKhRpaHFer1VAoZF/QTQHq0+fm\n5rbwRQAAAHbJvgl2DSKRiIgYhtH9LYODg0tLSyJy7tw553H169LSUvfDb1v4dAAAgN22X4Nd\nO5lMJhwO+3y+TCbTfFbTNF3XDcOwO5hYlmUYhq7rW+hI57zF5/M1zNg2HwEAANhV+2DxREsq\nt+m67jyYSCTsUbSxsbFr164133j69GnDMMLhcL1er9Vq4XBYHdzCpz/8cO++6QUAAA6gfRPs\nWo5+2f2BRSSfzxuGoWna+fPnBwcHK5XKQw891HzLwMBANpsNhUKWZb355psiks1mnR3puiwg\nnU43L8IAAABw0X6ditV1vWFR6vLysoiYpqkODg4Oqj3Bmo2MjOi6Hg6Hx8bGdF3fWj578cUX\nnVvWAgAAuG7fBDt7RWo0GtU07eTJkw1rHdSuYs7uJ4FAoN3T7C3Iut9Y1rkqtlAoFIvFoaEh\n5x6yAAAA7to3wc525swZy7KCwWBD15JN6evra/hhUwKBwPnz52WT/VYAAAB21f4Ldn6/3zRN\ncTtUqfFCOp4AAIDesf+CnYg8+OCDIhKPx51vuam0VywW7SO7Ok9aq9XknR1PnJiiBQAAe2/f\nbCnWUKfa2tW592uxWAwGgw2rYlW/upbfseVjW55qvrJWq5mmaRhGOp1WzYqd+5XZZ1s+n71i\nAQBexZZirtuvwa5Wq/X398s7N2xVaa/5Cbquj4+PN+wq2/xY+4jzlP1BzTRNW1hYUG/pqd1m\n7VOmaapKev/PCwAAPGNfTsWKSF9fn9qwdXFx0T5oj941MAzjyJEjW5geLRaL4+PjLU/dc889\nZ8+etddejIyMLC0tqZnZVCrVrhIAAIDdsw9G7Dal5fBeJpNRLevadbZrea+a2xWRVCr18Y9/\n3B7wK5fLlmXF43FN09Tms5vCVCwAwKuYinXdgQh2HY53uCYcDluWVSgUWvbDm5ubm5qaUi/V\nbarCUql0/q7PbeqWPfbQ1a+8+usRt6to67Y/yfx88lG3q2jr76Yfv+0bf+x2FZ28+tHfuPPd\nP3K7irZe+sl7e/l/GEql0k033eR2FZ288cYbr732mttVtPWrv/qrX/nKV9yuoq3Pfe5zjz7a\nu/96EZHHH3/8U5/6lNtVtPXss8/28j+/B8R+nYrdEZlMJhwO+3w+tferUz6ftyzLNM12XY5H\nR0fT6fQtt9xSq9V8Pl8sFmu4oN1xAACAXbJv9ordDpXbdF13HkwkEnYXurGxsWvXrjnPqg3K\n2nUzEZGBgQG1HlbeXipx5swZ5/qMb37zmyLymc98Zme+AwAAwPV4c8TO905jY2MiMjk5aV+Q\nz+cNw9A0bW1trV6vr62tXb582fmE5g3KOvjkJz8pIg1PuHr1qojcf//92/42AAAAXfFmsGug\n67qzK4q8PSBnmqY6ODg42HldheJrRUQCgUA0Gp2amlJdi5WpqaloNDowMLDz3wcAAKAVbwa7\n+tui0aimaSdPnnSmOmk1INfuXbouqc0wnnvuOfWraq3CPCwAANhL3gx2tjNnzliWFQwGq9Xq\npm5UG5Q1tL6rv5Pz1L333isic3Nz6tfV1VVhHhYAAOwtjwc7v9+vItrCwsKmbjx+/Lg0vTbX\ngWqYbFmW2qz26aefZh4WAADsMY8HO3l7kjQej1cqFfugSnsqhCkNg3PDw8Oapk1NTTmvcXK+\nTqc88MADIvLUU09VKpXZ2VnmYQEAwB7zfrAbHBxUMc65+Zhax5pIJFTaq1QqzZvMquUUwWAw\nmUw6Y1+5XM5kMmoDWWc/lMHBwWg0ahjGQw89JMzDAgCAPef9YCciExMT8s5Bu0AgYJqmZVlD\nQ0M+n29oaKi5ZV0gEFhdXdU0LR6PHzlyxF4Ge+TIEdU/JZVKNWwpFo1GRcSyLOZhAQDA3jsQ\nwU69ACfvHLSbnp5eWlpSeS6dTju73Nn8fv/S0tLKyoppmnby03V9aWlpfX29+ZZbb71V/cA8\nLAAA2Hte2yvWXcViMRgMisj6+nq7EbtSqbS3RQEAsEfYK9Z1B2JLsT3z1FNPiUg2m+0wD8t/\n6QEAwC4h2O0MtQWFiOi6PjIy4m4xAADgYPJssLOTlpOu6+Pj49fdAVbda09SNz+q+TmaplmW\nZZrm9PR054czFQsA8CpmpVzn2XfsWgY7ZXV1tXO2u26wUxr2n+1SqVQ6f9fnNnvXXnro6lde\n/fWI21W0ddufZH4++ajbVbT1d9OP3/aNP3a7ik5e/ehv3PnuH7ldRVsv/eS9t99+u9tVtPXy\nyy//7Jc/6HYVnfzcD7/7p3/1HreraOsT73vz4f93w+0q2nriv+wPn1t1u4pOlk4fuWf8S25X\n0dbzC18g2LnO46tiGzYBU2tjN7sLRfOj1HPsDcQAAAB6gceDXYNIJCIihmH0yHMAAAB20MEK\ndu1kMplwOOzz+TKZTPd3OXsaq97FzrPNRwAAAHaVZxdPtKRym67rzoOJRMIeexsbG7t27VqX\nz3n44Yd3oUYAAIAt8niwazlm5twxIp/PG4ahadr58+cHBwcrlYra6fW6j0qn07Q1AQAAPeVg\nTcXqut6wlHV5eVlETNNUBwcHB2dmZrp51IsvvmjvPAsAANALPB7s7HWs0WhU07STJ082NCiJ\nx+Mi4ux+EggEOj+qXq8XCoVisTg0NFQul3e1fgAAgO55PNjZzpw5Y1lWMBisVqvbf1ogEDh/\n/rxstXMKAADAbjgowc7v95umKTsXxdTIHx1PAABA7zgowU5EHnzwQRGJx+POd+NU2isWi/aR\nLmdXa7WavLPjiRNTtAAAYO95fEuxhm+XTCbj8bhzR9disRgMBhtWxVqWJU1bijkfVavVTNM0\nDCOdTqtmxbFYbHZ2NpvNjoyM2GebCxD2igUAeBdbirnuYAW7Wq3W398v79zmVaU9+5pUKjU1\nNSVd7BWradrCwkJfX5+I5HK5UChknzJNUz3Tq39eAADQgw7QVKyI9PX1qW1eFxcX7YPT09NL\nS0tqUjWdTju73LWjaVo6nbZTnYiMjIzYD0mlUvaIIAAAwJ7x7Iid0jzYpmmapmmjo6MDAwOb\nelStVnvmmWcuX748OzsrItFo9OMf//iJEyfsbNclpmIBAF7FVKzrDlywUzRNm5+f7z7b5fP5\nY8eOtTxVKBTatb5rqVQqPRX9te6v33snZ//iO//qXreraOvWf3nlpj/+hNtVtPXGb/zpndee\ncLuKTl46/PBHbvqu21W09eIbH7z99tvdrqKtl19++Rdu/AduV9HJ3/z0L1/9Vu/+W/22D/v+\n6A9W3a6ird/87SOPnfyq21V08sWnPnvqrt91u4q2Ll79PYKd6w7EVKyzt/DGxoau65Zl5XK5\nLm8vFosq1aXT6fX1dfWc9fV1NasbDAadi2oBAADcciCCnVNfX59a1nDx4sVurq/VasFgUEQK\nhUIkErEH+QYGBiKRSKFQEJFEIqG6nwAAALjowAU7EVFvxameJiLi8/kaZmydR5577jkRSafT\nLedbA4FAOp22LOvKlSvOe2u1WiwW8/l8yWRy974IAACA00EMdp17CzdQrYYfeOCBdhfcc889\nIvLCCy84D87Pz6s1FocPH95OqQAAAN07cMGuWq2q3SZOnTrVzfVq3tZuetfM7/fbl9k2NjY2\nNjbq9bpqXwwAALAH3uV2AXuhZdOTEydO7N4njo+Pb7YNCgAAwDYdiGDnZPex29XgpYbxAAAA\n9tKBCHbb6dWn67phGJVKpd1sbKVSERE1vQsAAOCiA/eO3Wbdd999IvL1r3+93QXqFEN0AADA\ndQS7RmoZrO1jH/uYiIyNjbXsQlwsFsfGxjRNU5cBAAC4iGAn0WhURNRGFLVabWFhwXm2r69P\ndSEOBoOZTKZararj1Wo1k8mo3sWPPvooSyUAAIDrDsResZ2/Yy6XC4VC9q+maareJc67isWi\nynDNVlZWhoeHu//EUqnUbfUAAOwr7BXrOoKdiIhlWXNzc5ZlpVKpycnJlnfVarXnnnvu6aef\nVp2Ho9HoZz7zmY997GMNY3VdfiIAAMCO83iwAwAAODi80+6kZRdi1bJuYGCgw2Uiouv6+Ph4\ny5WttVrtypUrL7zwgr23hGmad99998jIyNbqZCoWAOBVTMW6zjsjdi0Tm4homjY/P29nu3aX\nicjq6mpDtisWi4lEwrKs6z62e6VS6fRdv7vZu/bSuau/t/Df/N9uV9HW+P/1X3/997/ldhVt\nPfA7H379G3/rdhWdvP+j77rZ9ytuV9HW9+vfu/32292uoq2XX375hg+6XURHb31Xrr77ittV\ntHXXT+79X779JberaOuRO77w6X8/4XYVnXztH8/fdebTblfR1tWzXyPYuc5rq2LrDhsbG7qu\nW5alVry2u6xer6fTaRFpWA+rFkyoF+9WV1fti1dXV03TtCxrYqKn//kHAAAHjdeCnVNfX5+a\nP7148WLnKyORiIgYhuE8mEgkRKRQKExOTjpH8vx+//T0dCqVahkZAQAA3OLlYCciaslqy7nU\nzvL5vGVZpmkGAoGWF4yOjqbT6VtuuUX96vP5fD5ftVpNJpM+ny8cDm/hQwEAALbD48GuVquJ\niKZpnS/LZDIiouu6fWR5ebnzjQMDA5FIpOGdvImJCTVGaFkW2Q4AAOwxLwe7arVqmqaInDp1\nquGU753GxsZEZHJy0r5A5bPN7gAbCAQ2Njbq9Xo2mxWRp59+eptfAQAAoHteC3bOuHbo0CHD\nMDRNO3HiROe7dF1fW1sbHBzs/uE25wXj4+Nq8lc1Q1GtjAEAAPaGd/rYNbP72DVv5Go3eYnF\nYteuXTt58uR1U103NjvCBwAAsIO8Fuw225bvzJkzR44csSxrfX3d2ZRO7RhbLpedWa3h4R1a\n4gEAAOw9r03Fbpbf71fv4TU0sTt+/LiIXL582Z2yAAAANu+gBzsRefDBB0UkHo9XKhX74PDw\nsKZpU1NTxWKx5V1qvS0AAEDv8NqWYtf9Oi0vSyaT8XjcNM3p6Wn7oNp5QkRM09Q0zZ6TLZfL\nzz//vFpIq2na0tJSy8e2q4e9YgEAXsWWYq4j2ImI1Gq1/v5+EWlYG1sul+PxeLt2dKlUyu6Q\n0n2wAwAA2CVMxYqI9PX1qe1iFxcXncf9fv/S0tLKyooatFMHdV1fWlpaX1939r0DAABwnXdW\nxTaPjTWvWtU0LZVKjY6ONlxTr9ffeOONdk8eHh4eHh4WEcuynKN0IlKr1a5cufLCCy/YTzNN\n8+677x4ZGWk3VsdULADAq5iKdZ13pmKbtWtHomna/Py8am7inDCNxWKzs7PZbFa1F3bK5XKh\nUCgajV64cME+WCwWE4lEy4la50c0KJVKf/LRL271O+2FX//GY3/16SfdrqKt933t8+/5d992\nu4q23vwndwx+90a3q+ik8sGf3nnTLW5X0dZLb/yn22+/3e0q2nr55Zc/8KG/c7uKTn7wys9f\ne1/vbmZ4+K+0p9b/B7eraOvkof/t0T8Nu11FJ49/Yunkv/mv3K6iraf++f9DsHOd96di6w4b\nGxu6rluWlcvlnGfVz4888oiIhEKharXqfEK1Wg2FQvYFilpaocbwVldX7Y9YXV01TdOyrImJ\niT36hgAAACJyEIKdU19fn9oE9uLFi81nBwcH1RLXc+fOOY+rX5eWlpzrKhKJhIgUCoXJyUln\nE2O/3z89PZ1KpZzxEQAAYA945x27Lqntxez504a1q5qm6bpuGMZ9992nVktYlmUYhq7r9uIJ\nEcnn85ZlmaYZCARafsro6OhNN910yy29O+EFAAC852CN2MnbjYWdKa3B6dOnRSQcDquL1Q/q\noG15ebnzQwYGBiKRCFvHAgCAvXSwRuyq1aqaVz116lS7awYGBrLZbCgUsizrzTffFJFsNtuw\nDELN55LbAABAT/F+sGvZ9OTEiRMdbhkZGdF1XY3V6brevEi2yw8SGhQDAIA9dLCmYlUfu/n5\nefWmXQdqTM75AwAAQI/z/ojd1sbM7OTXMgKaphmPx8vlsnM2tuGD2nXRAwAA2CUHa8Rupxw/\nflxELl++7HYhAAAA/xnBbiuGh4c1TZuamioWiy0vUGtvAQAA9hLBbotmZmZEJBgMJpPJcrls\nHy+Xy5lMpr+/Xzr2QwEAANhx3t8rtvMX7HDNdW8vl8vxeLzlXrEikkqlJicnm4+XSqWOVQMA\nsF+xV6zrCHZbD3ZKPp9fXl6+fPmySni6rt933333339/Q+s7AACA3eblYAcAAHCgeLPdSXOr\nEV3Xx8fHu9krYlP31mq1K1euvPDCC3a7O9M077777g49jZmKBQB4FVOxrvPmiF27HnJra2uD\ng4M7dW+xWEwkEi3fsdM0bX5+vuVsbKlUemz8ns41uOuLC88/+z/d73YVbX3qf/6z7/+HtjvC\nue7mf3Txva8/6XYVnfzo/Z//h7+y4XYVbf3H7/V/6EMfcruKtl555ZUfHfplt6vo5L3rP/yj\nt37odhVt/eYNv/zf/n+X3K6irf/zv/jkR8/9a7er6OQbp//FXbFxt6to6+qFBYKd67y8Krbu\nkE6nRWRubm6n7i0Wi8Fg0LKsVCq1urpqX7y6umqapmVZExMTO/6NAAAAOvBysHOKRCIiYhjG\nTt2bSCREpFAoTE5OOmdp/X7/9PR0KpWyLCuXy22raAAAgM04KMFOcTaW8/l8Pp+vVqvFYjGf\nz5dMJru/N5/PW5ZlmmYgEGh58ejoaDqdvuWWW3akbAAAgG54c/FEs0wmIyIPP/xww/H5+fnZ\n2VkROXz4cPf3Li8vS8f+wwMDA2qcDwAAYM94Odg1LINIp9PNi1U3NjY2Njb6+vo2da9aA9vN\nGlsAAIA9c4CmYl988cVKpdJwcHx8vDnVdXlvA18rWy8XAABgk7wc7JwrWwuFQrFYHBoacu7r\nKu1H3bq5FwAAoKd4Odg5BQKB8+fPi8jCwsL27zVNU0Qacl79nXagaAAAgM04KMFORFR74a11\nPGm49/jx4yJy+fLlnasOAABguw5QsKvVatJxKWv39w4PD2uaNjU1VSwWO1wPAACwl7y8pZjz\nq9VqNdM0DcNIp9OqEUnzNd3fK2/vPCEipmlqmma/q1cul59//vmxsTER0TRtaWmp4fnsFQsA\n8Cq2FHOdl4NdM03TFhYW1DLYzsGu871KuVyOx+Mt94oVkVQqNTk5uYXiAQAAtuagBDtN006d\nOnXixAk7mXUf7Jrvdcrn88vLy5cvX1YJT9f1++677/777x8YGNiR7wIAANAlbwY7AACAA+gA\nLZ4AAADwNoIdAACARxDsAAAAPIJgBwAA4BEEOwAAAI8g2AEAAHgEwQ4AAMAjCHYAAAAeQbAD\nAADwCIIdAACARxDsAAAAPIJgBwAA4BEEOwAAAI8g2AEAAHgEwQ4AAMAjCHYAAAAeQbADAADw\nCIIdAACARxDsAAAAPIJgBwAA4BEEOwAAAI8g2AEAAHgEwQ4AAMAjCHYAAAAeQbADAADwCIId\nAACARxDsAAAAPIJgBwAA4BEEOwAAAI8g2AEAAHgEwQ4AAMAjCHYAAAAeQbADAADwCIIdAACA\nRxDsAAAAPIJgBwAA4BEEOwAAAI8g2AEAAHgEwQ4AAMAjCHYAAAAeQbADAADwCIIdAACARxDs\nAAAAPIJgBwAA4BEEOwAAAI8g2AEAAHgEwQ4AAMAjCHYAAAAeQbADAADwCIIdAACARxDsAAAA\nPIJgBwAA4BEEOwAAAI8g2AEAAHgEwQ4AAMAjCHYAAAAeQbADAADwCIIdAACARxDsAAAAPIJg\nBwAA4BEEOwAAAI8g2AEAAHgEwQ4AAMAjCHYAAAAeQbADAADwCIIdAACARxDsAAAAPIJgBwAA\n4BEEOwAAAI8g2AEAAHgEwQ4AAMAjCHYAAAAeQbADAADwCIIdAACARxDsAAAAPIJgBwAA4BEE\nOwAAAI8g2AEAAHgEwQ4AAMAjCHYAAAAeQbADAADwCIIdAHqqqdwAAAGISURBVACARxDsAAAA\nPIJgBwAA4BEEOwAAAI8g2AEAAHgEwQ4AAMAjCHYAAAAeQbADAADwCIIdAACARxDsAAAAPIJg\nBwAA4BEEOwAAAI8g2AEAAHgEwQ4AAMAjCHYAAAAeQbADAADwCIIdAACARxDsAAAAPIJgBwAA\n4BEEOwAAAI8g2AEAAHgEwQ4AAMAjCHYAAAAeQbADAADwCIIdAACARxDsAAAAPIJgBwAA4BEE\nOwAAAI8g2AEAAHgEwQ4AAMAjCHYAAAAeQbADAADwCIIdAACARxDsAAAAPIJgBwAA4BEEOwAA\nAI8g2AEAAHgEwQ4AAMAjCHYAAAAeQbADAADwCIIdAACARxDsAAAAPIJgBwAA4BEEOwAAAI8g\n2AEAAHgEwQ4AAMAjCHYAAAAeQbADAADwCIIdAACARxDsAAAAPIJgBwAA4BEEOwAAAI8g2AEA\nAHgEwQ4AAMAjCHYAAAAeQbADAADwCIIdAACARxDsAAAAPIJgBwAA4BEEOwAAAI8g2AEAAHjE\n/w9CPUk7TwVT+QAAAABJRU5ErkJggg==",
      "text/plain": [
       "plot without title"
      ]
     },
     "metadata": {
      "image/png": {
       "height": 420,
       "width": 420
      }
     },
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'#FFF5EB'</li><li>'#FEE6CE'</li><li>'#FDD0A2'</li><li>'#FDAE6B'</li><li>'#FD8D3C'</li><li>'#F16913'</li><li>'#D94801'</li><li>'#A63603'</li><li>'#7F2704'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item '\\#FFF5EB'\n",
       "\\item '\\#FEE6CE'\n",
       "\\item '\\#FDD0A2'\n",
       "\\item '\\#FDAE6B'\n",
       "\\item '\\#FD8D3C'\n",
       "\\item '\\#F16913'\n",
       "\\item '\\#D94801'\n",
       "\\item '\\#A63603'\n",
       "\\item '\\#7F2704'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. '#FFF5EB'\n",
       "2. '#FEE6CE'\n",
       "3. '#FDD0A2'\n",
       "4. '#FDAE6B'\n",
       "5. '#FD8D3C'\n",
       "6. '#F16913'\n",
       "7. '#D94801'\n",
       "8. '#A63603'\n",
       "9. '#7F2704'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"#FFF5EB\" \"#FEE6CE\" \"#FDD0A2\" \"#FDAE6B\" \"#FD8D3C\" \"#F16913\" \"#D94801\"\n",
       "[8] \"#A63603\" \"#7F2704\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0gAAANICAMAAADKOT/pAAAAS1BMVEUAAABNTU1oaGh8fHx/\nJwSMjIyampqmNgOnp6eysrK9vb3Hx8fQ0NDZSAHZ2dnh4eHp6enw8PDxaRP9jTz9rmv90KL+\n5s7/9ev///8SGHwUAAAACXBIWXMAABJ0AAASdAHeZh94AAAReklEQVR4nO3b61LiWgKA0Qij\n44VBmavv/6RjgpKATntaPs50tev7YaV30psUZhVJiMOzpLMb/t87IP0OgSQFgSQFgSQFgSQF\ngSQFgSQFgSQFgSQFgSQFgSQFgSQFgSQFgSQFgSQFgSQFgSQFgSQFgSQFgSQFgSQFgSQFgSQF\ngSQFgSQFgSQFgSQFgSQFgSQFgSQFgSQFgSQFgSQFgSQFgSQFgSQFgSQFgSQFgSQFgSQFgSQF\ngSQFgSQFgSQFgSQFgSQFgSQFgSQFgSQFgSQFgSQFgSQFgSQFgSQFgSQFgSQFgSQFgSQFgSQF\ngSQFgSQFgSQFgSQFgSQFgSQFgSQFgSQFgSQFgSQFgSQFgSQFgSQFgSQFgSQFgSQFgSQFgSQF\ngSQFgSQFgSQFgSQFgSQFgSQFgSQFgSQFgSQFgSQFgSQFgSQFgSQFgSQFgSQFgSQFgSQFgSQF\ngSQFgSQF/QmQ/lP277B/lf2z7B9lfy/7W9hfy/5S9oWjHCSQQAIJJJBAAgkkkP54IIEEUhBI\nIIEUBBJIIAWBBBJIQSCBBFIQSCCBFAQSSCAFgQQSSEEggQRSEEgggRQEEkggBYEEEkhBIIEE\nUhBIIIEUBBJIIAWBBBJIQSCBBFIQSCCBFAQSSCAFgQQSSEEggQRSEEgggRQEEkggBYEEEkhB\nIIEEUhBIIIEUBBJIIAWBBBJIQSCBBFIQSCCBFAQSSCAFgQQSSEEggQRSEEgggRQEEkggBYEE\nEkhBIIEEUhBIIIEUBBJIIAWBBBJIQSCBBFIQSCCBFAQSSCAFgQQSSEEggQRSEEgggRQEEkgg\nBYEEEkhBIIEEUhBIIIEUBBJIIAWBBBJIQSCBBFIQSCCBFAQSSCAFgQQSSEEggQRSEEgggRQE\nEkggBYEEEkhBIIEEUhBIIIEUBBJIIAWBBBJIQSCBBFIQSCCBFAQSSCAFgQQSSEEggQRSEEgg\ngRQEEkggBYEEEkhBIIEEUhBIIIEUBBJIIAWBBBJIQSCBBFIQSCCBFAQSSCAFgQQSSEEggQRS\nEEgggRQEEkggBYEEEkhBIIEEUhBIIIEUBBJIIAWBBBJIQSCBBFIQSCCBFAQSSCAFgQQSSEEg\ngQRSEEgggRQEEkggBYEEEkhBIIEEUhBIIIEUBBJIIAWBBBJIQSCBBFIQSCCBFAQSSCAFgQQS\nSEEggQRSEEgggRQEEkggBYEEEkhBIIEEUhBIIIEUBBJIIAWBBBJIQSCBBFIQSCCBFAQSSCAF\ngQQSSEEggQRSEEgggRQEEkggBYEEEkhBIIEEUhBIIIEUBBJIIAWBBBJIQSCBBFIQSCCBFAQS\nSCAFgQQSSEEggQRSEEgggRQEEkggBYEEEkhBIIEEUhBIIIEUBBJIIAWBBBJIQSCBBFIQSCCB\nFAQSSCAFgQQSSEEggQRSEEgggRQEEkggBYEEEkhBIIEEUhBIIIEUBBJIIAWBBBJIQSCBBFIQ\nSCCBFAQSSCAFgQQSSEEggQRSEEgggRQEEkggBYEEEkhBIIEEUhBIIIEUBBJIIAWBBBJIQSCB\nBFIQSCCBFAQSSCAFgQQSSEEggQRSEEgggRQEEkggBYEEEkhBIIEEUhBIIIEUBBJIIAWBBBJI\nQSCBBFIQSCCBFAQSSCAFgQQSSEEggQRSEEgggRQEEkggBYEEEkhBIIEEUhBIIIEUBBJIIAWB\nBBJIQSCBBFIQSCCBFAQSSCAFgQQSSEEggQRSEEgggRQEEkggBYEEEkhBIIEEUhBIIIEUBBJI\nIAWBBBJIQSCBBFIQSCCBFAQSSCAFgQQSSEEggQRSEEgggRQEEkggBYEEEkhBIIEEUhBIIIEU\nBBJIIAWBBBJIQSCBBFIQSCCBFAQSSCAFgQQSSEEggQRSEEgggRQEEkggBYEEEkhBIIEEUhBI\nIIEUBBJIIAWBBBJIQSCBBFIQSCCBFAQSSCAFgQQSSEEggQRSEEgggRQEEkggBYEEEkhBIIEE\nUhBIIIEUBBJIIAWBBBJIQSCBBFIQSCCBFAQSSCAFgQQSSEEggQRSEEgggRQEEkggBYEEEkhB\nIIEEUhBIIIEUBBJIIAWBBBJIQSCBBFIQSCCBFAQSSCAFgQQSSEEggQRSEEgggRQEEkggBYEE\nEkhBIIEEUhBIIIEUBBJIIAWBBBJIQSCBBFIQSCCBFAQSSCAFgQQSSEEggQRSEEgggRQEEkgg\nBYEEEkhBIIEEUhBIIIEUBBJIIAWBBBJIQSCBBFIQSCCBFAQSSCAFgQQSSEEggQRSEEgggRQE\nEkggBYEEEkhBIIEEUhBIIIEUBBJIIAWBBBJIQSCBBFIQSCCBFAQSSCAFgQQSSEEggQRSEEgg\ngRQEEkggBYEEEkhBIIEEUhBIIIEUBBJIIAWBBBJIQSCBBFIQSCCBFAQSSCAFgQQSSEEggQRS\nEEgggRQEEkggBYEEEkhBIIEEUhBIIIEUBBJIIAWBBBJIQSCBBFIQSCCBFAQSSCAFgQQSSEEg\ngQRSEEgggRQEEkggBYEEEkhBIIEEUhBIIIEUBBJIIAWBBBJIQSCBBFIQSCCBFAQSSCAFgQQS\nSEEggQRSEEgggRQEEkggBYEEEkhBIIEEUhBIIIEUBBJIIAWBBBJIQSCBBFIQSCCBFAQSSCAF\ngQQSSEEggQRSEEgggRQEEkggBYEEEkhBIIEEUhBIIIEUBBJIIAWBBBJIQSCBBFIQSCCBFAQS\nSCAFgQQSSEEggQRSEEgggRQEEkggBYEEEkhBIIEEUhBIIIEUBBJIIAWBBBJIQSCBBFIQSCCB\nFAQSSCAFgQQSSEEggQRSEEgggRQEEkggBYEEEkhBIIEEUhBIIIEUBBJIIAWBBBJIQSCBBFIQ\nSCCBFAQSSCAFgQQSSEEggQRSEEgggRQEEkggBYEEEkhBIIEEUhBIIIEUBBJIIAWBBBJIQSCB\nBFIQSCCBFAQSSCAFgQQSSEEggQRSEEgggRQEEkggBYEEEkhBIIEEUhBIIIEUBBJIIAWBBBJI\nQSCBBFIQSCCBFAQSSCAFgQQSSEEggQRSEEgggRQEEkggBYEEEkhBIIEEUhBIIIEUBBJIIAWB\nBBJIQSCBBFIQSCCBFAQSSCAFgQQSSEEggQRSEEgggRQEEkggBYEEEkhBIIEEUhBIIIEUBBJI\nIAWBBBJIQSCBBFIQSCCBFAQSSCAFgQQSSEEggQRSEEgggRQEEkggBYEEEkhBIIEEUhBIIIEU\nBBJIIAWBBBJIQSCBBFIQSCCBFAQSSCAFgQQSSEEggQRSEEgggRQEEkggBYEEEkhBIIEEUhBI\nIIEUBBJIIAWBBBJIQSCBBFIQSCCBFAQSSCAFgQQSSEEggQRSEEgggRQEEkggBYEEEkhBIIEE\nUhBIIIEUBBJIIAWBBBJIQSCBBFIQSCCBFAQSSCAFgQQSSEEggQRSEEgggRQEEkggBYEEEkhB\nIIEEUhBIIIEUBBJIIAWBBBJIQSCBBFIQSCCBFAQSSCAFgQQSSEEggQRSEEgggRQEEkggBYEE\nEkhBIIEEUhBIIIEUBBJIIAWBBBJIQSCBBFIQSCCBFAQSSCAFgQQSSEEggQRSEEgggRQEEkgg\nBYEEEkhBIIEEUhBIIIEUBBJIIAWBBBJIQSCBBFIQSCCBFAQSSCAFgQQSSEEggQRSEEgggRQE\nEkggBYEEEkhBIIEEUhBIIIEUBBJIIAWBBBJIQSCBBFIQSCCBFAQSSCAFgQQSSEEggQRSEEgg\ngRQEEkggBYEEEkhBIIEEUhBIIIEUBBJIIAWBBBJIQSCBBFIQSCCBFAQSSCAFgQQSSEEggQRS\nEEgggRQEEkggBYEEEkhBIIEEUhBIIIEUBBJIIAWBBBJIQSCBBFIQSCCBFAQSSCAFgQQSSEEg\ngQRSEEgggRQEEkggBYEEEkhBIIEEUhBIIIEUBBJIIAWBBBJIQSCBBFIQSCCBFAQSSCAFgQQS\nSEEggQRSEEgggRQEEkggBYEEEkhBIIEEUhBIIIEUBBJIIAWBBBJIQSCBBFIQSCCBFAQSSCAF\ngQQSSEEggQRSEEgggRQEEkggBYEEEkhBIIEEUhBIIIEUBBJIIAWBBBJIQSCBBFIQSCCBFAQS\nSCAFgQQSSEEggQRSEEgggRQEEkggBYEEEkhBIIEEUhBIIIEUBBJIIAWBBBJIQSCBBFIQSCCB\nFAQSSCAFgQQSSEEggQRSEEgggRQEEkggBYEEEkhBIIEEUhBIIIEUBBJIIAWBBBJIQSCBBFIQ\nSCCBFAQSSCAFgQQSSEEggQRSEEgggRQEEkggBYEEEkhBIIEEUhBIIIEUBBJIIAWBBBJIQSCB\nBFIQSCCBFAQSSCAFgQQSSEEggQRSEEgggRQEEkggBYEEEkhBIIEEUhBIIIEUBBJIIAWBBBJI\nQSCBBFIQSCCBFAQSSCAFgQQSSEEggQRSEEgggRQEEkggBYEEEkhBIIEEUhBIIIEUBBJIIEn6\nA4EkBYEkBYEkBYEkBYEkBYEkBYEkBYEkBYEkBYEkBYEkBYEkBYEkBYEkBYEkBYEkBYEkBYEk\nBYEkBYEkBYEkBYEkBYEkBYEkBYEkBYEkBYEkBYEkBYEkBYEkBYEkBYEkBYEkBYEkBYEkBYEk\nBYEkBYEkBYEkBYEkBYEkBYEkBYEkBYEkBYEkBYEkBYEkBYEkBYEkBYEkBYEkBYEkBYEkBYEk\nBYEkBYEkBYEkBYEkBYEkBYEkBYEkBYEkBYEkBYEkBYEkBYEkBYEkBYEkBYEkBYEkBYEkBYEk\nBYEkBYEkBYEkBYEkBYF0+Xa3q2FY3e66GTfXwRzjj+H49z/+83pz/tzfMJAu3sPwWnaEPg3n\no1xNv/kPIO2Gp7Mn/4aBdOk2w3D7cmg+3XaS1rfnzzF89JufBm/X58/+/QLpwu0OfjbDVXN2\n9xB8IP0A0m54OH/6bxdIF+5uOHx83A7347G6Ww3X42XOMFxNq4bh6Xq4uttvcvWy+f4Yv18N\nV/fT4GY9DOvFp9lqfTo8bzvOcDfNMCzO3Q7rD681nWu+rl7uyvhzvbrIO/F7B9KFux4e3xa3\nI6BheDlsb198TY2H78tRPC6Oktbjws10OF9P60cy9/tN7+dppsXF8Lztfoa7E0jz+sNrLSAd\n7cp+5u2f8978ToF04ZanUPtDd72bFh+m2xDPryP3w2o6+Xt8frwaRzfj4G49nhZejRQfhsPH\nxO2e5jy82PZhnmGGtFg/v9a8+mhXxsHHIbgG+26BdOHeQ9qerNyPTHeep8upzX5x5Lbbf4Yd\n36RY7y+R5uHFtosZZilHc7291vLMb7Erz9OWbjf8dCBduPeQXv/xtLlbL47exbr94mvjJ9Bw\n/fj4bsZ5eLHt0QwfzHX6WvufJ7vy8Y0I/TBv2YVbz9dIj/vPhLfx12P7U0jPd+N1zdXh2523\nGQ7D50I63RWQvpC37MId3bW7OxykN8PqfvP0I0jLSTa3q/kaaV71OrzY9n9AOll/vPrdroD0\nhbxlF+70e6TlsXoK6ega6eTb2/ngXi+/RjrZdr+4PUjZnqz/CNK7XXGN9IVAunSLJxvGLzoP\nkLbPj6fXSIu7dtPtt+f78WRwtb+rtrhrN92umIcX2y5mWA33462647mWfJ4OSye7snXX7ucD\n6eJt3i5RpgcG5lsFU9uPrlbmxfES6OGw5b7t9JXTcnjedj/vpGL6nun6ZK75tVYvA/uld7ty\n53uknw+ky7e7G5/+vtufkL2dot28HO/bzfGnxPRcwnq7X7x/+U830x2G6RGGxbG9fLJhPzxv\nOy5eP7/do7h5Pplrfq3t6g3Su13xZMMXAunX67NLlM2nz2efdbfgqXtM/RsF0i/UdPa3u/70\nEuXTp7/PguTp768E0i/U61NvV59t9+nfI50Dyd8jfSmQfqXu1+Of0n6+3ebmx+vPgXTjxO4r\ngSQFgSQFgSQFgSQFgSQFgSQFgSQFgSQFgSQFgSQFgSQFgSQFgSQFgSQFgSQFgSQFgSQFgSQF\ngSQFgSQFgSQFgSQFgSQFgSQFgSQFgSQFgSQF/ReCEwbFex8W2gAAAABJRU5ErkJggg==",
      "text/plain": [
       "plot without title"
      ]
     },
     "metadata": {
      "image/png": {
       "height": 420,
       "width": 420
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "library(RColorBrewer)\n",
    "# install.packages(\"RColorBrewer\")\n",
    "\n",
    "display.brewer.all()\n",
    "display.brewer.pal(9, \"Oranges\")\n",
    "brewer.pal(9, \"Oranges\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### 肠癌"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 差异分析"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(DESeq2)\n",
    "load(\"stat/hg38.quantify.FeatureCount.RData\")\n",
    "load(\"stat/hg38.gene.transcript.gtf.RData\")\n",
    "\n",
    "\n",
    "Sample_info <- read.table(\"stat/Sample_info2.txt\", header = T, stringsAsFactors = F)\n",
    "rownames(Sample_info) <- Sample_info$Samples\n",
    "IDs <- Sample_info$Samples[(Sample_info$CellLine %in% c(\"HCT116\", \"NCM460\")) & (Sample_info$SeqType %in% \"sgRNA\")]\n",
    "\n",
    "# sgRNA count >= 3 at least three sample\n",
    "sub.gene.count <- gene.count[, IDs]\n",
    "logic <- apply(sub.gene.count, 1, function(x){sum(x >= 3)>=3})\n",
    "sub.gene.count <- sub.gene.count[logic,]\n",
    "dim(sub.gene.count) # 16558\n",
    "sub.gene.count <- round(sub.gene.count)\n",
    "\n",
    "# run DESeq2\n",
    "colInfo <- Sample_info[IDs, c(\"CellLine\"), drop=F]\n",
    "dds <- DESeqDataSetFromMatrix(sub.gene.count, \n",
    "                                colInfo, \n",
    "                                ~CellLine)\n",
    "Object <- DESeq(dds\n",
    "                # parallel = T, BPPARAM=MulticoreParam(10)\n",
    "                )\n",
    "gene.norlCount <- counts(Object, normalized=TRUE)\n",
    "res <- results(Object, contrast=c(\"CellLine\", \"HCT116\", \"NCM460\"))\n",
    "# res <- res[complete.cases(res),]\n",
    "# dif_res <- res\n",
    "# dif_res <- res[res$padj<0.05 & abs(res$log2FoldChange)>1 & res$baseMean>10, ]\n",
    "\n",
    "save(Object, gene.norlCount, res, file=\"stat/HCT116_vs_NCM460_gene_object.RData\")\n",
    "\n",
    "\n",
    "\n",
    "out_res <- data.frame(gene.norlCount, res[, c(\"baseMean\", \"log2FoldChange\", \"pvalue\", \"padj\")])\n",
    "out_res$dif <- \"ns\"\n",
    "out_res$dif[res$padj<0.05 & res$log2FoldChange>1] <- \"Up\"\n",
    "out_res$dif[res$padj<0.05 & res$log2FoldChange < -1] <- \"Down\"\n",
    "table(out_res$dif)\n",
    "out_res <- out_res[order(out_res$dif, decreasing=T),]\n",
    "out_res <- data.frame(gene_id=rownames(out_res), out_res)\n",
    "out_res <- data.frame(out_res, gene_name=gene_gtf[rownames(out_res), \"gene_name\"])\n",
    "gene_res <- out_res\n",
    "# out_res <- gene_res[transcript_res$padj<0.05 & abs(transcript_res$log2FoldChange)>2 & transcript_res$baseMean>10, ]\n",
    "# write.table(out_res, \"stat/MDAMB231_vs_MCF10A_differential_genes.txt\", row.names = F, col.names = T, quote = F, sep=\"\\t\")\n",
    "write.table(out_res, \"stat/HCT116_vs_NCM460_all_genes.txt\", row.names = F, col.names = T, quote = F, sep=\"\\t\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": [],
    "toc-hr-collapsed": true
   },
   "source": [
    "# Aseembly: stringTie "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "heading_collapsed": "true",
    "tags": [],
    "toc-hr-collapsed": true
   },
   "source": [
    "## plotting assembly stat "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### sensitivity, precision and F1 score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in file(file, \"rt\"):\n",
      "\"无法打开文件'stringtie_res/totalRNA.gffcmp.stats_sensitivity_precision.txt': No such file or directory\"\n"
     ]
    },
    {
     "ename": "ERROR",
     "evalue": "Error in file(file, \"rt\"): 无法打开链结\n",
     "output_type": "error",
     "traceback": [
      "Error in file(file, \"rt\"): 无法打开链结\nTraceback:\n",
      "1. read.table(statFile, header = T)",
      "2. file(file, \"rt\")"
     ]
    }
   ],
   "source": [
    "setwd(\"/NAS/wg_looking/gRNA_ONT\")\n",
    "\n",
    "## assembly stat\n",
    "IDs = c(\"sgRNA_HCT116\", \"totalRNA_HCT116\", \"sgRNA_NCM460\", \"totalRNA_NCM460\", \n",
    "       \"totalRNA_K562\", \"sgRNA_K562\", \"sgRNA_MDAMB231\", \"sgRNA_MCF10A\")\n",
    "\n",
    "# IDs = c(\"sgRNA_HCT116_NCM460\")\n",
    "\n",
    "# minCoverage\n",
    "for(ID in IDs){\n",
    "    statFile = paste0(\"stringtie_res5/\", ID, \".gffcmp.stats_sensitivity_precision.txt\")\n",
    "    stat.tb <- read.table(statFile, header=T)\n",
    "    Types <- c(\"Base_level\", \"Exon_level\", \"Transcript_level\")\n",
    "    pp <- stat.tb[stat.tb$Type %in% Types,]\n",
    "    pp$Max <- pp$minCoverage\n",
    "    for(Type in Types){\n",
    "        tmp <- stat.tb[stat.tb$Type %in% Type & stat.tb$stats %in% \"F1_score\", ]\n",
    "        cov <- tmp$minCoverage[tmp$Percent==max(tmp$Percent)]\n",
    "        pp$Max[pp$Type %in% Type & !(pp$minCoverage %in% cov)] <- NA\n",
    "    }\n",
    "\n",
    "    pp$minCoverage <- factor(pp$minCoverage)\n",
    "    ggplot(pp, aes(x=minCoverage, y=Percent, color=stats)) +\n",
    "        geom_point()+\n",
    "        scale_color_manual(values=c(F1_score=\"#4678BA\", Sensitivity=\"#87B07A\", Precision=\"#EF8354\")) + \n",
    "        ggtitle(ID) +\n",
    "        geom_vline(aes(xintercept = Max), linetype=\"dashed\", color=\"red\") +\n",
    "        facet_wrap(~Type, nrow = 3) +\n",
    "        # guides(color=\"none\") +\n",
    "        # ylab(\"F1 score\") +\n",
    "        theme_bw() +\n",
    "        theme(\n",
    "          panel.grid=element_blank(),\n",
    "          axis.text.x = element_text(angle = 0, color = \"black\", \n",
    "                                     size = 8, hjust = 0.5),\n",
    "          plot.title=element_text(size=10, face=\"bold\", hjust=0.5)\n",
    "        ) \n",
    "\n",
    "    ggsave(paste0(\"stat/\", ID, \".stringtie.minCoverage.pdf\"), width=6, height=6)\n",
    "\n",
    "\n",
    "\n",
    "}\n",
    "\n",
    "\n",
    "# gap length\n",
    "for(ID in IDs){\n",
    "    statFile = paste0(\"stringtie_res5/\", ID, \".gap.gffcmp.stats_sensitivity_precision.txt\")\n",
    "    stat.tb <- read.table(statFile, header=T)\n",
    "    Types <- c(\"Base_level\", \"Exon_level\", \"Transcript_level\")\n",
    "    pp <- stat.tb[stat.tb$Type %in% Types,]\n",
    "    pp$Max <- pp$max_gap\n",
    "    for(Type in Types){\n",
    "        tmp <- stat.tb[stat.tb$Type %in% Type & stat.tb$stats %in% \"F1_score\", ]\n",
    "        gap <- tmp$max_gap[tmp$Percent==max(tmp$Percent)]\n",
    "        pp$Max[pp$Type %in% Type & !(pp$max_gap %in% gap)] <- NA\n",
    "    }\n",
    "\n",
    "    pp$max_gap <- factor(pp$max_gap)\n",
    "    ggplot(pp, aes(x=max_gap, y=Percent, color=stats)) +\n",
    "        geom_point()+\n",
    "        scale_color_manual(values=c(F1_score=\"#4678BA\", Sensitivity=\"#87B07A\", Precision=\"#EF8354\")) + \n",
    "        ggtitle(ID) +\n",
    "        geom_vline(aes(xintercept = Max), linetype=\"dashed\", color=\"red\") +\n",
    "        facet_wrap(~Type, nrow = 3) +\n",
    "        # guides(color=\"none\") +\n",
    "        # ylab(\"F1 score\") +\n",
    "        theme_bw() +\n",
    "        theme(\n",
    "          panel.grid=element_blank(),\n",
    "          axis.text.x = element_text(angle = 0, color = \"black\", \n",
    "                                     size = 8, hjust = 0.5),\n",
    "          plot.title=element_text(size=10, face=\"bold\", hjust=0.5)\n",
    "        ) \n",
    "\n",
    "    ggsave(paste0(\"stat/\", ID, \".stringtie.gap.pdf\"), width=6, height=6)\n",
    "\n",
    "\n",
    "\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### the count of assembly transcript "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### minCoverage:1-19, gap=0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(data.table)\n",
    "library(ggbreak)\n",
    "setwd(\"/NAS/wg_looking/gRNA_ONT\")\n",
    "\n",
    "\n",
    "IDs = c(\"sgRNA_HCT116\", \"totalRNA_HCT116\", \n",
    "        \"sgRNA_NCM460\", \"totalRNA_NCM460\", \n",
    "        \"sgRNA_K562\", \"totalRNA_K562\", \n",
    "        \"sgRNA_MDAMB231\", \"sgRNA_MCF10A\")\n",
    "\n",
    "# IDs = c(\"sgRNA_HCT116_NCM460\")\n",
    "\n",
    "# minCoverage\n",
    "count.list <- lapply(IDs, function(ID){\n",
    "    c <- lapply(1:19, function(minCoverage){\n",
    "        cellLine <- strsplit(ID, \"_\")[[1]][2]\n",
    "        # tmap\n",
    "        tmap <- paste0(ID, \".minCoverage_\", minCoverage, \".\", ID, \".minCoverage_\", minCoverage, \".gtf.tmap\")  \n",
    "        tmap <- file.path(\"stringtie_res5\", cellLine, tmap)\n",
    "        tb <- read.table(tmap, header = T, stringsAsFactors = F)\n",
    "        # new transcript with new junction\n",
    "        bed <- paste0(ID, \".minCoverage_\", minCoverage, \".annotated.intron.new.bed\") \n",
    "        bed <- file.path(\"stringtie_res5\", cellLine, bed)\n",
    "        bed.tb <- read.table(bed, stringsAsFactors = F)\n",
    "        splice_junction_id <- tstrsplit(bed.tb$V4, \"[|]\")[[3]]\n",
    "        splice_junction_id <- unique(splice_junction_id)\n",
    "        #\n",
    "        tb$classification <- \"other_exonic_overlap\"\n",
    "        tb$classification[tb$qry_id %in% splice_junction_id] <- \"new_splice_junction\"\n",
    "        tb$classification[tb$class_code %in% \"u\"] <- \"intergenic\"\n",
    "        tb$classification[tb$class_code %in% \"i\"] <- \"intragenic\"\n",
    "        tb$classification[tb$class_code %in% c(\"c\", \"=\")] <- \"match_reference\"\n",
    "        code.tb <- data.frame(table(tb$classification))\n",
    "        colnames(code.tb) <- c(\"classification\", \"Freq\")\n",
    "        code.tb <- data.frame(code.tb, minCoverage=minCoverage, ID=ID)\n",
    "    })\n",
    "    c <- rbindlist(c)  \n",
    "})\n",
    "\n",
    "\n",
    "pp <- rbindlist(count.list)\n",
    "\n",
    "\n",
    "# cols <- c(\"#7DCE82\", \"#87B07A\",\"#028090\",  \"#5AA9E6\", \"#4678BA\", \"#B3A394\", \n",
    "#           \"#FFE66D\", \"#F4B540\", \"#EF8354\", \"#ED6A5E\", \"#C95603\", \"#BA748A\",  \n",
    "#           \"#AA4465\", \"#7D6185\"\n",
    "#           )\n",
    "\n",
    "\n",
    "cols <- c(\"#F4B540\", \"#EF8354\", \"#4678BA\", \"#87B07A\", \"#BA748A\")\n",
    "names(cols) <- levels(pp$classification)\n",
    "\n",
    "ggplot(pp, aes(x=minCoverage, y=Freq, color=classification)) +\n",
    "    geom_line(size=1.5)+\n",
    "    scale_color_manual(values=cols) + \n",
    "    ggtitle(\"\") +\n",
    "    # guides(color=\"none\") +\n",
    "    # ylab(\"F1 score\") +\n",
    "    # scale_y_break(c(20000, 80000), scales=\"0.5\") +\n",
    "    facet_wrap(~ID, nrow = 4, scales=\"free\") +\n",
    "    theme_classic() +\n",
    "    theme(\n",
    "      panel.grid=element_blank(),\n",
    "      axis.text.x = element_text(angle = 0, color = \"black\", \n",
    "                                 size = 8, hjust = 0.5),\n",
    "      # legend.position = \"top\",\n",
    "      plot.title=element_text(size=10, face=\"bold\", hjust=0.5)\n",
    "    ) \n",
    "\n",
    "ggsave(paste0(\"stat/Assembly_classification_line.pdf\"), width=12, height=12)\n",
    "\n",
    "\n",
    "ggplot(pp, aes(x=minCoverage, y=Freq, color=classification)) +\n",
    "    geom_line(size=1.5)+\n",
    "    scale_color_manual(values=cols) + \n",
    "    ggtitle(\"\") +\n",
    "    # guides(color=\"none\") +\n",
    "    # ylab(\"F1 score\") +\n",
    "    # scale_y_break(c(20000, 80000), scales=\"0.5\") +\n",
    "    facet_wrap(~ID, nrow = 4, scales=\"free\") +\n",
    "    scale_y_log10() +\n",
    "    theme_classic() +\n",
    "    theme(\n",
    "      panel.grid=element_blank(),\n",
    "      axis.text.x = element_text(angle = 0, color = \"black\", \n",
    "                                 size = 8, hjust = 0.5),\n",
    "      # legend.position = \"top\",\n",
    "      plot.title=element_text(size=10, face=\"bold\", hjust=0.5)\n",
    "    ) \n",
    "\n",
    "ggsave(paste0(\"stat/Assembly_classification_line_log10.pdf\"), width=12, height=12)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tmp <- tb[,c(3, 4, 5, 6,9)]\n",
    "tmp[tmp$qry_id %in% \"STRG.171396.1\",]\n",
    "tmp[tmp$cov>3 & tmp$cov<5, ][100:110, ]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Max gap:0-50, coverage=5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(data.table)\n",
    "library(ggbreak)\n",
    "setwd(\"/NAS/wg_looking/gRNA_ONT\")\n",
    "\n",
    "\n",
    "IDs = c(\"sgRNA_HCT116\", \"totalRNA_HCT116\", \n",
    "        \"sgRNA_NCM460\", \"totalRNA_NCM460\", \n",
    "        \"sgRNA_K562\", \"totalRNA_K562\", \n",
    "        \"sgRNA_MDAMB231\", \"sgRNA_MCF10A\")\n",
    "\n",
    "# IDs = c(\"sgRNA_HCT116_NCM460\")\n",
    "\n",
    "# max gap\n",
    "count.list <- lapply(IDs, function(ID){\n",
    "    c <- lapply(seq(0,50,5), function(gap){\n",
    "        cellLine <- strsplit(ID, \"_\")[[1]][2]\n",
    "        # tmap\n",
    "        tmap <- paste0(ID, \".minCoverage_5.gap_\", gap, \".\", ID, \".minCoverage_5.gap_\", gap, \".gtf.tmap\")  \n",
    "        tmap <- file.path(\"stringtie_res5\", cellLine, tmap)\n",
    "        tb <- read.table(tmap, header = T, stringsAsFactors = F)\n",
    "        # new transcript with new junction\n",
    "        bed <- paste0(ID, \".minCoverage_5.gap_\", gap, \".annotated.intron.new.bed\") \n",
    "        bed <- file.path(\"stringtie_res5\", cellLine, bed)\n",
    "        bed.tb <- read.table(bed, stringsAsFactors = F)\n",
    "        splice_junction_id <- tstrsplit(bed.tb$V4, \"[|]\")[[3]]\n",
    "        splice_junction_id <- unique(splice_junction_id)\n",
    "        #\n",
    "        tb$classification <- \"other_exonic_overlap\"\n",
    "        tb$classification[tb$qry_id %in% splice_junction_id] <- \"new_splice_junction\"\n",
    "        tb$classification[tb$class_code %in% \"u\"] <- \"intergenic\"\n",
    "        tb$classification[tb$class_code %in% \"i\"] <- \"intragenic\"\n",
    "        tb$classification[tb$class_code %in% c(\"c\", \"=\")] <- \"match_reference\"\n",
    "        code.tb <- data.frame(table(tb$classification))\n",
    "        colnames(code.tb) <- c(\"classification\", \"Freq\")\n",
    "        code.tb <- data.frame(code.tb, max_gap=gap, ID=ID)\n",
    "\n",
    "    })\n",
    "    c <- rbindlist(c)  \n",
    "})\n",
    "\n",
    "\n",
    "pp <- rbindlist(count.list)\n",
    "\n",
    "\n",
    "cols <- c(\"#F4B540\", \"#EF8354\", \"#4678BA\", \"#87B07A\", \"#BA748A\")\n",
    "names(cols) <- levels(pp$classification)\n",
    "\n",
    "\n",
    "ggplot(pp, aes(x=max_gap, y=Freq, color=classification)) +\n",
    "    geom_line(size=1.5)+\n",
    "    scale_color_manual(values=cols) + \n",
    "    ggtitle(\"\") +\n",
    "    # guides(color=\"none\") +\n",
    "    # ylab(\"F1 score\") +\n",
    "    # scale_y_break(c(20000, 80000), scales=\"0.5\") +\n",
    "    facet_wrap(~ID, nrow = 4, scales=\"free\") +\n",
    "    theme_classic() +\n",
    "    theme(\n",
    "      panel.grid=element_blank(),\n",
    "      axis.text.x = element_text(angle = 0, color = \"black\", \n",
    "                                 size = 8, hjust = 0.5),\n",
    "      # legend.position = \"top\",\n",
    "      plot.title=element_text(size=10, face=\"bold\", hjust=0.5)\n",
    "    ) \n",
    "\n",
    "ggsave(paste0(\"stat/Assembly_max_gap_classification_line.pdf\"), width=12, height=12)\n",
    "\n",
    "\n",
    "ggplot(pp, aes(x=max_gap, y=Freq, color=classification)) +\n",
    "    geom_line(size=1.5)+\n",
    "    scale_color_manual(values=cols) + \n",
    "    ggtitle(\"\") +\n",
    "    # guides(color=\"none\") +\n",
    "    # ylab(\"F1 score\") +\n",
    "    # scale_y_break(c(20000, 80000), scales=\"0.5\") +\n",
    "    facet_wrap(~ID, nrow = 4, scales=\"free\") +\n",
    "    scale_y_log10() +\n",
    "    theme_classic() +\n",
    "    theme(\n",
    "      panel.grid=element_blank(),\n",
    "      axis.text.x = element_text(angle = 0, color = \"black\", \n",
    "                                 size = 8, hjust = 0.5),\n",
    "      # legend.position = \"top\",\n",
    "      plot.title=element_text(size=10, face=\"bold\", hjust=0.5)\n",
    "    ) \n",
    "\n",
    "ggsave(paste0(\"stat/Assembly_max_gap_classification_line_log10.pdf\"), width=12, height=12)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### minCoverage, gap=20"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(data.table)\n",
    "library(ggbreak)\n",
    "setwd(\"/NAS/wg_looking/gRNA_ONT\")\n",
    "\n",
    "\n",
    "IDs = c(\"sgRNA_HCT116\", \"totalRNA_HCT116\", \n",
    "        \"sgRNA_NCM460\", \"totalRNA_NCM460\", \n",
    "        \"sgRNA_K562\", \"totalRNA_K562\", \n",
    "        \"sgRNA_MDAMB231\", \"sgRNA_MCF10A\")\n",
    "\n",
    "# IDs = c(\"sgRNA_HCT116_NCM460\")\n",
    "\n",
    "# max gap\n",
    "count.list <- lapply(IDs, function(ID){\n",
    "    c <- lapply(1:19, function(minCoverage){\n",
    "        cellLine <- strsplit(ID, \"_\")[[1]][2]\n",
    "        # tmap\n",
    "        tmap <- paste0(ID, \".minCoverage_\", minCoverage,\".gap_20.\", ID, \".minCoverage_\", minCoverage,\".gap_20.gtf.tmap\")  \n",
    "        tmap <- file.path(\"stringtie_res5\", cellLine, tmap)\n",
    "        tb <- read.table(tmap, header = T, stringsAsFactors = F)\n",
    "        # new transcript with new junction\n",
    "        bed <- paste0(ID, \".minCoverage_\", minCoverage, \".gap_20.annotated.intron.new.bed\") \n",
    "        bed <- file.path(\"stringtie_res5\", cellLine, bed)\n",
    "        bed.tb <- read.table(bed, stringsAsFactors = F)\n",
    "        splice_junction_id <- tstrsplit(bed.tb$V4, \"[|]\")[[3]]\n",
    "        splice_junction_id <- unique(splice_junction_id)\n",
    "        #\n",
    "        tb$classification <- \"other_exonic_overlap\"\n",
    "        tb$classification[tb$qry_id %in% splice_junction_id] <- \"new_splice_junction\"\n",
    "        tb$classification[tb$class_code %in% \"u\"] <- \"intergenic\"\n",
    "        tb$classification[tb$class_code %in% \"i\"] <- \"intragenic\"\n",
    "        tb$classification[tb$class_code %in% c(\"c\", \"=\")] <- \"match_reference\"\n",
    "        code.tb <- data.frame(table(tb$classification))\n",
    "        colnames(code.tb) <- c(\"classification\", \"Freq\")\n",
    "        code.tb <- data.frame(code.tb, minCoverage=minCoverage, ID=ID)\n",
    "\n",
    "    })\n",
    "    c <- rbindlist(c)  \n",
    "})\n",
    "\n",
    "\n",
    "pp <- rbindlist(count.list)\n",
    "\n",
    "\n",
    "cols <- c(\"#F4B540\", \"#EF8354\", \"#4678BA\", \"#87B07A\", \"#BA748A\")\n",
    "names(cols) <- levels(pp$classification)\n",
    "\n",
    "\n",
    "ggplot(pp, aes(x=minCoverage, y=Freq, color=classification)) +\n",
    "    geom_line(size=1.5)+\n",
    "    scale_color_manual(values=cols) + \n",
    "    ggtitle(\"\") +\n",
    "    # guides(color=\"none\") +\n",
    "    # ylab(\"F1 score\") +\n",
    "    # scale_y_break(c(20000, 80000), scales=\"0.5\") +\n",
    "    facet_wrap(~ID, nrow = 4, scales=\"free\") +\n",
    "    theme_classic() +\n",
    "    theme(\n",
    "      panel.grid=element_blank(),\n",
    "      axis.text.x = element_text(angle = 0, color = \"black\", \n",
    "                                 size = 8, hjust = 0.5),\n",
    "      # legend.position = \"top\",\n",
    "      plot.title=element_text(size=10, face=\"bold\", hjust=0.5)\n",
    "    ) \n",
    "\n",
    "ggsave(paste0(\"stat/Assembly_minCoverage_gap20_classification_line.pdf\"), width=12, height=12)\n",
    "\n",
    "\n",
    "ggplot(pp, aes(x=minCoverage, y=Freq, color=classification)) +\n",
    "    geom_line(size=1.5)+\n",
    "    scale_color_manual(values=cols) + \n",
    "    ggtitle(\"\") +\n",
    "    # guides(color=\"none\") +\n",
    "    # ylab(\"F1 score\") +\n",
    "    # scale_y_break(c(20000, 80000), scales=\"0.5\") +\n",
    "    facet_wrap(~ID, nrow = 4, scales=\"free\") +\n",
    "    scale_y_log10() +\n",
    "    theme_classic() +\n",
    "    theme(\n",
    "      panel.grid=element_blank(),\n",
    "      axis.text.x = element_text(angle = 0, color = \"black\", \n",
    "                                 size = 8, hjust = 0.5),\n",
    "      # legend.position = \"top\",\n",
    "      plot.title=element_text(size=10, face=\"bold\", hjust=0.5)\n",
    "    ) \n",
    "\n",
    "ggsave(paste0(\"stat/Assembly_minCoverage_gap20_classification_line_log10.pdf\"), width=12, height=12)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Assembly transcripts"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": [],
    "toc-hr-collapsed": true
   },
   "source": [
    "### Function"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "heading_collapsed": "true"
   },
   "source": [
    "#### Run RepeatMasker with assemly fa and loading RepeatMasker results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "RM_run_for_assemly_fa <- function(ID, gtfFile, outFa){\n",
    "    gffread = \"/NAS/wg_looking/Tools/gffread-0.12.7.Linux_x86_64/gffread\"\n",
    "    hg38_fasta = \"/NAS/wg_looking/human_ref/GRCh38.p13.genome.fa\"\n",
    "    RepeatMasker = \"/NAS/wg_looking/Tools/RepeatMasker/RepeatMasker\"\n",
    "\n",
    "    ## extract transcript fa \n",
    "    cmd = paste0(gffread, \" -w \", outFa, \" -g \", hg38_fasta, \" \", gtfFile)\n",
    "    system(cmd)\n",
    "\n",
    "\n",
    "    ## run RepeatMasker\n",
    "    cmd = paste0(RepeatMasker, \" -pa 150 -a -species human \", outFa, \" stat\")\n",
    "    system(cmd)\n",
    "  \n",
    "}\n",
    "\n",
    "RM_results <- function(RM_file){\n",
    "    library(biomartr)\n",
    "    RM.tb <- read_rm(RM_file)\n",
    "    RM.tb <- data.frame(RM.tb)\n",
    "    RM.tb <- RM.tb[as.numeric(RM.tb$sw_score)>=225 & as.numeric(RM.tb$perc_div)<=18, ]\n",
    "    # RM.tb <- RM.tb[RM.tb$matching_class %in% c(\"SINE/Alu\", \"LINE/L1\"),]\n",
    "    RM.tb <- RM.tb[grepl(\"L1HS\", RM.tb$repeat_id) | \n",
    "                   grepl(\"L1P\", RM.tb$repeat_id) | \n",
    "                   RM.tb$matching_class %in% c(\"SINE/Alu\"), ]\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "heading_collapsed": "true"
   },
   "source": [
    "####  plot TE driver"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### plot TE driver\n",
    "trans_TE_func <- function(tb, totalTrans, ID, filename){\n",
    "  library(ggplot2)\n",
    "  library(dplyr)\n",
    "  library(reshape2)\n",
    "  library(scales)\n",
    "  library(RColorBrewer)\n",
    "  \n",
    "  \n",
    "  ###\n",
    "  paste_fun <- function(x){\n",
    "    # p <- paste0(x, collapse = \",\")\n",
    "    p <- paste0(unique(x), collapse = \",\")\n",
    "    return(p)}\n",
    "  #\n",
    "  \n",
    "  tb <- tb[order(tb$qry_id, tb$matching_class), ]\n",
    "  test <- aggregate(matching_class~qry_id,data=tb,FUN=paste_fun) \n",
    "  pp <- table(test$matching_class)\n",
    "  p <- data.frame(pp)\n",
    "  colnames(p)[1] <- \"Type\"\n",
    "  p <- rbind(p, data.frame(Type=\"without\", Freq=length(totalTrans)-dim(test)[1]))\n",
    "  p$Percentage <- p$Freq/sum(p$Freq)*100\n",
    "  \n",
    "  \n",
    "  ggplot(p, aes(x=\"\", y=Percentage, fill=Type))+\n",
    "    geom_bar(width = 1, stat = \"identity\") +\n",
    "    coord_polar(\"y\", start=0) +\n",
    "    scale_fill_manual(values = c(\"LINE/L1\"=\"#EE9248\", \"without\"=\"#EDC26D\", \n",
    "                                 \"SINE/Alu\"=\"#5CB1DD\", \"LINE/L1,SINE/Alu\"=\"#4A89D5\"))+\n",
    "    ggtitle(ID) + \n",
    "    theme_minimal() +\n",
    "    theme(\n",
    "      axis.title.x = element_blank(),\n",
    "      axis.title.y = element_blank(),\n",
    "      panel.border = element_blank(),\n",
    "      panel.grid=element_blank(),\n",
    "      axis.ticks = element_blank(),\n",
    "      plot.title=element_text(size=14, face=\"bold\", hjust = 0.5)\n",
    "    ) +\n",
    "    theme(axis.text.x=element_blank())+\n",
    "    geom_text(aes(x = 1.6, label = scales::percent(Percentage/100, accuracy = .1)), position = position_stack(vjust = .5))\n",
    "  \n",
    "  \n",
    "  ggsave(filename, width = 6, height = 4)\n",
    "  # ggsave(\"stat/sgRNA_TE_induced_new_trans.pdf\", width = 6, height = 4)\n",
    "  \n",
    " \n",
    "\n",
    "\n",
    "\n",
    " \n",
    "  \n",
    "}\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Main"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### loading data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## loading gtf\n",
    "gtf.list <- list(\n",
    "    sgK562 = \"stringtie_res5/K562/sgRNA_K562.minCoverage_5.gap_20.gtf\",\n",
    "    totalK562_1 = \"stringtie_res5/K562/totalRNA_K562_1.minCoverage_5.gap_20.gtf\",\n",
    "    totalK562 = \"stringtie_res5/K562/totalRNA_K562.minCoverage_5.gap_20.gtf\",\n",
    "    sgHCT116 = \"stringtie_res5/HCT116/sgRNA_HCT116.minCoverage_5.gap_20.gtf\",\n",
    "    totalHCT116 = \"stringtie_res5/HCT116/totalRNA_HCT116.minCoverage_5.gap_20.gtf\",\n",
    "    sgNCM460 = \"stringtie_res5/NCM460/sgRNA_NCM460.minCoverage_5.gap_20.gtf\",\n",
    "    totalNCM460 = \"stringtie_res5/NCM460/totalRNA_NCM460.minCoverage_5.gap_20.gtf\",\n",
    "    sgMDAMB231 = \"stringtie_res5/MDAMB231/sgRNA_MDAMB231.minCoverage_5.gap_20.gtf\",\n",
    "    sgMCF10A = \"stringtie_res5/MCF10A/sgRNA_MCF10A.minCoverage_5.gap_20.gtf\"\n",
    ")\n",
    "\n",
    "\n",
    "\n",
    "gtf.tb.list <- lapply(gtf.list, function(gtfFile){\n",
    "    sg.gtf <- rtracklayer::import(gtfFile)\n",
    "    sg.gtf <- as.data.frame(sg.gtf)\n",
    "    return(sg.gtf)\n",
    "})\n",
    "\n",
    "\n",
    "\n",
    "## loading RepeatMasker results\n",
    "RM.tb.list <- lapply(setNames(names(gtf.list), names(gtf.list)), function(ID){\n",
    "    # run RepeatMasker\n",
    "    gtfFile = gtf.list[[ID]]\n",
    "    outFa <- paste0(\"stat/\", ID, \"_assembly_transcripts.fa\")\n",
    "    RM_file <- paste0(outFa, \".out\")\n",
    "    if(!file.exists(RM_file)){ \n",
    "        print(RM_file)\n",
    "        RM_run_for_assemly_fa(ID, gtfFile, outFa) \n",
    "    }\n",
    "    # RM_run_for_assemly_fa(ID, gtfFile, outFa)\n",
    "    \n",
    "    # loading results\n",
    "    if(file.exists(RM_file)){ \n",
    "        RM.tb <- RM_results(RM_file) \n",
    "    }\n",
    "    \n",
    "})\n",
    "\n",
    "\n",
    "\n",
    "## loading TE annotation from hg38 genome \n",
    "library(bedtoolsr)\n",
    "RM.tb <- RM_results(\"/NAS/wg_looking/human_ref/GRCh38.p13.genome.fa.out\")\n",
    "RM.bed <- RM.tb[, c(\"qry_id\", \"qry_start\", \"qry_end\", \"repeat_id\", \"matching_class\")]\n",
    "\n",
    "write.table(RM.bed, \"stat/hg38.LINE.SINE.bed\", quote=F, sep=\"\\t\", row.names=F, col.names=F)\n",
    "write.table(RM.bed, \"/NAS/wg_looking/human_ref/GRCh38.p13.genome.fa.TE.bed\", quote=F, sep=\"\\t\", row.names=F, col.names=F)\n",
    "save(RM.bed, file=\"stat/hg38.LINE.SINE.RData\")\n",
    "\n",
    "\n",
    "RM.tb.list2 <- lapply(setNames(names(gtf.tb.list), names(gtf.tb.list)), function(ID){\n",
    "    # run RepeatMasker\n",
    "    gtf_info = gtf.tb.list[[ID]]\n",
    "    gtf_exon <- gtf_info[gtf_info$type %in% \"exon\", c(\"seqnames\", \"start\", \"end\", \"transcript_id\", \"type\")]\n",
    "    gtf_trans <- gtf_info[gtf_info$type %in% \"transcript\", ]\n",
    "    gtf_TSS_TES <- lapply(1:dim(gtf_trans)[1], function(i){\n",
    "        x <- gtf_trans[i, ]\n",
    "        cname <- c(\"seqnames\", \"start\", \"end\", \"transcript_id\", \"type\")\n",
    "        if(x[5]==\"*\" | x[5]==\"+\"){\n",
    "            TSS <- data.frame(seqnames=x[\"seqnames\"], start=x[\"start\"], end=x[2], \n",
    "                              transcript_id=x[\"transcript_id\"], type=\"TSS\")\n",
    "            TES <- data.frame(seqnames=x[\"seqnames\"], start=x[\"end\"], end=x[\"end\"],\n",
    "                              transcript_id=x[\"transcript_id\"], type=\"TES\")  \n",
    "        }else{\n",
    "            TSS <- data.frame(seqnames=x[\"seqnames\"], start=x[\"end\"], end=x[\"end\"],\n",
    "                              transcript_id=x[\"transcript_id\"], type=\"TSS\")\n",
    "            TES <- data.frame(seqnames=x[\"seqnames\"], start=x[\"start\"], end=x[\"start\"], \n",
    "                              transcript_id=x[\"transcript_id\"], type=\"TES\")  \n",
    "        }\n",
    "        colnames(TSS) <- cname\n",
    "        colnames(TES) <- cname\n",
    "        \n",
    "        return(list(TSS, TES))\n",
    "\n",
    "    })\n",
    "    \n",
    "    gtf_TSS_TES <- rbindlist(lapply(gtf_TSS_TES, rbindlist))\n",
    "    rownames(gtf_TSS_TES) <- NULL\n",
    "    gtf_bed <- rbind(gtf_exon, gtf_TSS_TES)\n",
    "  \n",
    "    res <- bt.intersect(gtf_bed, RM.bed, wa=T, wb=T)\n",
    "    \n",
    "    return(res)\n",
    "})\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "## loading compare\n",
    "tmap_file.list <- list(\n",
    "    K562=list(sg_vs_ref=\"stringtie_res5/K562/sgRNA_K562.minCoverage_5.gap_20.sgRNA_K562.minCoverage_5.gap_20.gtf.tmap\",\n",
    "              sg_vs_total=\"stringtie_res5/K562/sg_vs_total_K562.sgRNA_K562.minCoverage_5.gap_20.gtf.tmap\",\n",
    "              total_vs_ref=\"stringtie_res5/K562/totalRNA_K562_1.minCoverage_5.gap_20.totalRNA_K562_1.minCoverage_5.gap_20.gtf.tmap\"),\n",
    "    HCT116=list(sg_vs_ref=\"stringtie_res5/HCT116/sgRNA_HCT116.minCoverage_5.gap_20.sgRNA_HCT116.minCoverage_5.gap_20.gtf.tmap\",\n",
    "              sg_vs_total=\"stringtie_res5/HCT116/sg_vs_total_HCT116.sgRNA_HCT116.minCoverage_5.gap_20.gtf.tmap\",\n",
    "              total_vs_ref=\"stringtie_res5/HCT116/totalRNA_HCT116.minCoverage_5.gap_20.totalRNA_HCT116.minCoverage_5.gap_20.gtf.tmap\"),\n",
    "    NCM460=list(sg_vs_ref=\"stringtie_res5/NCM460/sgRNA_NCM460.minCoverage_5.gap_20.sgRNA_NCM460.minCoverage_5.gap_20.gtf.tmap\",\n",
    "              sg_vs_total=\"stringtie_res5/NCM460/sg_vs_total_NCM460.sgRNA_NCM460.minCoverage_5.gap_20.gtf.tmap\",\n",
    "              total_vs_ref=\"stringtie_res5/NCM460/totalRNA_NCM460.minCoverage_5.gap_20.totalRNA_NCM460.minCoverage_5.gap_20.gtf.tmap\"),\n",
    "    MDAMB231=list(sg_vs_ref=\"stringtie_res5/MDAMB231/sgRNA_MDAMB231.minCoverage_5.gap_20.sgRNA_MDAMB231.minCoverage_5.gap_20.gtf.tmap\"),\n",
    "    MCF10A=list(sg_vs_ref=\"stringtie_res5/MCF10A/sgRNA_MCF10A.minCoverage_5.gap_20.sgRNA_MCF10A.minCoverage_5.gap_20.gtf.tmap\")\n",
    ")\n",
    "\n",
    "tmap.tb.list <- lapply(tmap_file.list, function(cellLine_list){\n",
    "    tb.list <- lapply(cellLine_list, function(file){\n",
    "        print(file)\n",
    "        tb <- read.table(file, header = T, stringsAsFactors = F)\n",
    "        tb <- tb[, -which(colnames(tb)%in%c(\"FPKM\", \"TPM\", \"cov\"))]\n",
    "    })\n",
    "})\n",
    "\n",
    "\n",
    "\n",
    "save(gtf.tb.list, RM.tb.list2, tmap.tb.list, \n",
    "     file=\"stat/assembly_preData.RData\")\n",
    "\n",
    "\n",
    "# # loading new splicing bed \n",
    "# bed.list <- list(\n",
    "#     sgK562 = \"stringtie_res5/K562/sgRNA_K562.minCoverage_5.gap_20.annotated.intron.new.bed\",\n",
    "#     totalK562 = \"stringtie_res5/K562/totalRNA_K562.minCoverage_5.gap_20.annotated.intron.new.bed\",\n",
    "#     totalK562_1 = \"stringtie_res5/K562/totalRNA_K562_1.minCoverage_5.gap_20.annotated.intron.new.bed\",\n",
    "#     sgHCT116 = \"stringtie_res5/HCT116/sgRNA_HCT116.minCoverage_5.gap_20.annotated.intron.new.bed\",\n",
    "#     totalHCT116 = \"stringtie_res5/HCT116/totalRNA_HCT116.minCoverage_5.gap_20.annotated.intron.new.bed\",\n",
    "#     sgNCM460 = \"stringtie_res5/NCM460/sgRNA_NCM460.minCoverage_5.gap_20.annotated.intron.new.bed\",\n",
    "#     totalNCM460 = \"stringtie_res5/NCM460/totalRNA_NCM460.minCoverage_5.gap_20.annotated.intron.new.bed\",\n",
    "#     sgMDAMB231 = \"stringtie_res5/MDAMB231/sgRNA_MDAMB231.minCoverage_5.gap_20.annotated.intron.new.bed\",\n",
    "#     sgMCF10A = \"stringtie_res5/MCF10A/sgRNA_MCF10A.minCoverage_5.gap_20.annotated.intron.new.bed\"\n",
    "# )\n",
    "\n",
    "# new.splice.list <- lapply(bed.list, function(bed){\n",
    "#     # bed <- paste0(ID, \".minCoverage_\", minCoverage, \".gap_20.annotated.intron.new.bed\") \n",
    "#     # bed <- file.path(\"stringtie_res5\", cellLine, bed)\n",
    "#     bed.tb <- read.table(bed, stringsAsFactors = F)\n",
    "# })\n",
    "\n",
    "\n",
    "# # loading intron retained bed\n",
    "# bed.list2 <- list(\n",
    "#     sgK562 = \"stringtie_res5/K562/sgRNA_K562.minCoverage_5.gap_20.annotated.intron_retained.bed\",\n",
    "#     totalK562 = \"stringtie_res5/K562/totalRNA_K562.minCoverage_5.gap_20.annotated.intron_retained.bed\",\n",
    "#     totalK562_1 = \"stringtie_res5/K562/totalRNA_K562_1.minCoverage_5.gap_20.annotated.intron_retained.bed\",\n",
    "#     sgHCT116 = \"stringtie_res5/HCT116/sgRNA_HCT116.minCoverage_5.gap_20.annotated.intron_retained.bed\",\n",
    "#     totalHCT116 = \"stringtie_res5/HCT116/totalRNA_HCT116.minCoverage_5.gap_20.annotated.intron_retained.bed\",\n",
    "#     sgNCM460 = \"stringtie_res5/NCM460/sgRNA_NCM460.minCoverage_5.gap_20.annotated.intron_retained.bed\",\n",
    "#     totalNCM460 = \"stringtie_res5/NCM460/totalRNA_NCM460.minCoverage_5.gap_20.annotated.intron_retained.bed\",\n",
    "#     sgMDAMB231 = \"stringtie_res5/MDAMB231/sgRNA_MDAMB231.minCoverage_5.gap_20.annotated.intron_retained.bed\",\n",
    "#     sgMCF10A = \"stringtie_res5/MCF10A/sgRNA_MCF10A.minCoverage_5.gap_20.annotated.intron_retained.bed\"\n",
    "# )\n",
    "\n",
    "# intron.retained.list <- lapply(bed.list2, function(bed){\n",
    "#     # bed <- paste0(ID, \".minCoverage_\", minCoverage, \".gap_20.annotated.intron.new.bed\") \n",
    "#     # bed <- file.path(\"stringtie_res5\", cellLine, bed)\n",
    "#     bed.tb <- read.table(bed, stringsAsFactors = F)\n",
    "# })\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": [],
    "toc-hr-collapsed": true
   },
   "source": [
    "#### K562: sg VS total"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error in library(bedtoolsr): 不存在叫'bedtoolsr'这个名字的程辑包\n",
     "output_type": "error",
     "traceback": [
      "Error in library(bedtoolsr): 不存在叫'bedtoolsr'这个名字的程辑包\nTraceback:\n",
      "1. library(bedtoolsr)"
     ]
    }
   ],
   "source": [
    "\n",
    "## new transcripts from totalRNA\n",
    "totalDiff_func <- function(cellLine, tmap.tb.list, RM.tb.list){\n",
    "    #\n",
    "    total_ref.tb <- tmap.tb.list[[cellLine]][[\"total_vs_ref\"]]\n",
    "    total_ref.tb <- total_ref.tb[!(total_ref.tb$class_code %in% c(\"=\", \"c\")),]\n",
    "    # excluding possible false positives\n",
    "    total_ref.tb <- total_ref.tb[!(total_ref.tb$class_code %in% c(\"e\", \"s\", \"x\", \"p\")),]\n",
    "\n",
    "    #\n",
    "    total_ref.tb$classification <- \"other\"\n",
    "    total_ref.tb$classification[total_ref.tb$class_code %in% \"j\"] <- \"new_splicing\"\n",
    "    total_ref.tb$classification[total_ref.tb$class_code %in% c(\"m\", \"n\")] <- \"intron_retained\"\n",
    "    total_ref.tb$classification[total_ref.tb$class_code %in% c(\"o\", \"y\", \"k\")] <- \"span_intergenic_and_genic\"\n",
    "    total_ref.tb$classification[total_ref.tb$class_code %in% \"u\"] <- \"intergenic\"\n",
    "    total_ref.tb$classification[total_ref.tb$class_code %in% \"i\"] <- \"intronic\"\n",
    "    # total_ref.tb$classification[total_ref.tb$class_code %in% c(\"c\", \"=\")] <- \"match_reference\"\n",
    "\n",
    "    #\n",
    "    if(cellLine==\"K562\"){\n",
    "        RM.totalRNA <- RM.tb.list[[paste0(\"total\", cellLine, \"_1\")]]\n",
    "    }else{\n",
    "        RM.totalRNA <- RM.tb.list[[paste0(\"total\", cellLine)]]}\n",
    "    RM.totalRNA <- RM.totalRNA[, 4:10]\n",
    "    colnames(RM.totalRNA) <- c(\"qry_id\", \"type\", \"TE_chr\", \"TE_start\", \"TE_end\", \"repeat_id\", \"matching_class\")\n",
    "    \n",
    "    total_ref.tb.merge <- merge(total_ref.tb, RM.totalRNA, by=\"qry_id\")\n",
    "    # total_ref.tb.merge <- total_ref.tb.merge[, c(\"qry_id\", \"qry_start\", \"qry_end\", \"qry_left\", \n",
    "    #                                            \"repeat_id\", \"matching_class\", \"classification\")]\n",
    "    return(list(total_ref.tb, total_ref.tb.merge))\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "## new transcripts from sgRNA\n",
    "sgDiff_func <- function(cellLine, tmap.tb.list, RM.tb.list, hg38.gtf){\n",
    "    ref.tb <- tmap.tb.list[[cellLine]][[\"sg_vs_ref\"]]\n",
    "    # if(\"TRUE\" %in% grepl(\"total\",names(tmap.tb.list[[cellLine]]))){\n",
    "    #     total.tb <- tmap.tb.list[[cellLine]][[\"sg_vs_total\"]]\n",
    "    #     merge.tb <- merge(ref.tb, total.tb, by=c(\"qry_gene_id\", \"qry_id\", \"num_exons\", \"len\", \"major_iso_id\"), \n",
    "    #                       suffixes=c(\".ref\", \".total\"))\n",
    "    #     diff.tb <- merge.tb[!(merge.tb$class_code.ref %in%  c(\"=\", \"c\")),]\n",
    "    # }else{\n",
    "    #     diff.tb <- ref.tb[!(ref.tb$class_code %in%  c(\"=\", \"c\")),]\n",
    "    #     diff.tb$class_code.ref <- diff.tb$class_code\n",
    "    #     diff.tb$class_code.total <- NA\n",
    "    # }\n",
    "    diff.tb <- ref.tb[!(ref.tb$class_code %in%  c(\"=\", \"c\")),]\n",
    "    #\n",
    "    # excluding possible false positives\n",
    "    diff.tb <- diff.tb[!(diff.tb$class_code %in% c(\"e\", \"s\", \"x\", \"p\")),]\n",
    "    #\n",
    "    diff.tb$classification <- \"other\"\n",
    "    diff.tb$classification[diff.tb$class_code %in% \"j\"] <- \"new_splicing\"\n",
    "    diff.tb$classification[diff.tb$class_code %in% c(\"m\", \"n\")] <- \"intron_retained\"\n",
    "    diff.tb$classification[diff.tb$class_code %in% c(\"o\", \"y\", \"k\")] <- \"span_intergenic_and_genic\"\n",
    "    diff.tb$classification[diff.tb$class_code %in% \"u\"] <- \"intergenic\"\n",
    "    diff.tb$classification[diff.tb$class_code %in% \"i\"] <- \"intronic\"\n",
    "    \n",
    "    \n",
    "    ## gene_name anno\n",
    "    trans_gtf <- hg38.gtf[hg38.gtf$type %in% \"transcript\", ]\n",
    "    rownames(trans_gtf) <- trans_gtf$transcript_id\n",
    "    diff.tb$gene_name <- trans_gtf[diff.tb$ref_id, \"gene_name\"]\n",
    "    # diff.tb1 <- diff.tb[diff.tb$num_exons>1,]\n",
    "    # table(diff.tb$class_code.ref, diff.tb$class_code.total)\n",
    "    # table(diff.tb1$class_code.ref, diff.tb1$class_code.total)\n",
    "    #\n",
    "    \n",
    "    RM.sgRNA <- RM.tb.list[[paste0(\"sg\", cellLine)]]\n",
    "    RM.sgRNA <- RM.sgRNA[, 4:10]\n",
    "    colnames(RM.sgRNA) <- c(\"qry_id\", \"type\", \"TE_chr\", \"TE_start\", \"TE_end\", \"repeat_id\", \"matching_class\")\n",
    "    RM.sgRNA$TE_name <- paste0(paste(RM.sgRNA$TE_chr, RM.sgRNA$TE_start, RM.sgRNA$TE_end, sep=\"_\"), \":\", RM.sgRNA$repeat_id)\n",
    "    diff.tb.merge1 <- merge(diff.tb, RM.sgRNA, by=\"qry_id\")\n",
    "    diff.tb.merge2 <- merge(diff.tb, RM.sgRNA, by=\"qry_id\", all.x=T)\n",
    "    # diff.tb.merge <- diff.tb.merge[, c(\"qry_id\", \"qry_start\", \"qry_end\", \"qry_left\",\"repeat_id\", \n",
    "    #                                    \"matching_class\", \"classification\", \"class_code.total\")]\n",
    "    return(list(diff.tb, diff.tb.merge1, diff.tb.merge2))\n",
    "}\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(bedtoolsr)\n",
    "setwd(\"/NAS/wg_looking/gRNA_ONT\")\n",
    "load(\"stat/assembly_preData.RData\")\n",
    "load(\"stat/TE_location.RData\")\n",
    "load(\"stat/hg38.gtf.RData\")\n",
    "\n",
    "\n",
    "# run\n",
    "cellLine <- \"K562\"\n",
    "totalDiff <- totalDiff_func(cellLine, tmap.tb.list, RM.tb.list2)\n",
    "sgDiff <- sgDiff_func(cellLine, tmap.tb.list, RM.tb.list2, hg38.gtf)\n",
    "save(sgDiff, totalDiff, file=\"stat/K562_Diff.RData\")\n",
    "\n",
    "\n",
    "load(\"stat/K562_Diff.RData\")\n",
    "trans_TE_func(totalDiff[[2]], totalDiff[[1]]$qry_id, \"totalRNA_K562_1\", \n",
    "              paste0(\"stat/\", cellLine, \"_totalRNA_TE_induced_new_trans2.pdf\")) # plot\n",
    "trans_TE_func(sgDiff[[2]], sgDiff[[1]]$qry_id, \"sgRNA_K562_1\", \n",
    "              paste0(\"stat/\", cellLine, \"_sgRNA_TE_induced_new_trans2.pdf\")) # plot\n",
    "\n",
    "\n",
    "## Supplementary Table 2\n",
    "out <- sgDiff[[1]]\n",
    "out <- out[, c(\"qry_id\", \"classification\", \"class_code\", \"ref_gene_id\", \"gene_name\")]\n",
    "out$ref_gene_id[out$ref_gene_id ==\"-\"] <- NA\n",
    "gtf.tb <- gtf.tb.list[[\"sgK562\"]]\n",
    "gtf.tb <- gtf.tb[gtf.tb$type %in% \"transcript\", c(\"seqnames\", \"start\", \"end\", \"strand\", \"transcript_id\")]\n",
    "colnames(gtf.tb)[c(1, 5)] <- c(\"chrom\", \"qry_id\")\n",
    "out <- merge(gtf.tb, out, by=\"qry_id\")\n",
    "length(unique(out$qry_id)) # 18298\n",
    "\n",
    "diff.tb.merge <- sgDiff[[2]]\n",
    "qry_ids_terminal_TE <- unique(diff.tb.merge$qry_id[(diff.tb.merge$type %in% c(\"TES\", \"TSS\"))])\n",
    "qry_ids_internal_TE <- setdiff(diff.tb.merge$qry_id, qry_ids_terminal_TE)\n",
    "qry_ids_ALU <- unique(diff.tb.merge$qry_id[(diff.tb.merge$matching_class %in% c(\"SINE/Alu\"))])\n",
    "qry_ids_L1 <-  unique(diff.tb.merge$qry_id[(diff.tb.merge$matching_class %in% c(\"LINE/L1\"))])\n",
    "\n",
    "out$TE_location <- \"Without\"\n",
    "out$TE_location[out$qry_id %in% qry_ids_terminal_TE] <- \"Terminal\"\n",
    "out$TE_location[out$qry_id %in% qry_ids_internal_TE] <- \"Internal\"\n",
    "out$TE_type <- \"None\"\n",
    "out$TE_type[out$qry_id %in% qry_ids_ALU] <- \"Alu\"\n",
    "out$TE_type[out$qry_id %in% qry_ids_L1] <- \"L1\"\n",
    "out$TE_type[(out$qry_id %in% qry_ids_ALU) & (out$qry_id %in% qry_ids_L1)] <- \"Alu&L1\"\n",
    "\n",
    "\n",
    "colnames(out) <- c(\"New Transcript ID\", \"Chrom\", \"Start\", \"End\", \"Strand\", \n",
    "                   \"Classification\", \"Class Code\", \"Reference Gene ID\", \"Reference Gene Name\",\n",
    "                  \"TE Location\", \"TE Type\")\n",
    "out <- out[order(out[, \"New Transcript ID\"]), ]\n",
    "\n",
    "write.table(out, \"stat/Supplementary.Table.2.the_information_of_new_transcripts_of_K562_by_capTEs.txt\",\n",
    "           row.names=F, col.names=T, sep=\"\\t\", quote=F)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "## Supplementary Table 3\n",
    "out <- totalDiff[[1]]\n",
    "trans_gtf <- hg38.gtf[hg38.gtf$type %in% \"transcript\", ]\n",
    "rownames(trans_gtf) <- trans_gtf$transcript_id\n",
    "out$gene_name <- trans_gtf[out$ref_id, \"gene_name\"]\n",
    "\n",
    "out <- out[, c(\"qry_id\", \"classification\", \"class_code\", \"ref_gene_id\", \"gene_name\")]\n",
    "out$ref_gene_id[out$ref_gene_id ==\"-\"] <- NA\n",
    "gtf.tb <- gtf.tb.list[[\"totalK562_1\"]]\n",
    "gtf.tb <- gtf.tb[gtf.tb$type %in% \"transcript\", c(\"seqnames\", \"start\", \"end\", \"strand\", \"transcript_id\")]\n",
    "colnames(gtf.tb)[c(1, 5)] <- c(\"chrom\", \"qry_id\")\n",
    "out <- merge(gtf.tb, out, by=\"qry_id\")\n",
    "length(unique(out$qry_id)) # 18298\n",
    "\n",
    "\n",
    "diff.tb.merge <- totalDiff[[2]]\n",
    "qry_ids_terminal_TE <- unique(diff.tb.merge$qry_id[(diff.tb.merge$type %in% c(\"TES\", \"TSS\"))])\n",
    "qry_ids_internal_TE <- setdiff(diff.tb.merge$qry_id, qry_ids_terminal_TE)\n",
    "qry_ids_ALU <- unique(diff.tb.merge$qry_id[(diff.tb.merge$matching_class %in% c(\"SINE/Alu\"))])\n",
    "qry_ids_L1 <-  unique(diff.tb.merge$qry_id[(diff.tb.merge$matching_class %in% c(\"LINE/L1\"))])\n",
    "\n",
    "out$TE_location <- \"Without\"\n",
    "out$TE_location[out$qry_id %in% qry_ids_terminal_TE] <- \"Terminal\"\n",
    "out$TE_location[out$qry_id %in% qry_ids_internal_TE] <- \"Internal\"\n",
    "out$TE_type <- \"None\"\n",
    "out$TE_type[out$qry_id %in% qry_ids_ALU] <- \"Alu\"\n",
    "out$TE_type[out$qry_id %in% qry_ids_L1] <- \"L1\"\n",
    "out$TE_type[(out$qry_id %in% qry_ids_ALU) & (out$qry_id %in% qry_ids_L1)] <- \"Alu&L1\"\n",
    "\n",
    "\n",
    "colnames(out) <- c(\"New Transcript ID\", \"Chrom\", \"Start\", \"End\", \"Strand\", \n",
    "                   \"Classification\", \"Class Code\", \"Reference Gene ID\", \"Reference Gene Name\",\n",
    "                  \"TE Location\", \"TE Type\")\n",
    "out <- out[order(out[, \"New Transcript ID\"]), ]\n",
    "\n",
    "write.table(out, \"stat/Supplementary.Table.3.the_information_of_new_transcripts_of_K562_by_totalRNA.txt\",\n",
    "           row.names=F, col.names=T, sep=\"\\t\", quote=F)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "load(\"stat/hg38.gtf.RData\")\n",
    "hg38.gtf.gene <- hg38.gtf[hg38.gtf$type %in% \"gene\", ]\n",
    "rownames(hg38.gtf.gene) <- hg38.gtf.gene$gene_id\n",
    "\n",
    "tb <- sgDiff[[2]]\n",
    "tb$gene_name <- hg38.gtf.gene[tb$ref_gene_id.ref, \"gene_name\"]\n",
    "tb <- tb[order(tb$classification, tb$type),]\n",
    "\n",
    "write.table(tb, \"stat/k562_new_trans_with_TE.txt\", quote=F, sep=\"\\t\", row.names=F, col.names=T)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# the percentage of newTrans classification\n",
    "\n",
    "t1 <- data.frame(table(sgDiff[[1]]$classification), type=\"sgRNA\")\n",
    "t2 <- data.frame(table(totalDiff[[1]]$classification), type=\"totalRNA\")\n",
    "pp <- rbind(t1,t2)\n",
    "colnames(pp)[1] <- \"classification\"\n",
    "# p <- pp %>%\n",
    "#   group_by(type) %>%\n",
    "#   mutate(Percentage = Freq / sum(Freq)*100)\n",
    "p <- ddply(pp, .(type), mutate, Percentage = Freq / sum(Freq) * 100)\n",
    "\n",
    "cols <- c(\"#F4B540\", \"#EF8354\", \"#BA748A\", \"#7BB274\", \"#5AA9E6\", \"#4678BA\")\n",
    "names(cols) <- levels(pp$classification)\n",
    "\n",
    "\n",
    "ggplot(p, aes(x=type, y=Percentage, fill=classification)) +\n",
    "    geom_bar(stat=\"identity\")+\n",
    "    theme_bw() +\n",
    "    theme(\n",
    "      panel.grid=element_blank(),\n",
    "      axis.text.x = element_text(angle = 0, color = \"black\", \n",
    "                                 size = 8, hjust = 0.5, vjust=0.5),\n",
    "      plot.title=element_text(size=14, face=\"bold\")\n",
    "    ) +\n",
    "    # facet_wrap(~class, nrow = 2, scales=\"free\") +\n",
    "    scale_fill_manual(values = cols) +\n",
    "    ggtitle(cellLine)\n",
    "  \n",
    "ggsave(paste0(\"stat/\", cellLine, \"_newTrans_classification_pct.pdf\"),\n",
    "         width = 4, height = 5)\n",
    "  \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error in unique(Diff[[2]]$qry_id[Diff[[2]]$matching_class %in% \"SINE/Alu\"]): 找不到对象'sgDiff'\n",
     "output_type": "error",
     "traceback": [
      "Error in unique(Diff[[2]]$qry_id[Diff[[2]]$matching_class %in% \"SINE/Alu\"]): 找不到对象'sgDiff'\nTraceback:\n",
      "1. rbind(classification_stat(sgDiff, \"sgRNA\"), classification_stat(totalDiff, \n .     \"totalRNA\"))",
      "2. classification_stat(sgDiff, \"sgRNA\")",
      "3. unique(Diff[[2]]$qry_id[Diff[[2]]$matching_class %in% \"SINE/Alu\"])   # at line 4 of file <text>"
     ]
    }
   ],
   "source": [
    "# the percentage/Freq of TE-newTrans classification\n",
    "load(\"stat/K562_Diff.RData\")\n",
    "\n",
    "classification_stat <- function(Diff, SeqType){\n",
    "    alu_id <- unique(Diff[[2]]$qry_id[Diff[[2]]$matching_class %in% \"SINE/Alu\"])\n",
    "    L1_id <- unique(Diff[[2]]$qry_id[Diff[[2]]$matching_class %in% \"LINE/L1\"])\n",
    "    # only Alu\n",
    "    t1 <- Diff[[1]][Diff[[1]]$qry_id %in% setdiff(alu_id, L1_id),]\n",
    "    t1 <- data.frame(table(t1$classification), repeat_class=\"Alu\")\n",
    "    # only L1\n",
    "    t2 <- Diff[[1]][Diff[[1]]$qry_id %in% setdiff(L1_id, alu_id),]\n",
    "    t2 <- data.frame(table(t2$classification), repeat_class=\"L1\")\n",
    "    # Alu&L1\n",
    "    t3 <- Diff[[1]][Diff[[1]]$qry_id %in% intersect(L1_id, alu_id),]\n",
    "    t3 <- data.frame(table(t3$classification), repeat_class=\"Alu&L1\")\n",
    "    #\n",
    "    t <- rbindlist(list(t1,t2,t3))\n",
    "    t <- data.frame(t, SeqType=SeqType)\n",
    "    return(t)\n",
    "\n",
    "}\n",
    "\n",
    "\n",
    "pp <- rbind(classification_stat(sgDiff, \"sgRNA\"),\n",
    "            classification_stat(totalDiff, \"totalRNA\"))\n",
    "\n",
    "colnames(pp)[1] <- \"classification\"\n",
    "p <- aggregate(Freq~SeqType+classification, data=pp, sum)\n",
    "# p <- ddply(pp, .(SeqType, repeat_class), mutate, Percentage = Freq / sum(Freq) * 100)\n",
    "\n",
    "cols <- c(\"#EE9248\", \"#EDC26D\", \"#8C9FCA\", \"#5CB1DD\", \"#4A89D5\")\n",
    "names(cols) <- c(\"intronic\", \"intron_retained\", \"new_splicing\", \n",
    "                 \"span_intergenic_and_genic\", \"intergenic\")\n",
    "\n",
    "\n",
    "# freq plot\n",
    "p <- p[order(p$SeqType, p$repeat_class, p$Freq, decreasing=F), ]\n",
    "p$SeqType <- factor(p$SeqType, levels=c(\"totalRNA\", \"sgRNA\"))\n",
    "p$classification <- factor(p$classification, levels=rev(c(\"intronic\", \"intron_retained\", \"new_splicing\", \n",
    "                                                      \"span_intergenic_and_genic\", \n",
    "                                                      \"intergenic\")))\n",
    "ggplot(p, aes(x=SeqType, y=Freq, fill=classification)) +\n",
    "    geom_bar(stat=\"identity\",  position=position_dodge())+\n",
    "    theme_classic() +\n",
    "    theme(\n",
    "      panel.grid=element_blank(),\n",
    "      axis.text.x = element_text(angle = 0, color = \"black\", \n",
    "                                 size = 8, hjust = 0.5, vjust=0.5),\n",
    "      plot.title=element_text(size=14, face=\"bold\")\n",
    "    ) +\n",
    "    coord_flip() +\n",
    "    # facet_wrap(~repeat_class, nrow = 1, scales=\"free\") +\n",
    "    scale_fill_manual(values = cols) +\n",
    "    ggtitle(paste0(cellLine, \": TE-transcripts\"))\n",
    "  \n",
    "ggsave(paste0(\"stat/Fig2D.\", cellLine, \"_newTrans_with_TE_classification_Freq.pdf\"),\n",
    "         width = 10, height = 5)\n",
    "  \n",
    "\n",
    "\n",
    "\n",
    "# ggplot(p, aes(x=SeqType, y=Percentage, fill=classification)) +\n",
    "#     geom_bar(stat=\"identity\")+\n",
    "#     theme_bw() +\n",
    "#     theme(\n",
    "#       panel.grid=element_blank(),\n",
    "#       axis.text.x = element_text(angle = 0, color = \"black\", \n",
    "#                                  size = 8, hjust = 0.5, vjust=0.5),\n",
    "#       plot.title=element_text(size=14, face=\"bold\")\n",
    "#     ) +\n",
    "#     # facet_wrap(~repeat_class, nrow = 1, scales=\"free\") +\n",
    "#     scale_fill_manual(values = cols) +\n",
    "#     ggtitle(paste0(cellLine, \": TE-transcripts\"))\n",
    "  \n",
    "# ggsave(paste0(\"stat/\", cellLine, \"_newTrans_with_TE_classification_pct.pdf\"),\n",
    "#          width = 7, height = 5)\n",
    " \n",
    "# pie\n",
    "p2 <- aggregate(Freq~SeqType+classification, data=p, sum)\n",
    "p2 <- ddply(p2, .(SeqType), mutate, Percentage = Freq / sum(Freq) * 100)\n",
    "p2$SeqType <- factor(p2$SeqType, levels=rev(c(\"totalRNA\", \"sgRNA\")))\n",
    "p$classification <- factor(p$classification, levels=rev(c(\"intronic\", \"intron_retained\", \"new_splicing\", \n",
    "                                                      \"span_intergenic_and_genic\", \n",
    "                                                      \"intergenic\")))\n",
    "\n",
    "ggplot(p2, aes(x=\"\", y=Percentage, fill=classification))+\n",
    "    geom_bar(width = 1, stat = \"identity\") +\n",
    "    coord_polar(\"y\", start=0) +\n",
    "    scale_fill_manual(values = cols)+\n",
    "    # ggtitle(ID) + \n",
    "    theme_minimal() +\n",
    "    theme(\n",
    "      axis.title.x = element_blank(),\n",
    "      axis.title.y = element_blank(),\n",
    "      panel.border = element_blank(),\n",
    "      panel.grid=element_blank(),\n",
    "      axis.ticks = element_blank(),\n",
    "      axis.text.x=element_blank(),\n",
    "      plot.title=element_text(size=14, face=\"bold\", hjust = 0.5)\n",
    "    ) +\n",
    "    facet_wrap(~SeqType, nrow=2) + \n",
    "    geom_text(aes(x = 1.6, label = scales::percent(Percentage/100, accuracy = .1)), position = position_stack(vjust = .5))\n",
    "\n",
    "ggsave(paste0(\"stat/\", cellLine, \"_newTrans_with_TE_classification_pct.pdf\"),\n",
    "         width = 4.5, height = 6)\n",
    " \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# The percentage of new transcripts classied by TE location\n",
    "\n",
    "load(\"stat/K562_Diff.RData\") # sgDiff, totalDiff\n",
    "cellLine <- \"K562\"\n",
    "class_codes <- c(\"intronic\", \"intron_retained\", \"new_splicing\", \n",
    "                 \"span_intergenic_and_genic\", \"intergenic\")\n",
    "\n",
    "\n",
    "p <- lapply(class_codes, function(code){\n",
    "    # sgRNA\n",
    "    diff.tb <- sgDiff[[1]]\n",
    "    diff.tb.merge <- sgDiff[[2]]\n",
    "    qry_ids <- diff.tb$qry_id[diff.tb$classification %in% code]\n",
    "    qry_ids <- unique(qry_ids)\n",
    "    #\n",
    "    qry_ids_TE <- diff.tb.merge$qry_id[diff.tb.merge$classification %in% code]\n",
    "    qry_ids_TE <- unique(qry_ids_TE)\n",
    "    #\n",
    "#     left <- diff.tb.merge$qry_left\n",
    "#     left <- gsub(\"\\\\)\", \"\", gsub(\"\\\\(\", \"\", left))\n",
    "#     diff.tb.merge$left <- as.integer(left)\n",
    "\n",
    "#     qry_ids_terminal_TE <- unique(diff.tb.merge$qry_id[diff.tb.merge$classification %in% code & \n",
    "#                                                        (diff.tb.merge$qry_start<=5 | diff.tb.merge$left<=5)])\n",
    "    \n",
    "    qry_ids_terminal_TE <- unique(diff.tb.merge$qry_id[diff.tb.merge$classification %in% code & \n",
    "                                                       (diff.tb.merge$type %in% c(\"TES\", \"TSS\"))])\n",
    "    d1 <- c(without_TE=length(qry_ids)-length(qry_ids_TE),\n",
    "           terminal_TE=length(qry_ids_terminal_TE),\n",
    "           internal_TE=length(qry_ids_TE)-length(qry_ids_terminal_TE))\n",
    "    \n",
    "    # totalRNA\n",
    "    total_ref.tb <-  totalDiff[[1]]\n",
    "    total_ref.tb.merge <- totalDiff[[2]]\n",
    "    qry_ids <- unique(total_ref.tb$qry_id[total_ref.tb$classification %in% code])\n",
    "    qry_ids_TE <- unique(total_ref.tb.merge$qry_id[total_ref.tb.merge$classification %in% code])\n",
    "    # left <- total_ref.tb.merge$qry_left\n",
    "    # left <- gsub(\"\\\\)\", \"\", gsub(\"\\\\(\", \"\", left))\n",
    "    # total_ref.tb.merge$left <- as.integer(left)\n",
    "    # qry_ids_terminal_TE <- unique(total_ref.tb.merge$qry_id[total_ref.tb.merge$classification %in% code & \n",
    "    #                                          (total_ref.tb.merge$qry_start<=5 | total_ref.tb.merge$left<=5)])\n",
    "    qry_ids_terminal_TE <- unique(total_ref.tb.merge$qry_id[total_ref.tb.merge$classification %in% code & \n",
    "                                             (total_ref.tb.merge$type %in% c(\"TES\", \"TSS\"))])\n",
    "    d2 <- c(without_TE=length(qry_ids)-length(qry_ids_TE),\n",
    "            terminal_TE=length(qry_ids_terminal_TE),\n",
    "            internal_TE=length(qry_ids_TE)-length(qry_ids_terminal_TE))\n",
    "    \n",
    "    #\n",
    "    d <- data.frame(rbind(d1,d2), \n",
    "                    type=c(\"sgRNA\", \"totalRNA\"),\n",
    "                    class_code=code)\n",
    "\n",
    "    return(d)\n",
    "})\n",
    "\n",
    "\n",
    "pp <- rbindlist(p)\n",
    "pp <- melt(pp)\n",
    "pp <- ddply(pp, .(type, class_code), mutate, Percentage = value / sum(value)*100)\n",
    "\n",
    "\n",
    "cols <- c(\"#4678BA\", \"#FFE66D\", \"#EF8354\")\n",
    "ggplot(pp, aes(x=type, y=Percentage, fill=variable)) +\n",
    "    geom_bar(stat=\"identity\")+\n",
    "    theme_bw() +\n",
    "    theme(\n",
    "        legend.position=\"bottom\",\n",
    "        panel.grid=element_blank(),\n",
    "        axis.text.x = element_text(angle = 0, color = \"black\", \n",
    "                                 size = 8, hjust = 0.5, vjust=0.5),\n",
    "        plot.title=element_text(size=14, face=\"bold\")\n",
    "    ) +\n",
    "    facet_wrap(~class_code, nrow = 1, scales=\"free\") +\n",
    "    scale_fill_manual(values = cols) +\n",
    "    ggtitle(cellLine)\n",
    "\n",
    "ggsave(paste0(\"stat/Fig4S.A.\", cellLine, \"_newTrans_classification_TE_location_in_trans_pct.pdf\"),\n",
    "         width = 10, height = 5)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in readChar(con, 5L, useBytes = TRUE):\n",
      "\"无法打开压缩文件'stat/K562_Diff.RData'，可能是因为'No such file or directory'\"\n"
     ]
    },
    {
     "ename": "ERROR",
     "evalue": "Error in readChar(con, 5L, useBytes = TRUE): 无法打开链结\n",
     "output_type": "error",
     "traceback": [
      "Error in readChar(con, 5L, useBytes = TRUE): 无法打开链结\nTraceback:\n",
      "1. load(\"stat/K562_Diff.RData\")",
      "2. readChar(con, 5L, useBytes = TRUE)"
     ]
    }
   ],
   "source": [
    "# The percentage of new transcripts classied by TE location\n",
    "\n",
    "load(\"stat/K562_Diff.RData\") # sgDiff, totalDiff\n",
    "cellLine <- \"K562\"\n",
    "class_codes <- c(\"intergenic\", \"intronic\",\"new_splicing\",\n",
    "                 \"intron_retained\", \"new_splicing_&_intron_retained\", \"other\")\n",
    "\n",
    "\n",
    "p <- lapply(class_codes, function(code){\n",
    "    # sgRNA\n",
    "    diff.tb <- sgDiff[[1]]\n",
    "    diff.tb.merge <- sgDiff[[2]]\n",
    "    qry_ids <- diff.tb$qry_id[diff.tb$classification %in% code]\n",
    "    qry_ids <- unique(qry_ids)\n",
    "    #\n",
    "    qry_ids_TE <- diff.tb.merge$qry_id[diff.tb.merge$classification %in% code]\n",
    "    qry_ids_TE <- unique(qry_ids_TE)\n",
    "    #\n",
    "#     left <- diff.tb.merge$qry_left\n",
    "#     left <- gsub(\"\\\\)\", \"\", gsub(\"\\\\(\", \"\", left))\n",
    "#     diff.tb.merge$left <- as.integer(left)\n",
    "\n",
    "#     qry_ids_terminal_TE <- unique(diff.tb.merge$qry_id[diff.tb.merge$classification %in% code & \n",
    "#                                                        (diff.tb.merge$qry_start<=5 | diff.tb.merge$left<=5)])\n",
    "    \n",
    "    qry_ids_terminal_TE <- unique(diff.tb.merge$qry_id[diff.tb.merge$classification %in% code & \n",
    "                                                       (diff.tb.merge$type %in% c(\"TES\", \"TSS\"))])\n",
    "    d1 <- c(without_TE=length(qry_ids)-length(qry_ids_TE),\n",
    "           terminal_TE=length(qry_ids_terminal_TE),\n",
    "           internal_TE=length(qry_ids_TE)-length(qry_ids_terminal_TE))\n",
    "    \n",
    "    # totalRNA\n",
    "    total_ref.tb <-  totalDiff[[1]]\n",
    "    total_ref.tb.merge <- totalDiff[[2]]\n",
    "    qry_ids <- unique(total_ref.tb$qry_id[total_ref.tb$classification %in% code])\n",
    "    qry_ids_TE <- unique(total_ref.tb.merge$qry_id[total_ref.tb.merge$classification %in% code])\n",
    "    # left <- total_ref.tb.merge$qry_left\n",
    "    # left <- gsub(\"\\\\)\", \"\", gsub(\"\\\\(\", \"\", left))\n",
    "    # total_ref.tb.merge$left <- as.integer(left)\n",
    "    # qry_ids_terminal_TE <- unique(total_ref.tb.merge$qry_id[total_ref.tb.merge$classification %in% code & \n",
    "    #                                          (total_ref.tb.merge$qry_start<=5 | total_ref.tb.merge$left<=5)])\n",
    "    qry_ids_terminal_TE <- unique(total_ref.tb.merge$qry_id[total_ref.tb.merge$classification %in% code & \n",
    "                                             (total_ref.tb.merge$type %in% c(\"TES\", \"TSS\"))])\n",
    "    d2 <- c(without_TE=length(qry_ids)-length(qry_ids_TE),\n",
    "            terminal_TE=length(qry_ids_terminal_TE),\n",
    "            internal_TE=length(qry_ids_TE)-length(qry_ids_terminal_TE))\n",
    "    \n",
    "    #\n",
    "    d <- data.frame(rbind(d1,d2), \n",
    "                    type=c(\"sgRNA\", \"totalRNA\"),\n",
    "                    class_code=code)\n",
    "\n",
    "    return(d)\n",
    "})\n",
    "\n",
    "\n",
    "pp <- rbindlist(p)\n",
    "pp <- melt(pp)\n",
    "pp <- ddply(pp, .(type, class_code), mutate, Percentage = value / sum(value)*100)\n",
    "\n",
    "\n",
    "cols <- c(\"#4678BA\", \"#FFE66D\", \"#EF8354\")\n",
    "ggplot(pp, aes(x=type, y=Percentage, fill=variable)) +\n",
    "    geom_bar(stat=\"identity\")+\n",
    "    theme_bw() +\n",
    "    theme(\n",
    "        legend.position=\"bottom\",\n",
    "        panel.grid=element_blank(),\n",
    "        axis.text.x = element_text(angle = 0, color = \"black\", \n",
    "                                 size = 8, hjust = 0.5, vjust=0.5),\n",
    "        plot.title=element_text(size=14, face=\"bold\")\n",
    "    ) +\n",
    "    facet_wrap(~class_code, nrow = 1, scales=\"free\") +\n",
    "    scale_fill_manual(values = cols) +\n",
    "    ggtitle(cellLine)\n",
    "\n",
    "ggsave(paste0(\"stat/\", cellLine, \"_newTrans_classification_TE_pct.pdf\"),\n",
    "         width = 10, height = 5)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# The percentage of new transcripts classied by TE type\n",
    "\n",
    "load(\"stat/K562_Diff.RData\") # sgDiff, totalDiff\n",
    "cellLine <- \"K562\"\n",
    "class_codes <- c(\"intronic\", \"intron_retained\", \"new_splicing\", \"span_intergenic_and_genic\",  \"intergenic\")\n",
    "\n",
    "\n",
    "p <- lapply(class_codes, function(code){\n",
    "    # sgRNA\n",
    "    diff.tb <- sgDiff[[1]]\n",
    "    diff.tb.merge <- sgDiff[[2]]\n",
    "    qry_ids <- diff.tb$qry_id[diff.tb$classification %in% code]\n",
    "    qry_ids <- unique(qry_ids)\n",
    "    #\n",
    "    qry_ids_TE <- unique(diff.tb.merge$qry_id[diff.tb.merge$classification %in% code])\n",
    "    qry_Alu_id <- diff.tb.merge$qry_id[diff.tb.merge$classification %in% code &\n",
    "                                      diff.tb.merge$matching_class %in% \"SINE/Alu\"]\n",
    "    qry_Alu_id <- unique(qry_Alu_id)\n",
    "    qry_L1_id <- diff.tb.merge$qry_id[diff.tb.merge$classification %in% code &\n",
    "                                      diff.tb.merge$matching_class %in% \"LINE/L1\"]\n",
    "    qry_L1_id <- unique(qry_L1_id)\n",
    "    Alu_L1 <- intersect(qry_Alu_id, qry_L1_id)\n",
    "    \n",
    "    d1 <- c(without_TE=length(qry_ids)-length(qry_ids_TE),\n",
    "            Alu=length(setdiff(qry_Alu_id,Alu_L1)),\n",
    "            L1=length(setdiff(qry_L1_id,Alu_L1)), \n",
    "            \"Alu&L1\"=length(Alu_L1))\n",
    "    \n",
    "    # totalRNA\n",
    "    total_ref.tb <-  totalDiff[[1]]\n",
    "    total_ref.tb.merge <- totalDiff[[2]]\n",
    "    qry_ids <- unique(total_ref.tb$qry_id[total_ref.tb$classification %in% code])\n",
    "    qry_ids_TE <- unique(total_ref.tb.merge$qry_id[total_ref.tb.merge$classification %in% code])\n",
    "    qry_Alu_id <- total_ref.tb.merge$qry_id[total_ref.tb.merge$classification %in% code &\n",
    "                                      total_ref.tb.merge$matching_class %in% \"SINE/Alu\"]\n",
    "    qry_Alu_id <- unique(qry_Alu_id)\n",
    "    qry_L1_id <- total_ref.tb.merge$qry_id[total_ref.tb.merge$classification %in% code &\n",
    "                                      total_ref.tb.merge$matching_class %in% \"LINE/L1\"]\n",
    "    qry_L1_id <- unique(qry_L1_id)\n",
    "    Alu_L1 <- intersect(qry_Alu_id, qry_L1_id)\n",
    "    \n",
    "    d2 <- c(without_TE=length(qry_ids)-length(qry_ids_TE),\n",
    "            Alu=length(setdiff(qry_Alu_id,Alu_L1)),\n",
    "            L1=length(setdiff(qry_L1_id,Alu_L1)), \n",
    "            \"Alu&L1\"=length(Alu_L1))\n",
    "    \n",
    "    #\n",
    "    d <- data.frame(rbind(d1,d2), \n",
    "                    type=c(\"sgRNA\", \"totalRNA\"),\n",
    "                    class_code=code)\n",
    "\n",
    "    return(d)\n",
    "})\n",
    "\n",
    "\n",
    "pp <- rbindlist(p)\n",
    "pp <- melt(pp)\n",
    "pp <- ddply(pp, .(type, class_code), mutate, Percentage = value / sum(value)*100)\n",
    "\n",
    "\n",
    "cols <- c(\"#4678BA\", \"#FFE66D\", \"#BA748A\", \"#EF8354\")\n",
    "cols <- c(\"L1\"=\"#EE9248\", \"without_TE\"=\"#EDC26D\", \n",
    " \"Alu\"=\"#5CB1DD\", \"Alu.L1\"=\"#4A89D5\")\n",
    "\n",
    "ggplot(pp, aes(x=class_code, y=Percentage, fill=variable)) +\n",
    "    geom_bar(stat=\"identity\")+\n",
    "    theme_bw() +\n",
    "    theme(\n",
    "        legend.position=\"bottom\",\n",
    "        panel.grid=element_blank(),\n",
    "        axis.text.x = element_text(angle = 30, color = \"black\", \n",
    "                                 size = 8, hjust = 0.5, vjust=0.5),\n",
    "        plot.title=element_text(size=14, face=\"bold\")\n",
    "    ) +\n",
    "    facet_wrap(~type, nrow = 1, scales=\"free\") +\n",
    "    scale_fill_manual(values = cols) +\n",
    "    ggtitle(cellLine)\n",
    "\n",
    "ggsave(paste0(\"stat/Fig2E.\", cellLine, \"_newTrans_classification_Alu_L1_pct.pdf\"),\n",
    "         width = 8, height = 5)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### the distribution of new transcript length"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(patchwork)\n",
    "setwd(\"/NAS/wg_looking/gRNA_ONT\")\n",
    "load(\"stat/assembly_preData.RData\")\n",
    "load(\"stat/K562_Diff.RData\")\n",
    "\n",
    "\n",
    "names <- c(\"sgK562\", \"totalK562_1\")\n",
    "transcript_length <- lapply(setNames(names, names), function(ID){\n",
    "    assemblyFa <- paste0(\"stat/\", ID, \"_assembly_transcripts.fa\")\n",
    "    length_txt <- paste0(assemblyFa, \".length\")\n",
    "    cmd <- paste0(\"/NAS/wg_looking/Tools/seqkit fx2tab \", assemblyFa, \" -l -H -i -n > \", length_txt)\n",
    "    system(cmd)\n",
    "    L <- read.table(length_txt, stringsAsFactors=F)\n",
    "    colnames(L) <- c(\"qry_id\", \"length\")\n",
    "    rownames(L) <- L$qry_id\n",
    "    \n",
    "    return(L)\n",
    "})\n",
    "\n",
    "\n",
    "## sgRNA\n",
    "tb <- sgDiff[[1]][sgDiff[[1]]$qry_id %in% sgDiff[[2]]$qry_id, ]\n",
    "tb.length <- merge(tb, transcript_length[[\"sgK562\"]], by=\"qry_id\")\n",
    "\n",
    "cols <- c(\"#F4B540\", \"#EF8354\", \"#BA748A\", \"#7BB274\", \"#5AA9E6\", \"#4678BA\")\n",
    "names(cols) <- unique(tb.length$classification)\n",
    "p1 <- ggplot(tb.length, aes(x=length,  color=classification)) +\n",
    "    geom_density(size=2) + \n",
    "    theme_bw() +\n",
    "    theme(\n",
    "      panel.grid=element_blank(),\n",
    "      axis.text.x = element_text(angle = 30, color = \"black\", \n",
    "                                 size = 8, hjust = 0.5, vjust=0.5),\n",
    "      plot.title=element_text(size=14, face=\"bold\", hjust=0.5)\n",
    "    ) +\n",
    "    scale_color_manual(values = cols) +\n",
    "    scale_x_log10() +\n",
    "    ggtitle(\"K562: sgRNA\") \n",
    "  \n",
    "# ggsave(paste0(\"stat/sgK562_new_trans_length_density.pdf\"),\n",
    "#       width=9, height=5)\n",
    "\n",
    "\n",
    "\n",
    "## totalRNA\n",
    "tb <- totalDiff[[1]][totalDiff[[1]]$qry_id %in% totalDiff[[2]]$qry_id, ]\n",
    "tb.length <- merge(tb, transcript_length[[\"sgK562\"]], by=\"qry_id\")\n",
    "\n",
    "p2 <- ggplot(tb.length, aes(x=length,  color=classification)) +\n",
    "    geom_density(size=2) + \n",
    "    theme_bw() +\n",
    "    theme(\n",
    "      panel.grid=element_blank(),\n",
    "      axis.text.x = element_text(angle = 30, color = \"black\", \n",
    "                                 size = 8, hjust = 0.5, vjust=0.5),\n",
    "      plot.title=element_text(size=14, face=\"bold\", hjust=0.5)\n",
    "    ) +\n",
    "    scale_color_manual(values = cols) +\n",
    "    scale_x_log10() +\n",
    "    ggtitle(\"K562: totalRNA\") \n",
    "  \n",
    "# ggsave(paste0(\"stat/totalK562_new_trans_length_density.pdf\"),\n",
    "#       width=9, height=5)\n",
    "\n",
    "\n",
    "\n",
    "p <- p1 + p2 + plot_layout(ncol = 1, heights = c(1, 1))\n",
    "\n",
    "\n",
    "ggsave(paste0(\"stat/K562_new_trans_length_density.pdf\"),\n",
    "      width=9, height=8)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### new transcripts: intersect between sgRNA and totalRNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# new transcripts: intersect between sgRNA and totalRNA\n",
    "library(bedtoolsr)\n",
    "load(\"stat/assembly_preData.RData\")\n",
    "load(\"stat/K562_Diff.RData\") # sgDiff, totalDiff\n",
    "\n",
    "cellLine <- \"K562\"\n",
    "\n",
    "tmap.tb <- tmap.tb.list[[\"K562\"]][[\"sg_vs_total\"]]\n",
    "tmap.tb <- tmap.tb[!(tmap.tb$class_code %in% c(\"u\", \"i\", \"p\")), ]\n",
    "\n",
    "\n",
    "class_codes <- c(\"intergenic\",\"new_splicing\",\"intragenic\",\n",
    "                 \"intron_retained\", \"new_splicing_&_intron_retained\", \"other\")\n",
    "\n",
    "num.list <- lapply(class_codes, function(code){\n",
    "    sg.diff.merge <- sgDiff[[2]]\n",
    "    total.diff.merge <- totalDiff[[2]]\n",
    "    sg.diff.merge <- sg.diff.merge[sg.diff.merge$classification %in% code,]\n",
    "    total.diff.merge <- total.diff.merge[total.diff.merge$classification %in% code, ]\n",
    "    # overlap\n",
    "    sg.overlap <- unique(sg.diff.merge$qry_id[sg.diff.merge$qry_id %in% tmap.tb$qry_id])\n",
    "    total.overlap <- unique(total.diff.merge$qry_id[total.diff.merge$qry_id %in% tmap.tb$ref_id])\n",
    "\n",
    "    # sg.overlap <- unique(sg.diff.merge$qry_id[!(sg.diff.merge$class_code.total %in% c(\"u\", \"i\", \"p\"))])\n",
    "    # total.overlap <- unique(sgDiff[[1]]$ref_id.total[!(sgDiff[[1]]$class_code.total %in% c(\"u\", \"i\", \"p\")) &\n",
    "    #                     sgDiff[[1]]$ref_id.total %in% total.diff.merge$qry_id])\n",
    "    # sg specific\n",
    "    sg.specific <- unique(sg.diff.merge$qry_id[!(sg.diff.merge$qry_id %in% sg.overlap)])\n",
    "    # total specific\n",
    "    total.specific <- unique(total.diff.merge$qry_id[!(total.diff.merge$qry_id %in% total.overlap)])\n",
    "    #\n",
    "    data.frame(sg.overlap=length(sg.overlap),\n",
    "              total.overlap=length(total.overlap),\n",
    "              sg.specific=length(sg.specific),\n",
    "              total.specific=length(total.specific),\n",
    "              CellLine=cellLine,\n",
    "              Classfication=code)\n",
    "\n",
    "})\n",
    "\n",
    "\n",
    "pp.num <- rbindlist(num.list)\n",
    "pp <- melt(pp.num)\n",
    "pp <- pp[order(pp$value, pp$Classfication, decreasing=T),]\n",
    "pp$Classfication <- factor(pp$Classfication, levels=as.character(unique(pp$Classfication)))\n",
    "\n",
    "\n",
    "ggplot(pp, aes(x=Classfication, y=value,  fill=variable)) +\n",
    "    geom_bar(stat=\"identity\", position=position_dodge()) + \n",
    "    geom_text(aes(label=value), position=position_dodge(width=0.9), vjust=-0.25) +\n",
    "    theme_bw() +\n",
    "    theme(\n",
    "      panel.grid=element_blank(),\n",
    "      axis.text.x = element_text(angle = 30, color = \"black\", \n",
    "                                 size = 8, hjust = 0.5, vjust=0.5),\n",
    "      plot.title=element_text(size=14, face=\"bold\")\n",
    "    ) +\n",
    "    facet_wrap(~CellLine, nrow = 1, scales=\"free\") +\n",
    "    scale_fill_manual(values = c(\"#4678BA\", \"#FFE66D\", \"#EF8354\", \"#7BB274\")) +\n",
    "    ggtitle(\"\") \n",
    "  \n",
    "ggsave(paste0(\"stat/\", cellLine, \"_new_trans_sgRNA_vs_totalRNA.pdf\"),\n",
    "      width=10, height=5)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "for(code in class_codes){\n",
    "    \n",
    "    # plot venn\n",
    "#     library(VennDiagram)\n",
    "#     require(gridExtra)\n",
    "    \n",
    "    # pdf(paste0(\"stat/\", cellLine, \"_\", code, \"_new_trans_sgRNA_vs_totalRNA.pdf\"))\n",
    "#     g <- draw.pairwise.venn(area1=length(sg.specific), \n",
    "#                        area2=length(sg.overlap), \n",
    "#                        cross.area=length(common.id), \n",
    "#                        category = c(\"sgRNA\",\"totalRNA\"),\n",
    "#                        col = c(\"#4678BA\", \"#FDA56C\"), \n",
    "#                        fill = c(\"#4678BA\", \"#FDA56C\"),\n",
    "#                        main = cellLine,\n",
    "#                        lwd = c(3,3),\n",
    "#                        cex=c(3,3,3),\n",
    "#                        cat.cex =c(3,3),\n",
    "#                        cat.pos=c(-10,10),\n",
    "#                        alpha = rep(0.8, 2))\n",
    "    \n",
    "#     pdf(paste0(\"stat/\", cellLine, \"_\", code, \"_new_trans_sgRNA_vs_totalRNA.pdf\"))\n",
    "#     grid.arrange(gTree(children=g), \n",
    "#                  top=textGrob(cellLine, gp=gpar(fontsize=30)), \n",
    "#                  bottom=textGrob(code, gp=gpar(fontsize=30)))\n",
    "#     dev.off()  \n",
    "\n",
    "}\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###\n",
    "# library(pheatmap)\n",
    "# pp <- table(diff.tb$class_code.ref, diff.tb$class_code.total)\n",
    "# pp <- as.matrix(pp)\n",
    "# pheatmap(pp,\n",
    "#          display_numbers=T,\n",
    "#          number_format=\"%i\",\n",
    "#          main=\"New transcripts type\",\n",
    "#          # col=colorRampPalette(rev(brewer.pal(11,\"RdBu\")[1:5]))(1000),\n",
    "#          col=c(\"#F7F7F7\", colorRampPalette(c(\"#F4B540\", \"#FDA56C\",\"#EF8354\"))(1000)),\n",
    "#          fontsize=12,\n",
    "#          border=\"white\",\n",
    "#          # show_rownames=F,\n",
    "#          # show_colnames=F,\n",
    "#          cluster_rows=F,\n",
    "#          cluster_cols=F,\n",
    "#          width=6.5,\n",
    "#          height=6,\n",
    "#          filename=\"stat/K562_new_trans_type_heatmap.pdf\")\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### polyIII: Alu elements"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(bedtoolsr)\n",
    "load(\"stat/K562_Diff.RData\") # sgDiff, totalDiff\n",
    "\n",
    "cellLine <- \"K562\"\n",
    "\n",
    "\n",
    "ploy3.bed <- read.table(\"/NAS/wg_looking/gRNA_ONT/database/expressed_Alu_elements_ZhangXiaoOu.sorted.bed\")\n",
    "\n",
    "# sgRNA\n",
    "sg.gtf <- gtf.tb[[paste0(\"sg\", cellLine)]]\n",
    "sg.gtf.bed <- sg.gtf[sg.gtf$type==\"transcript\", c(\"seqnames\", \"start\", \"end\", \"transcript_id\")]\n",
    "sg.gtf.bed <- sg.gtf.bed[sg.gtf.bed$transcript_id %in% sgDiff[[2]]$qry_id, ]\n",
    "tb <- sgDiff[[1]][, c(\"qry_id\", \"classification.ref\")]\n",
    "rownames(tb) <- tb$qry_id\n",
    "sg.gtf.bed <- data.frame(sg.gtf.bed, tb[sg.gtf.bed$transcript_id, \"classification.ref\"])\n",
    "sg.res <- bt.intersect(sg.gtf.bed, ploy3.bed, wa=T)\n",
    "table(sg.res$V5)\n",
    "\n",
    "# total\n",
    "total.gtf <- gtf.tb[[paste0(\"total\", cellLine)]]\n",
    "total.gtf.bed <- total.gtf[total.gtf$type==\"transcript\", c(\"seqnames\", \"start\", \"end\", \"transcript_id\")]\n",
    "total.gtf.bed <- total.gtf.bed[total.gtf.bed$transcript_id %in% totalDiff[[2]]$qry_id, ]\n",
    "tb <- totalDiff[[1]][, c(\"qry_id\", \"classification\")]\n",
    "rownames(tb) <- tb$qry_id\n",
    "total.gtf.bed <- data.frame(total.gtf.bed, tb[total.gtf.bed$transcript_id, \"classification\"])\n",
    "total.res <- bt.intersect(total.gtf.bed, ploy3.bed, wa=T)\n",
    "table(total.res$V5)\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### HCT116 vs NCM460"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "load(\"stat/hg38.gtf.RData\")\n",
    "load(\"stat/assembly_preData.RData\") # gtf.tb, RM.tb.list, tmap.tb.list, new.splice.list, intron.retained.list\n",
    "HCT116_Diff <- sgDiff_func(\"HCT116\", tmap.tb.list, RM.tb.list2, hg38.gtf)\n",
    "NCM460_Diff <- sgDiff_func(\"NCM460\", tmap.tb.list, RM.tb.list2, hg38.gtf)\n",
    "save(HCT116_Diff, NCM460_Diff, file=\"stat/HCT116_NCM460_diff.Data\")\n",
    "write.table(HCT116_Diff[[3]], \"stat/HCT116_new_transcript_annotation.txt\", sep=\"\\t\", quote=F, col.names=T, row.names=F)\n",
    "write.table(NCM460_Diff[[3]], \"stat/NCM460_new_transcript_annotation.txt\", sep=\"\\t\", quote=F, col.names=T, row.names=F)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# plot\n",
    "trans_TE_func(HCT116_Diff[[2]], HCT116_Diff[[1]]$qry_id, \"HCT116\", paste0(\"stat/HCT116_sgRNA_TE_induced_new_trans.pdf\")) # plot\n",
    "trans_TE_func(NCM460_Diff[[2]], NCM460_Diff[[1]]$qry_id, \"NCM460\", paste0(\"stat/NCM460_sgRNA_TE_induced_new_trans.pdf\")) # plot\n",
    "\n",
    "\n",
    "\n",
    "t1 <- data.frame(table(HCT116_Diff[[1]]$classification.ref), type=\"HCT116\")\n",
    "t2 <- data.frame(table(NCM460_Diff[[1]]$classification.ref), type=\"NCM460\")\n",
    "pp <- rbind(t1,t2)\n",
    "colnames(pp)[1] <- \"classification\"\n",
    "p <- ddply(pp, .(type), mutate, Percentage = Freq / sum(Freq) * 100)\n",
    "\n",
    "\n",
    "cols <- c(\"#F4B540\", \"#EF8354\", \"#4678BA\", \"#87B07A\")\n",
    "names(cols) <- levels(p$classification)\n",
    "\n",
    "\n",
    "ggplot(p, aes(x=type, y=Percentage, fill=classification)) +\n",
    "    geom_bar(stat=\"identity\")+\n",
    "    theme_bw() +\n",
    "    theme(\n",
    "      panel.grid=element_blank(),\n",
    "      axis.text.x = element_text(angle = 0, color = \"black\", \n",
    "                                 size = 8, hjust = 0.5, vjust=0.5),\n",
    "      plot.title=element_text(size=14, face=\"bold\")\n",
    "    ) +\n",
    "    # facet_wrap(~class, nrow = 2, scales=\"free\") +\n",
    "    scale_fill_manual(values = cols) +\n",
    "    ggtitle(\"HCT116 vs NCM460\")\n",
    "  \n",
    "ggsave(paste0(\"stat/HCT116_vs_NCM460_newTrans_classification_pct.pdf\"),\n",
    "         width = 4, height = 5)\n",
    "  \n",
    "\n",
    "p <- p[order(p$type, p$Freq, decreasing=T), ]\n",
    "p$classification <- factor(p$classification, levels=p$classification[1:4])\n",
    "ggplot(p, aes(x=classification, y=Freq, fill=type)) +\n",
    "    geom_bar(stat=\"identity\", position=position_dodge())+\n",
    "    theme_bw() +\n",
    "    theme(\n",
    "      panel.grid=element_blank(),\n",
    "      axis.text.x = element_text(angle = 30, color = \"black\", \n",
    "                                 size = 8, hjust = 0.5, vjust=0.5),\n",
    "      plot.title=element_text(size=14, face=\"bold\")\n",
    "    ) +\n",
    "    # facet_wrap(~class, nrow = 2, scales=\"free\") +\n",
    "    scale_fill_manual(values = c(\"#4678BA\", \"#FDA56C\")) +\n",
    "    ggtitle(\"HCT116 vs NCM460\")\n",
    "  \n",
    "ggsave(paste0(\"stat/HCT116_vs_NCM460_newTrans_classification_pct_bar.pdf\"),\n",
    "         width = 6, height = 5)\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### new transcripts: intersect "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# new transcripts: intersect between sgRNA and totalRNA\n",
    "library(bedtoolsr)\n",
    "load(\"stat/HCT116_NCM460_diff.Data\") # HCT116_Diff, NCM460_Diff\n",
    "\n",
    "\n",
    "tmap <- \"stringtie_res5/HCT116/sgHCT116_vs_sgNCM460.sgRNA_HCT116.minCoverage_5.gap_20.gtf.tmap\"\n",
    "tmap.tb <- read.table(tmap, header = T, stringsAsFactors = F)\n",
    "tmap.tb <- tmap.tb[, -which(colnames(tmap.tb)%in%c(\"FPKM\", \"TPM\", \"cov\"))]\n",
    "tmap.tb <- tmap.tb[!(tmap.tb$class_code %in% c(\"u\", \"i\", \"p\")), ]\n",
    "\n",
    "\n",
    "class_codes <- c(\"intergenic\",\"new_splicing\",\"intragenic\",\n",
    "                 \"intron_retained\", \"new_splicing_&_intron_retained\", \"other\")\n",
    "\n",
    "num.list <- lapply(class_codes, function(code){\n",
    "    HCT116.diff.merge <- HCT116_Diff[[2]]\n",
    "    NCM460.diff.merge <- NCM460_Diff[[2]]\n",
    "    HCT116.diff.merge <- HCT116.diff.merge[HCT116.diff.merge$classification %in% code,]\n",
    "    NCM460.diff.merge <- NCM460.diff.merge[NCM460.diff.merge$classification %in% code, ]\n",
    "    # overlap\n",
    "    HCT116.overlap <- unique(HCT116.diff.merge$qry_id[HCT116.diff.merge$qry_id %in% tmap.tb$qry_id])\n",
    "    NCM460.overlap <- unique(NCM460.diff.merge$qry_id[NCM460.diff.merge$qry_id %in% tmap.tb$ref_id])\n",
    "    # HCT116 specific\n",
    "    HCT116.specific <- unique(HCT116.diff.merge$qry_id[!(HCT116.diff.merge$qry_id %in% HCT116.overlap)])\n",
    "    # NCM460 specific\n",
    "    NCM460.specific <- unique(NCM460.diff.merge$qry_id[!(NCM460.diff.merge$qry_id %in% NCM460.overlap)])\n",
    "    #\n",
    "    data.frame(HCT116.overlap=length(HCT116.overlap),\n",
    "              NCM460.overlap=length(NCM460.overlap),\n",
    "              HCT116.specific=length(HCT116.specific),\n",
    "              NCM460.specific=length(NCM460.specific),\n",
    "              Classfication=code)\n",
    "\n",
    "})\n",
    "\n",
    "\n",
    "pp.num <- rbindlist(num.list)\n",
    "pp <- melt(pp.num)\n",
    "pp <- pp[order(pp$value, pp$Classfication, decreasing=T),]\n",
    "pp$Classfication <- factor(pp$Classfication, levels=as.character(unique(pp$Classfication)))\n",
    "\n",
    "\n",
    "ggplot(pp, aes(x=Classfication, y=value,  fill=variable)) +\n",
    "    geom_bar(stat=\"identity\", position=position_dodge()) + \n",
    "    geom_text(aes(label=value), position=position_dodge(width=0.9), vjust=-0.25) +\n",
    "    theme_bw() +\n",
    "    theme(\n",
    "      panel.grid=element_blank(),\n",
    "      axis.text.x = element_text(angle = 30, color = \"black\", \n",
    "                                 size = 8, hjust = 0.5, vjust=0.5),\n",
    "      plot.title=element_text(size=14, face=\"bold\", hjust=0.5)\n",
    "    ) +\n",
    "    # facet_wrap(~CellLine, nrow = 1, scales=\"free\") +\n",
    "    scale_fill_manual(values = c(\"#4678BA\", \"#FFE66D\", \"#EF8354\", \"#7BB274\")) +\n",
    "    ggtitle(\"HCT116 vs NCM460\") \n",
    "  \n",
    "ggsave(paste0(\"stat/HCT116_vs_NCM460_new_trans_intersect.pdf\"),\n",
    "      width=10, height=5)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "#### MDA-MB-231 vs MCF-10A"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "load(\"stat/hg38.gtf.RData\")\n",
    "load(\"stat/assembly_preData.RData\") # gtf.tb, RM.tb.list, tmap.tb.list, new.splice.list, intron.retained.list\n",
    "MDAMB231_Diff <- sgDiff_func(\"MDAMB231\", tmap.tb.list , RM.tb.list2, hg38.gtf)\n",
    "MCF10A_Diff <- sgDiff_func(\"MCF10A\", tmap.tb.list , RM.tb.list2, hg38.gtf)\n",
    "\n",
    "save(MDAMB231_Diff, MCF10A_Diff, file=\"stat/MDAMB231_MCF10A_diff.Data\")\n",
    "\n",
    "\n",
    "write.table(MDAMB231_Diff[[3]], \"stat/MDAMB231_new_transcript_annotation.txt\", sep=\"\\t\", quote=F, col.names=T, row.names=F)\n",
    "write.table(MCF10A_Diff[[3]], \"stat/MCF10A_new_transcript_annotation.txt\", sep=\"\\t\", quote=F, col.names=T, row.names=F)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# the percentage/Freq of TE-newTrans classification\n",
    "\n",
    "classification_stat <- function(Diff, SeqType){\n",
    "    alu_id <- unique(Diff[[2]]$qry_id[Diff[[2]]$matching_class %in% \"SINE/Alu\"])\n",
    "    L1_id <- unique(Diff[[2]]$qry_id[Diff[[2]]$matching_class %in% \"LINE/L1\"])\n",
    "    # only Alu\n",
    "    t1 <- Diff[[1]][Diff[[1]]$qry_id %in% setdiff(alu_id, L1_id),]\n",
    "    t1 <- data.frame(table(t1$classification), repeat_class=\"Alu\")\n",
    "    # only L1\n",
    "    t2 <- Diff[[1]][Diff[[1]]$qry_id %in% setdiff(L1_id, alu_id),]\n",
    "    t2 <- data.frame(table(t2$classification), repeat_class=\"L1\")\n",
    "    # Alu&L1\n",
    "    t3 <- Diff[[1]][Diff[[1]]$qry_id %in% intersect(L1_id, alu_id),]\n",
    "    t3 <- data.frame(table(t3$classification), repeat_class=\"Alu&L1\")\n",
    "    #\n",
    "    t <- rbindlist(list(t1,t2,t3))\n",
    "    t <- data.frame(t, SeqType=SeqType)\n",
    "    return(t)\n",
    "\n",
    "}\n",
    "\n",
    "\n",
    "pp <- rbind(classification_stat(MDAMB231_Diff, \"MDAMB231\"),\n",
    "            classification_stat(MCF10A_Diff, \"MCF10A\"))\n",
    "colnames(pp)[1] <- \"classification\"\n",
    "p <- aggregate(Freq~SeqType+classification, data=pp, sum)\n",
    "# p <- ddply(pp, .(SeqType, classification), mutate, Count = sum(Freq) * 100)\n",
    "\n",
    "cols <- c(\"#F4B540\", \"#EF8354\", \"#BA748A\", \"#7BB274\", \"#5AA9E6\")\n",
    "names(cols) <- c(\"intronic\", \"intron_retained\", \"new_splicing\", \n",
    "                 \"span_intergenic_and_genic\", \"intergenic\")\n",
    "\n",
    "\n",
    "\n",
    "# freq plot\n",
    "p <- p[order(p$SeqType, p$Freq, decreasing=F), ]\n",
    "p$SeqType <- factor(p$SeqType, levels=c(\"MCF10A\", \"MDAMB231\"))\n",
    "p$classification <- factor(p$classification, levels=rev(c(\"intronic\", \"intron_retained\", \"new_splicing\", \n",
    "                                                      \"span_intergenic_and_genic\", \n",
    "                                                      \"intergenic\")))\n",
    "ggplot(p, aes(x=SeqType, y=Freq, fill=classification)) +\n",
    "    geom_bar(stat=\"identity\",  position=position_dodge())+\n",
    "    theme_classic() +\n",
    "    theme(\n",
    "      panel.grid=element_blank(),\n",
    "      axis.text.x = element_text(angle = 0, color = \"black\", \n",
    "                                 size = 8, hjust = 0.5, vjust=0.5),\n",
    "      plot.title=element_text(size=14, face=\"bold\")\n",
    "    ) +\n",
    "    coord_flip() +\n",
    "    # facet_wrap(~repeat_class, nrow = 1, scales=\"free\") +\n",
    "    scale_fill_manual(values = cols) +\n",
    "    ggtitle(paste0(cellLine, \": TE-transcripts\"))\n",
    "  \n",
    "ggsave(paste0(\"stat/MDAMB231_vs_MCF10A_newTrans_with_TE_classification_Freq.pdf\"),\n",
    "         width = 10, height = 5)\n",
    "  \n",
    "\n",
    "\n",
    "# pie\n",
    "p2 <- aggregate(Freq~SeqType+classification, data=p, sum)\n",
    "p2 <- ddply(p2, .(SeqType), mutate, Percentage = Freq / sum(Freq) * 100)\n",
    "p2$SeqType <- factor(p2$SeqType, levels=rev(c(\"totalRNA\", \"sgRNA\")))\n",
    "p2$classification <- factor(p2$classification, levels=rev(c(\"intronic\", \"intron_retained\", \"new_splicing\", \n",
    "                                                      \"span_intergenic_and_genic\", \n",
    "                                                      \"intergenic\")))\n",
    "\n",
    "ggplot(p2, aes(x=\"\", y=Percentage, fill=classification))+\n",
    "    geom_bar(width = 1, stat = \"identity\") +\n",
    "    coord_polar(\"y\", start=0) +\n",
    "    scale_fill_manual(values = cols)+\n",
    "    # ggtitle(ID) + \n",
    "    theme_minimal() +\n",
    "    theme(\n",
    "      axis.title.x = element_blank(),\n",
    "      axis.title.y = element_blank(),\n",
    "      panel.border = element_blank(),\n",
    "      panel.grid=element_blank(),\n",
    "      axis.ticks = element_blank(),\n",
    "      axis.text.x=element_blank(),\n",
    "      plot.title=element_text(size=14, face=\"bold\", hjust = 0.5)\n",
    "    ) +\n",
    "    facet_wrap(~SeqType, nrow=2) + \n",
    "    geom_text(aes(x = 1.6, label = scales::percent(Percentage/100, accuracy = .1)), position = position_stack(vjust = .5))\n",
    "\n",
    "ggsave(paste0(\"stat/\", cellLine, \"_newTrans_with_TE_classification_pct.pdf\"),\n",
    "         width = 4.5, height = 6)\n",
    " \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(bedtoolsr)\n",
    "load(\"stat/MDAMB231_MCF10A_diff.Data\") # MDAMB231_Diff, MCF10A_Diff\n",
    "\n",
    "\n",
    "tmap <- \"stringtie_res5/MDAMB231/sgMDAMB231_vs_sgMCF10A.sgRNA_MDAMB231.minCoverage_5.gap_20.gtf.tmap\"\n",
    "tmap.tb <- read.table(tmap, header = T, stringsAsFactors = F)\n",
    "tmap.tb <- tmap.tb[, -which(colnames(tmap.tb)%in%c(\"FPKM\", \"TPM\", \"cov\"))]\n",
    "tmap.tb <- tmap.tb[!(tmap.tb$class_code %in% c(\"u\", \"i\", \"p\")), ]\n",
    "\n",
    "\n",
    "class_codes <- c(\"intergenic\",\"new_splicing\",\"intragenic\",\n",
    "                 \"intron_retained\", \"new_splicing_&_intron_retained\", \"other\")\n",
    "\n",
    "num.list <- lapply(class_codes, function(code){\n",
    "    MDAMB231.diff.merge <- MDAMB231_Diff[[2]]\n",
    "    MCF10A.diff.merge <- MCF10A_Diff[[2]]\n",
    "    MDAMB231.diff.merge <- MDAMB231.diff.merge[MDAMB231.diff.merge$classification %in% code,]\n",
    "    MCF10A.diff.merge <- MCF10A.diff.merge[MCF10A.diff.merge$classification %in% code, ]\n",
    "    # overlap\n",
    "    MDAMB231.overlap <- unique(MDAMB231.diff.merge$qry_id[MDAMB231.diff.merge$qry_id %in% tmap.tb$qry_id])\n",
    "    MCF10A.overlap <- unique(MCF10A.diff.merge$qry_id[MCF10A.diff.merge$qry_id %in% tmap.tb$ref_id])\n",
    "    # MDAMB231 specific\n",
    "    MDAMB231.specific <- unique(MDAMB231.diff.merge$qry_id[!(MDAMB231.diff.merge$qry_id %in% MDAMB231.overlap)])\n",
    "    # MCF10A specific\n",
    "    MCF10A.specific <- unique(MCF10A.diff.merge$qry_id[!(MCF10A.diff.merge$qry_id %in% MCF10A.overlap)])\n",
    "    #\n",
    "    data.frame(MDAMB231.overlap=length(MDAMB231.overlap),\n",
    "              MCF10A.overlap=length(MCF10A.overlap),\n",
    "              MDAMB231.specific=length(MDAMB231.specific),\n",
    "              MCF10A.specific=length(MCF10A.specific),\n",
    "              Classfication=code)\n",
    "\n",
    "})\n",
    "\n",
    "\n",
    "pp.num <- rbindlist(num.list)\n",
    "pp <- melt(pp.num)\n",
    "pp <- pp[order(pp$value, pp$Classfication, decreasing=T),]\n",
    "pp$Classfication <- factor(pp$Classfication, levels=as.character(unique(pp$Classfication)))\n",
    "\n",
    "\n",
    "ggplot(pp, aes(x=Classfication, y=value,  fill=variable)) +\n",
    "    geom_bar(stat=\"identity\", position=position_dodge()) + \n",
    "    geom_text(aes(label=value), position=position_dodge(width=0.9), vjust=-0.25) +\n",
    "    theme_bw() +\n",
    "    theme(\n",
    "      panel.grid=element_blank(),\n",
    "      axis.text.x = element_text(angle = 30, color = \"black\", \n",
    "                                 size = 8, hjust = 0.5, vjust=0.5),\n",
    "      plot.title=element_text(size=14, face=\"bold\", hjust=0.5)\n",
    "    ) +\n",
    "    # facet_wrap(~CellLine, nrow = 1, scales=\"free\") +\n",
    "    scale_fill_manual(values = c(\"#4678BA\", \"#FFE66D\", \"#EF8354\", \"#7BB274\")) +\n",
    "    ggtitle(\"MDAMB231 vs MCF10A\") \n",
    "  \n",
    "ggsave(paste0(\"stat/MDAMB231_vs_MCF10A_new_trans_intersect.pdf\"),\n",
    "      width=10, height=5)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 细胞系之间的共性和差异性"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(ggplot2)\n",
    "library(ggvenn)\n",
    "setwd(\"/NAS/wg_looking/gRNA_ONT\")\n",
    "load(\"stat/MDAMB231_MCF10A_diff.Data\") # MDAMB231_Diff, MCF10A_Diff\n",
    "load(\"stat/HCT116_NCM460_diff.Data\") # HCT116_Diff, NCM460_Diff\n",
    "\n",
    "\n",
    "class_codes <- c(\"new_splicing\",\"intronic\", \"intron_retained\", \"new_splicing_&_intron_retained\", \"other\")\n",
    "Diff_list <- list(HCT116=HCT116_Diff, NCM460=NCM460_Diff,\n",
    "                  MDAMB231=MDAMB231_Diff, MCF10A=MCF10A_Diff)\n",
    "for(code in class_codes){\n",
    "    overlap.genes <- lapply(Diff_list, function(Diff){\n",
    "        tb <- Diff[[1]]\n",
    "        tb.TE <- Diff[[2]]\n",
    "        sub.tb <- tb[tb$classification %in% code &\n",
    "                    tb$qry_id %in% tb.TE$qry_id, ]\n",
    "        if(\"ref_gene_id\" %in% colnames(sub.tb)){\n",
    "            gene_ids <- unique(sub.tb$ref_gene_id)\n",
    "        }else{\n",
    "            gene_ids <- unique(sub.tb$ref_gene_id.ref)\n",
    "        }\n",
    "        \n",
    "        return(gene_ids)\n",
    "    })\n",
    "    \n",
    "    \n",
    "     # plot\n",
    "    ggvenn(overlap.genes, \n",
    "           fill_color = c(\"#4678BA\", \"#EF8354\", \"#87B07A\", \"#F4B540\"),\n",
    "           stroke_color = NA,\n",
    "           fill_alpha = 0.8,\n",
    "           stroke_size = 0.5, \n",
    "           set_name_size = 4,\n",
    "           show_outside = \"always\"\n",
    "    ) + \n",
    "    ggtitle(code) +\n",
    "    theme(\n",
    "      plot.title=element_text(size=14, face=\"bold\", hjust=0.5)\n",
    "    ) \n",
    "    \n",
    "    \n",
    "#     ggsave(paste0('stat/', code, '_', TE, '_coding_gene_venn.pdf'),\n",
    "#            width=10, height=5)\n",
    "       ggsave(paste0('stat/', code, '_TE_newTrans_overlapped_genes_venn.pdf'),\n",
    "           width=10, height=5)\n",
    "\n",
    "    \n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "load(\"stat/MDAMB231_MCF10A_diff.Data\") # MDAMB231_Diff, MCF10A_Diff\n",
    "load(\"stat/HCT116_NCM460_diff.Data\") # HCT116_Diff, NCM460_Diff\n",
    "\n",
    "\n",
    "## Supplementary Table 7\n",
    "Diff_list <- list()\n",
    "Diff_list[[\"MDAMB231\"]] <- MDAMB231_Diff\n",
    "Diff_list[[\"MCF10A\"]] <- MCF10A_Diff\n",
    "Diff_list[[\"HCT116\"]] <- HCT116_Diff\n",
    "Diff_list[[\"NCM460\"]] <- NCM460_Diff\n",
    "\n",
    "f <- lapply(names(Diff_list), function(CellLine){\n",
    "    out <- Diff_list[[CellLine]][[1]]\n",
    "    out <- out[, c(\"qry_id\", \"classification\", \"class_code\", \"ref_gene_id\", \"gene_name\")]\n",
    "    out$ref_gene_id[out$ref_gene_id ==\"-\"] <- NA\n",
    "    gtf.tb <- gtf.tb.list[[\"sgK562\"]]\n",
    "    gtf.tb <- gtf.tb[gtf.tb$type %in% \"transcript\", c(\"seqnames\", \"start\", \"end\", \"strand\", \"transcript_id\")]\n",
    "    colnames(gtf.tb)[c(1, 5)] <- c(\"chrom\", \"qry_id\")\n",
    "    out <- merge(gtf.tb, out, by=\"qry_id\")\n",
    "    length(unique(out$qry_id)) # 18298\n",
    "\n",
    "    diff.tb.merge <- Diff_list[[CellLine]][[2]]\n",
    "    qry_ids_terminal_TE <- unique(diff.tb.merge$qry_id[(diff.tb.merge$type %in% c(\"TES\", \"TSS\"))])\n",
    "    qry_ids_internal_TE <- setdiff(diff.tb.merge$qry_id, qry_ids_terminal_TE)\n",
    "    qry_ids_ALU <- unique(diff.tb.merge$qry_id[(diff.tb.merge$matching_class %in% c(\"SINE/Alu\"))])\n",
    "    qry_ids_L1 <-  unique(diff.tb.merge$qry_id[(diff.tb.merge$matching_class %in% c(\"LINE/L1\"))])\n",
    "\n",
    "    out$TE_location <- \"Without\"\n",
    "    out$TE_location[out$qry_id %in% qry_ids_terminal_TE] <- \"Terminal\"\n",
    "    out$TE_location[out$qry_id %in% qry_ids_internal_TE] <- \"Internal\"\n",
    "    out$TE_type <- \"None\"\n",
    "    out$TE_type[out$qry_id %in% qry_ids_ALU] <- \"Alu\"\n",
    "    out$TE_type[out$qry_id %in% qry_ids_L1] <- \"L1\"\n",
    "    out$TE_type[(out$qry_id %in% qry_ids_ALU) & (out$qry_id %in% qry_ids_L1)] <- \"Alu&L1\"\n",
    "\n",
    "\n",
    "    colnames(out) <- c(\"New Transcript ID\", \"Chrom\", \"Start\", \"End\", \"Strand\", \n",
    "                       \"Classification\", \"Class Code\", \"Reference Gene ID\", \"Reference Gene Name\",\n",
    "                      \"TE Location\", \"TE Type\")\n",
    "    out <- out[order(out[, \"New Transcript ID\"]), ]\n",
    "    out <- data.frame(CellLine=CellLine, out)\n",
    "\n",
    "})\n",
    "\n",
    "out <- rbindlist(f)\n",
    "write.table(out, \"stat/Supplementary.Table.7.the_information_of_new_transcripts_of_CRC_BC_by_capTEs.txt\",\n",
    "           row.names=F, col.names=T, sep=\"\\t\", quote=F)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "# non-ref TE insertion "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## K562"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error in parse(text = x, srcfile = src): <text>:3:45: 意外的')'\n2: TE_insertion <- lapply(IDs, function(ID){\n3:     a <- lapply(c(\"ALU\", \"L1\"), function(TE))\n                                               ^\n",
     "output_type": "error",
     "traceback": [
      "Error in parse(text = x, srcfile = src): <text>:3:45: 意外的')'\n2: TE_insertion <- lapply(IDs, function(ID){\n3:     a <- lapply(c(\"ALU\", \"L1\"), function(TE))\n                                               ^\nTraceback:\n"
     ]
    }
   ],
   "source": [
    "IDs <- c(\"sgRNA_K562_1\", \"totalRNA_K562_1\")\n",
    "TE_insertion_info <- lapply(IDs, function(ID){\n",
    "    a <- lapply(c(\"ALU\", \"LINE\"), function(TE){\n",
    "        file <- file.path(ID, \"PALMER_res\", TE, paste0(ID, \".PALMER.\", TE, \"_calls.txt\"))\n",
    "        tb <- read.table(file, header=T, stringsAsFactors = F)\n",
    "        bed1 <- tb[, c(\"chr\", \"start1\", \"end1\", \"cluster_id\", \"Potential_supporting_reads\")]\n",
    "        colnames(bed1)[1:3] <- c(\"chrom\", \"start\", \"end\")\n",
    "        bed2 <- tb[, c(\"chr\", \"start2\", \"end2\", \"cluster_id\", \"Potential_supporting_reads\")]\n",
    "        colnames(bed2)[1:3] <- c(\"chrom\", \"start\", \"end\")\n",
    "        bed <- data.frame(rbind(bed1, bed2), ID=ID, TE=TE)\n",
    "        bed <- bed[order(bed$Potential_supporting_reads, decreasing=T),]\n",
    "        bed <- bed[order(bed$chr, bed$start),]\n",
    "        bed$insertion_site <- paste(bed$chrom, bed$start, bed$end, sep=\"_\")\n",
    "        bed <- bed[!duplicated(bed$insertion_site),]\n",
    "\n",
    "        \n",
    "        b <- lapply(3:10, function(SR){\n",
    "            s <- sum(bed$Potential_supporting_reads>=SR)\n",
    "            s <- data.frame(supporting_reads=SR, count=s)\n",
    "            return(s)\n",
    "        })\n",
    "        \n",
    "        b <- rbindlist(b)\n",
    "        b <- data.frame(b, TE=TE, ID=ID)\n",
    "        \n",
    "        return(b) \n",
    "    })\n",
    "    \n",
    "    a <- rbindlist(a)\n",
    "})\n",
    "\n",
    "\n",
    "pp <- rbindlist(TE_insertion_info)\n",
    "\n",
    "\n",
    "ggplot(pp, aes(x=factor(supporting_reads), y=count, fill=ID)) +\n",
    "    geom_bar(stat=\"identity\", position=position_dodge())+\n",
    "    scale_fill_manual(values=c(\"#4A89D5\", \"#EE9248\")) + \n",
    "    # ggtitle(ID) +\n",
    "    facet_wrap(~TE, nrow = 1, scales=\"free\") +\n",
    "    # guides(color=\"none\") +\n",
    "    ylab(\"the count of non-ref TE insertion\") +\n",
    "    theme_bw() +\n",
    "    theme(\n",
    "        panel.grid=element_blank(),\n",
    "        axis.text.x = element_text(angle = 0, color = \"black\", \n",
    "                                 size = 8, hjust = 0.5),\n",
    "        plot.title=element_text(size=10, face=\"bold\", hjust=0.5),\n",
    "        legend.position = \"bottom\"\n",
    "    ) +\n",
    "    xlab(\"supporting reads (>=)\")\n",
    "\n",
    "ggsave(paste0(\"stat/Fig3S.E.sg_vs_total_TE_insertion_number.bar.pdf\"), width=6, height=3)\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### insertion anno"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "TE_insertion_func <-  function(TE, IDs, SupportReads){\n",
    "    a <- lapply(IDs, function(ID){\n",
    "        # file <- file.path(\"merge_PALMER\", ID, TE, paste0(ID, \".PALMER.\", TE, \"_calls.txt\"))\n",
    "        file <- file.path(ID, \"PALMER_res\", TE, paste0(ID, \".PALMER.\", TE, \"_calls.txt\"))\n",
    "        # print(file)\n",
    "        tb <- read.table(file, header=T, stringsAsFactors = F)\n",
    "        sub.tb <- tb[tb$Potential_supporting_reads >= SupportReads, ] \n",
    "        bed1 <- sub.tb[, c(\"chr\", \"start1\", \"end1\", \"cluster_id\", \"Potential_supporting_reads\")]\n",
    "        colnames(bed1)[1:3] <- c(\"chrom\", \"start\", \"end\")\n",
    "        bed2 <- sub.tb[, c(\"chr\", \"start2\", \"end2\", \"cluster_id\", \"Potential_supporting_reads\")]\n",
    "        colnames(bed2)[1:3] <- c(\"chrom\", \"start\", \"end\")\n",
    "        bed <- data.frame(rbind(bed1, bed2), ID=ID, TE=TE)\n",
    "        bed <- bed[order(bed$Potential_supporting_reads, decreasing=T),]\n",
    "        bed <- bed[order(bed$chr, bed$start),]\n",
    "        bed$insertion_site <- paste(bed$chrom, bed$start, bed$end, sep=\"_\")\n",
    "        bed <- bed[!duplicated(bed$insertion_site),]\n",
    "                \n",
    "        return(bed) \n",
    "    })\n",
    "    \n",
    "    \n",
    "    #\n",
    "    b <- rbindlist(a)\n",
    "    b <- b[order(b$chrom, b$start),]\n",
    "    # b$insertion_site <- paste(b$chrom, b$start, b$end, sep=\"_\")\n",
    "        \n",
    "    # anno\n",
    "    b.anno <- bt.intersect(a=b, b=gene_gtf, wao=T)\n",
    "    colnames(b.anno) <- c(colnames(b), colnames(gene_gtf))\n",
    "    b.anno <- b.anno[, c(colnames(b), \"gene_id\",\"gene_name\")]\n",
    "    \n",
    "    \n",
    "    # intron\n",
    "    intron_bed <- read.table(\"/NAS/wg_looking/human_ref/gencode.v37.chr_patch_hapl_scaff.annotation.intron.bed\", \n",
    "                             stringsAsFactors = F)\n",
    "    intron_res <- bt.intersect(a=b.anno, intron_bed, wa=T)\n",
    "\n",
    "    # exon\n",
    "    exon_bed <- hg38.gtf[hg38.gtf$type==\"exon\", c(\"seqnames\",\"start\",\"end\")]\n",
    "    exon_res <- bt.intersect(a=b.anno, exon_bed, wa=T)\n",
    "\n",
    "    # intergenic\n",
    "    intergenic_bed <- read.table(\"/NAS/wg_looking/human_ref/gencode.v37.chr_patch_hapl_scaff.annotation.intergenic.bed\", \n",
    "                             stringsAsFactors = F)\n",
    "    intergenic_res <- bt.intersect(a=b.anno, intergenic_bed, wa=T)\n",
    "    \n",
    "    # loading gff file\n",
    "    gffFile = \"/NAS/wg_looking/human_ref/gencode.v37.chr_patch_hapl_scaff.annotation.gff3\"\n",
    "    gff <- rtracklayer::import(gffFile)\n",
    "    gff <- as.data.frame(gff)\n",
    "    \n",
    "    # 5UTR\n",
    "    UTR5_bed <- gff[gff$type %in% \"five_prime_UTR\", c(\"seqnames\", \"start\", \"end\")]\n",
    "    UTR5_res <- bt.intersect(a=b.anno, UTR5_bed, wa=T)\n",
    "    \n",
    "    # 3UTR\n",
    "    UTR3_bed <- gff[gff$type %in% \"three_prime_UTR\", c(\"seqnames\", \"start\", \"end\")]\n",
    "    UTR3_res <- bt.intersect(a=b.anno, UTR3_bed, wa=T)\n",
    "    \n",
    "    # CDS\n",
    "    CDS_bed <- gff[gff$type %in% \"CDS\", c(\"seqnames\", \"start\", \"end\")]\n",
    "    CDS_res <- bt.intersect(a=b.anno, CDS_bed, wa=T)\n",
    "    \n",
    "    # TSS+-500bp\n",
    "    TSS_bed <- gff[gff$type %in% \"transcript\", c(\"seqnames\", \"start\", \"end\", \"strand\")]\n",
    "    TSS_bed$end[TSS_bed$strand %in% \"+\"] <- TSS_bed$start[TSS_bed$strand %in% \"+\"] + 500\n",
    "    TSS_bed$start[TSS_bed$strand %in% \"+\"] <- TSS_bed$start[TSS_bed$strand %in% \"+\"] - 500\n",
    "    TSS_bed$start[TSS_bed$strand %in% \"-\"] <- TSS_bed$end[TSS_bed$strand %in% \"-\"] - 500\n",
    "    TSS_bed$end[TSS_bed$strand %in% \"-\"] <- TSS_bed$end[TSS_bed$strand %in% \"-\"] + 500\n",
    "    TSS_bed$start[TSS_bed$start<0] <- 0\n",
    "    TSS_res <- bt.intersect(a=b.anno, TSS_bed, wa=T)\n",
    "    \n",
    "    \n",
    "    \n",
    "    # TES+-500bp\n",
    "    TES_bed <- gff[gff$type %in% \"transcript\", c(\"seqnames\", \"start\", \"end\")]\n",
    "    TES_bed$start[TES_bed$strand %in% \"+\"] <- TES_bed$end[TES_bed$strand %in% \"+\"] - 500\n",
    "    TES_bed$end[TES_bed$strand %in% \"+\"] <- TES_bed$end[TES_bed$strand %in% \"+\"] + 500\n",
    "    TES_bed$end[TES_bed$strand %in% \"-\"] <- TES_bed$start[TES_bed$strand %in% \"-\"] + 500\n",
    "    TES_bed$start[TES_bed$strand %in% \"-\"] <- TES_bed$start[TES_bed$strand %in% \"-\"] - 500\n",
    "    TES_bed$start[TES_bed$start<0] <- 0\n",
    "    TES_res <- bt.intersect(a=b.anno, TES_bed, wa=T)\n",
    "    \n",
    "\n",
    "    #\n",
    "    b.anno <- data.frame(b.anno, exon=0, intron=0, intergenic=0)\n",
    "    # b.anno <- data.frame(b.anno, exon=0, intron=0, intergenic=0, UTR5=0, UTR3=0, CDS=0, TSS500=0, TES500=0)\n",
    "    b.anno$exon[b.anno$insertion_site %in% exon_res$V8] <- 1\n",
    "    b.anno$intron[b.anno$insertion_site %in% intron_res$V8] <- 1\n",
    "    b.anno$intergenic[b.anno$insertion_site %in% intergenic_res$V8 ] <- 1\n",
    "    \n",
    "    return(b.anno)\n",
    "    \n",
    "    # b.anno$UTR5[b.anno$insertion_site %in% UTR5_res$V7] <- 1\n",
    "    # b.anno$UTR3[b.anno$insertion_site %in% UTR3_res$V7] <- 1\n",
    "    # l <- b.anno$insertion_site %in% CDS_res$V7\n",
    "    # b.anno$CDS[l] <- 1\n",
    "    # l <- b.anno$insertion_site %in% TSS_res$V7\n",
    "    # b.anno$TSS500[l] <- 1\n",
    "    # l <- b.anno$insertion_site %in% TES_res$V7\n",
    "    # b.anno$TES500[l] <- 1\n",
    "\n",
    "\n",
    "}\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##  insertion anno\n",
    "library(bedtoolsr)\n",
    "library(VennDiagram)\n",
    "load(\"stat/hg38.gene.transcript.gtf.RData\")\n",
    "load(\"stat/hg38.gtf.RData\")\n",
    "\n",
    "\n",
    "IDs <- c(\"sgRNA_K562_1\", \"totalRNA_K562_1\")\n",
    "\n",
    "\n",
    "TE_insertion.list <- list()\n",
    "TE_insertion.list[[\"ALU\"]] <- TE_insertion_func(\"ALU\", IDs, SupportReads=6, plot=F)\n",
    "TE_insertion.list[[\"LINE\"]] <- TE_insertion_func(\"LINE\", IDs, SupportReads=5, plot=F)\n",
    "\n",
    "save(TE_insertion.list, file=\"stat/K562_TE_insertion.anno.RData\")\n",
    "\n",
    "\n",
    "# Supplementary Table 4\n",
    "load(\"stat/K562_TE_insertion.anno.RData\")\n",
    "support <- read.table(\"stat/K562_inerstion_info_spaning_reads2.txt\", \n",
    "                      header=T, sep=\"\\t\", fileEncoding=\"UTF16\", stringsAsFactors=F)\n",
    "support <- support[order(support$insertion_site), ]\n",
    "support <- support[!duplicated(support$insertion_site),]\n",
    "rownames(support) <- support[,1]\n",
    "inerstion_info <- rbindlist(TE_insertion.list)\n",
    "inerstion_info$Total_spanning_reads <- support[as.character(inerstion_info$insertion_site), \"Total_spanning_reads\"]\n",
    "write.table(inerstion_info, \"stat/K562_inerstion_info.txt\", quote=FALSE, sep=\"\\t\", col.names=TRUE, row.names=FALSE)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(ggvenn)\n",
    "\n",
    "\n",
    "IDs <- c(\"sgRNA_K562_1\", \"totalRNA_K562_1\")\n",
    "\n",
    "insertion_genes.list <- lapply(setNames(IDs, IDs), function(ID){\n",
    "    i <- lapply(names(TE_insertion.list), function(TE){\n",
    "        b.anno <- TE_insertion.list[[TE]]\n",
    "        gene_id <- unique(b.anno[b.anno$ID %in% ID, \"gene_id\"])\n",
    "        if(\".\" %in% gene_id){ gene_id <- gene_id[-which(gene_id %in% \".\")] }\n",
    "        return(gene_id)\n",
    "    })\n",
    "    \n",
    "    gene_id <- unique(unlist(i))\n",
    "\n",
    "})\n",
    "\n",
    "\n",
    "plot <- venn.diagram(\n",
    "    x = insertion_genes.list,\n",
    "    # category.names = c(paste0(\"sgRNA: \", TE.name, \"-target related coding genes\") , \n",
    "    #                    paste0(\"Total RNA: include \", TE.name, \" coding genes\")),\n",
    "    filename = NULL,\n",
    "    # filename = paste0('stat/', ID, '_', TE.name, '_coding_gene_venn_diagramm.png'),\n",
    "    # imagetype = \"svg\",\n",
    "    # output=TRUE,\n",
    "    height = 1000 , \n",
    "    width = 1000 , \n",
    "    sub.cex = 0.5,\n",
    "    resolution = 300,\n",
    "    cat.pos = c(0, 1.05),\n",
    "    fill = c(\"#4A89D5\", \"#EE9248\"),\n",
    "    col = c(\"#4A89D5\", \"#EE9248\"),\n",
    "    )\n",
    "\n",
    "pdf(file=paste0('stat/Fig2G.K562_insertion_genes_venn.pdf'))\n",
    "grid.draw(plot)\n",
    "dev.off()\n",
    "\n",
    "\n",
    "\n",
    "common <- inerstion_info[inerstion_info$gene_id %in% intersect(insertion_genes.list[[1]], insertion_genes.list[[2]]),]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "load(\"stat/K562_TE_insertion.anno.RData\") # TE_insertion.list\n",
    "IDs <- c(\"sgRNA_K562_1\", \"totalRNA_K562_1\")\n",
    "\n",
    "ins.anno <- rbindlist(TE_insertion.list)\n",
    "ins.anno <- data.frame(ins.anno)\n",
    "pp <- lapply(IDs, function(ID){\n",
    "    sub.ins.anno <- ins.anno[which(ins.anno$ID %in% ID), ]\n",
    "    ins_genic <- unique(sub.ins.anno$insertion_site[sub.ins.anno$gene_id != \".\"])\n",
    "    ins_intergenic <- setdiff(sub.ins.anno$insertion_site, ins_genic)\n",
    "    \n",
    "    a <- data.frame(genic=length(ins_genic), intergenic=length(ins_intergenic), ID=ID)\n",
    "    \n",
    "})\n",
    "\n",
    "\n",
    "pp <- melt(pp)\n",
    "colnames(pp) <- c(\"ID\", \"Feature\", \"Insertion_Site_Count\")\n",
    "\n",
    "ggplot(pp, aes(x=ID, y=Insertion_Site_Count, fill=Feature)) +\n",
    "    # geom_bar(stat=\"identity\", position=position_dodge())+\n",
    "    geom_bar(stat=\"identity\")+\n",
    "    scale_fill_manual(values=c(\"#4A89D5\", \"#EE9248\")) + \n",
    "    # ggtitle(ID) +\n",
    "    # facet_wrap(~TE, nrow = 1, scales=\"free\") +\n",
    "    # guides(color=\"none\") +\n",
    "    ylab(\"the count of non-ref TE insertion\") +\n",
    "    theme_bw() +\n",
    "    theme(\n",
    "        panel.grid=element_blank(),\n",
    "        axis.text.x = element_text(angle = 30, color = \"black\", \n",
    "                                 size = 8, hjust = 0.5),\n",
    "        plot.title=element_text(size=10, face=\"bold\", hjust=0.5),\n",
    "        legend.position = \"bottom\"\n",
    "    ) \n",
    "\n",
    "ggsave(paste0(\"stat/K562_TE_insertion_sites_number_bar.pdf\"), width=2.5, height=4)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### compare with published ALU insertion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(bedtoolsr)\n",
    "setwd(\"/NAS/wg_looking/gRNA_ONT\")\n",
    "load(\"stat/K562_TE_insertion.anno.RData\")\n",
    "load(\"stat/hg38.gene.transcript.gtf.RData\")\n",
    "file = \"database/ALU_L1_insertion_in_K562_GR2019.txt\"\n",
    "\n",
    "\n",
    "########\n",
    "known <- read.table(file, stringsAsFactors = F, header=T, fileEncoding=\"UTF16\")\n",
    "known$site <- paste(known$Chromosome, known$Left_boundary, known$Right_boundary, sep=\"_\")\n",
    "known.anno <- bt.intersect(a=known, b=gene_gtf, wao=T)\n",
    "colnames(known.anno) <- c(colnames(known), colnames(gene_gtf))\n",
    "known.anno <- known.anno[, c(colnames(known), \"gene_id\",\"gene_name\")]\n",
    "\n",
    "table(known$Type)\n",
    "known.list <- list(ALU=known.anno[known.anno$Type %in% \"Alu\",],\n",
    "                  LINE=known.anno[known.anno$Type %in% \"L1\",])\n",
    "\n",
    "\n",
    "##########\n",
    "lapply(names(TE_insertion.list), function(TE){\n",
    "    tb <- TE_insertion.list[[TE]]\n",
    "    known.sub <- known.list[[TE]]\n",
    "    bed=bt.intersect(a=tb, known.sub, wa=T, wb=T) \n",
    "    pct=length(unique(bed$V8))/length(unique(tb$insertion_site))\n",
    "    cat(\"Total\", TE, \"insertion:\", length(unique(tb$insertion_site)), \"\\n\")\n",
    "    cat(\"known\", TE, \"insertion: number\", length(unique(bed$V8)), \"Percentage\", pct, \"\\n\")\n",
    "})\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# located at the same genes\n",
    "gene_bed <- hg38.gtf[hg38.gtf$type==\"gene\", ]\n",
    "exon_bed <- hg38.gtf[hg38.gtf$type==\"exon\", c(\"seqnames\",\"start\",\"end\")]\n",
    "\n",
    "\n",
    "lapply(names(TE_insertion.list), function(TE){\n",
    "    tb <- TE_insertion.list[[TE]]\n",
    "    known.sub <- known.list[[TE]]\n",
    "    intersect_genes <- intersect(unique(tb$gene_id), unique(known.sub$gene_id))\n",
    "    num <- length(unique(tb$insertion_site[tb$gene_id %in% intersect_genes]))\n",
    "    pct=num/length(unique(tb$insertion_site))\n",
    "    cat(\"Total\", TE, \"insertion:\", length(unique(tb$insertion_site)), \"\\n\")\n",
    "    cat(TE, \"located at the same genes as known\", TE ,\": number\", num, \", Percentage\", pct, \"\\n\")\n",
    "})\n",
    "\n",
    "\n",
    "\n",
    "lapply(names(TE_insertion.list), function(TE){\n",
    "    tb <- TE_insertion.list[[TE]]\n",
    "    tb.sub <- tb[tb$exon %in% 1, ]\n",
    "    known.sub <- known.list[[TE]]\n",
    "    bed=bt.intersect(a=known.sub, exon_bed, wa=T)\n",
    "    names(bed) <- names(known.sub)\n",
    "    \n",
    "    intersect_genes <- intersect(unique(tb.sub$gene_id), unique(bed$gene_id))\n",
    "    num <- length(unique(tb.sub$insertion_site[tb.sub$gene_id %in% intersect_genes]))\n",
    "    pct=num/length(unique(tb.sub$insertion_site))\n",
    "    cat(TE, \"insertion located at the exon:\", length(unique(tb.sub$insertion_site)), \"\\n\")\n",
    "    cat(TE, \"located at the same genes as known\", TE ,\": number\", num, \", Percentage\", pct, \"\\n\")\n",
    "\n",
    "})\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "load(\"stat/hg38.gtf.RData\")\n",
    "\n",
    "gene_bed <- hg38.gtf[hg38.gtf$type==\"gene\", ]\n",
    "exon_bed <- hg38.gtf[hg38.gtf$type==\"exon\", c(\"seqnames\",\"start\",\"end\")]\n",
    "\n",
    "lapply(names(known.list), function(TE){\n",
    "    known.sub <- known.list[[TE]]\n",
    "    bed=bt.intersect(a=known.sub, gene_bed, wa=T)\n",
    "    bed2=bt.intersect(a=known.sub, exon_bed, wa=T)\n",
    "    pct=length(unique(bed$V8))/dim(known.sub)[1]\n",
    "    pct2=length(unique(bed2$V8))/dim(known.sub)[1]\n",
    "    cat(\"Total known\", TE, \"insertion:\", dim(known.sub)[1], \"\\n\")\n",
    "    cat(\"located in gene: number\", length(unique(bed$V8)), \"Percentage\", pct, \"\\n\")\n",
    "    cat(\"located in exon: number\", length(unique(bed2$V8)), \"Percentage\", pct2, \"\\n\")\n",
    "})\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "## 乳腺癌和肠癌"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### insertion with support reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##  insertion with support reads\n",
    "IDs <- c(\"sgRNA_HCT116\", \"sgRNA_NCM460\", \"sgRNA_MDAMB231\", \"sgRNA_MCF10A\")\n",
    "\n",
    "TE_insertion_info <- lapply(IDs, function(ID){\n",
    "    a <- lapply(c(\"ALU\", \"LINE\"), function(TE){\n",
    "        file <- file.path(\"merge_PALMER\", ID, TE, paste0(ID, \".PALMER.\", TE, \"_calls.txt\"))\n",
    "        tb <- read.table(file, header=T, stringsAsFactors = F)\n",
    "        tb <- read.table(file, header=T, stringsAsFactors = F)\n",
    "        bed1 <- tb[, c(\"chr\", \"start1\", \"end1\", \"cluster_id\", \"Potential_supporting_reads\")]\n",
    "        colnames(bed1)[1:3] <- c(\"chrom\", \"start\", \"end\")\n",
    "        bed2 <- tb[, c(\"chr\", \"start2\", \"end2\", \"cluster_id\", \"Potential_supporting_reads\")]\n",
    "        colnames(bed2)[1:3] <- c(\"chrom\", \"start\", \"end\")\n",
    "        bed <- data.frame(rbind(bed1, bed2), ID=ID, TE=TE)\n",
    "        bed <- bed[order(bed$Potential_supporting_reads, decreasing=T),]\n",
    "        bed <- bed[order(bed$chr, bed$start),]\n",
    "        bed$insertion_site <- paste(bed$chrom, bed$start, bed$end, sep=\"_\")\n",
    "        bed <- bed[!duplicated(bed$insertion_site),]\n",
    "\n",
    "\n",
    "    \n",
    "        b <- lapply(3:10, function(SR){\n",
    "            s <- sum(bed$Potential_supporting_reads>=SR)\n",
    "            s <- data.frame(supporting_reads=SR, count=s)\n",
    "            return(s)\n",
    "        })\n",
    "        \n",
    "        b <- rbindlist(b)\n",
    "        b <- data.frame(b, TE=TE, ID=ID)\n",
    "        \n",
    "        return(b) \n",
    "    })\n",
    "    \n",
    "    a <- rbindlist(a)\n",
    "})\n",
    "\n",
    "\n",
    "pp <- rbindlist(TE_insertion_info)\n",
    "write.table(pp, \"stat/TE_insertion_count_with_different_supporting_reads.txt\", sep=\"\\t\", quote=F, row.names=F, col.names=T)\n",
    "\n",
    "ggplot(pp, aes(x=factor(supporting_reads), y=count, fill=ID)) +\n",
    "    geom_bar(stat=\"identity\", position=position_dodge())+\n",
    "    scale_fill_manual(values=c(\"#8683B9\", \"#5655FE\", \"#39C6AA\", \"#70E000\")) + \n",
    "    # ggtitle(ID) +\n",
    "    facet_wrap(~TE, nrow = 1, scales=\"free\") +\n",
    "    # guides(color=\"none\") +\n",
    "    ylab(\"the count of non-ref TE insertion\") +\n",
    "    theme_bw() +\n",
    "    theme(\n",
    "        panel.grid=element_blank(),\n",
    "        axis.text.x = element_text(angle = 0, color = \"black\", \n",
    "                                 size = 8, hjust = 0.5),\n",
    "        plot.title=element_text(size=10, face=\"bold\", hjust=0.5),\n",
    "        legend.position = \"bottom\"\n",
    "    ) +\n",
    "    xlab(\"supporting reads (>=)\")\n",
    "\n",
    "ggsave(paste0(\"stat/sg_CRC_BC_TE_insertion_number.bar.pdf\"), width=7, height=3)\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### insertion anno"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##  insertion anno\n",
    "library(bedtoolsr)\n",
    "library(VennDiagram)\n",
    "load(\"stat/hg38.gene.transcript.gtf.RData\")\n",
    "load(\"stat/hg38.gtf.RData\")\n",
    "\n",
    "\n",
    "IDs <- c(\"sgRNA_HCT116\", \"sgRNA_MDAMB231\", \"sgRNA_NCM460\", \"sgRNA_MCF10A\")\n",
    "# IDs <- c(\"sgRNA_MDAMB231\", \"sgRNA_MCF10A\")\n",
    "\n",
    "TE_insertion_func <-  function(TE, IDs, SupportReads, plot=T){\n",
    "    a <- lapply(IDs, function(ID){\n",
    "        file <- file.path(\"merge_PALMER\", ID, TE, paste0(ID, \".PALMER.\", TE, \"_calls.txt\"))\n",
    "        # print(file)\n",
    "        tb <- read.table(file, header=T, stringsAsFactors = F)\n",
    "        sub.tb <- tb[tb$Potential_supporting_reads >= SupportReads, ] \n",
    "        bed <- apply(sub.tb[, c('start1', 'start2', 'end1', 'end2')], 1, function(x){c(min(x), max(x))})\n",
    "        bed <- data.frame(sub.tb$chr, t(bed), sub.tb$cluster_id, ID=ID, TE=TE)\n",
    "                \n",
    "        return(bed) \n",
    "    })\n",
    "    \n",
    "    # b <- bt.window(a=a[[1]], b=a[[2]], w=50)\n",
    "    \n",
    "    i <- rbindlist(a)\n",
    "    i <- i[order(i$sub.tb.chr, i$X1),]\n",
    "    b <- bt.merge(i=i, d=50, c=\"4,5\", o=\"collapse\", , delim=\",\")\n",
    "    colnames(b) <- c(\"chrom\", \"start\", \"end\", \"insert_sites\", \"IDs\")\n",
    "    b$insertion_region <- paste(b$chrom, b$start, b$end, sep=\"_\")\n",
    "    \n",
    "    #\n",
    "    if(plot==T){\n",
    "        ## \n",
    "    insertion.list <- lapply(setNames(IDs, IDs), function(ID){b$insertion_region[grepl(ID, b$IDs)]})\n",
    "     # lengths(insertion.list)\n",
    "    ggvenn(data=insertion.list, \n",
    "           fill_color = c(\"#4678BA\", \"#EF8354\", \"#87B07A\", \"#F4B540\"),\n",
    "           stroke_color = NA,\n",
    "           fill_alpha = 0.6,\n",
    "           stroke_size = 0.5, \n",
    "           set_name_size = 4,\n",
    "           show_outside = \"auto\",\n",
    "           # auto_scale = FALSE\n",
    "        ) +\n",
    "        ggtitle(TE)\n",
    "    \n",
    "\n",
    "    ggsave(paste0('stat/CRC_BC_', TE, '_insertion_region_venn.pdf'),\n",
    "           width=5, height=5)\n",
    "    \n",
    "    \n",
    "    # plot <- venn.diagram(\n",
    "    #     x = insertion.list,\n",
    "    #     # category.names = c(paste0(\"sgRNA: \", TE.name, \"-target related coding genes\") , \n",
    "    #     #                    paste0(\"Total RNA: include \", TE.name, \" coding genes\")),\n",
    "    #     filename = NULL,\n",
    "    #     # filename = paste0('stat/', ID, '_', TE.name, '_coding_gene_venn_diagramm.png'),\n",
    "    #     # imagetype = \"svg\",\n",
    "    #     # output=TRUE,\n",
    "    #     height = 1000 , \n",
    "    #     width = 1000 , \n",
    "    #     sub.cex = 0.5,\n",
    "    #     resolution = 300,\n",
    "    #     # cat.pos = c(0, 1.05),\n",
    "    #     # fill = c(\"#4678BA\", \"#EF8354\"),\n",
    "    #     # col = c(\"#4678BA\", \"#EF8354\"),\n",
    "    #     )\n",
    "\n",
    "    # pdf(file=paste0('stat/', IDs[1], '_VS_', IDs[2], \"_\",TE, '_insertion_region_venn.pdf'))\n",
    "    # grid.draw(plot)\n",
    "    # dev.off()\n",
    "\n",
    "    }\n",
    "\n",
    "        \n",
    "    # anno\n",
    "    b.anno <- bt.intersect(a=b, b=gene_gtf, wao=T)\n",
    "    colnames(b.anno) <- c(colnames(b), colnames(gene_gtf))\n",
    "    b.anno <- b.anno[, c(colnames(b), \"gene_id\",\"gene_name\")]\n",
    "    \n",
    "    # intron\n",
    "    intron_bed <- read.table(\"/NAS/wg_looking/human_ref/gencode.v37.chr_patch_hapl_scaff.annotation.intron.bed\", \n",
    "                             stringsAsFactors = F)\n",
    "    intron_res <- bt.intersect(a=b.anno, intron_bed, wa=T)\n",
    "\n",
    "    # exon\n",
    "    exon_bed <- hg38.gtf[hg38.gtf$type==\"exon\", c(\"seqnames\",\"start\",\"end\")]\n",
    "    exon_res <- bt.intersect(a=b.anno, exon_bed, wa=T)\n",
    "\n",
    "    # intergenic\n",
    "    intergenic_bed <- read.table(\"/NAS/wg_looking/human_ref/gencode.v37.chr_patch_hapl_scaff.annotation.intergenic.bed\", \n",
    "                             stringsAsFactors = F)\n",
    "    intergenic_res <- bt.intersect(a=b.anno, intergenic_bed, wa=T)\n",
    "    \n",
    "    #\n",
    "    b.anno$exon <- 0\n",
    "    b.anno$intron <- 0\n",
    "    b.anno$intergenic <- 0\n",
    "    b.anno$exon[b.anno$insertion_region %in% exon_res$V6] <- 1\n",
    "    b.anno$intron[b.anno$insertion_region %in% intron_res$V6] <- 1\n",
    "    b.anno$intergenic[b.anno$insertion_region %in% intergenic_res$V6 ] <- 1\n",
    "      \n",
    "    return(b.anno)\n",
    "}\n",
    "\n",
    "\n",
    "TE_insertion.list <- list()\n",
    "TE_insertion.list[[\"ALU\"]] <- TE_insertion_func(\"ALU\", IDs, SupportReads=6, plot=T)\n",
    "TE_insertion.list[[\"LINE\"]] <- TE_insertion_func(\"LINE\", IDs, SupportReads=4, plot=T)\n",
    "\n",
    "save(TE_insertion.list, file=\"stat/TE_insertion.anno.RData\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "TE_info <- rbindlist(lapply(names(TE_insertion.list), function(TE){data.frame(TE_insertion.list[[TE]], TE=TE)}))\n",
    "write.table(TE_info, \"stat/TE_insertion_annotation.txt\", sep=\"\\t\", quote=F, row.names=F, col.names=T)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### insertion anno version2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error in library(bedtoolsr): 不存在叫'bedtoolsr'这个名字的程辑包\n",
     "output_type": "error",
     "traceback": [
      "Error in library(bedtoolsr): 不存在叫'bedtoolsr'这个名字的程辑包\nTraceback:\n",
      "1. library(bedtoolsr)"
     ]
    }
   ],
   "source": [
    "##  insertion anno\n",
    "library(bedtoolsr)\n",
    "library(VennDiagram)\n",
    "load(\"stat/hg38.gene.transcript.gtf.RData\")\n",
    "load(\"stat/hg38.gtf.RData\")\n",
    "\n",
    "\n",
    "IDs <- c(\"sgRNA_HCT116\", \"sgRNA_MDAMB231\", \"sgRNA_NCM460\", \"sgRNA_MCF10A\")\n",
    "\n",
    "\n",
    "TE_insertion_func <-  function(TE, IDs, SupportReads, plot=T){\n",
    "    a <- lapply(IDs, function(ID){\n",
    "        file <- file.path(\"merge_PALMER\", ID, TE, paste0(ID, \".PALMER.\", TE, \"_calls.txt\"))\n",
    "        # print(file)\n",
    "        tb <- read.table(file, header=T, stringsAsFactors = F)\n",
    "        sub.tb <- tb[tb$Potential_supporting_reads >= SupportReads, ] \n",
    "        bed1 <- sub.tb[, c(\"chr\", \"start1\", \"end1\", \"cluster_id\", \"Potential_supporting_reads\")]\n",
    "        colnames(bed1)[1:3] <- c(\"chrom\", \"start\", \"end\")\n",
    "        bed2 <- sub.tb[, c(\"chr\", \"start2\", \"end2\", \"cluster_id\", \"Potential_supporting_reads\")]\n",
    "        colnames(bed2)[1:3] <- c(\"chrom\", \"start\", \"end\")\n",
    "        bed <- data.frame(rbind(bed1, bed2), ID=ID, TE=TE)\n",
    "        bed <- bed[order(bed$Potential_supporting_reads, decreasing=T),]\n",
    "        bed <- bed[order(bed$chr, bed$start),]\n",
    "        bed$insertion_site <- paste(bed$chrom, bed$start, bed$end, sep=\"_\")\n",
    "        bed <- bed[!duplicated(bed$insertion_site),]\n",
    "                \n",
    "                \n",
    "        return(bed) \n",
    "    })\n",
    "    \n",
    "    \n",
    "    #\n",
    "    b <- rbindlist(a)\n",
    "    b <- b[order(b$chrom, b$start),]\n",
    "\n",
    "        \n",
    "    # anno\n",
    "    b.anno <- bt.intersect(a=b, b=gene_gtf, wao=T)\n",
    "    colnames(b.anno) <- c(colnames(b), colnames(gene_gtf))\n",
    "    b.anno <- b.anno[, c(colnames(b), \"gene_id\",\"gene_name\")]\n",
    "    \n",
    "    \n",
    "    # intron\n",
    "    intron_bed <- read.table(\"/NAS/wg_looking/human_ref/gencode.v37.chr_patch_hapl_scaff.annotation.intron.bed\", \n",
    "                             stringsAsFactors = F)\n",
    "    intron_res <- bt.intersect(a=b.anno, intron_bed, wa=T)\n",
    "\n",
    "    # exon\n",
    "    exon_bed <- hg38.gtf[hg38.gtf$type==\"exon\", c(\"seqnames\",\"start\",\"end\")]\n",
    "    exon_res <- bt.intersect(a=b.anno, exon_bed, wa=T)\n",
    "\n",
    "    # intergenic\n",
    "    intergenic_bed <- read.table(\"/NAS/wg_looking/human_ref/gencode.v37.chr_patch_hapl_scaff.annotation.intergenic.bed\", \n",
    "                             stringsAsFactors = F)\n",
    "    intergenic_res <- bt.intersect(a=b.anno, intergenic_bed, wa=T)\n",
    "    \n",
    "    # loading gff file\n",
    "    gffFile = \"/NAS/wg_looking/human_ref/gencode.v37.chr_patch_hapl_scaff.annotation.gff3\"\n",
    "    gff <- rtracklayer::import(gffFile)\n",
    "    gff <- as.data.frame(gff)\n",
    "    \n",
    "    # 5UTR\n",
    "    UTR5_bed <- gff[gff$type %in% \"five_prime_UTR\", c(\"seqnames\", \"start\", \"end\")]\n",
    "    UTR5_res <- bt.intersect(a=b.anno, UTR5_bed, wa=T)\n",
    "    \n",
    "    # 3UTR\n",
    "    UTR3_bed <- gff[gff$type %in% \"three_prime_UTR\", c(\"seqnames\", \"start\", \"end\")]\n",
    "    UTR3_res <- bt.intersect(a=b.anno, UTR3_bed, wa=T)\n",
    "    \n",
    "    # CDS\n",
    "    CDS_bed <- gff[gff$type %in% \"CDS\", c(\"seqnames\", \"start\", \"end\")]\n",
    "    CDS_res <- bt.intersect(a=b.anno, CDS_bed, wa=T)\n",
    "    \n",
    "    # TSS+-500bp\n",
    "    TSS_bed <- gff[gff$type %in% \"transcript\", c(\"seqnames\", \"start\", \"end\", \"strand\")]\n",
    "    TSS_bed$end[TSS_bed$strand %in% \"+\"] <- TSS_bed$start[TSS_bed$strand %in% \"+\"] + 500\n",
    "    TSS_bed$start[TSS_bed$strand %in% \"+\"] <- TSS_bed$start[TSS_bed$strand %in% \"+\"] - 500\n",
    "    TSS_bed$start[TSS_bed$strand %in% \"-\"] <- TSS_bed$end[TSS_bed$strand %in% \"-\"] - 500\n",
    "    TSS_bed$end[TSS_bed$strand %in% \"-\"] <- TSS_bed$end[TSS_bed$strand %in% \"-\"] + 500\n",
    "    TSS_bed$start[TSS_bed$start<0] <- 0\n",
    "    TSS_res <- bt.intersect(a=b.anno, TSS_bed, wa=T)\n",
    "    \n",
    "    \n",
    "    \n",
    "    # TES+-500bp\n",
    "    TES_bed <- gff[gff$type %in% \"transcript\", c(\"seqnames\", \"start\", \"end\")]\n",
    "    TES_bed$start[TES_bed$strand %in% \"+\"] <- TES_bed$end[TES_bed$strand %in% \"+\"] - 500\n",
    "    TES_bed$end[TES_bed$strand %in% \"+\"] <- TES_bed$end[TES_bed$strand %in% \"+\"] + 500\n",
    "    TES_bed$end[TES_bed$strand %in% \"-\"] <- TES_bed$start[TES_bed$strand %in% \"-\"] + 500\n",
    "    TES_bed$start[TES_bed$strand %in% \"-\"] <- TES_bed$start[TES_bed$strand %in% \"-\"] - 500\n",
    "    TES_bed$start[TES_bed$start<0] <- 0\n",
    "    TES_res <- bt.intersect(a=b.anno, TES_bed, wa=T)\n",
    "    \n",
    "\n",
    "\n",
    "\n",
    "    #\n",
    "    b.anno$exon <- 0\n",
    "    b.anno$intron <- 0\n",
    "    b.anno$intergenic <- 0\n",
    "    b.anno$UTR5 <- 0\n",
    "    b.anno$UTR3 <- 0\n",
    "    b.anno$CDS <- 0\n",
    "    b.anno$TSS500 <- 0\n",
    "    b.anno$TES500 <- 0\n",
    "    b.anno$exon[b.anno$insertion_site %in% exon_res$V8] <- 1\n",
    "    b.anno$intron[b.anno$insertion_site %in% intron_res$V8] <- 1\n",
    "    b.anno$intergenic[b.anno$insertion_site %in% intergenic_res$V8 ] <- 1\n",
    "    b.anno$UTR5[b.anno$insertion_site %in% UTR5_res$V8] <- 1\n",
    "    b.anno$UTR3[b.anno$insertion_site %in% UTR3_res$V8] <- 1\n",
    "    b.anno$CDS[b.anno$insertion_site %in% CDS_res$V8 ] <- 1\n",
    "    b.anno$TSS500[b.anno$insertion_site %in% TSS_res$V8] <- 1\n",
    "    b.anno$TES500[b.anno$insertion_site %in% TES_res$V8] <- 1\n",
    "   \n",
    "      \n",
    "    \n",
    "#     #############\n",
    "#     if(plot==T){\n",
    "#         ## \n",
    "#     insertion_genes.list <- lapply(setNames(IDs, IDs), function(ID){\n",
    "#         gene_id <- unique(b.anno[b.anno$ID %in% ID, \"gene_id\"])\n",
    "#     })\n",
    "#      # lengths(insertion.list)\n",
    "#     ggvenn(data=insertion_genes.list, \n",
    "#            fill_color = c(\"#4678BA\", \"#EF8354\", \"#87B07A\", \"#F4B540\"),\n",
    "#            stroke_color = NA,\n",
    "#            fill_alpha = 0.6,\n",
    "#            stroke_size = 0.5, \n",
    "#            set_name_size = 4,\n",
    "#            show_outside = \"auto\"\n",
    "#            # auto_scale = FALSE\n",
    "#         ) +\n",
    "#         ggtitle(TE)\n",
    "    \n",
    "\n",
    "#     ggsave(paste0('stat/CRC_BC_', TE, '_insertion_genes_venn.pdf'),\n",
    "#            width=5, height=5)\n",
    "    \n",
    "\n",
    "#     }\n",
    "    \n",
    "    \n",
    "    ##### return\n",
    "    return(b.anno)\n",
    "\n",
    "}\n",
    "\n",
    "\n",
    "TE_insertion.list <- list()\n",
    "TE_insertion.list[[\"ALU\"]] <- TE_insertion_func(\"ALU\", IDs, SupportReads=6, plot=F)\n",
    "TE_insertion.list[[\"LINE\"]] <- TE_insertion_func(\"LINE\", IDs, SupportReads=4, plot=F)\n",
    "\n",
    "save(TE_insertion.list, file=\"stat/TE_insertion.anno.RData\")\n",
    "\n",
    "\n",
    "\n",
    "# Supplementary Table 9\n",
    "load(\"stat/TE_insertion.anno.RData\")\n",
    "ins.anno <- rbindlist(TE_insertion.list)\n",
    "ins.anno <- data.frame(ins.anno)\n",
    "\n",
    "write.table(ins.anno, \"stat/CRC_BC_insertion_info.txt\", quote=F, sep=\"\\t\", col.names=T, row.names=F)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(data.table)\n",
    "load(\"stat/TE_insertion.anno.RData\") # TE_insertion.list\n",
    "IDs <- c(\"sgRNA_HCT116\", \"sgRNA_MDAMB231\", \"sgRNA_NCM460\", \"sgRNA_MCF10A\")\n",
    "\n",
    "ins.anno <- rbindlist(TE_insertion.list)\n",
    "ins.anno <- data.frame(ins.anno)\n",
    "pp <- lapply(IDs, function(ID){\n",
    "    sub.ins.anno <- ins.anno[which(ins.anno$ID %in% ID), ]\n",
    "    ins_genic <- unique(sub.ins.anno$insertion_site[sub.ins.anno$gene_id != \".\"])\n",
    "    ins_intergenic <- setdiff(sub.ins.anno$insertion_site, ins_genic)\n",
    "    \n",
    "    a <- data.frame(genic=length(ins_genic), intergenic=length(ins_intergenic), ID=ID)\n",
    "    \n",
    "})\n",
    "\n",
    "\n",
    "pp <- melt(pp)\n",
    "colnames(pp) <- c(\"ID\", \"Feature\", \"Insertion_Site_Count\")\n",
    "\n",
    "ggplot(pp, aes(x=ID, y=Insertion_Site_Count, fill=Feature)) +\n",
    "    # geom_bar(stat=\"identity\", position=position_dodge())+\n",
    "    geom_bar(stat=\"identity\")+\n",
    "    scale_fill_manual(values=c(\"#4A89D5\", \"#EE9248\")) + \n",
    "    # ggtitle(ID) +\n",
    "    # facet_wrap(~TE, nrow = 1, scales=\"free\") +\n",
    "    # guides(color=\"none\") +\n",
    "    ylab(\"the count of non-ref TE insertion\") +\n",
    "    theme_bw() +\n",
    "    theme(\n",
    "        panel.grid=element_blank(),\n",
    "        axis.text.x = element_text(angle = 30, color = \"black\", \n",
    "                                 size = 8, hjust = 0.5),\n",
    "        plot.title=element_text(size=10, face=\"bold\", hjust=0.5),\n",
    "        legend.position = \"bottom\"\n",
    "    ) \n",
    "\n",
    "ggsave(paste0(\"stat/Fig7S.sg_CRC_BC_TE_insertion_sites_number_bar.pdf\"), width=4, height=4)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# the genomic distribution reference TE\n",
    "load(\"stat/TE_insertion.anno.RData\") # TE_insertion.list\n",
    "\n",
    "ins.anno <- rbindlist(TE_insertion.list)\n",
    "ins.anno <- data.frame(ins.anno)\n",
    "gf <- ins.anno[, c(\"CDS\", \"intron\", \"intergenic\", \"UTR5\", \"UTR3\")]\n",
    "rs <- rowSums(gf)\n",
    "gf <- gf/rs\n",
    "cs <- colSums(gf, na.rm = TRUE)\n",
    "cs <- data.frame(Feature=names(cs), Freq=cs)\n",
    "cs$Percentage <- cs$Freq/sum(cs$Freq)*100\n",
    "\n",
    "\n",
    "cols <- c(UTR3=\"#EE9248\", UTR5=\"#EDC26D\", intergenic=\"#8C9FCA\", CDS=\"#5CB1DD\", intron=\"#4A89D5\")  \n",
    "plot <- ggplot(cs, aes(x=\"\", y=Percentage, fill=Feature))+\n",
    "    geom_bar(width = 1, stat = \"identity\") +\n",
    "    coord_polar(\"y\", start=0) +\n",
    "    scale_fill_manual(values = cols)+\n",
    "    theme_minimal() +\n",
    "    theme(\n",
    "      axis.title.x = element_blank(),\n",
    "      axis.title.y = element_blank(),\n",
    "      panel.border = element_blank(),\n",
    "      panel.grid=element_blank(),\n",
    "      axis.ticks = element_blank(),\n",
    "      plot.title=element_text(size=14, face=\"bold\", hjust=0.5)\n",
    "    ) +\n",
    "    theme(axis.text.x=element_blank())+\n",
    "    geom_text(aes(x = 1.6, label = scales::percent(Percentage/100, accuracy = .1)), position = position_stack(vjust = .5)) \n",
    "\n",
    "\n",
    "# ggsave(paste0(\"stat/\", ID, \"_Alu_L1_in_head_tail_pie.pdf\"), width = 5, height = 5)\n",
    "ggsave(paste0(\"stat/Fig4f.TE_insertion_in_genomic_feature_pct.pdf\"), width = 5, height = 5)\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">> binning method is used...2022-01-05 19:46:53\n",
      ">> preparing body regions by transcript... 2022-01-05 19:46:53\n",
      ">> preparing tag matrix by binning...  2022-01-05 19:46:53 \n",
      ">> preparing matrix with extension from (TSS-10%)~(TTS+10%)... 2022-01-05 19:46:53\n",
      ">> 5 peaks(0.05481254%), having lengths smaller than 800bp, are filtered... 2022-01-05 19:46:56\n",
      ">> Running bootstrapping for tag matrix...\t\t 2022-01-05 19:48:54 \n"
     ]
    },
    {
     "ename": "ERROR",
     "evalue": "Error in grDevices::pdf(file = filename, ..., version = version): 无法打开'E:/项目/转座子/sgRNA/results/insertion_density_around_genebody.pdf'文件\n",
     "output_type": "error",
     "traceback": [
      "Error in grDevices::pdf(file = filename, ..., version = version): 无法打开'E:/项目/转座子/sgRNA/results/insertion_density_around_genebody.pdf'文件\nTraceback:\n",
      "1. ggsave(\"E:/项目/转座子/sgRNA/results/insertion_density_around_genebody.pdf\", \n .     width = 7, height = 4)",
      "2. dev(filename = filename, width = dim[1], height = dim[2], bg = bg, \n .     ...)",
      "3. grDevices::pdf(file = filename, ..., version = version)"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0gAAANICAMAAADKOT/pAAAANlBMVEUAAAAzMzNNTU1oaGh8\nfHyMjIyampqnp6eysrK9vb3Hx8fQ0NDW1tbZ2dnh4eHp6enw8PD////2Gz4IAAAACXBIWXMA\nABJ0AAASdAHeZh94AAAgAElEQVR4nO3di3qbuBYFYB0naXrZ01Tv/7InNui2dRcSCFjrm6kx\nBiEJ/THCbiokgiCbI46uAIJcIYCEIB0CSAjSIYCEIB0CSAjSIYCEIB0CSAjSIYCEIB0yAtL/\nEOQmGQtpQJlL8P6JNGTcsAEk5EYBJB5AQhoCSDyAhDQEkHgACWkIIPEAEtIQQEKQqQNICNIh\ngIQgHXJWSJgjIQ3BHIkHkJCGABIPICENASQeQEIaAkg8gIQ0BJB4AAlpCCDxABLSEEBCkKkD\nSAjSIYCEIB1SA+nxHb4cetQbYY6EzJUp5kgP/YdZ5o/y4ewCSMhUASQeQEIaMj0kaT8Pld47\ngIQ05DyQ9BSJ/bKv3gEkpCGngPRg20i8IyGT5TSQ3AVAQubKGSA93A1Z6Qhy6XSD9PC2kYCE\n3Ca9IPF5Ei8dQS6dlm82POzl5fHxsBZCpfcO5khIQ6aYI20qvXcACWkIIPEAEtIQQOIBJKQh\ngMQDSEhDAIkHkJCGABIPICENASQEmTqAhCAdAkgI0iFnhYQ5EtKQ686RqLFkQEIaAkg8gIQ0\nBJB4AAlpCCDxABLSEEDiASSkIYDEA0hIQwAJQaYOICFIhwASgnTIWSFhjoQ0BHMkHkBCGnJh\nSI2SAAlpCCDxABLSEEDiASSkIYDEA0hIQwCJB5CQhgASgkwdQEKQDgEkBOmQs0LCHAlpCOZI\nPICENASQeAAJaQgg8QAS0hBA4gEkpCGAxANISEMAiQeQkIYAEoJMHUBCkA4BJATpkLNCwhwJ\naQjmSDyAhDQEkHgACWkIIPEAEtIQQOIBJKQhgMQDSEhDAIkHkJCGABKCTB1AQpAOASQE6ZCz\nQsIcCWkI5kg8gIQ0BJB4AAlpCCDxABLSEEDiASSkIYDEA0hIQwCJB5CQhgASgkwdQEKQDgEk\nBOmQs0LCHAlpCOZIPICENASQeAAJaQgg8QAS0hBA4gEkpCGAxANISEMAiQeQkIYAEoJMHUBC\nkA4BJATpkLNCwhwJaQjmSDyAhDQEkHgACWkIIPEAEtIQQOIBJKQhgMQDSEhDAIkHkJCGABKC\nTB1AQpAOASQE6ZCzQsIcCWkI5kg8gIQ0BJB4AAlpCCDxABLSEEDiASSkIYDEA0hIQwCJB5CQ\nhgASgkwdQEKQDgEkBOmQs0LCHAlpCOZIPICENASQeAAJaQgg8QAS0pArQ2qTBEhIQwCJB5CQ\nhgASDyAhDQEkBJk6gIQgHQJICNIhZ4WEORLSEMyReAAJaQgg8QAS0hBA4gEkpCHfw2bQnByQ\nkBsFkHgACWkIIPEAEtIQQEKQPgEkBOkQQEKQDgEkJ5gjIQ3BHIkHkJCGABIPICENASQeQEIa\nAkg8gIQ0BJB4AAlpCCDxABLSkBSkbcImgIQPkpAdA0gI0iGAhCAdAkhOMEdCGoI5Eg8gIQ0B\nJB5AQhoCSDyAhDQEkHgACWkIIPEAEtIQQOIBJKQhgIQgfQJICNIhgIQgHUIyYuaekDBHQhoi\nAIkFkJCGABIPICENASQeQEIaAkg814aE+y+DAkg8gIQ0BJB4AAlpCCDdK+iScXmON0C6SdAl\n4wJINwq6ZFwAyQnmSEhDBCCxABLSEEDiASSkIYDEA0hIQwCJB5CQhgASDyAhDQEkHkBCGgJI\n9wq6ZFwA6UZBl4zLBJAe3+HLsUdeeiyAFAq6ZFyOh/TQf5jl2KNXeiyYI4UCSIMyxRwJkHYL\nIA3K9JBk6DkgtQaQBuWEkP73Svb4gBQKIA3KKSA9JN6ROgWQBgWQeAAJacgZID34NhKQWgNI\ng3ICSFwTKz0WfI4UCrpkXF6QQh08B6SHv40EpNagS8bleEjm2wv28vL4eFgLodJjAaRQ0CXj\n0gKp4Hyc9bt2F58jQdKYCEBiASSkIYDEA0hIQwCJB5CQhgASDyAhDQEkHkBCGgJIPICENASQ\n7hX0ybgA0o2CPhkXQLpR0CfjAkhOMEdCGoI5Eg8gIQ0BJB5AQhoCSDyAhDQEkHgACWnItSG1\njBpAQhoCSDyAhDQEkO4VdMm4ANKNgi4ZF0C6UXbvkhudAkBygjlS3wPufLyjgjkSDyD1PNxd\nITnNBqTrBZA2Jd4aQOIBpJ6HA6TkXoB01pgu2WWEA1J6L0A6awBpUwCpPIDU83CAlNzrypCu\nHUDalGRrFkjkbwhI1wsgbUoOEgHSTQJImwJI5cEcqefhbgNJABLLbSDtMcQBKb0XIJ01gLQp\n7ZAiewLSSQNImwJI5QGknocDJHkNSPXnEZB6Hg6QJCBdMYC0KeWQ7C3vCenaAaRNKf8cCZCu\nHUDalBtDutaJ3BxA2hRAKg/mSF2PdhdIiTkSAdL1AkibkoFEgGRyE0i7fKMXkPQyIF0rgLQp\ngFSe60J6nU9A2hJAKg8gdTwiIElAulwAaXMACQGkDsl8jgRItwggbc52SKwEQDphAGlzOkB6\nLWgSF4aEOVLHI94HkjVHcvpZ0oUgVZ1MQOp4REACpOsFkDYn2hq6OCQCJBNA2pwKSObuAiBd\nLIC0OaWQCJAAqecRbwSJAMkNIHU8IiC9INElIF3tZDYHkDYnDknWQqJ0gSbzQMKHskuOgHSx\nvt8Cyf6kFpDOHEDaHEAqD+ZIHY94H0j5OZL9ARMgnTnHQLqOJOoGiQDpxCFA2hZAAqRnAGlj\ntkEiQLpIAGljAAmQngGkjQEkQHoGkDZmAyS6IaTLBpA2Jgkp/TlSBFLRFz8AabIA0sYA0nXO\n5ZYA0sZEIZEDiQDpFcyROh7xPpBEE6SS3gGkyQJI6YQqSuxJMyQ6LyRBZLWsPP1qO9kQAqR0\nACkcQGIBpHQAKRxAYgGkdPaFRIBUnMmG0N6Q6GaQKAjJXg1IbZlsCAFSOoAUTiukfplsCAFS\nOtsgOWIAqWsmG0KAlA4ghQNILICUDiCFY1+11pSMOVKv490K0nXnSIDEoiHR8gyQ3OQhxb4a\nB0ihXBoS7Qnp0Ovq+lRDImsBkLwAUrfj3RUSAZIEpI7HAyRA6pHJhhAgpQNI4QASCyClUwSJ\nQi9GIZEHaV0v6PqQ+mWyIQRI6WyD5ASQumayIQRI6RB75MuABEivAFIk5DyMg0T3goQ5Urfj\nnR0Sv7kQhLSsLJkjWesBqSaTDaH1RJZC2lz720NSb0GjIL39/K+whUWlxwJILIAUSQySd7t7\nMkhCiMePP6WtzJYeCzV+/RuQ1s07HO/WkGg0pH+/P74tifffX6UtTZYeCyCxAFIkR0Eq/E1x\nyTnSn8/Ht6W39vclQKrOTSHlj39mSPLrU7zelvLl5EqPJQopU/crQyJAim9xRkh/P15vR/+9\ni498QZnSY2mF1C9HDyEWQEpvsQ2SD2Y8pD/v+qpOtP783wRpp3N79BBisSARIFkbnBXSmxAf\nf9VLj3xBmdJjmQDS0WPIDSBFNjgrJPH5V25OJSS3/cnad5wjHTWGwscFpPDrfSAdMEf6l9+3\npvRY3KZZFY79rZI1gJQqpKYWgLRCEs56qbfeOkf6eK0Qbzt+jgRIgBR7PQaJZof0udxhEOJH\nvpCC0mOJQ0rfxLs3JAostdbivpCsUTcQ0kO8vmv3t/mOnVt6LIDE1wJS8PXTQlKAAGlUAMmp\nRu71Mkg0HaQP8eOflP8+27/WIBsgWSMEkG4GKVmF80L6ery+HSQeW+6Cj4PUL4B0I0i2F3vU\nDYT0/Wb0JsTb55abdnNDUmcGkI6HRJNDyl8AzfQ3ZAFJ2pAIkPTLOUhkb+lDoi2Q6NKQOtQW\nkKzj3QSSKIdEvSB9rpOkXe/a7Qzp1cmAdA1IpLbMQ+KYRkL6FGIXSLH734AESPbLCUg0NaSH\n+JXduaL0WACJrwWk4MunhbTpncgvPRZA4msBKfjyjpBUkX0gfYgO3/8GpPTxQ2sBKfhyEhIZ\nSDQbpK/H+6aPkFjpsdRDWpa2Q6JrQqptzXSQQjVpgmTtcCQkMevNhl6n/ASQoh+qx3ujujWz\nQiL2ch9IdCNIqv0Mkn52J0ixEQ5I1A4p8FnSSEhd0g6Jn9sRkAiQACkBSdwBUo85EiDJc0Mi\nmYdE3hxpd0i/Pr4v6943/QoUQEpVILwWkOwH/bIHiRgkmhTSv7fX/EiILf8oBSClKhBeC0j2\ng365EhJth0R9IP0Qn88PZX+P/ot9YUjeuQWk8K6A1ASJf8NzIKTn3Tr1f3MAKVWB8NoySHqu\n7e1eWwtAAqRNUYc4AySPyykgVZR5YUjrpd3n6F/HlYJE9nZ6KVTbhgCSfTxAGgXpn/qdDYN/\nQWQjpO0BJPt4V4JEcUi0PyQpf75+Z8Omr642QiJAykCiM0JKHCMgxd7r3JA6BJBSFQivBSR3\n4wpIlIFErZAKbhTMCcn0htlOL73+vPocie4BSTfD1GE97R0grW9R0oLUOEfaAunIL60CEiD1\nhkSAtG6nl0K1bQggucfrH0Ay+Xr/mS2iqPRYAImv5ZAosCkgnQ2S/Ce2SJoeEp0Nku6gSCHn\nhkSmbUdDot43Gw68tKM9II0ZQ0UVCK8FpHZINC+k383/ELNbeiwhSPYzs93ycCVIwQMHINlV\nvAckOgASX9f9ZsNntoyS0mNphbQ9oaPsmciBw5DsYWUGm7w4JP3W1A0SHQbpscURIKWPH1l9\nW0h0RUhdAkjJ40dWA1IKEqkdAMnO0XMkQLoPJDp2jrTlQ9nRkLac/tNAcuoISKWQqBaSd9vu\nepAIkPTL5jpFPWe711fjEpCoApJnZiQk+fn8BUJf75t+cXEJJLf+gJSARBYkSkIqb9NFIFEE\nEh0OSf0S/eG//ASQ+OogpHVgZSDFfCWrcSZIdDpI677/hn+zgUEiQLozJKqDxDm4t2jkDJDe\nxfOi7utdfOR7rKD0WAKQ7GfrRoCUhWSPMxoIqaK3AOmZv3v9zoY2SPqF5nhH2TmRA4c74gKQ\ngpWKQqIxkOiAmw3/Pp+/s2HT36JogMQauG4ESOWQ6NyQ1saSbm8ckvdjd05Ifh7f4ctm3UOv\nN19z7QKJAKkIkhpx54MUaOyVIT30H2bZrFv5uN8VL4QUpbRsgjlSC6TCZtV2wvGQaH1sgXTE\n50jev0aRhPSQgFR9/MhqfW7dsVULqaxdtZ0ASOFU/GsU6Xck52W/9FiCLTANXLsZkCTRsZCI\nPZYUzZ5NBMn/e+XDIPn/GkURJD1F+t8r2eMHW2APcboypNDQss4tGyRK0Y0gUSdIWsT+kPxf\nol/8jlR9s+G+kPxj2+eWDZKlHwBpfWQ7XQsSW+oBiQCJNCT7pWVzu7irQKIrQfL/NQpA6ho6\nEySaAxKdEJL/r1EcdGkXhiTVU+exJs5RDgglIIkIJHPa6fKQKAfJHy0eJOKQaH9I/r9GUQrJ\nunM3ABIBEiBtghT4V2THQvKiv81gLwe+2RAqPZZgC0wDAckaJIB0RkjvW/6lPr/0WIItcFto\n+sOC5M6RAOlikHQ7OkMS+0N69PhlKICUPH5vSARIzsiZAtLf988tf4GClx5LsAVuCwFJjwur\nU04BiaLPnNJKIVElJMpBEtbG43/5SXnPJUqPJdgCp28ASY8Ituq1uVMcILkjpxqSMxIBqShs\nRO4euhuk0DEmgbQsj4HUJb0gWeMGkJxqkx6MJZBCq4o64SyQ+BNA8vtm2XwdHwRI94NEGUiB\nkXM0pE3Xc8HSY6FGSOve0n2sCSt591AAErVBIulAci6GefmBahR0wlZINAkkOgJSB029IREg\nsWrbY+28kGInfyAksfazvf42kAiQiFUbkJIjJwVJANJl50i9IFEIEgGSmSOtbqyRd21IXt8A\nklXtdXN3rAGS6YoIpGVJQRLqpbtBevXMBSFRAyT1mtMcawRJr/xANQo64YKQxH0h6T66LiTq\nBSnQritA0lXNDxLTbYdC2v5vI7mlx0K3g6TrfFFIZC0dCMnqNkCK9c4L0utBldJQN6e8HdMG\nSY0KpxvM+t0hxbcHpOy29RkFiRZIBEhEZ4BECUiUgESAxEuPhRohyWtDEhNAcutWD4lKIaVP\ndldIdGlI/r8nnekdQFLdoNb3hmRfNy/Pzwlp3dx+CkhWz1wEEt0NkrdXDST7fCV34SW4WwOS\n0zOXmCMRIGUhBdx4KzL7W8uZOZIApPo4J2bH9INkLWyHRN6GgARIJXFOzI4hNU75sXeCxN97\n/K0BaV0j9F7b/1Vzaf+auuoUQSJAWl7IQYoNEAYp1i7Xh/vGSDtBCtYpe7oPgPR6YTukx54f\nyALS+sIGSJENpN01+0CiOkjUAqmgc0L9RPtD+mU5+pUto6T0WIgASb0wCJLVST6kUC8AUjdI\nstPfNp8YklPenimEVDFA7gYp3uJ0P1EeEhk7vSB1yThI8nyQ9NAFJOlsk21nN0iCQ9KfHQkX\nkugI6fMx8xwJkPKjSUMideAKSLQHJDoE0rq4E6TPuW82AFI5JPIgmb5jvUCqInNDqooDSewO\n6bHpLoNXeixEbZDOM0citqB/CPBjk6QRkPRRQpB4L1wRkpUjIOFmQ6cMg5TvpzAkOgISTQeJ\nkpBI9IL0If7JzRkDiQCpsJtUO6V6R5CAFIHk/o6ujpC+Hu87/bMut4JEJ4XEu3pXSN0SgiQG\nQ5r4mw2AVNpNsgyS0DUBJAZJXfdlTzMgAVIYEl0TkvUus8BRvy6S+ekGqUsAaSukLaISkOhE\nkDbG7kFAcuP+VJXe+SmJXdzQBCHRqSARIGVP81kv7WxIVG7BGhxWcUMzESQCpDQkAqTCzAGJ\n9oZEgLQsHwBpydf7z2wRRaXHsjSvoXtuCqmmq2ogCUDSXTFkjvRPbJE0HJIaDaUVmh6SOuGh\nYRB6nkodpIXPWtETQ0r1WAKSdVt8zM2GGS/t6AWJLEilGM4HifVNXVeZdkprhYJkv/+lIVk1\nLYRECUhOwXNDEt0g/Rbz/c6G1z6AVBC5DZLloB3SWok9IQn23CwXQhLUD5K+1/CZLaOk9FjU\niazurhNDovCxQ5AEl7MREqUg0bUhiTJI61dXe0N6bHE0GNIz0j3N2cM5Byaq2bcxR0KiDCRz\nNLE+Xyu67mP3lducs0Ki3SF1SRmklnu8gFSUsZBCPVcLiWaBZDrjVpAkINV00whI1AKJ9Bam\n4Lom5SLYJU4EEu0G6d/nmxBvn5v+VlIpJLctJQGkqm4qgaTrE4bkXtNNBYm/A00F6Wv93SeP\nLX8rCZCsYaNvjFD42FeARBwS2ZBMBeyCe0MS/SCJDpB+iOdf7Pt6Fz+yZZSUHouqfO0th9PM\nkWaDRB4kcQ9IIghpab1dAoO07pU9z9nf2bDDB7KAtG7onEx3gQ+LwngHtCDpIu2ht1Z03dbp\nK6cHqyHZNbDaTLWQQh0wHlL+PE8CiV/Y5vsTkEoShEQ5SNQJEp0UkugJad9LO0DaCVIg4yG5\nVTJb7AFJxCEJpWUopH1vNgCSD0nQXpBI9IRkXx6eHRJth7Tv7W/7XJb1Z0dIQyVdGFKg42og\n0ThILhdnvQ+JspBKRsgkH8gC0jyQiEGijZB4lTZCCvRAMyQCJAlINd2UytGQ6qJq6tw04M2J\nQqI8JDXTWG5GbIb06/W3J/778TdfRFHpseiuAKSJIDm1Gwepri26ts5jFhIJa/3ukN6FeBF6\nbPpbFBNDYoWNy2BImU4DpGMh/RaPP6+F/x7id76QgtJj0Y31hkymP28NSbDn0ZRCokpIFIOk\nNiSaEpKzLYMkTQndIL2LP+vSH/GeL6Sg9Fh0Y+8HSbiHPg6Sc5iTQmI34+KQ3IaPhmR9nWGX\nbzboN97y/rwMJDIbumd/vRBxuwWQfEj2B0EiAslbTwoSZSCJXpDG/86GuSFtZFYIiU4Jyesb\nvSEgvfIu1PcZvnb5ihAg+ZBEEyT2fCdIbreuG7wWo5BoGCT2/YUtkGgrpF+azw89W2pJISQC\nJANJTY12giSOgET1kLSUHSCRA0lQOyT5EB//fT/89yHe6sZNrPRYTG8A0jGQRAMkGgEpff5r\nILkdlIW0rg9CWg+UP88RSOorq9u+szoQEl0Nkm6SdYG+CZKweimRK0ESWUh+kYMhSfn747uC\nH1s+RJKAJC1IFIOk108IiawqW91BgMQzxXftdHsqAkhsiT/vBYn10mGQRPiR9HdL2yCZbZyx\neBtI55sj5SGpcXEEJEuS3TH7QcpdkZRC4l/5yc2RzDY3hbQ+yjNBogJI5rriIEhOx4Qg0XSQ\n1HcFU5AEAVKwX9dHdaYKKzQlJGf1/pBChzG9G4Vk91wUUqAGM0DyLioBCZDC/VIHKVDQ0jFt\nkKg3JJGAtPRVASRrllQIyTzNn2dAyu08FaR1+Njf2Az3y4khBWqiqZgeIA+S/U0EoQtqhiQA\nKXU4ftz8zntA4ie1G6SCjkxBEk2QKAmJvI5YD+VdszkvnhfSp/pMtn0QAZK+lxCFxM9rCSTr\nMsiDFH0l1pEngUQRSEIzK4TkvRcOhvQpxJ6QdAMqA0hudoFk1TncrfWQxAZIVqdFIAkFifaH\n9BC/to0gt/RY3PbUx4ZU0NypIAlRCsnoUINDOCvsQit7syMk2gTJvx1n16MBkpgD0qZ3Ir/0\nWNz21MeBlG8vIPGMhhRKPST9hsUfDSRhem4jJH1Iq8bZRCF9iE2/GpKXHovbnvKE5kiABEgZ\nSEVzJH1Iq8bZJH5l8fumL36z0mNx21OeO0Gyp0QGkrBWuHtX9WY1JDd2H22AJK4LScx9swGQ\nZoeUCCDVJg/Ju41VGEDqBSm1ubCrFug4CxKdABLVQKJOkLoEkLpAEg2QSn/h5hGQ/A2EiLVI\nV64U0tJp6/+GFyCFA0gXh+TU59SQ9vlnXeweZecyORAAKQJJD7D+kMjruEtD8puZytH/0Fgr\nJN3C60Far/OFutwXQUjCGXbO3qZ0nsSqrpDCt+yEvRt7aTMke3JE+qk1c1Il7A1pp3/6cisk\nOhckp305SEJDsi7YyLyoVvC9TemR10K93ApJN68OEn9JNVJsgqTBpCD5t+aHQtrpH2O+GiRv\n6ygkUQiJUpCsV8ze5L7ilOwFkO4Nib+4lJQ/mH/UXDftD0mNigQk2gyJAxoAKVAB53XnJX1Z\ntgmS1U9BSGJ/SJNf2t0KkjXcCyEJ6gTJWt4HUvAuShkkQcWQQnfo7nmz4ZyQnBYYSLxh5jql\nEyRhv+Z2YBkkNQbPCsnqgEJIrJnZzHX723s/Dz8JPF9Kyh/MP2qumw6HZDdXKCK7QxIrJLZd\nX0jChWQ7HglpeX0UpB6pg+T8FEqpCjwneS9I1AmSYM8jvXwgJGH9H4NEgHQ5SHzzBCTKQvKG\neQska0SeA5IwK6gCEk0ISYjdvrR6E0hUC8lgMc0VZCAZMO4WCUh6yk3eQxiSffw6SMEis5DW\nz57DkNwGC9LidFUBKdTV7APF4A9M1tR7Q1I7MEjWiCyFZNbtC4m2QNKVNPUNQKJ7XNrZfX0P\nSGJSSMKUqUrZE5KgHpBWQiWQQnVympkNIOW6aT5Iwh9Xa3+p0WND0nOt9dsQ1rWQ3a8ZSCIL\niX+V0N7dO0QEEl0Qkrqke4z9x5jjkPxPEiInRTd1akjemUpAogAkdb1TDkm9Dw2FpDt+WeQt\n9JtGga6wfLRCWtflIImdIT2ElaqBFCs9nkhfC6ePUidFNbWkvY2QyjHVQeKtIjNcopDEPpDY\n8TtD8lttQdK3CpKQSBRB0r2hdtsV0i/L0ZZfbwdIa63IP1MbIDkb6uIcSGa25EFy3m32gyT8\nc8ZaTdZx45DMoxBuoc5FjDCtOBCS3O332tVDYgPANLWkvaMh+YWRgcROXiUkNTASkBQfoX8g\n24ephSQAyW5kJlFI71u+rOqXHk+wQ+8AiXxIZpiEIQlrp1JItF4NGoh2h04Gia4H6XHUO5Lg\nS7pf7D7fERKdEpLavQ6SGaxdIQkKNJY13Nm0EZJdju6qgyH9ff/c4xdEApKzTjiQ7GG+BRJF\nIbkDuDska/wOgeS8B6UgWZvsP0c66K6dcHvM6hfe56ypJe2dGZIR4F6krCtIXBYSG+o2JHV1\n5tTNNKsWEgHS1SD5Y8uCREFIgkNyLlrqIJledrp1b0jqqKy1PSDpQx4MqUtaIanz7vSt0/dE\nrPNHQFo/ApoA0nrzzd7JDPd1A2HkeJAoC4kPVg6JRkLijTUDIAJJ6K0ASYW3SF/DuJ3E/DBn\npe1thlRIyS+MOkEiA0mvYpBEMSQhcpDYBGRaSKwFOUjiCEi/Pr4v697/lo2gXOnx8BYBktte\nvjmHpN+gHEi6DGE2NS8OgsR+3ZhVVdEISTiQlC0bklvYhJD+vb3mR0L8VzaEMqXH4/fFVJDo\nGEi8daY/GiGZblODWghih78tJL/yfiMzSfwWoc/ntxt+i/eyIZQpPR6/L/RFCiC5m7s9EYFk\nqamGpKnq/0z3P8cg7QxJpCB5+80J6Xm3Tv3fnGpI65yZ/ex1ljtA4h2Z2ppGQuIjuRGS/vnM\nIfmrHEjccRTSemkYh0QjIFk/Uq1KAFIkrA/DkKzO2AZpmfbsDSlwagZCsu8/qHIsSOptbBpI\ngVcikPQTYe1uFzYhpPXS7nPHXxCpOlVfrYQhCa/Tytt7DKTQqWmGZG0h1HgjBkn3pQ9J6BE5\nDyT/VBrwFZCcFUWQnPqOutmw/y+IXLtclEAK9eJoSAXd6RcWO0P9IK0sApAoA4kfvgKS2XP5\n/bEFkEInrAqSqIVEHiTaH5KUP/f+BZFr061LE90p94MU3NxptwPJKq4CknsI65J6CyTn8Obo\nbZDoApA6ZACk6M/sova2QyIfUmA/v7DYGToIkhLUCIlaIQk+nq1i0pCWZat+KUi85DAkeRNI\nqoNSkESwFwsO9vrzSEjWGNgOiWxI9ujRG/CdiiHZc6g0JOdPweqnig6fsRgkdczekEQcUvAG\nhHkxlzikXw8p/xOPn/kySkqPh7UfkKIJQqJukMhAUm905u0sB0kvhyFRPSR9cbIFEi/+AEi/\nhFj+RYotkk4CKbafgUQVkMhfw1qozrEPyR/d7gbWgwXJLcieRFZB0n2rPQGSfjGXKKQ38d/3\n/7/+ijsd9NUAAB68SURBVF1/HZcDidh50Mt0Hkj8dPWEFOoed71/Z6YCEoUhCae8IkhKdBRS\nuLkRSIEBEtrb/GE1VEPy7zKmIOVHcfID2T/i7YAPZN2fH+ujuXY5DJL/WdIQSPGzWQzJLjAE\nae284LB+vVwGaX1xP0j+uU/1lin+WEgP8fVD/H3OkgpKyZcej9979hNyuvBoSO5cyd9t3Sza\nOtUoXXvB1zdCst9d0pDinZeEpPbIQ3KnaE5Fbgnp5/PD2Ocb0mdBKfnS4/F7z35CThfanRm6\nCCo42OvPckhOF58YkvNiDBIxSPqIlgIb0mtxKKTMD9ESSJL1zd6Q5Kd4/Pl+Y9riaCAkNkbU\nT8f8wV5/1kEym3WB5FzmuI3dD1LqGHWQeAEzQXpGsj7YHVKPdITE3t4BKXxTwS4yBCmXvSGF\nV6vXAMkrPZ5Az/InMUjCHjNFDb4DJHt7QDoe0qf6TfoFpeRLjyfQs/zJ+q1MtzMvB8n+KRGM\niDwV8RHVAklQHlJcfT9IxCDpI26AJPaH9CnEPSCJ+H6TQfIKCi872zRBst4ClkJ8SNHKet8r\nIhsS36MOklNHSuxppwpSBFPBME7c/t7yz1B4pccT6Fn+5GqQRGgoLOsnhLTW6ghITimsjnqL\nTI6GdMy/RuFDEs7IGwFJFEOiEkhUAMn/fKcBUkkOgORciAGS/BCb/iYSLz2eQM+6XbaOu+kh\nURGkZVD5kIQNKXG5XpdGSGp5OyR1GbuI8e/Gpy5no5Cqft74kNazRftA+nq8H/BL9M8NKdDt\n3jm5EKTIrcK+kAJ71UJyi7TP1j6XdsffbFjOgANJHAuJxkLSxwifz/psgkRJSKIYkmiEpLfZ\nAZK8OiRhToXq1HUZkArSCZJePjskKoUkQ/ukMvUHsgYSrXjoIpBoXkjsB/h2SKI3pMhF5Rrv\nahyQyiGJHSFRDSR/b+G28cSQwoXrS28XEvWElGwWIK195D5x/1KkgST6QwrczR4EyW+ungmG\nTuOGDIIkyiBREaRcFbdCcmu3LyRhp6CUfOnx8D5yn9hXChYkcWZIgadTQ7KWt0JizRdFkPy3\nrH6Q9DkLQLKY5QNI+0LyJs5q9SSQ3D3UrYJl2bxeAEmodzMjazpI1hlaF3tD6pUSSHySxHss\nBcm9cJ8ckn2dxJs7KyTRA5Le3UxwA+U3V5IFkHRT3R47FFL4TOmXw1tH2paHZB2iU0ZBis9u\nrgFJ3gUSjYdE7ZDMC17L5ofk7u72PzmQgrfbAOkEkAQgVaZhkDq7b4ckRkBKtSsFSQLSimc9\nEQskQf0gqX61Nm2GRA4k3aAIpGWEnQiSSEASVZDE7pAoBckaF2MhPb7Dl826h7dNB0j2XTvr\nlMwOSW9BBZCEGm9XgCQKIAlr68j3I7IVi7+0DZJUT2kcpIf+wyybdY8H30b2gBTuxFNAogtA\nsuu7LHmQwpOqE0NaNqKDID1kJ0hFndgRkqApINEYSBsThKTvgPeB1IS99xzJnNCjIck9IS2U\ntkIy0y1AimRaSIlcDdL/XimoQVdIuSYLM8RFJ0hkdrC2keWQSJwKEukbdyFIJTcbbEi0MyR5\nOkis9ERqe+qakOiMkBxR6w4ckl2csLbQWwPSQZDU+RgNiTZAImkgmWFDUUhmg/UQc2Q7JAIk\nL1eFRAKQoqmGpC/4hL2XKcxQ2xuSzEJSp92CJNnL0UwAqf7abjJI+ueZvZG5pVEMyT7Erkkc\nrj8kQRYk/p7VIYBUHA5JFEOi/SCpZ+xrTB6k4Hz7RJBEAJKI+AhBokpIobrK1AYOJJoBkvk2\ng728/ZsNh0ESUUjqYhmQopCWF0UVJDoBJOlAkkMg1edQSIHmM0hiJCRV2XNAih0xCYnCkMTh\nkIJnApB4AKlly3QhUUh8UdjrgpCoEJJ/j6Wwrql1Mg1JSHaWASkOyR3n7pE8SGI/SMErnvCg\nmxOS+0ayrtJfc7A3j92MGwJJMkhua2Qekrw1JGHOqf1phyyHRIBkCqmB5KxqhRS6619YV2/N\nNkjytpCIQ7LOmqyDJFKQIsPLvsyWNqTg+b4SJEEJSDaz1MdDnSBJaw0gLanqyGe6QaJGSPZ7\n0XZI/lCaExL5kJQuDknEfRhIApBKc3ZIkVN5I0hWAKlsqJ8IEumr8hpIVAUpdSq3QAq1Yx9I\nqR8N2yCJtTNrIZXV26ur8whIJi39eUpIgYnDDJBkGSSqg5Qo5BBIq5ciSPJ2kJarvFZI+o+7\nQ4rcn+QphpS6XNPX5YBUnjkhkQ2JAKn4JFwSkvQhrf0BSGuxUvJ+mxCS/iO4dzbng0QjIFkT\noSQkZ5ADEk0CiQBpA6TMRrHKssehkCQgyQJIFIeUGqBTQ0rWO/Jq8UkIQSJAeuWekOg0kHKd\nA0gMEj97igIghWNDInuAboBEVo2ykIhD4hupihZAChzCX4pVJv2cnKG3G6RMOYGPpEoThKTO\nKSBVZzSk5KkEJL1iBkhSAtIrLf3pQCIfEpVBWsvg/ZU89HIZyCBFR/xxkPiQmwcS1UJKtMpc\nV7dBWnew13qQ6oY6IBVCoiCkaEVtSP4dX0DyapNeVw5JndN1c+ZFxiBJF5JzpZIIINmQqKxG\n9ZDENkjx4gFJnzF9w8iHJAEpeioaIYn1zdo+i2lIgdO9JyTpPo9v6T3fBZKZpdb8wxL7QFIn\nSJPg484MQH+92avIESBRF0iJirqjzX85lByk+KVfMSRnGlF4EsZCkl7FgpXlq/QZK4EUHuSA\nBEibIamBEqm+s30aEtVCYuUDUknpqRT2PjsV+rEFEiUg0U6QZKR0IqvA3SAlCtLPvb9mZDeN\nKiAFfqJ0hiRrIElAaoDECqBdIbFSYukBSfIht0KS/trQWaiE9Fp9BCQzhvpDWksuGsRzQKq3\n1BcSAdL0kPx6L4+69p0hSUBahncVJH1y5CBI4lhIUtXaDEgzMNVZiBRknjdDktafFN5yMyTZ\nDdK6iuIvBnIrSHQ0JF5KLAxScJR7g5M9V5ercjgkQZshySwkyV7Wz/UYAqQWSPqMHAZJncZ9\nIakhKfng9IahD0kCUmwEhlYB0q0huXtmIZk5AKtYEBIHMBCSdF8GpIKMg0RWfwyCRGlIqn4d\nIckEJNkASZ8F5xhu7aSG5PWN/sV2IyFJDkkmIEnZBVLixUDOCqlsjjQIkskekORWSNKDJKkA\nkl2/PSF5G3eEVDNHSrwYyAyQ+I+/kmhI6ym0+3xKSHx6sTsk1c89Ib1aOA6S0wuAlM9skIJH\nCuYckMwQIwZJdoJEfSFJXU3TCxlI+pyrBi9tUm1et7k8pGpJzZB4ATQ7pKXudhVVw5xCAAmQ\nTg8pXU8Nie+bLlYtLXW3q1gESS3ppqty9JWe2oofohaS+XMTJCMjBMm87DYPkOxsgUT9IYWP\nFEwxpEAxIyGZgambrsrhkCQ/RBwSZSBF+8qD5DdYGhlpSOwnhhlCAUiqGzZBKgsg7QHJv66j\nYZAWII2QpDkGIFVlHkj1mJZTVwJp7Uf7jeEckHQP2VXsAEn97LJPgi4lAYlikPxlv6k9IPHH\nSkjxAViwKhlA4pD8IyRyEkgyDMkZLaYUae+vSm+AJNlSGyTpvOw86pfMMNKLgFQXwS8qCiEF\nT3+oFqHZjUkAkuDPgoqoDpIeiX0hOSdBl6Ih2RKKIMUaWApplUHrGyYg+aUn0wBJsCUPEvWE\nFB0sPiT/22hjIEk2OHeClH7/jTVwDCRZDwlzJJ4QJLH08N6Q2PaDIekxUglJ72MgsZOgSzkK\nkjX8U5Dk+piCJDdDqs2lIAm6ASQZgaTGVzskSUFI60jM19pvoLvTAEjuLO/2kOQWSOa5C0n9\n9N4XEp9SieVjzOC+kSLtYq3+UfUrgCRbISmFO0PStQOkbOm57AGJNkBK39hlm3v12gWSGnEG\nknXpQ04/hyCZQjdBkvYiIHVKBSR5Tkh+vUKQgpuG2itvDkk90toOdXjTXn1u/SbkIElA8hOF\nRIdC8i4DM5D40XKQdF+ZVWZL6ghJqv1TkCI/CsziNJBUACmX60OS1AuSnB2SHAYpMfo2BpB2\ngBTeNgNJjZe1f0ogqXamIJnH+SDx4bAd0rp5dNh5GzcHkNzlK0DS/RkYQxRctFYS6fHYF5Ic\nDkkC0pJYP4dSNEeiS0GSbMDpQtaKO5CWBQKku82Rnon1cyijIHlnexgkGYYkI5Ckd7o1OzXW\n9oQkrWXWArsx1srtkNQSqeZ2g9QhgJT4uFUb2gXS+q5TDUk2QoqegwGQ1qvsSJoh+dUPLQFS\nIpWQ1lXHQAonCkn2gyS7QDI9amo8GJLVPkCSO0MyXyCQFiSaGZLT5l0glZwDD5IcBkm954Wr\nsly9URMkp0l6GZB4AGkYJNkbkuwFSQJSUaJjLpBaSLQNUum//NMHkrwXpMgFZwCStCHVNWkJ\nIOXyGu1q2ErntMvNkMr/meF+kGQRJKlHqOwDSbqQbFimR6kJUugE7w1pXC4KyT5RchSkwLoq\nSLIbJAdMN0iyPyS+ByDVll4S70xUZAgkYSCZjczWe0KSMgbJAzMPJFUAIPUrvSTemUgkOEca\nBsne0Ybk7TkVJDkFJDkTpHvMkSQ/E4n4Q1hEIdE2SPa/ySLSkPyPX1shyWUIud3DeysMiW1T\nmXkgSUAKlF4UfiYS6QkpOtpvC8kccXnaAEnWQIpUxYXkVKSqRYAUTxISNUHyx3wbpCKjfSDp\n9TtDkglI0moNIHUvvSimb2OnKT5AnX8T7xBIohck1Rs5SNZCbGBdFVJVi+T9IEmaHZKNpwBS\n4uMn6UOSM0DiJ8QqS3ft8kIxJFUuIG0vvTDUDIn2gCSmhOS/ZPVn9RlwnwQgyUpIplxA2l56\nYagUkp8oJLoKJN5XkWW3P6vPgPskAknuA0kd112V2CXbomGZDJJshxT8XGczJKE/oFqQ7Aip\nhsF0kCQgjSm9NLqv2WkqUTUQ0rovIB0MKbyiokXDclZIoREKSKHurD0B7hP7UlPVMgVJGkhu\nAQ2QzHF5FZsh3WaOZEGS00CiO0FixfqQ9HtNAyQp+dbrPslKbIRkB5AmgCSykIS3di9Iie7c\nWIBdlqp2HNKyCwHSsNJLswVSYG0YkmiEJDKfI3mQEo5KIG3P0ZBkD0j+fX9AymYIJH3OUpAC\nf+VIQdJP3D1HQOqc7W9p1qKudg7S+qpTgGmvtDcvaHYAUnMAaX5I5s0KkDpD6tcoQOoFiQQF\nIYkwJH4lJ9x5VhMkq4H3hOSeFECqKr00pZDCSc7tqQ6SKISkn90Rknc7W+4AacoAEo2CxI9i\nVUTqIWe/MDkkfWdgCyQJSNtKL81OkNwNSyERgyRcSGITpDlHTBTS+qrXuFpI+WbP2C1+zgqp\ncI7ECqiD5N3t1uvXpRgkwTZ0mmBBkleBJE0j5oZ0ozmSjEBikg6AZK1fl6LvSN59cF0DQKqG\n1DOA1AGSaIIUOEYEElmXdqGjXBCSNJDk2ohiSHLdHJCqSi9NHpJUgzWQOSBFSkxC2vtHc2mC\nkNiqJCSvIEDaVHppHEjsozt7PA6GpO40tEPy9wYktaC7wz/E4NwO0rpsjcK9IYUo6JfWpS6Q\n3BE4ddaKAlIgp4Aku0GSCySRgSQSkMxKB5J7X2L5xOmikPiqIkhmD0DaUnppSiGF0wjJmw91\ngsRrsDzcBFK0kCMhjcvMkOSekNyPjIT/LmW/qBbSkLwaLA+XhSSLmqEhSUBqKL00tLwRqWfm\nPE0MSZhdrAIDNVgepH2re22ZavvcISqBlCsEkLaUXppSSMIamvb4jcaCFLzVXQVJrAuNkGQY\n0vQhv6ZLqxoh6RW75V5zpIkghUraBknyu16A1LuaiQBSGBKXFHwHUbulIPGv1W2FZB2BVVwC\nEoO0Z24FSUp5IkgiAsk6MIfktBeQ9gwgbYZEOUhUDkmEIbGDOfWli0KSgLTkhpACnw+FIUXm\nXAMgbevl3RKkok5QKSQJSFtKL43buWYITgZp/eZCPSSvvZmPL6cKIMUzGyQ3HiTp3KpLQ+IK\nniN+AKS1MEA6A6RxuSAk/XnpbpAC8ylASpXz3PY8zS7KSSHpH2oxSCIByb8ErINEgGSv1Ceo\nfLIHSM2ltyUOKTFHykASRZASky0DiWxIfKPbQaq4a7Ju3q9mpcEcqRCS9jAWEvWDpP84QWKQ\nJCDND0nGIMkNkPhrVZAcSjFIJreAJAFp6DGOhyTcF0sghf4uURJSentACpYkAamp9LZoSNKF\nJJ+QzNoekPSfpZAEIC0rAemZa0GiJCSxQArJqIa03ELXkJJf8rsNpPKGAFJr6W0phWQ8bYFk\nf507cLMgAElf5AHSukFxScfk1pBkCBI1QwoNdh9SAIINiZogBcYhIF0kZ4AkTwAp6giQwiVd\nLaeDJMdDEmWQqAiSvD6k5SXzZ1FJV8tZITXPkYIRZgvvOg2QdM4P6b5zJNPnPiS99hBIZEFK\nXdldHpL3Z1FJxwSQxkKytiiBRJ0hnSUJSP5ypqRjcldI8kKQwuMQkHYNINVDEnoSMwKSPtyd\nIMlQfZtaAEitpTfG6vBySGJeSKkGniKAFMnVILkf8DBIGRXLd1cbICU/RgKkomL2CSCZbzmY\nz5HWtR0hUSOkRJYqyutCOrKYiXI2SDILSf91O2Kfro6ClE4G0tkCSJGcB5KZqZdDEsMh5Rxd\nDVKnXK8jzgRJLyTnSIB0gmCO1Fp6Y4Id7kGSZ4GErAGk1tIb0wmS2A8S3xyQZsptIQXDIclN\nkJw7foB07QCSk3JIGlQNJCqH5N+0A6SZA0hOyNQ2CUkAEuIEkHg6QmJf1KuD5BcHSBMHkHjs\nr7JKBolEGFJYBYPk/bbHPpAkIF07Z4WkUgDJ/MafHCStCZCQygASh0SAhNTnrJBScyQDiQAJ\nsYM5Ek8fSJIA6VYBJJ4oJDoUkoxCQmYIIPFYkNSjD4n2h+SVZtUROTyAxJOFpG55t0DytgGk\niwSQeCohqRXWdhKQ7hdA4ukASQLS7QJIsZRA0vMlQEJG5fSQJIe0YmKQ1u8JARIyJteEpDB1\ngJRzVgRpeDcgR+eskExtj4MU/JoEIM0czJF48pCoF6To64B0ugASTwUkikGSFZBCGwHS6QJI\nPJ0gSUC6VQCJpxgS9YDkf4WuCtKwXkAqA0g8hZAIkBArgMTTC5IcDMnUEJkggBRPHSRpD/EW\nSOzbeoCEvFID6fEdvhx61BvtAmlJbGTrbzkAEjIyFZAe+g+zzB/lw9nlBJBkLSSrhNDGgHTP\nnBUSry0b41NBQqbJFHOkEkiuo6kgiW2QpH4OSOfNeSDpKdL/XulbWSsTQpKANH3OA8naRk72\njsQZjILkrB7WAUhLTgNJ2o/XgiQB6fwBJJ46SBSGZO2bguR8VAVIZ85pIE14aed4SkEK7/3a\nwNoMkM6cU0Gy7tzt+znSeEgSkJBwWr7Z8LCXI4+89NHZC5LMQpKAdMec/7t2Syj0zesUJHff\nIkgyCMn+8pBTHhzdKWeFFJgjyTwkWQ3J3iwLyS0PkObLFHOkTaX3TjUkNa5rIElAulgAicer\nLSAh+QASzyGQJCCdPIDE0xOS+oQoA+m1pQSkMweQeOaAJGOQzDfzkJkCSDzjIclmSM5XXJGZ\nAkjZ7ANJSkBCQrkRJNkAKXykHCQJSLcLIMkGSOo2Qw7SwPYic+WskBrmSICEYI7E0xuSrIAk\nGSQN0S0SkOYLIPFMB0kC0gkCSDxjIek1oUNHIbEiAWm+ABLPgZBkIaTwzXPkyAASz+GQpAtJ\nAtIZAkg8fm1ngyQBab4AUj4HQJIZSGMaiswYQHrt60OSgIRU5K6Q2L7k/vMu66ZVkLwSAelO\nOSukrXMkti8g3SOYI/GEa1sIybsNUA1JAtIpA0g8e0CKUgCkswaQeAZCkoWQ1gNKv7j4nsix\nASQeQEIaAkg8h0KSgHTSABJPASTOagwkCUgnCiCVBZCQg3I/SDIFSZZDUjsCEiIBSe3XGxJy\ns5wVUvscafkKBN+PQSp9RyFzAw85QTBH4imFFPm6AiDdM4DEk4MkN0CStZCQswSQeCKQ5AGQ\nkPMEkHiaIQWN+JAKA0jnCiDxABLSEEDiaYIULU0Rq4aECdK5AkhlaYa0PgT/RkTmiAgiAUnt\ntzwAEtIYQFr2Uw+AhDTlrJA6z5H0AyBdOpgj8fSFpHevvg0HSKcKIPHMAgk5VQCJpxySBCRE\nBZB4YrVNQcoGkK4eQOKJ1tZc0gESwgNIPICENASQikP6S3OAhOyXe0CSgISMzaUhyUpIr/3H\nVg+5aM4KKTVHkqR/IUk9JOTKwRyJJwlJAhISDCDxABLSEEDiqYEkAQlZAkg8tZBwOw6RgOQH\nkJCGABJPFpJ+rn4BFyAhgFQeDkkCErJDAAlBOgSQEKRDzgopXtsApNefgIRgjuQlU1tfDSAh\ngOSnuraAhACSH0BCGgJIPICENASQeOohjagFcrIAEg8gIQ0BpM0BJGRkAAlBOgSQEKRDzgoJ\ncySkIZgj8YzrEeTCASQeQEIaAkg8gIQ0BJB4AAlpCCDxABLSEEDiASSkIYCEIFMHkBCkQwAJ\nQTrkrJAwR0IagjkSDyAhDQEkHkBCGgJIPICENASQeAAJaQgg8QAS0hBA4gEkpCGAhCBTB5AQ\npEMACUE65KyQMEdCGoI5Eg8gIQ0BJB5AQhoCSDyAhDQEkHgACWkIIPEAEtIQQOIBJKQhgIQg\nUweQEKRDAAlBOuSskDBHQhqCORIPICENASQeQEIaAkg8gIQ0BJB4AAlpCCDxABLSEEDiASSk\nIaeFhCA3yVBIBdR23m9IpqoMapPIHsMNkJozVWVQm0QAqdd+QzJVZVCbRK4LCUEuFkBCkA4B\nJATpEEBCkA4BJATpEEBCkA4BJATpEEBCkA7ZF9LraOI7y2NJBcSSdUGvGVrNdG2sKpQ1YWht\nRG2HDq2MmKtvdGpHTf1A3bWhSwevh9X/5XfTf5g9jjlBVt3N0nFv6lZXVHXo0ApN0jdOTWpH\nTcNA3bOlQtbXT0reE0dCcutw+GABpGxtJB8+Jfu0DNT9L+2c+hUd3n1nPXxWZ2pd3oRBFTEP\nVR06skKT9I2UIUhV+9T163GQXu+f5ZCkvsQ9boa0Ht+erh36hrRe/S+VOro2uhJT9I2qjvSG\nT/E+df16IKTloaB95s1I8IUDIlqaMKom+ofl8bWxu2WG2jBIsqAu+s5Ew0DdqZ3C3Plw61eJ\n/fgLPF6DsiaMC4d0YG0EXzq4bzik4jlSy0A99B1JXYAW7OXsIecZLIVNGBfR0KGjasIWj+4b\nG1LpqAlDKmgJIFXFOvjxg8W/tJsA0iy10bW4LCT9OZd6VrbXFB/I2p/umarN8PHwDLVxKnV0\nbaQLacsHskUtObCdCHKdABKCdAggIUiHABKCdAggIUiHABKCdAggIUiHANKO+f0hxOPHf7W7\n+Z9gqL+v9hErK/eRyZ8f9odiwbq+ibdfS0k//tTW+IYBpP3yvg7/z8r94pCEiEjKQPrvIeXf\nJKTfrxd/La8+qu3fL4C0W97F+/eA/Pf7If5uLWod+/8+xVvT/o9vIX8UaAPpw7z1vH0T/SMe\ny6u/HhvqepMA0l75owb9H/F9XSX//RDixz/5HKlfH+LxGtRfz3Vfr3XyQ3zIrzfx8W8d6Xoj\nKa2x/1rQRcmvd/H2R7/LWMU5e8vPJ4xf4vfy7OPDFPv2267ys5TXq4/aN9H7BZD2yg+hft6/\nxvzjeen0pPU9bVqv9/69Fh4vOt+zKfE9TRHixzKel9fUkHffkXRRy0YKkl2cOsSy22vpuz4f\n4u37zfFTK/n6+SYeP/+pp6/3rNern0KvRMIBpL3ycLr652uQLpOQ93/fbw+P59P35wXg53Pd\nj+cs5fP5x6Li83vNf9YbkTVHMkX9/C7g37vZxRSnDrEe+0n6Yyngy63l38+H+FwuPT+0229S\nP0d1y1UCSHvF+uq6eM5BXs8+Xtdd66tvz6Wv53vLa50Q62Xd8to/p6wlP55D3hS1FiCCxZnr\nwY/16Z/n9d0Pr6Kf64afb1rSlzGFhANIe8WFpCiY9d6S/sO/tfZ69uf5jiPjRYUKls6CW6vX\nc+sd6Xnr7jOwCxIKOmivfOg5UidIy7VbT0hsjhTZBQkFHbRX9F07aS7t1DP/WkwySPzS7vXw\n9pq6mKLil3bWTmphmbM598+tu3YP8c/fBYkHHbRbnremvx/+fCx3Aj6fl07v9ii37w5IBum5\n/V8+rr9X/LWLWgsI3GywdlJzpB/P2xO/nn/oWJ8jfT6N6hkU5kjZANJ+Ud9seH5RYL1T/dce\n5fb9askgfel73GrrZ34+15ii4re/7Z3Wu3ZuiTzrzus9Pdy1ywaQdsyfH9/D82O5enp+Wvr8\npoM9yu0PZBkk+fd9fU1v/czr4k4X9fpAVt8xZ8WZnZbPkZ4f8YrP6OdDeudn8DlSNoB0uYjs\nF3o+a8969dcD7xdAulBen89+Bj4a4nn8ym5iB9+1yweQLpTP4JcVAvmvTga+/Z0PIF0pv96s\niU0qf/LvWib4+0gFASQE6RBAQpAOASQE6RBAQpAOASQE6RBAQpAOASQE6RBAQpAO+T8LEgDv\nwY8Y0wAAAABJRU5ErkJggg==",
      "text/plain": [
       "plot without title"
      ]
     },
     "metadata": {
      "image/png": {
       "height": 420,
       "width": 420
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# run in local PC\n",
    "library(GenomicRanges)\n",
    "library(ChIPseeker)\n",
    "library(TxDb.Hsapiens.UCSC.hg38.knownGene)\n",
    "txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene\n",
    "load(\"E:/项目/转座子/sgRNA/results/TE_insertion.anno.RData\") # TE_insertion.list\n",
    "anno <- as.data.frame(rbindlist(TE_insertion.list))\n",
    "anno.bed <- anno[, c(\"chrom\", \"start\", \"end\", \"insertion_site\")]\n",
    "anno.bed <- anno.bed[order(anno.bed$chrom, anno.bed$start, anno.bed$end, anno.bed$insertion_site), ]\n",
    "anno.bed <- anno.bed[!duplicated(anno.bed$insertion_site),]\n",
    "write.table(anno.bed, \"stat/insertion.bed\", sep=\"\\t\", quote=F, col.names=F, row.names=F)\n",
    "insertion.GR <- GRanges(seqnames = anno.bed$chrom,\n",
    "                        ranges = IRanges(start = anno.bed$start-1,\n",
    "                                         end = anno.bed$end,\n",
    "                                         names = anno.bed$insertion_site))\n",
    "\n",
    "ChIPseeker::plotPeakProf2(\n",
    "    insertion.GR,\n",
    "    # upstream=1000,\n",
    "    # downstream=1000,\n",
    "    upstream = rel(0.1), \n",
    "    downstream = rel(0.1),\n",
    "    conf=0.95,\n",
    "    by=\"transcript\",\n",
    "    type=\"body\",\n",
    "    weightCol = NULL,\n",
    "    TxDb = txdb,\n",
    "    xlab = \"Genomic Region (5'->3')\",\n",
    "    ylab = \"Insertion Count Frequency\",\n",
    "    facet = \"none\",\n",
    "    free_y = TRUE,\n",
    "    verbose = TRUE,\n",
    "    nbin = 800,\n",
    "    ignore_strand = FALSE\n",
    "    ) + \n",
    "    theme_bw() +\n",
    "    theme(\n",
    "      panel.grid=element_blank(),\n",
    "      axis.text.x = element_text(angle = 0, color = \"black\", \n",
    "                                 size = 8, hjust = 0.5, vjust=0.5),\n",
    "      plot.title=element_text(size=14, face=\"bold\", hjust=0.5)\n",
    "    ) \n",
    "\n",
    "\n",
    "# ggsave(\"E:/项目/转座子/sgRNA/results/insertion_density_around_genebody.pdf\",\n",
    "#         width=7, height=4)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\n",
       "FALSE  TRUE \n",
       " 4544    18 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "table((anno.bed$end - anno.bed$start)>=800)\n",
    "table((anno.bed$end - anno.bed$start))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(ChIPseeker)\n",
    "library(GenomicRanges)\n",
    "library(TxDb.Hsapiens.UCSC.hg38.knownGene)\n",
    "txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene\n",
    "load(\"E:/项目/转座子/sgRNA/results/TE_insertion.anno.RData\") # TE_insertion.list\n",
    "\n",
    "anno <- as.data.frame(rbindlist(TE_insertion.list))\n",
    "anno.bed <- anno[, c(\"chrom\", \"start\", \"end\", \"insertion_site\")]\n",
    "anno.bed <- anno.bed[order(anno.bed$insertion_site), ]\n",
    "anno.bed <- anno.bed[!duplicated(anno.bed$insertion_site),]\n",
    "# length(unique(anno[!(anno$gene_id %in% \".\"), \"insertion_site\"]))/length(unique(anno[, \"insertion_site\"]))\n",
    "# anno <- anno[, c(\"insertion_site\", \"intron\", \"intergenic\", \"UTR5\", \"UTR3\", \"CDS\", \"TSS500\", \"TES500\")]\n",
    "# anno <- anno[order(anno$insertion_site), ]\n",
    "# anno <- anno[!duplicated(anno$insertion_site),]\n",
    "\n",
    "\n",
    "# convert bed to GRanges object\n",
    "\n",
    "\n",
    "\n",
    "insertion.GR <- GRanges(seqnames = anno.bed$chrom,\n",
    "                        ranges = IRanges(start = anno.bed$start,\n",
    "                                         end = anno.bed$end,\n",
    "                                         names = anno.bed$insertion_site))\n",
    "\n",
    "ChIPseeker::plotPeakProf2(\n",
    "    insertion.GR,\n",
    "    upstream=1000,\n",
    "    downstream=1000,\n",
    "    conf=0.95,\n",
    "    # by=\"transcript\",\n",
    "    # type=\"body\",\n",
    "    weightCol = NULL,\n",
    "    TxDb = txdb,\n",
    "    xlab = \"Genomic Region (5'->3')\",\n",
    "    ylab = \"Insertion Count Frequency\",\n",
    "    facet = \"none\",\n",
    "    free_y = FALSE,\n",
    "    verbose = TRUE,\n",
    "    # nbin = NULL,\n",
    "    # ignore_strand = TRUE\n",
    "    )\n",
    "\n",
    "\n",
    "ggsave(\"stat/insertion_density_around_genebody.pdf\",\n",
    "        width=5, height=4)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### insertion in genebody "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# /NAS/wg_looking/human_ref/gencode.v37.chr_patch_hapl_scaff.annotation.gtf\n",
    "\n",
    "nwin=100\n",
    "######### strand: +\n",
    "awk '$3==\"gene\" && $7==\"+\" {print $1\"\\t\"$4\"\\t\"$5}' \\\n",
    "/NAS/wg_looking/human_ref/gencode.v37.chr_patch_hapl_scaff.annotation.gtf \\\n",
    "> /NAS/wg_looking/gRNA_ONT/stat/gencode.v37.chr_patch_hapl_scaff.annotation.gene.positive.bed\n",
    "#\n",
    "bedtools makewindows -b gencode.v37.chr_patch_hapl_scaff.annotation.gene.positive.bed \\\n",
    "-n $nwin -i winnum > gencode.v37.chr_patch_hapl_scaff.annotation.positive.gene.bed\n",
    "#\n",
    "bedtools coverage -a gencode.v37.chr_patch_hapl_scaff.annotation.positive.gene.bed \\\n",
    "-b insertion.bed -counts > hg38.gene.positive.insertion.count.bed\n",
    "# upstream  \n",
    "bedtools flank -i gencode.v37.chr_patch_hapl_scaff.annotation.gene.positive.bed \\\n",
    "-g /NAS/wg_looking/human_ref/GRCh38.p13.genome.fa.fai -l 1000 -r 0 \\\n",
    "> gencode.v37.chr_patch_hapl_scaff.annotation.gene.positive.upstream.bed\n",
    "#\n",
    "bedtools makewindows -b gencode.v37.chr_patch_hapl_scaff.annotation.gene.positive.upstream.bed \\\n",
    "-n 10 -i winnum > gencode.v37.chr_patch_hapl_scaff.annotation.positive.upstream.bed\n",
    "#\n",
    "bedtools coverage -a gencode.v37.chr_patch_hapl_scaff.annotation.positive.upstream.bed \\\n",
    "-b insertion.bed -counts > hg38.gene.positive.upstream.insertion.count.bed\n",
    "# downstream\n",
    "bedtools flank -i gencode.v37.chr_patch_hapl_scaff.annotation.gene.positive.bed \\\n",
    "-g /NAS/wg_looking/human_ref/GRCh38.p13.genome.fa.fai -l 0 -r 1000 \\\n",
    "> gencode.v37.chr_patch_hapl_scaff.annotation.gene.positive.downstream.bed\n",
    "#\n",
    "bedtools makewindows -b gencode.v37.chr_patch_hapl_scaff.annotation.gene.positive.downstream.bed \\\n",
    "-n 10 -i winnum > gencode.v37.chr_patch_hapl_scaff.annotation.positive.downstream.bed\n",
    "#\n",
    "bedtools coverage -a gencode.v37.chr_patch_hapl_scaff.annotation.positive.downstream.bed \\\n",
    "-b insertion.bed -counts > hg38.gene.positive.downstream.insertion.count.bed\n",
    "\n",
    "\n",
    "######### strand: -\n",
    "awk '$3==\"gene\" && $7==\"-\" {print $1\"\\t\"$4\"\\t\"$5}' \\\n",
    "/NAS/wg_looking/human_ref/gencode.v37.chr_patch_hapl_scaff.annotation.gtf \\\n",
    "> /NAS/wg_looking/gRNA_ONT/stat/gencode.v37.chr_patch_hapl_scaff.annotation.gene.negative.bed\n",
    "#\n",
    "bedtools makewindows -b gencode.v37.chr_patch_hapl_scaff.annotation.gene.negative.bed \\\n",
    "-n $nwin -i winnum -reverse > gencode.v37.chr_patch_hapl_scaff.annotation.negative.gene.bed\n",
    "#\n",
    "bedtools coverage -a gencode.v37.chr_patch_hapl_scaff.annotation.negative.gene.bed \\\n",
    "-b insertion.bed -counts > hg38.gene.negative.insertion.count.bed\n",
    "# upstream  \n",
    "bedtools flank -i gencode.v37.chr_patch_hapl_scaff.annotation.gene.negative.bed \\\n",
    "-g /NAS/wg_looking/human_ref/GRCh38.p13.genome.fa.fai -l 0 -r 1000 \\\n",
    "> gencode.v37.chr_patch_hapl_scaff.annotation.gene.negative.upstream.bed\n",
    "#\n",
    "bedtools makewindows -b gencode.v37.chr_patch_hapl_scaff.annotation.gene.negative.upstream.bed \\\n",
    "-n 10 -i winnum -reverse > gencode.v37.chr_patch_hapl_scaff.annotation.negative.upstream.bed\n",
    "#\n",
    "bedtools coverage -a gencode.v37.chr_patch_hapl_scaff.annotation.negative.upstream.bed \\\n",
    "-b insertion.bed -counts > hg38.gene.negative.upstream.insertion.count.bed\n",
    "# downstream\n",
    "bedtools flank -i gencode.v37.chr_patch_hapl_scaff.annotation.gene.negative.bed \\\n",
    "-g /NAS/wg_looking/human_ref/GRCh38.p13.genome.fa.fai -l 1000 -r 0 \\\n",
    "> gencode.v37.chr_patch_hapl_scaff.annotation.gene.negative.downstream.bed\n",
    "#\n",
    "bedtools makewindows -b gencode.v37.chr_patch_hapl_scaff.annotation.gene.negative.downstream.bed \\\n",
    "-n 10 -i winnum -reverse > gencode.v37.chr_patch_hapl_scaff.annotation.negative.downstream.bed\n",
    "#\n",
    "bedtools coverage -a gencode.v37.chr_patch_hapl_scaff.annotation.negative.downstream.bed \\\n",
    "-b insertion.bed -counts > hg38.gene.negative.downstream.insertion.count.bed\n",
    "\n",
    "\n",
    "#### background using all TEs\n",
    "### strand: +\n",
    "bedtools coverage -a gencode.v37.chr_patch_hapl_scaff.annotation.positive.gene.bed \\\n",
    "-b hg38.LINE.SINE.bed -counts > hg38.gene.positive.TE.count.bed\n",
    "# upstream  \n",
    "bedtools coverage -a gencode.v37.chr_patch_hapl_scaff.annotation.positive.upstream.bed \\\n",
    "-b hg38.LINE.SINE.bed -counts > hg38.gene.positive.upstream.TE.count.bed\n",
    "# downstream\n",
    "bedtools coverage -a gencode.v37.chr_patch_hapl_scaff.annotation.positive.downstream.bed \\\n",
    "-b hg38.LINE.SINE.bed -counts > hg38.gene.positive.downstream.TE.count.bed\n",
    "\n",
    "\n",
    "### strand: -\n",
    "bedtools coverage -a gencode.v37.chr_patch_hapl_scaff.annotation.negative.gene.bed \\\n",
    "-b hg38.LINE.SINE.bed -counts > hg38.gene.negative.TE.count.bed\n",
    "# upstream  \n",
    "bedtools coverage -a gencode.v37.chr_patch_hapl_scaff.annotation.negative.upstream.bed \\\n",
    "-b hg38.LINE.SINE.bed -counts > hg38.gene.negative.upstream.TE.count.bed\n",
    "# downstream\n",
    "bedtools coverage -a gencode.v37.chr_patch_hapl_scaff.annotation.negative.downstream.bed \\\n",
    "-b hg38.LINE.SINE.bed -counts > hg38.gene.negative.downstream.TE.count.bed\n",
    "\n",
    "\n",
    "\n",
    "#### background using expressed TEs\n",
    "# load(\"stat/TE_location.RData\")\n",
    "# load(\"stat/MDAMB231_vs_MCF10A_TE_res.RData\") # TE.norlCount\n",
    "# expressedTEs <- rownames(res)\n",
    "# load(\"stat/HCT116_vs_NCM460_TE_res.RData\")\n",
    "# expressedTEs <- c(expressedTEs, rownames(res))\n",
    "# expressedTEs <- unique(expressedTEs)\n",
    "# expT <- TE_location[expressedTEs, c(\"Chr\", \"Start\", \"End\")]\n",
    "# write.table(expT, \"stat/BC_CRC_expressed_TEs.txt\", row.names=F, col.names=F, sep=\"\\t\", quote=F)\n",
    "\n",
    "### strand: +\n",
    "bedtools coverage -a gencode.v37.chr_patch_hapl_scaff.annotation.positive.gene.bed \\\n",
    "-b BC_CRC_expressed_TEs.txt -counts > hg38.gene.positive.expressedTE.count.bed\n",
    "# upstream  \n",
    "bedtools coverage -a gencode.v37.chr_patch_hapl_scaff.annotation.positive.upstream.bed \\\n",
    "-b BC_CRC_expressed_TEs.txt -counts > hg38.gene.positive.upstream.expressedTE.count.bed\n",
    "# downstream\n",
    "bedtools coverage -a gencode.v37.chr_patch_hapl_scaff.annotation.positive.downstream.bed \\\n",
    "-b BC_CRC_expressed_TEs.txt -counts > hg38.gene.positive.downstream.expressedTE.count.bed\n",
    "\n",
    "\n",
    "### strand: -\n",
    "bedtools coverage -a gencode.v37.chr_patch_hapl_scaff.annotation.negative.gene.bed \\\n",
    "-b BC_CRC_expressed_TEs.txt -counts > hg38.gene.negative.expressedTE.count.bed\n",
    "# upstream  \n",
    "bedtools coverage -a gencode.v37.chr_patch_hapl_scaff.annotation.negative.upstream.bed \\\n",
    "-b BC_CRC_expressed_TEs.txt -counts > hg38.gene.negative.upstream.expressedTE.count.bed\n",
    "# downstream\n",
    "bedtools coverage -a gencode.v37.chr_patch_hapl_scaff.annotation.negative.downstream.bed \\\n",
    "-b BC_CRC_expressed_TEs.txt -counts > hg38.gene.negative.downstream.expressedTE.count.bed\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### insertion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "body.Files <- list(\"stat/hg38.gene.positive.insertion.count.bed\",  \n",
    "                  \"stat/hg38.gene.negative.insertion.count.bed\"\n",
    "                  )\n",
    "upstream.Files <- list(\"stat/hg38.gene.positive.upstream.insertion.count.bed\", \n",
    "                      \"stat/hg38.gene.negative.upstream.insertion.count.bed\")\n",
    "downstream.Files <- list(\"stat/hg38.gene.positive.downstream.insertion.count.bed\",\n",
    "                        \"stat/hg38.gene.negative.downstream.insertion.count.bed\")\n",
    "\n",
    "\n",
    "nwin = 100\n",
    "body <- lapply(body.Files, function(file){\n",
    "    tb <- read.table(file)\n",
    "    w.count <- sapply(1:nwin, function(n){sum(tb[tb$V4 %in% n, \"V5\"])})\n",
    "    names(w.count) <- 1:nwin\n",
    "    return(w.count)\n",
    "})\n",
    "body <-  body[[1]]+body[[2]]\n",
    "\n",
    "\n",
    "upstream <- lapply(upstream.Files, function(file){\n",
    "    tb <- read.table(file)\n",
    "    w.count <- sapply(1:10, function(n){sum(tb[tb$V4 %in% n, \"V5\"])})\n",
    "    names(w.count) <- 1:10\n",
    "    return(w.count)\n",
    "})\n",
    "upstream <-  upstream[[1]]+upstream[[2]]\n",
    "\n",
    "\n",
    "\n",
    "downstream <- lapply(downstream.Files, function(file){\n",
    "    tb <- read.table(file)\n",
    "    w.count <- sapply(1:10, function(n){sum(tb[tb$V4 %in% n, \"V5\"])})\n",
    "    names(w.count) <- 1:10\n",
    "    return(w.count)\n",
    "})\n",
    "downstream <-  downstream[[1]]+downstream[[2]]\n",
    "\n",
    "\n",
    "pp <- data.frame(count=c(upstream, body, downstream))\n",
    "pp$wnum <- 1:(nwin+20)\n",
    "\n",
    "\n",
    "ggplot(pp, aes(x=wnum, y=count)) +\n",
    " geom_smooth(size=1.5, color=\"#8DA0CB\",se = FALSE,span = 0.1) +\n",
    " theme_bw() +\n",
    "    theme(\n",
    "      panel.grid=element_blank(),\n",
    "      axis.text.x = element_text(angle = 0, color = \"black\", \n",
    "                                 size = 8, hjust = 0.5, vjust=0.5),\n",
    "      plot.title=element_text(size=14, face=\"bold\", hjust=0.5)\n",
    "    ) +\n",
    "    geom_vline(aes(xintercept = 11), linetype=\"dashed\", color=\"red\") +\n",
    "   geom_vline(aes(xintercept = 10+nwin), linetype=\"dashed\", color=\"red\") +\n",
    "    ylab(\"insertion count\") +\n",
    "    xlab(\"Genomic Region (5'->3')\") +\n",
    "   scale_x_continuous(breaks=seq(0,120,10), labels=c(\"Upstream\", seq(0,100,10), \"Downstream\")) +\n",
    "   ggtitle(\"insertion\")\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "ggsave(\"stat/insertion_genebody_count.pdf\", width=5.5, height=3.5)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### TE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "body.Files <- list(\"stat/hg38.gene.positive.TE.count.bed\",  \n",
    "                  \"stat/hg38.gene.negative.TE.count.bed\"\n",
    "                  )\n",
    "upstream.Files <- list(\"stat/hg38.gene.positive.upstream.TE.count.bed\", \n",
    "                      \"stat/hg38.gene.negative.upstream.TE.count.bed\")\n",
    "downstream.Files <- list(\"stat/hg38.gene.positive.downstream.TE.count.bed\",\n",
    "                        \"stat/hg38.gene.negative.downstream.TE.count.bed\")\n",
    "\n",
    "\n",
    "nwin = 100\n",
    "body <- lapply(body.Files, function(file){\n",
    "    tb <- read.table(file)\n",
    "    w.count <- sapply(1:nwin, function(n){sum(tb[tb$V4 %in% n, \"V5\"])})\n",
    "    names(w.count) <- 1:nwin\n",
    "    return(w.count)\n",
    "})\n",
    "body <-  body[[1]]+body[[2]]\n",
    "\n",
    "\n",
    "upstream <- lapply(upstream.Files, function(file){\n",
    "    tb <- read.table(file)\n",
    "    w.count <- sapply(1:10, function(n){sum(tb[tb$V4 %in% n, \"V5\"])})\n",
    "    names(w.count) <- 1:10\n",
    "    return(w.count)\n",
    "})\n",
    "upstream <-  upstream[[1]]+upstream[[2]]\n",
    "\n",
    "\n",
    "\n",
    "downstream <- lapply(downstream.Files, function(file){\n",
    "    tb <- read.table(file)\n",
    "    w.count <- sapply(1:10, function(n){sum(tb[tb$V4 %in% n, \"V5\"])})\n",
    "    names(w.count) <- 1:10\n",
    "    return(w.count)\n",
    "})\n",
    "downstream <-  downstream[[1]]+downstream[[2]]\n",
    "\n",
    "\n",
    "pp <- data.frame(count=c(upstream, body, downstream))\n",
    "pp$wnum <- 1:(nwin+20)\n",
    "\n",
    "\n",
    "ggplot(pp, aes(x=wnum, y=count)) +\n",
    " geom_smooth(size=1.5, color=\"#8DA0CB\",se = FALSE,span = 0.1) +\n",
    " theme_bw() +\n",
    "    theme(\n",
    "      panel.grid=element_blank(),\n",
    "      axis.text.x = element_text(angle = 0, color = \"black\", \n",
    "                                 size = 8, hjust = 0.5, vjust=0.5),\n",
    "      plot.title=element_text(size=14, face=\"bold\", hjust=0.5)\n",
    "    ) +\n",
    "    geom_vline(aes(xintercept = 11), linetype=\"dashed\", color=\"red\") +\n",
    "   geom_vline(aes(xintercept = 10+nwin), linetype=\"dashed\", color=\"red\") +\n",
    "    ylab(\"TE count\") +\n",
    "    xlab(\"Genomic Region (5'->3')\") +\n",
    "   scale_x_continuous(breaks=seq(0,120,10), labels=c(\"Upstream\", seq(0,100,10), \"Downstream\")) +\n",
    "   ggtitle(\"TE\")\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "ggsave(\"stat/TE_genebody_count.pdf\", width=5.5, height=3.5)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### expressed TEs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "body.Files <- list(\"stat/hg38.gene.positive.expressedTE.count.bed\",  \n",
    "                  \"stat/hg38.gene.negative.expressedTE.count.bed\"\n",
    "                  )\n",
    "upstream.Files <- list(\"stat/hg38.gene.positive.upstream.expressedTE.count.bed\", \n",
    "                      \"stat/hg38.gene.negative.upstream.expressedTE.count.bed\")\n",
    "downstream.Files <- list(\"stat/hg38.gene.positive.downstream.expressedTE.count.bed\",\n",
    "                        \"stat/hg38.gene.negative.downstream.expressedTE.count.bed\")\n",
    "\n",
    "\n",
    "nwin = 100\n",
    "body <- lapply(body.Files, function(file){\n",
    "    tb <- read.table(file)\n",
    "    w.count <- sapply(1:nwin, function(n){sum(tb[tb$V4 %in% n, \"V5\"])})\n",
    "    names(w.count) <- 1:nwin\n",
    "    return(w.count)\n",
    "})\n",
    "body <-  body[[1]]+body[[2]]\n",
    "\n",
    "\n",
    "upstream <- lapply(upstream.Files, function(file){\n",
    "    tb <- read.table(file)\n",
    "    w.count <- sapply(1:10, function(n){sum(tb[tb$V4 %in% n, \"V5\"])})\n",
    "    names(w.count) <- 1:10\n",
    "    return(w.count)\n",
    "})\n",
    "upstream <-  upstream[[1]]+upstream[[2]]\n",
    "\n",
    "\n",
    "\n",
    "downstream <- lapply(downstream.Files, function(file){\n",
    "    tb <- read.table(file)\n",
    "    w.count <- sapply(1:10, function(n){sum(tb[tb$V4 %in% n, \"V5\"])})\n",
    "    names(w.count) <- 1:10\n",
    "    return(w.count)\n",
    "})\n",
    "downstream <-  downstream[[1]]+downstream[[2]]\n",
    "\n",
    "\n",
    "pp <- data.frame(count=c(upstream, body, downstream))\n",
    "pp$wnum <- 1:(nwin+20)\n",
    "\n",
    "\n",
    "ggplot(pp, aes(x=wnum, y=count)) +\n",
    " geom_smooth(size=1.5, color=\"#8DA0CB\",se = FALSE,span = 0.1) +\n",
    " theme_bw() +\n",
    "    theme(\n",
    "      panel.grid=element_blank(),\n",
    "      axis.text.x = element_text(angle = 0, color = \"black\", \n",
    "                                 size = 8, hjust = 0.5, vjust=0.5),\n",
    "      plot.title=element_text(size=14, face=\"bold\", hjust=0.5)\n",
    "    ) +\n",
    "    geom_vline(aes(xintercept = 11), linetype=\"dashed\", color=\"red\") +\n",
    "   geom_vline(aes(xintercept = 10+nwin), linetype=\"dashed\", color=\"red\") +\n",
    "    ylab(\"TE count\") +\n",
    "    xlab(\"Genomic Region (5'->3')\") +\n",
    "   scale_x_continuous(breaks=seq(0,120,10), labels=c(\"Upstream\", seq(0,100,10), \"Downstream\")) +\n",
    "   ggtitle(\"expressed TE\")\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "ggsave(\"stat/expressed_TE_genebody_count.pdf\", width=5.5, height=3.5)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### binary table of insertion "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### insertion region as row"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## binary table of insertion \n",
    "load(\"stat/TE_insertion.anno.RData\")\n",
    "tumor_genes <- read.table(\"database/NG_tumor_related_genes.txt\")\n",
    "\n",
    "TE <- \"ALU\"\n",
    "anno <- TE_insertion.list[[TE]]\n",
    "anno <- anno[order(anno$insertion_region), ]\n",
    "ins.region <- anno[!duplicated(anno$insertion_region), ]\n",
    "IDs <- c(\"sgRNA_HCT116\", \"sgRNA_MDAMB231\", \"sgRNA_NCM460\", \"sgRNA_MCF10A\")\n",
    "binary.tb <- lapply(setNames(IDs, IDs), function(ID){as.integer(grepl(ID, ins.region$IDs))})\n",
    "binary.tb <- data.frame(binary.tb)\n",
    "rownames(binary.tb) <- ins.region$insertion_region\n",
    "dim(binary.tb) # 1652\n",
    "\n",
    "\n",
    "rowInfo <- data.frame(insertion=rownames(binary.tb))\n",
    "rownames(rowInfo) <- rowInfo$insertion\n",
    "# tumor_genes$V1[tumor_genes$V1 %in% anno[anno$insertion_region %in% rowInfo$insertion, \"gene_name\" ]]\n",
    "rowInfo$locate_known_genes <- 0\n",
    "rowInfo$locate_known_genes[rowInfo$insertion %in% anno[!(anno$gene_id %in% \".\"), \"insertion_region\"]] <- 1\n",
    "rowInfo$locate_tumor_genes <- 0\n",
    "rowInfo$locate_tumor_genes[rowInfo$insertion %in% anno[anno$gene_name %in% tumor_genes$V1, \"insertion_region\"]] <- 1\n",
    "rowInfo <- rowInfo[,-1]\n",
    "\n",
    "pheatmap(binary.tb,\n",
    "         # display_numbers=T,\n",
    "         # annotation_col = colInfo,\n",
    "         annotation_row = rowInfo,\n",
    "         # annotation_colors = anno_colors,\n",
    "         main=\"insertion\",\n",
    "         col=c(\"#D9DADE\", \"#E05D4B\"),\n",
    "         # color=colorRampPalette(c(\"navy\", \"white\", \"firebrick3\"))(100),\n",
    "         # col=colorRampPalette(c(\"#4678BA\", \"#F7F7F7\", \"#EF8354\"))(11),\n",
    "         border=\"white\",\n",
    "         show_rownames=F,\n",
    "         # show_colnames=F,\n",
    "         # cluster_rows=F,\n",
    "         # cluster_cols=F,\n",
    "         width=5,\n",
    "         height=5,\n",
    "         filename=\"stat/CRC_BC_insertion_heatmap.pdf\")\n",
    "\n",
    "\n",
    "\n",
    "# length(unique(anno[!(anno$gene_id %in% \".\"), \"insertion_region\"]))\n",
    "# length(unique(anno[, \"insertion_region\"]))\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### gene_id as row"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "load(\"stat/hg38.gene.transcript.gtf.RData\")\n",
    "load(\"stat/TE_insertion.anno.RData\")\n",
    "tumor_genes <- read.table(\"database/NG_tumor_related_genes.txt\")\n",
    "\n",
    "\n",
    "anno <- rbindlist(TE_insertion.list)\n",
    "tb <- table(anno$gene_id, anno$ID)\n",
    "dim(tb) # 1420    4\n",
    "tb <- as.data.frame.matrix(tb)\n",
    "tb <- tb[!(rownames(tb) %in% \".\"), ]\n",
    "tb.all <- tb\n",
    "dim(tb.all) # 1419    4 \n",
    "# tb <- tb[tb$sgRNA_MDAMB231>0 & tb$sgRNA_MCF10A==0 |\n",
    "#         tb$sgRNA_HCT116>0 & tb$sgRNA_NCM460==0, ]\n",
    "\n",
    "\n",
    "# row order\n",
    "tb <- tb[order(rowSums(tb)),]\n",
    "cluster_rows <- hclust(dist(tb, method=\"binary\"), method = \"complete\")\n",
    "labels_row <- gene_gtf[rownames(tb), \"gene_name\"]\n",
    "# labels_row_check <- labels_row[cluster_rows$order]\n",
    "\n",
    "\n",
    "rowInfo <- data.frame(gene_id=rownames(tb))\n",
    "rownames(rowInfo) <- rowInfo$gene_id\n",
    "rowInfo$gene_name <- gene_gtf[rownames(rowInfo) , \"gene_name\"]\n",
    "rowInfo$locate_tumor_genes <- 0\n",
    "rowInfo$locate_tumor_genes[rowInfo$gene_name %in% tumor_genes$V1] <- 1\n",
    "rowInfo$ALU <- 0\n",
    "rowInfo$ALU[rowInfo$gene_id %in% anno$gene_id[anno$TE %in% \"ALU\"]] <- 1\n",
    "rowInfo$LINE <- 0\n",
    "rowInfo$LINE[rowInfo$gene_id %in% anno$gene_id[anno$TE %in% \"LINE\"]] <- 1\n",
    "rowInfo <- rowInfo[,-c(1, 2), drop=F]\n",
    "\n",
    "\n",
    "\n",
    "pheatmap(tb,\n",
    "         # display_numbers=T,\n",
    "         # annotation_col = colInfo,\n",
    "         annotation_row = rowInfo,\n",
    "         # clustering_distance_rows = \"binary\",\n",
    "         cluster_rows = cluster_rows,\n",
    "         labels_row=labels_row, \n",
    "         # annotation_colors = anno_colors,\n",
    "         main=\"insertion genes\",\n",
    "         # col=c(\"#D9DADE\", \"#E05D4B\"),\n",
    "         col=c(\"#F1F3F5\", colorRampPalette(c(\"#F6A496\",\"#CF2920\"))(25)),\n",
    "         # color=colorRampPalette(c(\"navy\", \"white\", \"firebrick3\"))(8)[2:7],\n",
    "         # col=colorRampPalette(c(\"#4678BA\", \"#F7F7F7\", \"#EF8354\"))(11),\n",
    "         # border=\"white\",\n",
    "         # show_rownames=F,\n",
    "         # show_colnames=F,\n",
    "         # cluster_rows=F,\n",
    "         # cluster_cols=F,\n",
    "         fontsize_row=0.5,\n",
    "         width=5,\n",
    "         height=5,\n",
    "         filename=\"stat/CRC_BC_insertion_heatmap_genes_as_row_all.pdf\")\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tb.tmp <- as.data.frame.matrix(tb[cluster_rows$order, ])\n",
    "tb.tmp <- data.frame(gene_id=rownames(tb.tmp), \n",
    "                     gene_name=labels_row[cluster_rows$order],\n",
    "                     tb.tmp)\n",
    "\n",
    "write.table(tb.tmp, \"stat/CRC_BC_insertion_genes.txt\", sep=\"\\t\", quote=F, col.names=T, row.names=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "################################################\n",
    "## MDAMB231 specific or HCT116 specifc insertion\n",
    "################################################\n",
    "\n",
    "tb <- tb[tb$sgRNA_MDAMB231>0 & tb$sgRNA_MCF10A==0 |\n",
    "        tb$sgRNA_HCT116>0 & tb$sgRNA_NCM460==0, ]\n",
    "\n",
    "\n",
    "# row order\n",
    "tb <- tb[order(rowSums(tb)),]\n",
    "dim(tb) # 672   4\n",
    "cluster_rows <- hclust(dist(tb, method=\"binary\"), method = \"complete\")\n",
    "labels_row <- gene_gtf[rownames(tb), \"gene_name\"]\n",
    "# labels_row_check <- labels_row[cluster_rows$order]\n",
    "\n",
    "\n",
    "rowInfo <- data.frame(gene_id=rownames(tb))\n",
    "rownames(rowInfo) <- rowInfo$gene_id\n",
    "rowInfo$gene_name <- gene_gtf[rownames(rowInfo) , \"gene_name\"]\n",
    "rowInfo$locate_tumor_genes <- 0\n",
    "rowInfo$locate_tumor_genes[rowInfo$gene_name %in% tumor_genes$V1] <- 1\n",
    "rowInfo$ALU <- 0\n",
    "rowInfo$ALU[rowInfo$gene_id %in% anno$gene_id[anno$TE %in% \"ALU\"]] <- 1\n",
    "rowInfo$LINE <- 0\n",
    "rowInfo$LINE[rowInfo$gene_id %in% anno$gene_id[anno$TE %in% \"LINE\"]] <- 1\n",
    "rowInfo <- rowInfo[,-c(1, 2), drop=F]\n",
    "\n",
    "\n",
    "\n",
    "pheatmap(tb,\n",
    "         # display_numbers=T,\n",
    "         # annotation_col = colInfo,\n",
    "         annotation_row = rowInfo,\n",
    "         # clustering_distance_rows = \"binary\",\n",
    "         cluster_rows = cluster_rows,\n",
    "         labels_row=labels_row, \n",
    "         # annotation_colors = anno_colors,\n",
    "         main=\"insertion genes\",\n",
    "         # col=c(\"#D9DADE\", \"#E05D4B\"),\n",
    "         col=c(\"#F1F3F5\", colorRampPalette(c(\"#F6A496\",\"#CF2920\"))(25)),\n",
    "         # color=colorRampPalette(c(\"navy\", \"white\", \"firebrick3\"))(8)[2:7],\n",
    "         # col=colorRampPalette(c(\"#4678BA\", \"#F7F7F7\", \"#EF8354\"))(11),\n",
    "         # border=\"white\",\n",
    "         # show_rownames=F,\n",
    "         # show_colnames=F,\n",
    "         # cluster_rows=F,\n",
    "         # cluster_cols=F,\n",
    "         fontsize_row=0.5,\n",
    "         width=5,\n",
    "         height=5,\n",
    "         filename=\"stat/CRC_BC_insertion_heatmap_genes_as_row.pdf\")\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#\n",
    "gene.Files <- list(BC=\"stat/MDAMB231_vs_MCF10A_all_genes.txt\",\n",
    "                   CRC=\"stat/HCT116_vs_NCM460_all_genes.txt\")\n",
    "tumor_genes <- read.table(\"database/NG_tumor_related_genes.txt\")\n",
    "tb.all\n",
    "\n",
    "\n",
    "# BC\n",
    "gtb <- read.table(gene.Files[[\"BC\"]], header=T, stringsAsFactors = F)\n",
    "g <- gtb[gtb$dif %in% c(\"Up\", \"Down\"), c(\"gene_id\", \"log2FoldChange\")]\n",
    "g2 <- data.frame(g, tb.all[g$gene_id,])\n",
    "g2[is.na(g2)] <- 0  \n",
    "g2$insertion_dif <- g2$sgRNA_MDAMB231 - g2$sgRNA_MCF10A    \n",
    "p1 <- g2[, c(\"insertion_dif\", \"log2FoldChange\")]\n",
    "p1 <- data.frame(p1, Tissue=\"BC\")\n",
    "\n",
    "# CRC\n",
    "gtb <- read.table(gene.Files[[\"CRC\"]], header=T, stringsAsFactors = F)\n",
    "g <- gtb[gtb$dif %in% c(\"Up\", \"Down\"), c(\"gene_id\", \"log2FoldChange\")]\n",
    "g2 <- data.frame(g, tb.all[g$gene_id,])\n",
    "g2[is.na(g2)] <- 0  \n",
    "g2$insertion_dif <- g2$sgRNA_HCT116 - g2$sgRNA_NCM460    \n",
    "p2 <- g2[, c(\"insertion_dif\", \"log2FoldChange\")]\n",
    "p2 <- data.frame(p2, Tissue=\"CRC\")\n",
    "\n",
    "p <- rbind(p1, p2)\n",
    "# cor(p$insertion_dif, p$log2FoldChange, method=\"spearman\") # 0.1527926\n",
    "# pearson 0.1348492\n",
    "\n",
    "ggplot(p, aes(x=log2FoldChange, y=insertion_dif, color=Tissue)) +\n",
    "      geom_point(alpha=0.5) +\n",
    "      theme_bw() +\n",
    "      theme(panel.grid = element_blank(),\n",
    "            plot.title = element_text(hjust = 0, size = 12),\n",
    "            axis.text.x  = element_text(angle = 0, hjust=1, vjust=0.5, \n",
    "                                        colour = \"black\", size = 11),\n",
    "            axis.text.y  = element_text(colour = \"black\"),\n",
    "            # legend.position = c(0.8,0.8),\n",
    "            strip.background = element_rect(fill=\"white\", colour = \"black\", size=1) ,\n",
    "            legend.title = element_text(face=\"bold\")) +\n",
    "      # geom_text(aes(label=symbol), vjust=0.5, hjust=0,size=4, colour=\"black\") +\n",
    "      # annotate(\"text\", x=pp$NAME[2], y=-1.5,size=3,\n",
    "      #          label=\"nominal P values: \\n*P < 0.05 \\n**P < 0.005 \\n***P < 0.0005 \\nNS: not significant\") +\n",
    "      # scale_x_discrete(limits=c(pp$term.name))+\n",
    "      scale_color_manual(values=c(\"#4678BA\", \"#EF8354\")) +\n",
    "      ggtitle(\"\")\n",
    "    # facet_grid(.~module, scales = \"free_x\", space = \"free_x\") \n",
    "\n",
    "\n",
    "ggsave(paste0(\"stat/Fig7S.Insertion_dif_along_gene_log2FoldChange_point.pdf\"),\n",
    "       height = 5, width = 5.5)\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### chisq test: insertion in oncogenes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tumor_genes <- read.table(\"database/NG_tumor_related_genes.txt\")\n",
    "ID.list <- list(BC=c(\"sgRNA_MDAMB231\", \"sgRNA_MCF10A\"),\n",
    "                CRC=c(\"sgRNA_HCT116\", \"sgRNA_NCM460\"))\n",
    "gene.Files <- list(BC=\"stat/MDAMB231_vs_MCF10A_all_genes.txt\",\n",
    "                   CRC=\"stat/HCT116_vs_NCM460_all_genes.txt\")\n",
    "\n",
    "\n",
    "test_res <- lapply(names(ID.list), function(T){\n",
    "    file <- gene.Files[[T]]\n",
    "    gene.tb <- read.table(file, header = TRUE)\n",
    "    bg <- unique(gene.tb$gene_name)\n",
    "    bg_tt <- table(bg %in% tumor_genes$V1)\n",
    "    IDs <- ID.list[[T]]\n",
    "    a <- lapply(IDs, function(ID){\n",
    "        tmp <- rownames(tb)[tb[, ID]>0 & tb[, setdiff(IDs, ID)]==0 ]\n",
    "        tt <- table(rowInfo[tmp, \"locate_tumor_genes\"])\n",
    "        x <- c(tt[\"1\"], tt[\"0\"], bg_tt[\"TRUE\"], bg_tt[\"FALSE\"])\n",
    "        x <- matrix(x, ncol = 2)\n",
    "        pvalue <- chisq.test(x)$p.value # 0.01910379\n",
    "        ratio= (tt[\"1\"]/sum(tt)) / (bg_tt[\"TRUE\"]/sum(bg_tt))\n",
    "        log2FC <-  log2(ratio) # 0.822162\n",
    "\n",
    "        res <- data.frame(ID=ID,\n",
    "                         pvalue=pvalue,\n",
    "                         log2FC=log2FC)\n",
    "    })\n",
    "    \n",
    "})\n",
    "\n",
    "pp <- rbindlist(lapply(test_res, rbindlist))\n",
    "pp$padj <- p.adjust(pp$pvalue, method = \"fdr\", n = length(pp$pvalue))\n",
    "pp <- pp[order(pp$padj, decreasing=TRUE),]\n",
    "pp$ID <- factor(pp$ID, levels=pp$ID)\n",
    "\n",
    "ggplot(pp, aes(x=ID, y=log2FC, fill=pvalue)) +\n",
    "      geom_bar(stat=\"identity\", width = 0.8) +\n",
    "      theme_bw() +\n",
    "      theme(panel.grid = element_blank(),\n",
    "            plot.title = element_text(hjust = 0, size = 12),\n",
    "            axis.text.x  = element_text(angle = 0, hjust=1, vjust=0.5, \n",
    "                                        colour = \"black\", size = 11),\n",
    "            axis.text.y  = element_text(colour = \"black\"),\n",
    "            # legend.position = c(0.8,0.8),\n",
    "            strip.background = element_rect(fill=\"white\", colour = \"black\", size=1) ,\n",
    "            legend.title = element_text(face=\"bold\")) +\n",
    "      # geom_text(aes(label=symbol), vjust=0.5, hjust=0,size=4, colour=\"black\") +\n",
    "      # annotate(\"text\", x=pp$NAME[2], y=-1.5,size=3,\n",
    "      #          label=\"nominal P values: \\n*P < 0.05 \\n**P < 0.005 \\n***P < 0.0005 \\nNS: not significant\") +\n",
    "      # scale_x_discrete(limits=c(pp$term.name))+\n",
    "      # scale_fill_manual(values=c(\"GO:BP\"=\"#4678BA\", \"KEGG\"=\"#EF8354\")) +\n",
    "      coord_flip() +\n",
    "      xlab(\"\") +\n",
    "      ggtitle(\"insertion related oncogenes\")\n",
    "    # facet_grid(.~module, scales = \"free_x\", space = \"free_x\") \n",
    "\n",
    "\n",
    "ggsave(paste0(\"stat/Insertion_oncogenes_chisq_test_bar.pdf\"),\n",
    "       height = 3, width = 6)\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "load(\"stat/hg38.quantify.FeatureCount.RData\") # trans.count, gene.count\n",
    "BC <- read.table(\"stat/MDAMB231_vs_MCF10A_all_genes.txt\", header=T)\n",
    "CRC <- read.table(\"stat/HCT116_vs_NCM460_all_genes.txt\", header=T)\n",
    "\n",
    "\n",
    "# load Sample info\n",
    "Sample_info <- read.table(\"stat/Sample_info2.txt\", header = T, stringsAsFactors = F)\n",
    "rownames(Sample_info) <- Sample_info$Samples\n",
    "\n",
    "SampleIDs <- Sample_info$Samples[Sample_info$CellLine %in% c(\"MDAMB231\", \"MCF10A\", \"HCT116\", \"NCM460\") &\n",
    "                                Sample_info$SeqType %in% c(\"sgRNA\")]\n",
    "\n",
    "\n",
    "# run DESeq2\n",
    "sub.gene.count <- gene.count[, SampleIDs]\n",
    "# sub.gene.count <- sub.gene.count[apply(sub.gene.count, 1, function(x){ (sum(x>=1))>=1 }),]\n",
    "sub.gene.count <- round(sub.gene.count)\n",
    "\n",
    "\n",
    "colInfo <- Sample_info[SampleIDs, c(\"CellLine\", \"TissueType\")]\n",
    "dds <- DESeqDataSetFromMatrix(sub.gene.count, \n",
    "                                colInfo, \n",
    "                                ~CellLine)\n",
    "Object <- DESeq(dds\n",
    "                # parallel = T, BPPARAM=MulticoreParam(10)\n",
    "                )\n",
    "gene.norlCount <- counts(Object, normalized=T)\n",
    "CRC_res <- results(Object, contrast=c(\"CellLine\", \"HCT116\", \"NCM460\"))\n",
    "BC_res <- results(Object, contrast=c(\"CellLine\", \"MDAMB231\", \"MCF10A\"))\n",
    "\n",
    "\n",
    "\n",
    "sub <- gene.norlCount[rownames(gene.norlCount) %in% rownames(tb), ]\n",
    "sub <- sub[rownames(tb)[cluster_rows$order], ]\n",
    "\n",
    "rowInfo <- data.frame(gene_id=rownames(sub))\n",
    "rownames(rowInfo) <- rowInfo$gene_id\n",
    "rowInfo$BC_dif <- 0\n",
    "rowInfo$BC_dif[rowInfo$gene_id %in% BC$gene_id[BC$dif %in% c(\"Up\", \"Down\")]] <- 1\n",
    "rowInfo$CRC_dif <- 0\n",
    "rowInfo$CRC_dif[rowInfo$gene_id %in% CRC$gene_id[CRC$dif %in% c(\"Up\", \"Down\")]] <- 1\n",
    "rowInfo <- rowInfo[,-1, drop=F]\n",
    "\n",
    "\n",
    "\n",
    "pheatmap(sub,\n",
    "         scale=\"row\",\n",
    "         # display_numbers=T,\n",
    "         # annotation_col = colInfo,\n",
    "         annotation_row = rowInfo,\n",
    "         # clustering_distance_rows = \"binary\",\n",
    "         # cluster_rows = cluster_rows,\n",
    "         # labels_row=labels_row, \n",
    "         # annotation_colors = anno_colors,\n",
    "         main=\"insertion genes expression\",\n",
    "         # col=c(\"#D9DADE\", \"#E05D4B\"),\n",
    "         # col=c(\"#F1F3F5\", colorRampPalette(c(\"#F6A496\",\"#CF2920\"))(25)),\n",
    "         # color=colorRampPalette(c(\"navy\", \"white\", \"firebrick3\"))(8)[2:7],\n",
    "         col=colorRampPalette(c(\"#4678BA\", \"#F7F7F7\", \"#EF8354\"))(11),\n",
    "         # border=\"white\",\n",
    "         show_rownames=F,\n",
    "         # show_colnames=F,\n",
    "         cluster_rows=F,\n",
    "         # cluster_cols=F,\n",
    "         fontsize_row=0.5,\n",
    "         width=3,\n",
    "         height=5,\n",
    "         filename=\"stat/CRC_BC_insertion_genes_expression_heatmap.pdf\")\n",
    "\n",
    "\n",
    "\n",
    "# lg2FC <- data.frame(CRC=CRC_res[rownames(sub), \"log2FoldChange\"], BC=BC_res[rownames(sub), \"log2FoldChange\"])\n",
    "# rownames(lg2FC) <- rownames(sub)\n",
    "BC <- read.table(\"stat/MDAMB231_vs_MCF10A_all_genes.txt\", header=T)\n",
    "rownames(BC) <- BC$gene_id\n",
    "CRC <- read.table(\"stat/HCT116_vs_NCM460_all_genes.txt\", header=T)\n",
    "rownames(CRC) <- CRC$gene_id\n",
    "lg2FC <- data.frame(CRC=CRC[rownames(tb), \"log2FoldChange\"], BC=BC[rownames(tb), \"log2FoldChange\"])\n",
    "rownames(lg2FC) <- rownames(tb)\n",
    "\n",
    "\n",
    "\n",
    "pheatmap(lg2FC[, c(\"BC\", \"CRC\")],\n",
    "         scale=\"row\",\n",
    "         # display_numbers=T,\n",
    "         # annotation_col = colInfo,\n",
    "         annotation_row = rowInfo,\n",
    "         # clustering_distance_rows = \"binary\",\n",
    "         # cluster_rows = cluster_rows,\n",
    "         # labels_row=labels_row, \n",
    "         # annotation_colors = anno_colors,\n",
    "         main=\"insertion genes logFC\",\n",
    "         # col=c(\"#D9DADE\", \"#E05D4B\"),\n",
    "         # col=c(\"#F1F3F5\", colorRampPalette(c(\"#F6A496\",\"#CF2920\"))(25)),\n",
    "         color=colorRampPalette(c(\"navy\", \"white\", \"firebrick3\"))(9)[4:8],\n",
    "         # col=colorRampPalette(c(\"#4678BA\", \"#F7F7F7\", \"#EF8354\"))(11),\n",
    "         # border=\"white\",\n",
    "         show_rownames=F,\n",
    "         # show_colnames=F,\n",
    "         cluster_rows=F,\n",
    "         cluster_cols=F,\n",
    "         fontsize_row=0.5,\n",
    "         width=3,\n",
    "         height=5,\n",
    "         filename=\"stat/CRC_BC_insertion_genes_log2FC_heatmap.pdf\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### correlation between insertion and logFC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## corrlation\n",
    "tb <- tb.all\n",
    "BC <- read.table(\"stat/MDAMB231_vs_MCF10A_all_genes.txt\", header=T)\n",
    "rownames(BC) <- BC$gene_id\n",
    "CRC <- read.table(\"stat/HCT116_vs_NCM460_all_genes.txt\", header=T)\n",
    "rownames(CRC) <- CRC$gene_id\n",
    "lg2FC <- data.frame(CRC=CRC[rownames(tb), \"log2FoldChange\"], BC=BC[rownames(tb), \"log2FoldChange\"])\n",
    "rownames(lg2FC) <- rownames(tb)\n",
    "\n",
    "\n",
    "\n",
    "# MDAMB231 vs MCF10A\n",
    "BC_tb <- tb[, c(\"sgRNA_MDAMB231\", \"sgRNA_MCF10A\")]\n",
    "BC_tb <- BC_tb[!rowSums(BC_tb)==0, ]\n",
    "BC_tmp <- BC_tb$sgRNA_MDAMB231 - BC_tb$sgRNA_MCF10A\n",
    "BC_lg2FC <- lg2FC[rownames(BC_tb), ]\n",
    "cor(BC_lg2FC$BC[complete.cases(BC_lg2FC$BC)], BC_tmp[complete.cases(BC_lg2FC$BC)], method=\"spearman\")\n",
    "\n",
    "\n",
    "# HCT116 vs NCM460\n",
    "CRC_tb <- tb[, c(\"sgRNA_HCT116\", \"sgRNA_NCM460\")]\n",
    "CRC_tb <- CRC_tb[!rowSums(CRC_tb)==0, ]\n",
    "CRC_tmp <- CRC_tb$sgRNA_HCT116 - CRC_tb$sgRNA_NCM460\n",
    "CRC_lg2FC <- lg2FC[rownames(CRC_tb), ]\n",
    "cor(CRC_lg2FC$CRC[complete.cases(CRC_lg2FC$CRC)], CRC_tmp[complete.cases(CRC_lg2FC$CRC)], method=\"spearman\")\n",
    "\n",
    "\n",
    "cor(c(BC_lg2FC$BC[complete.cases(BC_lg2FC$BC)], CRC_lg2FC$CRC[complete.cases(CRC_lg2FC$CRC)]),\n",
    "  c(BC_tmp[complete.cases(BC_lg2FC$BC)], CRC_tmp[complete.cases(CRC_lg2FC$CRC)]),\n",
    "                                                                           method=\"spearman\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ## corrlation\n",
    "# tb <- tb.all\n",
    "# lg2FC\n",
    "\n",
    "# # MDAMB231 vs MCF10A\n",
    "# BC_tb <- tb[, c(\"sgRNA_MDAMB231\", \"sgRNA_MCF10A\")]\n",
    "# BC_tb[BC_tb>0] <- 1\n",
    "# BC_tb <- BC_tb[rowSums(BC_tb)>0, ]\n",
    "# BC_tb[BC_tb==0] <- -1\n",
    "# BC_tmp <- BC_tb$sgRNA_MDAMB231\n",
    "# BC_tmp[BC_tmp==1 & BC_tb$sgRNA_MCF10A==1] <- 0\n",
    "# BC_lg2FC <- lg2FC[rownames(BC_tb), ]\n",
    "# cor(BC_lg2FC$BC[complete.cases(BC_lg2FC$BC)], BC_tmp[complete.cases(BC_lg2FC$BC)], method=\"spearman\")\n",
    "\n",
    "\n",
    "# # HCT116 vs NCM460\n",
    "# CRC_tb <- tb[, c(\"sgRNA_HCT116\", \"sgRNA_NCM460\")]\n",
    "# CRC_tb[CRC_tb>0] <- 1\n",
    "# CRC_tb <- CRC_tb[rowSums(CRC_tb)>0, ]\n",
    "# CRC_tb[CRC_tb==0] <- -1\n",
    "# CRC_tmp <- CRC_tb$sgRNA_HCT116\n",
    "# CRC_tmp[CRC_tmp==1 & CRC_tb$sgRNA_NCM460==1] <- 0\n",
    "# CRC_lg2FC <- lg2FC[rownames(CRC_tb), ]\n",
    "# cor(CRC_lg2FC$CRC[complete.cases(CRC_lg2FC$CRC)], CRC_tmp[complete.cases(CRC_lg2FC$CRC)], method=\"spearman\")\n",
    "\n",
    "\n",
    "# cor(c(BC_lg2FC$BC[complete.cases(BC_lg2FC$BC)], CRC_lg2FC$CRC[complete.cases(CRC_lg2FC$CRC)]),\n",
    "#   c(BC_tmp[complete.cases(BC_lg2FC$BC)], CRC_tmp[complete.cases(CRC_lg2FC$CRC)]),\n",
    "#                                                                            method=\"spearman\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 插入位置相关基因的功能"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Sys.which(\"python\")\n",
    "Sys.setenv(RETICULATE_PYTHON = \"/home/wg_looking/anaconda3/bin/python\")\n",
    "library(reticulate)\n",
    "source_python(\"/NAS/wg_looking/gRNA_ONT/code/NCBI_gene_summary.py\")\n",
    "load(\"stat/TE_insertion.anno.RData\") # TE_insertion.list\n",
    "\n",
    "gene_id <- unique(unlist(lapply(TE_insertion.list, function(x){unique(x$gene_id)})))\n",
    "genes <- setNames(as.character(sub$gene_id), as.character(sub$gene_id))\n",
    "# gene_function.list <- list()\n",
    "for(i in 2804:length(genes)){\n",
    "    gene_id <- genes[i]\n",
    "    Entrez_ID = ENSEMBL_ID_Entrez_ID_all[gene_id, \"ENTREZID\"]\n",
    "    gf = NCBI_gene_summary(as.character(Entrez_ID)) \n",
    "    gene_function.list[gene_id] <- gf\n",
    "}\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": [],
    "toc-hr-collapsed": true
   },
   "source": [
    "# Quantification : assembly gtf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## K562"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(DESeq2)\n",
    "library(tximport)\n",
    "setwd(\"/NAS/wg_looking/gRNA_ONT\")\n",
    "load(\"stat/K562_Diff.RData\") # sgDiff, totalDiff\n",
    "\n",
    "\n",
    "# load Sample info\n",
    "Sample_info <- read.table(\"stat/Sample_info2.txt\", header = T, stringsAsFactors = F)\n",
    "rownames(Sample_info) <- Sample_info$Samples\n",
    "\n",
    "SampleIDs <- c(\"sgRNA_K562_1\", \"totalRNA_K562_1\")\n",
    "\n",
    "\n",
    "# loading quantification\n",
    "files <- file.path(SampleIDs, paste0(SampleIDs,\".salmon_quant_with_assembly_gtf\"), \"quant.sf\")\n",
    "names(files) <- SampleIDs\n",
    "\n",
    "quant.list <- lapply(files, function(x){\n",
    "    quant <- read.table(x, header=T, stringsAsFactors=F)\n",
    "    colnames(quant)[1] <- \"qry_id\"\n",
    "    return(quant)\n",
    "})\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# sgRNA: TE-derived\n",
    "quant.merge <- merge(quant.list[[\"sgRNA_K562_1\"]], sgDiff[[2]], by=\"qry_id\")\n",
    "left <- quant.merge$qry_left\n",
    "left <- gsub(\"\\\\)\", \"\", gsub(\"\\\\(\", \"\", left))\n",
    "quant.merge$left <- as.integer(left)\n",
    "sub.quant.merge <- quant.merge[quant.merge$qry_start<=5 | quant.merge$left<=5,]\n",
    "sub.quant.merge$sub_repeat <- substring(sub.quant.merge$repeat_id, 1, 4)\n",
    "L1_index <- which(sub.quant.merge$matching_class %in% \"LINE/L1\")\n",
    "sub.quant.merge$sub_repeat[L1_index] <- substring(sub.quant.merge$sub_repeat[L1_index] , 1, 3)  \n",
    "sub.quant.merge$sub_repeat[sub.quant.merge$sub_repeat %in% \"L1H\"] <- \"L1HS\"\n",
    "sub.quant.merge <- sub.quant.merge[order(sub.quant.merge$qry_id, sub.quant.merge$sub_repeat),]\n",
    "sub.quant.merge <- sub.quant.merge[!duplicated(sub.quant.merge[,c(\"qry_id\", \"sub_repeat\")]), ]\n",
    "\n",
    "\n",
    "ggplot(sub.quant.merge, aes(x=sub_repeat, y=log(TPM), color=matching_class)) +\n",
    "    geom_point(alpha=0.5) +\n",
    "    geom_jitter(width = 0.25) +\n",
    "    theme_bw() +\n",
    "    theme(\n",
    "      panel.grid=element_blank(),\n",
    "      axis.text.x = element_text(angle = 0, color = \"black\", \n",
    "                                 size = 8, hjust = 0.5, vjust=0.5),\n",
    "      plot.title=element_text(size=14, face=\"bold\")\n",
    "    ) +\n",
    "    # facet_wrap(~class, nrow = 2, scales=\"free\") +\n",
    "    scale_color_manual(values = c(\"#EF8354\", \"#4678BA\")) +\n",
    "    ggtitle(\"sgRNA\")\n",
    "  \n",
    "ggsave(paste0(\"stat/sgRNA_TE_derived_transcripts_repeat_family_TPM.pdf\"),\n",
    "         width = 6, height = 5)\n",
    "  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# sgRNA: TE-derived\n",
    "quant.merge <- merge(quant.list[[\"totalRNA_K562_1\"]], totalDiff[[2]], by=\"qry_id\")\n",
    "left <- quant.merge$qry_left\n",
    "left <- gsub(\"\\\\)\", \"\", gsub(\"\\\\(\", \"\", left))\n",
    "quant.merge$left <- as.integer(left)\n",
    "sub.quant.merge <- quant.merge[quant.merge$qry_start<=5 | quant.merge$left<=5,]\n",
    "sub.quant.merge$sub_repeat <- substring(sub.quant.merge$repeat_id, 1, 4)\n",
    "L1_index <- which(sub.quant.merge$matching_class %in% \"LINE/L1\")\n",
    "sub.quant.merge$sub_repeat[L1_index] <- substring(sub.quant.merge$sub_repeat[L1_index] , 1, 3)  \n",
    "sub.quant.merge$sub_repeat[sub.quant.merge$sub_repeat %in% \"L1H\"] <- \"L1HS\"\n",
    "sub.quant.merge <- sub.quant.merge[order(sub.quant.merge$qry_id, sub.quant.merge$sub_repeat),]\n",
    "sub.quant.merge <- sub.quant.merge[!duplicated(sub.quant.merge[,c(\"qry_id\", \"sub_repeat\")]), ]\n",
    "\n",
    "\n",
    "ggplot(sub.quant.merge, aes(x=sub_repeat, y=log(TPM), color=matching_class)) +\n",
    "    geom_point(alpha=0.5) +\n",
    "    geom_jitter(width = 0.25) +\n",
    "    theme_bw() +\n",
    "    theme(\n",
    "      panel.grid=element_blank(),\n",
    "      axis.text.x = element_text(angle = 0, color = \"black\", \n",
    "                                 size = 8, hjust = 0.5, vjust=0.5),\n",
    "      plot.title=element_text(size=14, face=\"bold\")\n",
    "    ) +\n",
    "    # facet_wrap(~class, nrow = 2, scales=\"free\") +\n",
    "    scale_color_manual(values = c(\"#EF8354\", \"#4678BA\")) +\n",
    "    ggtitle(\"totalRNA\")\n",
    "  \n",
    "ggsave(paste0(\"stat/totalRNA_TE_derived_transcripts_repeat_family_TPM.pdf\"),\n",
    "         width = 6, height = 5)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## sgRNA: HCT116 vs NCM460"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(DESeq2)\n",
    "library(tximport)\n",
    "library(patchwork) # p <- p1 + p3 + plot_layout(ncol = 1, heights = c(1, 1))\n",
    "load(\"stat/HCT116_NCM460_diff.Data\") # HCT116_Diff, NCM460_Diff\n",
    "\n",
    "\n",
    "# load Sample info\n",
    "Sample_info <- read.table(\"stat/Sample_info2.txt\", header = T, stringsAsFactors = F)\n",
    "rownames(Sample_info) <- Sample_info$Samples\n",
    "SampleIDs <- Sample_info$Samples[(Sample_info$CellLine %in% c(\"HCT116\", \"NCM460\")) & (Sample_info$SeqType %in% \"sgRNA\")]\n",
    "\n",
    "\n",
    "# loading new junction information\n",
    "tmap = \"stringtie_res5/HCT116_NCM460/sgRNA_HCT116_NCM460.minCoverage_5.gap_20.sgRNA_HCT116_NCM460.minCoverage_5.gap_20.gtf.tmap\"\n",
    "tb <- read.table(tmap, header = T, stringsAsFactors = F)\n",
    "# new transcript with new junction\n",
    "bed <- \"stringtie_res5/HCT116_NCM460/sgRNA_HCT116_NCM460.minCoverage_5.gap_20.intron.new.bed\" \n",
    "bed.tb <- read.table(bed, stringsAsFactors = F)\n",
    "splice_junction_id <- tstrsplit(bed.tb$V4, \"[|]\")[[3]]\n",
    "splice_junction_id <- unique(splice_junction_id)\n",
    "#\n",
    "tb$classification <- \"other_exonic_overlap\"\n",
    "tb$classification[tb$qry_id %in% splice_junction_id] <- \"new_splice_junction\"\n",
    "tb$classification[tb$class_code %in% \"u\"] <- \"intergenic\"\n",
    "tb$classification[tb$class_code %in% \"i\"] <- \"intragenic\"\n",
    "tb$classification[tb$class_code %in% c(\"c\", \"=\")] <- \"match_reference\"\n",
    "tb <- tb[, c(\"qry_id\", \"classification\", \"class_code\")]\n",
    "colnames(tb) <- c(\"merge_transcript_id\", \"classification\", \"class_code\")\n",
    "\n",
    "\n",
    "## loading tmap: HCT116, NCM460 \n",
    "NCM460_tmap <- read.table(file.path(\"stringtie_res5\", \n",
    "                                    \"NCM460\", \n",
    "                \"sg_HCT116_NCM460_vs_sgNCM460.sgRNA_NCM460.minCoverage_5.gap_20.gtf.tmap\"),\n",
    "                         header=T, stringsAsFactors=F)\n",
    "NCM460_tmap <- NCM460_tmap[, c(\"ref_id\", \"qry_id\", \"class_code\")]\n",
    "colnames(NCM460_tmap) <- c(\"merge_transcript_id\", \"NCM460_transcript_id\", \"NCM460_class_code\")\n",
    "NCM460_tmap <- NCM460_tmap[NCM460_tmap$merge_transcript_id != \"-\", ]\n",
    "NCM460_tmap <- NCM460_tmap[NCM460_tmap$NCM460_transcript_id %in% NCM460_Diff[[1]]$qry_id, ]\n",
    "#\n",
    "HCT116_tmap <- read.table(file.path(\"stringtie_res5\", \n",
    "                                    \"HCT116\", \n",
    "                \"sg_HCT116_NCM460_vs_sgHCT116.sgRNA_HCT116.minCoverage_5.gap_20.gtf.tmap\"),\n",
    "                         header=T, stringsAsFactors=F)\n",
    "HCT116_tmap <- HCT116_tmap[, c(\"ref_id\", \"qry_id\", \"class_code\")]\n",
    "colnames(HCT116_tmap) <- c(\"merge_transcript_id\", \"HCT116_transcript_id\", \"HCT116_class_code\")\n",
    "HCT116_tmap <- HCT116_tmap[HCT116_tmap$merge_transcript_id != \"-\", ]\n",
    "HCT116_tmap <- HCT116_tmap[HCT116_tmap$HCT116_transcript_id %in% HCT116_Diff[[1]]$qry_id, ]\n",
    "#\n",
    "tracking <- merge(HCT116_tmap, NCM460_tmap, by=\"merge_transcript_id\")\n",
    "dim(tracking)\n",
    "tracking <- merge(tb, tracking, by=\"merge_transcript_id\")\n",
    "sub.tracking <- tracking[tracking$HCT116_class_code %in% c(\"=\", \"c\") |\n",
    "                         tracking$NCM460_class_code %in% c(\"=\", \"c\"), ]\n",
    "\n",
    "\n",
    "# loading quantification\n",
    "files <- file.path(SampleIDs, paste0(SampleIDs,\".salmon_quant_with_HCT116_NCM460_gtf\"), \"quant.sf\")\n",
    "names(files) <- SampleIDs\n",
    "txi <- tximport(files, type = \"salmon\", txOut = TRUE)\n",
    "names(txi)\n",
    "dds <- DESeqDataSetFromTximport(txi, \n",
    "                                Sample_info[SampleIDs, c(\"CellLine\"), drop=F], \n",
    "                                ~CellLine)\n",
    "Object <- DESeq(dds\n",
    "                # parallel = T, BPPARAM=MulticoreParam(10)\n",
    "                )\n",
    "norlCount <- counts(Object, normalized=T)\n",
    "res <- results(Object, contrast=c(\"CellLine\", \"HCT116\", \"NCM460\"))\n",
    "res <- res[complete.cases(res),]\n",
    "dif_res <- res[res$padj<0.05 & abs(res$log2FoldChange)>1 & res$baseMean>10, ]\n",
    "\n",
    "\n",
    "## new differential transcripts\n",
    "new.dif_res <- dif_res[rownames(dif_res) %in% sub.tracking$merge_transcript_id, ]\n",
    "new.dif_res$merge_transcript_id <- rownames(new.dif_res)\n",
    "new.dif_res.merge <- merge(data.frame(new.dif_res), sub.tracking, by=\"merge_transcript_id\")\n",
    "new.dif_res.merge$type <- \"\"\n",
    "new.dif_res.merge$type[new.dif_res.merge$HCT116_class_code %in% c(\"=\", \"c\") &\n",
    "                       !(new.dif_res.merge$NCM460_class_code %in% c(\"=\", \"c\"))] <- \"HCT116_specific\"\n",
    "new.dif_res.merge$type[new.dif_res.merge$NCM460_class_code %in% c(\"=\", \"c\") &\n",
    "                       !(new.dif_res.merge$HCT116_class_code %in% c(\"=\", \"c\"))] <- \"NCM460_specific\"\n",
    "new.dif_res.merge$type[new.dif_res.merge$NCM460_class_code %in% c(\"=\", \"c\") &\n",
    "                       (new.dif_res.merge$HCT116_class_code %in% c(\"=\", \"c\"))] <- \"Common\"\n",
    "table(new.dif_res.merge$type)\n",
    "\n",
    "new.dif_res.merge <- new.dif_res.merge[order(new.dif_res.merge$merge_transcript_id, new.dif_res.merge$type), ]\n",
    "new.dif_res.duplicated <- new.dif_res.merge[!duplicated(new.dif_res.merge$merge_transcript_id), ]\n",
    "\n",
    "new.dif_res.duplicated <- new.dif_res.duplicated[order(new.dif_res.duplicated$type, \n",
    "                                                       new.dif_res.duplicated$classification, \n",
    "                                                       new.dif_res.duplicated$log2FoldChange), ]\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(pheatmap)\n",
    "\n",
    "cols <- c(\"#F4B540\", \"#EF8354\", \"#4678BA\", \"#87B07A\", \"#BA748A\")\n",
    "names(cols) <- unique(new.dif_res.duplicated$classification)\n",
    "\n",
    "colInfo <- Sample_info[SampleIDs, c(\"CellLine\"), drop=F]\n",
    "rowInfo <- new.dif_res.duplicated[, c(\"type\", \"classification\")]\n",
    "rownames(rowInfo) <- new.dif_res.duplicated$merge_transcript_id\n",
    "anno_colors <- list(CellLine=c(NCM460=\"#4678BA\", HCT116=\"#F4B540\"),\n",
    "                   type=c(Common=\"#76A45B\", HCT116_specific=\"#F4B540\", NCM460_specific=\"#4678BA\"),\n",
    "                   classification=cols)\n",
    "\n",
    "\n",
    "\n",
    "pheatmap(norlCount[new.dif_res.duplicated$merge_transcript_id,],\n",
    "         # display_numbers=T,\n",
    "         scale=\"row\",\n",
    "         annotation_col = colInfo,\n",
    "         annotation_row = rowInfo,\n",
    "         annotation_colors = anno_colors,\n",
    "         main=\"differential assembly transcript expression\",\n",
    "         col=colorRampPalette(rev(brewer.pal(11,\"RdBu\")))(21),\n",
    "         # col=colorRampPalette(c(\"#4678BA\", \"#F7F7F7\", \"#EF8354\"))(11),\n",
    "         border=NA,\n",
    "         show_rownames=F,\n",
    "         show_colnames=F,\n",
    "         cluster_rows=F,\n",
    "         cluster_cols=F,\n",
    "         width=6.5,\n",
    "         height=6,\n",
    "         filename=\"stat/HCT116_NCM460_assembly_trans_dif_heatmap.pdf\")\n",
    "\n",
    "\n",
    "# out_res <- dif_res[, c(\"baseMean\", \"log2FoldChange\", \"padj\")]\n",
    "# out_res$dif <- \"HCT116_over\"\n",
    "# out_res$dif[out_res$log2FoldChange<0] <- \"NCM460_over\"\n",
    "# out_res <- out_res[order(out_res$dif),]\n",
    "# out_res <- data.frame(gene_id=rownames(out_res), out_res)\n",
    "# out_res <- data.frame(out_res, gene_name=transcript_gtf[rownames(out_res), \"gene_name\"])\n",
    "# transcript_res <- out_res\n",
    "# # write.table(out_res, \"stat/CRC_differential_transcripts.txt\", row.names = F, col.names = T, quote = F, sep=\"\\t\")\n",
    "# write.table(out_res, \"stat/CRC_all_transcripts.txt\", row.names = F, col.names = T, quote = F, sep=\"\\t\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### TE-derived"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(plyr)\n",
    "library(data.table)\n",
    "library(patchwork)\n",
    "setwd(\"/NAS/wg_looking/gRNA_ONT\")\n",
    "load(\"stat/HCT116_NCM460_diff.Data\") # HCT116_Diff, NCM460_Diff\n",
    "\n",
    "\n",
    "# load Sample info\n",
    "Sample_info <- read.table(\"stat/Sample_info2.txt\", header = T, stringsAsFactors = F)\n",
    "rownames(Sample_info) <- Sample_info$Samples\n",
    "\n",
    "SampleIDs <- Sample_info$Samples[Sample_info$CellLine %in% c(\"HCT116\", \"NCM460\") &\n",
    "                                Sample_info$SeqType %in% c(\"sgRNA\")]\n",
    "\n",
    "\n",
    "# loading quantification\n",
    "files <- file.path(SampleIDs, paste0(SampleIDs,\".salmon_quant_with_assembly_gtf\"), \"quant.sf\")\n",
    "names(files) <- SampleIDs\n",
    "\n",
    "quant.list <- lapply(files, function(x){\n",
    "    quant <- read.table(x, header=T, stringsAsFactors=F)\n",
    "    colnames(quant)[1] <- \"qry_id\"\n",
    "    quant$RPM <- (quant$NumReads)/sum(quant$NumReads)*10^6\n",
    "    \n",
    "    return(quant)\n",
    "})\n",
    "\n",
    "\n",
    "\n",
    "# files <- file.path(SampleIDs, \"prepDE_assembly_gtf_transcript_count.csv\")\n",
    "# names(files) <- SampleIDs\n",
    "\n",
    "# quant.list <- lapply(files, function(x){\n",
    "#     quant <- read.csv(x, header=T, stringsAsFactors=F)\n",
    "#     colnames(quant) <- c(\"qry_id\", \"NumReads\")\n",
    "#     quant$RPM <- (quant$NumReads)/sum(quant$NumReads)*10^6\n",
    "#     return(quant)\n",
    "# })\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "\n",
    "TE_RPM <- lapply(c(\"HCT116\", \"NCM460\"), function(CellLine){\n",
    "    IDs <- Sample_info$Samples[Sample_info$CellLine %in% CellLine &\n",
    "                                Sample_info$SeqType %in% c(\"sgRNA\")]\n",
    "    #\n",
    "    sub.quant.list <- lapply(IDs, function(ID){\n",
    "        q <- quant.list[[ID]][, c(\"qry_id\", \"RPM\")]\n",
    "        colnames(q) <- c(\"qry_id\", ID)\n",
    "        return(q)\n",
    "    })\n",
    "    #\n",
    "    quant <- join_all(sub.quant.list)\n",
    "    quant$RPM <- rowMeans(quant[, c(2:4)])\n",
    "    # \n",
    "    if(CellLine==\"HCT116\"){\n",
    "        quant.merge <- merge(quant, HCT116_Diff[[2]], by=\"qry_id\")\n",
    "    }\n",
    "    if(CellLine==\"NCM460\"){\n",
    "        quant.merge <- merge(quant, NCM460_Diff[[2]], by=\"qry_id\")\n",
    "    }\n",
    "    #\n",
    "    left <- quant.merge$qry_left\n",
    "    left <- gsub(\"\\\\)\", \"\", gsub(\"\\\\(\", \"\", left))\n",
    "    quant.merge$left <- as.integer(left)\n",
    "    sub.quant.merge <- quant.merge[quant.merge$qry_start<=5 | quant.merge$left<=5,]\n",
    "    sub.quant.merge$sub_repeat <- substring(sub.quant.merge$repeat_id, 1, 4)\n",
    "    L1_index <- which(sub.quant.merge$matching_class %in% \"LINE/L1\")\n",
    "    sub.quant.merge$sub_repeat[L1_index] <- substring(sub.quant.merge$sub_repeat[L1_index] , 1, 3)  \n",
    "    sub.quant.merge$sub_repeat[sub.quant.merge$sub_repeat %in% \"L1H\"] <- \"L1HS\"\n",
    "    sub.quant.merge <- sub.quant.merge[order(sub.quant.merge$qry_id, sub.quant.merge$sub_repeat),]\n",
    "    sub.quant.merge <- sub.quant.merge[!duplicated(sub.quant.merge[,c(\"qry_id\", \"sub_repeat\")]), ]\n",
    "    sub.quant.merge <- data.frame(sub.quant.merge[, !(colnames(sub.quant.merge) %in% IDs)], \n",
    "                                  CellLine=CellLine)\n",
    "    \n",
    "    return(sub.quant.merge)\n",
    "})\n",
    "\n",
    "pp <- rbindlist(TE_RPM)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "p1 <- ggplot(pp, aes(x=sub_repeat, y=log2(RPM+1), color=matching_class, fill=matching_class)) +\n",
    "    geom_violin(alpha=0.4) + \n",
    "    stat_summary(fun= median, geom = \"point\",\n",
    "               shape = 23, size = 2, color = \"red\") +\n",
    "    # geom_point(alpha=0.1) +\n",
    "    # geom_jitter(width = 0.25, alpha=0.1) +\n",
    "    theme_bw() +\n",
    "    theme(\n",
    "      panel.grid=element_blank(),\n",
    "      axis.text.x = element_text(angle = 0, color = \"black\", \n",
    "                                 size = 8, hjust = 0.5, vjust=0.5),\n",
    "      plot.title=element_text(size=14, face=\"bold\")\n",
    "    ) +\n",
    "    facet_wrap(~CellLine, nrow = 1, scales=\"free\") +\n",
    "    scale_color_manual(values = c(\"#EF8354\", \"#4678BA\")) +\n",
    "    scale_fill_manual(values = c(\"#EF8354\", \"#4678BA\")) +\n",
    "    ggtitle(\"TE-derived\")\n",
    "\n",
    "pp2 <- data.frame(table(pp$CellLine, pp$matching_class, pp$sub_repeat))\n",
    "colnames(pp2) <- c(\"CellLine\", \"matching_class\", \"sub_repeat\", \"Freq\")\n",
    "pp2 <- pp2[pp2$Freq>0,] \n",
    "\n",
    "p2 <- ggplot(pp2, aes(x=sub_repeat, y=Freq,  fill=matching_class)) +\n",
    "    geom_bar(stat=\"identity\") + \n",
    "    geom_text(aes(label=Freq), \n",
    "              =position_dodge(width=0.9), vjust=-0.25) +\n",
    "    theme_bw() +\n",
    "    theme(\n",
    "      panel.grid=element_blank(),\n",
    "      axis.text.x = element_text(angle = 0, color = \"black\", \n",
    "                                 size = 8, hjust = 0.5, vjust=0.5),\n",
    "      plot.title=element_text(size=14, face=\"bold\")\n",
    "    ) +\n",
    "    facet_wrap(~CellLine, nrow = 1, scales=\"free\") +\n",
    "    scale_fill_manual(values = c(\"#EF8354\", \"#4678BA\")) +\n",
    "    ggtitle(\"\") +\n",
    "    ylim(0,8500)\n",
    "  \n",
    "\n",
    "p <- p1 + p2 + plot_layout(ncol = 1, heights = c(1, 1))\n",
    "\n",
    "ggsave(paste0(\"stat/HCT116_vs_NCM460_TE_derived_transcripts_repeat_family_RPM.pdf\"),\n",
    "         width = 10, height = 6)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for common transcripts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# HCT116_Diff: TE-derived\n",
    "tmap <- \"stringtie_res5/HCT116/sgHCT116_vs_sgNCM460.sgRNA_HCT116.minCoverage_5.gap_20.gtf.tmap\"\n",
    "tmap.tb <- read.table(tmap, header = T, stringsAsFactors = F)\n",
    "tmap.tb <- tmap.tb[, -which(colnames(tmap.tb)%in%c(\"FPKM\", \"TPM\", \"cov\"))]\n",
    "\n",
    "index <- HCT116_Diff[[2]]$qry_id %in% tmap.tb[tmap.tb$class_code %in% \"=\", \"qry_id\"]\n",
    "length(unique(HCT116_Diff[[2]]$qry_id[index]))\n",
    "index <- HCT116_Diff[[2]]$qry_id %in% tmap.tb[tmap.tb$class_code %in% \"c\", \"qry_id\"]\n",
    "length(unique(HCT116_Diff[[2]]$qry_id[index]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## loading tmap: HCT116, NCM460 \n",
    "NCM460_tmap <- read.table(file.path(\"stringtie_res5\", \n",
    "                                    \"NCM460\", \n",
    "                \"sg_HCT116_NCM460_vs_sgNCM460.sgRNA_NCM460.minCoverage_5.gap_20.gtf.tmap\"),\n",
    "                         header=T, stringsAsFactors=F)\n",
    "NCM460_tmap <- NCM460_tmap[, c(\"ref_id\", \"qry_id\", \"class_code\")]\n",
    "colnames(NCM460_tmap) <- c(\"merge_transcript_id\", \"NCM460_transcript_id\", \"NCM460_class_code\")\n",
    "NCM460_tmap <- NCM460_tmap[NCM460_tmap$merge_transcript_id != \"-\", ]\n",
    "NCM460_tmap <- NCM460_tmap[NCM460_tmap$NCM460_transcript_id %in% NCM460_Diff[[2]]$qry_id, ]\n",
    "#\n",
    "HCT116_tmap <- read.table(file.path(\"stringtie_res5\", \n",
    "                                    \"HCT116\", \n",
    "                \"sg_HCT116_NCM460_vs_sgHCT116.sgRNA_HCT116.minCoverage_5.gap_20.gtf.tmap\"),\n",
    "                         header=T, stringsAsFactors=F)\n",
    "HCT116_tmap <- HCT116_tmap[, c(\"ref_id\", \"qry_id\", \"class_code\")]\n",
    "colnames(HCT116_tmap) <- c(\"merge_transcript_id\", \"HCT116_transcript_id\", \"HCT116_class_code\")\n",
    "HCT116_tmap <- HCT116_tmap[HCT116_tmap$merge_transcript_id != \"-\", ]\n",
    "HCT116_tmap <- HCT116_tmap[HCT116_tmap$HCT116_transcript_id %in% HCT116_Diff[[2]]$qry_id, ]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### TE autonomous"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pp <- rbindlist(TE_TPM)\n",
    "pp <- pp[pp$classification %in% c(\"intragenic\", \"intergenic\"),]\n",
    "\n",
    "p1 <- ggplot(pp, aes(x=sub_repeat, y=log2(TPM+1), color=matching_class, fill=matching_class)) +\n",
    "    geom_violin(alpha=0.4) + \n",
    "    stat_summary(fun= median, geom = \"point\",\n",
    "               shape = 23, size = 2, color = \"red\") +\n",
    "    # geom_point(alpha=0.1) +\n",
    "    # geom_jitter(width = 0.25, alpha=0.1) +\n",
    "    theme_bw() +\n",
    "    theme(\n",
    "      panel.grid=element_blank(),\n",
    "      axis.text.x = element_text(angle = 0, color = \"black\", \n",
    "                                 size = 8, hjust = 0.5, vjust=0.5),\n",
    "      plot.title=element_text(size=14, face=\"bold\", hjust=0.5)\n",
    "    ) +\n",
    "    facet_wrap(~CellLine, nrow = 1, scales=\"free\") +\n",
    "    scale_color_manual(values = c(\"#EF8354\", \"#4678BA\")) +\n",
    "    scale_fill_manual(values = c(\"#EF8354\", \"#4678BA\")) +\n",
    "    ggtitle(\"TE autonomous\")\n",
    "\n",
    "pp2 <- data.frame(table(pp$CellLine, pp$matching_class, pp$sub_repeat))\n",
    "colnames(pp2) <- c(\"CellLine\", \"matching_class\", \"sub_repeat\", \"Freq\")\n",
    "pp2 <- pp2[pp2$Freq>0,] \n",
    "\n",
    "p2 <- ggplot(pp2, aes(x=sub_repeat, y=Freq,  fill=matching_class)) +\n",
    "    geom_bar(stat=\"identity\") + \n",
    "    geom_text(aes(label=Freq), position=position_dodge(width=0.9), vjust=-0.25) +\n",
    "    theme_bw() +\n",
    "    theme(\n",
    "      panel.grid=element_blank(),\n",
    "      axis.text.x = element_text(angle = 0, color = \"black\", \n",
    "                                 size = 8, hjust = 0.5, vjust=0.5),\n",
    "      plot.title=element_text(size=14, face=\"bold\")\n",
    "    ) +\n",
    "    facet_wrap(~CellLine, nrow = 1, scales=\"free\") +\n",
    "    scale_fill_manual(values = c(\"#EF8354\", \"#4678BA\")) +\n",
    "    ggtitle(\"\") +\n",
    "    ylim(0,4500)\n",
    "  \n",
    "\n",
    "\n",
    "p <- p1 + p2 + plot_layout(ncol = 1, heights = c(1, 1))\n",
    "\n",
    "ggsave(paste0(\"stat/HCT116_vs_NCM460_TE_autonomous_transcripts_repeat_family_TPM.pdf\"),\n",
    "         width = 10, height = 6)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## sgRNA: MDAMB231 vs MCF10A"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### TE-derived"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(plyr)\n",
    "library(patchwork) # p <- p1 + p3 + plot_layout(ncol = 1, heights = c(1, 1))\n",
    "setwd(\"/NAS/wg_looking/gRNA_ONT\")\n",
    "load(\"stat/MDAMB231_MCF10A_diff.Data\") # HCT116_Diff, NCM460_Diff\n",
    "\n",
    "\n",
    "# load Sample info\n",
    "Sample_info <- read.table(\"stat/Sample_info2.txt\", header = T, stringsAsFactors = F)\n",
    "rownames(Sample_info) <- Sample_info$Samples\n",
    "\n",
    "SampleIDs <- Sample_info$Samples[Sample_info$CellLine %in% c(\"MDAMB231\", \"MCF10A\") &\n",
    "                                Sample_info$SeqType %in% c(\"sgRNA\")]\n",
    "\n",
    "\n",
    "# # loading quantification\n",
    "# files <- file.path(SampleIDs, paste0(SampleIDs,\".salmon_quant_with_assembly_gtf\"), \"quant.sf\")\n",
    "# names(files) <- SampleIDs\n",
    "\n",
    "# quant.list <- lapply(files, function(x){\n",
    "#     quant <- read.table(x, header=T, stringsAsFactors=F)\n",
    "#     colnames(quant)[1] <- \"qry_id\"\n",
    "#     return(quant)\n",
    "# })\n",
    "\n",
    "\n",
    "files <- file.path(SampleIDs, \"prepDE_assembly_gtf_transcript_count.csv\")\n",
    "names(files) <- SampleIDs\n",
    "\n",
    "quant.list <- lapply(files, function(x){\n",
    "    quant <- read.csv(x, header=T, stringsAsFactors=F)\n",
    "    colnames(quant) <- c(\"qry_id\", \"NumReads\")\n",
    "    quant$RPM <- (quant$NumReads)/sum(quant$NumReads)*10^6\n",
    "    return(quant)\n",
    "})\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# HCT116_Diff: TE-derived\n",
    "TE_RPM <- lapply(c(\"MDAMB231\", \"MCF10A\"), function(CellLine){\n",
    "    IDs <- Sample_info$Samples[Sample_info$CellLine %in% CellLine &\n",
    "                                Sample_info$SeqType %in% c(\"sgRNA\")]\n",
    "    #\n",
    "    sub.quant.list <- lapply(IDs, function(ID){\n",
    "        q <- quant.list[[ID]][, c(\"qry_id\", \"RPM\")]\n",
    "        colnames(q) <- c(\"qry_id\", ID)\n",
    "        return(q)\n",
    "    })\n",
    "    #\n",
    "    quant <- join_all(sub.quant.list)\n",
    "    quant$RPM <- rowMeans(quant[, c(2:4)])\n",
    "    # \n",
    "    if(CellLine==\"MDAMB231\"){\n",
    "        quant.merge <- merge(quant, MDAMB231_Diff[[2]], by=\"qry_id\")\n",
    "    }\n",
    "    if(CellLine==\"MCF10A\"){\n",
    "        quant.merge <- merge(quant, MCF10A_Diff[[2]], by=\"qry_id\")\n",
    "    }\n",
    "    #\n",
    "    left <- quant.merge$qry_left\n",
    "    left <- gsub(\"\\\\)\", \"\", gsub(\"\\\\(\", \"\", left))\n",
    "    quant.merge$left <- as.integer(left)\n",
    "    sub.quant.merge <- quant.merge[quant.merge$qry_start<=5 | quant.merge$left<=5,]\n",
    "    sub.quant.merge$sub_repeat <- substring(sub.quant.merge$repeat_id, 1, 4)\n",
    "    L1_index <- which(sub.quant.merge$matching_class %in% \"LINE/L1\")\n",
    "    sub.quant.merge$sub_repeat[L1_index] <- substring(sub.quant.merge$sub_repeat[L1_index] , 1, 3)  \n",
    "    sub.quant.merge$sub_repeat[sub.quant.merge$sub_repeat %in% \"L1H\"] <- \"L1HS\"\n",
    "    sub.quant.merge <- sub.quant.merge[order(sub.quant.merge$qry_id, sub.quant.merge$sub_repeat),]\n",
    "    sub.quant.merge <- sub.quant.merge[!duplicated(sub.quant.merge[,c(\"qry_id\", \"sub_repeat\")]), ]\n",
    "    sub.quant.merge <- data.frame(sub.quant.merge[, !(colnames(sub.quant.merge) %in% IDs)], \n",
    "                                  CellLine=CellLine)\n",
    "    \n",
    "    return(sub.quant.merge)\n",
    "})\n",
    "\n",
    "pp <- rbindlist(TE_RPM)\n",
    "\n",
    "\n",
    "  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "p1 <- ggplot(pp, aes(x=sub_repeat, y=log2(RPM+1), color=matching_class, fill=matching_class)) +\n",
    "    geom_violin(alpha=0.4) + \n",
    "    stat_summary(fun= median, geom = \"point\",\n",
    "               shape = 23, size = 2, color = \"red\") +\n",
    "    # geom_point(alpha=0.1) +\n",
    "    # geom_jitter(width = 0.25, alpha=0.1) +\n",
    "    theme_bw() +\n",
    "    theme(\n",
    "      panel.grid=element_blank(),\n",
    "      axis.text.x = element_text(angle = 0, color = \"black\", \n",
    "                                 size = 8, hjust = 0.5, vjust=0.5),\n",
    "      plot.title=element_text(size=14, face=\"bold\")\n",
    "    ) +\n",
    "    facet_wrap(~CellLine, nrow = 1, scales=\"free\") +\n",
    "    scale_color_manual(values = c(\"#EF8354\", \"#4678BA\")) +\n",
    "    scale_fill_manual(values = c(\"#EF8354\", \"#4678BA\")) +\n",
    "    ggtitle(\"TE-derived\")\n",
    "\n",
    "pp2 <- data.frame(table(pp$CellLine, pp$matching_class, pp$sub_repeat))\n",
    "colnames(pp2) <- c(\"CellLine\", \"matching_class\", \"sub_repeat\", \"Freq\")\n",
    "pp2 <- pp2[pp2$Freq>0,] \n",
    "\n",
    "p2 <- ggplot(pp2, aes(x=sub_repeat, y=Freq,  fill=matching_class)) +\n",
    "    geom_bar(stat=\"identity\") + \n",
    "    geom_text(aes(label=Freq), position=position_dodge(width=0.9), vjust=-0.25) +\n",
    "    theme_bw() +\n",
    "    theme(\n",
    "      panel.grid=element_blank(),\n",
    "      axis.text.x = element_text(angle = 0, color = \"black\", \n",
    "                                 size = 8, hjust = 0.5, vjust=0.5),\n",
    "      plot.title=element_text(size=14, face=\"bold\")\n",
    "    ) +\n",
    "    facet_wrap(~CellLine, nrow = 1, scales=\"free\") +\n",
    "    scale_fill_manual(values = c(\"#EF8354\", \"#4678BA\")) +\n",
    "    ggtitle(\"\") +\n",
    "    ylim(0,6500)\n",
    "  \n",
    "\n",
    "\n",
    "p <- p1 + p2 + plot_layout(ncol = 1, heights = c(1, 1))\n",
    "\n",
    "  \n",
    "ggsave(paste0(\"stat/MDAMB231_vs_MCF10A_TE_derived_transcripts_repeat_family_RPM.pdf\"),\n",
    "         width = 10, height = 6)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### TE autonomous"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pp <- data.frame(rbindlist(TE_TPM))\n",
    "pp <- pp[pp$classification %in% c(\"intragenic\", \"intergenic\"),]\n",
    "\n",
    "p1 <- ggplot(pp, aes(x=sub_repeat, y=log2(TPM+1), color=matching_class, fill=matching_class)) +\n",
    "    geom_violin(alpha=0.4) + \n",
    "    stat_summary(fun= median, geom = \"point\",\n",
    "               shape = 23, size = 2, color = \"red\") +\n",
    "    # geom_point(alpha=0.1) +\n",
    "    # geom_jitter(width = 0.25, alpha=0.1) +\n",
    "    theme_bw() +\n",
    "    theme(\n",
    "      panel.grid=element_blank(),\n",
    "      axis.text.x = element_text(angle = 0, color = \"black\", \n",
    "                                 size = 8, hjust = 0.5, vjust=0.5),\n",
    "      plot.title=element_text(size=14, face=\"bold\", hjust=0.5)\n",
    "    ) +\n",
    "    facet_wrap(~CellLine, nrow = 1, scales=\"free\") +\n",
    "    scale_color_manual(values = c(\"#EF8354\", \"#4678BA\")) +\n",
    "    scale_fill_manual(values = c(\"#EF8354\", \"#4678BA\")) +\n",
    "    ggtitle(\"TE autonomous\")\n",
    "\n",
    "pp2 <- data.frame(table(pp$CellLine, pp$matching_class, pp$sub_repeat))\n",
    "colnames(pp2) <- c(\"CellLine\", \"matching_class\", \"sub_repeat\", \"Freq\")\n",
    "pp2 <- pp2[pp2$Freq>0,] \n",
    "\n",
    "p2 <- ggplot(pp2, aes(x=sub_repeat, y=Freq,  fill=matching_class)) +\n",
    "    geom_bar(stat=\"identity\") + \n",
    "    geom_text(aes(label=Freq), position=position_dodge(width=0.9), vjust=-0.25) +\n",
    "    theme_bw() +\n",
    "    theme(\n",
    "      panel.grid=element_blank(),\n",
    "      axis.text.x = element_text(angle = 0, color = \"black\", \n",
    "                                 size = 8, hjust = 0.5, vjust=0.5),\n",
    "      plot.title=element_text(size=14, face=\"bold\")\n",
    "    ) +\n",
    "    facet_wrap(~CellLine, nrow = 1, scales=\"free\") +\n",
    "    scale_fill_manual(values = c(\"#EF8354\", \"#4678BA\")) +\n",
    "    ggtitle(\"\") +\n",
    "    ylim(0,3500)\n",
    "  \n",
    "\n",
    "\n",
    "p <- p1 + p2 + plot_layout(ncol = 1, heights = c(1, 1))\n",
    "\n",
    "ggsave(paste0(\"stat/MDAMB231_vs_MCF10A_TE_autonomous_transcripts_repeat_family_TPM.pdf\"),\n",
    "         width = 10, height = 6)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "# TE expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(\"Rsubread\")\n",
    "library(dplyr)\n",
    "library(limma)\n",
    "setwd(\"/NAS/wg_looking/gRNA_ONT\")\n",
    "load(\"stat/assembly_preData.RData\")\n",
    "load(\"stat/hg38.LINE.SINE.RData\") # RM.bed\n",
    "  \n",
    "Sample_info <- read.table(\"stat/Sample_info2.txt\", header = T, stringsAsFactors = F)\n",
    "rownames(Sample_info) <- Sample_info$Samples\n",
    "\n",
    "IDs <- setNames(Sample_info$Samples, Sample_info$Samples)\n",
    "\n",
    "RM.bed$TE_name <- paste0(paste(RM.bed$qry_id, RM.bed$qry_start, RM.bed$qry_end, sep=\"_\"), \":\", RM.bed$repeat_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "count.list <- lapply(IDs, function(ID){\n",
    "    cFile = file.path(ID, paste0(ID, \".TE.featureCounts\"))\n",
    "    tb <- read.table(cFile, header=T)\n",
    "    colnames(tb)[7] <- ID\n",
    "    \n",
    "    return(tb)\n",
    "})\n",
    "\n",
    "count.tb <- join_all(count.list)\n",
    "rownames(count.tb) <- count.tb$Geneid\n",
    "count.tb <- count.tb[count.tb$Geneid %in% RM.bed$TE_name, ]\n",
    "dim(count.tb) # 1234813\n",
    "\n",
    "save(count.tb, file=\"stat/TE.quantify.count.RData\")\n",
    "\n",
    "#\n",
    "load(\"stat/TE.quantify.count.RData\")\n",
    "v <- c(\"totalRNA_HCT116\", \"totalRNA_NCM460\", \"totalRNA_NCM460_2\")\n",
    "write.table(count.tb[, !(colnames(count.tb) %in% v)], \n",
    "                         \"stat/TE.quantify.count.txt\", row.names = F, col.names = T, sep=\"\\t\", quote=F)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## TE location"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(bedtoolsr)\n",
    "load(\"stat/TE.quantify.count.RData\")\n",
    "load(\"stat/hg38.gtf.RData\")\n",
    "# loading gff file\n",
    "gffFile = \"/NAS/wg_looking/human_ref/gencode.v37.chr_patch_hapl_scaff.annotation.gff3\"\n",
    "gff <- rtracklayer::import(gffFile)\n",
    "gff <- as.data.frame(gff)\n",
    "# intergenic\n",
    "# bedtools sort -i gencode.v37.chr_patch_hapl_scaff.transcript.bed \\\n",
    "# | bedtools merge -i - \\\n",
    "# | bedtools subtract -a GRCh38.p13.genome.fa.bed -b - > gencode.v37.chr_patch_hapl_scaff.annotation.intergenic.bed\n",
    "\n",
    "\n",
    "\n",
    "###### intron exon intergenic 5UTR 3UTR\n",
    "TE_location <- count.tb[, c(\"Chr\", \"Start\", \"End\", \"Geneid\")]\n",
    "\n",
    "# intron\n",
    "intron_bed <- read.table(\"/NAS/wg_looking/human_ref/gencode.v37.chr_patch_hapl_scaff.annotation.intron.bed\", \n",
    "                         stringsAsFactors = F)\n",
    "intron_res <- bt.intersect(a=TE_location, intron_bed, wa=T)\n",
    "\n",
    "# exon\n",
    "exon_bed <- hg38.gtf[hg38.gtf$type==\"exon\", c(\"seqnames\",\"start\",\"end\")]\n",
    "exon_res <- bt.intersect(a=TE_location, exon_bed, wa=T)\n",
    "\n",
    "\n",
    "# intergenic\n",
    "intergenic_bed <- read.table(\"/NAS/wg_looking/human_ref/gencode.v37.chr_patch_hapl_scaff.annotation.intergenic.bed\", \n",
    "                         stringsAsFactors = F)\n",
    "intergenic_res <- bt.intersect(a=TE_location, intergenic_bed, wa=T)\n",
    "\n",
    "\n",
    "\n",
    "# 5UTR\n",
    "UTR5_bed <- gff[gff$type %in% \"five_prime_UTR\", c(\"seqnames\", \"start\", \"end\")]\n",
    "UTR5_res <- bt.intersect(a=TE_location, UTR5_bed, wa=T)\n",
    "\n",
    "# 3UTR\n",
    "UTR3_bed <- gff[gff$type %in% \"three_prime_UTR\", c(\"seqnames\", \"start\", \"end\")]\n",
    "UTR3_res <- bt.intersect(a=TE_location, UTR3_bed, wa=T)\n",
    "\n",
    "# CDS\n",
    "CDS_bed <- gff[gff$type %in% \"CDS\", c(\"seqnames\", \"start\", \"end\")]\n",
    "CDS_res <- bt.intersect(a=b.anno, CDS_bed, wa=T)\n",
    "\n",
    "#\n",
    "TE_location$exon <- 0\n",
    "TE_location$intron <- 0\n",
    "TE_location$intergenic <- 0\n",
    "TE_location$UTR5 <- 0\n",
    "TE_location$UTR3 <- 0\n",
    "TE_location$CDS <- 0\n",
    "TE_location$exon[rownames(TE_location) %in% exon_res$V4] <- 1\n",
    "TE_location$intron[rownames(TE_location) %in% intron_res$V4] <- 1\n",
    "TE_location$intergenic[rownames(TE_location) %in% intergenic_res$V4 ] <- 1\n",
    "TE_location$UTR5[rownames(TE_location) %in% UTR5_res$V4] <- 1\n",
    "TE_location$UTR3[rownames(TE_location) %in% UTR3_res$V4 ] <- 1\n",
    "TE_location$CDS[rownames(TE_location) %in% CDS_res$V4 ] <- 1\n",
    "colnames(TE_location)[4] <- \"TE_name\"\n",
    "\n",
    "\n",
    "# gene annotation\n",
    "gene_bed <- hg38.gtf[hg38.gtf$type==\"gene\", c(\"seqnames\", \"start\",\"end\", \"gene_id\", \"gene_name\")]\n",
    "TE_location_gene <- bt.intersect(a=TE_location, gene_bed, wa=T, wb=T)\n",
    "colnames(TE_location_gene) <- c(colnames(TE_location), colnames(gene_bed))\n",
    "colnames(TE_location_gene)[4] <- \"TE_name\"\n",
    "\n",
    "paste_func <- function(x){paste(x, collapse=\",\")}\n",
    "test1 <- aggregate(gene_id~TE_name, data=TE_location_gene, paste_func)\n",
    "test2 <- aggregate(gene_name~TE_name, data=TE_location_gene, paste_func)\n",
    "test1$gene_id <- unlist(test1$gene_id)\n",
    "test2$gene_name <- unlist(test2$gene_name)\n",
    "\n",
    "\n",
    "TE_location.anno <- join_all(list(TE_location, test1, test2))\n",
    "TE_location <- TE_location.anno\n",
    "TE_location$sub_repeat <- tstrsplit(TE_location$TE_name, \":\")[[2]]\n",
    "TE_location$sub_repeat <- substring(TE_location$sub_repeat, 1, 4)\n",
    "TE_location$sub_repeat[grepl(\"L1P\",TE_location$sub_repeat)] <- \"L1P\"\n",
    "\n",
    "rownames(TE_location) <- TE_location$TE_name\n",
    "\n",
    "save(TE_location, file=\"stat/TE_location.RData\")\n",
    "\n",
    "\n",
    "table(rowSums(TE_location[, c(\"exon\", \"intron\", \"intergenic\")]))\n",
    "which(rowSums(TE_location[, c(\"exon\", \"intron\", \"intergenic\")])==0)[1:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# the genomic distribution reference TE\n",
    "load(\"stat/TE_location.RData\")\n",
    "\n",
    "gf <- TE_location[, c(\"CDS\", \"intron\", \"intergenic\", \"UTR5\", \"UTR3\")]\n",
    "rs <- rowSums(gf)\n",
    "gf <- gf/rs\n",
    "cs <- colSums(gf, na.rm = TRUE)\n",
    "cs <- data.frame(Feature=names(cs), Freq=cs)\n",
    "cs$Percentage <- cs$Freq/sum(cs$Freq)*100\n",
    "\n",
    "\n",
    "\n",
    "cols <- c(UTR3=\"#EE9248\", UTR5=\"#EDC26D\", intergenic=\"#8C9FCA\", CDS=\"#5CB1DD\", intron=\"#4A89D5\")\n",
    "plot <- ggplot(cs, aes(x=\"\", y=Percentage, fill=Feature))+\n",
    "    geom_bar(width = 1, stat = \"identity\") +\n",
    "    coord_polar(\"y\", start=0) +\n",
    "    scale_fill_manual(values = cols)+\n",
    "    theme_minimal() +\n",
    "    theme(\n",
    "      axis.title.x = element_blank(),\n",
    "      axis.title.y = element_blank(),\n",
    "      panel.border = element_blank(),\n",
    "      panel.grid=element_blank(),\n",
    "      axis.ticks = element_blank(),\n",
    "      plot.title=element_text(size=14, face=\"bold\", hjust=0.5)\n",
    "    ) +\n",
    "    theme(axis.text.x=element_blank())+\n",
    "    geom_text(aes(x = 1.6, label = scales::percent(Percentage/100, accuracy = .1)), position = position_stack(vjust = .5)) \n",
    "\n",
    "\n",
    "# ggsave(paste0(\"stat/\", ID, \"_Alu_L1_in_head_tail_pie.pdf\"), width = 5, height = 5)\n",
    "ggsave(paste0(\"stat/Fig4f.ref_Alu_L1_in_genomic_feature_pct.pdf\"), width = 5, height = 5)\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# the genomic distribution expressed TEs\n",
    "load(\"stat/TE_location.RData\")\n",
    "load(\"stat/MDAMB231_vs_MCF10A_TE_res.RData\") # TE.norlCount\n",
    "expressedTEs <- rownames(res)\n",
    "load(\"stat/HCT116_vs_NCM460_TE_res.RData\")\n",
    "expressedTEs <- c(expressedTEs, rownames(res))\n",
    "expressedTEs <- unique(expressedTEs)\n",
    "\n",
    "\n",
    "gf <- TE_location[expressedTEs, c(\"CDS\", \"intron\", \"intergenic\", \"UTR5\", \"UTR3\")]\n",
    "rs <- rowSums(gf)\n",
    "gf <- gf/rs\n",
    "cs <- colSums(gf, na.rm = TRUE)\n",
    "cs <- data.frame(Feature=names(cs), Freq=cs)\n",
    "cs$Percentage <- cs$Freq/sum(cs$Freq)*100\n",
    "\n",
    "\n",
    "\n",
    "cols <- c(UTR3=\"#EE9248\", UTR5=\"#EDC26D\", intergenic=\"#8C9FCA\", CDS=\"#5CB1DD\", intron=\"#4A89D5\")\n",
    "plot <- ggplot(cs, aes(x=\"\", y=Percentage, fill=Feature))+\n",
    "    geom_bar(width = 1, stat = \"identity\") +\n",
    "    coord_polar(\"y\", start=0) +\n",
    "    scale_fill_manual(values = cols)+\n",
    "    theme_minimal() +\n",
    "    theme(\n",
    "      axis.title.x = element_blank(),\n",
    "      axis.title.y = element_blank(),\n",
    "      panel.border = element_blank(),\n",
    "      panel.grid=element_blank(),\n",
    "      axis.ticks = element_blank(),\n",
    "      plot.title=element_text(size=14, face=\"bold\", hjust=0.5)\n",
    "    ) +\n",
    "    theme(axis.text.x=element_blank())+\n",
    "    geom_text(aes(x = 1.6, label = scales::percent(Percentage/100, accuracy = .1)), position = position_stack(vjust = .5)) \n",
    "\n",
    "\n",
    "# ggsave(paste0(\"stat/\", ID, \"_Alu_L1_in_head_tail_pie.pdf\"), width = 5, height = 5)\n",
    "ggsave(paste0(\"stat/Fig4g.sup.expressed_Alu_L1_in_genomic_feature_pct.pdf\"), width = 5, height = 5)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# sgRNA count > 3 at least 3 samples\n",
    "load(\"stat/TE.quantify.count.RData\")\n",
    "\n",
    "IDs <- Sample_info$Samples[Sample_info$SeqType %in% c(\"sgRNA\")]\n",
    "sub.count.tb <- count.tb[, IDs]\n",
    "sub.count.tb <- sub.count.tb[apply(sub.count.tb, 1, function(x){ (sum(x>=3))>=3 }),]\n",
    "dim(sub.count.tb) # 121092\n",
    "sub.count.tb <- round(sub.count.tb)\n",
    "\n",
    "# run DESeq2\n",
    "colInfo <- Sample_info[IDs, c(\"CellLine\"), drop=F]\n",
    "dds <- DESeqDataSetFromMatrix(sub.count.tb, \n",
    "                                colInfo, \n",
    "                                ~CellLine)\n",
    "Object <- DESeq(dds\n",
    "                # parallel = T, BPPARAM=MulticoreParam(10)\n",
    "                )\n",
    "TE.norlCount <- counts(Object, normalized=T)\n",
    "\n",
    "save(Object, file=\"stat/TE.count.DESeq2.Object.RData\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# sgRNA count > 1 at least one sample\n",
    "\n",
    "\n",
    "pheatmap(cor(TE.norlCount, method = \"pearson\"),\n",
    "         display_numbers=T,\n",
    "         main=\"Count correlation\",\n",
    "         # col=colorRampPalette(rev(brewer.pal(11,\"RdYlBu\")[3:10]))(21),\n",
    "         col=colorRampPalette(c(\"#4678BA\", \"#F7F7F7\", \"#EF8354\"))(100),\n",
    "         border=NA,\n",
    "         filename=\"stat/TE_count_correlation.pdf\")\n",
    "\n",
    "\n",
    "pheatmap(cor(TE.norlCount, method = \"spearman\"),\n",
    "         display_numbers=T,\n",
    "         main=\"Count correlation\",\n",
    "         # col=colorRampPalette(rev(brewer.pal(11,\"RdYlBu\")[3:10]))(21),\n",
    "         col=colorRampPalette(c(\"#4678BA\", \"#F7F7F7\", \"#EF8354\"))(100),\n",
    "         border=NA,\n",
    "         filename=\"stat/TE_count_correlation_spearman.pdf\")\n",
    "\n",
    "\n",
    "\n",
    "pheatmap(cor(norlCount, method = \"spearman\"),\n",
    "         display_numbers=T,\n",
    "         annotation_col = colInfo,\n",
    "         annotation_row = colInfo,\n",
    "         annotation_colors = anno_colors,\n",
    "         main=\"Correlation of Transcript normalized Count \",\n",
    "         # col=colorRampPalette(rev(brewer.pal(11,\"RdYlBu\")[3:10]))(21),\n",
    "         col=colorRampPalette(c(\"#4678BA\", \"#F7F7F7\", \"#EF8354\"))(100),\n",
    "         border=\"white\",\n",
    "         width=6.5,\n",
    "         height=6,\n",
    "         show_rownames=F,\n",
    "         show_colnames=F,\n",
    "         filename=\"stat/Fig3A.MDAMB231_MCF10A_FeatureCount_spearman_correlation.pdf\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## MDAMB231 vs MCF10A"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "setwd(\"/NAS/wg_looking/gRNA_ONT\")\n",
    "load(\"stat/HCT116_NCM460_diff.Data\")\n",
    "load(\"stat/MDAMB231_MCF10A_diff.Data\")\n",
    "load(\"stat/assembly_preData.RData\")\n",
    "load(\"stat/TE.quantify.count.RData\")\n",
    "\n",
    "\n",
    "# load Sample info\n",
    "Sample_info <- read.table(\"stat/Sample_info2.txt\", header = T, stringsAsFactors = F)\n",
    "rownames(Sample_info) <- Sample_info$Samples\n",
    "\n",
    "SampleIDs <- Sample_info$Samples[Sample_info$CellLine %in% c(\"MDAMB231\", \"MCF10A\") &\n",
    "                                Sample_info$SeqType %in% c(\"sgRNA\")]\n",
    "\n",
    "## new transcript with TE\n",
    "trans_TE <- MDAMB231_Diff[[2]]\n",
    "trans_TE$TE_name <- paste0(paste(trans_TE$TE_chr, trans_TE$TE_start, trans_TE$TE_end, sep=\"_\"), \":\", trans_TE$repeat_id)\n",
    "trans_TE <- trans_TE[trans_TE$classification %in% c(\"intergenic\", \"intronic\"),]\n",
    "trans_TE2 <- MCF10A_Diff[[2]]\n",
    "trans_TE2$TE_name <- paste0(paste(trans_TE2$TE_chr, \n",
    "                                  trans_TE2$TE_start, \n",
    "                                  trans_TE2$TE_end, sep=\"_\"), \":\", trans_TE2$repeat_id)\n",
    "trans_TE2 <- trans_TE2[trans_TE2$classification %in% c(\"intergenic\", \"intronic\"),]\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# sgRNA count > 3 at least 3 samples\n",
    "library(DESeq2)\n",
    "\n",
    "SampleIDs <- Sample_info$Samples[Sample_info$CellLine %in% c(\"MDAMB231\", \"MCF10A\") &\n",
    "                                Sample_info$SeqType %in% c(\"sgRNA\")]\n",
    "sub.count.tb <- count.tb[, SampleIDs]\n",
    "sub.count.tb <- sub.count.tb[apply(sub.count.tb, 1, function(x){ (sum(x>=3))>=3 }),]\n",
    "dim(sub.count.tb) # 65631\n",
    "sub.count.tb <- round(sub.count.tb)\n",
    "\n",
    "# run DESeq2\n",
    "colInfo <- Sample_info[SampleIDs, c(\"CellLine\"), drop=F]\n",
    "dds <- DESeqDataSetFromMatrix(sub.count.tb, \n",
    "                                colInfo, \n",
    "                                ~CellLine)\n",
    "Object <- DESeq(dds\n",
    "                # parallel = T, BPPARAM=MulticoreParam(10)\n",
    "                )\n",
    "TE.norlCount <- counts(Object, normalized=T)\n",
    "\n",
    "res <- results(Object, contrast=c(\"CellLine\", \"MDAMB231\", \"MCF10A\"))\n",
    "# res <- res[complete.cases(res),]\n",
    "save(TE.norlCount, res, file=\"stat/MDAMB231_vs_MCF10A_TE_res.RData\")\n",
    "\n",
    "dif_res <- res\n",
    "dif_res <- res[res$padj<0.05 & abs(res$log2FoldChange)>1, ]\n",
    "table(dif_res$log2FoldChange>1)\n",
    "\n",
    "\n",
    "dif_res2 <- dif_res[rownames(dif_res) %in% trans_TE$TE_name |\n",
    "       rownames(dif_res) %in% trans_TE2$TE_name,]\n",
    "table(dif_res2$log2FoldChange>1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "load(\"stat/TE_location.RData\")\n",
    "load(\"stat/MDAMB231_vs_MCF10A_TE_res.RData\")\n",
    "\n",
    "\n",
    "out_res <- data.frame(TE.norlCount, res[, c(\"baseMean\", \"log2FoldChange\", \"pvalue\", \"padj\")])\n",
    "out_res$dif <- \"ns\"\n",
    "out_res$dif[res$padj<0.05 & res$log2FoldChange>1] <- \"Up\"\n",
    "out_res$dif[res$padj<0.05 & res$log2FoldChange < -1] <- \"Down\"\n",
    "table(out_res$dif)\n",
    "out_res <- out_res[order(out_res$dif, decreasing=T),]\n",
    "out_res <- data.frame(gene_id=rownames(out_res), out_res)\n",
    "colnames(out_res)[1] <- \"TE_name\"\n",
    "\n",
    "# out_res <- data.frame(out_res, TE_location[rownames(out_res), c(\"sub_repeat\", \"exon\", \"intron\", \"intergenic\",\n",
    "#                                                                 \"gene_id\", \"gene_name\")])\n",
    "out_res <- merge(out_res, TE_location, by=\"TE_name\")\n",
    "out_res$gene_id <- as.character(out_res$gene_id)\n",
    "\n",
    "# out_res <- gene_res[transcript_res$padj<0.05 & abs(transcript_res$log2FoldChange)>2 & transcript_res$baseMean>10, ]\n",
    "# write.table(out_res, \"stat/MDAMB231_vs_MCF10A_differential_genes.txt\", row.names = F, col.names = T, quote = F, sep=\"\\t\")\n",
    "write.table(out_res, \"stat/MDAMB231_vs_MCF10A_all_TEs.txt\", row.names = F, col.names = T, quote = F, sep=\"\\t\")\n",
    "\n",
    "\n",
    "\n",
    "# TE.norlCount <- counts(Object, normalized=T)\n",
    "# n <- data.frame(TE_name=rownames(TE.norlCount), \n",
    "#                 TE_location[rownames(TE.norlCount), c(\"exon\", \"intron\", \"intergenic\")], \n",
    "#                 TE.norlCount)\n",
    "\n",
    "# write.table(n, \"stat/MDAMB231_MCF10A.norlCount.txt\", row.names = F, col.names = T, sep=\"\\t\", quote=F)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "load(\"stat/MDAMB231_vs_MCF10A_TE_res.RData\")\n",
    "\n",
    "SampleIDs <- Sample_info$Samples[Sample_info$CellLine %in% c(\"MDAMB231\", \"MCF10A\") &\n",
    "                                Sample_info$SeqType %in% c(\"sgRNA\")]\n",
    "## plot correlation\n",
    "anno_colors <- list(CellLine=c(MDAMB231=\"#87B07A\", MCF10A=\"#F4B540\"))\n",
    "colInfo <- Sample_info[SampleIDs, c(\"CellLine\"), drop=F]\n",
    "\n",
    "pheatmap(cor(TE.norlCount, method = \"spearman\"),\n",
    "         display_numbers=T,\n",
    "         annotation_col = colInfo,\n",
    "         annotation_row = colInfo,\n",
    "         annotation_colors = anno_colors,\n",
    "         main=\"Correlation of TE normalized Count \",\n",
    "         # col=colorRampPalette(rev(brewer.pal(11,\"RdYlBu\")[3:10]))(21),\n",
    "         col=colorRampPalette(c(\"#4678BA\", \"#F7F7F7\", \"#EF8354\"))(100),\n",
    "         border=\"white\",\n",
    "         width=6.5,\n",
    "         height=6,\n",
    "         show_rownames=F,\n",
    "         show_colnames=F,\n",
    "         filename=\"stat/MDAMB231_MCF10A_TE_count_spearman_correlation.pdf\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# volcano plot\n",
    "res$dif <- \"ns\"\n",
    "res$dif[res$log2FoldChange>1 & res$padj<0.05] <- \"Up\"\n",
    "res$dif[res$log2FoldChange< -1 & res$padj<0.05] <- \"Down\"\n",
    "\n",
    "res <- data.frame(res)\n",
    "\n",
    "ggplot(res, aes(x=log2FoldChange, y=-log10(padj), color=dif)) +\n",
    "    geom_point(size=0.3, alpha=0.4) +\n",
    "    scale_color_manual(values = c(MCF10A_hyper=\"#327FFF\", MDAMB231_hyper=\"#FD7E29\", ns=\"#AFAFAF\")) +\n",
    "    scale_y_log10() +\n",
    "    theme_bw() +\n",
    "    theme(\n",
    "    panel.grid=element_blank(),\n",
    "    axis.text.x = element_text(angle = 0, color = \"black\", \n",
    "                               size = 8, hjust = 0.5, vjust = 0.5),\n",
    "    plot.title=element_text(size=14, face=\"bold\")\n",
    "    ) + \n",
    "    ggtitle(\"\")\n",
    "\n",
    "\n",
    "ggsave(\"stat/MDAMB231_VS_MCF10A_TE_volcano_plot.pdf\",\n",
    "       width = 5, height = 3.2)\n",
    "\n",
    "\n",
    "\n",
    "# pie\n",
    "pp <- data.frame(table(res$dif))\n",
    "\n",
    "p <- pp %>%\n",
    "    mutate(Percentage = Freq / sum(Freq)*100)\n",
    "  \n",
    "  \n",
    "plot <- ggplot(p, aes(x=\"\", y=Percentage, fill=Var1))+\n",
    "    geom_bar(width = 1, stat = \"identity\") +\n",
    "    coord_polar(\"y\", start=0) +\n",
    "    scale_fill_manual(values = c(MCF10A_hyper=\"#327FFF\", MDAMB231_hyper=\"#FD7E29\", ns=\"#AFAFAF\"))+\n",
    "    geom_text(aes(x = 1.6, label = paste0(Freq, \"(\",scales::percent(Percentage/100, accuracy = .1), \")\")), \n",
    "              position = position_stack(vjust = .5)) +\n",
    "    # facet_wrap(~Method)\n",
    "    ggtitle(\"\") +\n",
    "    theme_minimal() +\n",
    "    theme(\n",
    "      axis.title.x = element_blank(),\n",
    "      axis.title.y = element_blank(),\n",
    "      panel.border = element_blank(),\n",
    "      panel.grid=element_blank(),\n",
    "      axis.ticks = element_blank(),\n",
    "      plot.title=element_text(size=14, face=\"bold\", hjust=0.5),\n",
    "      axis.text.x=element_blank()\n",
    "    ) \n",
    "\n",
    "\n",
    "ggsave(\"stat/MDAMB231_VS_MCF10A_TE_dif_pie.pdf\",\n",
    "       width = 6, height = 3)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 差异的TE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "library(ggpubr)\n",
    "load(\"stat/TE_location.RData\")\n",
    "load(\"stat/MDAMB231_vs_MCF10A_TE_res.RData\")\n",
    "\n",
    "\n",
    "sub.TE.norlCount <- TE.norlCount[rownames(res)[abs(res$log2FoldChange)>1 & res$padj<0.05], ]\n",
    "pp <- melt(sub.TE.norlCount)\n",
    "colnames(pp) <- c(\"TE_name\", \"ID\", \"norlCount\")\n",
    "pp$stat <- \"Up\"\n",
    "pp$stat[pp$TE_name %in% rownames(res)[res$log2FoldChange< -1 & res$padj<0.05]] <- \"Down\"\n",
    "pp$repeat_id <- tstrsplit(pp$TE_name, \":\")[[2]]\n",
    "pp$sub_repeat <- substring(pp$repeat_id, 1, 4)\n",
    "pp$sub_repeat[grepl(\"L1P\", pp$sub_repeat)] <- \"L1P\"\n",
    "pp$CellLine <- tstrsplit(pp$ID, \"_\")[[2]]\n",
    "\n",
    "location <- TE_location[,c(\"intergenic\", \"exon\", \"intron\")]\n",
    "location <- melt(t(location))\n",
    "location <- location[location$value==1, 1:2]\n",
    "colnames(location) <- c(\"genomic_feature\", \"TE_name\")\n",
    "dim(location) # 1265136\n",
    "\n",
    "\n",
    "p <- merge(pp, location, by=\"TE_name\")\n",
    "\n",
    "\n",
    "\n",
    "## jitter\n",
    "p1 <- ggplot(p, aes(x=sub_repeat, y=log2(norlCount+1), color=CellLine, fill=CellLine)) +\n",
    "    # geom_violin(alpha=0.4) + \n",
    "    geom_point(alpha=0.3, size=0.5, position=position_jitterdodge(dodge.width=0.95)) +\n",
    "    # geom_jitter(position = position_jitter(width = .2), alpha = 0.5, size=0.5)  +\n",
    "    stat_summary(fun= median, geom = \"point\",position = position_dodge(width = 0.9),\n",
    "               shape = 23, size = 2, color = \"red\") +\n",
    "    # geom_point(alpha=0.1) +\n",
    "    # geom_jitter(width = 0.25, alpha=0.1) +\n",
    "    theme_bw() +\n",
    "    theme(\n",
    "      panel.grid=element_blank(),\n",
    "      axis.text.x = element_text(angle = 0, color = \"black\", \n",
    "                                 size = 8, hjust = 0.5, vjust=0.5),\n",
    "      plot.title=element_text(size=14, face=\"bold\", hjust=0.5)\n",
    "    ) +\n",
    "    facet_wrap(~genomic_feature, nrow = 1, scale=\"free\") +\n",
    "    scale_color_manual(values = c(\"#EF8354\", \"#4678BA\")) +\n",
    "    scale_fill_manual(values = c(\"#EF8354\", \"#4678BA\")) +\n",
    "    ggtitle(\"\") + \n",
    "    stat_compare_means(aes(group = CellLine), method = \"wilcox.test\", label = \"p.signif\", size=2)\n",
    "\n",
    "\n",
    "ggsave(paste0(\"stat/Fig3.MDAMB231_vs_MCF10A_TE_dif_repeat_genomic_feature_normalCount.pdf\"),\n",
    "         width = 14, height = 3)\n",
    "\n",
    "\n",
    "# bar\n",
    "pp <- pp[order(pp$TE_name),]\n",
    "pp2 <- pp[!(duplicated(pp$TE_name)),]\n",
    "pp2 <- data.frame(table(pp2$stat, pp2$sub_repeat))\n",
    "colnames(pp2) <- c(\"stat\", \"sub_repeat\", \"Freq\")\n",
    "p2 <- ggplot(pp2, aes(x=sub_repeat, y=Freq,  fill=stat)) +\n",
    "    geom_bar(stat=\"identity\", position=position_dodge(width=0.9)) + \n",
    "    geom_text(aes(label=Freq), position=position_dodge(width=0.9), vjust=-0.25) +\n",
    "    theme_bw() +\n",
    "    theme(\n",
    "      panel.grid=element_blank(),\n",
    "      axis.text.x = element_text(angle = 0, color = \"black\", \n",
    "                                 size = 8, hjust = 0.5, vjust=0.5),\n",
    "      plot.title=element_text(size=14, face=\"bold\")\n",
    "    ) +\n",
    "    # facet_wrap(~CellLine, nrow = 1, scales=\"free\") +\n",
    "    scale_fill_manual(values = c(\"#EF8354\", \"#4678BA\")) +\n",
    "    # ylim(0,3200) +\n",
    "    ggtitle(\"the count of differential expression TEs\") \n",
    "   \n",
    "  \n",
    "\n",
    "ggsave(paste0(\"stat/MDAMB231_vs_MCF10A_TE_dif_count_bar.pdf\"),\n",
    "         width = 7, height = 4)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 差异TE相关的基因的注释"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 所有差异TE相关的基因的注释\n",
    "library(gprofiler2)\n",
    "library(bedtoolsr)\n",
    "library(data.table)\n",
    "Sys.which(\"python\")\n",
    "Sys.setenv(RETICULATE_PYTHON = \"/home/wg_looking/anaconda3/bin/python\")\n",
    "library(reticulate)\n",
    "source_python(\"/NAS/wg_looking/gRNA_ONT/code/NCBI_gene_summary.py\")\n",
    "load(\"stat/TE_location.RData\")\n",
    "load(\"stat/hg38.gtf.RData\")\n",
    "load(\"stat/MDAMB231_vs_MCF10A_res.RData\")\n",
    "\n",
    "transcript_bed <- hg38.gtf[hg38.gtf$type==\"transcript\", c(\"seqnames\", \"start\",\"end\", \"gene_id\", \"transcript_id\", \"gene_name\")]\n",
    "TE_location_transcript <- bt.intersect(a=TE_location, transcript_bed, wa=T, wb=T)\n",
    "colnames(TE_location_transcript) <- c(colnames(TE_location), colnames(exon_bed))\n",
    "sub <- TE_location_transcript[TE_location_transcript$Geneid %in% rownames(dif_res),]\n",
    "\n",
    "\n",
    "genes <- setNames(as.character(sub$gene_id), as.character(sub$gene_id))\n",
    "genes <- sort(genes)\n",
    "genes <- genes[!duplicated(genes)]\n",
    "# gene_function.list <- list()\n",
    "for(i in 1:length(genes)){\n",
    "    gene_id <- genes[i]\n",
    "    Entrez_ID = ENSEMBL_ID_Entrez_ID_all[gene_id, \"ENTREZID\"]\n",
    "    gf = NCBI_gene_summary(as.character(Entrez_ID)) \n",
    "    gene_function.list[gene_id] <- gf\n",
    "}\n",
    "\n",
    "save(gene_function.list, file=\"stat/MDAMB231_vs_MCF10A_dif_TE_related_gene_annotation.RData\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### correlation logFC vs new trans count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(ggpubr)\n",
    "load(\"stat/TE_location.RData\")\n",
    "load(\"stat/MDAMB231_vs_MCF10A_TE_res.RData\")\n",
    "load(\"stat/HCT116_NCM460_diff.Data\")\n",
    "load(\"stat/MDAMB231_MCF10A_diff.Data\")\n",
    "BC_genes <- read.table(\"stat/MDAMB231_vs_MCF10A_all_genes.txt\", header=T)\n",
    "CRC_genes <- read.table(\"stat/HCT116_vs_NCM460_all_genes.txt\", header=T)\n",
    "Diff.list <- list(MDAMB231=MDAMB231_Diff, MCF10A=MCF10A_Diff,\n",
    "                  HCT116=HCT116_Diff, NCM460=NCM460_Diff)\n",
    "\n",
    "\n",
    "## for BC\n",
    "dif_res <- res[res$padj<0.05 & abs(res$log2FoldChange)>1, ]\n",
    "dif_res$dif <- \"Up\"\n",
    "dif_res$dif[dif_res$log2FoldChange< -1] <- \"Down\"\n",
    "dif_TE <- data.frame(TE_name=rownames(dif_res), dif=dif_res$dif )\n",
    "\n",
    "sub.TE_location <- merge(dif_TE, TE_location[, c(\"TE_name\", \"gene_id\")], by=\"TE_name\")\n",
    "relate_genes <- sub.TE_location$gene_id\n",
    "relate_genes <- lapply(relate_genes, function(x){strsplit(x, \",\")[[1]]})\n",
    "names(relate_genes) <- sub.TE_location$TE_name\n",
    "TE.genes <- reshape2::melt(relate_genes)\n",
    "colnames(TE.genes) <- c(\"gene_id\", \"TE_name\")\n",
    "TE.genes <- TE.genes[complete.cases(TE.genes), ]\n",
    "df <- merge(TE.genes, BC_genes[, c(\"gene_id\", \"log2FoldChange\", \"dif\")], by=\"gene_id\")\n",
    "colnames(df)[4] <- \"gene_dif\"\n",
    "df <- merge(df, dif_TE, by=\"TE_name\")\n",
    "\n",
    "nc <- data.frame(table(MDAMB231_Diff[[1]]$ref_gene_id))\n",
    "rownames(nc) <- nc$Var1\n",
    "df$MDAMB231_gene_include_newTrans <- nc[as.character(df$gene_id), \"Freq\"]\n",
    "df$MDAMB231_gene_include_newTrans[is.na(df$MDAMB231_gene_include_newTrans)] <- 0\n",
    "\n",
    "nc <- data.frame(table(MCF10A_Diff[[1]]$ref_gene_id))\n",
    "rownames(nc) <- nc$Var1\n",
    "df$MCF10A_gene_include_newTrans <- nc[as.character(df$gene_id), \"Freq\"]\n",
    "df$MCF10A_gene_include_newTrans[is.na(df$MCF10A_gene_include_newTrans)] <- 0\n",
    "df$newTrans_count_diff <- df$MDAMB231_gene_include_newTrans-df$MCF10A_gene_include_newTrans\n",
    "BC_df <- df\n",
    "\n",
    "pp <- BC_df\n",
    "cor <- cor(pp$log2FoldChange, pp$newTrans_count_diff, method=\"spearman\") # 0.5725984\n",
    "\n",
    "ggplot(pp, aes(x=log2FoldChange, y=newTrans_count_diff, shape=dif, color=gene_dif)) +\n",
    "    geom_point() + \n",
    "    theme_bw() +\n",
    "    theme(\n",
    "      panel.grid=element_blank(),\n",
    "      axis.text.x = element_text(angle = 0, color = \"black\", \n",
    "                                 size = 8, hjust = 0.5, vjust=0.5),\n",
    "      plot.title=element_text(size=14, face=\"bold\", hjust=0.5)\n",
    "    ) +\n",
    "    scale_color_manual(values = c(Up=\"#FC8D62\", ns=\"#D3D3D3\", Down=\"#66C2A5\")) +\n",
    "    # facet_wrap(~ID, scales=\"free\", ncol=1) +\n",
    "    # facet_grid(scales=\"free\", rows=vars(ID)) +\n",
    "    # scale_x_log10() +\n",
    "    # scale_x_continuous(expand = c(0,0), limits=c(-10,10)) +\n",
    "    # scale_y_continuous(expand = c(0,0)) +\n",
    "    geom_text(aes(x = -8, y=10, label = paste0(\"spearman cor: \", round(cor, 3))), color=\"black\") \n",
    "    ggtitle(\"\") \n",
    "  \n",
    "ggsave(\"stat/Fig6S.BC_TE_genes_logFC_newTrans.pdf\", width=6, height=5)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### the location of differential TE "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in readChar(con, 5L, useBytes = TRUE):\n",
      "\"无法打开压缩文件'stat/TE_location.RData'，可能是因为'No such file or directory'\"\n"
     ]
    },
    {
     "ename": "ERROR",
     "evalue": "Error in readChar(con, 5L, useBytes = TRUE): 无法打开链结\n",
     "output_type": "error",
     "traceback": [
      "Error in readChar(con, 5L, useBytes = TRUE): 无法打开链结\nTraceback:\n",
      "1. load(\"stat/TE_location.RData\")",
      "2. readChar(con, 5L, useBytes = TRUE)"
     ]
    }
   ],
   "source": [
    "## the location of differential TE \n",
    "load(\"stat/TE_location.RData\")\n",
    "load(\"stat/MDAMB231_vs_MCF10A_TE_res.RData\")\n",
    "\n",
    "res <- results(Object, contrast=c(\"CellLine\", \"MDAMB231\", \"MCF10A\"))\n",
    "res <- res[complete.cases(res),]\n",
    "res$dif <- \"ns\"\n",
    "res$dif[res$log2FoldChange>1 & res$padj<0.05] <- \"Up\"\n",
    "res$dif[res$log2FoldChange< -1 & res$padj<0.05] <- \"Down\"\n",
    "\n",
    "TE_location_modif <- TE_location[, c(\"exon\", \"intron\", \"intergenic\")]\n",
    "TE_location_modif <- TE_location_modif/rowSums(TE_location_modif)\n",
    "res_ex <- data.frame(res, TE_location_modif[rownames(res), c(\"exon\", \"intron\", \"intergenic\")])\n",
    "location_count <- colSums(res_ex[,  c(\"exon\", \"intron\", \"intergenic\")])\n",
    "location_count_up <- colSums(res_ex[res_ex$log2FoldChange> 1 & res_ex$padj<0.05,  c(\"exon\", \"intron\", \"intergenic\")])\n",
    "location_count_down <- colSums(res_ex[res_ex$log2FoldChange< -1 & res_ex$padj<0.05,  c(\"exon\", \"intron\", \"intergenic\")])\n",
    "\n",
    "\n",
    "pp <- data.frame(background=location_count, up=location_count_up, down=location_count_down)\n",
    "pp <- melt(t(pp))\n",
    "colnames(pp) <- c(\"stat\", \"Feature\", \"Freq\")\n",
    "pp <- ddply(pp, .(stat), mutate, Percentage=Freq/sum(Freq)*100)\n",
    "\n",
    "\n",
    "ggplot(pp, aes(x=stat, y=Percentage,  fill=Feature)) +\n",
    "    geom_bar(stat=\"identity\") + \n",
    "    theme_bw() +\n",
    "    theme(\n",
    "      panel.grid=element_blank(),\n",
    "      axis.text.x = element_text(angle = 0, color = \"black\", \n",
    "                                 size = 8, hjust = 0.5, vjust=0.5),\n",
    "      plot.title=element_text(size=14, face=\"bold\")\n",
    "    ) +\n",
    "    # facet_wrap(~CellLine, nrow = 1, scales=\"free\") +\n",
    "    scale_fill_manual(values = c(\"#8683B9\", \"#5655FE\", \"#39C6AA\")) +\n",
    "    coord_flip() +\n",
    "    # ylim(0,3200) +\n",
    "    ggtitle(\"the location of differential expression TEs\") \n",
    "   \n",
    "\n",
    "ggsave(paste0(\"stat/Fig3.MDAMB231_vs_MCF10A_TE_dif_location_percentage.pdf\"),\n",
    "         width = 5, height = 3)\n",
    "    \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# chisq.test for Feature Distribution\n",
    "library(gtools)\n",
    "\n",
    "pp.list <- list(background=location_count, up=location_count_up, down=location_count_down)\n",
    "comb <- combinations(length(names(pp.list)), 2, names(pp.list))\n",
    "\n",
    "chisq_test_res_list <- list()\n",
    "for(i in 1:dim(comb)[1]){\n",
    "  type1 <- comb[i, 1]\n",
    "  type2 <- comb[i, 2]\n",
    "  comb_type <- paste(type1, type2, sep = \"-\")\n",
    "  for(feature in c(\"exon\", \"intron\", \"intergenic\")){\n",
    "      testInfo1 <- pp.list[[type1]]\n",
    "      testInfo2 <- pp.list[[type2]]\n",
    "      testInfo1 <- c(testInfo1[feature], Other=sum(testInfo1[!(names(testInfo1) %in% feature)]))\n",
    "      testInfo2 <- c(testInfo2[feature], Other=sum(testInfo2[!(names(testInfo2) %in% feature)]))\n",
    "      x <- matrix(c(testInfo1, testInfo2), ncol = 2)\n",
    "      pvalue <- chisq.test(x)$p.value\n",
    "      log2FC <- log2( (testInfo1[1]/sum(testInfo1))/(testInfo2[1]/sum(testInfo2)) )\n",
    "      chisq_test_res_list[[comb_type]][[feature]] <- data.frame(VS=comb_type ,\n",
    "                                                                genomic_feature=feature,\n",
    "                                                                pvalue=pvalue,\n",
    "                                                                log2FC=log2FC                                                              \n",
    "                                                               )\n",
    "  }\n",
    "}\n",
    "\n",
    "rbindlist(lapply(chisq_test_res_list, rbindlist))\n",
    "\n",
    "#                VS genomic_feature       pvalue     log2FC\n",
    "# 1: background-down            exon 9.176394e-13 -0.6360895\n",
    "# 2: background-down          intron 1.179018e-43  0.1881517\n",
    "# 3: background-down      intergenic 1.453233e-30 -0.8790435\n",
    "# 4:   background-up            exon 0.000000e+00 -2.0624271\n",
    "# 5:   background-up          intron 0.000000e+00  0.5056756\n",
    "# 6:   background-up      intergenic 8.714962e-13 -0.4731949\n",
    "# 7:         down-up            exon 5.727681e-64 -1.4263376\n",
    "# 8:         down-up          intron 4.156872e-33  0.3175240\n",
    "# 9:         down-up      intergenic 3.398676e-05  0.4058486\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(ggplot2)\n",
    "load(\"stat/TE_location.RData\")\n",
    "load(\"stat/MDAMB231_vs_MCF10A_TE_res.RData\")\n",
    "\n",
    "res <- res[complete.cases(res),]\n",
    "res$dif <- \"ns\"\n",
    "res$dif[res$log2FoldChange>1 & res$padj<0.05] <- \"Up\"\n",
    "res$dif[res$log2FoldChange< -1 & res$padj<0.05] <- \"Down\"\n",
    "res <- data.frame(TE_name=rownames(res), res)\n",
    "\n",
    "location <- TE_location[,c(\"intergenic\", \"exon\", \"intron\")]\n",
    "location <- melt(t(location))\n",
    "location <- location[location$value==1, 1:2]\n",
    "colnames(location) <- c(\"genomic_feature\", \"TE_name\")\n",
    "dim(location) # 1265136\n",
    "\n",
    "pp <- merge(res, location, by=\"TE_name\")\n",
    "pp <- merge(pp, TE_location[, c(\"TE_name\", \"sub_repeat\")], by=\"TE_name\")\n",
    "pp <- data.frame(table(pp$dif, pp$genomic_feature, pp$sub_repeat))\n",
    "colnames(pp) <- c(\"stat\", \"genomic_feature\", \"sub_repeat\", \"Freq\")\n",
    "background = aggregate(Freq~genomic_feature+sub_repeat, data=pp, sum)\n",
    "background = data.frame(stat=\"background\", background)\n",
    "pp <- rbind(pp, background)\n",
    "pp <- pp[pp$stat!=\"ns\", ]\n",
    "\n",
    "p <- ddply(pp, .(stat, genomic_feature), mutate, Percentage=Freq/sum(Freq)*100)\n",
    "ggplot(p, aes(x=sub_repeat, y=Percentage,  color=stat, shape=stat)) +\n",
    "    # geom_point() + \n",
    "    geom_point(size=2, position=position_jitterdodge(dodge.width=0.5)) +\n",
    "    # geom_jitter(position = position_jitter(width = .2), alpha = 0.5, size=0.5)  +\n",
    "    theme_bw() +\n",
    "    theme(\n",
    "      panel.grid=element_blank(),\n",
    "      axis.text.x = element_text(angle = 0, color = \"black\", \n",
    "                                 size = 8, hjust = 0.5, vjust=0.5),\n",
    "      plot.title=element_text(size=14, face=\"bold\")\n",
    "    ) +\n",
    "    facet_wrap(~genomic_feature, nrow = 1, scales=\"free\") +\n",
    "    scale_color_manual(values = c(Down=\"#39C6AA\", Up=\"#5655FE\", background=\"#8683B9\")) +\n",
    "    # coord_flip() +\n",
    "    # ylim(0,3200) +\n",
    "    ggtitle(\"\") \n",
    "   \n",
    "\n",
    "ggsave(paste0(\"stat/Fig4S.MDAMB231_vs_MCF10A_TE_location_repeat_percentage.pdf\"),\n",
    "         width = 12, height = 3)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pp = aggregate(Freq~stat+sub_repeat, data=pp, sum)\n",
    "\n",
    "p <- ddply(pp, .(stat), mutate, Percentage=Freq/sum(Freq)*100)\n",
    "\n",
    "ggplot(p, aes(x=sub_repeat, y=Percentage,  color=stat, shape=stat)) +\n",
    "    # geom_point() + \n",
    "    geom_point(size=2, position=position_jitterdodge(dodge.width=0.5)) +\n",
    "    # geom_jitter(position = position_jitter(width = .2), alpha = 0.5, size=0.5)  +\n",
    "    theme_bw() +\n",
    "    theme(\n",
    "      panel.grid=element_blank(),\n",
    "      axis.text.x = element_text(angle = 0, color = \"black\", \n",
    "                                 size = 8, hjust = 0.5, vjust=0.5),\n",
    "      plot.title=element_text(size=14, face=\"bold\")\n",
    "    ) +\n",
    "    # facet_wrap(~genomic_feature, nrow = 1, scales=\"free\") +\n",
    "    scale_color_manual(values = c(Down=\"#39C6AA\", Up=\"#5655FE\", background=\"#8683B9\")) +\n",
    "    # coord_flip() +\n",
    "    # ylim(0,3200) +\n",
    "    ggtitle(\"\") \n",
    "   \n",
    "\n",
    "ggsave(paste0(\"stat/Fig3.MDAMB231_vs_MCF10A_TE_sub_repeat_percentage.pdf\"),\n",
    "         width = 5, height = 3)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### dif TE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(ggplot2)\n",
    "load(\"stat/TE_location.RData\")\n",
    "load(\"stat/MDAMB231_vs_MCF10A_TE_res.RData\")\n",
    "\n",
    "res <- res[complete.cases(res),]\n",
    "res$dif <- \"ns\"\n",
    "res$dif[res$log2FoldChange>1 & res$padj<0.05] <- \"Up\"\n",
    "res$dif[res$log2FoldChange< -1 & res$padj<0.05] <- \"Down\"\n",
    "res <- data.frame(TE_name=rownames(res), res)\n",
    "res$RepeatType <- tstrsplit(rownames(res), \":\")[[2]]\n",
    "pp.bg <- data.frame(table(res$RepeatType))\n",
    "colnames(pp.bg) <- c(\"RepeatType\", \"Freq\")\n",
    "pp.dif <- data.frame(table(res[res$dif %in% c(\"Down\", \"Up\"),]$RepeatType))\n",
    "colnames(pp.dif) <- c(\"RepeatType\", \"Freq\")\n",
    "\n",
    "\n",
    "chisq_test_res_list <- list()\n",
    "for(i in 1:dim(pp.dif)[1]){\n",
    "    repeat_type <- as.character(pp.dif[i,1])\n",
    "    dif <- pp.dif[i,2]\n",
    "    bg <- pp.bg[pp.bg$RepeatType %in% repeat_type, 2]\n",
    "    dif.other <- sum(pp.dif$Freq) - dif\n",
    "    bg.other <- sum(pp.bg$Freq) - bg\n",
    "    x <- matrix(c(dif, dif.other, bg, bg.other), ncol = 2)\n",
    "    # pvalue <- chisq.test(x)$p.value # 2.417162e-85(promoter) 7.399233e-64(Distal intergenic)\n",
    "     pvalue <- fisher.test(x)$p.value\n",
    "    chisq_test_res_list[[repeat_type]] <- data.frame(repeat_type=repeat_type, \n",
    "                                                     dif_pct=dif/sum(pp.dif$Freq), \n",
    "                                                     bg_pct=bg/sum(pp.bg$Freq), \n",
    "                                                     pvalue=pvalue)\n",
    "  }\n",
    "\n",
    "chisq_test_res <- rbindlist(chisq_test_res_list)\n",
    "chisq_test_res$padj <- p.adjust(chisq_test_res$pvalue, method = \"fdr\", n = length(chisq_test_res$pvalue))\n",
    "chisq_test_res$significant <- \"NO\"\n",
    "chisq_test_res$significant[chisq_test_res$padj<0.05] <- \"YES\"\n",
    "chisq_test_res$logFC <- log2(chisq_test_res$dif_pct/chisq_test_res$bg_pct)\n",
    "chisq_test_res[abs(chisq_test_res$logFC)>0.3 & chisq_test_res$padj<0.05, c(\"repeat_type\", \"padj\", \"logFC\")]\n",
    "chisq_test_res[chisq_test_res$padj<0.2, c(\"repeat_type\", \"padj\", \"logFC\")]\n",
    "\n",
    "\n",
    "\n",
    "# p <- ddply(pp, .(stat, genomic_feature), mutate, Percentage=Freq/sum(Freq)*100)\n",
    "ggplot(chisq_test_res, aes(x=dif_pct, y=bg_pct, size=-log10(padj), color=significant)) +\n",
    "    geom_point() + \n",
    "    # geom_point(size=2, position=position_jitterdodge(dodge.width=0.5)) +\n",
    "    # geom_jitter(position = position_jitter(width = .2), alpha = 0.5, size=0.5)  +\n",
    "    theme_bw() +\n",
    "    theme(\n",
    "      panel.grid=element_blank(),\n",
    "      axis.text.x = element_text(angle = 0, color = \"black\", \n",
    "                                 size = 8, hjust = 0.5, vjust=0.5),\n",
    "      plot.title=element_text(size=14, face=\"bold\")\n",
    "    ) +\n",
    "    # facet_wrap(~genomic_feature, nrow = 1, scales=\"free\") +\n",
    "    scale_color_manual(values = c(NO=\"navy\", YES=\"firebrick3\")) +\n",
    "    # coord_flip() +\n",
    "    # ylim(0,3200) +\n",
    "    scale_x_log10() +\n",
    "    scale_y_log10() +\n",
    "    ggtitle(\"\") \n",
    "   \n",
    "\n",
    "ggsave(paste0(\"stat/Fig4S.MDAMB231_vs_MCF10A_TE_repeat_chisq_test.pdf\"),\n",
    "         width = 6, height = 5)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "##### intergenic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(ggplot2)\n",
    "load(\"stat/TE_location.RData\")\n",
    "load(\"stat/MDAMB231_vs_MCF10A_TE_res.RData\")\n",
    "\n",
    "res <- res[complete.cases(res),]\n",
    "res$dif <- \"ns\"\n",
    "res$dif[res$log2FoldChange>1 & res$padj<0.05] <- \"Up\"\n",
    "res$dif[res$log2FoldChange< -1 & res$padj<0.05] <- \"Down\"\n",
    "res <- data.frame(TE_name=rownames(res), res)\n",
    "res$RepeatType <- tstrsplit(rownames(res), \":\")[[2]]\n",
    "res <- merge(res, TE_location[,c(\"TE_name\", \"intergenic\")], by=\"TE_name\")\n",
    "res.intergenic <- res[res$intergenic %in% 1, ]\n",
    "\n",
    "pp.bg <- data.frame(table(res.intergenic$RepeatType))\n",
    "colnames(pp.bg) <- c(\"RepeatType\", \"Freq\")\n",
    "\n",
    "# intergenic\n",
    "pp.dif <- data.frame(table(res.intergenic[res.intergenic$dif %in% c(\"Down\", \"Up\"),]$RepeatType))\n",
    "colnames(pp.dif) <- c(\"RepeatType\", \"Freq\")\n",
    "\n",
    "\n",
    "chisq_test_res_list <- list()\n",
    "for(i in 1:dim(pp.dif)[1]){\n",
    "    repeat_type <- as.character(pp.dif[i,1])\n",
    "    dif <- pp.dif[i,2]\n",
    "    bg <- pp.bg[pp.bg$RepeatType %in% repeat_type, 2]\n",
    "    dif.other <- sum(pp.dif$Freq) - dif\n",
    "    bg.other <- sum(pp.bg$Freq) - bg\n",
    "    x <- matrix(c(dif, dif.other, bg, bg.other), ncol = 2)\n",
    "    # pvalue <- chisq.test(x)$p.value # 2.417162e-85(promoter) 7.399233e-64(Distal intergenic)\n",
    "     pvalue <- fisher.test(x)$p.value\n",
    "    chisq_test_res_list[[repeat_type]] <- data.frame(repeat_type=repeat_type, \n",
    "                                                     dif_pct=dif/sum(pp.dif$Freq), \n",
    "                                                     bg_pct=bg/sum(pp.bg$Freq), \n",
    "                                                     pvalue=pvalue)\n",
    "  }\n",
    "\n",
    "chisq_test_res <- rbindlist(chisq_test_res_list)\n",
    "chisq_test_res$padj <- p.adjust(chisq_test_res$pvalue, method = \"fdr\", n = length(chisq_test_res$pvalue))\n",
    "chisq_test_res$logFC <- log2(chisq_test_res$dif_pct/chisq_test_res$bg_pct)\n",
    "chisq_test_res$significant <- \"NO\"\n",
    "chisq_test_res$significant[chisq_test_res$padj<0.05 & abs(chisq_test_res$logFC)>0.3] <- \"YES\"\n",
    "\n",
    "# chisq_test_res[abs(chisq_test_res$logFC)>0.3 & chisq_test_res$padj<0.05, c(\"repeat_type\", \"padj\", \"logFC\")]\n",
    "# chisq_test_res[chisq_test_res$padj<0.2, c(\"repeat_type\", \"padj\", \"logFC\")]\n",
    "\n",
    "\n",
    "\n",
    "# p <- ddply(pp, .(stat, genomic_feature), mutate, Percentage=Freq/sum(Freq)*100)\n",
    "ggplot(chisq_test_res, aes(x=dif_pct, y=bg_pct, color=significant)) +\n",
    "    geom_point(size=2.5, alpha=0.7) + \n",
    "    # geom_point(size=2, position=position_jitterdodge(dodge.width=0.5)) +\n",
    "    # geom_jitter(position = position_jitter(width = .2), alpha = 0.5, size=0.5)  +\n",
    "    theme_bw() +\n",
    "    theme(\n",
    "      panel.grid=element_blank(),\n",
    "      axis.text.x = element_text(angle = 0, color = \"black\", \n",
    "                                 size = 8, hjust = 0.5, vjust=0.5),\n",
    "      plot.title=element_text(size=14, face=\"bold\")\n",
    "    ) +\n",
    "    # facet_wrap(~genomic_feature, nrow = 1, scales=\"free\") +\n",
    "    geom_text(aes(label=repeat_type), vjust=-1.2, size=1.5, colour=\"black\") +\n",
    "    scale_color_manual(values = c(NO=\"navy\", YES=\"firebrick3\")) +\n",
    "    # coord_flip() +\n",
    "    # ylim(0,3200) +\n",
    "    scale_x_log10() +\n",
    "    scale_y_log10() +\n",
    "    ggtitle(\"the percentage of TE subfamily\") +\n",
    "    xlab(\"intergenic differential TE in BC\") +\n",
    "    ylab(\"TE detected in Breast\")\n",
    "   \n",
    "\n",
    "ggsave(paste0(\"stat/Fig5S.F.MDAMB231_vs_MCF10A_intergenic_dif_TE_repeat_chisq_test.pdf\"),\n",
    "         width = 6, height = 5)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Up TE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(ggplot2)\n",
    "load(\"stat/TE_location.RData\")\n",
    "load(\"stat/MDAMB231_vs_MCF10A_TE_res.RData\")\n",
    "\n",
    "res <- res[complete.cases(res),]\n",
    "res$dif <- \"ns\"\n",
    "res$dif[res$log2FoldChange>1 & res$padj<0.05] <- \"Up\"\n",
    "res$dif[res$log2FoldChange< -1 & res$padj<0.05] <- \"Down\"\n",
    "res <- data.frame(TE_name=rownames(res), res)\n",
    "res$RepeatType <- tstrsplit(rownames(res), \":\")[[2]]\n",
    "pp.bg <- data.frame(table(res$RepeatType))\n",
    "colnames(pp.bg) <- c(\"RepeatType\", \"Freq\")\n",
    "pp.dif <- data.frame(table(res[res$dif %in% c(\"Up\"),]$RepeatType))\n",
    "colnames(pp.dif) <- c(\"RepeatType\", \"Freq\")\n",
    "\n",
    "\n",
    "chisq_test_res_list <- list()\n",
    "for(i in 1:dim(pp.dif)[1]){\n",
    "    repeat_type <- as.character(pp.dif[i,1])\n",
    "    dif <- pp.dif[i,2]\n",
    "    bg <- pp.bg[pp.bg$RepeatType %in% repeat_type, 2]\n",
    "    dif.other <- sum(pp.dif$Freq) - dif\n",
    "    bg.other <- sum(pp.bg$Freq) - bg\n",
    "    x <- matrix(c(dif, dif.other, bg, bg.other), ncol = 2)\n",
    "    # pvalue <- chisq.test(x)$p.value # 2.417162e-85(promoter) 7.399233e-64(Distal intergenic)\n",
    "    pvalue <- fisher.test(x)$p.value\n",
    "    chisq_test_res_list[[repeat_type]] <- data.frame(repeat_type=repeat_type, \n",
    "                                                     dif_pct=dif/sum(pp.dif$Freq), \n",
    "                                                     bg_pct=bg/sum(pp.bg$Freq), \n",
    "                                                     pvalue=pvalue)\n",
    "  }\n",
    "\n",
    "chisq_test_res <- rbindlist(chisq_test_res_list)\n",
    "chisq_test_res$padj <- p.adjust(chisq_test_res$pvalue, method = \"fdr\", n = length(chisq_test_res$pvalue))\n",
    "chisq_test_res$significant <- \"NO\"\n",
    "chisq_test_res$significant[chisq_test_res$padj<0.05] <- \"YES\"\n",
    "chisq_test_res$logFC <- log2(chisq_test_res$dif_pct/chisq_test_res$bg_pct)\n",
    "# chisq_test_res[abs(chisq_test_res$logFC)>0.3 & chisq_test_res$padj<0.05, c(\"repeat_type\", \"padj\", \"logFC\")]\n",
    "# chisq_test_res[chisq_test_res$padj<0.2, c(\"repeat_type\", \"padj\", \"logFC\")]\n",
    "\n",
    "\n",
    "\n",
    "ggplot(chisq_test_res, aes(x=dif_pct, y=bg_pct, color=significant)) +\n",
    "    geom_point() + \n",
    "    # geom_point(size=2, position=position_jitterdodge(dodge.width=0.5)) +\n",
    "    # geom_jitter(position = position_jitter(width = .2), alpha = 0.5, size=0.5)  +\n",
    "    theme_bw() +\n",
    "    theme(\n",
    "      panel.grid=element_blank(),\n",
    "      axis.text.x = element_text(angle = 0, color = \"black\", \n",
    "                                 size = 8, hjust = 0.5, vjust=0.5),\n",
    "      plot.title=element_text(size=14, face=\"bold\")\n",
    "    ) +\n",
    "    # facet_wrap(~genomic_feature, nrow = 1, scales=\"free\") +\n",
    "    scale_color_manual(values = c(NO=\"navy\", YES=\"firebrick3\")) +\n",
    "    geom_text(aes(label=repeat_type), vjust=-1.2, size=1.5, colour=\"black\") +\n",
    "    # coord_flip() +\n",
    "    # ylim(0,3200) +\n",
    "    scale_x_log10() +\n",
    "    scale_y_log10() +\n",
    "    ggtitle(\"the percentage of TE subfamily\") +\n",
    "    xlab(\"TE over-expressed in BC\") +\n",
    "    ylab(\"TE detected in Breast\")\n",
    "   \n",
    "\n",
    "ggsave(paste0(\"stat/Fig3C.MDAMB231_vs_MCF10A_Up_TE_repeat_fisher_test.pdf\"),\n",
    "         width = 6, height = 5)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### located in exon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "##  differential TE located in exon\n",
    "library(gprofiler2)\n",
    "library(bedtoolsr)\n",
    "library(data.table)\n",
    "library(pheatmap)\n",
    "library(RColorBrewer)\n",
    "load(\"stat/TE_location.RData\")\n",
    "load(\"stat/hg38.gtf.RData\")\n",
    "load(\"stat/MDAMB231_vs_MCF10A_res.RData\") # res, TE.norlCount\n",
    "load(\"stat/MDAMB231_MCF10A_diff.Data\") # MDAMB231_Diff, MCF10A_Diff,\n",
    "gene_comp <- read.table(\"stat/MDAMB231_vs_MCF10A_all_genes.txt\", header=T, row.names=1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dif_res <- res[abs(res$log2FoldChange)>1 & res$padj<0.05, ]\n",
    "dif_res$dif <- 'Up'\n",
    "dif_res$dif[dif_res$log2FoldChange < -1 ] <- 'Down'\n",
    "dif_res$TE_name <- rownames(dif_res)\n",
    "dif_res.TE <- merge(data.frame(dif_res), TE_location, by=\"TE_name\")\n",
    "dif_res.exon <- dif_res.TE[dif_res.TE$exon==1,]\n",
    "\n",
    "\n",
    "gene_ids <- lapply(dif_res.exon$gene_id, function(x){strsplit(x, \",\")[[1]]})\n",
    "names(gene_ids) <- dif_res.exon$TE_name\n",
    "gene_ids <- melt(gene_ids)\n",
    "colnames(gene_ids) <- c(\"gene_id\", \"TE_name\")\n",
    "\n",
    "df <- merge(dif_res.exon, gene_ids, by=\"TE_name\")\n",
    "df$gene_dif <- gene_comp[as.character(df$gene_id.y), \"dif\"]\n",
    "df$MDAMB231_TE_in_newTrans <- 0\n",
    "df$MDAMB231_TE_in_newTrans[df$TE_name %in% MDAMB231_Diff[[2]]$TE_name] <- 1\n",
    "df$MCF10A_TE_in_newTrans <- 0\n",
    "df$MCF10A_TE_in_newTrans[df$TE_name %in% MCF10A_Diff[[2]]$TE_name] <- 1\n",
    "\n",
    "nc <- data.frame(table(MDAMB231_Diff[[1]]$ref_gene_id))\n",
    "rownames(nc) <- nc$Var1\n",
    "df$MDAMB231_gene_include_newTrans <- nc[as.character(df$gene_id.y), \"Freq\"]\n",
    "df$MDAMB231_gene_include_newTrans[is.na(df$MDAMB231_gene_include_newTrans)] <- 0\n",
    "\n",
    "nc <- data.frame(table(MCF10A_Diff[[1]]$ref_gene_id))\n",
    "rownames(nc) <- nc$Var1\n",
    "df$MCF10A_gene_include_newTrans <- nc[as.character(df$gene_id.y), \"Freq\"]\n",
    "df$MCF10A_gene_include_newTrans[is.na(df$MCF10A_gene_include_newTrans)] <- 0\n",
    "df$newTrans_count_diff <- df$MDAMB231_gene_include_newTrans-df$MCF10A_gene_include_newTrans\n",
    "\n",
    "\n",
    "\n",
    "df <- df[order(df$dif, df$gene_dif, \n",
    "               # df$MDAMB231_gene_include_newTrans, df$MCF10A_gene_include_newTrans,\n",
    "               df$newTrans_count_diff\n",
    "               df$MDAMB231_TE_in_newTrans, df$MCF10A_TE_in_newTrans\n",
    "               ),]\n",
    "\n",
    "extract_name <- c(\"dif\", \"gene_dif\", \n",
    "                  \"newTrans_count_diff\",\n",
    "                  # \"MDAMB231_gene_include_newTrans\", \"MCF10A_gene_include_newTrans\",\n",
    "                    \"MDAMB231_TE_in_newTrans\", \"MCF10A_TE_in_newTrans\"\n",
    "                    )\n",
    "\n",
    "rowInfo <- data.frame(df[, extract_name])\n",
    "rowInfo$newTrans_count_diff <- factor(rowInfo$newTrans_count_diff)\n",
    "rownames(rowInfo) <- paste(df$TE_name, df$gene_id.y, sep=\"-\")\n",
    "colInfo <- Sample_info[SampleIDs, c(\"CellLine\"), drop=F]\n",
    "\n",
    "l <- levels(rowInfo$newTrans_count_diff)\n",
    "anno_colors <- list(CellLine=c(MDAMB231=\"#87B07A\", MCF10A=\"#F4B540\"),\n",
    "                   dif=c(Up=\"#FD7E29\", Down=\"#327FFF\"),\n",
    "                   gene_dif=c(Up=\"#5655FE\", Down=\"#39C6AA\", ns=\"#AFAFAF\"),\n",
    "                   newTrans_count_diff=setNames(rev(colorRampPalette(brewer.pal(11,\"RdGy\")[c(1:3,6,9:11)])(length(l))), \n",
    "                                       l)\n",
    "                   )\n",
    "\n",
    "\n",
    "\n",
    "sub.TE.norlCount <- TE.norlCount[df$TE_name, ]\n",
    "rownames(sub.TE.norlCount) <- rownames(rowInfo)\n",
    "pheatmap(sub.TE.norlCount,\n",
    "         # display_numbers=T,\n",
    "         scale=\"row\",\n",
    "         annotation_col = colInfo,\n",
    "         annotation_row = rowInfo,\n",
    "         annotation_colors = anno_colors,\n",
    "         main=\"differential TE expression\",\n",
    "         col=colorRampPalette(rev(brewer.pal(11,\"RdBu\")))(21),\n",
    "         # col=colorRampPalette(c(\"#4678BA\", \"#F7F7F7\", \"#EF8354\"))(11),\n",
    "         border=NA,\n",
    "         show_rownames=F,\n",
    "         show_colnames=F,\n",
    "         cluster_row=F,\n",
    "         width=7,\n",
    "         height=8,\n",
    "         filename=\"stat/MDAMB231_vs_MCF10A_TE_dif_exon_heatmap2.pdf\")\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA0gAAANICAIAAAByhViMAAAACXBIWXMAABJ0AAASdAHeZh94\nAAAgAElEQVR4nOzdcXxb533f+x8a12mTOuSalk5qhUwzl4ptOoCcG4myVnsBuqTKdqC+UisF\ndIembUgOXO1cacR6r9ODuR3R7aY9bLWIm3BB5rZ7cVfALCevFmdXajMRvXRLi/cmkoCIskPW\nS0OmbiKsqXHqJk3cdrh/PPPZMQiAIAnygAef91/kwTkPftSrdb99nvP8Hl+1WhUAAADsf9/l\ndgEAAABoD4IdAACARxDsAAAAPIJgBwAA4BEEOwAAAI8g2AEAAHgEwQ4AAMAjCHYAAAAeQbAD\nAADwCIIdAACARxDsAAAAPIJgBwAA4BEEOwAAAI8g2AEAAHgEwQ4AAMAjCHYAAAAeQbADAADw\nCIIdAACARxDsAAAAPIJgBwAA4BEEOwAAAI8g2AEAAHgEwQ4AAMAjCHYAAAAeQbADAADwCIId\nAACARxDsAAAAPIJgBwAA4BEEOwAAAI8g2AEAAHgEwQ4AAMAjCHYAAAAeQbADAADwCIIdAACA\nRxDsAAAAPIJgBwAA4BEEOwAAAI8g2AEAAHgEwQ4AAMAjCHYAAAAeQbADAADwCIIdAACARxDs\nAAAAPIJgBwAA4BEEOwAAAI8g2AEAAHgEwQ4AAMAjCHYAAAAeQbADAADwCIIdAACARxDsAAAA\nPIJgBwAA4BEEOwAAAI8g2AEAAHgEwQ4AAMAjCHYAAAAeQbADAADwCIIdAACARxDsAAAAPIJg\nBwAA4BEEOwAAAI8g2AEAAHgEwQ4AAMAjCHYAAAAeQbADAADwCIIdAACARxDsAAAAPIJgBwAA\n4BEEOwAAAI8g2AEAAHgEwQ4AAMAjCHYAAAAeQbADAADwCIIdAACARxDsAAAAPIJgBwAA4BEE\nOwAAAI8g2AEAAHgEwQ4AAMAjCHYAAAAeQbADAADwCIIdAACARxDsAAAAPIJgBwAA4BEEOwAA\nAI8g2AEAAHgEwQ4AAMAjCHYAAAAeQbADAADwCIIdAACARxDsAAAAPIJgBwAA4BEEOwAAAI8g\n2AEAAHgEwQ4AAMAjCHYAAAAeQbADAADwCIIdAACARxDsAAAAPIJgBwAA4BEEOwAAAI8g2AEA\nAHgEwQ4AAMAjCHYAAAAeQbADAADwCIIdAACARxDsAAAAPIJgBwAA4BEEOwAAAI8g2AEAAHgE\nwQ4AAMAjCHYAAAAeQbADAADwCIIdAACARxDsAAAAPIJgBwAA4BEEOwAAAI8g2AEAAHgEwQ4A\nAMAjCHYAAAAeQbADAADwCIIdAACARxDsAAAAPOIOtwtoZmZmZmxszDCMiYmJujdMTU0lEolM\nJjM6Ourz+USkWq06b7As6/LlywsLC+l0WkTi8fijjz56/Pjxnp6eVgpQY9bQdT0Wiw0ODrby\nbE09IrK8vNzKVwMA9t7Q0JDbJQA74tuYPDrK+Ph4Op2en58PBoM1HxUKhVAoFI/Hz58/L/WC\n1NLS0tGjR+sOWywW/X7/pt9eN9gpKysrzbNdk2A3NPSdTb/aRcvLbxz6oWtuV9HQ8p++94E7\n59yuoqFbr8Ye+OMn3a6imVs//K/v/9y421U09PwHzt/3iz/ldhUNvfAr/7H3g8fdrqKZyu9d\nLj/4k25X0VDfzc98poPL+8mbnyHYYb/r9KXYJ598UkRCoVC5XHZeL5fLoVDIvmGjUqmkUl02\nm719+3a1Wq1Wq7dv385msyISCARKpVKLNVRfT40wN9e52QIAAHSnTg92/f39+XxeRM6dO+e8\nrn7N5/P9/f0bn7IsKxAIiEixWIxEIn19fep6X19fJBIpFosikkwmLcvaRkmRSEREUqnUNp4F\nAADYPZ0e7ERE0zRd11OplGma6oppmqlUStd1TdPqPvLss8+KSDabrbve6vf7s9msaZrXrv33\nBUefz+fz+SzLGh8f9/l8U1NT26gzl8uFw2Gfz5fL5bbxOAAAwA7tg2AnIk888YSIhMNhEbEs\nS/2gLta1uroqIg8//HCjGx566CERuXHjhvPi7Oys2mNxzz33NClG5TZd150Xk8lkNBpV0TMa\njW4vGgIAAOzE/gh2fX198/PzImKa5uXLl0Vkfn7eXmDdKJFIiEjdVVpF7XtQt9kqlUqlUqlW\nq2qx1eZ7vWg0KiKjo6P2DUtLS6lUStO0tbW1arW6tra2sLCwvb8UAABg2/ZHsBORYDCo63o4\nHI5Go7qub9wku3OxWKyVNii6rq+trTlT4+LioogYhqEu9vf3T05Otr08AACA5vZNsBPHBFvN\nTFu7NGpfYu+HjcfjmqY99thjNXOBqh7n4630UgEAAGiv/RTs7Om0TefV1Atw6+vrjW5QHxmG\nsaUCzpw5Y5pmIBCo6b0CAADQCfZTsGvd4cOHReS5555rdIP6aNPTI2oMDg6qLEgTOwAA0IG8\nGeweeeQREYlGo3W7EJdKpWg0qmmaum1LTp48KSKJRMI5HajSnvO71LZcAACAvdTpR4rVaHRO\n18aPSqWS6lGczWaDwaDaQlsulwuFgtrWevXq1eHh4ebD1r2uDqh1nmCrvkvTtOnp6f7+/vX1\n9ccff1y1PuGsWADYRzhSDPudZ4OdOLLdRs5U12TYutcty+rt7RUR595YlfbsezKZzNjYWKNS\nAQAAdoM3l2IVv99fqVTy+Xw8HldX4vF4Pp+vVCrOVLdVPT096rjYixcv2hcnJiby+bw6CSOb\nzTq73AEAAOyNfTZj5wEsxQJAx2IpFvvdHW4XsAUzMzNjY2POl9tqqPXQTCYzOjraaBX18uXL\nCwsL6uiweDz+6KOPHj9+vJW+xPLayqyTruuxWGyru2uHhr5/S/fvseXlPx+697bbVTS0/OLd\nD/QW3K6ioVuV4AOWvvl97rnVk3rguY+4XUVDtx5++u2ffr/bVTT0tY/9/nee7OgulW/816Xn\nfvaQ21U09PBv3jj34fe6XUVDT3z2mtslADu1n5ZiR0dH4/F4IpEoFOr83/VCoZBIJOLxeKNl\n0KWlpd7e3mg0qlKdiKTT6Wg02tvbW3fzbCtSqdTBgwebNMwDAADYM/sp2InIk08+KSKhUKim\nRXC5XA6FQvYNG5VKpaNHj4pINpu9ffu2Okni9u3b6m25QCDQerarOqjHZ2ZmdvA3AQAAtMc+\nC3b9/f35fF5Ezp0757yufs3n8zWHfSmWZantscViMRKJqNYnItLX1xeJRIrFoogkk0nLsrZa\nTyQSEZFUKrXlvwQAAKDd9lmwExFN03RdT6VSqlGciJimmUqldF1Xm1I3evbZZ0Ukm83WPcLV\n7/dns1nTNK9d++9vV/h8Pp/PZ1nW+Pi4z+ebmpratCT7560+CwAA0C77L9iJyBNPPCEi4XBY\nRCzLUj+oi3WpcyAefvjhRjc89NBDInLjxg3nxdnZWfU23j333NPowVwuJyKnT5+uud7KswAA\nAO21n3bF2vr6+ubn50OhkGma3/zmN0Vkfn7eXmDdSLUOrrtKq6htrYlEwrnftlKpVCqVjRtm\na/bGqpMtau5p9CwAAMDu2ZfBTkSCwaCu62quTtf1jdFq52KxWCvJ7NatW+vr6zWpscVnAQAA\n2mhfLsUq9hFezrO82qhRdzrnrthisVgqlQYGBtRq76bPAgAA7J59HOzsKbFN58Z0XReRJt3m\n1EeGYWy1Br/fPz09LSJzc3NbfRYAAKC99nGwa93hw4dF5Lnnnmt0g/poe9NsahGWjicAAMB1\nXRHsHnnkERGJRqN1uxCXSqVoNKppmrptq1T3u0adVgAAAPaMr+Y01f2l7oGwdT8qlUqqR7Ha\nxKq20JbL5UKhEI1GReTq1avDw8PNh9143bIswzBSqVQ2m1XNipuUpCwvL2/zrwUA7LKhoSG3\nSwB2pFuCnTiy3UbOVNdk2JpGJzZN0+bm5tSrfpsGOwAAgF3SFUuxit/vr1Qq+Xw+Ho+rK/F4\nPJ/PVyoVZ6rbEk3TstmsneoAAABc1Ll97GZmZsbGxgzDcDYNdlKndWUyGak3T1atVi3LKhQK\nN27csPuhGIZx6NCh8+fPnz9/vslX18y3bZyo03U9FovV3Wyx6VwdS7EA0LFYisV+19FLsePj\n4+l0en5+fmP/4UKhEAqF4vG4imh1F16TyaR9nqyTpmmzs7NNTqqo0WgFdm1trclpFo0sLy8P\nDd251af20vLyqz/0I192u4qG/vSP3vWXfWtuV9HQ95UHnnn1625X0cxjd77t41+13K6ioU+9\no+dDhW+4XUVDl4JvPfFrt9yuopnf+ecPfPhnF9yuoqHP/uaj7/vQr7hdRUOfv/SLBDvsdx29\nFPvkk0+KSCgUKpfLzuvlcjkUCtk3bKRepzNNM5PJrKys2P2EV1ZWDMMwTXNkZGSrxTj7Emez\nWRGZmZnZzl8FAACwOzo62PX39+fzeRE5d+6c87r6NZ/PN5owSyaTIlIsFkdHR50LpoODgxMT\nE5lMxjTNQqGw7cLUBlh61wEAgI7S0cFORDRN03U9lUrZi6qmaaZSKV3XG7WOW1paMk3TMAy/\n31/3hhMnTmSz2QMHDqhffT6fz+crl8tTU1M+ny8cDtddwK1bm/2zGsT56cYrAAAAu6rTg52I\nPPHEEyISDodFxLIs9YO6WNfi4qI07Rjc19cXiURqtj6MjIyoPRamaW6a7XK5nIicPn16S38I\nAADArtoHwa6vr29+fl5ETNO8fPmyiMzPzzfZ+qDy2VbPB1PNUKrVqvquS5cu1dzgc4hGo6rR\n8Vb/FgAAgN2zD4KdiASDQV3Xw+FwNBrVdX0bicpXj/OGWCymetGpwdPpdPMBb926tb6+vtUy\nAAAAds/+CHby2jyc84f22nSGz7krtlgslkqlgYGB1dXV3SgGAABgG/ZNsLOPdtj0jAfDMESk\nJnJVX2+Hxfj9/unpaRGZm5vb4VAAAADtsm+CXeuOHTsmIgsLu9uiU3VaoeMJAADoHB4MdsPD\nw5qmjY2NlUqlujdYVhva7qtBGu29ZYkWAADsvY4+UqzGxnPDGn2kTp4QEcMwNE2z359bXV29\nfv16NBoVEU3TVPfjjcPWXNl4g2VZhmGkUqlsNquaFTtPP7M/rVstZ8UCQMfiSDHsd94MdiKy\nurqaSCQataPLZDKjo6ONnq0b7DbSNG1ubk6986fOrrU/MgxDbfLYR/+8AABgv/NssFOWlpYW\nFxcXFhZUwtN1/fDhw0eOHHG2wdtGsNM07dSpU8ePH3fu5DBNc2ZmRh1QOzo62qRaAACA3bCf\ngp03sBQLAB2LpVjsd3e4XcDumpmZGRsbMwxjYmKi7g1TU1OJRKLJHJtlWdeuXbtx44bdP88w\njEOHDu3k2Imhoe/f9rN7YHn5z++970/drqKhF1/4oTvfUXS7ioZe/Wpg9U3/2e0qmhn81j/4\nv1654HYVDf3Pd536+Jd+1e0qGvrUu3/hJ36no88S/O0TZx/8tyNuV9HQzZ+fffCffcTtKhq6\n+etPu10CsFMeD3ajo6PXr19PJBJ1o1ihUEgkEvF43H7frkapVEomkzUv6qmEp2na7Oxsk5PN\nAAAA9pgH253UePLJJ0UkFAqVy2Xn9XK5rLY7qBs2Ultr1TtzKysrdnPjlZUVwzBM0xwZ6dz/\ntxgAAHQh7we7/v5+1dbk3Llzzuvq13w+r1oNb5RMJkWkWCyOjo46DxwbHBycmJjIZDKmaRYK\nhV0sHQAAYCu8H+xERNM0XddTqZS9qGqaZiqV0nW9UYfhpaUl0zQNw/D7/XVvOHHiRDabPXDg\ngPrV5/P5fL5yuTw1NeXz+cLhcKNOKwAAALukK4KdiDzxxBMiEg6HRcSyLPWDuljX4uKiND5Y\nQkT6+voikYhzJk9ERkZG1Bt4pmmS7QAAwB7rlmDX19c3Pz8vIqZpXr58WUTm5+ebbH1Q+awm\nt23K7/dXKpVqtaq+69KlSzsqGgAAYCu6JdiJSDAY1HU9HA5Ho1Fd17fRr8RXj/OGWCymWhar\nwdPpdLuKBwAA2FQXBTt5bR7O+UN7bXWGDwAAoI26K9jZJ4A5jwKryzAMEVldXXVerL7eLhUJ\nAACwPd0V7Fp37NgxEVlYWHC7EAAAgFYR7OobHh7WNG1sbKxUKtW9wbKsPS4JAACgOV+3LSnW\nPRC27kfq5AkRMQxD0zT7/bnV1dXr169Ho1ER0TRNdT/eOGyjL1peXm7znwQAaJOhoSG3SwB2\nhGDX7KPV1dVEItGoHV0mk7EPmW092AEAAOwSgt3mHy0tLS0uLi4sLKiEp+v64cOHjxw54myD\nR7ADAACu83iwq+kzp+i6HovFdt6aZHvRjaVYAOhYLMViv+vGYKesrKzsMNttO9gNvesdO/ne\n3bb85a8+cFfn/k/FrVd89339uttVNPTC2x667/d/0+0qmnnh/T/77vP/q9tVNPSl8U8eiEfd\nrqKhP0lnv/0PPuR2Fc18z3++9MW/t+Xu63vmPX9YmH7wYberaOjxm88R7LDfdcWu2Jr+c9ls\nVkTm5ubcrgsAAKCduiLY1YhEIiKSSqXcLgQAAKCdujHYNVIoFMbHx30+XzgcLhQKG2/I5XLh\ncNjn8+VyuY2flkqlZDKpDpBNJpONGuABAADskjvcLsAFKpbpuu68ODU1ZR8ga5qmaZq6rk9O\nTto3JJNJe5IvGo2+9NJLzscLhUIoFLJ/TaVSqVRqfn4+GOzcl10AAIDHdMWMne/1VG9huwWd\niJRKpUQioet6pVKpVquVSkXX9VQqZc+6LS0tpVIpTdPW1taq1era2lrNaWNnz54VEfVptVq9\nevWqiFy8eHHv/kgAAND1uiLY1dB1fW1trb+/375y5coVEUkkEj09PSLS09OjZu/UdRFZXFwU\nEcMw1FP9/f3OyTwRUS3uXn75ZfXr8PBwtVo9f/78Hvw5AAAASlcEO3s/bDwe1zTtsccec6Y6\nEVExrre3157V6+3tta/bPzjbo/j9fucIaqdtIBCYmppaX19fX1/f5b8JAACgVlcEO9uZM2dM\n0wwEAuVyub0jRyKRfD6vaVoikRgYGBgYGAiHw23/FgAAgCa6K9gNDg4ahiENmthV62l9cE3T\n8vl8sVjMZrPxeNw0zaeeeqptpQMAAGymu4KdiJw8eVJEEomEc7VUpb0mDUo23rC6ulr3Tr/f\nH4lE1Nt16XS6TVUDAABsriuOFKv5G1VnE8MwJiYm1JWlpaWjR49qmjY9Pa1ev1PtS+x7SqVS\nIBCwb1hfX3/88cfVhgk1+Pj4eDqdvnr16vDwsD1gPB7fuH+Cs2IBoGNxpBj2u24MdpZlqb0R\nzr2xzjZ1iqZps7OzfX196ldnozsRyWQyY2Nj9uAqydV8+86PowUAAGhd1y3FikhPT4/axOrs\nMzc5OTk/Px+Px9WvmUzGmepEZGJiQm2PEJFsNutsgyciw8PDxWLRbnqs6zqpDgAA7DGPz9h1\nIJZiAaBjsRSL/c6zR4qpRdgauq7HYrFNJ9JqFnA3DqVpmqZpJ06ccE7ptW7o3ndu46k9s/zi\nV+6/601uV9HQ8698676/+IbbVTT0wlveevALHd3FcOV/6r/33153u4qGXvz5h970s3/odhUN\nfes3/94fvaejT5T5kS+ezL7nP7hdRUPRL/7jX3iwczu3/+rNcbdLAHaqu5ZiU6nUwYMHG21o\nbZ1pmmNjYyMjI3SqAwAAncPjwa6mKZ16ta5uE7stDaUOkzVNs1AotLtkAACAbfJ4sKsRiURE\npGb36zbYh8leuHChDWUBAAC0Q3cFu0ZyuVw4HPb5fLlcrsVHenp6RES1shMRdcKs84aNVwAA\nAHaVZzdP1KVym92URHF2sItGoy+99FIrQ1mWJSKq+wkAAEAn8PiMne/1otGoiDhb0C0tLaVS\nKU3T1tbWqtXq2trawsLCpsOWy2V1yNipU6d2r3gAAIAt6a4ZO13XR0dH7dMmRGRxcVFEDMNQ\nF/v7+ycnJ+0FVqe6TU+OHz++m/UCAABsgceDnd2Lbnx8/KWXXnrsscecqU5E1B4IZ2c7v9+/\n6bB2Hzv1ph0AAEAn8Hiws505c+bgwYOmad6+fXt7XYU5ogMAAHQ4j79jZxscHFRvxW2viR0A\nAEDn65ZgJyInT54UkUQisb7+P058UmmvVCrZV3Z+LkW7BgEAANgSn1dXGGvOe1WmpqYSiYRh\nGBMTE+pKqVQKBAKapk1PT/f396+vrz/++ONq80TNWbHN/6HGx8fT6fT8/HwwGLQsyzAM1UJl\n41PLy8vt+QsBAO02NDTkdgnAjuyDYFd3O6rau9Dkbbm6acyyrN7eXhFZW1uzd1GotGffk8lk\nxsbG7F+r1Worwa5QKIRCIftXwzDUmJ3/zwsAADxjXy7FmqY5NjY2MjJSLpe39GBPT486Lvbi\nxYv2xYmJiXw+r1oNZ7NZZ5e71gWDQXuQTCZjzwgCAADsmX0zY+es017rzGaz6vjXvfnetmAp\nFgA6Fkux2O/2ZbCT1xZVNU3L5/N7+b07t7y8fN+772/vmO31wpeeP/BD73K7iob+5E+//MY3\nvs3tKhr6zne+frv8vW5X0czdfX/1+avfdLuKht539M2fefprblfR0E9+5O3n/0XR7SqaGf+X\ngV9+7IrbVTT01DM/9viDT7tdRUPTNz9CsMN+ty+XYkVEdQa2j4hQJ4Y5b3BeUT9bljU+Pu7z\n+aamppw35HI5n88XDofVSbJNFAoFNUI4HC4UCjWflkqlZDKphk0mk86dtgAAAHtgvwY7y7JE\nRL3T1qLZ2dl0Oi0i99xzj33RNE11gKz6IZlMNnp8amoqFAqpEUzTDIVCzpsLhUIgEFA7YUUk\nlUoFAoGN4Q8AAGD37MtgVy6XVf+5U6dOtf5UpVKpVCrVatX5Wt7MzMza2lq1Wl1bW9M0LZVK\nLS0tbXy2VColEgld19UIlUpF1/VUKmVPy509e1ZE1FDVavXq1avy+i0aAAAAu23fBDufw913\n351KpTRNO378eOsjxGKxjUe7Tk5Oqr4n/f39k5OTIrK4uLjx2StXrohIIpFQI/T09KhuJuq6\nvLYo/PLLL6tfh4eHq9Xq+fPnt/hXAgAAbN++CXZOmqZlMpnZ2dmNQa2JwcHBjRf9fn/Nz86e\ndjZ1sbe31w6Xqh+efbPqohIIBKamptbX152HWwAAAOyNO9wuoFUdvns3Eom8+c1vnpmZSSQS\nKu1pmjY7O9ukhTIAAEB77csZO7dU67E/Vb1XisViNpuNx+OmaT711FMuVgsAALqNN4Pd6urq\nNu5UOyEymczG29RejVY6mPj9/kgkot6uU1toAQAA9oZHgl08HhcR1V7Esqy5ubkWH0wkEup9\nuPX1ddW+5PDhwxtvO3bsmIgkk0n75blCoWC3xBMR1d/O3lGrflBVAQAA7A2PBLuTJ0+KSCgU\nUtsa1M6GVvj9/oGBAZ/PNzAwYJqmYRjO7RS24eFhXddN01Q3+3y+UCikaVosFlM3fPSjHxWR\no0ePqk+PHj0qImfOnGnPnwcAANCC/Xqk2Eamac7MzJimmclkRkdHnU/VHcG+ODU1lUgk4vH4\nyZMng8Fgk+8tFAoXL15UC6yZTObEiRPOvRGlUumZZ55RPYp1XY/FYnX34XJWLAB0LI4Uw363\nD4LdLtmlo2ABAADc4pGlWAAAAOybPnaewVIsAHQslmKx33k22KmVVidN0zRNq3kxbquDSNP3\n5zY+W3ep994f3uRZd734x6s/8JYDblfR0J/9xZ/4/vSNblfRUPWHvvOthb90u4pm3vTo933f\np/6L21U09Jcf/7s/8JHLblfR0J89ffyvH510u4pmvnshufjgaberaOjYzbPTD/4jt6to6PGb\n/8ntEoCd6qKlWNM0x8bGRkZGyuWyvNZteBvjpFKpgwcPtt4qDwAAYG94PNg5j4ioVCqqZYlq\nd7e9QarVqjoWtvVWeQAAAHvD48HOqaenR53ieuHChZ2ME4lERES1NQEAAOgcXRTsRKSnp0dE\nTNNUv6pmws4bNl5pXS6XC4fDPp8vl8vtsE4AAIBt8OzmibosyxIRTdN2MojKbbquOy8mk0l7\nDi8ajb700ks7+QoAAIBt6KIZu3K5bBiGiJw6dWpLD/peLxqNisjo6Kh9w9LSUiqV0jRtbW2t\nWq2ura0tLCy0t3gAAIBNeXzGrm7Tk+PHj+9kTF3XR0dH+/v77SuLi4siYhiGutjf3z85OWkv\n+AIAAOwNjwc7J7uPnXrTrnV2V5Tx8fGXXnrpsccec6Y6EVF7Mpyd7fx+/47rBQAA2BqPB7v2\nHgV75syZgwcPmqZ5+/bt1rscAwAA7I0uesdu5wYHB9VbejSxAwAAHYhg9z+0cpjEyZMnRSSR\nSKyvr9sXVdorlUpbGgoAAKC9fO1drOwcTY5qtY2Pj6fT6fn5+WAwaFmWYRiqZYn9VN1Bpqam\nEomEYRgTExPqSqlUCgQCmqZNT0/39/evr68//vjjavPExgKWl5fb8xcCANptaGjI7RKAHenq\nYFcoFEKhkP2rYRhqG0TzYGdZVm9vr4isra3ZuyhU2rPvyWQyY2NjmxYAAADQRl29FBsMBvP5\nvOpXnMlk7Bm45np6etRxsRcvXrQvTkxM2ENls1lnlzsAAIC90XEzdnU7z6k2Ja1vRN04iK7r\nsVjM2ZGkRZZlXbt27caNG/aEnGEYhw4dCgaDWx1KYSkWADoWS7HY7/ZBsFM0TZudnW0x2zUa\nxLl42opSqZRMJuu2Gt5SPU7Ly8tD9757q0/tpeUXv/T2nne4XUVDX7O++rf/7a1uV9HQG77r\nGzfL3+N2Fc082Pft6ef/xu0qGnr8/js+Nv8Nt6to6NOht/o/e9ntKpopffj4e/6Pf+V2FQ19\n8Z984j0TH3O7ioa+OPVpgh32uw5diq06VCoVXddN0ywUCtseRC2ezszMtP642hJhmmYmk1lZ\nWbGHWllZMQzDNM2RkZGt/VUAAAC7qUODnVNPT49aBr1w4cK2B4lEIiKiNr22KJlMikixWBwd\nHXWu4Q4ODk5MTGQymW1kTQAAgN2zD4KdiKhDwOwlUZ/PV7PYuvFKXWpzQyuDLP+8n20AACAA\nSURBVC0tmaZpGEajw8FOnDiRzWYPHDjgfLZcLk9NTfl8vnA4zFmxAABgj+2PYGdZlrw+lm1V\nLpcTkdOnT7d4/+LiYvNv7Ovri0QiNbsxRkZG1OSiaZpkOwAAsMf2QbArl8vqaIdTp05t6UGf\nQzQazWazrW9lVflsq7to/X5/pVKpVqvz8/MicunSpS09DgAAsBN3uF1AfXWbnhw/fnwnY966\ndWt9fX1Lu2I3rUpe34I4FoupVWOVINPp9Pnz57f9dQAAAFuyD2bsNE3LZDKzs7MqM7XOuSu2\nWCyWSqWBgYFdPcV1G33yAAAA2qVDg50zk+Xz+dHR0W10jHPy+/3T09MiMjc318r9avG3JgVW\nX28n9QAAALRdhwa73aAWYVvseHLs2DERWVhY2N2aAAAA2scLwa7F1dXmW2trBhkeHtY0bWxs\nrFQqNRkNAACgc3To5onm4vF4Op0uFArBYNCyrFZWVy3Lqtlau+kgk5OTpmkGAgHDMDRNs9+f\nW11dvX79ejQale12YFl+8UvbeGovfc36qtslNPOG7+rcI6dE5MG+b7tdwiYev7+j/xf/06HO\nPTJOREof3tEurj3wxX/yCbdLaOaLU592uwTAyzr0rNjmVRUKhVAoZP9qGIbqTmI/1eTA2bm5\nObUJY9NBRGR1dTWRSDRqR5fJZEZHRxuV3cofAgAA0Eb7MtiJiGmaMzMz6iDX0dHRmqfqdks5\nderU8ePHnVtrmw9iW1paWlxcXFhYUAlP1/XDhw8fOXLEuZ+DYAcAAFzXccHO85aXl90uAQBQ\n39DQkNslADvS0a/a7FzdqTtN006cONF6/5SNg+i6HovFtt21bujed2/vwb2x/OKX3v4DA25X\n0dDX/mztb/9b576D9Ybv+sbN8ve4XUUzD/Z9+98//zduV9HQR++/42PznfsO5adDb33o0//F\n7Sqauf6xv3voX3/O7SoauvHkB/yjv+J2FQ2VZn7R7RKAnfLCrtgtMU1zbGxsZGSkXC5ve5BU\nKnXw4MH19fU2FgYAALBDXRHsnF2FK5WKruumaRYKhW0Pks1mRWRmZmZ36gUAANiOrgh2Tj09\nPWr364ULF7Y9SCQSkZZ7HQMAAOyNrgt2IqI2xtpNTHw+X81bdBuv1OVsYrftQQAAANqlG4Nd\n8yMoWpHL5UTk9OnTbasJAABgxzy+K3ajcrl87tw5cRxB0aKa6bdsNhsMBttZGQAAwM50RbCr\n2/Tk+PEdnQt069at9fX1/v7+nQwCAADQRl23FKtpWiaTmZ2ddR5B0QrnrthisVgqlQYGBlZX\nV3epTgAAgK3qihm7tp+u4ff7p6enTdOcm5ubnJxs7+AAAADb03Uzdu2iFmHpeAIAADoHwa5W\ni6urzbfWskQLAAD2XlcsxTYXj8fT6XShUAgGg5Zlzc3NbfqIZVmGYYhja+2WBll+8UttqXz3\nfO3P1twuoZk3fFfnniUqIg/2fdvtEjbx0fs7+n/xPx3q3LOAReT6x/6u2yVs4saTH3C7hGY4\njxXYVb62v3/WUdR+2OZ/Y6FQCIVC9q+GYaijKeynGvUZ1jRtbm5ObcLYdBAAAIDdxlKsBIPB\nfD6vFlUzmczExMSmj2ials1m7VS3vUEAAADay+MzdpZlXb58eWFhIZ1Oi0g8Hn/00UePHz/e\nYq+TunN1uq7HYrHBwcFWnt34z7u8vNxS6QCAPTc0NOR2CcCOeDnYLS0tHT16tO5HxWLR7/dv\nOkKTw15XVlaaZ7smwW7o3rdv+tUuWn7xaw/84DfdrqKhW//1zff7Pu92FQ09X33f/V//Hber\naOb5t52473P/yu0qGnrhA5848Kl/7nYVDf3Jx3/tjtg/cruKZv5m7j995cEH3a6ioXfevPlM\nB5f32M2bBDvsd55dii2VSirVZbPZ27dvq8bCt2/fzmazIhIIBEqlUotDVV9PjdDKHgsAAIC9\n5M1gZ1lWIBAQkWKxGIlE+vr61PW+vr5IJFIsFkUkmUyqliVbFYlEhA52AACg83gz2D377LMi\nks1m6663+v3+bDZrmua1a9fUFZ/P5/P5LMsaHx/3+XxTU1Pb+NJcLhcOh30+Xy6X20nxAAAA\n29PR7ay2TfUHfvjhhxvd8NBDD4nIjRs3gsGgfXF2dlbtsbjnnnuaDK5ym67rzovJZNKew4tG\noy+99NL2qwcAANgWb87YqR5y6tSvutS+B3WbrVKpVCqVarWqFlttvteLRqMiMjo6at+wtLSU\nSqU0TVtbW6tWq2trawsLC+39iwAAADblzWC3PbFYrJU2KLqur62tOVPj4uKiiBiGoS729/dP\nTk7uXp0AAAB1eXMpdnsatS+xW5aMj4+/9NJLjz32WM1coJr5cz7eSi8VAACA9vLmjJ16AW59\nfb3RDeojdd5r686cOWOaZiAQKJfLO6wQAACg7bwZ7A4fPiwizz33XKMb1Eebnh5RY3BwUGVB\nmtgBAIAO5M1g98gjj4hINBqt24W4VCpFo1FN09RtW3Ly5EkRSSQSzulAlfac36W25QIAAOwl\nzx4pViqVVI/ibDYbDAZVj+JyuVwoFNS21qtXrw4PD6ubGx3/Vff61NRUIpEwDGNiYsL5XZqm\nTU9P9/f3r6+vP/7446Zp1h2Ts2IBoGNxpBj2O88GO3Fku42cqU62GOwsy+rt7RUR595Ylfbs\nezKZzNjYWN0xAQAAdkkHLcValpXL5dTZDz6fb3x8PJfLbe/UL8Xv91cqlXw+H4/H1ZV4PJ7P\n5yuVijPVbVVPT486LvbixYv2xYmJiXw+r2maiGSzWWeXOwAAgL3RKTN2S0tLR48erftRsVj0\nUvcQlmIBoGOxFIv9riP62JVKJZXq6r4PFwgEPJbtht71NrdLaGb5y19/4K1/7XYVDd36xnff\n/7c33K6ioeffcOj+P0q7XUUzz/9I/P7PjbtdRUPPf+B8/7/7+25X0dD6P/1/7vzn73G7imZe\n/bUv3vxY/VdQOsGDny4+89ght6to6LFnOve/LUCL3F+KtSxLvQlXLBYjkYhKdSLS19cXiUSK\nxaKIJJPJnazJAgAAdAP3g92zzz4rItlstu6cnN/vz2azpmleu3ZNXVFv4FmWpd7Gm5qaUtdL\npdLU1JT6NBwO53I55zjqerlcVveEw2G1a9Upl8uFw2Gfz6eeVY84bygUCupLw+FwoVCoebxU\nKiWTSfVUMpms22kFAABg97i/FKtavj388MONbnjooYdE5MaNG8Fg0L44OzubTqdF5J577hER\n0zTD4bD9qWmaKrdFIhHnUCMjI+q6usHe7iAiyWQylUqpn6PR6CuvvFJThnPfq3pc13X7TNhC\noRAKheybU6lUKpWan5931gwAALCr3J+xU2mp5vRVJ3U+hLOZiIhUKpVKpVKtVlV0U6nu6tWr\n1Wq1Wq1evXpVRFS/Oie1T7Zarc7Pz4vIpUuX1PVSqZRKpTRNW1tbq1ara2trNfN5pVIpkUjo\nuq4er1Qquq6nUil7Wu7s2bMioh63C3BumwUAANht7ge77YnFYj09PfavKk7ZTUwadTOxn1IT\naWrOT0SuXLkiIpOTkypf9vf321NxzhsSiYR6vKenRwVNdV1EVBB8+eWX7QKq1er58+fb89cC\nAAC0YL8Gu7rHvJbL5VKpZJpmMpls/Sl5bTrQ+ZJfzQt/6obe3l7fa1SPYnseUXW2CwQCU1NT\n6+vrzgPHAAAA9ob7wU7XdRFpkoTUR+o81iaSyeTdd98dCATC4bD9ttyeiUQi6o29RCIxMDAw\nMDAQDofL5fIelwEAALqZ+8Hu8OHDIvLcc881ukF91GiyTZmZmUmlUvF4fH5+vlgs3r59u+11\nymsLvjXsTzVNy+fzxWIxm83G43HTNJ966qndKAMAAKAu94PdI488IiLRaLRuf5BSqRSNRjVN\nU7c1og5mPX/+fDAY3EYrYzUd6CygppiNNzTi9/sjkYh6u85+hw8AAGAPuB/senp6VBfiQCCQ\ny+Xs5ctyuZzL5VTv4k984hPOrRKNqM4p6+vr586d21INP/ZjPyYiyWRSLfuur6/XvKV37Ngx\n5w0iUigUnF30VH+7paUl9av6wT6jFgAAYA+438dORPx+f7FYDAQCGxuUiMjVq1cb7XK1ZbPZ\naDR68ODBmuurq6vN13DtAlT7ko1di5Xh4eGNN2iaFovF1M8f/ehH0+l0zXG3Z86c2fSrAQAA\n2sXnfEvMXZZlPfvss5cuXVIrmPF4/EMf+tAjjzxSM1enToPYWPbMzIxakNV1PRaL/dVf/VUg\nEDAMY2Jiou5TG6/kcrkLFy6ozsOxWEzFROcNhULh4sWLqrxMJnPixAn7ADQRKZVKzzzzjNq3\noUaomymXl5e39+8DANhtQ0NDbpcA7EgHBbudsyzr2rVrN27csLuQGIZx6NCh7R3/4PP5nGdL\nAAAAdLiOWIptC3VUa81aqkp4mqbNzs46Z9dqqNk7e83XsqzZ2VkReeCBB3a3aAAAgPbxyIxd\nqVRS2ywymcyjjz5qr4Gurq6applIJFQvkkaP1xw1qzR/ZNtYigWAjsVSLPY7jwS7cDhsmmax\nWKzb60S9fjc/P99kTbZQKPz+7/++ekMuHo8/+uij6hTatlteXu7w/3AsLy+/4x3vcLuKhr76\n1a/eeeedblfR0Kuvvtrhjan7+vr+4A/+wO0qGvrRH/3R3/qt33K7ioZ+5md+psP7U/7yL//y\nhz/8YberaOizn/3sgw8+6HYVDd28ebPD//sMbMr9dic7t7S0ZJqmYRiNOtidOHEim80eOHBA\n/arOBLMsS/UoUS1LgsHg5OTk/Px8PB5Pp9MXLlwoFArqfsuyfD7f+Ph4zbA119VasBo8mUy2\n0vQOAACgjbwQ7BYXF0VE07RGN/T19UUikZo9qrOzs2p/6z333KOuTE1NhUIhddE0zVAopLrZ\n9fT0GIaRTqdVnzzbCy+8ICIf+tCHRKRQKAQCAfsos1QqFQgE7GgIAACwB7wQ7NQOiVb61TlV\nKpVKpVKtVtWSa6lUSiQSuq6ri5VKRTWuUxNvqoPxwsKCc4SbN2+KyJEjR0Tk7NmzIrK2tqbO\nGbt69aqIXLx4sS1/IAAAQCu8EOzq8tXjvCEWizk75F25ckVEEomEutjT06Pyorru9/vj8fjY\n2JhlWfYjY2Nj8XhcbbZVu3Fffvll9dHw8HC1WlUHiwEAAOwNzwa7TdXM8KkY19vba6fA3t5e\n+7qInDx5UkSeffZZ9atallXrsCKSzWZFJBAITE1Nra+v2yePAQAA7BkvBDvDMOS1pGWrvt7O\nv+W9732viMzMzKhfV1ZW5LV1WBGJRCL5fF7TtEQiMTAwMDAwEA6HO3x3JAAA8BgvBLtjx47J\nhhfgtqdaj/qop6cnm82apqneurt06ZK9DquovnfFYjGbzcbjcdM0O7wtAgAA8BgvBLvh4WFN\n08bGxhp1GHG+GNeImvZr3qPk4YcfFpFnnnlmfX09nU7b67BOfr8/Eomot+vUBlsAAIC94YVg\nJyLqRFf1iptzTXZ1dTWXy6m35Zr0Q5HXpv2SyaT9elyhULC73Cn9/f3xeDyVSj3++OPiWIcV\nEdUSb2lpSf2qfojH4236+wAAADbnkWDn9/tXVlbUK24HDx60N0AcPHgwGo2KSCaTaX4+2PDw\nsK7rpmkODAyoZ0OhkKZpsVjMeZvKaqZp1qzDfvSjHxWRo0ePqmePHj0qImfOnNmNPxYAAKAu\njxwpZltaWlpcXFxYWFD9R3RdP3z48JEjR5whTPU9qfuHFwqFixcvqiXUTCZz4sQJ54MiYlmW\nmv9TWyWcH5VKpWeeeUb1KNZ1PRaL1W2tx1mxANCxOFIM+53Xgt1uK5VKgUBARG7fvl2T+QAA\nANzlkaXYPfPMM8+IyPz8PKkOAAB0mjvcLmDfsA+u0HU9GAxuexyWYgGgY7EUi/1uPwU7y7Ku\nXbt248YN+zQIwzAOHTq0k5jVOk3TTNM0DGNiYmKHQ91737vaUtIuefGFL3//u+5yu4qG/vzL\nr/xl3zfcrqKh7yu/9fqdS25X0cxDrw7/e2vO7Soa+mhP7ONf+lW3q2joU+/+hZ/43X/mdhXN\n/PaP//qDmajbVTR0cyz74Jkfd7uKhm7+xu+6XQKwU/sm2JVKpWQyqbZE2FTC0zRtdnZ2t9dG\nm2+qBQAAcN3+eMdObVkwTTOTyaysrNhnQqysrBiGYZrmyMiI2zUCAAC4bH/M2CWTSREpFot+\nv995fXBwcGJi4i1vecvY2FihUNibNVkAAIDOtA9m7JaWltTLbTWpznbixIlsNnvgwAH1q2oR\nbFmWOg3CeXREoVBQF8PhcKFQ2DhUkxvUsOVyeWpqSt1Qsy6sFovVbclksvnpZAAAAG23D4Ld\n4uKiND0QrK+vLxKJ1HQDnp2dVX2G77nnHnVlamoqFAqpi6ZphkIhNRFo2/QGERkZGVEv9pmm\n6cx2hUIhEAio7sQikkqlAoFA3ewIAACwS/ZBsFNBqu4pDk1UKpVKpVKtViORiIiUSqVEIqHr\nurpYqVR0XU+lUva82qY3KH6/X90wPz8vIpcuXVLXz549KyJra2vq5b+rV6+KyMWLF3f6xwMA\nALRsHwS7unz1OG+IxWI9PT32r1euXBGRRCKhLvb09Ki8qK63ckPNsOp9PjW9JyJq6u7ll19W\nvw4PD1er1fPnz7f/LwcAAGhgvwa7TdXM8KmU1tvba6dAdeSr3RJv0xvqDmvLZrMiEggEpqam\n1tfX19fX2/8nAQAANLUPgp1hGCKyurrqvFh9PZdK+x8ikUg+n9c0LZFIDAwMDAwMhMPhcrns\ndl0AAKCL7INgd+zYMRFZWFjY+VDVerZ0QxOapuXz+WKxmM1m4/G4aZpPPfXUzmsGAABo0T4I\ndsPDw5qmjY2NNWogYlnWpoOoab8mLUg2vaFFfr8/Eomot+vsN/AAAAD2gK8T1jE3pU6eEBHD\nMDRNs190W11dvX79ejQaldcmzERE7aKo+buWlpaOHj2qadr09HR/f7+IFAqFUChkn/266Q0b\nh3VeGR8fT6fTV69eHR4etkeLx+Mb908sLy/vxj8RAGDnhoaG3C4B2JH9EexEZHV1NZFI1PQE\ntmUymdHRUfVz3WAnIslk0u4zp9QcMtv8hubBTiW5mm9cWVnZapcWAACAbds3wU5ZWlpaXFxc\nWFhQCU/X9cOHDx85csQOZ9I42IlIoVC4ePGiWiHNZDInTpxwPtj8hubBTkRKpdIzzzyjoqGu\n67FYjFQHAAD20j4LdltiWda1a9du3LhhtywxDOPQoUNbPVK2pkOe7Cy3sRQLAB2LpVjsd54N\ndurk1rpLtzUrsJvaGOyUtbU19TbeliwvL9973zu2+tReevGFr971zu9xu4qGXvnKt7/ztr9w\nu4qG3vj1t9x6U0d3MXzgW/2572y+38gtkTf2/OJX3C6isV95p0R+77+6XUUzuQ/+4Pv+7Yrb\nVTT0+Z8/+L4nPut2FQ19/tyHCXbY7/bBrthtUJstTNPMZDIrKyt245KVlRXDMEzTHBkZ2eqY\nzgYoqh3xzMzMLtQOAACwTd4MdslkUkSKxeLo6KhzwXRwcHBiYiKTyZimWSgUtj2+On+2ZqcF\nAACAuzwY7JaWlkzTNAzD7/fXveHEiRPZbPbAgQPqV3WAWLlcnpqa8vl84XC40d7bGpqm2T9v\nPKx24xUAAIBd5cFgt7i4KK9PXTX6+voikUjN1oeRkRG1x8I0zU2zXS6XE5HTp0+3p2IAAIB2\n8GCwU/lsq1tW/X5/pVKpVqvz8/MicunSpZobfA7RaDSbzW51dy0AAMCu8mCwq8tXj/OGWCzW\n09MjIiqubXoa2K1bt9bXO3rzIwAA6DbdEuw2tekMn3NXbLFYLJVKAwMDq6ure1MeAADApjwY\n7AzDEJGayFV9vR1+hd/vn56eFpG5ubkdDgUAANAuHgx2x44dE5GFhYVd/RbVmpiOJwAAoHN4\nMNgNDw9rmjY2NlYqlereYFltaLuvBmm095YlWgAAsPe8eaSYOnlCRAzD0DTNfn9udXX1+vXr\n0WhURDRNy+fz8tqJYc5/h5orG2+wLMswjFQqlc1mVbPi8fHxdDo9Pz8fDAbtT2ueUjgrFgA6\nFkeKYb/zZrATkdXV1UQi0agdXSaTGR0dVT+3GOw20jRtbm5O7aUtFAqhUMj+yDAM1XXFq/+8\nAACgA3k22ClLS0uLi4sLCwsq4em6fvjw4SNHjvT19dn3bCPYaZp26tSp48ePq1SnmKY5MzOj\nDqgdHR3dOCwAAMCu8niw60AsxQJAx2IpFvvdHW4XsLs2Trbpuh6LxbZ0LkXdGTtN006cOOGc\n+Wvdvffdv42n9syLLzx/19t/2O0qGnrla3/8Z/JWt6to6AfkG5+//Ua3q2jmfXd/5/f+4BW3\nq2jogz961++lv+x2FQ19MP6u//vj/5/bVTTzDz91+LN/7z+6XUVDH/7Dn/rEg/+b21U09K9u\n/u9ulwDslAd3xTaXSqUOHjy4w0MjTNMcGxsbGRkpl8vtKgwAAGCHuiLYOVsTZ7NZEZmZmdnJ\nIJVKRdd10zQLhcIu1AsAALAdXRHsnFR3kh02Fu7p6VGbXi9cuNCesgAAAHas64Kd4mws7PP5\nat6i23hlI7Uf1m6nsr1BAAAA2qjrgl0ulxOR06dP73Cc5idPAAAA7D2P74pVambOstlsMBjc\nyYDlcvncuXMicurUqR1VBgAA0D5dEexq3Lp1a319vb+/f0tP1W16cvz48fbVBQAAsCNdsRTr\n3NBaLBZLpdLAwMDq6uq2B9Q0LZPJzM7OOk+eAAAAcFfXzdj5/f7p6WnTNOfm5iYnJ1t/kCM6\nAABAh+uKGbsaahF2hx1PAAAAOk03BrvmG1p3skTb3kEAAAC2pOuWYi3LMgxDHBta4/F4Op0u\nFArBYNCyrLm5uW0Mu6VBXnzh+W18xV565Wt/7HYJzfyAfMPtEpp5393fcbuETXzwR+9yu4Rm\nPhh/l9slNPMPP3XY7RI28eE//Cm3S2iG81iBXeXz9qtjjVoEa5o2Nzentj4UCoVQKGR/ZBiG\nOlXC/pdRgzT/h9p0EAAAgN3WdUuxmqZls1k71YlIMBjM5/NqZTaTyUxMTGxj2LYMAgAAsBPe\nnLHbOFGn63osFhscHNzqUJZlXbt27caNG2oGTkQMwzh06NC2WxwvLy9v70EAwG4bGhpyuwRg\nR7ol2Clra2tb6ktcKpWSyaR9IKyTpmmzs7N9fX1brW15efne++7f6lN76cUXnv/+d205Ae+Z\nP//y6nfe/la3q2jojV/7xnPf/S23q2jm4b9+07mXS25X0dATf8f/8T/aznuue+NTPxL7id/9\nZ25X0cxv//ivP5iJul1FQzfHsg+e+XG3q2jo5m/8LsEO+52Xl2KdfYmz2ayIzMzMtP54qVQK\nBAKmaWYymZWVFXuolZUVwzBM0xwZGdm12gEAALbMy8HOKRKJyBZ71yWTSREpFoujo6PONdzB\nwcGJiYlMJmOaZqFQaHupAAAA29MtwU5x9q7z+Xw1K7bOK0tLS6ZpGobh9/vrDnXixIlsNnvg\nwAHns+VyeWpqyufzhcPhugu4AAAAu6dbgl0ulxOR06dPt3j/4uKiNG5iLCJ9fX2RSKRmN8bI\nyIjaY2GaJtkOAADsMS8HO59DNBrNZrOtb2VV+Wyru2j9fn+lUqlWq/Pz8yJy6dKlrdYMAACw\nbV4OdjVu3bq1vr6+kxF89ThviMViqj2eSpDpdHonXwcAALAlXg52zl2xxWKxVCoNDAzs6imu\n2+iTBwAA0C5eDnZOfr9/enpaRFo8CladJ1uTAquvtxt1AgAAbFu3BDsRUa2JW+x4cuzYMRFZ\nWFjY3ZoAAADap4uCnWVZ0nija83k3PDwsKZpY2NjpVL9Hv1qNAAAgM7h5SPFnH+aZVmGYaRS\nqWw2q5oVj4+Pp9Pp+fn5YDBof+p8Sp08ISKGYWiaZr8/t7q6ev369Wg0KiKapuXz+brfuPGK\nwlmxANCxOFIM+52Xg91GmqbNzc2pjauFQiEUCtkfGYahWpw4/0FWV1cTiUSjdnSZTGZ0dNT5\nja0EOwAAgF3SLcFO07RTp04dP35cpTrFNM2ZmRl1Guzo6GijKLa0tLS4uLiwsKASnq7rhw8f\nPnLkSF9fX803EuwAAICLvBnsOhlLsQDQsViKxX53h9sF7JaNk3a6rsdiMWerObcm1e59d0f/\nh+PFLy2/+W33ul1FQ9/8+ouvfM8Bt6to6K5v/8nn/+otblfRzPu+9y/Sf/QGt6toKP4jf/u/\n/P4rblfR0L95/13h/3NHfc53W/7n+t/3L/5ft6to6PP/8sj7Yp3buf3zc3G3SwB2qot2xaZS\nqYMHD+7w8AkAAICO5fFg5+wnnM1mRWRmZsbtogAAAHaFx4Odk+py0mKDYgAAgH2ni4Kd0qhB\nsc/nq3ktb+OVQqEwPj7u8/nC4XChUKgZoVQqJZNJ9VQymWzU2RgAAGCXdFGwy+VyInL69Ont\nPT41NRUKhdLptIiYphkKhZLJpP1poVAIBAL2dGAqlQoEAhvDHwAAwO7xeLDzOUSj0Ww2GwwG\ntzFOqVRKJBK6rlcqlWq1WqlUdF1PpVL2tNzZs2dFZG1tTb3Pd/XqVRG5ePFiG/8WAACA5jwe\n7GrcunVre7tir1y5IiKJREL1N+7p6VHHVKjrIqJ6F7/88svq1+Hh4Wq1ev78+baUDQAA0AqP\nBzvnrthisVgqlQYGBlZXV7c6jopxvb299vxfb2+vfV1E1JbbQCAwNTW1vr5OUxUAALD3PB7s\nnPx+//T0tIjMzc21ffBIJJLP5zVNSyQSAwMDAwMD4XC4XC63/YsAAAAa6aJgJyL9/f2yg44n\n1XrsTzVNy+fzxWIxm83G43HTNJ966qn21A0AANCC7gp2lmVJ444nTjXLtYZhiEgrHUz8fn8k\nElFv16kttAAAAHvDs2fFbmRZlspnp06d2vhpPB5Pp9OFQiEYDFqWVbNcTmZTYAAAIABJREFU\ne+zYMRFJJpPT09Nq2q9QKIRCIcMwJiYmRGR8fDydTl+9enV4eFhElpaW1Jh1K3nxS8tt/tva\n7Ztff9HtEpq569t/4nYJzbzve//C7RI2Ef+Rv3W7hGb+zfvvcruEZvI/1+92CZv4/L884nYJ\nzXAeK7CrfM7FRC+p6S1s0zRtbm5ObW5V96h/ARXU7NsMw1AbI+x/n2QyWbOGq2na7OxsX1+f\niCwtLR09erTmu1ZWVgYHB9v1FwEAADTXRUuxmqZls1k71dUIBoNq94OIZDIZNQ/nNDk5OT8/\nb0/CZTIZO9WJyPDwcLFY1HVd/arrOqkOAADsMc/O2AEAAHSbLpqxAwAA8DaCHQAAgEcQ7AAA\nADyCYAcAAOARBDsAAACPINgBAAB4BMEOAADAIwh2AAAAHkGwAwAA8AiCHQAAgEcQ7AAAADyC\nYAcAAOARBDsAAACPINgBAAB4xB1uF9B1lpeX3S4BAIBuNDQ05HYJu67rgp1lWZcvX15YWEin\n0yKi6/r73//+YDC4pUFKpZLf768Z9umnnx4bGxORTCbzwQ9+sL+/v9Hjn5O/3Fbte+QD8n1D\nQ19xu4qGlpff+V+/9ka3q2joB9/+nXd/5jfdrqKZL/3kz37my24X0dhPvkvuX/sNt6to6PmB\nM195879zu4pm3vnNf3rfq9/vdhUNvXDnn995551uV9HQq6++evszZberaObun+z7ztv+wu0q\nGnrj198y9EOzblfR0PKfjrhdwl7orqVYy7JisVg0GlWpTkRSqVQoFAqHw5ZltTjI+Ph4IBCo\nuRiLxVSqE5GxsbGBgYH19fV2lQ0AANCK7gp2Tz/9tGmauq7fvn27Wq1Wq9Xbt2/rum6a5uXL\nl1scxA6FtlwuZ5pmJpNRY2azWRG5ePFim6sHAABoqruCnZpUSyQSfX196kpfX18ikRCRaDS6\n7WEvXLggIh/5yEfUr5FIRH3LDqsFAADYku4KdkpPT0/Nr2qmzXmxUCiMj4/7fL5wOFwoFOzr\nPp/P/sH+OZ/PV6vVmmE1TduV6gEAABrormCnFklnZmbK5Wavx05NTYVCIbXkappmKBRKJpOt\nf0sulxOR06dP76xYAACAremuYBeJRHRdHxsbu/vuu8fHx3O53Orqas09pVIpkUjoul6pVKrV\naqVS0XU9lUqVSiURsSf2Nk7yicjS0pLP54tGo9lsdqs7bQEAAHaou4KdiExOThaLxUwmk06n\no9HowYMHw+GwCm3KlStXRCSRSKil1Z6eHvW2nLre3Le+9S3DMDRNi0ajpmnu2h8BAABQR9cF\nOxHx+/2jo6PVanVlZSWfz4tIIBCwp+5UjOvt7fW9pre3V1rbDBEMBicmJvL5fCaTqXk5DwAA\nYLd1Y7CzDQ4Oapo2PT0tIr/xG+3siap2yJ49e7aNYwIAADTXRcHOuY/VSR0RUdOdrlpP69+l\nlnFZjQUAAHupi4KdYRgisnF5VJ05YXcnUbc537rbVDgc9vl8zrMr1K7beDy+s5IBAAC2wLel\niah9bXV19eDBgyIyPz//3ve+V02qra6uzs3NpVKp+fl5tY91aWnp6NGjaolWTeYVCoVQKGQY\nxsTEhLzWyu727dt2l+NcLheNRjOZzOjoqIhYljU7O5tIJOwxnZaXl/fubwYAAK8ZGhpyu4Td\nV3fN0atUH7uNdF133qbres0NmqbZp5DZ83CaptmPbGxHXDMmAADAbnvDL/3SL7U7K3auoaGh\nn/u5n7v//vtFRG2D1XX9k5/85E//9E87bwsGg4888sib3vSmL3zhCyKSyWR+4Rd+wZ6fGxwc\n/Ju/+ZsvfOELBw8etA8ii0aj7373u1999dXV1dV4PP7JT37yYx/72J7+bQAAoOvd4XYBe62n\np+euu+6655577Cvf+ta3Nt4WDAaDweD58+c3fjQ4OBiPxzd+1NfXp4Z96aWXyuWyZVk1h4wp\nLMUC6GZdsRYGuKeL3rETEcuyYrHYxs2qmqbNzc3VzWEbjY+Pp9Ppmn839ZpdK2MuLy+fe/mL\nW6997zzxd95z3+fa2fylvV74wJnPXfu221U09IH3fs8PXK64XUUzf3a89+Z3/7LbVTT04F8/\nNXTnP3a7ioaWX/0P328tuF1FM3/e8+hXvvIVt6to6J3vfCfBDthVXbQrVkSefvpp0zR1Xbdf\nmLt9+7au66ZpXr58ucVBahqjiMj6+rraPKFOIatWq/Pz86ZpPv300+3+CwAAABrqrmA3NjYm\nIolEwn5hrq+vTx0pUTPftiUXL14UkdHRUXt+Tm2GVV8HAACwN7or2Ck1y6M9PT1qms15sVAo\njI+P+3y+mpPB7BbHznbHExMTXbWiDQAAOlN3BTvV7mRmZkY1EG5kamoqFAqpJVfTNEOhUDKZ\n3NIXqWbFG9umAAAA7J7uCnaRSETX9bGxsbvvvnt8fDyXy6mmJ06lUimRSOi6rl6Yq1Qquq6n\nUil1FoU9M7dxks/p2rVrIvLYY4/t2p8CAABQq7uCnYhMTk4Wi8VMJpNOp6PR6MGDB8PhsPMA\nsStXrohIIpFQK7Y9PT3qJTx1vRXlcvns2bOGYfj9/l34CwAAAOrrumAnIn6/f3R0tFqtrqys\n5PN5EQkEAvbUnYpxvb29vtf09vb+/+3df3Aj533n+S8SR3Jiy6RjhbOpkUnJK2Fc8siApIrE\n0VzZZ8I5e2w3eNmMfODsMpu6kDzgVnOrMbF1kbZxLh17dbnahkd7M3vDA5mrONwd4ERlc8W2\nJesyQG5URxNJaiTAA8kmrJVMeKcUwj9ERFZsK4lxfzxRbwu/BvzZYPP9+sNF9i98wbLkj5+n\nn+9jH7+uWq02MTERCATU/mMAAAB75iAGO5vf71d7worI2bM707nNNM3Dhw/PzMzsyNMAAAC6\nd4CCnXMdq9Pg4KA0dadruf9a5+dXq9VYLCYijz3Wu91fAQCAhx2gYGeapog4e5coagWrpmnO\ny5xv3XWjWCweOnTo5ptvnpmZsZvkAQAA7KUDtKVYuVw+cuSIiGSz2XvvvVetjSiXywsLC4Zh\nZLNZ1VU4n88fO3ZMTdGqwbxcLhcKhUzTVK/NqWG/9fV1O8BVq9VDhw7put7NDCx7xQI4yNhS\nDNhVByjYSasdXZWGTJZIJAzDcF6gadr8/LxKcmqvWHVQrb2Ym5trt8nEgfrzAgAAdx2sYCci\nlUrl2WeftSzLsiwR0XX9s5/97PDwcMNluVxucXFRBbhUKjU6OmqPz5XL5bNnz87OztrBruWr\ne8pB+/MCAAAXHbhg5zqmYgGgZzFTjP3uXW4XsNdqtdozzzxz+fJlNRqn6/onPvEJ9XZd94rF\nYrvmw+qFvM5x+f2fOrGpj9tjrz/7zMTiNberaGv+wcOPvvw/ul1FW4/f/r8u/8uC21V0cvxf\nBY8+dZfbVbRVOnn16NHfdruKtkqlP7g686DbVXRyV2Lx9v/wfberaOvlf3Tz17/+dberaOuB\nBx5wuwRguw7QqlgRqdVq4+PjY2NjdnMTwzBCoVA4HFZrY7sRi8WCwWDLU5VKJRQK7UytAAAA\nm3Swgt2TTz5pWZau6+vr66o13fr6uq7rlmU988wzXT6koeOdrVKpPPTQQztXLAAAwOYcrGCn\n1q7G43F7JcTAwIDaK6zlatnu5XK5oaEhNocFAAAuOljBTlEd7Jy/Nm8skcvlYrGYz+cLh8PO\nnsb2AtiGfSxCoVA6nWYnMQAA4KKDFezS6bSIzM3NVavVDpclk8lQKKSmXC3LCoVCiUSi85PX\n1tYikcgOlgoAALBZByvYRSIRXdenpqYOHToUi8UymUy5XG64plgsxuNxXdc3Njbq9frGxoau\n64ZhqE3G7IG9hkE+tUcFAACAiw5WsBORmZmZQqGQSqVmZ2fHxsaOHDkSDoedO8NeunRJROLx\nuJqx7evrUy/hqeMAAAA968AFOxEJBAKTk5P1en11dVVtHREMBu2hOxXj+vv7fW/r7++3jwMA\nAPSsgxjsbH6/X9O08+fPi8jZs2fdLgcAAGBbDlCwa1jHalOvxzV0p6u3skeFAgAAbMkBCnam\naYqIs3eJovac0DTNeZnzrTsAAIB9wXdwBqLK5fKRI0dEJJvN3nvvvWptRLlcXlhYMAwjm82q\nHWPz+fyxY8fUFK0azFPbv5qmOT09LW+3sltfX7e7HDupsx3+qqVSaVe+HgBg244ePep2CcC2\nHKBgJyKZTKblDhO6rjt7CycSCcMwnBdomjY/P6+SXCwWU/O2mqaptRdO1w12AAAAu+QATcWK\nSCQSWVtbS6VS9sSrrusrKysNO0bMzMxks9loNKp+TaVSdqoTkTNnztinAAAAesc+GLFrXvGg\naZqmaaOjoy0nQ7emWCxuaqfX5pG5XC63uLg4OzuradqpU6dOnDjRsHeZwlQsgIOMuU5gV+3L\nYKc4p0e3Sc2ubupP0RDsmid5NU1bWFhoznalUuk7fW9tr97ddWvthjvfWnW7irZeuuHIT37u\nV9yuoq13/+x7t/34/W5X0cmrv/j6Bz70125X0dYPXvmlo0PfdruKtkprdwz2/dTtKjqp1G78\n6U97t8Ibb7yRYAfsqn0zFetsO6K2+bIsq3mJ69Y09DrZrEqlMjY2lkql1C5k9Xo9m81alvXk\nk0/uSHkAAADd2DfBzsne5uvixYtu1yIisri4KCKTk5P2+JxaYDs1NeVmWQAA4IDZl8FORFSE\nsizLPlIsFpPJpOpCHA6HM5lMwy3FYjGRSKgLEomE3anOnupt7mCcy+VisZh6YIfRwenp6d6f\n0QYAAJ63X4NdQ1dhy7KCwaC9natlWWNjY85sl8vlgsGg3cTEMIxgMNh5JjeZTIZCITVLa1lW\nKBRKJBKbKk/X9c19KwAAgG3Yl8GuWq2q/SFOnTqljoTDYRFZWVlRr7itrKyIiHM1wxNPPCEi\na2trzgvUFKo92ObcN6xYLMbjcV3X1Wtz6q0+wzC63JHiypUrInLy5Mmd+cIAAABdeJfbBXSr\nZdOTEydOqJ8bZkKHh4cbLlaTtq+//rraTGJ4eLjz5OmlS5dEJB6Pqzlf9VafYRiXLl26bleU\narX6xBNPmKa5qf4pAAAA27QvR+w0TVNNgxuaiVSr1WKxaFlW85xpOp0WkWAwmEwmK5VKpVLp\n/BFqVre/v9/3tv7+fvt4B7VabWJiIhAIqP3HAAAA9sy+GbG77uqE5n3AnCKRyHve8565ubl4\nPK7C2Q62wXMyTfPw4cMNW1kAAADsgX05Ytdsbm7OMIxoNJrNZguFwvr6evM1amvXQqGQTqej\n0ahlWV/84hc7P7beSruLq9VqLBYTkccee2ybXwcAAGALPBLsVMe4CxcujIyMdH6zLRAIRCKR\nCxcuSMe+xGpxRpdLJdSVhw4duvnmm2dmZnZ8FBAAAKAb+2Yqthvlctnv91cqlbm5uYZTatOw\nlZUVta4in8+LSDQadV5TrVbtTHb8+HERSSQS58+fV+stcrlcKBQyTbP55blqtRoMBnVd73IG\n9tbaDVv5envopRuOuF1CJ+/+2ffcLqGTV3/xdbdLuI4fvPJLbpfQSWntDrdL6KRSu9HtEq7j\nxht7vUIAu2ff7BXbuc7mrVqV1dVVv98vIvl8/tixY+3Oqtgnb0/XqrPNL+05X8tzVjU3N9du\nk4ne//MCAADP8MhUbCQSSaVS6mdd11dXVwuFgji2phgeHi4UCnbHYHWNSnUicubMmYbROxGZ\nmZnJZrP2cbUOt+U0K1uHAQCAXrAPRuw8plQquV0CAKC1o0ePul0CsC374B27lq2JNU0bHR3d\nwWUKxWJxU/2E200Qq1fxOsflD699ewsV7plvDd3xnbf63a6irVtv2PjpP/grt6to68a/fN97\nfvJ+t6vo5M13v/6RG666XUVbL75119GjP3W7irZKpRt/+ReqblfRyQ//ZuDOG3/V7Sraeumn\nrw0NDbldRVtra2tulwBs176cirUsa2pqamJiolrdmX/DxmKxYDC4/edUKpVQKLT95wAAAGzB\nvgl2zk5yaudWy7JyudyOPLxD35PuVSqVhx56aPvPAQAA2Jp9E+yc1M6tInLx4kW3a/l7uVxu\naGiIzWEBAICL9mWwExG1S6y96FVEisViMplU+7qGw+FMJtNwS7FYTCQS6oJEImE3H7bf4VOn\nnLfkcrlYLKYe2Hl0MBQKpdNpdhIDAAAu2geLJ1qq1Woiomma+tWyrHA4bJ+1LEtlvkgkoo6o\nNQ32BYZhGIaRzWZHRkbafUQymVTjgvYDO7QgXltbU32MAQAA3LIvR+yq1ara8uvUqVPqiEp1\nKysr6iW8lZUVEXG2LH7iiSdEZG1tzXnB4uKiOFa2OreCLRaL8Xhc1/WNjQ37rT7DMNptMkaq\nAwAArts3I3Ytm56cOHFC/dzQXkTtG+akBvBef/11lcCGh4c7dyS5dOmSiMTjcTXnq97qMwzj\n0qVLvEgHAAB6074csdM0Te0DoVKXrVqtFotFy7ISiUTDLel0WkSCwWAymaxUKpVKpfNHqEnY\n/v5+39v6+/vt4wAAAD1o34zYXXeHjOatXZ0ikch73vOeubm5eDyuwplz41cAAAAP2Jcjds3m\n5uYMw4hGo9lstlAorK+vN1+jadrS0lKhUEin09Fo1LKsL37xi50fW29ld74BAADAdnkk2E1N\nTYnIhQsXRkZGOr8DFwgEIpHIhQsXpGNfYrU4o91SCQAAgB60b6Ziu1Eul/1+f6VSmZubazgV\ni8VmZ2dXVlbUuop8Pi8i0WjUeU21WrVnZo8fPy4iiUTi/Pnzar2Faphimub09PQ26/zW0B3b\nfMJuu/WGDbdL6OTGv3yf2yV08ua7X3e7hOt48a273C6hk1LpRrdL6OSHf9Pr72+89NPX3C6h\nE/ZjBXaVr/fnFtV62M51ZjIZZ3MT2+rqqt/vF5F8Pn/s2LF2Z1Xsk7ena9XZ5pf2nK/ltauq\nm2oBAAB2g0emYiORSCqVUj/rur66ulooFMSxNcXw8HChUNB13XmNSnUicubMmYbROxGZmZnJ\nZrP2cbUOl8UWAACgZ+2DEbsOWja30zRtdHS0+wS2S2Ns7R5bKpV29oMAAEA3jh496nYJu85r\nwU7ZVCuTvQ92V/+mf2c/a2fd9Qsbn/v3vfsSzFf+8dCr733G7Srauu1HJ377Sou3AnrHH9yb\n/vDK192uoq1vHXug/1Mn3K6irY1nn7nz+/NuV9HJSzdPHH2qd9+hLJ28+uNXf97tKtr6xdv+\n7sJ/v+J2FZ3E/vdjR4/27j+/pdIDrzz+oNtVtPWhRxcPQrDzwlSssxeJ2vvLsqxcLrep23e1\nQgAAgD3ghWDnpPb+EpGLFy+6XQsAAMCe8lqwExG1z5i9bEJEisViMplUO4OFw+FMJuO8Xh13\n/lyr1WKxmM/nSyaT9mW5XE4dDIfDLYcDM5lMOBz2+XwNzwcAANgbnupjp9RqNRHRNE39allW\nOBy2z1qWpTJfJBJp94T5+XnV/eTw4cPqSDKZtHeJVU/QdX1mZsa+xdkbZWxs7Nq1azv4jQAA\nALrhtRG7arWqNo04deqUOqJS3crKinqXbmVlRURaNr2zbWxsbGxs1Ot1Ff6KxWI8Htd1XR1U\nr/EZhmHvS5HP5w3D0DRtbW2tXq+vra1dvnx5V78mAABAMy8EO5/DoUOHVMY6ceLvF9apPKc2\nnBAR+4cOxsfH1XyucunSJRGJx+PqoP0anzouIsvLyyJimqbao2JwcNA5mAcAALA3vDYVa/ex\ncyYzEalWq6+99lqlUvnzP//z6z7EblysqBjX39/YoyQej6vtxdQFzrs671cLAACwG7wQ7K7b\nrKR5czAAAADv8cJUbGdzc3OGYUSj0Ww2WygU1tfXt/aceis7WyoAAMB2eD/YTU1NiciFCxdG\nRka2NkOqVmPYSyW6uaBcLm/hgwAAALbDC1uKdf4K6prV1VW/31+pVNQAnvMu50NaPjCfzx87\ndkzTtPPnz6vlEblcLhQKmaap3rErFovBYNC+oFKpPPTQQ6qpCnvFAgDQIw7ClmLeD3aZTKZl\ncxMV9aSLYCet3tJr2I7W2ehORFKplBop3Nd/XgAAsL94fyo2EomkUin1s67rq6urhUJB3rk1\nxXXNzMxks9loNKp+TaVSzlQnItPT00tLS6orcjqdnpyc3LEvAAAA0J39PWK3HzEVC+AgOwhz\nYYCLvNDuRLH3e7VFo9GPf/zjJ06caOhpt+WHdw7BuVxucXFxdnZW07RTp051+NxP/8W/22Y9\nu+prv/ZPbjB7tw/fW/HiH/7+d9yuoq3f+p1bb/78M25X0cn3nzxx7upn3a6irdN3ffUD3+rd\n//r94MPFqwufcbuKTu4af/rcuXNuV9HW6dOn3S4B8DgvT8XOzs6OjY2Nj49Xq9Xd/qxMJhMK\nhdQOs5Zlqc9Vu9YCAADsDa8FO2eTufX1dV3XLcva7f//WqlUxsbGUqmU2ky2Xq9ns1nLsp58\n8sld/VwAAAAnrwU7p4GBgZmZmWg0ahhGhy5027e4uCgik5OT9tzryMiIvN1CDwAAYG94Odgp\naimrc4vYYrGYTCZ9Pp/P5wuHw5lMxnm9Ol6r1WKxmM/nSyaTLR+byWR8Pp9aWjs9Pc0aFAAA\n4DrvLJ5o59ZbbxWR559/Xv1qWVY4HLbPWpalwlkkEnHeNT8/r16YO3z4cPMzVW88XddVf5Nm\n6u06Xdd35CsAAAB0w/sjdmp6VKU0EVGpbmVlRb0Mt7KyIiLNHYw3NjbUC3MNgU8cqW5mZqbd\nh165ckVETp48uXPfAwAA4Dq8P2LXoGHOdHh4uOVl4+PjLZuVdJPqqtXqE088YZrm1ramBQAA\n2JoDF+yUarX62muvVSoV57t3Tmq3sQb2xmLj4+Ptnlyr1SYmJgKBgNpGFgAAYM94fypWNbFz\nvu6WSCQOHToUDAbD4XDDDrCdGYahXqo7e/Zsu2tM0zx8+HCH8TwAAIBd4v1g98orr4jIRz7y\nEfXr3NycYRjRaDSbzRYKhfX19e4ftbq6ev78eRGZnZ3N5/MNZ6vVaiwWE5HHHntsZ0oHAADY\nDO8Huy9/+csics8996hfVW+5CxcujIyMbPYdOL/fPzg4uLS0JCKPP/6481SxWDx06NDNN988\nMzMzMDCwM6UDAABshs8zDdiat3OtVqvnzp0zDMO51kFdtrq66vf7K5WKGsBz3thyW9iGg7FY\nbHZ2dmlpSc3MVqvVQ4cOdV5RYSuVStv9qgCwbx09etTtEgAv81qwa6Zp2vz8vD2Kppa1Nl+m\nop50F+wqlcrQ0JCIbGxs9PX1zc3NtdtkwjN/XgAA0Pu8PBWr6/rS0tLS0pJzbjQSiaRSKfuC\n1dXVQqEgIqpNcZfsCdn5+Xlh6zAAANAb9veIXTgctixL0zQVs3pHsVhs9wIfU7EADjKmYoFd\ntY+DXbFYDAaD6udCodA73YDVG3jt/rClUumpv35qj0valJO/dPJ73249r90LfuWO+p0/2MTw\n6h576QPa0aNfd7uKTkqlByb/j//odhVtzf13//CDgd79B+S7xZPf/71Pul1FJzf/7qVz5865\nXUVbp0+fJtgBu2ofT8Wq3sLpdNr+uUfY25cBAADspX0c7KampjRNi0QimqbxlhsAAMB+DXaq\nP/CpU6fs/2zuGCwimUwmHA77fL5EIqG2oOj+bC6Xi8ViPp8vHA7ncjnnKZ/P5/P5qtVqMplU\nF9hrL+zFueqaHfiqAAAA3dmvwe6rX/2qvN12WP2nOuKUSCTGxsZU5DIMY2JiolardXk2mUyG\nQiE1qWpZVigUSiQSDc+fmJiIx+PqAme2AwAAcMW+DHa1Wk1t26o6z/n9fk3TDMNwJrN8Pq9a\nE29sbNTr9VQqZVnWM888083ZYrEYj8ftsxsbG7quG4ZRLBadZQQCAXVBNpsVkaefflocjevq\n9fr+XZgCAAD2o30Z7L75zW/K2zOwivpZHVeWl5dFZHx8vK+vT0Q+//nPi8jFixe7OXvp0iUR\nicfj6mxfX58amVPHbfbtIyMjwpoJAADgtne5XcBWNGz/KiIPPPCAOj48PKyOqCimhvREpK+v\nzzl+1s3Z/v7+hs+Nx+PT09P2r/btAAAAvWD/jdhVq1U1NnbkyBHf29QGX7Ozs81rIAAAAA6I\n/Rfs/uzP/mzLZzel3spOPRwAAGDH7b9gp9YorK+vN0Su9fV1+6yI6LouIuVyWf1aq9VUX5Ju\nzpqmKSINSyUAAAB63D7bUqxSqQwNDZmm6XzXzZZIJAzDWFtbGxwczOVyoVBI13W1BiKTyYyN\njaXT6UgkIiKdz+bz+WPHjmmadv78+cHBQft6+3NVgzrnn855RP28vr4+MDDQXCR7xQI4yNhS\nDNhdLScce5baQKxQKLQ8WygURCSdTqtf1bCcTdM01Z1kC2fVBfYwYfOfznkkGo3at+zgdwcA\nAOhsn43YqdnSpaWlLi9QQ3EiYprm+Ph4wxBa57O5XG5xcVEt1EilUqOjo/YFnUfsyuXy2bNn\nZ2dnNU3rUCoAAMDO2mfBDgAAAO3sv8UTAAAAaIlgBwAA4BEEOwAAAI8g2AEAAHgEwQ4AAMAj\nCHYAAAAeQbADAADwCIIdAACARxDsAAAAPIJgBwAA4BEEOwAAAI8g2AEAAHgEwQ4AAMAjei7Y\n1Wq1TCYTi8V8Pp/P50skErlcbo9rKBaL23+Iqn/7zwEAAOiSr16vu13Df1ar1cbHxy3Lajiu\nadrCwkJfX98e1BCLxWZnZ7f/Z1Gprqf+vAAAwNt6K9jNzc1NTU3pun769OmBgQERqVar586d\nMwwjnU5HIpE9qGGnAlm755RKpW0+GQCA3nT06FG3SzjoeivYqTC0sbHhHJyr1Wr9/f2yV6Nf\nexDs/uKj/3ybD99Vv/aNf/Nz/0x3u4q2fvZvjdtSD7tdRVuvTj1x57e/5HYVnbx0xxc+ctPX\n3K6irRff+PTRo993u4q2SqWbP/ChZber6OQHrxz/1i/twMsku+R0144xAAAgAElEQVTDfx04\n9/o33K6irdPv/+g//WrV7So6+fJnB/61/oLbVbT1L4y7CXau67l37ESkYcq1r6+vXq/bCcl+\ndy2Tyfh8vnA4nMlkmh+Sy+XUi3rhcLjlW3qZTCYcDqvX+KrVqv3whk+xf67VauqByWRSHS8W\ni8lkUp1tVwYAAMCe6a1gl06nRWRubs5OWu1YljU2Nmb/kEgknGeTyWQoFJqdnVUXhEKhhgsS\nicTY2Jh6mc8wjImJiVqt1vkT5+fn1QMPHz6sHhsMBuPxuLMesh0AAHBRbwW7SCSi6/rU1NSh\nQ4disVgmkymXyy2vnJubW1tbq9fra2trmqYZhpHP59WpYrEYj8d1Xd/Y2KjX6xsbG7quG4Zh\nr3XN5/OGYdgXpFIpy7KeeeYZccycOscIlY2NDXW9etUvHA6LyMrKirpyZWVFRFTWBAAAcEVv\nBTsRmZmZKRQKqVRqdnZ2bGzsyJEj4XC4uf/IzMzM4OCgiAwODs7MzIjI8vLfv/hy6dIlEYnH\n42pKt6+vT42rqeP2lePj4+qCz3/+8yJy8eLFzoXZ1ysqzw0PD6tf7R8AAADc0nPBTkQCgcDk\n5GS9Xl9dXV1aWhKRYDDYMHQXCAQafrZnRdUP/f39vreptRcNF/j9fvWreodPfVAH9vVO1Wq1\nWCxaltUw1QsAALD33uV2AZ34/X6/3x8IBCzLOnv27IULF9yu6B0SiYRhGG5XAQAA8Pd6aMSu\n3VYNaspVLVzoXr2VnSlURETm5uYMw4hGo9lstlAorK+v7+DDAQAAtqCHgp1pmiLS3JpErVfV\nNM150Dkzq97AS6VSzud02BZM13XnE2q1mupXsqlqp6amROTChQsjIyPOeWEAAAC39FCwU9Et\nFArlcjm7+Ui5XFZB7eGH39GTNh6PVyoVEalUKur9tvvuu0+dOn78uIgkEgl1gYjkcjln/7lP\nfOITIrKwsKA+Ra2HPXXqlPP51+23Ypenajh37twWvjIAAMAO6qF37Px+fzqdHhsbC4VCDad0\nXR8ZGXEeCQQCQ0ND9q+madrDZsPDw6q/iXPPWU3TxsfH1c8jIyPqAvsNOU3TTpw4oX6ORqOz\ns7OHDh3SNK3DigpV6pEjRxqOl8vllsssAAAAdltvbSkmIpVK5dlnn7UsS8UyXdc/+9nPOpuJ\n2Ft1JZPJeDwejUYffPDBhtgnIrlcbnFxUb2Zl0qlRkdH1eaztkwmo9rOmaY5Pj5uny2Xy2fP\nnp2dnbWDXbvNwdTOtqrI8fHxH//4x8Fg0DTN6enpDnexVywAwKvYUsx1PRfsrmun9nIFAADw\nmB56xw4AAADb0UPv2DVo2fpELWjt8t56va5mS+3p0WZqPjeVSk1OTrYcC6zValeuXHnhhRfs\n/samad59993Nk79dYioWAOBVTMW6rnenYlsGO1vnsp0RLRaLzc7OZrPZlu/hhUKhaDSqWh83\nB7tisZhIJJyLMGyaps3Pzze8t9eNUqn0jbt+c7N37aWPXv2jd42fdLuKtv524anbUg9f/zqX\nvDr1xJ3f/pLbVXTy0h1f+MhNX3O7irZefOPTR49+3+0q2iqVbr7hlt7964nIW//p01dv6K1e\n7k53vRU7V33C7SraOj3w8KeWe3oboWePz9z1byfcrqKtq/9snmDnul6fim3oMJxOp6XrcTvl\nkUceEZFQKNTQwaRararlt+qCZsViMRgMWpaVSqVWV1ftGlZXV03TtCxrYqJ3/+kCAAAHUK8H\nuwaRSERENrWR1+DgoFrc2tBqTv26tLSkdrZoptrjFQqFyclJZwcTv98/PT2dSqUsy2pupwwA\nAOCWfRbs2slkMuFw2OfzZTKZ5rOapjV0trMsyzAMXdcbNrSw5fN5y7Kc7fEajI6OptPpW265\nRf2q9kOrVqvJZFLtY9FyAhcAAGD39O7iiZZUbmuYik0kEvYY3tjY2LVr15pvPH36tGEY4XC4\nXq/XajW1gdjp06fbfdDy8rI07WPmNDAwoIYPnSYmJlSeU334lpaWOjwBAABgZ/X6iJ3vnVRL\n4cnJSfuCfD5vGIamaWtra/V6fW1t7fLly83PGRgYyGazImJZltpDLJvNdlj6oNbAbnYPiUAg\nsLGxUa/X1Wc9/fTTm7odAABgO3o92DXQdX1tbc35VpwaWjNNUx0cHBycmZlpea/aSSwcDo+N\njTXvUdYNXyvOC8bHx/v6+tRniYja9wIAAGBv9Hqws9eiRqNRTdNOnjzZsNaheWit3Vtx9sXO\nH3YWu8QCAAAX9Xqws505c8ayrGAw2NC1ZFPUcJrzh3ZM0xSRcrnsPNjQe2XLZQAAAOyGfRPs\n/H6/ClsLCwt78HHHjx8XkZav6wEAAPSmfRPsROTBBx8UkXg8XqlU7IMq7RWLRftIwzDb1gwP\nD2uaNjU15XyyU61W2/6nAAAA7KD9FOwGBwdVjFtcXLQPfvKTnxSRRCKh0l6lUtmp9+fUIoxg\nMJhMJp1hsVwuZzKZ/v5+6dgPBQAAYI/1+l6xDeXVajWVqJxrY5PJpDPMpVKpqamp5ns7PLbd\nqXK5HI/H27UaTqVSduOV5nvbfVCpVGr5NAAA9jv2inXdPgt2IpLJZMbGxkzTnJ6etg9aljU3\nN2dZVjqdjkQim0pv1z2Vz+eXl5cvX76sEp6u6/fdd9/999/vbIPXfbADAADYJb0b7AAAALAp\n+2BLsYYmwIqu6+Pj49ftG2cPm83NzU1NTTWM8zmp+Vw1u9ow2NZcQJef3hJTsQAAr2Iq1nX7\nYMSuZbBTVldXO6crZ0SLxWKzs7PZbLZ5z4lcLhcKhaLR6IULF6RpFrVdAQ17YHSpVCp9467f\n3Oxde+mjV//oXeMn3a6irb9deOq21MNuV9HWq1NP3PntL7ldRScv3fGFj9yUdbuKtl58I3T0\n6PfdrqKtUunmm269w+0qOnnjO9/+1g1tN0t03Yffqn65/DO3q2jrn/p/7gt/uOZ2FZ186beG\nfut3nnO7irb+8Pc/RrBz3b5ZFdvQHDidTssme9o98sgjIhIKhRpaHFer1VAoZF/QTQHq0+fm\n5rbwRQAAAHbJvgl2DSKRiIgYhtH9LYODg0tLSyJy7tw553H169LSUvfDb1v4dAAAgN22X4Nd\nO5lMJhwO+3y+TCbTfFbTNF3XDcOwO5hYlmUYhq7rW+hI57zF5/M1zNg2HwEAANhV+2DxREsq\nt+m67jyYSCTsUbSxsbFr164133j69GnDMMLhcL1er9Vq4XBYHdzCpz/8cO++6QUAAA6gfRPs\nWo5+2f2BRSSfzxuGoWna+fPnBwcHK5XKQw891HzLwMBANpsNhUKWZb355psiks1mnR3puiwg\nnU43L8IAAABw0X6ditV1vWFR6vLysoiYpqkODg4Oqj3Bmo2MjOi6Hg6Hx8bGdF3fWj578cUX\nnVvWAgAAuG7fBDt7RWo0GtU07eTJkw1rHdSuYs7uJ4FAoN3T7C3Iut9Y1rkqtlAoFIvFoaEh\n5x6yAAAA7to3wc525swZy7KCwWBD15JN6evra/hhUwKBwPnz52WT/VYAAAB21f4Ldn6/3zRN\ncTtUqfFCOp4AAIDesf+CnYg8+OCDIhKPx51vuam0VywW7SO7Ok9aq9XknR1PnJiiBQAAe2/f\nbCnWUKfa2tW592uxWAwGgw2rYlW/upbfseVjW55qvrJWq5mmaRhGOp1WzYqd+5XZZ1s+n71i\nAQBexZZirtuvwa5Wq/X398s7N2xVaa/5Cbquj4+PN+wq2/xY+4jzlP1BzTRNW1hYUG/pqd1m\n7VOmaapKev/PCwAAPGNfTsWKSF9fn9qwdXFx0T5oj941MAzjyJEjW5geLRaL4+PjLU/dc889\nZ8+etddejIyMLC0tqZnZVCrVrhIAAIDdsw9G7Dal5fBeJpNRLevadbZrea+a2xWRVCr18Y9/\n3B7wK5fLlmXF43FN09Tms5vCVCwAwKuYinXdgQh2HY53uCYcDluWVSgUWvbDm5ubm5qaUi/V\nbarCUql0/q7PbeqWPfbQ1a+8+usRt6to67Y/yfx88lG3q2jr76Yfv+0bf+x2FZ28+tHfuPPd\nP3K7irZe+sl7e/l/GEql0k033eR2FZ288cYbr732mttVtPWrv/qrX/nKV9yuoq3Pfe5zjz7a\nu/96EZHHH3/8U5/6lNtVtPXss8/28j+/B8R+nYrdEZlMJhwO+3w+tferUz6ftyzLNM12XY5H\nR0fT6fQtt9xSq9V8Pl8sFmu4oN1xAACAXbJv9ordDpXbdF13HkwkEnYXurGxsWvXrjnPqg3K\n2nUzEZGBgQG1HlbeXipx5swZ5/qMb37zmyLymc98Zme+AwAAwPV4c8TO905jY2MiMjk5aV+Q\nz+cNw9A0bW1trV6vr62tXb582fmE5g3KOvjkJz8pIg1PuHr1qojcf//92/42AAAAXfFmsGug\n67qzK4q8PSBnmqY6ODg42HldheJrRUQCgUA0Gp2amlJdi5WpqaloNDowMLDz3wcAAKAVbwa7\n+tui0aimaSdPnnSmOmk1INfuXbouqc0wnnvuOfWraq3CPCwAANhL3gx2tjNnzliWFQwGq9Xq\npm5UG5Q1tL6rv5Pz1L333isic3Nz6tfV1VVhHhYAAOwtjwc7v9+vItrCwsKmbjx+/Lg0vTbX\ngWqYbFmW2qz26aefZh4WAADsMY8HO3l7kjQej1cqFfugSnsqhCkNg3PDw8Oapk1NTTmvcXK+\nTqc88MADIvLUU09VKpXZ2VnmYQEAwB7zfrAbHBxUMc65+Zhax5pIJFTaq1QqzZvMquUUwWAw\nmUw6Y1+5XM5kMmoDWWc/lMHBwWg0ahjGQw89JMzDAgCAPef9YCciExMT8s5Bu0AgYJqmZVlD\nQ0M+n29oaKi5ZV0gEFhdXdU0LR6PHzlyxF4Ge+TIEdU/JZVKNWwpFo1GRcSyLOZhAQDA3jsQ\nwU69ACfvHLSbnp5eWlpSeS6dTju73Nn8fv/S0tLKyoppmnby03V9aWlpfX29+ZZbb71V/cA8\nLAAA2Hte2yvWXcViMRgMisj6+nq7EbtSqbS3RQEAsEfYK9Z1B2JLsT3z1FNPiUg2m+0wD8t/\n6QEAwC4h2O0MtQWFiOi6PjIy4m4xAADgYPJssLOTlpOu6+Pj49fdAVbda09SNz+q+TmaplmW\nZZrm9PR054czFQsA8CpmpVzn2XfsWgY7ZXV1tXO2u26wUxr2n+1SqVQ6f9fnNnvXXnro6lde\n/fWI21W0ddufZH4++ajbVbT1d9OP3/aNP3a7ik5e/ehv3PnuH7ldRVsv/eS9t99+u9tVtPXy\nyy//7Jc/6HYVnfzcD7/7p3/1HreraOsT73vz4f93w+0q2nriv+wPn1t1u4pOlk4fuWf8S25X\n0dbzC18g2LnO46tiGzYBU2tjN7sLRfOj1HPsDcQAAAB6gceDXYNIJCIihmH0yHMAAAB20MEK\ndu1kMplwOOzz+TKZTPd3OXsaq97FzrPNRwAAAHaVZxdPtKRym67rzoOJRMIeexsbG7t27VqX\nz3n44Yd3oUYAAIAt8niwazlm5twxIp/PG4ahadr58+cHBwcrlYra6fW6j0qn07Q1AQAAPeVg\nTcXqut6wlHV5eVlETNNUBwcHB2dmZrp51IsvvmjvPAsAANALPB7s7HWs0WhU07STJ082NCiJ\nx+Mi4ux+EggEOj+qXq8XCoVisTg0NFQul3e1fgAAgO55PNjZzpw5Y1lWMBisVqvbf1ogEDh/\n/rxstXMKAADAbjgowc7v95umKTsXxdTIHx1PAABA7zgowU5EHnzwQRGJx+POd+NU2isWi/aR\nLmdXa7WavLPjiRNTtAAAYO95fEuxhm+XTCbj8bhzR9disRgMBhtWxVqWJU1bijkfVavVTNM0\nDCOdTqtmxbFYbHZ2NpvNjoyM2GebCxD2igUAeBdbirnuYAW7Wq3W398v79zmVaU9+5pUKjU1\nNSVd7BWradrCwkJfX5+I5HK5UChknzJNUz3Tq39eAADQgw7QVKyI9PX1qW1eFxcX7YPT09NL\nS0tqUjWdTju73LWjaVo6nbZTnYiMjIzYD0mlUvaIIAAAwJ7x7Iid0jzYpmmapmmjo6MDAwOb\nelStVnvmmWcuX748OzsrItFo9OMf//iJEyfsbNclpmIBAF7FVKzrDlywUzRNm5+f7z7b5fP5\nY8eOtTxVKBTatb5rqVQqPRX9te6v33snZ//iO//qXreraOvWf3nlpj/+hNtVtPXGb/zpndee\ncLuKTl46/PBHbvqu21W09eIbH7z99tvdrqKtl19++Rdu/AduV9HJ3/z0L1/9Vu/+W/22D/v+\n6A9W3a6ird/87SOPnfyq21V08sWnPnvqrt91u4q2Ll79PYKd6w7EVKyzt/DGxoau65Zl5XK5\nLm8vFosq1aXT6fX1dfWc9fV1NasbDAadi2oBAADcciCCnVNfX59a1nDx4sVurq/VasFgUEQK\nhUIkErEH+QYGBiKRSKFQEJFEIqG6nwAAALjowAU7EVFvxameJiLi8/kaZmydR5577jkRSafT\nLedbA4FAOp22LOvKlSvOe2u1WiwW8/l8yWRy974IAACA00EMdp17CzdQrYYfeOCBdhfcc889\nIvLCCy84D87Pz6s1FocPH95OqQAAAN07cMGuWq2q3SZOnTrVzfVq3tZuetfM7/fbl9k2NjY2\nNjbq9bpqXwwAALAH3uV2AXuhZdOTEydO7N4njo+Pb7YNCgAAwDYdiGDnZPex29XgpYbxAAAA\n9tKBCHbb6dWn67phGJVKpd1sbKVSERE1vQsAAOCiA/eO3Wbdd999IvL1r3+93QXqFEN0AADA\ndQS7RmoZrO1jH/uYiIyNjbXsQlwsFsfGxjRNU5cBAAC4iGAn0WhURNRGFLVabWFhwXm2r69P\ndSEOBoOZTKZararj1Wo1k8mo3sWPPvooSyUAAIDrDsResZ2/Yy6XC4VC9q+maareJc67isWi\nynDNVlZWhoeHu//EUqnUbfUAAOwr7BXrOoKdiIhlWXNzc5ZlpVKpycnJlnfVarXnnnvu6aef\nVp2Ho9HoZz7zmY997GMNY3VdfiIAAMCO83iwAwAAODi80+6kZRdi1bJuYGCgw2Uiouv6+Ph4\ny5WttVrtypUrL7zwgr23hGmad99998jIyNbqZCoWAOBVTMW6zjsjdi0Tm4homjY/P29nu3aX\nicjq6mpDtisWi4lEwrKs6z62e6VS6fRdv7vZu/bSuau/t/Df/N9uV9HW+P/1X3/997/ldhVt\nPfA7H379G3/rdhWdvP+j77rZ9ytuV9HW9+vfu/32292uoq2XX375hg+6XURHb31Xrr77ittV\ntHXXT+79X779JberaOuRO77w6X8/4XYVnXztH8/fdebTblfR1tWzXyPYuc5rq2LrDhsbG7qu\nW5alVry2u6xer6fTaRFpWA+rFkyoF+9WV1fti1dXV03TtCxrYqKn//kHAAAHjdeCnVNfX5+a\nP7148WLnKyORiIgYhuE8mEgkRKRQKExOTjpH8vx+//T0dCqVahkZAQAA3OLlYCciaslqy7nU\nzvL5vGVZpmkGAoGWF4yOjqbT6VtuuUX96vP5fD5ftVpNJpM+ny8cDm/hQwEAALbD48GuVquJ\niKZpnS/LZDIiouu6fWR5ebnzjQMDA5FIpOGdvImJCTVGaFkW2Q4AAOwxLwe7arVqmqaInDp1\nquGU753GxsZEZHJy0r5A5bPN7gAbCAQ2Njbq9Xo2mxWRp59+eptfAQAAoHteC3bOuHbo0CHD\nMDRNO3HiROe7dF1fW1sbHBzs/uE25wXj4+Nq8lc1Q1GtjAEAAPaGd/rYNbP72DVv5Go3eYnF\nYteuXTt58uR1U103NjvCBwAAsIO8Fuw225bvzJkzR44csSxrfX3d2ZRO7RhbLpedWa3h4R1a\n4gEAAOw9r03Fbpbf71fv4TU0sTt+/LiIXL582Z2yAAAANu+gBzsRefDBB0UkHo9XKhX74PDw\nsKZpU1NTxWKx5V1qvS0AAEDv8NqWYtf9Oi0vSyaT8XjcNM3p6Wn7oNp5QkRM09Q0zZ6TLZfL\nzz//vFpIq2na0tJSy8e2q4e9YgEAXsWWYq4j2ImI1Gq1/v5+EWlYG1sul+PxeLt2dKlUyu6Q\n0n2wAwAA2CVMxYqI9PX1qe1iFxcXncf9fv/S0tLKyooatFMHdV1fWlpaX1939r0DAABwnXdW\nxTaPjTWvWtU0LZVKjY6ONlxTr9ffeOONdk8eHh4eHh4WEcuynKN0IlKr1a5cufLCCy/YTzNN\n8+677x4ZGWk3VsdULADAq5iKdZ13pmKbtWtHomna/Py8am7inDCNxWKzs7PZbFa1F3bK5XKh\nUCgajV64cME+WCwWE4lEy4la50c0KJVKf/LRL271O+2FX//GY3/16SfdrqKt933t8+/5d992\nu4q23vwndwx+90a3q+ik8sGf3nnTLW5X0dZLb/yn22+/3e0q2nr55Zc/8KG/c7uKTn7wys9f\ne1/vbmZ4+K+0p9b/B7eraOvkof/t0T8Nu11FJ49/Yunkv/mv3K6iraf++f9DsHOd96di6w4b\nGxu6rluWlcvlnGfVz4888oiIhEKharXqfEK1Wg2FQvYFilpaocbwVldX7Y9YXV01TdOyrImJ\niT36hgAAACJyEIKdU19fn9oE9uLFi81nBwcH1RLXc+fOOY+rX5eWlpzrKhKJhIgUCoXJyUln\nE2O/3z89PZ1KpZzxEQAAYA945x27Lqntxez504a1q5qm6bpuGMZ9992nVktYlmUYhq7r9uIJ\nEcnn85ZlmaYZCARafsro6OhNN910yy29O+EFAAC852CN2MnbjYWdKa3B6dOnRSQcDquL1Q/q\noG15ebnzQwYGBiKRCFvHAgCAvXSwRuyq1aqaVz116lS7awYGBrLZbCgUsizrzTffFJFsNtuw\nDELN55LbAABAT/F+sGvZ9OTEiRMdbhkZGdF1XY3V6brevEi2yw8SGhQDAIA9dLCmYlUfu/n5\nefWmXQdqTM75AwAAQI/z/ojd1sbM7OTXMgKaphmPx8vlsnM2tuGD2nXRAwAA2CUHa8Rupxw/\nflxELl++7HYhAAAA/xnBbiuGh4c1TZuamioWiy0vUGtvAQAA9hLBbotmZmZEJBgMJpPJcrls\nHy+Xy5lMpr+/Xzr2QwEAANhx3t8rtvMX7HDNdW8vl8vxeLzlXrEikkqlJicnm4+XSqWOVQMA\nsF+xV6zrCHZbD3ZKPp9fXl6+fPmySni6rt933333339/Q+s7AACA3eblYAcAAHCgeLPdSXOr\nEV3Xx8fHu9krYlP31mq1K1euvPDCC3a7O9M077777g49jZmKBQB4FVOxrvPmiF27HnJra2uD\ng4M7dW+xWEwkEi3fsdM0bX5+vuVsbKlUemz8ns41uOuLC88/+z/d73YVbX3qf/6z7/+HtjvC\nue7mf3Txva8/6XYVnfzo/Z//h7+y4XYVbf3H7/V/6EMfcruKtl555ZUfHfplt6vo5L3rP/yj\nt37odhVt/eYNv/zf/n+X3K6irf/zv/jkR8/9a7er6OQbp//FXbFxt6to6+qFBYKd67y8Krbu\nkE6nRWRubm6n7i0Wi8Fg0LKsVCq1urpqX7y6umqapmVZExMTO/6NAAAAOvBysHOKRCIiYhjG\nTt2bSCREpFAoTE5OOmdp/X7/9PR0KpWyLCuXy22raAAAgM04KMFOcTaW8/l8Pp+vVqvFYjGf\nz5dMJru/N5/PW5ZlmmYgEGh58ejoaDqdvuWWW3akbAAAgG54c/FEs0wmIyIPP/xww/H5+fnZ\n2VkROXz4cPf3Li8vS8f+wwMDA2qcDwAAYM94Odg1LINIp9PNi1U3NjY2Njb6+vo2da9aA9vN\nGlsAAIA9c4CmYl988cVKpdJwcHx8vDnVdXlvA18rWy8XAABgk7wc7JwrWwuFQrFYHBoacu7r\nKu1H3bq5FwAAoKd4Odg5BQKB8+fPi8jCwsL27zVNU0Qacl79nXagaAAAgM04KMFORFR74a11\nPGm49/jx4yJy+fLlnasOAABguw5QsKvVatJxKWv39w4PD2uaNjU1VSwWO1wPAACwl7y8pZjz\nq9VqNdM0DcNIp9OqEUnzNd3fK2/vPCEipmlqmma/q1cul59//vmxsTER0TRtaWmp4fnsFQsA\n8Cq2FHOdl4NdM03TFhYW1DLYzsGu871KuVyOx+Mt94oVkVQqNTk5uYXiAQAAtuagBDtN006d\nOnXixAk7mXUf7Jrvdcrn88vLy5cvX1YJT9f1++677/777x8YGNiR7wIAANAlbwY7AACAA+gA\nLZ4AAADwNoIdAACARxDsAAAAPIJgBwAA4BEEOwAAAI8g2AEAAHgEwQ4AAMAjCHYAAAAeQbAD\nAADwCIIdAACARxDsAAAAPIJgBwAA4BEEOwAAAI8g2AEAAHgEwQ4AAMAjCHYAAAAeQbADAADw\nCIIdAACARxDsAAAAPIJgBwAA4BEEOwAAAI8g2AEAAHgEwQ4AAMAjCHYAAAAeQbADAADwCIId\nAACARxDsAAAAPIJgBwAA4BEEOwAAAI8g2AEAAHgEwQ4AAMAjCHYAAAAeQbADAADwCIIdAACA\nRxDsAAAAPIJgBwAA4BEEOwAAAI8g2AEAAHgEwQ4AAMAjCHYAAAAeQbADAADwCIIdAACARxDs\nAAAAPIJgBwAA4BEEOwAAAI8g2AEAAHgEwQ4AAMAjCHYAAAAeQbADAADwCIIdAACARxDsAAAA\nPIJgBwAA4BEEOwAAAI8g2AEAAHgEwQ4AAMAjCHYAAAAeQbADAADwCIIdAACARxDsAAAAPIJg\nBwAA4BEEOwAAAI8g2AEAAHgEwQ4AAMAjCHYAAAAeQbADAADwCIIdAACARxDsAAAAPIJgBwAA\n4BEEOwAAAI8g2AEAAHgEwQ4AAMAjCHYAAAAeQbADAADwCIIdAACARxDsAAAAPIJgBwAA4BEE\nOwAAAI8g2AEAAHgEwQ4AAMAjCHYAAAAeQbADAADwCIIdAHqqqdwAAAGISURBVACARxDsAAAA\nPIJgBwAA4BEEOwAAAI8g2AEAAHgEwQ4AAMAjCHYAAAAeQbADAADwCIIdAACARxDsAAAAPIJg\nBwAA4BEEOwAAAI8g2AEAAHgEwQ4AAMAjCHYAAAAeQbADAADwCIIdAACARxDsAAAAPIJgBwAA\n4BEEOwAAAI8g2AEAAHgEwQ4AAMAjCHYAAAAeQbADAADwCIIdAACARxDsAAAAPIJgBwAA4BEE\nOwAAAI8g2AEAAHgEwQ4AAMAjCHYAAAAeQbADAADwCIIdAACARxDsAAAAPIJgBwAA4BEEOwAA\nAI8g2AEAAHgEwQ4AAMAjCHYAAAAeQbADAADwCIIdAACARxDsAAAAPIJgBwAA4BEEOwAAAI8g\n2AEAAHgEwQ4AAMAjCHYAAAAeQbADAADwCIIdAACARxDsAAAAPIJgBwAA4BEEOwAAAI8g2AEA\nAHgEwQ4AAMAjCHYAAAAeQbADAADwCIIdAACARxDsAAAAPIJgBwAA4BEEOwAAAI8g2AEAAHjE\n/w9CPUk7TwVT+QAAAABJRU5ErkJggg==",
      "text/plain": [
       "plot without title"
      ]
     },
     "metadata": {
      "image/png": {
       "height": 420,
       "width": 420
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "display.brewer.all()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### gene function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Sys.which(\"python\")\n",
    "Sys.setenv(RETICULATE_PYTHON = \"/home/wg_looking/anaconda3/bin/python\")\n",
    "library(reticulate)\n",
    "source_python(\"/NAS/wg_looking/gRNA_ONT/code/NCBI_gene_summary.py\")\n",
    "setwd(\"/NAS/wg_looking/gRNA_ONT\")\n",
    "load(\"stat/test_genes.RData\")\n",
    "\n",
    "\n",
    "ENSEMBL_ID_Entrez_ID_all <- read.table(\"/NAS/wg_looking/gRNA_ONT/database/ENSEMBL_ID_Entrez_ID_all.txt\",\n",
    "                                      header=T)\n",
    "rownames(ENSEMBL_ID_Entrez_ID_all) <- ENSEMBL_ID_Entrez_ID_all$gene_id\n",
    "\n",
    "\n",
    "sub <- sub[1:2,]\n",
    "\n",
    "\n",
    "genes <- setNames(as.character(sub$gene_id), as.character(sub$gene_name))\n",
    "genes <- sort(genes)\n",
    "genes <- genes[!duplicated(genes)]\n",
    "gene_function <- lapply(genes, function(gene_id){\n",
    "     Entrez_ID = ENSEMBL_ID_Entrez_ID_all[gene_id, \"ENTREZID\"]\n",
    "    gf = NCBI_gene_summary(as.character(Entrez_ID))  \n",
    "    return(gf)\n",
    "})\n",
    "\n",
    "pp <- melt(gene_function)\n",
    "colnames(pp) <- c(\"function\", \"gene_name\")\n",
    "pp <- pp[, c(\"gene_name\", \"function\")]\n",
    "write.table(pp, \"stat/diff_TE_up_exon_gene_function.txt\", col.names=T, row.names=F, sep=\"\\t\", quote=F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "load(\"E:/项目/转座子/sgRNA/results/Figure/test_genes.RData\")\n",
    "\n",
    "gostres <- gost(query=unique(genes),\n",
    "                # custom_bg=totalGenes,\n",
    "                organism = \"hsapiens\",\n",
    "                user_threshold = 0.05,\n",
    "                correction_method=\"gSCS\")\n",
    "\n",
    "gostres <- data.frame(gostres$result)\n",
    "gostres[1:5,\"term_name\",drop=F]\n",
    "gostres[gostres$source %in% c(\"GO:BP\",\"KEGG\"), c(\"p_value\", \"term_name\", \"source\")]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 位于i和u上的TE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## new transcript with TE\n",
    "trans_TE <- MDAMB231_Diff[[2]]\n",
    "trans_TE$TE_name <- paste0(paste(trans_TE$TE_chr, trans_TE$TE_start, trans_TE$TE_end, sep=\"_\"), \":\", trans_TE$repeat_id)\n",
    "trans_TE <- trans_TE[trans_TE$classification %in% c(\"intergenic\", \"intronic\"),]\n",
    "trans_TE2 <- MCF10A_Diff[[2]]\n",
    "trans_TE2$TE_name <- paste0(paste(trans_TE2$TE_chr, \n",
    "                                  trans_TE2$TE_start, \n",
    "                                  trans_TE2$TE_end, sep=\"_\"), \":\", trans_TE2$repeat_id)\n",
    "trans_TE2 <- trans_TE2[trans_TE2$classification %in% c(\"intergenic\", \"intronic\"),]\n",
    "\n",
    "\n",
    "\n",
    "length(intersect(trans_TE$TE_name, trans_TE2$TE_name))\n",
    "length(setdiff(trans_TE$TE_name, trans_TE2$TE_name))\n",
    "length(setdiff(trans_TE2$TE_name, trans_TE$TE_name))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Diff <- list()\n",
    "Diff[[\"MDAMB231\"]] <- MDAMB231_Diff\n",
    "Diff[[\"MCF10A\"]] <- MCF10A_Diff\n",
    "res$dif <- \"ns\"\n",
    "res$dif[res$log2FoldChange>1 & res$padj<0.05] <- \"Up\"\n",
    "res$dif[res$log2FoldChange< -1 & res$padj<0.05] <- \"Down\"\n",
    "sub_res <- res[rownames(res) %in% trans_TE$TE_name | rownames(res) %in% trans_TE2$TE_name, ]\n",
    "\n",
    "sub_res <- res[rownames(res) %in% trans_TE$TE_name & rownames(res) %in% trans_TE2$TE_name, ]\n",
    "table(sub_res$dif)\n",
    "sub_res <- res[rownames(res) %in% trans_TE$TE_name & !(rownames(res) %in% trans_TE2$TE_name), ]\n",
    "table(sub_res$dif)\n",
    "sub_res <- res[!(rownames(res) %in% trans_TE$TE_name) & (rownames(res) %in% trans_TE2$TE_name), ]\n",
    "table(sub_res$dif)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# pie\n",
    "pp <- data.frame(table(sub_res$dif))\n",
    "\n",
    "p <- pp %>%\n",
    "    mutate(Percentage = Freq / sum(Freq)*100)\n",
    "  \n",
    "  \n",
    "plot <- ggplot(p, aes(x=\"\", y=Percentage, fill=Var1))+\n",
    "    geom_bar(width = 1, stat = \"identity\") +\n",
    "    coord_polar(\"y\", start=0) +\n",
    "    scale_fill_manual(values = c(Down=\"#327FFF\", Up=\"#FD7E29\", ns=\"#AFAFAF\"))+\n",
    "    geom_text(aes(x = 1.6, label = paste0(Freq, \"(\",scales::percent(Percentage/100, accuracy = .1), \")\")), \n",
    "              position = position_stack(vjust = .5)) +\n",
    "    # facet_wrap(~Method)\n",
    "    ggtitle(\"\") +\n",
    "    theme_minimal() +\n",
    "    theme(\n",
    "      axis.title.x = element_blank(),\n",
    "      axis.title.y = element_blank(),\n",
    "      panel.border = element_blank(),\n",
    "      panel.grid=element_blank(),\n",
    "      axis.ticks = element_blank(),\n",
    "      plot.title=element_text(size=14, face=\"bold\", hjust=0.5),\n",
    "      axis.text.x=element_blank()\n",
    "    ) \n",
    "\n",
    "\n",
    "ggsave(\"stat/MDAMB231_VS_MCF10A_TE_i_u_dif_pie.pdf\",\n",
    "       width = 6, height = 3)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sub.TE.norlCount <- TE.norlCount[rownames(res)[abs(res$log2FoldChange)>1 & res$padj<0.05], ]\n",
    "pp <- melt(sub.TE.norlCount)\n",
    "colnames(pp) <- c(\"TE_name\", \"ID\", \"norlCount\")\n",
    "pp$stat <- \"Up\"\n",
    "pp$stat[pp$TE_name %in% rownames(res)[res$log2FoldChange< -1 & res$padj<0.05]] <- \"Down\"\n",
    "pp$repeat_id <- tstrsplit(pp$TE_name, \":\")[[2]]\n",
    "pp$sub_repeat <- substring(pp$repeat_id, 1, 4)\n",
    "pp$sub_repeat[grepl(\"L1P\", pp$sub_repeat)] <- \"L1P\"\n",
    "pp$CellLine <- tstrsplit(pp$ID, \"_\")[[2]]\n",
    "\n",
    "\n",
    "## violin\n",
    "p1 <- ggplot(pp, aes(x=sub_repeat, y=log2(norlCount+1), color=CellLine, fill=CellLine)) +\n",
    "    geom_violin(alpha=0.4) + \n",
    "    stat_summary(fun= median, geom = \"point\",position = position_dodge(width = 0.9),\n",
    "               shape = 23, size = 2, color = \"red\") +\n",
    "    # geom_point(alpha=0.1) +\n",
    "    # geom_jitter(width = 0.25, alpha=0.1) +\n",
    "    theme_bw() +\n",
    "    theme(\n",
    "      panel.grid=element_blank(),\n",
    "      axis.text.x = element_text(angle = 0, color = \"black\", \n",
    "                                 size = 8, hjust = 0.5, vjust=0.5),\n",
    "      plot.title=element_text(size=14, face=\"bold\", hjust=0.5)\n",
    "    ) +\n",
    "    facet_wrap(~stat, nrow = 2) +\n",
    "    scale_color_manual(values = c(\"#EF8354\", \"#4678BA\")) +\n",
    "    scale_fill_manual(values = c(\"#EF8354\", \"#4678BA\")) +\n",
    "    ggtitle(\"\")\n",
    "\n",
    "\n",
    "ggsave(paste0(\"stat/MDAMB231_vs_MCF10A_TE_repeat_family_normalCount.pdf\"),\n",
    "         width = 10, height = 6)\n",
    "\n",
    "\n",
    "# bar\n",
    "pp <- pp[order(pp$TE_name),]\n",
    "pp2 <- pp[!(duplicated(pp$TE_name)),]\n",
    "pp2 <- data.frame(table(pp2$stat, pp2$sub_repeat))\n",
    "colnames(pp2) <- c(\"stat\", \"sub_repeat\", \"Freq\")\n",
    "p2 <- ggplot(pp2, aes(x=sub_repeat, y=Freq,  fill=stat)) +\n",
    "    geom_bar(stat=\"identity\", position=position_dodge(width=0.9)) + \n",
    "    geom_text(aes(label=Freq), position=position_dodge(width=0.9), vjust=-0.25) +\n",
    "    theme_bw() +\n",
    "    theme(\n",
    "      panel.grid=element_blank(),\n",
    "      axis.text.x = element_text(angle = 0, color = \"black\", \n",
    "                                 size = 8, hjust = 0.5, vjust=0.5),\n",
    "      plot.title=element_text(size=14, face=\"bold\")\n",
    "    ) +\n",
    "    # facet_wrap(~CellLine, nrow = 1, scales=\"free\") +\n",
    "    scale_fill_manual(values = c(\"#EF8354\", \"#4678BA\")) +\n",
    "    # ylim(0,3200) +\n",
    "    ggtitle(\"the count of differential expression TEs\") \n",
    "   \n",
    "  \n",
    "\n",
    "ggsave(paste0(\"stat/MDAMB231_vs_MCF10A_TE_i_u_dif_count_bar.pdf\"),\n",
    "         width = 7, height = 4)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "TEtype_normalCount <- lapply(c(\"MDAMB231\", \"MCF10A\"), function(CellLine){\n",
    "    IDs <- Sample_info$Samples[Sample_info$CellLine %in% CellLine &\n",
    "                                Sample_info$SeqType %in% c(\"sgRNA\")]\n",
    "    #\n",
    "    quant <- TE.norlCount[, IDs]\n",
    "    quant <- data.frame(TE_name=rownames(quant), quant, normalCount=rowMeans(quant))\n",
    "    sub.TE <- rownames(res)[grepl(CellLine, res$dif)]\n",
    "    quant <- quant[sub.TE, ]\n",
    "    # \n",
    "    trans_TE <- Diff[[CellLine]][[2]]\n",
    "    trans_TE$TE_name <- paste0(paste(trans_TE$TE_chr, trans_TE$TE_start, trans_TE$TE_end, sep=\"_\"), \":\", trans_TE$repeat_id)\n",
    "    trans_TE <- trans_TE[trans_TE$classification %in% c(\"intergenic\", \"intronic\"),]\n",
    "\n",
    "\n",
    "    sub.quant <- merge(quant, trans_TE[, c(\"classification\", \"repeat_id\", \"TE_name\", \"matching_class\")], by=\"TE_name\")\n",
    "    sub.quant <- sub.quant[order(sub.quant$TE_name),]\n",
    "    sub.quant <- sub.quant[!duplicated(sub.quant),]\n",
    "    \n",
    "    #\n",
    "    sub.quant$sub_repeat <- substring(sub.quant$repeat_id, 1, 4)\n",
    "    sub.quant$sub_repeat[grepl(\"L1P\", sub.quant$sub_repeat)] <- \"L1P\"\n",
    "    #\n",
    "    sub.quant <- data.frame(sub.quant[, !(colnames(sub.quant) %in% IDs)], \n",
    "                                  CellLine=CellLine)\n",
    "    \n",
    "    return(sub.quant)\n",
    "})\n",
    "\n",
    "pp <- rbindlist(TEtype_normalCount)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(patchwork)\n",
    "\n",
    "pp <- rbindlist(TEtype_normalCount)\n",
    "\n",
    "\n",
    "p1 <- ggplot(pp, aes(x=sub_repeat, y=log2(normalCount+1), color=matching_class, fill=matching_class)) +\n",
    "    geom_violin(alpha=0.4) + \n",
    "    stat_summary(fun= median, geom = \"point\",\n",
    "               shape = 23, size = 2, color = \"red\") +\n",
    "    # geom_point(alpha=0.1) +\n",
    "    # geom_jitter(width = 0.25, alpha=0.1) +\n",
    "    theme_bw() +\n",
    "    theme(\n",
    "      panel.grid=element_blank(),\n",
    "      axis.text.x = element_text(angle = 0, color = \"black\", \n",
    "                                 size = 8, hjust = 0.5, vjust=0.5),\n",
    "      plot.title=element_text(size=14, face=\"bold\", hjust=0.5)\n",
    "    ) +\n",
    "    facet_wrap(classification~CellLine, nrow = 2) +\n",
    "    scale_color_manual(values = c(\"#EF8354\", \"#4678BA\")) +\n",
    "    scale_fill_manual(values = c(\"#EF8354\", \"#4678BA\")) +\n",
    "    ggtitle(\"\")\n",
    "\n",
    "\n",
    "ggsave(paste0(\"stat/MDAMB231_vs_MCF10A_TE_in_i_u_transcripts_repeat_family_normalCount.pdf\"),\n",
    "         width = 10, height = 6)\n",
    "\n",
    "\n",
    "pp2 <- data.frame(table(pp$CellLine, pp$matching_class, pp$sub_repeat))\n",
    "colnames(pp2) <- c(\"CellLine\", \"matching_class\", \"sub_repeat\", \"Freq\")\n",
    "pp2 <- pp2[pp2$Freq>0,] \n",
    "\n",
    "p2 <- ggplot(pp2, aes(x=sub_repeat, y=Freq,  fill=matching_class)) +\n",
    "    geom_bar(stat=\"identity\") + \n",
    "    geom_text(aes(label=Freq), position=position_dodge(width=0.9), vjust=-0.25) +\n",
    "    theme_bw() +\n",
    "    theme(\n",
    "      panel.grid=element_blank(),\n",
    "      axis.text.x = element_text(angle = 0, color = \"black\", \n",
    "                                 size = 8, hjust = 0.5, vjust=0.5),\n",
    "      plot.title=element_text(size=14, face=\"bold\")\n",
    "    ) +\n",
    "    facet_wrap(~CellLine, nrow = 1, scales=\"free\") +\n",
    "    scale_fill_manual(values = c(\"#EF8354\", \"#4678BA\")) +\n",
    "    ggtitle(\"\") +\n",
    "    ylim(0,7000)\n",
    "  \n",
    "\n",
    "\n",
    "p <- p1 + p2 + plot_layout(ncol = 1, heights = c(1, 1))\n",
    "\n",
    "ggsave(paste0(\"stat/MDAMB231_vs_MCF10A_TE_in_i_u_transcripts_repeat_family_normalCount.pdf\"),\n",
    "         width = 10, height = 6)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "## HCT116 vs NCM460"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "setwd(\"/NAS/wg_looking/gRNA_ONT\")\n",
    "load(\"stat/HCT116_NCM460_diff.Data\")\n",
    "load(\"stat/MDAMB231_MCF10A_diff.Data\")\n",
    "load(\"stat/assembly_preData.RData\")\n",
    "load(\"stat/TE.quantify.count.RData\")\n",
    "\n",
    "\n",
    "# load Sample info\n",
    "Sample_info <- read.table(\"stat/Sample_info2.txt\", header = T, stringsAsFactors = F)\n",
    "rownames(Sample_info) <- Sample_info$Samples\n",
    "\n",
    "SampleIDs <- Sample_info$Samples[Sample_info$CellLine %in% c(\"HCT116\", \"NCM460\") &\n",
    "                                Sample_info$SeqType %in% c(\"sgRNA\")]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# sgRNA count > 3 at least 3 samples\n",
    "library(DESeq2)\n",
    "\n",
    "SampleIDs <- Sample_info$Samples[Sample_info$CellLine %in% c(\"HCT116\", \"NCM460\") &\n",
    "                                Sample_info$SeqType %in% c(\"sgRNA\")]\n",
    "sub.count.tb <- count.tb[, SampleIDs]\n",
    "sub.count.tb <- sub.count.tb[apply(sub.count.tb, 1, function(x){ (sum(x>=3))>=3 }),]\n",
    "dim(sub.count.tb) # 75933\n",
    "sub.count.tb <- round(sub.count.tb)\n",
    "\n",
    "# run DESeq2\n",
    "colInfo <- Sample_info[SampleIDs, c(\"CellLine\"), drop=F]\n",
    "dds <- DESeqDataSetFromMatrix(sub.count.tb, \n",
    "                                colInfo, \n",
    "                                ~CellLine)\n",
    "Object <- DESeq(dds\n",
    "                # parallel = T, BPPARAM=MulticoreParam(10)\n",
    "                )\n",
    "TE.norlCount <- counts(Object, normalized=T)\n",
    "\n",
    "res <- results(Object, contrast=c(\"CellLine\", \"HCT116\", \"NCM460\"))\n",
    "# res <- res[complete.cases(res),]\n",
    "save(TE.norlCount, res, file=\"stat/HCT116_vs_NCM460_TE_res.RData\")\n",
    "\n",
    "dif_res <- res\n",
    "dif_res <- res[res$padj<0.05 & abs(res$log2FoldChange)>1, ]\n",
    "table(dif_res$log2FoldChange>1)\n",
    "\n",
    "\n",
    "dif_res2 <- dif_res[rownames(dif_res) %in% trans_TE$TE_name |\n",
    "       rownames(dif_res) %in% trans_TE2$TE_name,]\n",
    "table(dif_res2$log2FoldChange>1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "load(\"stat/TE_location.RData\")\n",
    "load(\"stat/HCT116_vs_NCM460_TE_res.RData\")\n",
    "\n",
    "\n",
    "out_res <- data.frame(TE.norlCount, res[, c(\"baseMean\", \"log2FoldChange\", \"pvalue\", \"padj\")])\n",
    "out_res$dif <- \"ns\"\n",
    "out_res$dif[res$padj<0.05 & res$log2FoldChange>1] <- \"Up\"\n",
    "out_res$dif[res$padj<0.05 & res$log2FoldChange < -1] <- \"Down\"\n",
    "table(out_res$dif)\n",
    "out_res <- out_res[order(out_res$dif, decreasing=T),]\n",
    "out_res <- data.frame(gene_id=rownames(out_res), out_res)\n",
    "colnames(out_res)[1] <- \"TE_name\"\n",
    "\n",
    "# out_res <- data.frame(out_res, TE_location[rownames(out_res), c(\"sub_repeat\", \"exon\", \"intron\", \"intergenic\",\n",
    "#                                                                 \"gene_id\", \"gene_name\")])\n",
    "out_res <- merge(out_res, TE_location, by=\"TE_name\")\n",
    "out_res$gene_id <- as.character(out_res$gene_id)\n",
    "\n",
    "# out_res <- gene_res[transcript_res$padj<0.05 & abs(transcript_res$log2FoldChange)>2 & transcript_res$baseMean>10, ]\n",
    "# write.table(out_res, \"stat/MDAMB231_vs_MCF10A_differential_genes.txt\", row.names = F, col.names = T, quote = F, sep=\"\\t\")\n",
    "write.table(out_res, \"stat/HCT116_vs_NCM460_all_TEs.txt\", row.names = F, col.names = T, quote = F, sep=\"\\t\")\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## correlation: CRC and BC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "load(\"stat/TE.count.DESeq2.Object.RData\")\n",
    "load(\"stat/MDAMB231_vs_MCF10A_TE_res.RData\")\n",
    "BC_res <- res\n",
    "load(\"stat/HCT116_vs_NCM460_TE_res.RData\")\n",
    "CRC_res <- res\n",
    "\n",
    "\n",
    "TE.norlCount <- counts(Object, normalized=T)\n",
    "cancer_Up <- intersect(rownames(BC_res)[BC_res$log2FoldChange>1 & BC_res$padj<0.05],\n",
    "                      rownames(CRC_res)[CRC_res$log2FoldChange>1 & CRC_res$padj<0.05])\n",
    "\n",
    "cancer_Down <- intersect(rownames(BC_res)[BC_res$log2FoldChange< -1 & BC_res$padj<0.05],\n",
    "                      rownames(CRC_res)[CRC_res$log2FoldChange< -1 & CRC_res$padj<0.05])\n",
    "\n",
    "\n",
    "\n",
    "pheatmap(cor(TE.norlCount[cancer_Up, ], method = \"spearman\"),\n",
    "         display_numbers=T,\n",
    "         # annotation_col = colInfo,\n",
    "         # annotation_row = colInfo,\n",
    "         # annotation_colors = anno_colors,\n",
    "         main=\"Correlation of TE normalized Count \",\n",
    "         # col=colorRampPalette(rev(brewer.pal(11,\"RdYlBu\")[3:10]))(21),\n",
    "         col=colorRampPalette(c(\"#4678BA\", \"#F7F7F7\", \"#EF8354\"))(100),\n",
    "         border=\"white\",\n",
    "         width=6.5,\n",
    "         height=6,\n",
    "         # show_rownames=F,\n",
    "         # show_colnames=F,\n",
    "         filename=\"stat/TE_count_spearman_correlation_cancer_up.pdf\")\n",
    "\n",
    "\n",
    "pheatmap(cor(TE.norlCount[cancer_Down, ], method = \"spearman\"),\n",
    "         display_numbers=T,\n",
    "         # annotation_col = colInfo,\n",
    "         # annotation_row = colInfo,\n",
    "         # annotation_colors = anno_colors,\n",
    "         main=\"Correlation of TE normalized Count \",\n",
    "         # col=colorRampPalette(rev(brewer.pal(11,\"RdYlBu\")[3:10]))(21),\n",
    "         col=colorRampPalette(c(\"#4678BA\", \"#F7F7F7\", \"#EF8354\"))(100),\n",
    "         border=\"white\",\n",
    "         width=6.5,\n",
    "         height=6,\n",
    "         # show_rownames=F,\n",
    "         # show_colnames=F,\n",
    "         filename=\"stat/TE_count_spearman_correlation_cancer_Down.pdf\")\n",
    "\n",
    "\n",
    "pheatmap(cor(TE.norlCount[c(cancer_Down, cancer_Up), ], method = \"spearman\"),\n",
    "         display_numbers=T,\n",
    "         # annotation_col = colInfo,\n",
    "         # annotation_row = colInfo,\n",
    "         # annotation_colors = anno_colors,\n",
    "         main=\"Correlation of TE normalized Count \",\n",
    "         # col=colorRampPalette(rev(brewer.pal(11,\"RdYlBu\")[3:10]))(21),\n",
    "         col=colorRampPalette(c(\"#4678BA\", \"#F7F7F7\", \"#EF8354\"))(100),\n",
    "         border=\"white\",\n",
    "         width=6.5,\n",
    "         height=6,\n",
    "         # show_rownames=F,\n",
    "         # show_colnames=F,\n",
    "         filename=\"stat/TE_count_spearman_correlation_cancer.pdf\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## TE expression in i, u and span"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# TE activity\n",
    "library(bedtoolsr)\n",
    "library(DESeq2)\n",
    "setwd(\"/NAS/wg_looking/gRNA_ONT\")\n",
    "load(\"stat/assembly_preData.RData\") # \"gtf.tb.list\"  \"RM.tb.list\"   \"RM.tb.list2\"  \"tmap.tb.list\"\n",
    "load(\"stat/HCT116_NCM460_diff.Data\")\n",
    "load(\"stat/MDAMB231_MCF10A_diff.Data\")\n",
    "load(\"stat/hg38.gtf.RData\")\n",
    "load(\"stat/hg38.gene.transcript.gtf.RData\")\n",
    "load(\"stat/TE_location.RData\")\n",
    "load(\"stat/TE.quantify.count.RData\")\n",
    "\n",
    "\n",
    "# load Sample info\n",
    "Sample_info <- read.table(\"stat/Sample_info2.txt\", header = T, stringsAsFactors = F)\n",
    "rownames(Sample_info) <- Sample_info$Samples\n",
    "\n",
    "SampleIDs <- Sample_info$Samples[Sample_info$CellLine %in% c(\"MDAMB231\", \"MCF10A\", \"HCT116\", \"NCM460\") &\n",
    "                                Sample_info$SeqType %in% c(\"sgRNA\")]\n",
    "\n",
    "\n",
    "\n",
    "## activity TE\n",
    "Diff.list <- list(MDAMB231=MDAMB231_Diff, MCF10A=MCF10A_Diff,\n",
    "                  HCT116=HCT116_Diff, NCM460=NCM460_Diff)\n",
    "\n",
    "activity_TE.list <- lapply(names(Diff.list), function(CellLine){\n",
    "    Diff <- Diff.list[[CellLine]] \n",
    "    sub.diff <- Diff[[2]][Diff[[2]]$type %in% c(\"TES\", \"TSS\"), ]\n",
    "    # sub.diff <- sub.diff[sub.diff$classification %in% c(\"intronic\", \"span_intergenic_and_genic\", \"intergenic\"), ]\n",
    "    #\n",
    "    sub <- sub.diff[sub.diff$classification %in% c(\"intergenic\"), ]\n",
    "    TE <- sub[, c(\"TE_chr\", \"TE_start\", \"TE_end\", \"TE_name\")]\n",
    "    TE <- TE[order(TE$TE_chr, TE$TE_start),]\n",
    "    intergenic_bed <- read.table(\"/NAS/wg_looking/human_ref/gencode.v37.chr_patch_hapl_scaff.annotation.intergenic.bed\", \n",
    "                         stringsAsFactors = F)\n",
    "    intergenic_res <- bt.intersect(a=TE, intergenic_bed, wa=T)\n",
    "    colnames(intergenic_res) <- colnames(TE)\n",
    "    sub <- sub[sub$TE_name %in% intergenic_res$TE_name, ]\n",
    "    #\n",
    "    sub.diff <- rbind(sub.diff[sub.diff$classification %in% c(\"intronic\", \"intergenic\"), ],\n",
    "                sub)\n",
    "    \n",
    "    TE_name <- data.frame(TE_name=unique(sub.diff$TE_name), CellLine=CellLine)\n",
    "    \n",
    "    return(TE_name)\n",
    "\n",
    "})\n",
    "\n",
    "\n",
    "activity_TE <- rbindlist(activity_TE.list)\n",
    "activity_TE <- as.data.frame.matrix(table(activity_TE$TE_name, activity_TE$CellLine))\n",
    "table(rowSums(activity_TE))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "library(gtools)\n",
    "\n",
    "TE_expr <- count.tb[rownames(activity_TE), SampleIDs]\n",
    "dim(TE_expr) # 30579 12\n",
    "TE_expr <- round(TE_expr)\n",
    "\n",
    "# run DESeq2\n",
    "colInfo <- Sample_info[SampleIDs, c(\"CellLine\", \"TissueType\")]\n",
    "dds <- DESeqDataSetFromMatrix(TE_expr, \n",
    "                                colInfo, \n",
    "                                ~CellLine)\n",
    "Object <- DESeq(dds\n",
    "                # parallel = T, BPPARAM=MulticoreParam(10)\n",
    "                )\n",
    "TE.norlCount <- counts(Object, normalized=T)\n",
    "\n",
    "# combs <- combinations(length(unique(colInfo$CellLine)), 2, unique(colInfo$CellLine))\n",
    "# combs <- t(data.frame(comb1=c(\"HCT116\",\"MDAMB231\"), \n",
    "#                     comb2=c(\"MCF10A\",\"NCM460\"), \n",
    "#                     comb3=c(\"HCT116\",\"NCM460\"), \n",
    "#                     comb4=c(\"MDAMB231\", \"MCF10A\")))\n",
    "combs <- t(data.frame(comb2=c(\"HCT116\",\"NCM460\"), \n",
    "                    comb4=c(\"MDAMB231\", \"MCF10A\")))\n",
    "dif_TE <- c()\n",
    "for(i in 1:dim(combs)[1]){\n",
    "    comb <- combs[i,]\n",
    "    res <- results(Object, contrast=c(\"CellLine\", comb[1], comb[2]))\n",
    "    res <- res[complete.cases(res), ]\n",
    "    dif_res <- res[res$padj<0.05 & abs(res$log2FoldChange)>1, ]\n",
    "    dif_TE <- c(dif_TE, rownames(dif_res))\n",
    "}\n",
    "dif_TE <- unique(dif_TE)\n",
    "length(dif_TE)\n",
    "\n",
    "\n",
    "# heatmap plot\n",
    "rowInfo <- activity_TE[dif_TE, ]\n",
    "rowInfo <- data.frame(apply(rowInfo, 2, as.character))\n",
    "rownames(rowInfo) <- dif_TE\n",
    "colInfo <- Sample_info[Sample_info$Samples %in% SampleIDs, \"CellLine\", drop=F]\n",
    "anno_colors <- list(CellLine=c(HCT116=\"#FC8D62\", NCM460=\"#31A354\", MDAMB231=\"#FFD92F\", MCF10A=\"#43A2CA\"),\n",
    "                   HCT116=c(\"1\"=\"#FC8D62\", \"0\"=\"#F0F9E8\"),\n",
    "                   NCM460=c(\"1\"=\"#31A354\", \"0\"=\"#F0F9E8\"),\n",
    "                   MDAMB231=c(\"1\"=\"#FFD92F\", \"0\"=\"#F0F9E8\"),\n",
    "                   MCF10A=c(\"1\"=\"#43A2CA\", \"0\"=\"#F0F9E8\"))\n",
    "\n",
    "pheatmap(TE.norlCount[dif_TE,],\n",
    "         # display_numbers=T,\n",
    "         scale=\"row\",\n",
    "         annotation_col = colInfo,\n",
    "         annotation_row = rowInfo,\n",
    "         annotation_colors = anno_colors,\n",
    "         main=\"differential TE expression\",\n",
    "         col=colorRampPalette(rev(brewer.pal(11,\"RdBu\")))(21),\n",
    "         # col=colorRampPalette(c(\"#4678BA\", \"#F7F7F7\", \"#EF8354\"))(11),\n",
    "         border=NA,\n",
    "         show_rownames=F,\n",
    "         show_colnames=F,\n",
    "         # cluster_row=F,\n",
    "         width=5,\n",
    "         height=6,\n",
    "         filename=\"stat/CRC_BC_TE_dif_heatmap.pdf\")\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "\n",
    "combs <- t(data.frame(comb2=c(\"HCT116\",\"NCM460\"), \n",
    "                    comb4=c(\"MDAMB231\", \"MCF10A\")))\n",
    "\n",
    "\n",
    "dif_TE.list <- list()\n",
    "for(i in 1:dim(combs)[1]){\n",
    "    comb <- combs[i,]\n",
    "    n <- paste0(comb[1], \"_vs_\", comb[2])\n",
    "    res <- results(Object, contrast=c(\"CellLine\", comb[1], comb[2]))\n",
    "    res <- res[complete.cases(res), ]\n",
    "    dif_TE <- rownames(res)[res$padj<0.05 & res$log2FoldChange>1]\n",
    "    dif_TE.list[[n]] <- dif_TE\n",
    "}\n",
    "\n",
    "\n",
    "up <- intersect(dif_TE.list[[1]], dif_TE.list[[2]])\n",
    "table(TE_location[up, \"sub_repeat\"])\n",
    "c <- colSums(TE_location[up, c(\"intron\", \"intergenic\")])\n",
    "genomic_location_pct <- c/sum(c)\n",
    "\n",
    "\n",
    "\n",
    "TE_location[up, ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Tumor vs Normal\n",
    "load(\"stat/TE_location.RData\")\n",
    "\n",
    "dds <- DESeqDataSetFromMatrix(TE_expr, \n",
    "                                colInfo, \n",
    "                                ~TissueType)\n",
    "Object <- DESeq(dds\n",
    "                # parallel = T, BPPARAM=MulticoreParam(10)\n",
    "                )\n",
    "\n",
    "res <- results(Object, contrast=c(\"TissueType\", \"Tumor\", \"Normal\"))\n",
    "res <- res[complete.cases(res), ]\n",
    "Up <- res[res$padj<0.05 & res$log2FoldChange>1, ]\n",
    "Up <- data.frame(Up, TE_location[rownames(Up),])\n",
    "up_pct <- table(Up$sub_repeat)/sum(table(Up$sub_repeat))\n",
    "bg_pct <- table(TE_location[rownames(activity_TE), \"sub_repeat\"])/sum(table(TE_location[rownames(activity_TE), \"sub_repeat\"]))\n",
    "\n",
    "c <- colSums(Up[, c(\"intron\", \"intergenic\")])\n",
    "genomic_location_pct <- c/sum(c)\n",
    "\n",
    "\n",
    "c <- colSums(TE_location[rownames(activity_TE), c(\"intron\", \"intergenic\")])\n",
    "genomic_location_pct <- c/sum(c)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# MDAMB231 MCF10A HCT116 NCM460\n",
    "\n",
    "activity_TE[,]\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## TE expression in BC and CRC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TE activity\n",
    "library(bedtoolsr)\n",
    "library(DESeq2)\n",
    "library(data.table)\n",
    "setwd(\"/NAS/wg_looking/gRNA_ONT\")\n",
    "load(\"stat/assembly_preData.RData\") # \"gtf.tb.list\"  \"RM.tb.list\"   \"RM.tb.list2\"  \"tmap.tb.list\"\n",
    "load(\"stat/HCT116_NCM460_diff.Data\")\n",
    "load(\"stat/MDAMB231_MCF10A_diff.Data\")\n",
    "load(\"stat/hg38.gtf.RData\")\n",
    "load(\"stat/hg38.gene.transcript.gtf.RData\")\n",
    "load(\"stat/TE_location.RData\")\n",
    "load(\"stat/TE.quantify.count.RData\")\n",
    "\n",
    "\n",
    "# load Sample info\n",
    "Sample_info <- read.table(\"stat/Sample_info2.txt\", header = T, stringsAsFactors = F)\n",
    "rownames(Sample_info) <- Sample_info$Samples\n",
    "\n",
    "SampleIDs <- Sample_info$Samples[Sample_info$CellLine %in% c(\"MDAMB231\", \"MCF10A\", \"HCT116\", \"NCM460\") &\n",
    "                                Sample_info$SeqType %in% c(\"sgRNA\")]\n",
    "\n",
    "\n",
    "\n",
    "## activity TE\n",
    "Diff.list <- list(MDAMB231=MDAMB231_Diff, MCF10A=MCF10A_Diff,\n",
    "                  HCT116=HCT116_Diff, NCM460=NCM460_Diff)\n",
    "\n",
    "activity_TE.list <- lapply(names(Diff.list), function(CellLine){\n",
    "    Diff <- Diff.list[[CellLine]] \n",
    "    sub.diff <- Diff[[2]][Diff[[2]]$type %in% c(\"TES\", \"TSS\"), ]\n",
    "    # sub.diff <- Diff[[2]]\n",
    "    sub.diff <- sub.diff[!(sub.diff$ref_gene_id %in% \"-\"), ]\n",
    "    \n",
    "    TE_name <- data.frame(TE_name=unique(sub.diff$TE_name), CellLine=CellLine)\n",
    "    \n",
    "    return(TE_name)\n",
    "\n",
    "})\n",
    "\n",
    "\n",
    "activity_TE <- rbindlist(activity_TE.list)\n",
    "activity_TE <- as.data.frame.matrix(table(activity_TE$TE_name, activity_TE$CellLine))\n",
    "table(rowSums(activity_TE))\n",
    "dim(activity_TE) # 43085\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(gtools)\n",
    "\n",
    "TE_expr <- count.tb[, SampleIDs]\n",
    "TE_expr <- TE_expr[apply(TE_expr, 1, function(x){sum(x>=3)>=3}), ]\n",
    "dim(TE_expr) # 105771     12\n",
    "TE_expr <- round(TE_expr)\n",
    "\n",
    "# run DESeq2\n",
    "colInfo <- Sample_info[SampleIDs, c(\"CellLine\", \"TissueType\")]\n",
    "dds <- DESeqDataSetFromMatrix(TE_expr, \n",
    "                                colInfo, \n",
    "                                ~CellLine)\n",
    "Object <- DESeq(dds\n",
    "                # parallel = T, BPPARAM=MulticoreParam(10)\n",
    "                )\n",
    "TE.norlCount <- counts(Object, normalized=TRUE)\n",
    "\n",
    "# combs <- combinations(length(unique(colInfo$CellLine)), 2, unique(colInfo$CellLine))\n",
    "# combs <- t(data.frame(comb1=c(\"HCT116\",\"MDAMB231\"), \n",
    "#                     comb2=c(\"MCF10A\",\"NCM460\"), \n",
    "#                     comb3=c(\"HCT116\",\"NCM460\"), \n",
    "#                     comb4=c(\"MDAMB231\", \"MCF10A\")))\n",
    "combs <- list(c(\"HCT116\",\"NCM460\"), \n",
    "                    c(\"MDAMB231\", \"MCF10A\"))\n",
    "res.list <- list()\n",
    "dif_TE.list <- list()\n",
    "for(comb in combs){\n",
    "    print(comb)\n",
    "    res <- results(Object, contrast=c(\"CellLine\", comb[1], comb[2]))\n",
    "    res <- res[complete.cases(res), ]\n",
    "    Up <- rownames(res[res$padj<0.05 & res$log2FoldChange>1, ])\n",
    "    Down <- rownames(res[res$padj<0.05 & res$log2FoldChange< -1, ])\n",
    "    n <- paste0(comb[1], \"_vs_\", comb[2])\n",
    "    \n",
    "    dif_TE.list[[n]] <- list(Up=Up, Down=Down)\n",
    "    res.list[[n]] <- res\n",
    "}\n",
    "\n",
    "\n",
    "length(intersect(dif_TE.list[[1]]$Up, dif_TE.list[[2]]$Up))\n",
    "length(intersect(dif_TE.list[[1]]$Down, dif_TE.list[[2]]$Down))\n",
    "dif_TE <- unique(unlist(dif_TE.list))\n",
    "length(dif_TE) # 17108\n",
    "save(TE.norlCount, res.list, dif_TE.list, file=\"stat/BC_CRC_TE_count_and_dif.RData\")\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(pheatmap)\n",
    "setwd(\"/NAS/wg_looking/gRNA_ONT\")\n",
    "load(\"stat/BC_CRC_TE_count_and_dif.RData\") # TE.norlCount, dif_TE.list\n",
    "dif_TE <- unique(unlist(dif_TE.list))\n",
    "length(dif_TE) # 17108\n",
    "\n",
    "\n",
    "\n",
    "# heatmap plot\n",
    "rowInfo <- data.frame(TE_name=dif_TE)\n",
    "rownames(rowInfo) <- dif_TE\n",
    "# rowInfo <- data.frame(rowInfo, activity_TE[rownames(rowInfo),])\n",
    "# rowInfo[is.na(rowInfo)] <- 0\n",
    "rowInfo$HCT116_up <- 0\n",
    "rowInfo$HCT116_up[rownames(rowInfo) %in% dif_TE.list[[\"HCT116_vs_NCM460\"]][[\"Up\"]]] <- 1\n",
    "rowInfo$MDAMB231_up <- 0\n",
    "rowInfo$MDAMB231_up[rownames(rowInfo) %in% dif_TE.list[[\"MDAMB231_vs_MCF10A\"]][[\"Up\"]]] <- 1\n",
    "rowInfo$MCF10A_up <- 0\n",
    "rowInfo$MCF10A_up[rownames(rowInfo) %in% dif_TE.list[[\"MDAMB231_vs_MCF10A\"]][[\"Down\"]]] <- 1\n",
    "rowInfo$NCM460_up <- 0\n",
    "rowInfo$NCM460_up[rownames(rowInfo) %in% dif_TE.list[[\"HCT116_vs_NCM460\"]][[\"Down\"]]] <- 1\n",
    "####\n",
    "rowInfo$log2FC_BC <- res.list[[\"MDAMB231_vs_MCF10A\"]][rownames(rowInfo), \"log2FoldChange\"]\n",
    "rowInfo$log2FC_CRC <- res.list[[\"HCT116_vs_NCM460\"]][rownames(rowInfo), \"log2FoldChange\"]\n",
    "rowInfo <- rowInfo[,-1]\n",
    "rowInfo <- rowInfo[order(rowInfo$HCT116_up, \n",
    "                         # rowInfo$log2FC_CRC, \n",
    "                         rowInfo$MDAMB231_up, \n",
    "                         # rowInfo$log2FC_BC, \n",
    "                         rowInfo$MCF10A_up, \n",
    "                         rowInfo$NCM460_up, \n",
    "                         decreasing=T), ]\n",
    "# rowInfo <- data.frame(apply(rowInfo, 2, as.character))\n",
    "# rownames(rowInfo) <- dif_TE\n",
    "\n",
    "colInfo <- Sample_info[Sample_info$Samples %in% SampleIDs, \"CellLine\", drop=F]\n",
    "\n",
    "sub.TE.norlCount <- TE.norlCount[rownames(rowInfo),]\n",
    "\n",
    "k=10\n",
    "pheatmap.p <- pheatmap(sub.TE.norlCount, scale=\"row\", cutree_rows=k)\n",
    "hc <- hclust(dist(rowInfo[,5:8], method=\"binary\"), method = \"complete\")\n",
    "clusters <- cutree(pheatmap.p$tree_row, k=k)\n",
    "rowInfo$cluster <- as.character(cluster)\n",
    "table(rowInfo$cluster)\n",
    "\n",
    "anno_colors <- list(CellLine=c(HCT116=\"#FC8D62\", NCM460=\"#31A354\", MDAMB231=\"#FFD92F\", MCF10A=\"#43A2CA\"),\n",
    "                   cluster=setNames(brewer.pal(12, \"Paired\")[1:k], unique(rowInfo$cluster)))\n",
    "\n",
    "\n",
    "pheatmap(t(sub.TE.norlCount),\n",
    "         # display_numbers=T,\n",
    "         scale=\"column\",\n",
    "         annotation_col = rowInfo[, 1:4, drop=F],\n",
    "         annotation_row = colInfo,\n",
    "         annotation_colors = anno_colors,\n",
    "         # cutree_cols=10,\n",
    "         main=\"differential TE expression\",\n",
    "         col=colorRampPalette(rev(brewer.pal(11,\"RdBu\")))(21),\n",
    "         # col=colorRampPalette(c(\"#4678BA\", \"#F7F7F7\", \"#EF8354\"))(11),\n",
    "         border=NA,\n",
    "         show_rownames=F,\n",
    "         show_colnames=F,\n",
    "         cluster_col=F,\n",
    "         # cluster_col=hc,\n",
    "         # cluster_row=F,\n",
    "         width=7,\n",
    "         height=7,\n",
    "         filename=\"stat/CRC_BC_TE_dif_heatmap.pdf\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library(dendsort)\n",
    "method1 = c(\"euclidean\", \"maximum\", \"manhattan\", \"canberra\", \"minkowski\", \"binary\")\n",
    "method2 = c(\"ward.D\", \"ward.D2\", \"single\",  \"complete\", \"average\")\n",
    "method3 = c(\"average\",\"min\")\n",
    "for(i in method1){\n",
    "  for(j in method2){\n",
    "    for(z in method3){\n",
    "      filename= paste(\"stat/CRC_BC_TE_dif_heatmap_\", i, \"_\", j, \"_\",z,\".pdf\", sep=\"\")\n",
    "      title = paste(\"distance:\", i, \" hclust:\",j, \" type:\",z, sep = \"\")\n",
    "      if(file.exists(filename)==FALSE){\n",
    "          try(\n",
    "              pheatmap(\n",
    "                t(sub.TE.norlCount),\n",
    "                col=colorRampPalette(rev(brewer.pal(11,\"RdBu\")))(21),\n",
    "                breaks = c(-4,seq(-3,3,length=20),4),\n",
    "                legend_breaks = c(-3,-1,1,3),\n",
    "                filename = filename,\n",
    "                scale=\"column\",\n",
    "                main = title,\n",
    "                width = 7,\n",
    "                height = 3,\n",
    "                # clustering_method = \"complete\",\n",
    "                border_color=NA,\n",
    "                # fontsize_col = 0.5,\n",
    "                # annotation_row = geneCluster,\n",
    "                # annotation_col = colInfo,\n",
    "                annotation_row = colInfo,\n",
    "                annotation_colors = anno_colors,\n",
    "                annotation_legend = T,\n",
    "                show_rownames = F,\n",
    "                show_colnames = F,\n",
    "                # cluster_rows = F,\n",
    "                cluster_cols = as.hclust(dendsort(hclust(dist(sub.TE.norlCount, method=i), method = j), type=z))\n",
    "              ), \n",
    "              silent=TRUE)\n",
    "                  \n",
    "\n",
    "      \n",
    "      }\n",
    "    }\n",
    "  }\n",
    "}\n",
    "##\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# new transcript density\n",
    "rowInfo <- data.frame(TE_name=dif_TE)\n",
    "rownames(rowInfo) <- dif_TE\n",
    "rowInfo <- data.frame(rowInfo, activity_TE[rownames(rowInfo),])\n",
    "rowInfo[is.na(rowInfo)] <- 0\n",
    "rowInfo <- data.frame(apply(rowInfo, 2, as.character))\n",
    "rownames(rowInfo) <- dif_TE\n",
    "\n",
    "\n",
    "pp <- rowInfo[pheatmap.p$tree_row$order, ]\n",
    "pp <- rbindlist(list(data.frame(logic=pp$MDAMB231, ID=\"MDAMB231\"),\n",
    "               data.frame(logic=pp$MCF10A, ID=\"MCF10A\"),\n",
    "               data.frame(logic=pp$HCT116, ID=\"HCT116\"),\n",
    "               data.frame(logic=pp$NCM460, ID=\"NCM460\")))\n",
    "step <- 50\n",
    "winSet <- rep(1:ceiling(dim(pp)[1]/4/step), each = step)[1:(dim(pp)[1]/4)]\n",
    "pp$order <- rep(winSet, 4)\n",
    "# pp.sum <- aggregate(logic~ID+order, data=pp, sum)\n",
    "pp$logic <- as.numeric(as.character(pp$logic))\n",
    "# pp$order <- rep(1:(dim(pp)[1]/4), 4)\n",
    "ppp <- pp[!(pp$logic==0), ]\n",
    "ppp$ID <- factor(ppp$ID, levels=c(\"HCT116\", \"MDAMB231\", \"MCF10A\", \"NCM460\"))\n",
    "\n",
    "cols <- c(HCT116=\"#FC8D62\", NCM460=\"#31A354\", MDAMB231=\"#FFD92F\", MCF10A=\"#43A2CA\")\n",
    "ggplot(ppp, aes(x=order, color=ID)) +\n",
    "    geom_density() + \n",
    "    theme_bw() +\n",
    "    theme(\n",
    "      panel.grid=element_blank(),\n",
    "      axis.text.x = element_text(angle = 0, color = \"black\", \n",
    "                                 size = 8, hjust = 0.5, vjust=0.5),\n",
    "      plot.title=element_text(size=14, face=\"bold\", hjust=0.5)\n",
    "    ) +\n",
    "    scale_color_manual(values = cols) +\n",
    "    # facet_wrap(~ID, scales=\"free\", ncol=1) +\n",
    "    # facet_grid(scales=\"free\", rows=vars(ID)) +\n",
    "    # scale_x_log10() +\n",
    "    ggtitle(\"\") \n",
    "  \n",
    "ggsave(\"stat/Fig4.TE_activity_new_transcripts_density.pdf\", width=7, height=2)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### cluster TE in oncogenes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# oncogenes\n",
    "load(\"stat/TE_location.RData\")\n",
    "load(\"stat/hg38.gene.transcript.gtf.RData\")\n",
    "load(\"stat/BC_CRC_TE_count_and_dif.RData\") # TE.norlCount, dif_TE.list\n",
    "tumor_genes <- read.table(\"database/NG_tumor_related_genes.txt\")\n",
    "ID.list <- list(BC=c(\"MDAMB231\", \"MCF10A\"),\n",
    "                CRC=c(\"HCT116\", \"NCM460\"))\n",
    "\n",
    "# background\n",
    "bg_genes <- unique(TE_location[rownames(TE.norlCount), \"gene_name\"])\n",
    "bg_genes <- bg_genes[complete.cases(bg_genes)]\n",
    "bg_tt <- table(bg_genes %in% tumor_genes$V1)\n",
    "# dif TE anno\n",
    "dif_TE <- reshape2::melt(dif_TE.list)\n",
    "colnames(dif_TE) <- c(\"TE_name\", \"dif\", \"Compare\")\n",
    "dif_TE <- merge(dif_TE, TE_location, by=\"TE_name\")\n",
    "\n",
    "gene_names <- lapply(dif_TE$gene_name, function(x){strsplit(x, \",\")[[1]]})\n",
    "names(gene_names) <- dif_TE$TE_name\n",
    "gene_names <- reshape2::melt(gene_names)\n",
    "colnames(gene_names) <- c(\"gene_name\", \"TE_name\")\n",
    "dif_TE <- merge(dif_TE, gene_names, by=\"TE_name\")\n",
    "dif_TE$Tumor_genes <- 0\n",
    "dif_TE$Tumor_genes[dif_TE$gene_name.y %in% tumor_genes$V1] <- 1\n",
    "    \n",
    "\n",
    "# cluster\n",
    "k<- 10\n",
    "clusters <- cutree(pheatmap.p$tree_row, k=k)\n",
    "test_res <- lapply(1:k, function(cluster){\n",
    "    cluster_TE <- names(clusters)[clusters %in% cluster]\n",
    "    a1 <- unique(dif_TE$TE_name[dif_TE$TE_name %in% cluster_TE & dif_TE$Tumor_genes %in% 1])\n",
    "    a2 <- unique(dif_TE$TE_name[dif_TE$TE_name %in% cluster_TE & dif_TE$Tumor_genes %in% 0])\n",
    "    # if(a1>0)\n",
    "    x <- c(length(a1), length(a2), bg_tt[\"TRUE\"], bg_tt[\"FALSE\"])\n",
    "    x <- matrix(x, ncol = 2)\n",
    "    pvalue <- chisq.test(x)$p.value # 0.01910379\n",
    "    ratio= (length(a1)/sum(c(length(a1), length(a2)))) / (bg_tt[\"TRUE\"]/sum(bg_tt))\n",
    "    log2FC <-  log2(ratio) # 0.822162\n",
    "\n",
    "    res <- data.frame(cluster=cluster,\n",
    "                     pvalue=pvalue,\n",
    "                     log2FC=log2FC)  \n",
    "\n",
    "})\n",
    "\n",
    "\n",
    "\n",
    "pp <- rbindlist(test_res)\n",
    "pp$padj <- p.adjust(pp$pvalue, method = \"fdr\", n = length(pp$pvalue))\n",
    "# pp <- pp[order(pp$pvalue, decreasing=TRUE),]\n",
    "pp <- pp[order(pp$log2FC),]\n",
    "pp$cluster <- paste0(\"Cluster\", pp$cluster)\n",
    "pp$cluster <- factor(pp$cluster, levels=pp$cluster)\n",
    "\n",
    "ggplot(pp, aes(x=cluster, y=log2FC, fill=padj)) +\n",
    "      geom_bar(stat=\"identity\", width = 0.8) +\n",
    "      theme_bw() +\n",
    "      theme(panel.grid = element_blank(),\n",
    "            plot.title = element_text(hjust = 0, size = 12),\n",
    "            axis.text.x  = element_text(angle = 0, hjust=0.5, vjust=0.5, \n",
    "                                        colour = \"black\", size = 11),\n",
    "            axis.text.y  = element_text(colour = \"black\"),\n",
    "            # legend.position = c(0.8,0.8),\n",
    "            strip.background = element_rect(fill=\"white\", colour = \"black\", size=1) ,\n",
    "            legend.title = element_text(face=\"bold\")) +\n",
    "      # geom_text(aes(label=symbol), vjust=0.5, hjust=0,size=4, colour=\"black\") +\n",
    "      # annotate(\"text\", x=pp$NAME[2], y=-1.5,size=3,\n",
    "      #          label=\"nominal P values: \\n*P < 0.05 \\n**P < 0.005 \\n***P < 0.0005 \\nNS: not significant\") +\n",
    "      # scale_x_discrete(limits=c(pp$term.name))+\n",
    "      # scale_fill_manual(values=c(\"GO:BP\"=\"#4678BA\", \"KEGG\"=\"#EF8354\")) +\n",
    "      coord_flip() +\n",
    "      geom_hline(yintercept = 0.3, color=\"red\", linetype=\"dashed\") +\n",
    "      xlab(\"\") +\n",
    "      ggtitle(\"TE related oncogenes\")\n",
    "    # facet_grid(.~module, scales = \"free_x\", space = \"free_x\") \n",
    "\n",
    "\n",
    "ggsave(paste0(\"stat/BC_CRC_cluster_TE_oncogenes_chisq_test_bar.pdf\"),\n",
    "       height = 3, width = 6)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "##\n",
    "k <- 10\n",
    "clusters <- cutree(pheatmap.p$tree_row, k=k)\n",
    "clusters.tb <- data.frame(TE_name=names(clusters), cluster=clusters)\n",
    "out_TE <- merge(dif_TE, clusters.tb, by=\"TE_name\")\n",
    "write.table(out_TE, file=\"stat/BC_CRC_dif_TE_oncogenes.txt\", sep=\"\\t\", quote=FALSE, row.names=FALSE, col.names=TRUE)\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### upTE in oncogenes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# oncogenes\n",
    "load(\"stat/TE_location.RData\")\n",
    "load(\"stat/hg38.gene.transcript.gtf.RData\")\n",
    "tumor_genes <- read.table(\"database/NG_tumor_related_genes.txt\")\n",
    "ID.list <- list(BC=c(\"MDAMB231\", \"MCF10A\"),\n",
    "                CRC=c(\"HCT116\", \"NCM460\"))\n",
    "gene.Files <- list(BC=\"stat/MDAMB231_vs_MCF10A_all_genes.txt\",\n",
    "                   CRC=\"stat/HCT116_vs_NCM460_all_genes.txt\")\n",
    "\n",
    "\n",
    "test_res <- lapply(names(ID.list), function(T){\n",
    "    file <- gene.Files[[T]]\n",
    "    gene.tb <- read.table(file, header = TRUE)\n",
    "    bg <- unique(gene.tb$gene_name)\n",
    "    bg_tt <- table(bg %in% tumor_genes$V1)\n",
    "    \n",
    "    IDs <- ID.list[[T]]\n",
    "    vs <- paste(IDs[1], IDs[2], sep=\"_vs_\")\n",
    "    dif_TE <- reshape2::melt(dif_TE.list[[vs]])\n",
    "    colnames(dif_TE) <- c(\"TE_name\", \"dif\")\n",
    "    dif_TE <- merge(dif_TE, TE_location, by=\"TE_name\")\n",
    "\n",
    "    gene_names <- lapply(dif_TE$gene_name, function(x){strsplit(x, \",\")[[1]]})\n",
    "    names(gene_names) <- dif_TE$TE_name\n",
    "    gene_names <- reshape2::melt(gene_names)\n",
    "    colnames(gene_names) <- c(\"gene_name\", \"TE_name\")\n",
    "    dif_TE <- merge(dif_TE, gene_names, by=\"TE_name\")\n",
    "    dif_TE$Tumor_genes <- 0\n",
    "    dif_TE$Tumor_genes[dif_TE$gene_name.y %in% tumor_genes$V1] <- 1\n",
    "    \n",
    "    \n",
    "    a <- lapply(c(\"Up\", \"Down\"), function(stat){   \n",
    "        a1 <- unique(dif_TE$TE_name[dif_TE$dif %in% stat & dif_TE$Tumor_genes %in% 1])\n",
    "        a2 <- unique(dif_TE$TE_name[dif_TE$dif %in% stat & dif_TE$Tumor_genes %in% 0])\n",
    "        x <- c(length(a1), length(a2), bg_tt[\"TRUE\"], bg_tt[\"FALSE\"])\n",
    "        x <- matrix(x, ncol = 2)\n",
    "        pvalue <- chisq.test(x)$p.value # 0.01910379\n",
    "        ratio= (length(a1)/sum(c(length(a1), length(a2)))) / (bg_tt[\"TRUE\"]/sum(bg_tt))\n",
    "        log2FC <-  log2(ratio) # 0.822162\n",
    "\n",
    "        res <- data.frame(vs=vs,\n",
    "                          stat=stat,\n",
    "                         pvalue=pvalue,\n",
    "                         log2FC=log2FC)\n",
    "    \n",
    "    })\n",
    "\n",
    "    \n",
    "})\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "pp <- rbindlist(lapply(test_res, rbindlist))\n",
    "pp$padj <- p.adjust(pp$pvalue, method = \"fdr\", n = length(pp$pvalue))\n",
    "pp <- pp[order(pp$pvalue, decreasing=TRUE),]\n",
    "pp$ID <- paste(pp$vs, pp$stat, sep=\":\")\n",
    "pp$ID <- factor(pp$ID, levels=pp$ID)\n",
    "\n",
    "ggplot(pp, aes(x=ID, y=log2FC, fill=pvalue)) +\n",
    "      geom_bar(stat=\"identity\", width = 0.8) +\n",
    "      theme_bw() +\n",
    "      theme(panel.grid = element_blank(),\n",
    "            plot.title = element_text(hjust = 0, size = 12),\n",
    "            axis.text.x  = element_text(angle = 0, hjust=0.5, vjust=0.5, \n",
    "                                        colour = \"black\", size = 11),\n",
    "            axis.text.y  = element_text(colour = \"black\"),\n",
    "            # legend.position = c(0.8,0.8),\n",
    "            strip.background = element_rect(fill=\"white\", colour = \"black\", size=1) ,\n",
    "            legend.title = element_text(face=\"bold\")) +\n",
    "      # geom_text(aes(label=symbol), vjust=0.5, hjust=0,size=4, colour=\"black\") +\n",
    "      # annotate(\"text\", x=pp$NAME[2], y=-1.5,size=3,\n",
    "      #          label=\"nominal P values: \\n*P < 0.05 \\n**P < 0.005 \\n***P < 0.0005 \\nNS: not significant\") +\n",
    "      # scale_x_discrete(limits=c(pp$term.name))+\n",
    "      # scale_fill_manual(values=c(\"GO:BP\"=\"#4678BA\", \"KEGG\"=\"#EF8354\")) +\n",
    "      coord_flip() +\n",
    "      geom_hline(yintercept = 0.3, color=\"red\", linetype=\"dashed\") +\n",
    "      xlab(\"\") +\n",
    "      ggtitle(\"TE related oncogenes\")\n",
    "    # facet_grid(.~module, scales = \"free_x\", space = \"free_x\") \n",
    "\n",
    "\n",
    "ggsave(paste0(\"stat/BC_CRC_TE_oncogenes_chisq_test_bar.pdf\"),\n",
    "       height = 3, width = 6)\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### (gene log2FC) vs (gene newtrans count)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error in parse(text = x, srcfile = src): <text>:17:26: unexpected symbol\n16: TE.genes <- reshape2::melt(relate_genes)\n17: colnames(TE.genes) <- c(\"gene_id\n                             ^\n",
     "output_type": "error",
     "traceback": [
      "Error in parse(text = x, srcfile = src): <text>:17:26: unexpected symbol\n16: TE.genes <- reshape2::melt(relate_genes)\n17: colnames(TE.genes) <- c(\"gene_id\n                             ^\nTraceback:\n"
     ]
    }
   ],
   "source": [
    "library(ggpubr)\n",
    "\n",
    "## activity TE\n",
    "load(\"stat/TE_location.RData\")\n",
    "BC_genes <- read.table(\"stat/MDAMB231_vs_MCF10A_all_genes.txt\", header=T)\n",
    "CRC_genes <- read.table(\"stat/HCT116_vs_NCM460_all_genes.txt\", header=T)\n",
    "Diff.list <- list(MDAMB231=MDAMB231_Diff, MCF10A=MCF10A_Diff,\n",
    "                  HCT116=HCT116_Diff, NCM460=NCM460_Diff)\n",
    "\n",
    "\n",
    "## for BC\n",
    "dif_TE <- reshape2::melt(dif_TE.list[[\"MDAMB231_vs_MCF10A\"]])\n",
    "colnames(dif_TE) <- c(\"TE_name\", \"dif\")\n",
    "sub.TE_location <- merge(dif_TE, TE_location[, c(\"TE_name\", \"gene_id\")], by=\"TE_name\")\n",
    "relate_genes <- sub.TE_location$gene_id\n",
    "relate_genes <- lapply(relate_genes, function(x){strsplit(x, \",\")[[1]]})\n",
    "names(relate_genes) <- sub.TE_location$TE_name\n",
    "TE.genes <- reshape2::melt(relate_genes)\n",
    "colnames(TE.genes) <- c(\"gene_id\", \"TE_name\")\n",
    "TE.genes <- TE.genes[complete.cases(TE.genes), ]\n",
    "df <- merge(TE.genes, BC_genes[, c(\"gene_id\", \"log2FoldChange\", \"dif\")], by=\"gene_id\")\n",
    "colnames(df)[4] <- \"gene_dif\"\n",
    "df <- merge(df, dif_TE, by=\"TE_name\")\n",
    "\n",
    "nc <- data.frame(table(MDAMB231_Diff[[1]]$ref_gene_id))\n",
    "rownames(nc) <- nc$Var1\n",
    "df$MDAMB231_gene_include_newTrans <- nc[as.character(df$gene_id), \"Freq\"]\n",
    "df$MDAMB231_gene_include_newTrans[is.na(df$MDAMB231_gene_include_newTrans)] <- 0\n",
    "\n",
    "nc <- data.frame(table(MCF10A_Diff[[1]]$ref_gene_id))\n",
    "rownames(nc) <- nc$Var1\n",
    "df$MCF10A_gene_include_newTrans <- nc[as.character(df$gene_id), \"Freq\"]\n",
    "df$MCF10A_gene_include_newTrans[is.na(df$MCF10A_gene_include_newTrans)] <- 0\n",
    "df$newTrans_count_diff <- df$MDAMB231_gene_include_newTrans-df$MCF10A_gene_include_newTrans\n",
    "BC_df <- df\n",
    "\n",
    "\n",
    "## for CRC\n",
    "dif_TE <- reshape2::melt(dif_TE.list[[\"HCT116_vs_NCM460\"]])\n",
    "colnames(dif_TE) <- c(\"TE_name\", \"dif\")\n",
    "sub.TE_location <- merge(dif_TE, TE_location[, c(\"TE_name\", \"gene_id\")], by=\"TE_name\")\n",
    "relate_genes <- sub.TE_location$gene_id\n",
    "relate_genes <- lapply(relate_genes, function(x){strsplit(x, \",\")[[1]]})\n",
    "names(relate_genes) <- sub.TE_location$TE_name\n",
    "TE.genes <- reshape2::melt(relate_genes)\n",
    "colnames(TE.genes) <- c(\"gene_id\", \"TE_name\")\n",
    "TE.genes <- TE.genes[complete.cases(TE.genes), ]\n",
    "df <- merge(TE.genes, CRC_genes[, c(\"gene_id\", \"log2FoldChange\", \"dif\")], by=\"gene_id\")\n",
    "colnames(df)[4] <- \"gene_dif\"\n",
    "df <- merge(df, dif_TE, by=\"TE_name\")\n",
    "\n",
    "nc <- data.frame(table(HCT116_Diff[[1]]$ref_gene_id))\n",
    "rownames(nc) <- nc$Var1\n",
    "df$HCT116_gene_include_newTrans <- nc[as.character(df$gene_id), \"Freq\"]\n",
    "df$HCT116_gene_include_newTrans[is.na(df$HCT116_gene_include_newTrans)] <- 0\n",
    "\n",
    "nc <- data.frame(table(NCM460_Diff[[1]]$ref_gene_id))\n",
    "rownames(nc) <- nc$Var1\n",
    "df$NCM460_gene_include_newTrans <- nc[as.character(df$gene_id), \"Freq\"]\n",
    "df$NCM460_gene_include_newTrans[is.na(df$NCM460_gene_include_newTrans)] <- 0\n",
    "df$newTrans_count_diff <- df$HCT116_gene_include_newTrans-df$NCM460_gene_include_newTrans\n",
    "CRC_df <- df\n",
    "\n",
    "extract_name <- intersect(colnames(BC_df), colnames(CRC_df))\n",
    "pp <- rbind(BC_df[,extract_name], CRC_df[,extract_name])\n",
    "\n",
    "cor <- cor(pp$log2FoldChange, pp$newTrans_count_diff, method=\"spearman\") # 0.5686579\n",
    "# cor <- cor(BC_df$log2FoldChange, BC_df$newTrans_count_diff, method=\"spearman\") # 0.5716111\n",
    "# cor <- cor(CRC_df$log2FoldChange, CRC_df$newTrans_count_diff, method=\"spearman\") # 0.5584396\n",
    "\n",
    "ggplot(pp, aes(x=log2FoldChange, y=newTrans_count_diff, shape=dif, color=gene_dif)) +\n",
    "    geom_point() + \n",
    "    theme_bw() +\n",
    "    theme(\n",
    "      panel.grid=element_blank(),\n",
    "      axis.text.x = element_text(angle = 0, color = \"black\", \n",
    "                                 size = 8, hjust = 0.5, vjust=0.5),\n",
    "      plot.title=element_text(size=14, face=\"bold\", hjust=0.5)\n",
    "    ) +\n",
    "    scale_color_manual(values = c(Up=\"#FC8D62\", ns=\"#D3D3D3\", Down=\"#66C2A5\")) +\n",
    "    # facet_wrap(~ID, scales=\"free\", ncol=1) +\n",
    "    # facet_grid(scales=\"free\", rows=vars(ID)) +\n",
    "    # scale_x_log10() +\n",
    "    # scale_x_continuous(expand = c(0,0), limits=c(-10,10)) +\n",
    "    # scale_y_continuous(expand = c(0,0)) +\n",
    "    geom_text(aes(x = -8, y=10, label = paste0(\"spearman cor: \", round(cor, 3))), color=\"black\") \n",
    "    ggtitle(\"\") \n",
    "  \n",
    "ggsave(\"stat/CRC_TE_genes_logFC_newTrans.pdf\", width=6, height=5)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 影响基因的TE所位于新转录本上的位置"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sub.BC_df <- BC_df[BC_df$gene_dif %in% c(\"Up\") & BC_df$dif %in% c(\"Up\"), ]\n",
    "BC_upTE_in_newTrans <- sum(unique(sub.BC_df$TE_name) %in% MDAMB231_Diff[[2]]$TE_name)\n",
    "BC_upTE_in_TSS_TES <- sum(unique(sub.BC_df$TE_name) %in% MDAMB231_Diff[[2]][MDAMB231_Diff[[2]]$type %in% c(\"TSS\", \"TES\"), ]$TE_name)\n",
    "BC_TE_in_TSS_TES <- length(unique(MDAMB231_Diff[[2]][MDAMB231_Diff[[2]]$type %in% c(\"TSS\", \"TES\"), ]$TE_name))\n",
    "BC_TE_in_newTrans <- length(unique(MDAMB231_Diff[[2]]$TE_name))\n",
    "\n",
    "\n",
    "sub.CRC_df <- CRC_df[CRC_df$gene_dif %in% c(\"Up\") & CRC_df$dif %in% c(\"Up\"), ]\n",
    "CRC_upTE_in_newTrans <- sum(unique(sub.CRC_df$TE_name) %in% HCT116_Diff[[2]]$TE_name)\n",
    "CRC_upTE_in_TSS_TES <- sum(unique(sub.CRC_df$TE_name) %in% HCT116_Diff[[2]][HCT116_Diff[[2]]$type %in% c(\"TSS\", \"TES\"), ]$TE_name)\n",
    "CRC_TE_in_TSS_TES <- length(unique(HCT116_Diff[[2]][HCT116_Diff[[2]]$type %in% c(\"TSS\", \"TES\"), ]$TE_name))\n",
    "CRC_TE_in_newTrans <- length(unique(HCT116_Diff[[2]]$TE_name))\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# pie\n",
    "# here, upTE represent : MDAMB231/HCT116 up in genes and TE, simultaneously\n",
    "BC <- data.frame(upTE=c(TE_in_newTrans=BC_upTE_in_newTrans-BC_upTE_in_TSS_TES, \n",
    "                        TE_in_TSS_TES=BC_upTE_in_TSS_TES),\n",
    "                 bg=c(TE_in_newTrans=BC_TE_in_newTrans-BC_TE_in_TSS_TES, \n",
    "                      TE_in_TSS_TES=BC_TE_in_TSS_TES))\n",
    "CRC <- data.frame(upTE=c(TE_in_newTrans=CRC_upTE_in_newTrans-CRC_upTE_in_TSS_TES, \n",
    "                         TE_in_TSS_TES=CRC_upTE_in_TSS_TES),\n",
    "                 bg=c(TE_in_newTrans=CRC_TE_in_newTrans-CRC_TE_in_TSS_TES,\n",
    "                      TE_in_TSS_TES=CRC_TE_in_TSS_TES))\n",
    "chisq.test(BC)$p.value # 0.00463643\n",
    "chisq.test(CRC)$p.value # 1.24659e-05\n",
    "\n",
    "\n",
    "pp <- rbind(data.frame(reshape2::melt(t(BC)), Cancer=\"BC\"),\n",
    "           data.frame(reshape2::melt(t(CRC)), Cancer=\"CRC\"))\n",
    "colnames(pp) <- c(\"stat\", \"TE_condition\", \"Count\", \"Cancer\")\n",
    "p <- ddply(pp, .(Cancer, stat), mutate, Percentage = Count / sum(Count)*100)\n",
    "\n",
    "\n",
    "plot <- ggplot(p, aes(x=\"\", y=Percentage, fill=TE_condition))+\n",
    "    geom_bar(width = 1, stat = \"identity\") +\n",
    "    coord_polar(\"y\", start=0) +\n",
    "    scale_fill_manual(values = c(\"#F4B540\", \"#BA748A\", \"#87B07A\", \"#4678BA\", \"#FDA56C\"))+\n",
    "    theme_minimal() +\n",
    "    theme(\n",
    "      axis.title.x = element_blank(),\n",
    "      axis.title.y = element_blank(),\n",
    "      panel.border = element_blank(),\n",
    "      panel.grid=element_blank(),\n",
    "      # axis.ticks = element_blank(),\n",
    "      plot.title=element_text(size=14, face=\"bold\", hjust=0.5)\n",
    "    ) +\n",
    "    theme(axis.text.x=element_blank())+\n",
    "    # geom_text(aes(x = 1.6, label = scales::percent(Percentage/100, accuracy = .1)), position = position_stack(vjust = .5)) +\n",
    "    geom_text(aes(x = 1.6, label = paste0(Count, \"(\",scales::percent(Percentage/100, accuracy = .1), \")\")), \n",
    "              position = position_stack(vjust = .5)) +\n",
    "    facet_grid(stat~Cancer) +\n",
    "    ggtitle(\"\")\n",
    "\n",
    "\n",
    "ggsave(paste0(\"stat/BC_CRC_upTE_in_newTrans_pie.pdf\"), width = 5, height = 5)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### oncogenes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "load(\"stat/hg38.gene.transcript.gtf.RData\")\n",
    "tumor_genes <- read.table(\"database/NG_tumor_related_genes.txt\")\n",
    "\n",
    "### BC\n",
    "sub.BC_df$TE_in_newTrans <- 0\n",
    "sub.BC_df$TE_in_TSS_TES <- 0\n",
    "sub.BC_df$TE_in_newTrans[sub.BC_df$TE_name %in% MDAMB231_Diff[[2]]$TE_name] <- 1\n",
    "sub.BC_df$TE_in_TSS_TES[sub.BC_df$TE_name %in% MDAMB231_Diff[[2]][MDAMB231_Diff[[2]]$type %in% c(\"TSS\", \"TES\"), ]$TE_name] <- 1\n",
    "sub.BC_df$gene_name <- gene_gtf[as.character(sub.BC_df$gene_id), \"gene_name\"]\n",
    "sub.BC_df$tumor_genes <- 0\n",
    "sub.BC_df$tumor_genes[sub.BC_df$gene_name %in% tumor_genes$V1] <- 1\n",
    "\n",
    "\n",
    "\n",
    "a1 <- unique(sub.BC_df$TE_name[sub.BC_df$TE_in_newTrans==1 & !(sub.BC_df$TE_in_TSS_TES==1) & sub.BC_df$tumor_genes==1])\n",
    "a2 <- unique(sub.BC_df$TE_name[sub.BC_df$TE_in_newTrans==1 & !(sub.BC_df$TE_in_TSS_TES==1) & sub.BC_df$tumor_genes==0])\n",
    "b1 <- unique(sub.BC_df$TE_name[(sub.BC_df$TE_in_TSS_TES==1) & sub.BC_df$tumor_genes==1])\n",
    "b2 <- unique(sub.BC_df$TE_name[(sub.BC_df$TE_in_TSS_TES==1) & sub.BC_df$tumor_genes==0])\n",
    "\n",
    "\n",
    "x <- matrix(c(length(a1), length(a2), length(b1), length(b2)), ncol = 2)\n",
    "pvalue <- chisq.test(x)$p.value\n",
    "\n",
    "### CRC\n",
    "sub.CRC_df$TE_in_newTrans <- 0\n",
    "sub.CRC_df$TE_in_TSS_TES <- 0\n",
    "sub.CRC_df$TE_in_newTrans[sub.CRC_df$TE_name %in% HCT116_Diff[[2]]$TE_name] <- 1\n",
    "sub.CRC_df$TE_in_TSS_TES[sub.CRC_df$TE_name %in% HCT116_Diff[[2]][HCT116_Diff[[2]]$type %in% c(\"TSS\", \"TES\"), ]$TE_name] <- 1\n",
    "sub.CRC_df$gene_name <- gene_gtf[as.character(sub.CRC_df$gene_id), \"gene_name\"]\n",
    "sub.CRC_df$tumor_genes <- 0\n",
    "sub.CRC_df$tumor_genes[sub.CRC_df$gene_name %in% tumor_genes$V1] <- 1\n",
    "\n",
    "\n",
    "\n",
    "a1 <- unique(sub.CRC_df$TE_name[sub.CRC_df$TE_in_newTrans==1 & !(sub.CRC_df$TE_in_TSS_TES==1) & sub.CRC_df$tumor_genes==1])\n",
    "a2 <- unique(sub.CRC_df$TE_name[sub.CRC_df$TE_in_newTrans==1 & !(sub.CRC_df$TE_in_TSS_TES==1) & sub.CRC_df$tumor_genes==0])\n",
    "b1 <- unique(sub.CRC_df$TE_name[(sub.CRC_df$TE_in_TSS_TES==1) & sub.CRC_df$tumor_genes==1])\n",
    "b2 <- unique(sub.CRC_df$TE_name[(sub.CRC_df$TE_in_TSS_TES==1) & sub.CRC_df$tumor_genes==0])\n",
    "\n",
    "\n",
    "x <- matrix(c(length(a1), length(a2), length(b1), length(b2)), ncol = 2)\n",
    "pvalue <- chisq.test(x)$p.value\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "# new transcripts density"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "load(\"stat/assembly_preData.RData\") # \"gtf.tb.list\"  \"RM.tb.list\"   \"RM.tb.list2\"  \"tmap.tb.list\"\n",
    "load(\"stat/HCT116_NCM460_diff.Data\")\n",
    "load(\"stat/MDAMB231_MCF10A_diff.Data\")\n",
    "load(\"stat/hg38.gtf.RData\")\n",
    "load(\"stat/hg38.gene.transcript.gtf.RData\")\n",
    "\n",
    "\n",
    "\n",
    "MDAMB231_Diff[[1]][1,]\n",
    "\n",
    "new_count.list <- list(MDAMB231=table(MDAMB231_Diff[[1]]$ref_gene_id),\n",
    "    MCF10A=table(MCF10A_Diff[[1]]$ref_gene_id),\n",
    "    HCT116=table(HCT116_Diff[[1]]$ref_gene_id.ref),\n",
    "    NCM460=table(NCM460_Diff[[1]]$ref_gene_id.ref))\n",
    "\n",
    "\n",
    "new_count.list <- lapply(names(new_count.list), function(id){\n",
    "    x <- new_count.list[[id]]\n",
    "    x <- data.frame(x)\n",
    "    colnames(x) <- c(\"gene_id\", id)\n",
    "    x <- x[x$gene_id != \"-\", ]\n",
    "    \n",
    "    return(x)\n",
    "\n",
    "})\n",
    "\n",
    "pp <- join_all(new_count.list, type=\"full\")\n",
    "rownames(pp) <- pp$gene_id\n",
    "pp <- pp[,-1]\n",
    "pp[is.na(pp)] <- 0\n",
    "sub.pp <- pp[apply(pp, 1, function(x){any(x>=10)}), ]\n",
    "rownames(sub.pp) <- gene_gtf[rownames(sub.pp), 'gene_name']\n",
    "\n",
    "\n",
    "\n",
    "pheatmap(sub.pp,\n",
    "         # display_numbers=T,\n",
    "         # scale=\"row\",\n",
    "         # annotation_col = colInfo,\n",
    "         # annotation_row = rowInfo,\n",
    "         # annotation_colors = anno_colors,\n",
    "         main=\"new trans count\",\n",
    "         col=colorRampPalette(rev(brewer.pal(11,\"RdBu\")))(21),\n",
    "         # col=colorRampPalette(c(\"#4678BA\", \"#F7F7F7\", \"#EF8354\"))(11),\n",
    "         border=NA,\n",
    "         fontsize_row=1.5,\n",
    "         # show_rownames=F,\n",
    "         # show_colnames=F,\n",
    "         # cluster_row=F,\n",
    "         width=5,\n",
    "         height=6,\n",
    "         filename=\"stat/new_trans_density_in_known_genes_heatmap.pdf\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# disable"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# sub.count.tb <- count.tb[count.tb$Geneid %in% trans_TE$TE_name | \n",
    "#                          count.tb$Geneid %in% trans_TE2$TE_name, ]\n",
    "# sub.count.tb <- count.tb[, SampleIDs]\n",
    "# sub.count.tb <- sub.count.tb[apply(sub.count.tb, 1, function(x){ (sum(x>=3))>=3 }),]\n",
    "\n",
    "# CellLine <- factor(Sample_info[SampleIDs, ]$CellLine)\n",
    "# design <- model.matrix(~0+CellLine)\n",
    "# colnames(design) <- levels(CellLine)\n",
    "\n",
    "# y <- voom(sub.count.tb, design, plot=F)\n",
    "# fit <- lmFit(y, design)\n",
    "# contr <- makeContrasts(MDAMB231vsMCF10A=MDAMB231-MCF10A,levels=design)\n",
    "# fit.contr <- eBayes(contrasts.fit(fit, contr))\n",
    "# dt <- decideTests(fit.contr)\n",
    "# summary(dt)\n",
    "# top <- topTable(fit.contr, number=Inf, sort.by=\"P\")\n",
    "# top <- top[rownames(top) %in% trans_TE$TE_name | \n",
    "#           rownames(top) %in% trans_TE2$TE_name, ]\n",
    "# Down <- top[top$logFC< -1 & top$adj.P.Val<0.05, ]\n",
    "# Up <- top[top$logFC> 1 & top$adj.P.Val<0.05, ]\n",
    "# dim(Down)\n",
    "# dim(Up)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# ###\n",
    "# Down_TE <- tstrsplit(rownames(Down), \":\")[[2]]\n",
    "# Down_TE <- substring(Down_TE, 1, 4)\n",
    "# Down_TE[grepl(\"L1\", Down_TE)] <- substring(Down_TE[grepl(\"L1\", Down_TE)], 1, 3)\n",
    "# Down_TE[Down_TE %in% \"HAL\"] <- \"HAL1\"\n",
    "# Down_TE[Down_TE %in% \"L1H\"] <- \"L1HS\"\n",
    "# table(Down_TE)\n",
    "\n",
    "\n",
    "# ###\n",
    "# Up_TE <- tstrsplit(rownames(Up), \":\")[[2]]\n",
    "# Up_TE <- substring(Up_TE, 1, 4)\n",
    "# Up_TE[grepl(\"L1\", Up_TE)] <- substring(Up_TE[grepl(\"L1\", Up_TE)], 1, 3)\n",
    "# Up_TE[Up_TE %in% \"HAL\"] <- \"HAL1\"\n",
    "# Up_TE[Up_TE %in% \"L1H\"] <- \"L1HS\"\n",
    "# table(Up_TE)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "anno.tb <- read.table(\"/NAS/wg_looking/human_ref/GRCh38.p13.genome.fa.TE.saf\")\n",
    "colnames(anno.tb) <- c(\"GeneID\", \"Chr\", \"Start\", \"End\", \"Strand\")\n",
    "\n",
    "fcounts <- lapply(IDs, function(ID){\n",
    "    bam = file.path(ID, paste0(ID, \".Q7.L300.sort.bam\"))\n",
    "    fc <- featureCounts(bam,\n",
    "\n",
    "\t# annotation\n",
    "\t# annot.inbuilt=\"hg38\",\n",
    "\tannot.ext=anno.tb,\n",
    "\tisGTFAnnotationFile=FALSE,\n",
    "\tGTF.featureType=\"transcript\",\n",
    "\tGTF.attrType=\"transcript_id\",\n",
    "\tchrAliases=NULL,\n",
    "\t\n",
    "\t# level of summarization\n",
    "\tuseMetaFeatures=TRUE,\n",
    "\t\n",
    "\t# overlap between reads and features\n",
    "\tallowMultiOverlap=TRUE,\n",
    "\tminOverlap=20,\n",
    "\tlargestOverlap=FALSE,\n",
    "\treadExtension5=0,\n",
    "\treadExtension3=0,\n",
    "\tread2pos=NULL,\n",
    "\t\n",
    "\t# multi-mapping reads\n",
    "\tcountMultiMappingReads=TRUE,\n",
    "\tfraction=TRUE,\n",
    "    isLongRead=TRUE,\n",
    "                    \n",
    "\t# read filtering\n",
    "\tminMQS=0,\n",
    "\tsplitOnly=FALSE,\n",
    "\tnonSplitOnly=FALSE,\n",
    "\tprimaryOnly=FALSE,\n",
    "\tignoreDup=FALSE,\n",
    "\t\n",
    "\t# strandness\n",
    "\tstrandSpecific=0,\n",
    "\t\n",
    "\t# exon-exon junctions\n",
    "\tjuncCounts=FALSE,\n",
    "\tgenome=NULL,\n",
    "\t\n",
    "\t# miscellaneous\n",
    "\tnthreads=10,\n",
    "\tmaxMOp=10)\n",
    "\n",
    "    return(fc)\n",
    "\n",
    "})\n"
   ]
  }
 ],
 "metadata": {
  "julynter-results": {
   "filteredId": [],
   "filteredIndividual": [],
   "filteredRestart": [],
   "filteredType": [],
   "hash": "f2519e84065770948b91926e5274aa6a017b25ac",
   "visible": [
    {
     "cellId": "group",
     "hash": "c0de46ba034d710c2b2406a37ee6816be9634aa0",
     "reason": "This groups other lint messages",
     "reportId": "group",
     "reportType": "hiddenstate",
     "suggestion": null,
     "text": "Hidden State"
    },
    {
     "cellId": 0,
     "hash": "ab635f27d8307c7ec20c7344dcd83860849b066f",
     "reason": "A skip in the execution count might indicate the presence of a hidden state caused by a cell that does not exist anymore. Hidden states might prevent cells from executing or producing the same results, hampering the reproducibility.",
     "reportId": "h4",
     "reportType": "hiddenstate",
     "suggestion": "Please consider re-running the notebook to guarantee the reproducibility.",
     "text": "Cell 0 skips the execution count"
    },
    {
     "cellId": 1,
     "hash": "610f85a5c06f0a661da15dfd426b9b45f6e5aa14",
     "reason": "A skip in the execution count might indicate the presence of a hidden state caused by a cell that does not exist anymore. Hidden states might prevent cells from executing or producing the same results, hampering the reproducibility.",
     "reportId": "h4",
     "reportType": "hiddenstate",
     "suggestion": "Please consider re-running the notebook to guarantee the reproducibility.",
     "text": "Cell 1 skips the execution count"
    },
    {
     "cellId": "group",
     "hash": "19d3d3d153f22e318cbe061df9e685ba5165b0aa",
     "reason": "This groups other lint messages",
     "reportId": "group",
     "reportType": "confusenotebook",
     "suggestion": null,
     "text": "Confuse Notebook"
    },
    {
     "cellId": 0,
     "hash": "558be4586258f1aebfed8fbd0a40471ebb6669a5",
     "reason": "A markdown cell at the beginning of the notebook can provide a human-friendly title with no constraints and introduce the notebook, indicating its purpose and external requirements.",
     "reportId": "c4",
     "reportType": "confusenotebook",
     "suggestion": "Please consider adding a markdown cell to describe the notebook.",
     "text": "The first cell of the notebook is not a markdown cell"
    },
    {
     "cellId": 1,
     "hash": "0aeb1ac1775b418b738b6d5daef9d37a59fb76a3",
     "reason": "A markdown cell at the end of the notebook can conclude it, presenting a summary of the obtained results.",
     "reportId": "c5",
     "reportType": "confusenotebook",
     "suggestion": "Please consider adding a markdown cell to conclude the notebook.",
     "text": "The last cell of the notebook is not a markdown cell"
    }
   ]
  },
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.1"
  },
  "toc-autonumbering": true,
  "toc-showmarkdowntxt": true,
  "toc-showtags": false
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
